<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Darren Irwin">
<meta name="dcterms.date" content="2023-09-10">

<title>Greenish Warbler Genomic Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="GreenishWarblerGenomics2023_files/libs/clipboard/clipboard.min.js"></script>
<script src="GreenishWarblerGenomics2023_files/libs/quarto-html/quarto.js"></script>
<script src="GreenishWarblerGenomics2023_files/libs/quarto-html/popper.min.js"></script>
<script src="GreenishWarblerGenomics2023_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="GreenishWarblerGenomics2023_files/libs/quarto-html/anchor.min.js"></script>
<link href="GreenishWarblerGenomics2023_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="GreenishWarblerGenomics2023_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="GreenishWarblerGenomics2023_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="GreenishWarblerGenomics2023_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="GreenishWarblerGenomics2023_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="GreenishWarblerGenomics2023.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Greenish Warbler Genomic Analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Darren Irwin </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 10, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>This page contains notes and code describing the data analysis for a manuscript on Greenish Warbler genomics. I’ve been working with the data for several years, and the R and then Julia code has been in development for a while. This is a Quarto notebook, which can run and display the results of Julia (or other) code blocks, along with text narration, and output in html, pdf, Word, etc.</p>
<p>The Julia code here is loosely based on R code written for Greenish Warbler analysis (Irwin et al.&nbsp;2016), and then the North American warbler analyses (Irwin et al.&nbsp;2019), and then my (unpublished) 2019 Greenish Warbler analysis. Most recently, this was adapted from the scripts called GW2022_R_analysis_script.R and IrwinLabGenomicsAnalysisScript.jl but has had a lot of opimizations since then. The SNP data here are a result of GBS reads mapped to our new 2022 Biozeron genome assembly for a greenish warbler from southern China.</p>
<section id="load-packages" class="level2">
<h2 class="anchored" data-anchor-id="load-packages">Load packages</h2>
<p>If running this for the first time, you will need to load packages used in the script, so run what is in this section below. It will take some time to install and precompile the packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> <span class="bu">Pkg; Pkg</span>.<span class="fu">add</span>(<span class="st">"CSV"</span>) <span class="co"># took less than a minute</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"DataFrames"</span>) <span class="co"># took about a minute</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"Plots"</span>) <span class="co"># seems to install and working more simply than Makie (but less powerful)</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"Haversine"</span>) <span class="co"># for great circle (Haversine) distances</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"Distributions"</span>) <span class="co"># this seemed to fix a problem installing GLMakie</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"MultivariateStats"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"StatsBase"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"Impute"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"JLD2"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"CairoMakie"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="bu">Pkg</span>.<span class="fu">add</span>(<span class="st">"PrettyTables"</span>) <span class="co"># for printing nice tables to REPL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now actually load those packages into the Julia session:</p>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV </span><span class="co"># for reading in delimited files</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DataFrames </span><span class="co"># for storing data as type DataFrame</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Haversine </span><span class="co"># for calculating Great Circle (haversine) distances between sites</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">MultivariateStats </span><span class="co"># for Principal Coordinates Analysis (multidimensional scaling)</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DelimitedFiles </span><span class="co"># for reading delimited files (the genotypic data)</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Impute </span><span class="co"># for imputing missing genotypes</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">JLD2 </span><span class="co"># for saving data</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CairoMakie </span><span class="co"># for plots</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">PrettyTables</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">activate!</span>()  <span class="co"># this makes CairoMakie the main package for figures (in case another already loaded)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load my custom package <code>SNPlots</code>:</p>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include</span>(<span class="st">"SNPlots.jl"</span>) <span class="co"># load file containing custom-built functions</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">.SNPlots </span><span class="co"># actually make SNPlots module available with SNPlots.functionName(),</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or if functions are exported from SNPlots then they are available.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Test Julia:</p>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> <span class="fl">1</span>; y <span class="op">=</span> <span class="fl">2</span>; z <span class="op">=</span> x<span class="op">+</span>y</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"z = "</span>, z)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>z = 3</code></pre>
</div>
</div>
<p>(If Quarto is calling Julia properly, you will see <code>z = 3</code> as the output of the code block above.)</p>
<p>Choose working directory:</p>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>repoDirectory <span class="op">=</span> <span class="fu">pwd</span>() <span class="co"># this gets the starting working directory, for later use</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>dataDirectory <span class="op">=</span> <span class="st">"/Users/darrenirwin/Dropbox/Darren's current work/"</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ok-lets-load-the-genomic-data" class="level2">
<h2 class="anchored" data-anchor-id="ok-lets-load-the-genomic-data"><em>OK, let’s load the genomic data!</em></h2>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose path and filename for the 012NA files</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>baseName <span class="op">=</span> <span class="st">"GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>filenameTextMiddle <span class="op">=</span> <span class="st">".max2allele_noindel.vcf.maxmiss"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># indicate percent threshold for missing genotypes for each SNP--</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># this was set by earlier filtering, and is just a record-keeper for the filenames:</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>missingGenotypeThreshold <span class="op">=</span> <span class="fl">60</span> </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>filenameTextEnd <span class="op">=</span> <span class="st">".MQ20.lowHet.tab"</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>tagName <span class="op">=</span> <span class="st">".Sept2023."</span>   <span class="co"># choose a tag name for this analysis</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># indicate name of metadata file, a text file with these column headings:</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ID    location    group   Fst_group   plot_order</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>metadataFile <span class="op">=</span> <span class="st">"GW_genomics_2022_with_new_genome/GW_all4plates.Fst_groups.txt"</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load metadata</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(metadataFile)) <span class="co"># the CSV.File function interprets the correct delimiter</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>num_metadata_cols <span class="op">=</span> <span class="fu">ncol</span>(metadata)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>num_individuals <span class="op">=</span> <span class="fu">nrow</span>(metadata) </span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co"># read in individual names for this dataset</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>individuals_file_name <span class="op">=</span> <span class="fu">string</span>(baseName, filenameTextMiddle, missingGenotypeThreshold, filenameTextEnd, <span class="st">".012.indv"</span>)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>ind <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(individuals_file_name; header<span class="op">=</span>[<span class="st">"ind"</span>], types<span class="op">=</span>[<span class="dt">String</span>])) </span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>indNum <span class="op">=</span> <span class="fu">size</span>(ind, <span class="fl">1</span>) <span class="co"># number of individuals</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> num_individuals <span class="op">!=</span> indNum</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: number of rows in metadata file different than number of individuals in .indv file"</span>)</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="co"># read in position data for this dataset</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>position_file_name <span class="op">=</span> <span class="fu">string</span>(baseName, filenameTextMiddle, missingGenotypeThreshold, filenameTextEnd, <span class="st">".012.pos"</span>)</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>pos_whole_genome <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(position_file_name; header<span class="op">=</span>[<span class="st">"chrom"</span>, <span class="st">"position"</span>], types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int</span>]))</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="co"># read in genotype data</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>column_names <span class="op">=</span> [<span class="st">"null"</span>; <span class="fu">string</span>.(<span class="st">"c."</span>, pos_whole_genome.chrom, <span class="st">"."</span>, pos_whole_genome.position)]    </span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>genotype_file_name <span class="op">=</span> <span class="fu">string</span>(baseName, filenameTextMiddle, missingGenotypeThreshold, filenameTextEnd, <span class="st">".012minus1"</span>) </span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a><span class="pp">@time</span> <span class="cf">if</span> <span class="fl">1</span> <span class="op">&lt;=</span> indNum <span class="op">&lt;=</span> <span class="fl">127</span>   </span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>    geno <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name, <span class="ch">'\t'</span>, <span class="dt">Int8</span>, <span class="ch">'\n'</span>); <span class="co"># this has been sped up dramatically, by first coverting "NA" to -1</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> <span class="fl">128</span> <span class="op">&lt;=</span> indNum <span class="op">&lt;=</span> <span class="fl">32767</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    geno <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name, <span class="ch">'\t'</span>, <span class="dt">Int16</span>, <span class="ch">'\n'</span>); <span class="co"># this needed for first column, which is number of individual; Int16 not much slower on import than Int8</span></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Error: Number of individuals in .indv appears outside of range from 1 to 32767"</span>)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>loci_count <span class="op">=</span> <span class="fu">size</span>(geno, <span class="fl">2</span>) <span class="op">-</span> <span class="fl">1</span>   <span class="co"># because the first column is not a SNP (just a count from zero)</span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">string</span>(<span class="st">"Read in genotypic data at "</span>, loci_count,<span class="st">" loci for "</span>, indNum, <span class="st">" individuals. </span><span class="sc">\n</span><span class="st">"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 62.411180 seconds (3.94 M allocations: 15.994 GiB, 0.27% gc time, 1.07% compilation time)
Read in genotypic data at 2431709 loci for 310 individuals. </code></pre>
</div>
</div>
<section id="check-that-individuals-are-same-in-genotype-data-and-metadata" class="level3">
<h3 class="anchored" data-anchor-id="check-that-individuals-are-same-in-genotype-data-and-metadata">Check that individuals are same in genotype data and metadata</h3>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata <span class="op">=</span> <span class="fu">hcat</span>(ind, metadata)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(ind_with_metadata)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>()  <span class="co"># prints a line break </span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">isequal</span>(ind_with_metadata.ind, ind_with_metadata.ID)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly."</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: names of individuals in metadata file and genotype ind file do not completely match."</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>310×6 DataFrame
 Row │ ind                             ID                              location  group           Fst_group       plot_order 
     │ String                          String31                        String7   String15        String15        Float64    
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1           GW_Armando_plate1_AB1           AB        vir             vir                 20.01
   2 │ GW_Armando_plate1_JF07G02       GW_Armando_plate1_JF07G02       ST        plumb           plumb               87.001
   3 │ GW_Armando_plate1_JF07G03       GW_Armando_plate1_JF07G03       ST        plumb           plumb               87.002
   4 │ GW_Armando_plate1_JF07G04       GW_Armando_plate1_JF07G04       ST        plumb           plumb               87.003
   5 │ GW_Armando_plate1_JF08G02       GW_Armando_plate1_JF08G02       ST        plumb           plumb               87.004
   6 │ GW_Armando_plate1_JF09G01       GW_Armando_plate1_JF09G01       ST        plumb           plumb               87.005
   7 │ GW_Armando_plate1_JF09G02       GW_Armando_plate1_JF09G02       ST        plumb           plumb               87.006
   8 │ GW_Armando_plate1_JF10G03       GW_Armando_plate1_JF10G03       ST        plumb_vir       plumb_vir           98.0
   9 │ GW_Armando_plate1_JF11G01       GW_Armando_plate1_JF11G01       ST        plumb           plumb               87.008
  10 │ GW_Armando_plate1_JF12G01       GW_Armando_plate1_JF12G01       ST        plumb           plumb               87.009
  11 │ GW_Armando_plate1_JF12G02       GW_Armando_plate1_JF12G02       ST        plumb           plumb               87.01
  12 │ GW_Armando_plate1_JF12G04       GW_Armando_plate1_JF12G04       ST_vi     vir             vir                 24.001
  13 │ GW_Armando_plate1_JF13G01       GW_Armando_plate1_JF13G01       ST        plumb           plumb               87.011
  14 │ GW_Armando_plate1_JF15G03       GW_Armando_plate1_JF15G03       KK        plumb           plumb               87.012
  15 │ GW_Armando_plate1_JF16G01       GW_Armando_plate1_JF16G01       KK_vi     plumb_vir       vir                 24.041
  16 │ GW_Armando_plate1_JF20G01       GW_Armando_plate1_JF20G01       KK        plumb           plumb               87.014
  17 │ GW_Armando_plate1_JF22G01       GW_Armando_plate1_JF22G01       KK        plumb           plumb               87.015
  18 │ GW_Armando_plate1_JF23G01       GW_Armando_plate1_JF23G01       KK        plumb           plumb               87.016
  19 │ GW_Armando_plate1_JF23G02       GW_Armando_plate1_JF23G02       KK        plumb           plumb               87.017
  20 │ GW_Armando_plate1_JF24G02       GW_Armando_plate1_JF24G02       KK        plumb           plumb               87.018
  21 │ GW_Armando_plate1_JF26G01       GW_Armando_plate1_JF26G01       ST        plumb           plumb               87.019
  22 │ GW_Armando_plate1_JF27G01       GW_Armando_plate1_JF27G01       ST        plumb           plumb               87.02
  23 │ GW_Armando_plate1_JF29G01       GW_Armando_plate1_JF29G01       ST        plumb           plumb               87.021
  24 │ GW_Armando_plate1_JF29G02       GW_Armando_plate1_JF29G02       ST        plumb           plumb               87.022
  25 │ GW_Armando_plate1_JF29G03       GW_Armando_plate1_JF29G03       ST        plumb           plumb               87.023
  26 │ GW_Armando_plate1_JG02G02       GW_Armando_plate1_JG02G02       KK        plumb           plumb               87.024
  27 │ GW_Armando_plate1_JG02G04       GW_Armando_plate1_JG02G04       KK        plumb           plumb               87.025
  28 │ GW_Armando_plate1_JG08G01       GW_Armando_plate1_JG08G01       ST        plumb           plumb               87.026
  29 │ GW_Armando_plate1_JG08G02       GW_Armando_plate1_JG08G02       ST        plumb           plumb               87.027
  30 │ GW_Armando_plate1_JG10G01       GW_Armando_plate1_JG10G01       ST        plumb           plumb               87.028
  31 │ GW_Armando_plate1_JG12G01       GW_Armando_plate1_JG12G01       ST        plumb           plumb               87.029
  32 │ GW_Armando_plate1_JG17G01       GW_Armando_plate1_JG17G01       ST        plumb_vir       plumb               77.92
  33 │ GW_Armando_plate1_NO_BC_TTGW05  GW_Armando_plate1_NO_BC_TTGW05  blank     blank           blank              -99.0
  34 │ GW_Armando_plate1_NO_DNA        GW_Armando_plate1_NO_DNA        blank     blank           blank              -99.0
  35 │ GW_Armando_plate1_RF20G01       GW_Armando_plate1_RF20G01       BJ        obs_plumb       plumb_BJ            77.501
  36 │ GW_Armando_plate1_RF29G02       GW_Armando_plate1_RF29G02       BJ        obs_plumb       plumb_BJ            77.502
  37 │ GW_Armando_plate1_TL3           GW_Armando_plate1_TL3           TL        vir             vir                 11.01
  38 │ GW_Armando_plate1_TTGW01        GW_Armando_plate1_TTGW01        MN        troch_MN        troch_west          53.0
  39 │ GW_Armando_plate1_TTGW05_rep1   GW_Armando_plate1_TTGW05_rep1   MN_rep    troch_MN_rep    troch_west_rep      53.0
  40 │ GW_Armando_plate1_TTGW05_rep2   GW_Armando_plate1_TTGW05_rep2   MN        troch_MN        troch_west          53.0
  41 │ GW_Armando_plate1_TTGW06        GW_Armando_plate1_TTGW06        SU        lud_Sukhto      lud_central         47.0
  42 │ GW_Armando_plate1_TTGW07        GW_Armando_plate1_TTGW07        SU        lud_Sukhto      lud_central         47.0
  43 │ GW_Armando_plate1_TTGW10        GW_Armando_plate1_TTGW10        SU        lud_Sukhto      lud_central         47.0
  44 │ GW_Armando_plate1_TTGW11        GW_Armando_plate1_TTGW11        SU        lud_Sukhto      lud_central         47.0
  45 │ GW_Armando_plate1_TTGW13        GW_Armando_plate1_TTGW13        TH        lud_Thallighar  lud_central         43.0
  46 │ GW_Armando_plate1_TTGW17        GW_Armando_plate1_TTGW17        TH        lud_Thallighar  lud_central         43.0
  47 │ GW_Armando_plate1_TTGW19        GW_Armando_plate1_TTGW19        TH        lud_Thallighar  lud_central         43.0
  48 │ GW_Armando_plate1_TTGW21        GW_Armando_plate1_TTGW21        SR        lud_Sural       lud_central         45.0
  49 │ GW_Armando_plate1_TTGW22        GW_Armando_plate1_TTGW22        SR        lud_Sural       lud_central         45.0
  50 │ GW_Armando_plate1_TTGW23        GW_Armando_plate1_TTGW23        SR        lud_Sural       lud_central         45.0
  51 │ GW_Armando_plate1_TTGW29        GW_Armando_plate1_TTGW29        SR        lud_Sural       lud_central         45.0
  52 │ GW_Armando_plate1_TTGW52        GW_Armando_plate1_TTGW52        NG        lud_Nainaghar   lud_central         49.0
  53 │ GW_Armando_plate1_TTGW53        GW_Armando_plate1_TTGW53        NG        lud_Nainaghar   lud_central         49.0
  54 │ GW_Armando_plate1_TTGW55        GW_Armando_plate1_TTGW55        NG        lud_Nainaghar   lud_central         49.0
  55 │ GW_Armando_plate1_TTGW57        GW_Armando_plate1_TTGW57        NG        lud_Nainaghar   lud_central         49.0
  56 │ GW_Armando_plate1_TTGW58        GW_Armando_plate1_TTGW58        NG        lud_Nainaghar   lud_central         49.0
  57 │ GW_Armando_plate1_TTGW59        GW_Armando_plate1_TTGW59        NG        lud_Nainaghar   lud_central         49.0
  58 │ GW_Armando_plate1_TTGW63        GW_Armando_plate1_TTGW63        SP        lud_Spiti       troch_west          55.0
  59 │ GW_Armando_plate1_TTGW64        GW_Armando_plate1_TTGW64        SP        lud_Spiti       troch_west          55.0
  60 │ GW_Armando_plate1_TTGW65        GW_Armando_plate1_TTGW65        SP        lud_Spiti       troch_west          55.0
  61 │ GW_Armando_plate1_TTGW66        GW_Armando_plate1_TTGW66        SP        lud_Spiti       troch_west          55.0
  62 │ GW_Armando_plate1_TTGW68        GW_Armando_plate1_TTGW68        SP        lud_Spiti       troch_west          55.0
  63 │ GW_Armando_plate1_TTGW70        GW_Armando_plate1_TTGW70        SA        lud_Sathrundi   lud_Sath            41.0
  64 │ GW_Armando_plate1_TTGW71        GW_Armando_plate1_TTGW71        SA        lud_Sathrundi   lud_Sath            41.0
  65 │ GW_Armando_plate1_TTGW72        GW_Armando_plate1_TTGW72        SA        lud_Sathrundi   lud_Sath            41.0
  66 │ GW_Armando_plate1_TTGW74        GW_Armando_plate1_TTGW74        SA        lud_Sathrundi   lud_Sath            41.0
  67 │ GW_Armando_plate1_TTGW78        GW_Armando_plate1_TTGW78        SA        lud_Sathrundi   lud_Sath            41.0
  68 │ GW_Armando_plate1_TTGW_15_05    GW_Armando_plate1_TTGW_15_05    SR        lud_Sural       lud_central         45.0
  69 │ GW_Armando_plate1_TTGW_15_07    GW_Armando_plate1_TTGW_15_07    SR        lud_Sural       lud_central         45.0
  70 │ GW_Armando_plate1_TTGW_15_08    GW_Armando_plate1_TTGW_15_08    SR        lud_Sural       lud_central         45.0
  71 │ GW_Armando_plate1_TTGW_15_09    GW_Armando_plate1_TTGW_15_09    SR        lud_Sural       lud_central         45.0
  72 │ GW_Armando_plate1_UY1           GW_Armando_plate1_UY1           UY        plumb           plumb               88.01
  73 │ GW_Armando_plate2_IL2           GW_Armando_plate2_IL2           IL_rep    plumb_rep       plumb_rep           83.01
  74 │ GW_Armando_plate2_JE31G01       GW_Armando_plate2_JE31G01       KK_vi     vir_misID       vir                 24.002
  75 │ GW_Armando_plate2_JF03G01       GW_Armando_plate2_JF03G01       ST_vi     vir_misID       vir                 24.003
  76 │ GW_Armando_plate2_JF03G02       GW_Armando_plate2_JF03G02       KK_vi     vir_misID       vir                 24.004
  77 │ GW_Armando_plate2_JF07G01       GW_Armando_plate2_JF07G01       ST        plumb           plumb               87.031
  78 │ GW_Armando_plate2_JF08G04       GW_Armando_plate2_JF08G04       ST        plumb           plumb               87.032
  79 │ GW_Armando_plate2_JF10G02       GW_Armando_plate2_JF10G02       ST        plumb           plumb               87.033
  80 │ GW_Armando_plate2_JF11G02       GW_Armando_plate2_JF11G02       ST        plumb           plumb               87.034
  81 │ GW_Armando_plate2_JF12G03       GW_Armando_plate2_JF12G03       ST        plumb           plumb               87.035
  82 │ GW_Armando_plate2_JF12G05       GW_Armando_plate2_JF12G05       ST        plumb           plumb               87.036
  83 │ GW_Armando_plate2_JF13G02       GW_Armando_plate2_JF13G02       ST        plumb           plumb               87.037
  84 │ GW_Armando_plate2_JF14G01       GW_Armando_plate2_JF14G01       KK        plumb           plumb               87.038
  85 │ GW_Armando_plate2_JF14G02       GW_Armando_plate2_JF14G02       KK        plumb           plumb               87.039
  86 │ GW_Armando_plate2_JF15G01       GW_Armando_plate2_JF15G01       KK        plumb           plumb               87.04
  87 │ GW_Armando_plate2_JF15G02       GW_Armando_plate2_JF15G02       KK        plumb           plumb               87.041
  88 │ GW_Armando_plate2_JF16G02       GW_Armando_plate2_JF16G02       KK_vi     plumb_vir       vir                 24.042
  89 │ GW_Armando_plate2_JF19G01       GW_Armando_plate2_JF19G01       KK        plumb           plumb               87.043
  90 │ GW_Armando_plate2_JF20G02       GW_Armando_plate2_JF20G02       KK        plumb           plumb               87.044
  91 │ GW_Armando_plate2_JF24G01       GW_Armando_plate2_JF24G01       KK        plumb           plumb               87.045
  92 │ GW_Armando_plate2_JF24G03       GW_Armando_plate2_JF24G03       ST        plumb           plumb               87.046
  93 │ GW_Armando_plate2_JF25G01       GW_Armando_plate2_JF25G01       KK        plumb           plumb               87.047
  94 │ GW_Armando_plate2_JF26G02       GW_Armando_plate2_JF26G02       KK        plumb           plumb               87.048
  95 │ GW_Armando_plate2_JF27G02       GW_Armando_plate2_JF27G02       KK        plumb           plumb               87.049
  96 │ GW_Armando_plate2_JF30G01       GW_Armando_plate2_JF30G01       ST_vi     vir_misID       vir                 24.005
  97 │ GW_Armando_plate2_JG01G01       GW_Armando_plate2_JG01G01       KK        plumb           plumb               87.05
  98 │ GW_Armando_plate2_JG02G01       GW_Armando_plate2_JG02G01       KK        plumb           plumb               87.051
  99 │ GW_Armando_plate2_JG02G03       GW_Armando_plate2_JG02G03       KK        plumb           plumb               87.052
 100 │ GW_Armando_plate2_JG10G02       GW_Armando_plate2_JG10G02       ST        plumb           plumb               87.053
 101 │ GW_Armando_plate2_JG10G03       GW_Armando_plate2_JG10G03       ST        plumb           plumb               87.054
 102 │ GW_Armando_plate2_JG12G02       GW_Armando_plate2_JG12G02       ST        plumb           plumb               87.055
 103 │ GW_Armando_plate2_JG12G03       GW_Armando_plate2_JG12G03       ST        plumb           plumb               87.056
 104 │ GW_Armando_plate2_LN11          GW_Armando_plate2_LN11          LN_rep    troch_LN_rep    troch_LN_rep        65.01
 105 │ GW_Armando_plate2_LN2           GW_Armando_plate2_LN2           LN        troch_LN        troch_LN            58.01
 106 │ GW_Armando_plate2_NO_BC_TTGW05  GW_Armando_plate2_NO_BC_TTGW05  blank     blank           blank              -99.0
 107 │ GW_Armando_plate2_NO_DNA        GW_Armando_plate2_NO_DNA        blank     blank           blank              -99.0
 108 │ GW_Armando_plate2_RF29G01       GW_Armando_plate2_RF29G01       BJ        obs_plumb       plumb_BJ            77.503
 109 │ GW_Armando_plate2_TTGW02        GW_Armando_plate2_TTGW02        MN        troch_MN        troch_west          53.0
 110 │ GW_Armando_plate2_TTGW03        GW_Armando_plate2_TTGW03        MN        troch_MN        troch_west          53.0
 111 │ GW_Armando_plate2_TTGW05_rep3   GW_Armando_plate2_TTGW05_rep3   MN_rep    troch_MN_rep    troch_west_rep      53.0
 112 │ GW_Armando_plate2_TTGW05_rep4   GW_Armando_plate2_TTGW05_rep4   MN_rep    troch_MN_rep    troch_west_rep      53.0
 113 │ GW_Armando_plate2_TTGW08        GW_Armando_plate2_TTGW08        SU        lud_Sukhto      lud_central         47.0
 114 │ GW_Armando_plate2_TTGW09        GW_Armando_plate2_TTGW09        SU        lud_Sukhto      lud_central         47.0
 115 │ GW_Armando_plate2_TTGW12        GW_Armando_plate2_TTGW12        TH        lud_Thallighar  lud_central         43.0
 116 │ GW_Armando_plate2_TTGW14        GW_Armando_plate2_TTGW14        TH        lud_Thallighar  lud_central         43.0
 117 │ GW_Armando_plate2_TTGW15        GW_Armando_plate2_TTGW15        TH        lud_Thallighar  lud_central         43.0
 118 │ GW_Armando_plate2_TTGW16        GW_Armando_plate2_TTGW16        TH        lud_Thallighar  lud_central         43.0
 119 │ GW_Armando_plate2_TTGW18        GW_Armando_plate2_TTGW18        TH        lud_Thallighar  lud_central         43.0
 120 │ GW_Armando_plate2_TTGW20        GW_Armando_plate2_TTGW20        SR        lud_Sural       lud_central         45.0
 121 │ GW_Armando_plate2_TTGW24        GW_Armando_plate2_TTGW24        SR        lud_Sural       lud_central         45.0
 122 │ GW_Armando_plate2_TTGW25        GW_Armando_plate2_TTGW25        SR        lud_Sural       lud_central         45.0
 123 │ GW_Armando_plate2_TTGW27        GW_Armando_plate2_TTGW27        SR        lud_Sural       lud_central         45.0
 124 │ GW_Armando_plate2_TTGW28        GW_Armando_plate2_TTGW28        SR        lud_Sural       lud_central         45.0
 125 │ GW_Armando_plate2_TTGW50        GW_Armando_plate2_TTGW50        NG        lud_Nainaghar   lud_central         49.0
 126 │ GW_Armando_plate2_TTGW51        GW_Armando_plate2_TTGW51        NG        lud_Nainaghar   lud_central         49.0
 127 │ GW_Armando_plate2_TTGW54        GW_Armando_plate2_TTGW54        NG        lud_Nainaghar   lud_central         49.0
 128 │ GW_Armando_plate2_TTGW56        GW_Armando_plate2_TTGW56        NG        lud_Nainaghar   lud_central         49.0
 129 │ GW_Armando_plate2_TTGW60        GW_Armando_plate2_TTGW60        SP        lud_Spiti       troch_west          55.0
 130 │ GW_Armando_plate2_TTGW61        GW_Armando_plate2_TTGW61        SP        lud_Spiti       troch_west          55.0
 131 │ GW_Armando_plate2_TTGW62        GW_Armando_plate2_TTGW62        SP        lud_Spiti       troch_west          55.0
 132 │ GW_Armando_plate2_TTGW67        GW_Armando_plate2_TTGW67        SP        lud_Spiti       troch_west          55.0
 133 │ GW_Armando_plate2_TTGW69        GW_Armando_plate2_TTGW69        SP        lud_Spiti       troch_west          55.0
 134 │ GW_Armando_plate2_TTGW73        GW_Armando_plate2_TTGW73        SA        lud_Sathrundi   lud_Sath            41.0
 135 │ GW_Armando_plate2_TTGW75        GW_Armando_plate2_TTGW75        SA        lud_Sathrundi   lud_Sath            41.0
 136 │ GW_Armando_plate2_TTGW77        GW_Armando_plate2_TTGW77        SA        lud_Sathrundi   lud_Sath            41.0
 137 │ GW_Armando_plate2_TTGW79        GW_Armando_plate2_TTGW79        SA        lud_Sathrundi   lud_Sath            41.0
 138 │ GW_Armando_plate2_TTGW80        GW_Armando_plate2_TTGW80        SA        lud_Sathrundi   lud_Sath            41.0
 139 │ GW_Armando_plate2_TTGW_15_01    GW_Armando_plate2_TTGW_15_01    SR        lud_Sural       lud_central         45.0
 140 │ GW_Armando_plate2_TTGW_15_02    GW_Armando_plate2_TTGW_15_02    SR        lud_Sural       lud_central         45.0
 141 │ GW_Armando_plate2_TTGW_15_03    GW_Armando_plate2_TTGW_15_03    SR        lud_Sural       lud_central         45.0
 142 │ GW_Armando_plate2_TTGW_15_04    GW_Armando_plate2_TTGW_15_04    SR        lud_Sural       lud_central         45.0
 143 │ GW_Armando_plate2_TTGW_15_06    GW_Armando_plate2_TTGW_15_06    SR        lud_Sural       lud_central         45.0
 144 │ GW_Armando_plate2_TTGW_15_10    GW_Armando_plate2_TTGW_15_10    SR        lud_Sural       lud_central         45.0
 145 │ GW_Lane5_AA1                    GW_Lane5_AA1                    AA        vir_S           vir_S               25.0
 146 │ GW_Lane5_AA10                   GW_Lane5_AA10                   AA        vir_S           vir_S               33.0
 147 │ GW_Lane5_AA11                   GW_Lane5_AA11                   AA        vir_S           vir_S               34.0
 148 │ GW_Lane5_AA3                    GW_Lane5_AA3                    AA        vir_S           vir_S               26.0
 149 │ GW_Lane5_AA4                    GW_Lane5_AA4                    AA        vir_S           vir_S               27.0
 150 │ GW_Lane5_AA5                    GW_Lane5_AA5                    AA        vir_S           vir_S               28.0
 151 │ GW_Lane5_AA6                    GW_Lane5_AA6                    AA        vir_S           vir_S               29.0
 152 │ GW_Lane5_AA7                    GW_Lane5_AA7                    AA        vir_S           vir_S               30.0
 153 │ GW_Lane5_AA8                    GW_Lane5_AA8                    AA        vir_S           vir_S               31.0
 154 │ GW_Lane5_AA9                    GW_Lane5_AA9                    AA        vir_S           vir_S               32.0
 155 │ GW_Lane5_AB1                    GW_Lane5_AB1                    AB_rep    vir_rep         vir_rep             20.0
 156 │ GW_Lane5_AB2                    GW_Lane5_AB2                    AB        vir             vir                 21.0
 157 │ GW_Lane5_AN1                    GW_Lane5_AN1                    AN        plumb           plumb               80.0
 158 │ GW_Lane5_AN2                    GW_Lane5_AN2                    AN        plumb           plumb               81.0
 159 │ GW_Lane5_BK2                    GW_Lane5_BK2                    BK        plumb           plumb               78.0
 160 │ GW_Lane5_BK3                    GW_Lane5_BK3                    BK        plumb           plumb               79.0
 161 │ GW_Lane5_DA2                    GW_Lane5_DA2                    XN        obs             obs                 73.0
 162 │ GW_Lane5_DA3                    GW_Lane5_DA3                    XN        obs             obs                 74.0
 163 │ GW_Lane5_DA4                    GW_Lane5_DA4                    XN        obs             obs                 75.0
 164 │ GW_Lane5_DA6                    GW_Lane5_DA6                    XN        obs             low_reads           76.0
 165 │ GW_Lane5_DA7                    GW_Lane5_DA7                    XN        obs             obs                 77.0
 166 │ GW_Lane5_EM1                    GW_Lane5_EM1                    EM        troch_EM        troch_EM            72.0
 167 │ GW_Lane5_IL1                    GW_Lane5_IL1                    IL        plumb           plumb               82.0
 168 │ GW_Lane5_IL2                    GW_Lane5_IL2                    IL_rep    plumb_rep       plumb_rep           83.0
 169 │ GW_Lane5_IL4                    GW_Lane5_IL4                    IL        plumb           plumb               84.0
 170 │ GW_Lane5_KS1                    GW_Lane5_KS1                    OV        lud_KS          lud_KS              40.0
 171 │ GW_Lane5_KS2                    GW_Lane5_KS2                    OV        lud_KS          lud_KS              40.0
 172 │ GW_Lane5_LN1                    GW_Lane5_LN1                    LN        troch_LN        troch_LN            57.0
 173 │ GW_Lane5_LN10                   GW_Lane5_LN10                   LN        troch_LN        troch_LN            64.0
 174 │ GW_Lane5_LN11                   GW_Lane5_LN11                   LN        troch_LN        troch_LN            65.0
 175 │ GW_Lane5_LN12                   GW_Lane5_LN12                   LN        troch_LN        troch_LN            66.0
 176 │ GW_Lane5_LN14                   GW_Lane5_LN14                   LN        troch_LN        troch_LN            67.0
 177 │ GW_Lane5_LN16                   GW_Lane5_LN16                   LN        troch_LN        troch_LN            68.0
 178 │ GW_Lane5_LN18                   GW_Lane5_LN18                   LN        troch_LN        troch_LN            69.0
 179 │ GW_Lane5_LN19                   GW_Lane5_LN19                   LN        troch_LN        troch_LN            70.0
 180 │ GW_Lane5_LN2                    GW_Lane5_LN2                    LN_rep    troch_LN_rep    troch_LN_rep        58.0
 181 │ GW_Lane5_LN20                   GW_Lane5_LN20                   LN        troch_LN        troch_LN            71.0
 182 │ GW_Lane5_LN3                    GW_Lane5_LN3                    LN        troch_LN        troch_LN            59.0
 183 │ GW_Lane5_LN4                    GW_Lane5_LN4                    LN        troch_LN        troch_LN            60.0
 184 │ GW_Lane5_LN6                    GW_Lane5_LN6                    LN        troch_LN        troch_LN            61.0
 185 │ GW_Lane5_LN7                    GW_Lane5_LN7                    LN        troch_LN        troch_LN            62.0
 186 │ GW_Lane5_LN8                    GW_Lane5_LN8                    LN        troch_LN        troch_LN            63.0
 187 │ GW_Lane5_MN1                    GW_Lane5_MN1                    MN        troch_MN        troch_west          51.0
 188 │ GW_Lane5_MN12                   GW_Lane5_MN12                   MN        troch_MN        troch_west          56.0
 189 │ GW_Lane5_MN3                    GW_Lane5_MN3                    MN        troch_MN        troch_west          52.0
 190 │ GW_Lane5_MN5                    GW_Lane5_MN5                    MN        troch_MN        troch_west          53.0
 191 │ GW_Lane5_MN8                    GW_Lane5_MN8                    MN        troch_MN        troch_west          54.0
 192 │ GW_Lane5_MN9                    GW_Lane5_MN9                    MN        troch_MN        troch_west          55.0
 193 │ GW_Lane5_NA1                    GW_Lane5_NA1                    NR        lud_PK          lud_PK              39.2
 194 │ GW_Lane5_NA3-3ul                GW_Lane5_NA3-3ul                NR        lud_PK          lud_PK              39.2
 195 │ GW_Lane5_PT11                   GW_Lane5_PT11                   KL        lud_KL          lud_central         42.0
 196 │ GW_Lane5_PT12                   GW_Lane5_PT12                   KL        lud_KL          lud_central         42.0
 197 │ GW_Lane5_PT2                    GW_Lane5_PT2                    ML        lud_ML          lud_ML              51.0
 198 │ GW_Lane5_PT3                    GW_Lane5_PT3                    PA        lud_PA          lud_central         46.0
 199 │ GW_Lane5_PT4                    GW_Lane5_PT4                    PA        lud_PA          lud_central         46.0
 200 │ GW_Lane5_PT6                    GW_Lane5_PT6                    KL        lud_KL          lud_central         42.0
 201 │ GW_Lane5_SH1                    GW_Lane5_SH1                    PK        lud_PK          lud_PK              39.1
 202 │ GW_Lane5_SH2                    GW_Lane5_SH2                    PK        lud_PK          lud_PK              39.1
 203 │ GW_Lane5_SH4                    GW_Lane5_SH4                    PK        lud_PK          lud_PK              39.1
 204 │ GW_Lane5_SH5                    GW_Lane5_SH5                    PK        lud_PK          lud_PK              39.1
 205 │ GW_Lane5_SL1                    GW_Lane5_SL1                    SL        plumb           plumb               95.0
 206 │ GW_Lane5_SL2                    GW_Lane5_SL2                    SL        plumb           plumb               96.0
 207 │ GW_Lane5_ST1                    GW_Lane5_ST1                    ST        plumb           plumb               85.0
 208 │ GW_Lane5_ST12                   GW_Lane5_ST12                   ST        plumb           plumb               87.0
 209 │ GW_Lane5_ST3                    GW_Lane5_ST3                    ST        plumb           plumb               86.0
 210 │ GW_Lane5_STvi1                  GW_Lane5_STvi1                  STvi      vir             vir                 22.0
 211 │ GW_Lane5_STvi2                  GW_Lane5_STvi2                  STvi      vir             vir                 23.0
 212 │ GW_Lane5_STvi3                  GW_Lane5_STvi3                  STvi      vir             vir                 24.0
 213 │ GW_Lane5_TA1                    GW_Lane5_TA1                    TA        plumb           plumb               94.0
 214 │ GW_Lane5_TL1                    GW_Lane5_TL1                    TL        vir             vir                  9.0
 215 │ GW_Lane5_TL10                   GW_Lane5_TL10                   TL        vir             vir                 17.0
 216 │ GW_Lane5_TL11                   GW_Lane5_TL11                   TL        vir             vir                 18.0
 217 │ GW_Lane5_TL12                   GW_Lane5_TL12                   TL        vir             vir                 19.0
 218 │ GW_Lane5_TL2                    GW_Lane5_TL2                    TL        vir             vir                 10.0
 219 │ GW_Lane5_TL3                    GW_Lane5_TL3                    TL_rep    vir_rep         vir_rep             11.0
 220 │ GW_Lane5_TL4                    GW_Lane5_TL4                    TL        vir             vir                 12.0
 221 │ GW_Lane5_TL5                    GW_Lane5_TL5                    TL        vir             vir                 13.0
 222 │ GW_Lane5_TL7                    GW_Lane5_TL7                    TL        vir             vir                 14.0
 223 │ GW_Lane5_TL8                    GW_Lane5_TL8                    TL        vir             vir                 15.0
 224 │ GW_Lane5_TL9                    GW_Lane5_TL9                    TL        vir             vir                 16.0
 225 │ GW_Lane5_TU1                    GW_Lane5_TU1                    TU        nit             nit                 35.0
 226 │ GW_Lane5_TU2                    GW_Lane5_TU2                    TU        nit             nit                 36.0
 227 │ GW_Lane5_UY1                    GW_Lane5_UY1                    UY_rep    plumb_rep       plumb_rep           88.0
 228 │ GW_Lane5_UY2                    GW_Lane5_UY2                    UY        plumb           plumb               89.0
 229 │ GW_Lane5_UY3                    GW_Lane5_UY3                    UY        plumb           plumb               90.0
 230 │ GW_Lane5_UY4                    GW_Lane5_UY4                    UY        plumb           plumb               91.0
 231 │ GW_Lane5_UY5                    GW_Lane5_UY5                    UY        plumb           plumb               92.0
 232 │ GW_Lane5_UY6                    GW_Lane5_UY6                    UY        plumb           plumb               93.0
 233 │ GW_Lane5_YK1                    GW_Lane5_YK1                    YK        vir             vir                  1.0
 234 │ GW_Lane5_YK11                   GW_Lane5_YK11                   YK        vir             vir                  8.0
 235 │ GW_Lane5_YK3                    GW_Lane5_YK3                    YK        vir             vir                  2.0
 236 │ GW_Lane5_YK4                    GW_Lane5_YK4                    YK        vir             vir                  3.0
 237 │ GW_Lane5_YK5                    GW_Lane5_YK5                    YK        vir             vir                  4.0
 238 │ GW_Lane5_YK6                    GW_Lane5_YK6                    YK        vir             vir                  5.0
 239 │ GW_Lane5_YK7                    GW_Lane5_YK7                    YK        vir             vir                  6.0
 240 │ GW_Lane5_YK9                    GW_Lane5_YK9                    YK        vir             vir                  7.0
 241 │ GW_Liz_GBS_Liz10045             GW_Liz_GBS_Liz10045             ML        lud             lud_ML              51.01
 242 │ GW_Liz_GBS_Liz10094             GW_Liz_GBS_Liz10094             ML        lud             lud_ML              51.02
 243 │ GW_Liz_GBS_Liz5101              GW_Liz_GBS_Liz5101              ML        lud             lud_ML              51.03
 244 │ GW_Liz_GBS_Liz5101_R            GW_Liz_GBS_Liz5101_R            ML_rep    lud_rep         lud_ML_rep          51.04
 245 │ GW_Liz_GBS_Liz5118              GW_Liz_GBS_Liz5118              ML        lud             lud_ML              51.05
 246 │ GW_Liz_GBS_Liz5139              GW_Liz_GBS_Liz5139              ML        lud             lud_ML              51.06
 247 │ GW_Liz_GBS_Liz5142              GW_Liz_GBS_Liz5142              ML        lud             lud_ML              51.07
 248 │ GW_Liz_GBS_Liz5144              GW_Liz_GBS_Liz5144              ML        lud             lud_ML              51.08
 249 │ GW_Liz_GBS_Liz5150              GW_Liz_GBS_Liz5150              ML        lud             lud_ML              51.09
 250 │ GW_Liz_GBS_Liz5159              GW_Liz_GBS_Liz5159              ML        lud_chick       lud_ML              51.1
 251 │ GW_Liz_GBS_Liz5162              GW_Liz_GBS_Liz5162              ML        lud_chick       lud_ML              51.11
 252 │ GW_Liz_GBS_Liz5163              GW_Liz_GBS_Liz5163              ML        lud_chick       lud_ML              51.12
 253 │ GW_Liz_GBS_Liz5164              GW_Liz_GBS_Liz5164              ML        lud_chick       lud_ML              51.13
 254 │ GW_Liz_GBS_Liz5165              GW_Liz_GBS_Liz5165              ML        lud             lud_ML              51.14
 255 │ GW_Liz_GBS_Liz5167              GW_Liz_GBS_Liz5167              ML        lud_chick       lud_ML              51.15
 256 │ GW_Liz_GBS_Liz5168              GW_Liz_GBS_Liz5168              ML        lud_chick       lud_ML              51.16
 257 │ GW_Liz_GBS_Liz5169              GW_Liz_GBS_Liz5169              ML        lud_chick       lud_ML              51.17
 258 │ GW_Liz_GBS_Liz5171              GW_Liz_GBS_Liz5171              ML        lud             lud_ML              51.18
 259 │ GW_Liz_GBS_Liz5172              GW_Liz_GBS_Liz5172              ML        lud_chick       lud_ML              51.19
 260 │ GW_Liz_GBS_Liz5173              GW_Liz_GBS_Liz5173              ML        lud_chick       lud_ML              51.2
 261 │ GW_Liz_GBS_Liz5174              GW_Liz_GBS_Liz5174              ML        lud             lud_ML              51.21
 262 │ GW_Liz_GBS_Liz5175              GW_Liz_GBS_Liz5175              ML        lud             lud_ML              51.22
 263 │ GW_Liz_GBS_Liz5176              GW_Liz_GBS_Liz5176              ML        lud             lud_ML              51.23
 264 │ GW_Liz_GBS_Liz5177              GW_Liz_GBS_Liz5177              ML        lud_chick       lud_ML              51.24
 265 │ GW_Liz_GBS_Liz5178              GW_Liz_GBS_Liz5178              ML        lud_chick       lud_ML              51.25
 266 │ GW_Liz_GBS_Liz5179              GW_Liz_GBS_Liz5179              ML        lud_chick       lud_ML              51.26
 267 │ GW_Liz_GBS_Liz5180              GW_Liz_GBS_Liz5180              ML        lud             lud_ML              51.27
 268 │ GW_Liz_GBS_Liz5182              GW_Liz_GBS_Liz5182              ML        lud_chick       lud_ML              51.28
 269 │ GW_Liz_GBS_Liz5184              GW_Liz_GBS_Liz5184              ML        lud_chick       lud_ML              51.29
 270 │ GW_Liz_GBS_Liz5185              GW_Liz_GBS_Liz5185              ML        lud             lud_ML              51.3
 271 │ GW_Liz_GBS_Liz5186              GW_Liz_GBS_Liz5186              ML        lud_chick       lud_ML              51.31
 272 │ GW_Liz_GBS_Liz5187              GW_Liz_GBS_Liz5187              ML        lud_chick       lud_ML              51.32
 273 │ GW_Liz_GBS_Liz5188              GW_Liz_GBS_Liz5188              ML        lud             lud_ML              51.33
 274 │ GW_Liz_GBS_Liz5189              GW_Liz_GBS_Liz5189              ML        lud_chick       lud_ML              51.34
 275 │ GW_Liz_GBS_Liz5190              GW_Liz_GBS_Liz5190              ML        lud_chick       lud_ML              51.35
 276 │ GW_Liz_GBS_Liz5191              GW_Liz_GBS_Liz5191              ML        lud_chick       lud_ML              51.36
 277 │ GW_Liz_GBS_Liz5192              GW_Liz_GBS_Liz5192              ML        lud_chick       lud_ML              51.37
 278 │ GW_Liz_GBS_Liz5193              GW_Liz_GBS_Liz5193              ML        lud_chick       lud_ML              51.38
 279 │ GW_Liz_GBS_Liz5194              GW_Liz_GBS_Liz5194              ML        lud_chick       lud_ML              51.39
 280 │ GW_Liz_GBS_Liz5195              GW_Liz_GBS_Liz5195              ML        lud             lud_ML              51.4
 281 │ GW_Liz_GBS_Liz5197              GW_Liz_GBS_Liz5197              ML        lud             lud_ML              51.41
 282 │ GW_Liz_GBS_Liz5199              GW_Liz_GBS_Liz5199              ML        lud_chick       lud_ML              51.42
 283 │ GW_Liz_GBS_Liz6002              GW_Liz_GBS_Liz6002              ML        lud             lud_ML              51.43
 284 │ GW_Liz_GBS_Liz6006              GW_Liz_GBS_Liz6006              ML        lud             lud_ML              51.44
 285 │ GW_Liz_GBS_Liz6008              GW_Liz_GBS_Liz6008              ML        lud             lud_ML              51.45
 286 │ GW_Liz_GBS_Liz6009              GW_Liz_GBS_Liz6009              ML        lud             lud_ML              51.46
 287 │ GW_Liz_GBS_Liz6010              GW_Liz_GBS_Liz6010              ML        lud             lud_ML              51.47
 288 │ GW_Liz_GBS_Liz6012              GW_Liz_GBS_Liz6012              ML        lud             lud_ML              51.48
 289 │ GW_Liz_GBS_Liz6014              GW_Liz_GBS_Liz6014              ML        lud             lud_ML              51.49
 290 │ GW_Liz_GBS_Liz6055              GW_Liz_GBS_Liz6055              ML        lud             lud_ML              51.5
 291 │ GW_Liz_GBS_Liz6057              GW_Liz_GBS_Liz6057              ML        lud             lud_ML              51.51
 292 │ GW_Liz_GBS_Liz6060              GW_Liz_GBS_Liz6060              ML        lud             lud_ML              51.52
 293 │ GW_Liz_GBS_Liz6062              GW_Liz_GBS_Liz6062              ML        lud             lud_ML              51.53
 294 │ GW_Liz_GBS_Liz6063              GW_Liz_GBS_Liz6063              ML        lud             lud_ML              51.54
 295 │ GW_Liz_GBS_Liz6066              GW_Liz_GBS_Liz6066              ML        lud             lud_ML              51.55
 296 │ GW_Liz_GBS_Liz6072              GW_Liz_GBS_Liz6072              ML        lud             lud_ML              51.56
 297 │ GW_Liz_GBS_Liz6079              GW_Liz_GBS_Liz6079              ML        lud             lud_ML              51.57
 298 │ GW_Liz_GBS_Liz6203              GW_Liz_GBS_Liz6203              ML        lud_chick       lud_ML              51.58
 299 │ GW_Liz_GBS_Liz6204              GW_Liz_GBS_Liz6204              ML        lud_chick       lud_ML              51.59
 300 │ GW_Liz_GBS_Liz6461              GW_Liz_GBS_Liz6461              ML        lud             lud_ML              51.6
 301 │ GW_Liz_GBS_Liz6472              GW_Liz_GBS_Liz6472              ML        lud             lud_ML              51.61
 302 │ GW_Liz_GBS_Liz6478              GW_Liz_GBS_Liz6478              ML        lud             lud_ML              51.62
 303 │ GW_Liz_GBS_Liz6766              GW_Liz_GBS_Liz6766              ML        lud             lud_ML              51.63
 304 │ GW_Liz_GBS_Liz6776              GW_Liz_GBS_Liz6776              ML        lud             lud_ML              51.64
 305 │ GW_Liz_GBS_Liz6794              GW_Liz_GBS_Liz6794              ML        lud             lud_ML              51.65
 306 │ GW_Liz_GBS_P_fusc               GW_Liz_GBS_P_fusc               fusc      fusc            fusc               101.0
 307 │ GW_Liz_GBS_P_h_man              GW_Liz_GBS_P_h_man              hmand     hmand           hmand              102.0
 308 │ GW_Liz_GBS_P_humei              GW_Liz_GBS_P_humei              hume      hume            hume               103.0
 309 │ GW_Liz_GBS_P_inor               GW_Liz_GBS_P_inor               inor      inor            inor               104.0
 310 │ GW_Liz_GBS_S_burk               GW_Liz_GBS_S_burk               burk      burk            burk               105.0

GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly.</code></pre>
</div>
</div>
</section>
</section>
<section id="filtering" class="level2">
<h2 class="anchored" data-anchor-id="filtering">Filtering</h2>
<section id="filter-out-duplicate-runs-indicated-with-_rep-in-fst_group-column" class="level4">
<h4 class="anchored" data-anchor-id="filter-out-duplicate-runs-indicated-with-_rep-in-fst_group-column">Filter out duplicate runs (indicated with <code>_rep</code> in <code>Fst_group</code> column)</h4>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> <span class="fu">occursin</span>.(<span class="st">"_rep"</span>, ind_with_metadata.Fst_group)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"""Filtering out these runs because they are duplicates of another,</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="st">    according to having "rep" in Fst_group: """</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(ind_with_metadata.ind[selection])</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered <span class="op">=</span> ind_with_metadata[<span class="fu">Not</span>(selection), <span class="op">:</span>];</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    geno_indFiltered <span class="op">=</span> <span class="fu">view</span>(geno, <span class="fu">Not</span>(selection), <span class="op">:</span>);  <span class="co"># use of view() avoids copying large memory object</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering out these runs because they are duplicates of another,
according to having "rep" in Fst_group: </code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>11-element Vector{String}:
 "GW_Armando_plate1_TTGW05_rep1"
 "GW_Armando_plate2_IL2"
 "GW_Armando_plate2_LN11"
 "GW_Armando_plate2_TTGW05_rep3"
 "GW_Armando_plate2_TTGW05_rep4"
 "GW_Lane5_AB1"
 "GW_Lane5_IL2"
 "GW_Lane5_LN2"
 "GW_Lane5_TL3"
 "GW_Lane5_UY1"
 "GW_Liz_GBS_Liz5101_R"</code></pre>
</div>
</div>
</section>
<section id="filter-specific-individuals" class="level4">
<h4 class="anchored" data-anchor-id="filter-specific-individuals">Filter specific individuals</h4>
<p>If there are certain individuals that we want to filter out prior to any additional analysis, we can do so here by setting filter to <code>true</code> and specifying the individual row numbers in <code>filter_out_inds</code>:</p>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>filter <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Specify individuals to filter out:</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>filter_out_inds <span class="op">=</span> [<span class="st">"GW_Liz_GBS_P_fusc"</span>, <span class="st">"GW_Liz_GBS_P_h_man"</span>, <span class="st">"GW_Liz_GBS_P_humei"</span>, <span class="st">"GW_Liz_GBS_P_inor"</span>, <span class="st">"GW_Liz_GBS_S_burk"</span>]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> filter</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(filter_out_inds), ind_with_metadata_indFiltered.ind)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    filtered_out <span class="op">=</span> ind_with_metadata_indFiltered.ind[selection]</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered <span class="op">=</span> ind_with_metadata_indFiltered[<span class="fu">Not</span>(selection), <span class="op">:</span>]</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    geno_indFiltered <span class="op">=</span> <span class="fu">view</span>(geno_indFiltered, <span class="fu">Not</span>(selection), <span class="op">:</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Specific individuals filtered out as requested: "</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(filtered_out)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"No specific individuals filtered (because filter not true)"</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Specific individuals filtered out as requested: </code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>5-element Vector{String}:
 "GW_Liz_GBS_P_fusc"
 "GW_Liz_GBS_P_h_man"
 "GW_Liz_GBS_P_humei"
 "GW_Liz_GBS_P_inor"
 "GW_Liz_GBS_S_burk"</code></pre>
</div>
</div>
</section>
<section id="filter-individuals-based-on-missing-genotypes" class="level4">
<h4 class="anchored" data-anchor-id="filter-individuals-based-on-missing-genotypes">Filter individuals based on missing genotypes</h4>
<p>Here we determine number of missing SNPs per individual (40% for this round), and filter out those individual datasets with more than a certain percent of missing SNPs:</p>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>SNPmissing_percent_allowed_per_ind <span class="op">=</span> <span class="fl">40</span>   <span class="co"># this is the percentage threshold</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>threshold_missing <span class="op">=</span> loci_count <span class="op">*</span> SNPmissing_percent_allowed_per_ind<span class="op">/</span><span class="fl">100</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>numMissings <span class="op">=</span> <span class="fu">sum</span>(geno_indFiltered <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.numMissings <span class="op">.=</span> numMissings</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">vec</span>(numMissings <span class="op">.&lt;=</span> threshold_missing) <span class="co"># the vec command converts to BitVector rather than BitMatrix--important below</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Filtering out these individuals based on too many missing genotypes: "</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>filtered_inds <span class="op">=</span> ind_with_metadata_indFiltered.ind[selection<span class="op">.==</span><span class="cn">false</span>]</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="fu">DataFrame</span>(filtered_inds <span class="op">=</span> filtered_inds)) <span class="co"># did this to print all lines</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered <span class="op">=</span> ind_with_metadata_indFiltered[selection, <span class="op">:</span>]</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>geno_indFiltered <span class="op">=</span> <span class="fu">view</span>(geno_indFiltered, selection, <span class="op">:</span>);</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>()</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Here are the remaining individuals: "</span>)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="fu">DataFrame</span>(ind_with_metadata_indFiltered))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering out these individuals based on too many missing genotypes: 
33×1 DataFrame
 Row │ filtered_inds                  
     │ String                         
─────┼────────────────────────────────
   1 │ GW_Armando_plate1_JG08G02
   2 │ GW_Armando_plate1_JG10G01
   3 │ GW_Armando_plate1_NO_BC_TTGW05
   4 │ GW_Armando_plate1_NO_DNA
   5 │ GW_Armando_plate1_TTGW21
   6 │ GW_Armando_plate1_TTGW71
   7 │ GW_Armando_plate2_NO_BC_TTGW05
   8 │ GW_Armando_plate2_NO_DNA
   9 │ GW_Armando_plate2_TTGW15
  10 │ GW_Lane5_AA10
  11 │ GW_Lane5_DA6
  12 │ GW_Lane5_LN11
  13 │ GW_Liz_GBS_Liz5101
  14 │ GW_Liz_GBS_Liz5118
  15 │ GW_Liz_GBS_Liz5139
  16 │ GW_Liz_GBS_Liz5142
  17 │ GW_Liz_GBS_Liz5150
  18 │ GW_Liz_GBS_Liz5159
  19 │ GW_Liz_GBS_Liz5162
  20 │ GW_Liz_GBS_Liz5169
  21 │ GW_Liz_GBS_Liz5171
  22 │ GW_Liz_GBS_Liz5172
  23 │ GW_Liz_GBS_Liz5174
  24 │ GW_Liz_GBS_Liz5176
  25 │ GW_Liz_GBS_Liz5177
  26 │ GW_Liz_GBS_Liz5180
  27 │ GW_Liz_GBS_Liz5186
  28 │ GW_Liz_GBS_Liz5187
  29 │ GW_Liz_GBS_Liz5192
  30 │ GW_Liz_GBS_Liz5195
  31 │ GW_Liz_GBS_Liz6012
  32 │ GW_Liz_GBS_Liz6203
  33 │ GW_Liz_GBS_Liz6766

Here are the remaining individuals: 
261×7 DataFrame
 Row │ ind                            ID                             location  group           Fst_group    plot_order  numMissings 
     │ String                         String31                       String7   String15        String15     Float64     Int64       
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1          GW_Armando_plate1_AB1          AB        vir             vir              20.01         99270
   2 │ GW_Armando_plate1_JF07G02      GW_Armando_plate1_JF07G02      ST        plumb           plumb            87.001       181208
   3 │ GW_Armando_plate1_JF07G03      GW_Armando_plate1_JF07G03      ST        plumb           plumb            87.002       116334
   4 │ GW_Armando_plate1_JF07G04      GW_Armando_plate1_JF07G04      ST        plumb           plumb            87.003       116065
   5 │ GW_Armando_plate1_JF08G02      GW_Armando_plate1_JF08G02      ST        plumb           plumb            87.004        87194
   6 │ GW_Armando_plate1_JF09G01      GW_Armando_plate1_JF09G01      ST        plumb           plumb            87.005       237912
   7 │ GW_Armando_plate1_JF09G02      GW_Armando_plate1_JF09G02      ST        plumb           plumb            87.006       110439
   8 │ GW_Armando_plate1_JF10G03      GW_Armando_plate1_JF10G03      ST        plumb_vir       plumb_vir        98.0         202891
   9 │ GW_Armando_plate1_JF11G01      GW_Armando_plate1_JF11G01      ST        plumb           plumb            87.008       112391
  10 │ GW_Armando_plate1_JF12G01      GW_Armando_plate1_JF12G01      ST        plumb           plumb            87.009       129490
  11 │ GW_Armando_plate1_JF12G02      GW_Armando_plate1_JF12G02      ST        plumb           plumb            87.01        130718
  12 │ GW_Armando_plate1_JF12G04      GW_Armando_plate1_JF12G04      ST_vi     vir             vir              24.001       133969
  13 │ GW_Armando_plate1_JF13G01      GW_Armando_plate1_JF13G01      ST        plumb           plumb            87.011       107508
  14 │ GW_Armando_plate1_JF15G03      GW_Armando_plate1_JF15G03      KK        plumb           plumb            87.012       110933
  15 │ GW_Armando_plate1_JF16G01      GW_Armando_plate1_JF16G01      KK_vi     plumb_vir       vir              24.041       122599
  16 │ GW_Armando_plate1_JF20G01      GW_Armando_plate1_JF20G01      KK        plumb           plumb            87.014       102740
  17 │ GW_Armando_plate1_JF22G01      GW_Armando_plate1_JF22G01      KK        plumb           plumb            87.015        99501
  18 │ GW_Armando_plate1_JF23G01      GW_Armando_plate1_JF23G01      KK        plumb           plumb            87.016       113052
  19 │ GW_Armando_plate1_JF23G02      GW_Armando_plate1_JF23G02      KK        plumb           plumb            87.017        87860
  20 │ GW_Armando_plate1_JF24G02      GW_Armando_plate1_JF24G02      KK        plumb           plumb            87.018        88105
  21 │ GW_Armando_plate1_JF26G01      GW_Armando_plate1_JF26G01      ST        plumb           plumb            87.019       104432
  22 │ GW_Armando_plate1_JF27G01      GW_Armando_plate1_JF27G01      ST        plumb           plumb            87.02         96001
  23 │ GW_Armando_plate1_JF29G01      GW_Armando_plate1_JF29G01      ST        plumb           plumb            87.021        86530
  24 │ GW_Armando_plate1_JF29G02      GW_Armando_plate1_JF29G02      ST        plumb           plumb            87.022       101762
  25 │ GW_Armando_plate1_JF29G03      GW_Armando_plate1_JF29G03      ST        plumb           plumb            87.023       107333
  26 │ GW_Armando_plate1_JG02G02      GW_Armando_plate1_JG02G02      KK        plumb           plumb            87.024        81977
  27 │ GW_Armando_plate1_JG02G04      GW_Armando_plate1_JG02G04      KK        plumb           plumb            87.025       140445
  28 │ GW_Armando_plate1_JG08G01      GW_Armando_plate1_JG08G01      ST        plumb           plumb            87.026       131448
  29 │ GW_Armando_plate1_JG12G01      GW_Armando_plate1_JG12G01      ST        plumb           plumb            87.029       132475
  30 │ GW_Armando_plate1_JG17G01      GW_Armando_plate1_JG17G01      ST        plumb_vir       plumb            77.92        106045
  31 │ GW_Armando_plate1_RF20G01      GW_Armando_plate1_RF20G01      BJ        obs_plumb       plumb_BJ         77.501       102115
  32 │ GW_Armando_plate1_RF29G02      GW_Armando_plate1_RF29G02      BJ        obs_plumb       plumb_BJ         77.502        99695
  33 │ GW_Armando_plate1_TL3          GW_Armando_plate1_TL3          TL        vir             vir              11.01        119043
  34 │ GW_Armando_plate1_TTGW01       GW_Armando_plate1_TTGW01       MN        troch_MN        troch_west       53.0          80329
  35 │ GW_Armando_plate1_TTGW05_rep2  GW_Armando_plate1_TTGW05_rep2  MN        troch_MN        troch_west       53.0          95518
  36 │ GW_Armando_plate1_TTGW06       GW_Armando_plate1_TTGW06       SU        lud_Sukhto      lud_central      47.0          80282
  37 │ GW_Armando_plate1_TTGW07       GW_Armando_plate1_TTGW07       SU        lud_Sukhto      lud_central      47.0         153267
  38 │ GW_Armando_plate1_TTGW10       GW_Armando_plate1_TTGW10       SU        lud_Sukhto      lud_central      47.0          97805
  39 │ GW_Armando_plate1_TTGW11       GW_Armando_plate1_TTGW11       SU        lud_Sukhto      lud_central      47.0          93712
  40 │ GW_Armando_plate1_TTGW13       GW_Armando_plate1_TTGW13       TH        lud_Thallighar  lud_central      43.0          87677
  41 │ GW_Armando_plate1_TTGW17       GW_Armando_plate1_TTGW17       TH        lud_Thallighar  lud_central      43.0         100310
  42 │ GW_Armando_plate1_TTGW19       GW_Armando_plate1_TTGW19       TH        lud_Thallighar  lud_central      43.0         100019
  43 │ GW_Armando_plate1_TTGW22       GW_Armando_plate1_TTGW22       SR        lud_Sural       lud_central      45.0          95151
  44 │ GW_Armando_plate1_TTGW23       GW_Armando_plate1_TTGW23       SR        lud_Sural       lud_central      45.0         218168
  45 │ GW_Armando_plate1_TTGW29       GW_Armando_plate1_TTGW29       SR        lud_Sural       lud_central      45.0         101068
  46 │ GW_Armando_plate1_TTGW52       GW_Armando_plate1_TTGW52       NG        lud_Nainaghar   lud_central      49.0          89055
  47 │ GW_Armando_plate1_TTGW53       GW_Armando_plate1_TTGW53       NG        lud_Nainaghar   lud_central      49.0          76154
  48 │ GW_Armando_plate1_TTGW55       GW_Armando_plate1_TTGW55       NG        lud_Nainaghar   lud_central      49.0          84162
  49 │ GW_Armando_plate1_TTGW57       GW_Armando_plate1_TTGW57       NG        lud_Nainaghar   lud_central      49.0          93413
  50 │ GW_Armando_plate1_TTGW58       GW_Armando_plate1_TTGW58       NG        lud_Nainaghar   lud_central      49.0         102386
  51 │ GW_Armando_plate1_TTGW59       GW_Armando_plate1_TTGW59       NG        lud_Nainaghar   lud_central      49.0          98471
  52 │ GW_Armando_plate1_TTGW63       GW_Armando_plate1_TTGW63       SP        lud_Spiti       troch_west       55.0          98425
  53 │ GW_Armando_plate1_TTGW64       GW_Armando_plate1_TTGW64       SP        lud_Spiti       troch_west       55.0         111234
  54 │ GW_Armando_plate1_TTGW65       GW_Armando_plate1_TTGW65       SP        lud_Spiti       troch_west       55.0         119729
  55 │ GW_Armando_plate1_TTGW66       GW_Armando_plate1_TTGW66       SP        lud_Spiti       troch_west       55.0         105443
  56 │ GW_Armando_plate1_TTGW68       GW_Armando_plate1_TTGW68       SP        lud_Spiti       troch_west       55.0          98099
  57 │ GW_Armando_plate1_TTGW70       GW_Armando_plate1_TTGW70       SA        lud_Sathrundi   lud_Sath         41.0          84930
  58 │ GW_Armando_plate1_TTGW72       GW_Armando_plate1_TTGW72       SA        lud_Sathrundi   lud_Sath         41.0          92370
  59 │ GW_Armando_plate1_TTGW74       GW_Armando_plate1_TTGW74       SA        lud_Sathrundi   lud_Sath         41.0         325539
  60 │ GW_Armando_plate1_TTGW78       GW_Armando_plate1_TTGW78       SA        lud_Sathrundi   lud_Sath         41.0          96540
  61 │ GW_Armando_plate1_TTGW_15_05   GW_Armando_plate1_TTGW_15_05   SR        lud_Sural       lud_central      45.0         173819
  62 │ GW_Armando_plate1_TTGW_15_07   GW_Armando_plate1_TTGW_15_07   SR        lud_Sural       lud_central      45.0          84861
  63 │ GW_Armando_plate1_TTGW_15_08   GW_Armando_plate1_TTGW_15_08   SR        lud_Sural       lud_central      45.0          88322
  64 │ GW_Armando_plate1_TTGW_15_09   GW_Armando_plate1_TTGW_15_09   SR        lud_Sural       lud_central      45.0          87883
  65 │ GW_Armando_plate1_UY1          GW_Armando_plate1_UY1          UY        plumb           plumb            88.01         97598
  66 │ GW_Armando_plate2_JE31G01      GW_Armando_plate2_JE31G01      KK_vi     vir_misID       vir              24.002       101749
  67 │ GW_Armando_plate2_JF03G01      GW_Armando_plate2_JF03G01      ST_vi     vir_misID       vir              24.003       214510
  68 │ GW_Armando_plate2_JF03G02      GW_Armando_plate2_JF03G02      KK_vi     vir_misID       vir              24.004       296518
  69 │ GW_Armando_plate2_JF07G01      GW_Armando_plate2_JF07G01      ST        plumb           plumb            87.031       104578
  70 │ GW_Armando_plate2_JF08G04      GW_Armando_plate2_JF08G04      ST        plumb           plumb            87.032       101010
  71 │ GW_Armando_plate2_JF10G02      GW_Armando_plate2_JF10G02      ST        plumb           plumb            87.033        95393
  72 │ GW_Armando_plate2_JF11G02      GW_Armando_plate2_JF11G02      ST        plumb           plumb            87.034        89277
  73 │ GW_Armando_plate2_JF12G03      GW_Armando_plate2_JF12G03      ST        plumb           plumb            87.035       118134
  74 │ GW_Armando_plate2_JF12G05      GW_Armando_plate2_JF12G05      ST        plumb           plumb            87.036        85768
  75 │ GW_Armando_plate2_JF13G02      GW_Armando_plate2_JF13G02      ST        plumb           plumb            87.037       113146
  76 │ GW_Armando_plate2_JF14G01      GW_Armando_plate2_JF14G01      KK        plumb           plumb            87.038        97185
  77 │ GW_Armando_plate2_JF14G02      GW_Armando_plate2_JF14G02      KK        plumb           plumb            87.039        94078
  78 │ GW_Armando_plate2_JF15G01      GW_Armando_plate2_JF15G01      KK        plumb           plumb            87.04         89126
  79 │ GW_Armando_plate2_JF15G02      GW_Armando_plate2_JF15G02      KK        plumb           plumb            87.041       101787
  80 │ GW_Armando_plate2_JF16G02      GW_Armando_plate2_JF16G02      KK_vi     plumb_vir       vir              24.042       125147
  81 │ GW_Armando_plate2_JF19G01      GW_Armando_plate2_JF19G01      KK        plumb           plumb            87.043        94354
  82 │ GW_Armando_plate2_JF20G02      GW_Armando_plate2_JF20G02      KK        plumb           plumb            87.044        78032
  83 │ GW_Armando_plate2_JF24G01      GW_Armando_plate2_JF24G01      KK        plumb           plumb            87.045       113526
  84 │ GW_Armando_plate2_JF24G03      GW_Armando_plate2_JF24G03      ST        plumb           plumb            87.046        77983
  85 │ GW_Armando_plate2_JF25G01      GW_Armando_plate2_JF25G01      KK        plumb           plumb            87.047        86955
  86 │ GW_Armando_plate2_JF26G02      GW_Armando_plate2_JF26G02      KK        plumb           plumb            87.048        79995
  87 │ GW_Armando_plate2_JF27G02      GW_Armando_plate2_JF27G02      KK        plumb           plumb            87.049        97159
  88 │ GW_Armando_plate2_JF30G01      GW_Armando_plate2_JF30G01      ST_vi     vir_misID       vir              24.005       117010
  89 │ GW_Armando_plate2_JG01G01      GW_Armando_plate2_JG01G01      KK        plumb           plumb            87.05         97961
  90 │ GW_Armando_plate2_JG02G01      GW_Armando_plate2_JG02G01      KK        plumb           plumb            87.051        81494
  91 │ GW_Armando_plate2_JG02G03      GW_Armando_plate2_JG02G03      KK        plumb           plumb            87.052        96990
  92 │ GW_Armando_plate2_JG10G02      GW_Armando_plate2_JG10G02      ST        plumb           plumb            87.053        78807
  93 │ GW_Armando_plate2_JG10G03      GW_Armando_plate2_JG10G03      ST        plumb           plumb            87.054        91258
  94 │ GW_Armando_plate2_JG12G02      GW_Armando_plate2_JG12G02      ST        plumb           plumb            87.055       102563
  95 │ GW_Armando_plate2_JG12G03      GW_Armando_plate2_JG12G03      ST        plumb           plumb            87.056        91523
  96 │ GW_Armando_plate2_LN2          GW_Armando_plate2_LN2          LN        troch_LN        troch_LN         58.01         77555
  97 │ GW_Armando_plate2_RF29G01      GW_Armando_plate2_RF29G01      BJ        obs_plumb       plumb_BJ         77.503        94504
  98 │ GW_Armando_plate2_TTGW02       GW_Armando_plate2_TTGW02       MN        troch_MN        troch_west       53.0          82355
  99 │ GW_Armando_plate2_TTGW03       GW_Armando_plate2_TTGW03       MN        troch_MN        troch_west       53.0          93915
 100 │ GW_Armando_plate2_TTGW08       GW_Armando_plate2_TTGW08       SU        lud_Sukhto      lud_central      47.0          79030
 101 │ GW_Armando_plate2_TTGW09       GW_Armando_plate2_TTGW09       SU        lud_Sukhto      lud_central      47.0         124074
 102 │ GW_Armando_plate2_TTGW12       GW_Armando_plate2_TTGW12       TH        lud_Thallighar  lud_central      43.0          80774
 103 │ GW_Armando_plate2_TTGW14       GW_Armando_plate2_TTGW14       TH        lud_Thallighar  lud_central      43.0          91726
 104 │ GW_Armando_plate2_TTGW16       GW_Armando_plate2_TTGW16       TH        lud_Thallighar  lud_central      43.0          78610
 105 │ GW_Armando_plate2_TTGW18       GW_Armando_plate2_TTGW18       TH        lud_Thallighar  lud_central      43.0          90831
 106 │ GW_Armando_plate2_TTGW20       GW_Armando_plate2_TTGW20       SR        lud_Sural       lud_central      45.0          70473
 107 │ GW_Armando_plate2_TTGW24       GW_Armando_plate2_TTGW24       SR        lud_Sural       lud_central      45.0          88081
 108 │ GW_Armando_plate2_TTGW25       GW_Armando_plate2_TTGW25       SR        lud_Sural       lud_central      45.0         113545
 109 │ GW_Armando_plate2_TTGW27       GW_Armando_plate2_TTGW27       SR        lud_Sural       lud_central      45.0          92690
 110 │ GW_Armando_plate2_TTGW28       GW_Armando_plate2_TTGW28       SR        lud_Sural       lud_central      45.0          90248
 111 │ GW_Armando_plate2_TTGW50       GW_Armando_plate2_TTGW50       NG        lud_Nainaghar   lud_central      49.0          73271
 112 │ GW_Armando_plate2_TTGW51       GW_Armando_plate2_TTGW51       NG        lud_Nainaghar   lud_central      49.0          82100
 113 │ GW_Armando_plate2_TTGW54       GW_Armando_plate2_TTGW54       NG        lud_Nainaghar   lud_central      49.0         680557
 114 │ GW_Armando_plate2_TTGW56       GW_Armando_plate2_TTGW56       NG        lud_Nainaghar   lud_central      49.0          77725
 115 │ GW_Armando_plate2_TTGW60       GW_Armando_plate2_TTGW60       SP        lud_Spiti       troch_west       55.0          80730
 116 │ GW_Armando_plate2_TTGW61       GW_Armando_plate2_TTGW61       SP        lud_Spiti       troch_west       55.0          96107
 117 │ GW_Armando_plate2_TTGW62       GW_Armando_plate2_TTGW62       SP        lud_Spiti       troch_west       55.0         101894
 118 │ GW_Armando_plate2_TTGW67       GW_Armando_plate2_TTGW67       SP        lud_Spiti       troch_west       55.0         145905
 119 │ GW_Armando_plate2_TTGW69       GW_Armando_plate2_TTGW69       SP        lud_Spiti       troch_west       55.0          91534
 120 │ GW_Armando_plate2_TTGW73       GW_Armando_plate2_TTGW73       SA        lud_Sathrundi   lud_Sath         41.0          72217
 121 │ GW_Armando_plate2_TTGW75       GW_Armando_plate2_TTGW75       SA        lud_Sathrundi   lud_Sath         41.0          92128
 122 │ GW_Armando_plate2_TTGW77       GW_Armando_plate2_TTGW77       SA        lud_Sathrundi   lud_Sath         41.0          75475
 123 │ GW_Armando_plate2_TTGW79       GW_Armando_plate2_TTGW79       SA        lud_Sathrundi   lud_Sath         41.0          96324
 124 │ GW_Armando_plate2_TTGW80       GW_Armando_plate2_TTGW80       SA        lud_Sathrundi   lud_Sath         41.0          75357
 125 │ GW_Armando_plate2_TTGW_15_01   GW_Armando_plate2_TTGW_15_01   SR        lud_Sural       lud_central      45.0          91072
 126 │ GW_Armando_plate2_TTGW_15_02   GW_Armando_plate2_TTGW_15_02   SR        lud_Sural       lud_central      45.0          72691
 127 │ GW_Armando_plate2_TTGW_15_03   GW_Armando_plate2_TTGW_15_03   SR        lud_Sural       lud_central      45.0          74750
 128 │ GW_Armando_plate2_TTGW_15_04   GW_Armando_plate2_TTGW_15_04   SR        lud_Sural       lud_central      45.0         193531
 129 │ GW_Armando_plate2_TTGW_15_06   GW_Armando_plate2_TTGW_15_06   SR        lud_Sural       lud_central      45.0          76538
 130 │ GW_Armando_plate2_TTGW_15_10   GW_Armando_plate2_TTGW_15_10   SR        lud_Sural       lud_central      45.0         219999
 131 │ GW_Lane5_AA1                   GW_Lane5_AA1                   AA        vir_S           vir_S            25.0         717676
 132 │ GW_Lane5_AA11                  GW_Lane5_AA11                  AA        vir_S           vir_S            34.0         736085
 133 │ GW_Lane5_AA3                   GW_Lane5_AA3                   AA        vir_S           vir_S            26.0         662552
 134 │ GW_Lane5_AA4                   GW_Lane5_AA4                   AA        vir_S           vir_S            27.0         665969
 135 │ GW_Lane5_AA5                   GW_Lane5_AA5                   AA        vir_S           vir_S            28.0         718610
 136 │ GW_Lane5_AA6                   GW_Lane5_AA6                   AA        vir_S           vir_S            29.0         778821
 137 │ GW_Lane5_AA7                   GW_Lane5_AA7                   AA        vir_S           vir_S            30.0         724717
 138 │ GW_Lane5_AA8                   GW_Lane5_AA8                   AA        vir_S           vir_S            31.0         922313
 139 │ GW_Lane5_AA9                   GW_Lane5_AA9                   AA        vir_S           vir_S            32.0         791514
 140 │ GW_Lane5_AB2                   GW_Lane5_AB2                   AB        vir             vir              21.0         915622
 141 │ GW_Lane5_AN1                   GW_Lane5_AN1                   AN        plumb           plumb            80.0         719041
 142 │ GW_Lane5_AN2                   GW_Lane5_AN2                   AN        plumb           plumb            81.0         684348
 143 │ GW_Lane5_BK2                   GW_Lane5_BK2                   BK        plumb           plumb            78.0         666018
 144 │ GW_Lane5_BK3                   GW_Lane5_BK3                   BK        plumb           plumb            79.0         665090
 145 │ GW_Lane5_DA2                   GW_Lane5_DA2                   XN        obs             obs              73.0         655980
 146 │ GW_Lane5_DA3                   GW_Lane5_DA3                   XN        obs             obs              74.0         696722
 147 │ GW_Lane5_DA4                   GW_Lane5_DA4                   XN        obs             obs              75.0         634940
 148 │ GW_Lane5_DA7                   GW_Lane5_DA7                   XN        obs             obs              77.0         733664
 149 │ GW_Lane5_EM1                   GW_Lane5_EM1                   EM        troch_EM        troch_EM         72.0         711636
 150 │ GW_Lane5_IL1                   GW_Lane5_IL1                   IL        plumb           plumb            82.0         698039
 151 │ GW_Lane5_IL4                   GW_Lane5_IL4                   IL        plumb           plumb            84.0         726050
 152 │ GW_Lane5_KS1                   GW_Lane5_KS1                   OV        lud_KS          lud_KS           40.0         751642
 153 │ GW_Lane5_KS2                   GW_Lane5_KS2                   OV        lud_KS          lud_KS           40.0         820241
 154 │ GW_Lane5_LN1                   GW_Lane5_LN1                   LN        troch_LN        troch_LN         57.0         707607
 155 │ GW_Lane5_LN10                  GW_Lane5_LN10                  LN        troch_LN        troch_LN         64.0         783282
 156 │ GW_Lane5_LN12                  GW_Lane5_LN12                  LN        troch_LN        troch_LN         66.0         884653
 157 │ GW_Lane5_LN14                  GW_Lane5_LN14                  LN        troch_LN        troch_LN         67.0         753482
 158 │ GW_Lane5_LN16                  GW_Lane5_LN16                  LN        troch_LN        troch_LN         68.0         738272
 159 │ GW_Lane5_LN18                  GW_Lane5_LN18                  LN        troch_LN        troch_LN         69.0         711612
 160 │ GW_Lane5_LN19                  GW_Lane5_LN19                  LN        troch_LN        troch_LN         70.0         731751
 161 │ GW_Lane5_LN20                  GW_Lane5_LN20                  LN        troch_LN        troch_LN         71.0         738170
 162 │ GW_Lane5_LN3                   GW_Lane5_LN3                   LN        troch_LN        troch_LN         59.0         702316
 163 │ GW_Lane5_LN4                   GW_Lane5_LN4                   LN        troch_LN        troch_LN         60.0         683211
 164 │ GW_Lane5_LN6                   GW_Lane5_LN6                   LN        troch_LN        troch_LN         61.0         690808
 165 │ GW_Lane5_LN7                   GW_Lane5_LN7                   LN        troch_LN        troch_LN         62.0         758021
 166 │ GW_Lane5_LN8                   GW_Lane5_LN8                   LN        troch_LN        troch_LN         63.0         662407
 167 │ GW_Lane5_MN1                   GW_Lane5_MN1                   MN        troch_MN        troch_west       51.0         943574
 168 │ GW_Lane5_MN12                  GW_Lane5_MN12                  MN        troch_MN        troch_west       56.0         671074
 169 │ GW_Lane5_MN3                   GW_Lane5_MN3                   MN        troch_MN        troch_west       52.0         946478
 170 │ GW_Lane5_MN5                   GW_Lane5_MN5                   MN        troch_MN        troch_west       53.0         752979
 171 │ GW_Lane5_MN8                   GW_Lane5_MN8                   MN        troch_MN        troch_west       54.0         771625
 172 │ GW_Lane5_MN9                   GW_Lane5_MN9                   MN        troch_MN        troch_west       55.0         897944
 173 │ GW_Lane5_NA1                   GW_Lane5_NA1                   NR        lud_PK          lud_PK           39.2         919218
 174 │ GW_Lane5_NA3-3ul               GW_Lane5_NA3-3ul               NR        lud_PK          lud_PK           39.2         798370
 175 │ GW_Lane5_PT11                  GW_Lane5_PT11                  KL        lud_KL          lud_central      42.0         771797
 176 │ GW_Lane5_PT12                  GW_Lane5_PT12                  KL        lud_KL          lud_central      42.0         792683
 177 │ GW_Lane5_PT2                   GW_Lane5_PT2                   ML        lud_ML          lud_ML           51.0         760347
 178 │ GW_Lane5_PT3                   GW_Lane5_PT3                   PA        lud_PA          lud_central      46.0         722177
 179 │ GW_Lane5_PT4                   GW_Lane5_PT4                   PA        lud_PA          lud_central      46.0         705413
 180 │ GW_Lane5_PT6                   GW_Lane5_PT6                   KL        lud_KL          lud_central      42.0         763723
 181 │ GW_Lane5_SH1                   GW_Lane5_SH1                   PK        lud_PK          lud_PK           39.1         966761
 182 │ GW_Lane5_SH2                   GW_Lane5_SH2                   PK        lud_PK          lud_PK           39.1         768646
 183 │ GW_Lane5_SH4                   GW_Lane5_SH4                   PK        lud_PK          lud_PK           39.1         849641
 184 │ GW_Lane5_SH5                   GW_Lane5_SH5                   PK        lud_PK          lud_PK           39.1         929394
 185 │ GW_Lane5_SL1                   GW_Lane5_SL1                   SL        plumb           plumb            95.0         648883
 186 │ GW_Lane5_SL2                   GW_Lane5_SL2                   SL        plumb           plumb            96.0         654733
 187 │ GW_Lane5_ST1                   GW_Lane5_ST1                   ST        plumb           plumb            85.0         606242
 188 │ GW_Lane5_ST12                  GW_Lane5_ST12                  ST        plumb           plumb            87.0         691266
 189 │ GW_Lane5_ST3                   GW_Lane5_ST3                   ST        plumb           plumb            86.0         696992
 190 │ GW_Lane5_STvi1                 GW_Lane5_STvi1                 STvi      vir             vir              22.0         690090
 191 │ GW_Lane5_STvi2                 GW_Lane5_STvi2                 STvi      vir             vir              23.0         897337
 192 │ GW_Lane5_STvi3                 GW_Lane5_STvi3                 STvi      vir             vir              24.0         768162
 193 │ GW_Lane5_TA1                   GW_Lane5_TA1                   TA        plumb           plumb            94.0         711909
 194 │ GW_Lane5_TL1                   GW_Lane5_TL1                   TL        vir             vir               9.0         743509
 195 │ GW_Lane5_TL10                  GW_Lane5_TL10                  TL        vir             vir              17.0         669934
 196 │ GW_Lane5_TL11                  GW_Lane5_TL11                  TL        vir             vir              18.0         638402
 197 │ GW_Lane5_TL12                  GW_Lane5_TL12                  TL        vir             vir              19.0         585697
 198 │ GW_Lane5_TL2                   GW_Lane5_TL2                   TL        vir             vir              10.0         770857
 199 │ GW_Lane5_TL4                   GW_Lane5_TL4                   TL        vir             vir              12.0         758037
 200 │ GW_Lane5_TL5                   GW_Lane5_TL5                   TL        vir             vir              13.0         867165
 201 │ GW_Lane5_TL7                   GW_Lane5_TL7                   TL        vir             vir              14.0         803407
 202 │ GW_Lane5_TL8                   GW_Lane5_TL8                   TL        vir             vir              15.0         698745
 203 │ GW_Lane5_TL9                   GW_Lane5_TL9                   TL        vir             vir              16.0         606969
 204 │ GW_Lane5_TU1                   GW_Lane5_TU1                   TU        nit             nit              35.0         793640
 205 │ GW_Lane5_TU2                   GW_Lane5_TU2                   TU        nit             nit              36.0         736785
 206 │ GW_Lane5_UY2                   GW_Lane5_UY2                   UY        plumb           plumb            89.0         729002
 207 │ GW_Lane5_UY3                   GW_Lane5_UY3                   UY        plumb           plumb            90.0         677527
 208 │ GW_Lane5_UY4                   GW_Lane5_UY4                   UY        plumb           plumb            91.0         749848
 209 │ GW_Lane5_UY5                   GW_Lane5_UY5                   UY        plumb           plumb            92.0         718373
 210 │ GW_Lane5_UY6                   GW_Lane5_UY6                   UY        plumb           plumb            93.0         713675
 211 │ GW_Lane5_YK1                   GW_Lane5_YK1                   YK        vir             vir               1.0         831245
 212 │ GW_Lane5_YK11                  GW_Lane5_YK11                  YK        vir             vir               8.0         730798
 213 │ GW_Lane5_YK3                   GW_Lane5_YK3                   YK        vir             vir               2.0         731944
 214 │ GW_Lane5_YK4                   GW_Lane5_YK4                   YK        vir             vir               3.0         740051
 215 │ GW_Lane5_YK5                   GW_Lane5_YK5                   YK        vir             vir               4.0         738740
 216 │ GW_Lane5_YK6                   GW_Lane5_YK6                   YK        vir             vir               5.0         697420
 217 │ GW_Lane5_YK7                   GW_Lane5_YK7                   YK        vir             vir               6.0         692052
 218 │ GW_Lane5_YK9                   GW_Lane5_YK9                   YK        vir             vir               7.0         768722
 219 │ GW_Liz_GBS_Liz10045            GW_Liz_GBS_Liz10045            ML        lud             lud_ML           51.01        818407
 220 │ GW_Liz_GBS_Liz10094            GW_Liz_GBS_Liz10094            ML        lud             lud_ML           51.02        905297
 221 │ GW_Liz_GBS_Liz5144             GW_Liz_GBS_Liz5144             ML        lud             lud_ML           51.08        942026
 222 │ GW_Liz_GBS_Liz5163             GW_Liz_GBS_Liz5163             ML        lud_chick       lud_ML           51.12        851723
 223 │ GW_Liz_GBS_Liz5164             GW_Liz_GBS_Liz5164             ML        lud_chick       lud_ML           51.13        866612
 224 │ GW_Liz_GBS_Liz5165             GW_Liz_GBS_Liz5165             ML        lud             lud_ML           51.14        939782
 225 │ GW_Liz_GBS_Liz5167             GW_Liz_GBS_Liz5167             ML        lud_chick       lud_ML           51.15        932442
 226 │ GW_Liz_GBS_Liz5168             GW_Liz_GBS_Liz5168             ML        lud_chick       lud_ML           51.16        944373
 227 │ GW_Liz_GBS_Liz5173             GW_Liz_GBS_Liz5173             ML        lud_chick       lud_ML           51.2         961268
 228 │ GW_Liz_GBS_Liz5175             GW_Liz_GBS_Liz5175             ML        lud             lud_ML           51.22        947557
 229 │ GW_Liz_GBS_Liz5178             GW_Liz_GBS_Liz5178             ML        lud_chick       lud_ML           51.25        942983
 230 │ GW_Liz_GBS_Liz5179             GW_Liz_GBS_Liz5179             ML        lud_chick       lud_ML           51.26        933796
 231 │ GW_Liz_GBS_Liz5182             GW_Liz_GBS_Liz5182             ML        lud_chick       lud_ML           51.28        950069
 232 │ GW_Liz_GBS_Liz5184             GW_Liz_GBS_Liz5184             ML        lud_chick       lud_ML           51.29        875068
 233 │ GW_Liz_GBS_Liz5185             GW_Liz_GBS_Liz5185             ML        lud             lud_ML           51.3         948876
 234 │ GW_Liz_GBS_Liz5188             GW_Liz_GBS_Liz5188             ML        lud             lud_ML           51.33        853346
 235 │ GW_Liz_GBS_Liz5189             GW_Liz_GBS_Liz5189             ML        lud_chick       lud_ML           51.34        972180
 236 │ GW_Liz_GBS_Liz5190             GW_Liz_GBS_Liz5190             ML        lud_chick       lud_ML           51.35        922566
 237 │ GW_Liz_GBS_Liz5191             GW_Liz_GBS_Liz5191             ML        lud_chick       lud_ML           51.36        836413
 238 │ GW_Liz_GBS_Liz5193             GW_Liz_GBS_Liz5193             ML        lud_chick       lud_ML           51.38        895586
 239 │ GW_Liz_GBS_Liz5194             GW_Liz_GBS_Liz5194             ML        lud_chick       lud_ML           51.39        900099
 240 │ GW_Liz_GBS_Liz5197             GW_Liz_GBS_Liz5197             ML        lud             lud_ML           51.41        808093
 241 │ GW_Liz_GBS_Liz5199             GW_Liz_GBS_Liz5199             ML        lud_chick       lud_ML           51.42        953161
 242 │ GW_Liz_GBS_Liz6002             GW_Liz_GBS_Liz6002             ML        lud             lud_ML           51.43        903717
 243 │ GW_Liz_GBS_Liz6006             GW_Liz_GBS_Liz6006             ML        lud             lud_ML           51.44        872123
 244 │ GW_Liz_GBS_Liz6008             GW_Liz_GBS_Liz6008             ML        lud             lud_ML           51.45        932032
 245 │ GW_Liz_GBS_Liz6009             GW_Liz_GBS_Liz6009             ML        lud             lud_ML           51.46        946736
 246 │ GW_Liz_GBS_Liz6010             GW_Liz_GBS_Liz6010             ML        lud             lud_ML           51.47        844804
 247 │ GW_Liz_GBS_Liz6014             GW_Liz_GBS_Liz6014             ML        lud             lud_ML           51.49        828632
 248 │ GW_Liz_GBS_Liz6055             GW_Liz_GBS_Liz6055             ML        lud             lud_ML           51.5         878978
 249 │ GW_Liz_GBS_Liz6057             GW_Liz_GBS_Liz6057             ML        lud             lud_ML           51.51        850572
 250 │ GW_Liz_GBS_Liz6060             GW_Liz_GBS_Liz6060             ML        lud             lud_ML           51.52        865228
 251 │ GW_Liz_GBS_Liz6062             GW_Liz_GBS_Liz6062             ML        lud             lud_ML           51.53        834482
 252 │ GW_Liz_GBS_Liz6063             GW_Liz_GBS_Liz6063             ML        lud             lud_ML           51.54        879233
 253 │ GW_Liz_GBS_Liz6066             GW_Liz_GBS_Liz6066             ML        lud             lud_ML           51.55        876902
 254 │ GW_Liz_GBS_Liz6072             GW_Liz_GBS_Liz6072             ML        lud             lud_ML           51.56        845266
 255 │ GW_Liz_GBS_Liz6079             GW_Liz_GBS_Liz6079             ML        lud             lud_ML           51.57        820880
 256 │ GW_Liz_GBS_Liz6204             GW_Liz_GBS_Liz6204             ML        lud_chick       lud_ML           51.59        840138
 257 │ GW_Liz_GBS_Liz6461             GW_Liz_GBS_Liz6461             ML        lud             lud_ML           51.6         815684
 258 │ GW_Liz_GBS_Liz6472             GW_Liz_GBS_Liz6472             ML        lud             lud_ML           51.61        907391
 259 │ GW_Liz_GBS_Liz6478             GW_Liz_GBS_Liz6478             ML        lud             lud_ML           51.62        779125
 260 │ GW_Liz_GBS_Liz6776             GW_Liz_GBS_Liz6776             ML        lud             lud_ML           51.64        962187
 261 │ GW_Liz_GBS_Liz6794             GW_Liz_GBS_Liz6794             ML        lud             lud_ML           51.65        841253</code></pre>
</div>
</div>
</section>
<section id="filter-snps-with-too-many-missing-genotypes" class="level4">
<h4 class="anchored" data-anchor-id="filter-snps-with-too-many-missing-genotypes">Filter SNPs with too many missing genotypes:</h4>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (remember that first column is arbitrary row number in input file)</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>missing_genotypes_per_SNP <span class="op">=</span> <span class="fu">sum</span>(geno_indFiltered <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>missing_genotypes_percent_allowed_per_site <span class="op">=</span> <span class="fl">5</span>   <span class="co"># this is the percentage threshold</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>threshold_genotypes_missing <span class="op">=</span> <span class="fu">size</span>(geno_indFiltered)[<span class="fl">1</span>] <span class="op">*</span> missing_genotypes_percent_allowed_per_site<span class="op">/</span><span class="fl">100</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">vec</span>(missing_genotypes_per_SNP <span class="op">.&lt;=</span> threshold_genotypes_missing)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>geno_ind_SNP_filtered <span class="op">=</span> geno_indFiltered[<span class="op">:</span>, selection] </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered <span class="op">=</span> pos_whole_genome[selection[<span class="fu">Not</span>(<span class="fl">1</span>)],<span class="op">:</span>]  <span class="co"># the Not(1) is needed because first column in geno is arbitrary row number</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Started with "</span>, <span class="fu">size</span>(geno_indFiltered, <span class="fl">2</span>)<span class="op">-</span><span class="fl">1</span>, <span class="st">" SNPs.</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>After filtering SNPs <span class="cf">for</span> no more than <span class="st">", missing_genotypes_percent_allowed_per_site, "</span><span class="op">%</span> <span class="cn">missing</span> genotypes, <span class="st">", size(geno_ind_SNP_filtered, 2)-1, "</span> SNPs remain.<span class="st">" )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Started with 2431709 SNPs.
After filtering SNPs for no more than 5% missing genotypes, 1017581 SNPs remain.</code></pre>
</div>
</div>
</section>
<section id="nd-round-of-filtering-individuals" class="level4">
<h4 class="anchored" data-anchor-id="nd-round-of-filtering-individuals">2nd round of filtering individuals</h4>
<p>I added this in August 2023, to improve accuracy of imputation-based PCA, because I noticed outliers tended to have more missing data. Now I only allow up to 10% missing SNPs per individual.</p>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>SNPmissing_percent_allowed_per_ind_round2 <span class="op">=</span> <span class="fl">10</span>   <span class="co"># this is the percentage threshold</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>threshold_missing <span class="op">=</span> (<span class="fu">size</span>(geno_ind_SNP_filtered, <span class="fl">2</span>) <span class="op">-</span> <span class="fl">1</span>) <span class="op">*</span> SNPmissing_percent_allowed_per_ind_round2<span class="op">/</span><span class="fl">100</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>numMissings <span class="op">=</span> <span class="fu">sum</span>(geno_ind_SNP_filtered <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">vec</span>(numMissings <span class="op">.&lt;=</span> threshold_missing) <span class="co"># the vec command converts to BitVector rather than BitMatrix--important below</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>geno_ind_SNP_ind_filtered <span class="op">=</span> geno_ind_SNP_filtered[selection, <span class="op">:</span>]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Filtering out these individuals based on too many missing genotypes: "</span>)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>filtered_inds <span class="op">=</span> ind_with_metadata_indFiltered.ind[selection<span class="op">.==</span><span class="cn">false</span>]</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="fu">DataFrame</span>(filtered_inds <span class="op">=</span> filtered_inds)) <span class="co"># did this to print all lines</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered <span class="op">=</span> ind_with_metadata_indFiltered[selection, <span class="op">:</span>]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"This leaves "</span>, <span class="fu">size</span>(geno_ind_SNP_ind_filtered, <span class="fl">1</span>), <span class="st">" individuals and "</span>, <span class="fu">size</span>(geno_ind_SNP_ind_filtered, <span class="fl">2</span>)<span class="op">-</span><span class="fl">1</span>, <span class="st">" loci, </span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>with no individuals <span class="cn">missing</span> more than <span class="st">", SNPmissing_percent_allowed_per_ind_round2, "</span><span class="op">%</span> of genotypes</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>and no loci <span class="cn">missing</span> <span class="kw">in</span> more than <span class="st">", missing_genotypes_percent_allowed_per_site, "</span><span class="op">%</span> of individuals.<span class="st">")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering out these individuals based on too many missing genotypes: 
4×1 DataFrame
 Row │ filtered_inds            
     │ String                   
─────┼──────────────────────────
   1 │ GW_Armando_plate1_TTGW74
   2 │ GW_Armando_plate2_TTGW54
   3 │ GW_Lane5_AA8
   4 │ GW_Lane5_YK1
This leaves 257 individuals and 1017581 loci, 
with no individuals missing more than 10% of genotypes
and no loci missing in more than 5% of individuals.</code></pre>
</div>
</div>
</section>
</section>
<section id="estimate-relationships-of-individuals-using-pca" class="level2">
<h2 class="anchored" data-anchor-id="estimate-relationships-of-individuals-using-pca">Estimate relationships of individuals using PCA</h2>
<p>Our goal is to produce plots showing individuals in genotype space, using Principal Components Analysis. First we need to do a couple changes to our data matrix:</p>
<p>Remove the first column of the genotype matrix (which was an initial row number):</p>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>genosOnly <span class="op">=</span> geno_ind_SNP_ind_filtered[<span class="op">:</span>, <span class="fu">Not</span>(<span class="fl">1</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>257×1017581 Matrix{Int16}:
 0  0  0  0  0  1  0   0  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   1  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  1
 0  0  0  0  0  0  0   0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 ⋮              ⋮               ⋮     ⋱     ⋮              ⋮              ⋮
 0  0  0  0  0  0  0   0  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  1  0     1  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  -1  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   0  0  0  0  0     0  0  0  0  0  0  0  0  0  0  0  0</code></pre>
</div>
</div>
<p>For missing genotypes, change our code of <code>-1</code> to <code>missing</code>:</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Int16}}</span>(genosOnly)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing[genosOnly_with_missing <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>] <span class="op">.=</span> <span class="cn">missing</span>;</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="impute-and-save-genotypes-for-each-scaffold" class="level4">
<h4 class="anchored" data-anchor-id="impute-and-save-genotypes-for-each-scaffold">Impute and save genotypes for each scaffold</h4>
<p>PCA requires imputation of missing genotypes. I did imputation for each scaffold above a certain size threshold. Those scaffolds (many of which correspond to whole chromosomes) are listed here:</p>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>chromosomes_to_process <span class="op">=</span> <span class="fu">vec</span>([<span class="st">"gw2"</span>,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw1"</span>,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw3"</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gwZ"</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw1A"</span>,</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw4"</span>,</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw5"</span>,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw7"</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw6"</span>,</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw8"</span>,</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw9"</span>,</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw11"</span>,</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw12"</span>,</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw10"</span>,</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw13"</span>,</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw14"</span>,</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw18"</span>,</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw20"</span>,</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw15"</span>,</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw1B"</span>,</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws100"</span>,</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw17"</span>,</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw19"</span>,</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws101"</span>,</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw4A"</span>,</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw21"</span>,</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw26"</span>,</span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws102"</span>,</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw23"</span>,</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw25"</span>,</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws103"</span>,</span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw22"</span>,</span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws104"</span>,</span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw28"</span>,</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw27"</span>,</span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gw24"</span>,</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws105"</span>,</span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws106"</span>,</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws107"</span>,</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws108"</span>,</span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws109"</span>,</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws110"</span>,</span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gws112"</span>]);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Imputation can take several minutes per scaffold, so I ran this imputation step separately from this Quarto notebook (otherwise render would take long) and saved the genotype data for each scaffold for loading in the next step. This is the code I used for imputing:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_process)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_process[i]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chrom)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> pos_SNP_filtered[loci_selection,<span class="op">:</span>]</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_region_for_imputing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genosOnly_with_missing[<span class="op">:</span>,loci_selection])</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@time</span> imputed_genos <span class="op">=</span> Impute.<span class="fu">svd</span>(genosOnly_region_for_imputing)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".imputedMissing.jld2"</span>)</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">jldsave</span>(filename; imputed_genos, ind_with_metadata_indFiltered, pos_SNP_filtered_region)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Chromosome "</span>, chrom, <span class="st">": Saved real and imputed genotypes for "</span>, <span class="fu">size</span>(pos_SNP_filtered_region, <span class="fl">1</span>),<span class="st">" SNPs and "</span>, <span class="fu">size</span>(genosOnly_region_for_imputing, <span class="fl">1</span>),<span class="st">" filtered individuals."</span>))</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now we can cycle through a set of chromosomes and plot a PCA for each. We need to first specify some groups to include in the plot, and their colors:</p>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_misID"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>group_colors_PCA <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>];</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we’ll actually do the PCA and make the plot for each scaffold. The code block below is what I initially used, based on the SVD method of imputing. I’ve now decided that KNN is better (see further below).</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_process)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_process[i]</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".imputedMissing.jld2"</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    imputed_genos <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>, filename))</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genos, <span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genos, <span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>    flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>    flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>    PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(imputed_genos, ind_with_metadata_indFiltered,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>        groups_to_plot_PCA, group_colors_PCA;</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>        sampleSet<span class="op">=</span><span class="st">"greenish warblers"</span>, regionText<span class="op">=</span>regionText,</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>        flip1<span class="op">=</span>flipPC1, flip2<span class="op">=</span>flipPC2,</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>        showPlot<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add position of reference genome</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>    refGenomePCAposition <span class="op">=</span> <span class="fu">predict</span>(PCAmodel.model, <span class="fu">zeros</span>(<span class="fu">size</span>(imputed_genos, <span class="fl">2</span>)))</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>    flipPC1 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">1</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># this flips PC1 if flipPC1 = true</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>    flipPC2 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">2</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># same for PC2</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(refGenomePCAposition[<span class="fl">1</span>], refGenomePCAposition[<span class="fl">2</span>], marker<span class="op">=:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">15</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">catch</span></span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"NOTICE: Figure for "</span>, regionText, <span class="st">" could not be shown due to an unknown error."</span>)</span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="imputation-using-knn" class="level4">
<h4 class="anchored" data-anchor-id="imputation-using-knn">Imputation using KNN</h4>
<p>Troyanskaya <em>et al.</em> (2001) recommend imputation using K-nearest neighbors approach as being better than SVD, both of which are better than other methods for DNA genotyping. They also recommend using Euclidian distance. So I will try KNN with Euclidian distance, which like SVD is provided by Impute.jl. I am going with the default of setting <code>dims</code> to <code>:rows</code>, as that seems to run much faster and produces PCAs that make a lot of sense. I’ve already run this next code cell, which does the imputing and saves the imputed data matrix for each scaffold:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_process)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_process[i]</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chrom)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> pos_SNP_filtered[loci_selection,<span class="op">:</span>]</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_region_for_imputing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genosOnly_with_missing[<span class="op">:</span>,loci_selection])</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@time</span> imputed_genos <span class="op">=</span> Impute.<span class="fu">knn</span>(genosOnly_region_for_imputing; k<span class="op">=</span><span class="fl">1</span>, dims<span class="op">=:</span>rows)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">jldsave</span>(filename; imputed_genos, ind_with_metadata_indFiltered, pos_SNP_filtered_region)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Chromosome "</span>, chrom, <span class="st">": Saved real and imputed genotypes for "</span>, <span class="fu">size</span>(pos_SNP_filtered_region, <span class="fl">1</span>),<span class="st">" SNPs and "</span>, <span class="fu">size</span>(genosOnly_region_for_imputing, <span class="fl">1</span>),<span class="st">" filtered individuals."</span>))</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now do the KNN PCA:</p>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_process)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_process[i]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    imputed_genos <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>,filename))</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genos,<span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genos,<span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>    flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>    flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>    PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(imputed_genos, ind_with_metadata_indFiltered, </span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText<span class="op">=</span>regionText,</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> <span class="cn">true</span>, flip2 <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add position of reference genome</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>    refGenomePCAposition <span class="op">=</span> <span class="fu">predict</span>(PCAmodel.model, <span class="fu">zeros</span>(<span class="fu">size</span>(imputed_genos, <span class="fl">2</span>)))</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>    flipPC1 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">1</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># this flips PC1 if flipPC1 = true</span></span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>    flipPC2 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">2</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># same for PC2</span></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(refGenomePCAposition[<span class="fl">1</span>], refGenomePCAposition[<span class="fl">2</span>], marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">15</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">catch</span></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"NOTICE: Figure for "</span>, regionText, <span class="st">" could not be shown due to an unknown error."</span>)</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw2.KNNimputedMissing.jld2
chrgw2: 93292 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw1.KNNimputedMissing.jld2
chrgw1: 80862 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw3.KNNimputedMissing.jld2
chrgw3: 82372 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgwZ.KNNimputedMissing.jld2
chrgwZ: 53336 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw1A.KNNimputedMissing.jld2
chrgw1A: 50051 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw4.KNNimputedMissing.jld2
chrgw4: 49980 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw5.KNNimputedMissing.jld2
chrgw5: 55329 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw7.KNNimputedMissing.jld2
chrgw7: 36575 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw6.KNNimputedMissing.jld2
chrgw6: 40175 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw8.KNNimputedMissing.jld2
chrgw8: 37818 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw9.KNNimputedMissing.jld2
chrgw9: 38180 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw11.KNNimputedMissing.jld2
chrgw11: 27683 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw12.KNNimputedMissing.jld2
chrgw12: 33294 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw10.KNNimputedMissing.jld2
chrgw10: 26962 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw13.KNNimputedMissing.jld2
chrgw13: 33543 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw14.KNNimputedMissing.jld2
chrgw14: 30969 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw18.KNNimputedMissing.jld2
chrgw18: 19359 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw20.KNNimputedMissing.jld2
chrgw20: 32739 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw15.KNNimputedMissing.jld2
chrgw15: 27517 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw1B.KNNimputedMissing.jld2
chrgw1B: 638 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws100.KNNimputedMissing.jld2
chrgws100: 208 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw17.KNNimputedMissing.jld2
chrgw17: 26313 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw19.KNNimputedMissing.jld2
chrgw19: 25414 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws101.KNNimputedMissing.jld2
chrgws101: 158 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw4A.KNNimputedMissing.jld2
chrgw4A: 18467 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw21.KNNimputedMissing.jld2
chrgw21: 13321 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw26.KNNimputedMissing.jld2
chrgw26: 14303 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws102.KNNimputedMissing.jld2
chrgws102: 302 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw23.KNNimputedMissing.jld2
chrgw23: 13949 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw25.KNNimputedMissing.jld2
chrgw25: 3794 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws103.KNNimputedMissing.jld2
chrgws103: 322 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw22.KNNimputedMissing.jld2
chrgw22: 5473 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws104.KNNimputedMissing.jld2
chrgws104: 369 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw28.KNNimputedMissing.jld2
chrgw28: 11180 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw27.KNNimputedMissing.jld2
chrgw27: 9684 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw24.KNNimputedMissing.jld2
chrgw24: 13821 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws105.KNNimputedMissing.jld2
chrgws105: 475 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws106.KNNimputedMissing.jld2
chrgws106: 115 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws107.KNNimputedMissing.jld2
chrgws107: 260 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws108.KNNimputedMissing.jld2
chrgws108: 160 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws109.KNNimputedMissing.jld2
chrgws109: 310 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws110.KNNimputedMissing.jld2
chrgws110: 175 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgws112.KNNimputedMissing.jld2
chrgws112: 1884 SNPs from 257 individuals</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-4.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-5.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-6.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-7.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-8.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-9.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-10.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-11.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-12.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-13.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-14.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-15.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-16.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-17.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-18.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-19.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-20.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-21.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-22.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-23.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-24.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-25.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-26.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-27.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-28.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-29.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-30.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-31.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-32.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-33.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-34.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-35.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-36.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-37.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-38.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-39.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-40.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-41.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-42.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-43.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-17-output-44.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="whole-genome-pca" class="level3">
<h3 class="anchored" data-anchor-id="whole-genome-pca">Whole-genome PCA</h3>
<p>In addition to making PCA plots for each scaffold, we can do one for the whole genome. The imputing for the whole genome takes some time (almost 2 hours!) for this dataset because the earlier GBS plates had low read depth so more missing genotypes, so I did this in advance and saved a file. This is incorporated into the code below–to actually do the imputing, set <code>do_imputing = true</code>. Otherwise this code will load the previously-imputed data.</p>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>genosOnly_for_imputing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genosOnly_with_missing)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>regionText <span class="op">=</span> <span class="st">"wholeGenome"</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># to do the imputing, do this by setting to true, but TAKES A LONG TIME (e.g. took 8 hours on laptop with KNN method by rows):</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>do_imputing <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> do_imputing</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># @time imputed_genosOnly = Impute.svd(genosOnly_for_imputing) # old method; knn should work better</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@time</span> imputed_genosOnly <span class="op">=</span> Impute.<span class="fu">knn</span>(genosOnly_for_imputing; k<span class="op">=</span><span class="fl">1</span>, dims<span class="op">=:</span>rows)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">jldsave</span>(filename; imputed_genosOnly, ind_with_metadata_indFiltered, pos_SNP_filtered)</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>    imputed_genosOnly_wholeGenome <span class="op">=</span> imputed_genosOnly</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered_wholeGenome <span class="op">=</span> ind_with_metadata_indFiltered</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_wholeGenome <span class="op">=</span> pos_SNP_filtered</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Saved matrix of real and imputed genotypes for filtered individuals. </span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> <span class="co"># load the already saved imputing</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>    imputed_genosOnly_wholeGenome <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genosOnly"</span>)</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered_wholeGenome <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_wholeGenome <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered"</span>)</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>,filename))</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genosOnly_wholeGenome, <span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genosOnly_wholeGenome, <span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.wholeGenome.KNNimputedMissing.jld2
wholeGenome: 1017581 SNPs from 257 individuals</code></pre>
</div>
</div>
<p>Now make the whole-genome PCA:</p>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>GW_wholeGenome_PCA <span class="op">=</span> <span class="fu">plotPCA</span>(imputed_genosOnly_wholeGenome,          </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>        ind_with_metadata_indFiltered_wholeGenome, </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>        groups_to_plot_PCA, group_colors_PCA;</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText<span class="op">=</span>regionText,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_wholeGenome.PC1 <span class="op">=</span> GW_wholeGenome_PCA.PC1</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_wholeGenome.PC2 <span class="op">=</span> GW_wholeGenome_PCA.PC2</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co"># add position of reference genome</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>refGenomePCAposition <span class="op">=</span> <span class="fu">predict</span>(GW_wholeGenome_PCA.model, <span class="fu">zeros</span>(<span class="fu">size</span>(imputed_genosOnly_wholeGenome, <span class="fl">2</span>)))</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">1</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># this flips PC1 if flipPC1 = true</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">2</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># same for PC2</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(refGenomePCAposition[<span class="fl">1</span>], refGenomePCAposition[<span class="fl">2</span>], marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">15</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>try</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(GW_wholeGenome_PCA.PCAfig)</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>catch</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"NOTICE: Figure for "</span>, regionText, <span class="st">" could not be shown due to an unknown error."</span>)</span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-19-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-19-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
</section>
<section id="genotype-by-individual-plots" class="level2">
<h2 class="anchored" data-anchor-id="genotype-by-individual-plots">Genotype-by-individual plots</h2>
<p>Now, show individual genotypes for subsets of the dataset. Can choose individuals and genomic regions to plot, along with an Fst cutoff (only show SNPs with greater Fst than the cutoff).</p>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>set <span class="op">=</span> <span class="st">"vir_plumb"</span>    <span class="co">#"59_inds_around_ring"  #"east_side_of_ring"    #"67_inds_around_ring"  # "west_side_of_ring"</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> set <span class="op">==</span> <span class="st">"59_inds_around_ring"</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)   </span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"lud_PK"</span>,<span class="st">"lud_KS"</span>,<span class="st">"lud_central"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>, <span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>, <span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">15</span>, <span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">10</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_plumb"</span>   <span class="co">#"Fst_among"  #"vir_troch_LN"       #"vir_plumb"      #"troch_LN_plumb"      #"vir_troch_LN"</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.95</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"37_inds_around_ring_plusAllVirPlumb"</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"lud"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>, <span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"seagreen4"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>, <span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">100</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"Fst_among"</span></span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.7</span></span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"west_side_of_ring"</span></span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_misID"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb37-26"><a href="#cb37-26" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb37-27"><a href="#cb37-27" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-28"><a href="#cb37-28" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-29"><a href="#cb37-29" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"troch_LN"</span></span>
<span id="cb37-30"><a href="#cb37-30" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_troch_LN"</span> <span class="co"># "Fst_among"</span></span>
<span id="cb37-31"><a href="#cb37-31" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.6</span></span>
<span id="cb37-32"><a href="#cb37-32" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-33"><a href="#cb37-33" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"all_ludlowi_plus_a_few_other"</span></span>
<span id="cb37-34"><a href="#cb37-34" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb37-35"><a href="#cb37-35" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-36"><a href="#cb37-36" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"red"</span>]</span>
<span id="cb37-37"><a href="#cb37-37" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">4</span>, <span class="fl">4</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-38"><a href="#cb37-38" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-39"><a href="#cb37-39" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"troch_LN"</span></span>
<span id="cb37-40"><a href="#cb37-40" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_troch_LN"</span> <span class="co"># "Fst_among"</span></span>
<span id="cb37-41"><a href="#cb37-41" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.6</span></span>
<span id="cb37-42"><a href="#cb37-42" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-43"><a href="#cb37-43" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"east_side_of_ring"</span></span>
<span id="cb37-44"><a href="#cb37-44" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb37-45"><a href="#cb37-45" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-46"><a href="#cb37-46" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb37-47"><a href="#cb37-47" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-48"><a href="#cb37-48" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"troch_LN"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-49"><a href="#cb37-49" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb37-50"><a href="#cb37-50" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"troch_LN_plumb"</span></span>
<span id="cb37-51"><a href="#cb37-51" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.7</span></span>
<span id="cb37-52"><a href="#cb37-52" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-53"><a href="#cb37-53" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"vir_plumb"</span></span>
<span id="cb37-54"><a href="#cb37-54" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-55"><a href="#cb37-55" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"plumb_vir"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb37-56"><a href="#cb37-56" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"purple"</span>,<span class="st">"red"</span>]</span>
<span id="cb37-57"><a href="#cb37-57" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">100</span>,<span class="fl">100</span>,<span class="fl">100</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb37-58"><a href="#cb37-58" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb37-59"><a href="#cb37-59" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb37-60"><a href="#cb37-60" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_plumb"</span></span>
<span id="cb37-61"><a href="#cb37-61" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span>  <span class="fl">0.8</span></span>
<span id="cb37-62"><a href="#cb37-62" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb37-63"><a href="#cb37-63" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>0.2</code></pre>
</div>
</div>
<section id="calculate-allele-freqs-and-sample-sizes-use-column-fst_group" class="level4">
<h4 class="anchored" data-anchor-id="calculate-allele-freqs-and-sample-sizes-use-column-fst_group">Calculate allele freqs and sample sizes (use column Fst_group)</h4>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_with_missing, ind_with_metadata_indFiltered.Fst_group, groups)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
</div>
</section>
<section id="calculate-fst" class="level4">
<h4 class="anchored" data-anchor-id="calculate-fst">calculate Fst</h4>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated Fst values</code></pre>
</div>
</div>
</section>
<section id="limit-the-individuals-to-include-in-plot" class="level4">
<h4 class="anchored" data-anchor-id="limit-the-individuals-to-include-in-plot">limit the individuals to include in plot</h4>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For this figure only, filter out individuals with lots of missing genotypes</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>numMissings_threshold <span class="op">=</span> <span class="fl">800_000</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> ind_with_metadata_indFiltered.numMissings <span class="op">.&lt;</span> numMissings_threshold</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing_selected <span class="op">=</span> <span class="fu">view</span>(genosOnly_with_missing, selection, <span class="op">:</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_selected <span class="op">=</span> <span class="fu">view</span>(ind_with_metadata_indFiltered, selection, <span class="op">:</span>)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co"># now limit each group to specified numbers</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups, numIndsToPlot, genosOnly_with_missing_selected, ind_with_metadata_indFiltered_selected);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="enter-the-chromosome-scaffold-lengths" class="level4">
<h4 class="anchored" data-anchor-id="enter-the-chromosome-scaffold-lengths">Enter the chromosome (scaffold) lengths</h4>
<p>The code below will load a fasta index file for the reference genome, and then make a dictionary (a look-up table) where the key is the scaffold name and value is the chromosome length. This will be used in the genotype-by-individual figures.</p>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(repoDirectory)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>scaffold_info_filepath <span class="op">=</span> <span class="st">"metadata/GW2022ref.fa.fai"</span> <span class="co"># a fasta index file for the reference genome</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>scaffold_info <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(scaffold_info_filepath; header<span class="op">=</span>[<span class="st">"name"</span>, <span class="st">"length"</span>, <span class="st">"offset"</span>, <span class="st">"linebases"</span>, <span class="st">"linewidth"</span>], types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int</span>, <span class="dt">Int</span>, <span class="dt">Int</span>, <span class="dt">Int</span>]))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>scaffold_lengths <span class="op">=</span> <span class="fu">Dict</span>(scaffold_info.name <span class="op">.=&gt;</span> scaffold_info.length)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="choose-the-scaffold-and-region-to-show" class="level4">
<h4 class="anchored" data-anchor-id="choose-the-scaffold-and-region-to-show">Choose the scaffold and region to show</h4>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw23"</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>("gw23", 1, 8435933, "chr gw23 1 to 8435933")</code></pre>
</div>
</div>
</section>
<section id="now-actually-make-the-plot" class="level4">
<h4 class="anchored" data-anchor-id="now-actually-make-the-plot">Now actually make the plot</h4>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, missingFractionAllowed,</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    regionInfo, pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included, freqs, plotGroups, plotGroupColors);</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-26-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="choose-another-chromosome-and-plot-similarly-to-above" class="level4">
<h4 class="anchored" data-anchor-id="choose-another-chromosome-and-plot-similarly-to-above">choose another chromosome, and plot similarly to above</h4>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw26"</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, missingFractionAllowed,</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>    regionInfo, pos_SNP_filtered, Fst, pairwiseNamesFst,</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included, freqs, plotGroups, plotGroupColors);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-27-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="make-a-gbi-plot-to-illustrate-variation-along-west-side-of-ring" class="level4">
<h4 class="anchored" data-anchor-id="make-a-gbi-plot-to-illustrate-variation-along-west-side-of-ring">Make a GBI plot to illustrate variation along west side of ring</h4>
<div class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> [<span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">10</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>group2 <span class="op">=</span> <span class="st">"troch_LN"</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>groupsToCompare <span class="op">=</span> <span class="st">"vir_troch_LN"</span> <span class="co"># "Fst_among"</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>Fst_cutoff <span class="op">=</span> <span class="fl">0.9</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_with_missing, ind_with_metadata_indFiltered.Fst_group, groups)</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate Fst </span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the individuals to include in plot</span></span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a><span class="co"># For this figure only, filter out individuals with lots of missing genotypes</span></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a>numMissings_threshold <span class="op">=</span> <span class="fl">800_000</span></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> ind_with_metadata_indFiltered.numMissings <span class="op">.&lt;</span> numMissings_threshold</span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing_selected <span class="op">=</span> <span class="fu">view</span>(genosOnly_with_missing, selection, <span class="op">:</span>)</span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_selected <span class="op">=</span> <span class="fu">view</span>(ind_with_metadata_indFiltered, selection, <span class="op">:</span>)</span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a><span class="co"># now limit each group to specified numbers</span></span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a>genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups, numIndsToPlot, genosOnly_with_missing_selected, ind_with_metadata_indFiltered_selected);</span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a><span class="co"># choose the scaffold and region to show</span></span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw26"</span></span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) </span>
<span id="cb49-31"><a href="#cb49-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-32"><a href="#cb49-32" aria-hidden="true" tabindex="-1"></a><span class="co">#### Now actually make the plot</span></span>
<span id="cb49-33"><a href="#cb49-33" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, missingFractionAllowed,</span>
<span id="cb49-34"><a href="#cb49-34" aria-hidden="true" tabindex="-1"></a>    regionInfo, pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb49-35"><a href="#cb49-35" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included, freqs, plotGroups, plotGroupColors);</span>
<span id="cb49-36"><a href="#cb49-36" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-28-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="make-a-gbi-plot-to-illustrate-variation-along-east-side-of-ring" class="level4">
<h4 class="anchored" data-anchor-id="make-a-gbi-plot-to-illustrate-variation-along-east-side-of-ring">Make a GBI plot to illustrate variation along east side of ring</h4>
<div class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">17</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>group1 <span class="op">=</span> <span class="st">"troch_LN"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>groupsToCompare <span class="op">=</span> <span class="st">"troch_LN_plumb"</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>Fst_cutoff <span class="op">=</span> <span class="fl">0.9</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_with_missing, ind_with_metadata_indFiltered.Fst_group, groups)</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate Fst </span></span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the individuals to include in plot</span></span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a><span class="co"># For this figure only, filter out individuals with lots of missing genotypes</span></span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>numMissings_threshold <span class="op">=</span> <span class="fl">800_000</span></span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> ind_with_metadata_indFiltered.numMissings <span class="op">.&lt;</span> numMissings_threshold</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing_selected <span class="op">=</span> <span class="fu">view</span>(genosOnly_with_missing, selection, <span class="op">:</span>)</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_selected <span class="op">=</span> <span class="fu">view</span>(ind_with_metadata_indFiltered, selection, <span class="op">:</span>)</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a><span class="co"># now limit each group to specified numbers</span></span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups, numIndsToPlot, genosOnly_with_missing_selected, ind_with_metadata_indFiltered_selected);</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">Not</span>(ind_with_metadata_included.ind <span class="op">.==</span> <span class="st">"GW_Armando_plate1_JF09G01"</span>)</span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-31"><a href="#cb51-31" aria-hidden="true" tabindex="-1"></a><span class="co"># choose the scaffold and region to show</span></span>
<span id="cb51-32"><a href="#cb51-32" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw28"</span></span>
<span id="cb51-33"><a href="#cb51-33" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb51-34"><a href="#cb51-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-35"><a href="#cb51-35" aria-hidden="true" tabindex="-1"></a><span class="co"># need to remove two plumbeitarsus individuals that have introgression from viridanus, as that would otherwise be misleading in this figure since viridanus are not shown</span></span>
<span id="cb51-36"><a href="#cb51-36" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">Not</span>(ind_with_metadata_included.ind <span class="op">.∈</span> <span class="fu">Ref</span>([<span class="st">"GW_Armando_plate1_JF09G01"</span>, <span class="st">"GW_Armando_plate1_JF24G02"</span>]))</span>
<span id="cb51-37"><a href="#cb51-37" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included <span class="op">=</span> ind_with_metadata_included[selection,<span class="op">:</span>]</span>
<span id="cb51-38"><a href="#cb51-38" aria-hidden="true" tabindex="-1"></a>genosOnly_included <span class="op">=</span> genosOnly_included[selection,<span class="op">:</span>]</span>
<span id="cb51-39"><a href="#cb51-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-40"><a href="#cb51-40" aria-hidden="true" tabindex="-1"></a><span class="co">#### Now actually make the plot</span></span>
<span id="cb51-41"><a href="#cb51-41" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, missingFractionAllowed,</span>
<span id="cb51-42"><a href="#cb51-42" aria-hidden="true" tabindex="-1"></a>    regionInfo, pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb51-43"><a href="#cb51-43" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included, freqs, plotGroups, plotGroupColors);</span>
<span id="cb51-44"><a href="#cb51-44" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-29-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="pca-plot-of-the-east-side-of-the-ring-for-one-chromosome" class="level4">
<h4 class="anchored" data-anchor-id="pca-plot-of-the-east-side-of-the-ring-for-one-chromosome">PCA plot of the east side of the ring, for one chromosome</h4>
<div class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>eastern_groups_to_plot_PCA <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>eastern_group_colors_PCA <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>];</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>chrom <span class="op">=</span> <span class="st">"gw28"</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>imputed_genos <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered_region <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>,filename))</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genos,<span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genos,<span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(imputed_genos, ind_with_metadata_indFiltered, </span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>        eastern_groups_to_plot_PCA, eastern_group_colors_PCA; </span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText<span class="op">=</span>regionText,</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> <span class="cn">false</span>, flip2 <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgw28.KNNimputedMissing.jld2
chrgw28: 11180 SNPs from 257 individuals</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-30-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>(model = PCA(indim = 11180, outdim = 3, principalratio = 0.16508746), values = Float32[3.4456518 2.861559 … 3.3472672 2.746452; -2.8029943 -1.5361849 … -1.6793878 -1.6296741; 0.44376191 -2.3090105 … -0.34599677 2.28425], PC1 = Float32[3.4456518, 2.861559, 3.4070215, 3.7611568, -0.3116923, 3.176039, 3.6195168, 3.3295372, 3.6839056, 4.377373  …  3.6361964, 3.0057902, 3.6028526, 3.6254516, 3.7723784, 2.6325006, 4.629251, 3.147118, 3.3472672, 2.746452], PC2 = Float32[2.8029943, 1.5361849, 0.6433394, -0.08207402, -0.65680426, 0.7644048, -0.18662581, 0.8106848, 0.37520826, 0.74327254  …  0.81521857, -0.19739068, 1.3837582, -0.082504876, 0.28881872, 1.3875047, 2.444236, 0.91859967, 1.6793878, 1.6296741], PCAfig = Scene (672px, 480px):
  0 Plots
  1 Child Scene:
    └ Scene (672px, 480px))</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="calculate-distances-around-ring" class="level2">
<h2 class="anchored" data-anchor-id="calculate-distances-around-ring">Calculate distances around ring</h2>
<p>The locations around the ring (assuming barrier in North) can be graphed against genomic PC1 (or other variables).</p>
<section id="load-latlong-data" class="level4">
<h4 class="anchored" data-anchor-id="load-latlong-data">Load lat/long data</h4>
<div class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(repoDirectory)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>latlong_filepath <span class="op">=</span> <span class="st">"metadata/GW_locations_LatLong_2023.txt"</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>latlongs <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(latlong_filepath))</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(latlongs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>32×5 DataFrame
 Row │ Location_name               location_short  lat_N    long_E    subspecies    
     │ String31                    String7         Float64  Float64   String15      
─────┼──────────────────────────────────────────────────────────────────────────────
   1 │ Yekaterinburg               YK               56.8     60.6     viridanus
   2 │ Abakan                      AB               52.0     89.5     viridanus
   3 │ Teletsk                     TL               51.7     87.6     viridanus
   4 │ Stolbi                      ST_vi            55.9     92.6     viridanus
   5 │ Krasnoyarski_Krai_vi        KK_vi            55.9     92.6     viridanus
   6 │ Turkey                      TU               41.0     42.0     nitidus
   7 │ Ala_Archa                   AA               42.54    74.5     viridanus
   8 │ Naran_Pakistan              NR               34.884   73.691   ludlowi
   9 │ Shogran_Pakistan            SH               34.594   73.466   ludlowi
  10 │ Overa_Kashmir               OV               33.991   75.243   ludlowi
  11 │ Satharundhi_ChambaDistrict  SA               32.974   76.222   ludlowi
  12 │ KL_Killar_HP                KL               33.106   76.409   ludlowi
  13 │ Thalighar                   TH               32.828   76.45    ludlowi
  14 │ Sural                       SR               33.134   76.455   ludlowi
  15 │ PA_Tindi_HP                 PA               32.771   76.472   ludlowi
  16 │ Sukhto                      SU               32.868   76.855   ludlowi
  17 │ Nainaghar                   NG               32.728   76.8594  ludlowi
  18 │ Mooling_and_Keylong         ML               32.508   76.981   ludlowi
  19 │ Manali                      MN               32.237   77.13    ludlowi
  20 │ Spiti                       SP               32.377   77.281   ludlowi
  21 │ Langtang                    LN               28.25    85.5     trochiloides
  22 │ Emeishan                    EM               29.5    103.3     obscuratus
  23 │ Xining                      XN               37.0    102.0     obscuratus
  24 │ Beijing                     BJ               40.0    115.5     plumbeitarsus
  25 │ Baikal                      BK               51.9    104.9     plumbeitarsus
  26 │ Arshan                      AN               51.9    102.5     plumbeitarsus
  27 │ Ilinka                      IL               51.1     95.5     plumbeitarsus
  28 │ Tuva                        TA               51.3     92.0     plumbeitarsus
  29 │ Stolbi                      ST               55.9     92.6     plumbeitarsus
  30 │ Krasnoyarski_Krai           KK               55.9     92.6     plumbeitarsus
  31 │ Uyukski                     UY               51.9     94.1     plumbeitarsus
  32 │ Solgonski                   SL               55.7     91.0     plumbeitarsus</code></pre>
</div>
</div>
</section>
<section id="make-a-quick-plot-to-inspect-latlong-data" class="level4">
<h4 class="anchored" data-anchor-id="make-a-quick-plot-to-inspect-latlong-data">Make a quick plot to inspect latlong data:</h4>
<div class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Research locations"</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"longitude (E)"</span>,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"latitude (N)"</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">scatter!</span>(latlongs.long_E, latlongs.lat_N)</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>f</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="32">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-32-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="remove-green-warbler-nitidus" class="level4">
<h4 class="anchored" data-anchor-id="remove-green-warbler-nitidus">remove “Green warbler” <em>nitidus</em></h4>
<p><em>Phylloscopus [t.] nitidus</em> is outside of the main ring, so remove these samples from this analysis:</p>
<div class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>latlongs2 <span class="op">=</span> latlongs[<span class="fu">Not</span>(latlongs.subspecies <span class="op">.==</span> <span class="st">"nitidus"</span>), <span class="op">:</span>];</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(latlongs2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>31×5 DataFrame
 Row │ Location_name               location_short  lat_N    long_E    subspecies    
     │ String31                    String7         Float64  Float64   String15      
─────┼──────────────────────────────────────────────────────────────────────────────
   1 │ Yekaterinburg               YK               56.8     60.6     viridanus
   2 │ Abakan                      AB               52.0     89.5     viridanus
   3 │ Teletsk                     TL               51.7     87.6     viridanus
   4 │ Stolbi                      ST_vi            55.9     92.6     viridanus
   5 │ Krasnoyarski_Krai_vi        KK_vi            55.9     92.6     viridanus
   6 │ Ala_Archa                   AA               42.54    74.5     viridanus
   7 │ Naran_Pakistan              NR               34.884   73.691   ludlowi
   8 │ Shogran_Pakistan            SH               34.594   73.466   ludlowi
   9 │ Overa_Kashmir               OV               33.991   75.243   ludlowi
  10 │ Satharundhi_ChambaDistrict  SA               32.974   76.222   ludlowi
  11 │ KL_Killar_HP                KL               33.106   76.409   ludlowi
  12 │ Thalighar                   TH               32.828   76.45    ludlowi
  13 │ Sural                       SR               33.134   76.455   ludlowi
  14 │ PA_Tindi_HP                 PA               32.771   76.472   ludlowi
  15 │ Sukhto                      SU               32.868   76.855   ludlowi
  16 │ Nainaghar                   NG               32.728   76.8594  ludlowi
  17 │ Mooling_and_Keylong         ML               32.508   76.981   ludlowi
  18 │ Manali                      MN               32.237   77.13    ludlowi
  19 │ Spiti                       SP               32.377   77.281   ludlowi
  20 │ Langtang                    LN               28.25    85.5     trochiloides
  21 │ Emeishan                    EM               29.5    103.3     obscuratus
  22 │ Xining                      XN               37.0    102.0     obscuratus
  23 │ Beijing                     BJ               40.0    115.5     plumbeitarsus
  24 │ Baikal                      BK               51.9    104.9     plumbeitarsus
  25 │ Arshan                      AN               51.9    102.5     plumbeitarsus
  26 │ Ilinka                      IL               51.1     95.5     plumbeitarsus
  27 │ Tuva                        TA               51.3     92.0     plumbeitarsus
  28 │ Stolbi                      ST               55.9     92.6     plumbeitarsus
  29 │ Krasnoyarski_Krai           KK               55.9     92.6     plumbeitarsus
  30 │ Uyukski                     UY               51.9     94.1     plumbeitarsus
  31 │ Solgonski                   SL               55.7     91.0     plumbeitarsus</code></pre>
</div>
</div>
</section>
<section id="make-a-matrix-of-great-circle-distances" class="level4">
<h4 class="anchored" data-anchor-id="make-a-matrix-of-great-circle-distances">Make a matrix of great circle distances</h4>
<p>These are Haversine distances, assuming spherical Earth which is really close:</p>
<div class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>geoPoints <span class="op">=</span> <span class="fu">GeoLocation</span>.(latlongs2.long_E, latlongs2.lat_N)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># this next line is so neat--uses list comprehension to make a matrix of pairwise calculations</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>distances <span class="op">=</span> [(<span class="fu">HaversineDistance</span>(geoPoints[i], geoPoints[j])<span class="op">/</span><span class="fl">1000</span>) for i <span class="kw">in</span> <span class="fu">eachindex</span>(geoPoints), j <span class="kw">in</span> <span class="fu">eachindex</span>(geoPoints)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="34">
<pre><code>31×31 Matrix{Float64}:
    0.0   1929.03   1829.5    1956.21   …  1956.21   2214.04   1866.47
 1929.03     0.0     134.698   478.643      478.643   315.404   422.996
 1829.5    134.698     0.0     570.583      570.583   447.369   497.776
 1956.21   478.643   570.583     0.0          0.0     455.478   102.442
 1956.21   478.643   570.583     0.0          0.0     455.478   102.442
 1866.88  1539.57   1417.12   1971.61   …  1971.61   1800.23   1883.0
 2629.03  2280.66   2174.21   2744.17      2744.17   2492.37   2664.81
 2653.46  2318.81   2212.24   2782.15      2782.15   2530.6    2702.67
 2768.57  2304.58   2204.89   2775.3       2775.3    2498.42   2700.28
 2905.38  2370.69   2276.33   2845.24      2845.24   2550.34   2773.58
 2897.44  2350.4    2256.5    2825.21   …  2825.21   2529.04   2753.85
 2927.82  2377.48   2284.21   2852.64      2852.64   2554.28   2781.67
 2895.96  2345.88   2252.09   2820.75      2820.75   2524.24   2749.47
    ⋮                                   ⋱                         ⋮
 3728.75  2661.7    2613.39   3126.13      3126.13   2725.19   3083.74
 4484.94  2747.98   2788.34   3054.87   …  3054.87   2603.84   3071.78
 3751.29  1934.14   1987.84   2216.8       2216.8    1768.98   2237.85
 4317.27  2390.71   2499.26   2434.21      2434.21   2104.82   2502.82
 2869.29  1053.52   1187.02    918.603      918.603   740.322  1003.57
 2724.34   889.835  1023.03    785.302      785.302   576.015   863.881
 2341.58   426.63    551.809   567.027  …   567.027   131.542   591.859
 2117.8    189.217   307.752   513.021      513.021   159.644   493.694
 1956.21   478.643   570.583     0.0          0.0     455.478   102.442
 1956.21   478.643   570.583     0.0          0.0     455.478   102.442
 2214.04   315.404   447.369   455.478      455.478     0.0     468.914
 1866.47   422.996   497.776   102.442  …   102.442   468.914     0.0</code></pre>
</div>
</div>
</section>
<section id="now-adjust-distances-to-assume-no-gene-flow-through-centre-of-ring." class="level4">
<h4 class="anchored" data-anchor-id="now-adjust-distances-to-assume-no-gene-flow-through-centre-of-ring.">Now adjust distances to assume no gene flow through centre of ring.</h4>
<div class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get some key distances</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">getIndex</span>(name, nameVector <span class="op">=</span> latlongs2.Location_name)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">findfirst</span>(<span class="fu">isequal</span>(name), nameVector)</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>index_AA <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Ala_Archa"</span>)</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>index_PK <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Naran_Pakistan"</span>)</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>index_LN <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Langtang"</span>)</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>index_EM <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Emeishan"</span>)</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>index_XN <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Xining"</span>)</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>index_BJ <span class="op">=</span> <span class="fu">getIndex</span>(<span class="st">"Beijing"</span>)</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>index_last <span class="op">=</span> <span class="fu">nrow</span>(latlongs2)</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>dist_PK_to_LN <span class="op">=</span> distances[index_PK, index_LN]</span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>dist_LN_to_EM <span class="op">=</span> distances[index_LN, index_EM]</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>dist_EM_to_BJ <span class="op">=</span> distances[index_EM, index_BJ]</span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a><span class="co"># This next part will assume locations in the input file are arranged in order around ring:</span></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Float32}</span>(<span class="cn">undef</span>, <span class="fu">size</span>(distances)[<span class="fl">1</span>], <span class="fu">size</span>(distances)[<span class="fl">2</span>])</span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a><span class="co"># accept all distances within viridanus:</span></span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a><span class="co"># function for accepting straight-line great circle dists as distances between sets of sites</span></span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a>acceptDists <span class="op">=</span> <span class="kw">function</span>(straightGreatCircleDists, start, finish, distsAroundRing)</span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a>    distsAroundRing[start<span class="op">:</span>finish, start<span class="op">:</span>finish] <span class="op">=</span> straightGreatCircleDists[start<span class="op">:</span>finish, start<span class="op">:</span>finish]</span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(distsAroundRing)</span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a><span class="co"># accept all distances within viridanus:</span></span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, <span class="fl">1</span>, index_AA, distsAroundRing)</span>
<span id="cb63-30"><a href="#cb63-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-31"><a href="#cb63-31" aria-hidden="true" tabindex="-1"></a><span class="co"># accept dist from AA to PK:</span></span>
<span id="cb63-32"><a href="#cb63-32" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, index_AA, index_PK, distsAroundRing)</span>
<span id="cb63-33"><a href="#cb63-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-34"><a href="#cb63-34" aria-hidden="true" tabindex="-1"></a><span class="co"># accept all distances from PK to LN:</span></span>
<span id="cb63-35"><a href="#cb63-35" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, index_PK, index_LN, distsAroundRing)</span>
<span id="cb63-36"><a href="#cb63-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-37"><a href="#cb63-37" aria-hidden="true" tabindex="-1"></a><span class="co"># accept dist from LN to EM:</span></span>
<span id="cb63-38"><a href="#cb63-38" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, index_LN, index_EM, distsAroundRing)</span>
<span id="cb63-39"><a href="#cb63-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-40"><a href="#cb63-40" aria-hidden="true" tabindex="-1"></a><span class="co"># accept dists between EM, XN, BJ:</span></span>
<span id="cb63-41"><a href="#cb63-41" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, index_EM, index_BJ, distsAroundRing)</span>
<span id="cb63-42"><a href="#cb63-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-43"><a href="#cb63-43" aria-hidden="true" tabindex="-1"></a><span class="co"># accept all distances within plumbeitarsus:</span></span>
<span id="cb63-44"><a href="#cb63-44" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">acceptDists</span>(distances, index_BJ, index_last, distsAroundRing)</span>
<span id="cb63-45"><a href="#cb63-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-46"><a href="#cb63-46" aria-hidden="true" tabindex="-1"></a><span class="co"># function for adding up distances measured through certain sites:</span></span>
<span id="cb63-47"><a href="#cb63-47" aria-hidden="true" tabindex="-1"></a>addDists <span class="op">=</span> <span class="kw">function</span>(set1start, set1end, set2start, set2end, distsAroundRing)</span>
<span id="cb63-48"><a href="#cb63-48" aria-hidden="true" tabindex="-1"></a>    firstDists <span class="op">=</span> <span class="fu">repeat</span>(distsAroundRing[set1start<span class="op">:</span>(set1end<span class="op">-</span><span class="fl">1</span>), set1end], <span class="fl">1</span>, set2end<span class="op">-</span>set2start<span class="op">+</span><span class="fl">1</span>)</span>
<span id="cb63-49"><a href="#cb63-49" aria-hidden="true" tabindex="-1"></a>    secondDists <span class="op">=</span> <span class="fu">repeat</span>(<span class="fu">transpose</span>(distsAroundRing[set1end, set2start<span class="op">:</span>set2end]), set1end<span class="op">-</span>set1start, <span class="fl">1</span>)</span>
<span id="cb63-50"><a href="#cb63-50" aria-hidden="true" tabindex="-1"></a>    totalDists <span class="op">=</span> firstDists <span class="op">+</span> secondDists</span>
<span id="cb63-51"><a href="#cb63-51" aria-hidden="true" tabindex="-1"></a>    distsAroundRing[set1start<span class="op">:</span>(set1end<span class="op">-</span><span class="fl">1</span>), set2start<span class="op">:</span>set2end] <span class="op">=</span> totalDists</span>
<span id="cb63-52"><a href="#cb63-52" aria-hidden="true" tabindex="-1"></a>    distsAroundRing[set2start<span class="op">:</span>set2end, set1start<span class="op">:</span>(set1end<span class="op">-</span><span class="fl">1</span>)] <span class="op">=</span> <span class="fu">transpose</span>(totalDists)</span>
<span id="cb63-53"><a href="#cb63-53" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(distsAroundRing)</span>
<span id="cb63-54"><a href="#cb63-54" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb63-55"><a href="#cb63-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-56"><a href="#cb63-56" aria-hidden="true" tabindex="-1"></a><span class="co"># dists from viridanus to PK are sum of dists to AA plus AA to PK:</span></span>
<span id="cb63-57"><a href="#cb63-57" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">addDists</span>(<span class="fl">1</span>, index_AA, index_PK, index_PK, distsAroundRing)</span>
<span id="cb63-58"><a href="#cb63-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-59"><a href="#cb63-59" aria-hidden="true" tabindex="-1"></a><span class="co"># dists from "northwest of PK" to Himalayas are sum of ringdists to PK plus PK to locations up to LN:</span></span>
<span id="cb63-60"><a href="#cb63-60" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">addDists</span>(<span class="fl">1</span>, index_PK, index_PK<span class="op">+</span><span class="fl">1</span>, index_LN, distsAroundRing)</span>
<span id="cb63-61"><a href="#cb63-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-62"><a href="#cb63-62" aria-hidden="true" tabindex="-1"></a><span class="co"># dists from "west / northwest of LN" to EM are sum of dists to LN plus LN to EM:</span></span>
<span id="cb63-63"><a href="#cb63-63" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">addDists</span>(<span class="fl">1</span>, index_LN, index_EM, index_EM, distsAroundRing)</span>
<span id="cb63-64"><a href="#cb63-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-65"><a href="#cb63-65" aria-hidden="true" tabindex="-1"></a><span class="co"># dists from "west / northwest of EM" to China are sum of dists to EM plus EM to (XN, BJ):</span></span>
<span id="cb63-66"><a href="#cb63-66" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">addDists</span>(<span class="fl">1</span>, index_EM, index_XN, index_BJ, distsAroundRing)</span>
<span id="cb63-67"><a href="#cb63-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-68"><a href="#cb63-68" aria-hidden="true" tabindex="-1"></a><span class="co"># dists from "west of BJ" to east Siberia are sum of dists to BJ plus BJ to other plumbeitarsus:</span></span>
<span id="cb63-69"><a href="#cb63-69" aria-hidden="true" tabindex="-1"></a>distsAroundRing <span class="op">=</span> <span class="fu">addDists</span>(<span class="fl">1</span>, index_BJ, index_BJ<span class="op">+</span><span class="fl">1</span>, index_last, distsAroundRing);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="do-principal-coordinates-analysis-on-the-distances-around-the-ring" class="level4">
<h4 class="anchored" data-anchor-id="do-principal-coordinates-analysis-on-the-distances-around-the-ring">Do Principal Coordinates Analysis on the distances around the ring</h4>
<p>This produces a single location axis around ring, going from west Siberia south, then east, then north to east Siberia.</p>
<div class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>PCO_around_ring <span class="op">=</span> <span class="fu">fit</span>(MDS, distsAroundRing; distances<span class="op">=</span><span class="cn">true</span>, maxoutdim<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add this as a column to the data frame:</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>latlongs2.LocationAroundRing <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">-predict</span>(PCO_around_ring))</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co"># another way: </span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="co"># latlongs2[:, :LocationAroundRing] = vec(-predict(PCO_around_ring))</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>latlongs2[<span class="op">:</span>, [<span class="op">:</span>location_short, <span class="op">:</span>LocationAroundRing]]</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(latlongs2[<span class="op">:</span>, [<span class="op">:</span>location_short, <span class="op">:</span>LocationAroundRing]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>31×2 DataFrame
 Row │ location_short  LocationAroundRing 
     │ String7         Float32            
─────┼────────────────────────────────────
   1 │ YK                       -4484.22
   2 │ AB                       -4223.99
   3 │ TL                       -4104.65
   4 │ ST_vi                    -4655.8
   5 │ KK_vi                    -4655.8
   6 │ AA                       -2705.3
   7 │ NR                       -1849.44
   8 │ SH                       -1844.6
   9 │ OV                       -1675.97
  10 │ SA                       -1539.62
  11 │ KL                       -1532.33
  12 │ TH                       -1513.07
  13 │ SR                       -1530.28
  14 │ PA                       -1508.06
  15 │ SU                       -1483.0
  16 │ NG                       -1474.58
  17 │ ML                       -1452.26
  18 │ MN                       -1424.96
  19 │ SP                       -1420.66
  20 │ LN                        -507.711
  21 │ EM                        1233.65
  22 │ XN                        1914.05
  23 │ BJ                        2846.22
  24 │ BK                        4390.92
  25 │ AN                        4498.45
  26 │ IL                        4817.46
  27 │ TA                        5039.28
  28 │ ST                        5276.86
  29 │ KK                        5276.86
  30 │ UY                        4946.53
  31 │ SL                        5346.01</code></pre>
</div>
</div>
<p>Add these ring locations to the metadata table:</p>
<div class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_wholeGenome.ring_km <span class="op">.=</span> <span class="cn">NaN</span>  <span class="co"># pre-allocate the column</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(latlongs2, <span class="fl">1</span>)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>    match_indices <span class="op">=</span> <span class="fu">findall</span>(ind_with_metadata_indFiltered_wholeGenome.location <span class="op">.==</span> latlongs2.location_short[i])</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered_wholeGenome.ring_km[match_indices] <span class="op">.=</span> latlongs2.LocationAroundRing[i];</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-location-around-ring-vs.-pc1" class="level4">
<h4 class="anchored" data-anchor-id="plot-location-around-ring-vs.-pc1">Plot location around ring vs.&nbsp;PC1:</h4>
<div class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(ind_with_metadata_indFiltered_wholeGenome.ring_km, ind_with_metadata_indFiltered_wholeGenome.PC1)</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Genomic PC1 around ring"</span>,</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Location around ring (km)"</span>,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Genomic PC1"</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>jitterSize <span class="op">=</span> <span class="fl">100</span>   <span class="co"># in km</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> ind_with_metadata_indFiltered_wholeGenome.ring_km <span class="op">.+</span> jitterSize <span class="op">.*</span> (<span class="fu">rand</span>(<span class="fu">length</span>(ind_with_metadata_indFiltered_wholeGenome.PC1)) <span class="op">.-</span> <span class="fl">0.5</span>)</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> ind_with_metadata_indFiltered_wholeGenome.PC1</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> ind_with_metadata_indFiltered_wholeGenome.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values[selection], y_plot_values[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span>group_colors_PCA[i], markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-38-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="z-chromosome-differentiation" class="level1">
<h1>Z chromosome differentiation</h1>
<p>The PCA of Z chromosome variation (see far above) showed some unexpected structure, which I think it likely due to females being shifted (especially on PC2) compared to males. I think this is likely due to females having W chromosomes, reads from which are sometimes incorrectly mapped to the Z. Hence a small fraction of SNPs differ between females and males (with females called as heterozygotes, males as homozygotes, at those SNPs). However, at loci that are only on the Z, males can be heterozygous (2 alleles) whereas females can only be hemizygous (one allele). I am not sure of the net effect.</p>
<p>To remove this problem, we could take several approaches (e.g., remove problematic SNPs, or look at one sex at a time). Here we’ll plot only males as that will remove the W chromosome. To infer sex of the birds, we’ll try plotting heterozygosity of Z chromosome SNPs per bird.</p>
<div class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>chrom <span class="op">=</span> <span class="st">"gwZ"</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chrom)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered_gwZ <span class="op">=</span> pos_SNP_filtered[loci_selection,<span class="op">:</span>]</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>genotypes_gwZ <span class="op">=</span> genosOnly[<span class="op">:</span>,loci_selection]</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>numHetSNPs <span class="op">=</span> <span class="fu">sum</span>(genotypes_gwZ <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>numGenotypedSNPs <span class="op">=</span> <span class="fu">sum</span>(<span class="fu">map</span>(<span class="fu">in</span>([<span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>]), genotypes_gwZ), dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>hetFraction <span class="op">=</span> numHetSNPs <span class="op">./</span> numGenotypedSNPs</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered[!, <span class="op">:</span>hetFractionZ] <span class="op">.=</span> hetFraction</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Average heterozygosity at SNPs on the Z chromosome"</span>,</span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Individual (order as in metadata)"</span>,</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Heterozygosity (as fraction of genotypes SNPs)"</span></span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> <span class="fu">eachindex</span>(ind_with_metadata_indFiltered.hetFractionZ)</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> ind_with_metadata_indFiltered.hetFractionZ</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values, y_plot_values, marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-39-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>That doesn’t provide super clear separation.</p>
<p>Let’s try looking at number of missing genotypes:</p>
<div class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>missingGenotypeCount <span class="op">=</span> <span class="fu">sum</span>(genotypes_gwZ <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered[!, <span class="op">:</span>missingZgenotypeCount] <span class="op">.=</span> missingGenotypeCount</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Number of missing SNP genotypes on the Z chromosome"</span>,</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Individual (order as in metadata)"</span>,</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Number of missing SNPs (out of 53336 SNPs)"</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> <span class="fu">eachindex</span>(ind_with_metadata_indFiltered.missingZgenotypeCount)</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> ind_with_metadata_indFiltered.missingZgenotypeCount</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values, y_plot_values, marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-40-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Now let’s plot those together:</p>
<div class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Missing vs. fraction heterozygous Z SNPs"</span>,</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Number missing genotypes"</span>,</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Fraction heterozygous"</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> ind_with_metadata_indFiltered.missingZgenotypeCount</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> ind_with_metadata_indFiltered.hetFractionZ</span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values, y_plot_values, marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-41-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>I don’t think number missing is very helpful–probably similar in females and males.</p>
<p>I realized that Z-chromosome read depth is probably the best way I presently have to infer sex of individuals. So I used vcftools to compare average SNP read depth on chr gwZ and the largest of the autosomes, gw2. These commands were run in the Terminal (I’ve simplified the file paths here):</p>
<pre><code>vcftools --vcf GW2022_all4plates.genotypes.SNPs_only.whole_genome.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf --chr gwZ --depth --out GW2022_all4plates.genotypes.SNPs_only.chrgwZ.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf

# After filtering, kept 310 out of 310 Individuals
# Outputting Mean Depth by Individual
# After filtering, kept 134926 out of a possible 2431709 Sites
# Run Time = 155.00 seconds

vcftools --vcf GW2022_all4plates.genotypes.SNPs_only.whole_genome.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf --chr gw2 --depth --out GW2022_all4plates.genotypes.SNPs_only.chrgw2.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf

# After filtering, kept 310 out of 310 Individuals
# Outputting Mean Depth by Individual
# After filtering, kept 229227 out of a possible 2431709 Sites
# Run Time = 155.00 seconds</code></pre>
<p>The plan is to load these read depth files into Julia and determine sex of individuals from the ratio of read depth on chromosome Z to chromosome 2:</p>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>filename_gwZ <span class="op">=</span> <span class="st">"metadata/GW2022_all4plates.genotypes.SNPs_only.chrgwZ.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf.idepth"</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>readDepthZ <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(filename_gwZ))</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>filename_gw2 <span class="op">=</span> <span class="st">"metadata/GW2022_all4plates.genotypes.SNPs_only.chrgw2.max2allele_noindel.vcf.maxmiss60.MQ20.lowHet.vcf.idepth"</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>readDepth2 <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(filename_gw2))</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>readDepthRatioZto2 <span class="op">=</span> <span class="fu">innerjoin</span>(readDepthZ, readDepth2, on <span class="op">=</span> <span class="op">:</span>INDV, renamecols <span class="op">=</span> <span class="st">"_gwZ"</span> <span class="op">=&gt;</span> <span class="st">"_gw2"</span>)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>readDepthRatioZto2[!, <span class="op">:</span>depthRatio] <span class="op">=</span> readDepthRatioZto2.MEAN_DEPTH_gwZ <span class="op">./</span> readDepthRatioZto2.MEAN_DEPTH_gw2</span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_sex <span class="op">=</span> <span class="fu">leftjoin</span>(ind_with_metadata_indFiltered, readDepthRatioZto2, on <span class="op">=</span> <span class="op">:</span>ind <span class="op">=&gt;</span> <span class="op">:</span>INDV)</span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_indFiltered_sex.depthRatio)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="42">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-42-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>That is quite a clear difference. The cluster with the lower Z / autosome read depth ratio should correspond to females (one copy of Z), whereas the higher Z / autosome ratio should be males. Interestingly, the ratios are a little higher than the expected 0.5 and 1, likely due to some repetitive elements (on the Z) and some W fragments being mapped to Z. Additional, the pseudoautosomal region, which is homologous and recombines between Z and W, will drive up Z read depth in females.</p>
<p>It is these SNPs with a W copy that I will now try to detect and remove. They should show high heterozygosity in females.</p>
<div class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_sex.sex <span class="op">.=</span> <span class="st">"na"</span> </span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>females <span class="op">=</span> ind_with_metadata_indFiltered_sex.depthRatio <span class="op">.&lt;</span> <span class="fl">0.8</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_sex.sex[females] <span class="op">.=</span> <span class="st">"F"</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>num_females <span class="op">=</span> <span class="fu">sum</span>(females)</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>females_genotypes_gwZ <span class="op">=</span> <span class="fu">view</span>(genotypes_gwZ, females, <span class="op">:</span>)</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>numHetsPerSNP_females <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">sum</span>(females_genotypes_gwZ <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>))</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>female_heterozygosity <span class="op">=</span> numHetsPerSNP_females <span class="op">./</span> num_females</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">hist</span>(female_heterozygosity)</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">plot</span>(female_heterozygosity)</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a><span class="co"># compare with males:</span></span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>males <span class="op">=</span> ind_with_metadata_indFiltered_sex.depthRatio <span class="op">.&gt;</span> <span class="fl">0.9</span></span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_sex.sex[males] <span class="op">.=</span> <span class="st">"M"</span></span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a>num_males <span class="op">=</span> <span class="fu">sum</span>(males)</span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a>males_genotypes_gwZ <span class="op">=</span> <span class="fu">view</span>(genotypes_gwZ, males, <span class="op">:</span>)</span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a>numHetsPerSNP_males <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">sum</span>(males_genotypes_gwZ <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>))</span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a>male_heterozygosity <span class="op">=</span> numHetsPerSNP_males <span class="op">./</span> num_males</span>
<span id="cb73-18"><a href="#cb73-18" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">hist</span>(male_heterozygosity)</span>
<span id="cb73-19"><a href="#cb73-19" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">plot</span>(male_heterozygosity)</span>
<span id="cb73-20"><a href="#cb73-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-21"><a href="#cb73-21" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter</span>(<span class="fu">vec</span>(numHetsPerSNP_females <span class="op">./</span> num_females), <span class="fu">vec</span>(numHetsPerSNP_males <span class="op">./</span> num_males),</span>
<span id="cb73-22"><a href="#cb73-22" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Proportion females heterozygous"</span>,</span>
<span id="cb73-23"><a href="#cb73-23" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Proportion males heterozygous"</span>)</span>
<span id="cb73-24"><a href="#cb73-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-25"><a href="#cb73-25" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb73-26"><a href="#cb73-26" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb73-27"><a href="#cb73-27" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"Heterozygosity by sex of putative Z-chromosome SNPs"</span>,</span>
<span id="cb73-28"><a href="#cb73-28" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Proportion females heterozygous"</span>,</span>
<span id="cb73-29"><a href="#cb73-29" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Proportion males heterozygous"</span></span>
<span id="cb73-30"><a href="#cb73-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb73-31"><a href="#cb73-31" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> female_heterozygosity</span>
<span id="cb73-32"><a href="#cb73-32" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> male_heterozygosity</span>
<span id="cb73-33"><a href="#cb73-33" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values, y_plot_values, marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb73-34"><a href="#cb73-34" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-43-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>This really clarifies things. The SNPs in the middle of the graph, with similar heterozygosity in females and males, are consistent with being pseudoautosomal SNPs. But the SNPs that have much higher female heterozygosity than male heterozygosity could be non-pseudoautosomal SNPs with reads from the W mapping to the Z, and giving appearance of Z heterozygosity in females (and not males). It is these SNPs we should remove. Also, there are a few with &gt; 0.5 heterozygosity in males, which should be removed because they are likely due to paralogs.</p>
<p>So, going to remove SNPs according to these rules: If (female heterozygosity is above 0.05) AND (ratio of male to female heterozygosity is less than 1/2), REMOVE the SNP from consideration. If (male heterozygosity is above 0.5), REMOVE the SNP.</p>
<div class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>removed_SNPs <span class="op">=</span> (male_heterozygosity <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.||</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                ((female_heterozygosity <span class="op">.&gt;</span> <span class="fl">0.05</span>) <span class="op">.&amp;&amp;</span> </span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>                ((male_heterozygosity <span class="op">./</span> female_heterozygosity) <span class="op">.&lt;</span> <span class="fl">0.5</span> ))</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>genotypes_gwZ_SNPfiltered <span class="op">=</span> genotypes_gwZ[<span class="op">:</span>, <span class="fu">Not</span>(removed_SNPs)]</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered_gwZ_SNPfiltered <span class="op">=</span> pos_SNP_filtered_gwZ[<span class="fu">Not</span>(removed_SNPs),<span class="op">:</span>]</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"This filtering removed "</span>, <span class="fu">sum</span>(removed_SNPs), <span class="st">" SNPs."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>This filtering removed 1831 SNPs.</code></pre>
</div>
</div>
<p>Check that the filtering did remove the problematic SNPs:</p>
<div class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>females_genotypes_gwZ <span class="op">=</span> <span class="fu">view</span>(genotypes_gwZ_SNPfiltered, females, <span class="op">:</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>numHetsPerSNP_females <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">sum</span>(females_genotypes_gwZ <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>))</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>female_heterozygosity <span class="op">=</span> numHetsPerSNP_females <span class="op">./</span> num_females</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">hist</span>(female_heterozygosity)</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">plot</span>(female_heterozygosity)</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a><span class="co"># compare with males:</span></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>males_genotypes_gwZ <span class="op">=</span> <span class="fu">view</span>(genotypes_gwZ_SNPfiltered, males, <span class="op">:</span>)</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>numHetsPerSNP_males <span class="op">=</span> <span class="fu">vec</span>(<span class="fu">sum</span>(males_genotypes_gwZ <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>))</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>male_heterozygosity <span class="op">=</span> numHetsPerSNP_males <span class="op">./</span> num_males</span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">hist</span>(male_heterozygosity)</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">plot</span>(male_heterozygosity)</span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter</span>(<span class="fu">vec</span>(numHetsPerSNP_females <span class="op">./</span> num_females), <span class="fu">vec</span>(numHetsPerSNP_males <span class="op">./</span> num_males),</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Proportion females heterozygous"</span>,</span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Proportion males heterozygous"</span>)</span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"After filtering of SNPs: Heterozygosity by sex of putative Z-chromosome SNPs"</span>,</span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Proportion females heterozygous"</span>,</span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Proportion males heterozygous"</span></span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a>x_plot_values <span class="op">=</span> female_heterozygosity</span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a>y_plot_values <span class="op">=</span> male_heterozygosity</span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(ax, x_plot_values, y_plot_values, marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">10</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb76-27"><a href="#cb76-27" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-45-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Looks good. Now impute and do the PCA. I did the imputing previously (to do it again, set <code>do_imputing = true</code> below).</p>
<div class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>genotypes_gwZ_SNPfiltered_with_missing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genotypes_gwZ_SNPfiltered)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># change "-1" to "missing":</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>genotypes_gwZ_SNPfiltered_with_missing[genotypes_gwZ_SNPfiltered_with_missing <span class="op">.==</span> <span class="op">-</span><span class="fl">1</span>] <span class="op">.=</span> <span class="cn">missing</span>;</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>regionText <span class="op">=</span> <span class="st">"chrgwZ_cleaned"</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="co"># to do the imputing, do this by setting to true:</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>do_imputing <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> do_imputing</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@time</span> imputed_genos <span class="op">=</span> Impute.<span class="fu">knn</span>(genotypes_gwZ_SNPfiltered_with_missing; k<span class="op">=</span><span class="fl">1</span>, dims<span class="op">=:</span>rows)</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># took 46 sec</span></span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">jldsave</span>(filename; imputed_genos, ind_with_metadata_indFiltered <span class="op">=</span> ind_with_metadata_indFiltered_sex, pos_SNP_filtered_region <span class="op">=</span> pos_SNP_filtered_gwZ_SNPfiltered)</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>    imputed_genos_chrZcleaned <span class="op">=</span> imputed_genos</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered_sex_chrZcleaned <span class="op">=</span> ind_with_metadata_indFiltered_sex</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_chZcleaned <span class="op">=</span> pos_SNP_filtered_gwZ_SNPfiltered</span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Saved matrix of real and imputed Z chromosome genotypes. </span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> <span class="co"># load the already saved imputing</span></span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a>    imputed_genos_chrZcleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_indFiltered_sex_chrZcleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb77-20"><a href="#cb77-20" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_chZcleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb77-21"><a href="#cb77-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>,filename))</span>
<span id="cb77-22"><a href="#cb77-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genos_chrZcleaned, <span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genos_chrZcleaned, <span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb77-23"><a href="#cb77-23" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb77-24"><a href="#cb77-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(imputed_genos_chrZcleaned, ind_with_metadata_indFiltered_sex_chrZcleaned, </span>
<span id="cb77-25"><a href="#cb77-25" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb77-26"><a href="#cb77-26" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText<span class="op">=</span>regionText,</span>
<span id="cb77-27"><a href="#cb77-27" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> <span class="cn">true</span>, flip2 <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Sept2023.chrgwZ_cleaned.KNNimputedMissing.jld2
chrgwZ_cleaned: 51505 SNPs from 257 individuals</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-46-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display" data-execution_count="46">
<pre><code>(model = PCA(indim = 51505, outdim = 3, principalratio = 0.29669666), values = Float32[26.49374 -27.183502 … 2.318793 -1.6987975; -16.098654 -11.69909 … 16.598045 19.148478; 11.760423 0.39466065 … 2.8587532 1.6929432], PC1 = Float32[-26.49374, 27.183502, 25.092571, 26.133352, 26.256199, 24.569853, 26.721336, 9.051643, 27.068913, 25.516523  …  -0.21135885, -9.800941, -4.118909, 1.4636457, -1.21454, 1.5925986, -19.479843, 1.205171, -2.318793, 1.6987975], PC2 = Float32[16.098654, 11.69909, 11.81376, 11.603577, 10.503201, 11.48941, 10.991947, 12.992962, 10.841367, 11.300306  …  -19.243574, -7.4317665, -14.9319515, -18.592932, -20.60289, -19.11184, 7.1143894, -21.90177, -16.598045, -19.148478], PCAfig = Scene (672px, 480px):
  0 Plots
  1 Child Scene:
    └ Scene (672px, 480px))</code></pre>
</div>
</div>
<p>Looks Great!! Now have a Z chromosome PCA with both males and females, based on only Z-chromosome markers.</p>
<section id="genotype-by-individual-plots-for-z-chromosome" class="level4">
<h4 class="anchored" data-anchor-id="genotype-by-individual-plots-for-z-chromosome">Genotype-by-individual plots for Z chromosome</h4>
<div class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>set <span class="op">=</span> <span class="st">"55_inds_around_ring"</span>  <span class="co">#"east_side_of_ring"    #"67_inds_around_ring"  # "west_side_of_ring"</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> set <span class="op">==</span> <span class="st">"55_inds_around_ring"</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)   </span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"lud_PK"</span>,<span class="st">"lud_KS"</span>,<span class="st">"lud_central"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>, <span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>]</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">10</span>, <span class="fl">1</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_plumb"</span>   <span class="co">#"Fst_among"  #"vir_troch_LN"       #"vir_plumb"      #"troch_LN_plumb"      #"vir_troch_LN"</span></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.95</span></span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"37_inds_around_ring_plusAllVirPlumb"</span></span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"lud"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>, <span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"seagreen4"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>, <span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">100</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-18"><a href="#cb80-18" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-19"><a href="#cb80-19" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb80-20"><a href="#cb80-20" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"Fst_among"</span></span>
<span id="cb80-21"><a href="#cb80-21" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.7</span></span>
<span id="cb80-22"><a href="#cb80-22" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-23"><a href="#cb80-23" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"west_side_of_ring"</span></span>
<span id="cb80-24"><a href="#cb80-24" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb80-25"><a href="#cb80-25" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_misID"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb80-26"><a href="#cb80-26" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb80-27"><a href="#cb80-27" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-28"><a href="#cb80-28" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-29"><a href="#cb80-29" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"troch_LN"</span></span>
<span id="cb80-30"><a href="#cb80-30" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_troch_LN"</span> <span class="co"># "Fst_among"</span></span>
<span id="cb80-31"><a href="#cb80-31" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.6</span></span>
<span id="cb80-32"><a href="#cb80-32" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-33"><a href="#cb80-33" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"all_ludlowi_plus_a_few_other"</span></span>
<span id="cb80-34"><a href="#cb80-34" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb80-35"><a href="#cb80-35" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb80-36"><a href="#cb80-36" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"red"</span>]</span>
<span id="cb80-37"><a href="#cb80-37" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">4</span>, <span class="fl">4</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-38"><a href="#cb80-38" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-39"><a href="#cb80-39" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"troch_LN"</span></span>
<span id="cb80-40"><a href="#cb80-40" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_troch_LN"</span> <span class="co"># "Fst_among"</span></span>
<span id="cb80-41"><a href="#cb80-41" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.6</span></span>
<span id="cb80-42"><a href="#cb80-42" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-43"><a href="#cb80-43" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"east_side_of_ring"</span></span>
<span id="cb80-44"><a href="#cb80-44" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)</span></span>
<span id="cb80-45"><a href="#cb80-45" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb80-46"><a href="#cb80-46" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb80-47"><a href="#cb80-47" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-48"><a href="#cb80-48" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"troch_LN"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-49"><a href="#cb80-49" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb80-50"><a href="#cb80-50" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"troch_LN_plumb"</span></span>
<span id="cb80-51"><a href="#cb80-51" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span> <span class="fl">0.7</span></span>
<span id="cb80-52"><a href="#cb80-52" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-53"><a href="#cb80-53" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> set <span class="op">==</span> <span class="st">"vir_plumb"</span></span>
<span id="cb80-54"><a href="#cb80-54" aria-hidden="true" tabindex="-1"></a>    groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb80-55"><a href="#cb80-55" aria-hidden="true" tabindex="-1"></a>    plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"plumb_vir"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb80-56"><a href="#cb80-56" aria-hidden="true" tabindex="-1"></a>    plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"purple"</span>,<span class="st">"red"</span>]</span>
<span id="cb80-57"><a href="#cb80-57" aria-hidden="true" tabindex="-1"></a>    numIndsToPlot <span class="op">=</span> [<span class="fl">100</span>,<span class="fl">100</span>,<span class="fl">100</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb80-58"><a href="#cb80-58" aria-hidden="true" tabindex="-1"></a>    group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb80-59"><a href="#cb80-59" aria-hidden="true" tabindex="-1"></a>    group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb80-60"><a href="#cb80-60" aria-hidden="true" tabindex="-1"></a>    groupsToCompare <span class="op">=</span> <span class="st">"vir_plumb"</span></span>
<span id="cb80-61"><a href="#cb80-61" aria-hidden="true" tabindex="-1"></a>    Fst_cutoff <span class="op">=</span>  <span class="fl">0.7</span></span>
<span id="cb80-62"><a href="#cb80-62" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb80-63"><a href="#cb80-63" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb80-64"><a href="#cb80-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-65"><a href="#cb80-65" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb80-66"><a href="#cb80-66" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genotypes_gwZ_SNPfiltered_with_missing, ind_with_metadata_indFiltered_sex_chrZcleaned.Fst_group, groups)</span>
<span id="cb80-67"><a href="#cb80-67" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb80-68"><a href="#cb80-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-69"><a href="#cb80-69" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb80-70"><a href="#cb80-70" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span>
<span id="cb80-71"><a href="#cb80-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-72"><a href="#cb80-72" aria-hidden="true" tabindex="-1"></a><span class="co"># For this Z-chromosome genotype-by-individual plot, include only males (so 2 Z chromosomes) and filter out individuals with lots of missing genotypes</span></span>
<span id="cb80-73"><a href="#cb80-73" aria-hidden="true" tabindex="-1"></a>numMissings_threshold <span class="op">=</span> <span class="fl">800_000</span></span>
<span id="cb80-74"><a href="#cb80-74" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (ind_with_metadata_indFiltered_sex_chrZcleaned.numMissings <span class="op">.&lt;</span> numMissings_threshold) <span class="op">.&amp;&amp;</span> (ind_with_metadata_indFiltered_sex_chrZcleaned.sex <span class="op">.==</span> <span class="st">"M"</span>) </span>
<span id="cb80-75"><a href="#cb80-75" aria-hidden="true" tabindex="-1"></a>genotypes_gwZ_SNPfiltered_with_missing_selected <span class="op">=</span> <span class="fu">view</span>(genotypes_gwZ_SNPfiltered_with_missing, selection, <span class="op">:</span>)</span>
<span id="cb80-76"><a href="#cb80-76" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_sex_chrZcleaned_selected <span class="op">=</span> <span class="fu">view</span>(ind_with_metadata_indFiltered_sex_chrZcleaned, selection, <span class="op">:</span>)</span>
<span id="cb80-77"><a href="#cb80-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-78"><a href="#cb80-78" aria-hidden="true" tabindex="-1"></a><span class="co"># now limit each group to specified numbers</span></span>
<span id="cb80-79"><a href="#cb80-79" aria-hidden="true" tabindex="-1"></a>genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups, numIndsToPlot, genotypes_gwZ_SNPfiltered_with_missing_selected, ind_with_metadata_indFiltered_sex_chrZcleaned_selected);</span>
<span id="cb80-80"><a href="#cb80-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-81"><a href="#cb80-81" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gwZ"</span></span>
<span id="cb80-82"><a href="#cb80-82" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered_chZcleaned, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr])</span>
<span id="cb80-83"><a href="#cb80-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-84"><a href="#cb80-84" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, missingFractionAllowed,</span>
<span id="cb80-85"><a href="#cb80-85" aria-hidden="true" tabindex="-1"></a>    regionInfo, pos_SNP_filtered_chZcleaned, Fst, pairwiseNamesFst, </span>
<span id="cb80-86"><a href="#cb80-86" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included, freqs, plotGroups, plotGroupColors);</span>
<span id="cb80-87"><a href="#cb80-87" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-47-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="make-final-whole-genome-pca" class="level1">
<h1>Make final whole-genome PCA</h1>
<p>Now that the Z-chromosome problem has been solved (by removing SNPs that had a divergent W sequence mapped onto that Z location), we can construct a new whole-genome PCA. I will combine the saved imputed genotypes for each chromosome into a large data matrix, and conduct PCA on that:</p>
<div class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>chromosomes_to_process[chromosomes_to_process <span class="op">.==</span> <span class="st">"gwZ"</span>] <span class="op">.=</span> <span class="st">"gwZ_cleaned"</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize data structures for gneotypes and positions</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>genos_imputed_loaded <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(<span class="cn">undef</span>, <span class="fu">nrow</span>(ind_with_metadata_indFiltered), <span class="fl">0</span>)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>pos_SNP_loaded <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="dt">String</span>[], position <span class="op">=</span> <span class="dt">Int64</span>[])</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_process)</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_process[i]</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>    imputed_genos_one_chr <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>    genos_imputed_loaded <span class="op">=</span> <span class="fu">hcat</span>(genos_imputed_loaded, imputed_genos_one_chr)</span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> ind_with_metadata_indFiltered.ind <span class="op">!=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)[<span class="op">:</span>, <span class="op">:</span>ind]</span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"""Warning: "ind" columns in loaded data and memory data don't match."""</span>)</span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>    pos_SNP_loaded <span class="op">=</span> <span class="fu">vcat</span>(pos_SNP_loaded, pos_SNP_filtered_region)</span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># println(string("Loaded ",filename))</span></span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># println(string(regionText, ": ", size(imputed_genos_one_chr,2), " SNPs from ", size(imputed_genos_one_chr,1), " individuals"))</span></span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb82-19"><a href="#cb82-19" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb82-20"><a href="#cb82-20" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb82-21"><a href="#cb82-21" aria-hidden="true" tabindex="-1"></a>PCA_wholeGenome <span class="op">=</span> <span class="fu">plotPCA</span>(genos_imputed_loaded, ind_with_metadata_indFiltered, </span>
<span id="cb82-22"><a href="#cb82-22" aria-hidden="true" tabindex="-1"></a>        groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb82-23"><a href="#cb82-23" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> <span class="st">"wholeGenome"</span>,</span>
<span id="cb82-24"><a href="#cb82-24" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2)</span>
<span id="cb82-25"><a href="#cb82-25" aria-hidden="true" tabindex="-1"></a>totalObservationVariance <span class="op">=</span> <span class="fu">var</span>(PCA_wholeGenome.model) </span>
<span id="cb82-26"><a href="#cb82-26" aria-hidden="true" tabindex="-1"></a>PC1_variance, PC2_variance <span class="op">=</span> <span class="fu">principalvars</span>(PCA_wholeGenome.model)[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>]</span>
<span id="cb82-27"><a href="#cb82-27" aria-hidden="true" tabindex="-1"></a>PC1_prop_variance <span class="op">=</span> PC1_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb82-28"><a href="#cb82-28" aria-hidden="true" tabindex="-1"></a>PC2_prop_variance <span class="op">=</span> PC2_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb82-29"><a href="#cb82-29" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"PC1 explains "</span>, <span class="fl">100</span><span class="op">*</span>PC1_prop_variance, <span class="st">"% of the total variance.</span></span>
<span id="cb82-30"><a href="#cb82-30" aria-hidden="true" tabindex="-1"></a>PC2 explains <span class="st">", 100*PC2_prop_variance, "</span><span class="op">%</span>.<span class="st">")</span></span>
<span id="cb82-31"><a href="#cb82-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-32"><a href="#cb82-32" aria-hidden="true" tabindex="-1"></a><span class="co"># add position of reference genome</span></span>
<span id="cb82-33"><a href="#cb82-33" aria-hidden="true" tabindex="-1"></a>refGenomePCAposition <span class="op">=</span> <span class="fu">predict</span>(PCA_wholeGenome.model, <span class="fu">zeros</span>(<span class="fu">size</span>(genos_imputed_loaded, <span class="fl">2</span>)))</span>
<span id="cb82-34"><a href="#cb82-34" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">1</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># this flips PC1 if flipPC1 = true</span></span>
<span id="cb82-35"><a href="#cb82-35" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">2</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># same for PC2</span></span>
<span id="cb82-36"><a href="#cb82-36" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(refGenomePCAposition[<span class="fl">1</span>], refGenomePCAposition[<span class="fl">2</span>], marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">15</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb82-37"><a href="#cb82-37" aria-hidden="true" tabindex="-1"></a>try</span>
<span id="cb82-38"><a href="#cb82-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(PCA_wholeGenome.PCAfig)</span>
<span id="cb82-39"><a href="#cb82-39" aria-hidden="true" tabindex="-1"></a>catch</span>
<span id="cb82-40"><a href="#cb82-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"NOTICE: Figure for "</span>, regionText, <span class="st">" could not be shown due to an unknown error."</span>)</span>
<span id="cb82-41"><a href="#cb82-41" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-48-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>PC1 explains 12.106356% of the total variance.
PC2 explains 6.406127%.</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-48-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display" data-execution_count="48">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="make-a-whole-genome-pca-just-for-the-eastern-side-of-the-ring" class="level4">
<h4 class="anchored" data-anchor-id="make-a-whole-genome-pca-just-for-the-eastern-side-of-the-ring">Make a whole-genome PCA just for the eastern side of the ring</h4>
<div class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>eastern_groups_to_plot_PCA <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>eastern_group_colors_PCA <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>PCA_wholeGenome <span class="op">=</span> <span class="fu">plotPCA</span>(genos_imputed_loaded, ind_with_metadata_indFiltered, </span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>        eastern_groups_to_plot_PCA, eastern_group_colors_PCA; </span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> <span class="st">"wholeGenome"</span>,</span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2)</span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>totalObservationVariance <span class="op">=</span> <span class="fu">var</span>(PCA_wholeGenome.model) </span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a>PC1_variance, PC2_variance <span class="op">=</span> <span class="fu">principalvars</span>(PCA_wholeGenome.model)[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>]</span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>PC1_prop_variance <span class="op">=</span> PC1_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>PC2_prop_variance <span class="op">=</span> PC2_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"PC1 explains "</span>, <span class="fl">100</span><span class="op">*</span>PC1_prop_variance, <span class="st">"% of the total variance.</span></span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>PC2 explains <span class="st">", 100*PC2_prop_variance, "</span><span class="op">%</span>.<span class="st">")</span></span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a><span class="co"># add position of reference genome</span></span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a>refGenomePCAposition <span class="op">=</span> <span class="fu">predict</span>(PCA_wholeGenome.model, <span class="fu">zeros</span>(<span class="fu">size</span>(genos_imputed_loaded, <span class="fl">2</span>)))</span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">1</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># this flips PC1 if flipPC1 = true</span></span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">&amp;&amp;</span> (refGenomePCAposition[<span class="fl">2</span>] <span class="op">*=</span> <span class="op">-</span><span class="fl">1</span>)  <span class="co"># same for PC2</span></span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>CairoMakie.<span class="fu">scatter!</span>(refGenomePCAposition[<span class="fl">1</span>], refGenomePCAposition[<span class="fl">2</span>], marker <span class="op">=</span> <span class="op">:</span>diamond, color<span class="op">=</span><span class="st">"black"</span>, markersize<span class="op">=</span><span class="fl">15</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a>try</span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(PCA_wholeGenome.PCAfig)</span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>catch</span>
<span id="cb85-24"><a href="#cb85-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"NOTICE: Figure for "</span>, regionText, <span class="st">" could not be shown due to an unknown error."</span>)</span>
<span id="cb85-25"><a href="#cb85-25" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-49-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>PC1 explains 9.776807% of the total variance.
PC2 explains 1.8531415%.</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-49-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display" data-execution_count="49">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="make-a-whole-genome-pca-just-for-the-western-side-of-the-ring" class="level4">
<h4 class="anchored" data-anchor-id="make-a-whole-genome-pca-just-for-the-western-side-of-the-ring">Make a whole-genome PCA just for the western side of the ring</h4>
<div class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>western_groups_to_plot_PCA <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"vir_misID"</span>, <span class="st">"vir_S"</span>, <span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>, <span class="st">"troch_west"</span>, <span class="st">"troch_LN"</span>]</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>western_group_colors_PCA <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"blue"</span>, <span class="st">"turquoise1"</span>, <span class="st">"grey"</span>, <span class="st">"seagreen4"</span>, <span class="st">"seagreen3"</span>, <span class="st">"seagreen2"</span>, <span class="st">"olivedrab3"</span>, <span class="st">"olivedrab2"</span>, <span class="st">"olivedrab1"</span>, <span class="st">"yellow"</span>]</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>PCA_wholeGenome <span class="op">=</span> <span class="fu">plotPCA</span>(genos_imputed_loaded, ind_with_metadata_indFiltered,</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>    western_groups_to_plot_PCA, western_group_colors_PCA;</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>    sampleSet<span class="op">=</span><span class="st">"greenish warblers"</span>, regionText<span class="op">=</span><span class="st">"wholeGenome"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>    flip1<span class="op">=</span><span class="cn">true</span>, flip2<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>totalObservationVariance <span class="op">=</span> <span class="fu">var</span>(PCA_wholeGenome.model)</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>PC1_variance, PC2_variance <span class="op">=</span> <span class="fu">principalvars</span>(PCA_wholeGenome.model)[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>]</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>PC1_prop_variance <span class="op">=</span> PC1_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>PC2_prop_variance <span class="op">=</span> PC2_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"PC1 explains "</span>, <span class="fl">100</span> <span class="op">*</span> PC1_prop_variance, <span class="st">"% of the total variance.</span></span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>  PC2 explains <span class="st">", 100 * PC2_prop_variance, "</span><span class="op">%</span>.<span class="st">")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-50-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>PC1 explains 11.392442% of the total variance.
  PC2 explains 1.7764572%.</code></pre>
</div>
</div>
</section>
</section>
<section id="make-gbi-plots-showing-all-individuals" class="level1">
<h1>Make GBI plots showing all individuals</h1>
<p>For Z chromosome:</p>
<div class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"lud_PK"</span>,<span class="st">"troch_LN"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)   </span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> groups_to_plot_PCA <span class="co"># all GW individuals</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> group_colors_PCA</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>groupsToCompare <span class="op">=</span> <span class="st">"Fst_among"</span> <span class="co">#"vir_plumb" </span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>Fst_cutoff <span class="op">=</span> <span class="fl">0.7</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genotypes_gwZ_SNPfiltered, ind_with_metadata_indFiltered_sex.Fst_group, groups)</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gwZ"</span></span>
<span id="cb90-18"><a href="#cb90-18" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered_gwZ_SNPfiltered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb90-19"><a href="#cb90-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-20"><a href="#cb90-20" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff,</span>
<span id="cb90-21"><a href="#cb90-21" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed, regionInfo, </span>
<span id="cb90-22"><a href="#cb90-22" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_gwZ_SNPfiltered, Fst, pairwiseNamesFst, </span>
<span id="cb90-23"><a href="#cb90-23" aria-hidden="true" tabindex="-1"></a>    genotypes_gwZ_SNPfiltered, ind_with_metadata_indFiltered_sex, freqs, plotGroups, plotGroupColors;</span>
<span id="cb90-24"><a href="#cb90-24" aria-hidden="true" tabindex="-1"></a>    indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>));</span>
<span id="cb90-25"><a href="#cb90-25" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-51-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>For chr 28:</p>
<div class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"lud_PK"</span>,<span class="st">"troch_LN"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)   </span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> groups_to_plot_PCA <span class="co"># all GW individuals</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> group_colors_PCA</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>groupsToCompare <span class="op">=</span> <span class="st">"Fst_among"</span> <span class="co">#"vir_plumb" </span></span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>Fst_cutoff <span class="op">=</span> <span class="fl">0.6</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_with_missing, ind_with_metadata_indFiltered.Fst_group, groups)</span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-14"><a href="#cb92-14" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb92-15"><a href="#cb92-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span>
<span id="cb92-16"><a href="#cb92-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-17"><a href="#cb92-17" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw26"</span></span>
<span id="cb92-18"><a href="#cb92-18" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb92-19"><a href="#cb92-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-20"><a href="#cb92-20" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb92-21"><a href="#cb92-21" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed, regionInfo,</span>
<span id="cb92-22"><a href="#cb92-22" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb92-23"><a href="#cb92-23" aria-hidden="true" tabindex="-1"></a>    genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb92-24"><a href="#cb92-24" aria-hidden="true" tabindex="-1"></a>    plotGroups, plotGroupColors;</span>
<span id="cb92-25"><a href="#cb92-25" aria-hidden="true" tabindex="-1"></a>    indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>));</span>
<span id="cb92-26"><a href="#cb92-26" aria-hidden="true" tabindex="-1"></a><span class="co"># plotInfo contains a tuple with: (f, plottedGenotype, locations, plottedMetadata)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-52-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>For chr 28:</p>
<div class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw28"</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr])</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed, regionInfo,</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>    plotGroups, plotGroupColors;</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>    indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>));</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-53-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>For chr 23:</p>
<div class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw23"</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr])</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed, regionInfo,</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>    plotGroups, plotGroupColors;</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>    indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>));</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-54-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>For chr 25:</p>
<div class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw25"</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr])</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>    missingFractionAllowed, regionInfo,</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>    plotGroups, plotGroupColors;</span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>    indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>));</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="GreenishWarblerGenomics2023_files/figure-html/cell-55-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="make-a-table-of-locations-and-groups-of-individuals-included-after-filtering" class="level1">
<h1>Make a table of locations and groups of individuals included after filtering</h1>
<div class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>gdf <span class="op">=</span> <span class="fu">groupby</span>(ind_with_metadata_indFiltered, [<span class="op">:</span>location, <span class="op">:</span>Fst_group])</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>sample_origin_summary <span class="op">=</span> <span class="fu">combine</span>(gdf, nrow)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(sample_origin_summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>34×3 DataFrame
 Row │ location  Fst_group    nrow  
     │ String7   String15     Int64 
─────┼──────────────────────────────
   1 │ AB        vir              2
   2 │ ST        plumb           33
   3 │ ST        plumb_vir        1
   4 │ ST_vi     vir              3
   5 │ KK        plumb           21
   6 │ KK_vi     vir              4
   7 │ BJ        plumb_BJ         3
   8 │ TL        vir             11
   9 │ MN        troch_west      10
  10 │ SU        lud_central      6
  11 │ TH        lud_central      7
  12 │ SR        lud_central     18
  13 │ NG        lud_central      9
  14 │ SP        troch_west      10
  15 │ SA        lud_Sath         8
  16 │ UY        plumb            6
  17 │ LN        troch_LN        14
  18 │ AA        vir_S            8
  19 │ AN        plumb            2
  20 │ BK        plumb            2
  21 │ XN        obs              4
  22 │ EM        troch_EM         1
  23 │ IL        plumb            2
  24 │ OV        lud_KS           2
  25 │ NR        lud_PK           2
  26 │ KL        lud_central      3
  27 │ ML        lud_ML          44
  28 │ PA        lud_central      2
  29 │ PK        lud_PK           4
  30 │ SL        plumb            2
  31 │ STvi      vir              3
  32 │ TA        plumb            1
  33 │ TU        nit              2
  34 │ YK        vir              7</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>