<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Darren Irwin">
<meta name="dcterms.date" content="2025-01-05">

<title>Greenish Warbler heterozygosity variance analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="GW_Heterozygosity_Variance_files/libs/clipboard/clipboard.min.js"></script>
<script src="GW_Heterozygosity_Variance_files/libs/quarto-html/quarto.js"></script>
<script src="GW_Heterozygosity_Variance_files/libs/quarto-html/popper.min.js"></script>
<script src="GW_Heterozygosity_Variance_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="GW_Heterozygosity_Variance_files/libs/quarto-html/anchor.min.js"></script>
<link href="GW_Heterozygosity_Variance_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="GW_Heterozygosity_Variance_files/libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="GW_Heterozygosity_Variance_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="GW_Heterozygosity_Variance_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="GW_Heterozygosity_Variance_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="GW_Heterozygosity_Variance.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Greenish Warbler heterozygosity variance analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Darren Irwin </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 5, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>This script depends on <code>GreenishWarblerGenomics2023.qmd</code> and <code>GW_Zchromosome_analysis.qmd</code></p>
<section id="load-packages" class="level2">
<h2 class="anchored" data-anchor-id="load-packages">Load packages</h2>
<div id="09bbe5ae" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">JLD2</span> <span class="co"># for loading saved data</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DataFrames</span> <span class="co"># for storing data as type DataFrame</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">GenomicDiversity</span> <span class="co"># for plotPCA (my custom package)</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CairoMakie</span> <span class="co"># for plots</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Impute</span> <span class="co"># for imputing missing genotypes</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Statistics</span> <span class="co"># for var() function</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">MultivariateStats</span> <span class="co"># for getting variances from PCA model</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV</span> <span class="co"># for reading in delimited files</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DelimitedFiles</span> <span class="co"># for reading delimited files (the genotypic data)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="choose-working-directory" class="level2">
<h2 class="anchored" data-anchor-id="choose-working-directory">Choose working directory</h2>
<div id="54269c52" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dataDirectory <span class="op">=</span> <span class="st">"/Users/darrenirwin/Dropbox/Darren's current work/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-the-filtered-dataset" class="level2">
<h2 class="anchored" data-anchor-id="load-the-filtered-dataset">Load the filtered dataset</h2>
<p>This dataset was produced through filtering in GreenishWarblerGenomics2023.qmd</p>
<div id="89466f98" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>baseName <span class="op">=</span> <span class="st">"GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>tagName <span class="op">=</span> <span class="st">".Jan2025."</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, <span class="st">"ind_SNP_ind_filtered.jld2"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load info into a dictionary:</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>d <span class="op">=</span> <span class="fu">load</span>(filename)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> baseName <span class="op">!=</span> d[<span class="st">"baseName"</span>]</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: baseNames don't match between that defined above and in the saved file"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> tagName <span class="op">!=</span> d[<span class="st">"tagName"</span>]</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: tagNames don't match don't match between that defined above and in the saved file"</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>genosOnly <span class="op">=</span> d[<span class="st">"genosOnly_ind_SNP_ind_filtered"</span>]</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered <span class="op">=</span> d[<span class="st">"ind_with_metadata_indFiltered"</span>]</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered <span class="op">=</span> d[<span class="st">"pos_SNP_filtered"</span>]</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>repoDirectory <span class="op">=</span> d[<span class="st">"repoDirectory"</span>]</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>dataDirectory <span class="op">=</span> d[<span class="st">"dataDirectory"</span>]</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>scaffold_info <span class="op">=</span> d[<span class="st">"scaffold_info"</span>]</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>scaffold_lengths <span class="op">=</span> d[<span class="st">"scaffold_lengths"</span>]</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>filenameTextMiddle <span class="op">=</span> d[<span class="st">"filenameTextMiddle"</span>]</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>missingGenotypeThreshold <span class="op">=</span> d[<span class="st">"missingGenotypeThreshold"</span>]</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>filenameTextEnd <span class="op">=</span> d[<span class="st">"filenameTextEnd"</span>]</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>chromosomes_to_process <span class="op">=</span>d[<span class="st">"chromosomes_to_process"</span>]</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>metadataFile <span class="op">=</span> d[<span class="st">"metadataFile"</span>]</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Loaded the filtered data."</span>)</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co">#= filename = string(baseName, tagName, "ind_SNP_ind_filtered.jld2")</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co">genosOnly = load(filename, "genosOnly_ind_SNP_ind_filtered")</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="co">ind_with_metadata_indFiltered = load(filename, "ind_with_metadata_indFiltered")</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co">pos_SNP_filtered = load(filename, "pos_SNP_filtered")</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co">println("Loaded the filtered data.") =#</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded the filtered data.</code></pre>
</div>
</div>
<p>Also define <code>correctNames()</code> function as in main script, to correct some names:</p>
<div id="194a135e" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">correctNames</span>(metadataColumn)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>        metadataColumn_corrected <span class="op">=</span> <span class="fu">replace</span>(metadataColumn, <span class="st">"GW_Armando_plate1_TTGW05_rep2"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW05r2"</span>,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Lane5_NA3-3ul"</span> <span class="op">=&gt;</span> <span class="st">"GW_Lane5_NA3"</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_05"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-05"</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_07"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-07"</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_08"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-08"</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_09"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-09"</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_01"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-01"</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_02"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-02"</span>,   </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_03"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-03"</span>,</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_04"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-04"</span>,</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_06"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-06"</span>,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate1_TTGW_15_10"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate1_TTGW-15-10"</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_01"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-01"</span>,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_02"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-02"</span>,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_03"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-03"</span>,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_04"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-04"</span>,</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_06"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-06"</span>,</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>        <span class="st">"GW_Armando_plate2_TTGW_15_10"</span> <span class="op">=&gt;</span> <span class="st">"GW_Armando_plate2_TTGW-15-10"</span>) </span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>correctNames (generic function with 1 method)</code></pre>
</div>
</div>
</section>
<section id="replace-the-z-chromosome-snps-with-the-filtered-z-chromosome-snps" class="level2">
<h2 class="anchored" data-anchor-id="replace-the-z-chromosome-snps-with-the-filtered-z-chromosome-snps">Replace the Z chromosome SNPs with the filtered Z chromosome SNPs</h2>
<div id="61a0ca86" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the Z SNPs from the big dataset loaded above:</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.!=</span> <span class="st">"gwZ"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered <span class="op">=</span> pos_SNP_filtered[selection, <span class="op">:</span>]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>genosOnly <span class="op">=</span> genosOnly[<span class="op">:</span>,selection]</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># load and add the Z filtered SNPs:</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, <span class="st">"chrgwZ_cleaned.notImputed.jld2"</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>genosOnly_chrgwZ_cleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"genotypes_gwZ_SNPfiltered"</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered_chrgwZ_cleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered_chrgwZ_cleaned <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_region"</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> ind_with_metadata_indFiltered.ind <span class="op">!=</span> ind_with_metadata_indFiltered_chrgwZ_cleaned.ind</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: the list of individuals in the big file and Z file are not completely identical."</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered <span class="op">=</span> <span class="fu">vcat</span>(pos_SNP_filtered, pos_SNP_filtered_chrgwZ_cleaned)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>genosOnly <span class="op">=</span> <span class="fu">hcat</span>(genosOnly, genosOnly_chrgwZ_cleaned)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Replaced the Z chromosome data with the filtered Z data."</span>)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="co"># copy the sex column so we can use later:</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.sex <span class="op">=</span> ind_with_metadata_indFiltered_chrgwZ_cleaned.sex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Replaced the Z chromosome data with the filtered Z data.</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>257-element Vector{String}:
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 "M"
 ⋮
 "M"
 "M"
 "M"
 "M"
 "M"
 "F"
 "M"
 "M"
 "F"
 "M"
 "M"
 "M"</code></pre>
</div>
</div>
</section>
<section id="switch-plot-order-to-ring_km-and-set-nitidus-set-to--2500-km-and-the-siberian-hybrid-set-to-5000-km-plot-order" class="level2">
<h2 class="anchored" data-anchor-id="switch-plot-order-to-ring_km-and-set-nitidus-set-to--2500-km-and-the-siberian-hybrid-set-to-5000-km-plot-order">switch plot order to ring_km, and set nitidus (set to -2500 km) and the Siberian hybrid (set to 5000 km) plot order:</h2>
<div id="0ea9d68e" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.original_plot_order <span class="op">=</span> ind_with_metadata_indFiltered.plot_order</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.plot_order <span class="op">=</span> ind_with_metadata_indFiltered.ring_km</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.plot_order[ind_with_metadata_indFiltered.Fst_group <span class="op">.==</span> <span class="st">"nit"</span>] <span class="op">.=</span> <span class="op">-</span><span class="fl">2500</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_indFiltered.plot_order[ind_with_metadata_indFiltered.Fst_group <span class="op">.==</span> <span class="st">"plumb_vir"</span>] <span class="op">.=</span> <span class="fl">5000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>1-element view(::Vector{Float64}, [8]) with eltype Float64:
 5000.0</code></pre>
</div>
</div>
</section>
<section id="prepare-data-for-genotype-by-individual-plots-and-pca" class="level2">
<h2 class="anchored" data-anchor-id="prepare-data-for-genotype-by-individual-plots-and-pca">Prepare data for Genotype-by-individual plots and PCA</h2>
<p>For missing genotypes, change our code of <code>-1</code> to <code>missing</code>:</p>
<div id="b79b5271" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Int16}}</span>(genosOnly)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>genosOnly_with_missing <span class="op">=</span> <span class="fu">replace</span>(genosOnly_with_missing, <span class="op">-</span><span class="fl">1</span> <span class="op">=&gt;</span> <span class="cn">missing</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co">#genosOnly_with_missing[genosOnly_with_missing .== -1] .= missing;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>257×1015750 Matrix{Union{Missing, Int16}}:
 0  0  0  0  0  1  0  0         0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  1         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 ⋮              ⋮                     ⋱        ⋮              ⋮           
 0  0  0  0  0  0  0  0         0  0  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0   missing  0  0  …  0  0  0  2  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0  …  0  0  0  1  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0         0  0     0  0  0  0  0  0  0  0  0  0  0  0</code></pre>
</div>
</div>
</section>
<section id="make-list-of-scaffolds-to-plot" class="level2">
<h2 class="anchored" data-anchor-id="make-list-of-scaffolds-to-plot">Make list of scaffolds to plot:</h2>
<div id="97e64c5f" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>scaffolds_to_plot <span class="op">=</span> <span class="st">"gw"</span> <span class="op">.*</span> <span class="fu">string</span>.(<span class="fu">vcat</span>(<span class="fl">1</span>, <span class="st">"1A"</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, <span class="st">"4A"</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">17</span><span class="op">:</span><span class="fl">28</span>, <span class="st">"Z"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<pre><code>30-element Vector{String}:
 "gw1"
 "gw1A"
 "gw2"
 "gw3"
 "gw4"
 "gw4A"
 "gw5"
 "gw6"
 "gw7"
 "gw8"
 "gw9"
 "gw10"
 "gw11"
 ⋮
 "gw18"
 "gw19"
 "gw20"
 "gw21"
 "gw22"
 "gw23"
 "gw24"
 "gw25"
 "gw26"
 "gw27"
 "gw28"
 "gwZ"</code></pre>
</div>
</div>
</section>
<section id="determine-number-of-snps-in-the-chromosomes-above" class="level2">
<h2 class="anchored" data-anchor-id="determine-number-of-snps-in-the-chromosomes-above">Determine number of SNPs in the chromosomes above</h2>
<div id="43559de0" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">map</span>(<span class="fu">in</span>(scaffolds_to_plot), pos_SNP_filtered.chrom))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>1003924</code></pre>
</div>
</div>
<p>This reports that 1003924 SNPs are within the listed chromosomes. (Good because that matches the number as determined in GW_PCAplots.qmd)</p>
</section>
<section id="choose-groups-and-colors" class="level2">
<h2 class="anchored" data-anchor-id="choose-groups-and-colors">Choose groups and colors</h2>
<div id="51b16152" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>group_colors_PCA <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>];</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="section" class="level15">
<p class="heading"></p>
</section>
</section>
<section id="show-windowed-heterozygosity-for-individuals" class="level2">
<h2 class="anchored" data-anchor-id="show-windowed-heterozygosity-for-individuals">Show windowed heterozygosity for individuals</h2>
<div id="c9db71f1" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># option to select a subset of individuals</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>filterGroups <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#groupsToInclude = ["vir"]</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">#groupsToInclude = ["troch_LN"]</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#numIndsToPlot = [1000]</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># groupsToInclude = ["vir", "plumb", "plumb_vir"]</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># numIndsToPlot = [1000, 1000, 1000]</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># groupsToInclude = ["troch_LN","troch_EM","obs","plumb_BJ"]</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co"># numIndsToPlot = [1000, 1000, 1000, 1000]</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>groupsToInclude <span class="op">=</span> [<span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>]</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> [<span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>, <span class="fl">1000</span>]</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> filterGroups</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(groupsToInclude, numIndsToPlot, genosOnly_with_missing, ind_with_metadata_indFiltered)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>    genosOnly_included <span class="op">=</span> genosOnly_with_missing</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_included <span class="op">=</span> ind_with_metadata_indFiltered</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw15"</span></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>windowSize <span class="op">=</span> <span class="fl">500</span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chr)</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>pos_region <span class="op">=</span> pos_SNP_filtered[loci_selection, <span class="op">:</span>]</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>genotypes_region <span class="op">=</span> genosOnly_included[<span class="op">:</span>, loci_selection]</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>windowedPos, windowedIndHet <span class="op">=</span> <span class="fu">getWindowedIndHet</span>(genotypes_region, pos_region, windowSize)</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>plotTitle <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Windowed heterozygosity of "</span>, <span class="fu">size</span>(windowedIndHet, <span class="fl">1</span>), <span class="st">" individuals"</span>)</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>titleSize <span class="op">=</span> <span class="fl">24</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Location on scaffold "</span>, chr)</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>yLabelText <span class="op">=</span> <span class="st">"Heterozygosity"</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>labelSize <span class="op">=</span> <span class="fl">24</span></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>plotTitle, titlesize<span class="op">=</span>titleSize,</span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span>xLabelText, xlabelsize<span class="op">=</span>labelSize,</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span>yLabelText, ylabelsize<span class="op">=</span>labelSize)</span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a><span class="fu">lines!</span>(windowedPos, windowedIndHet[<span class="fl">1</span>, <span class="op">:</span>])</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="fu">size</span>(windowedIndHet, <span class="fl">1</span>)</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lines!</span>(windowedPos, windowedIndHet[i, <span class="op">:</span>])</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>meanPerWindow_windowedIndHet <span class="op">=</span> <span class="fu">sum</span>(windowedIndHet, dims<span class="op">=</span><span class="fl">1</span>) <span class="op">./</span> <span class="fu">size</span>(windowedIndHet, <span class="fl">1</span>)</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a><span class="fu">lines!</span>(windowedPos, <span class="fu">vec</span>(meanPerWindow_windowedIndHet), linewidth<span class="op">=</span><span class="fl">10</span>, color<span class="op">=:</span>red)</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>windowedIndHet_standardized <span class="op">=</span> <span class="fu">standardizeIndHet</span>(windowedIndHet)</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>plotTitle <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Standardized heterozygosity of "</span>, <span class="fu">size</span>(windowedIndHet, <span class="fl">1</span>), <span class="st">" individuals"</span>)</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>titleSize <span class="op">=</span> <span class="fl">24</span></span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a>xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Location on scaffold "</span>, chr)</span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>yLabelText <span class="op">=</span> <span class="st">"Standardized heterozygosity"</span></span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>labelSize <span class="op">=</span> <span class="fl">24</span></span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(g[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>plotTitle, titlesize<span class="op">=</span>titleSize,</span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span>xLabelText, xlabelsize<span class="op">=</span>labelSize,</span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span>yLabelText, ylabelsize<span class="op">=</span>labelSize)</span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">size</span>(windowedIndHet_standardized, <span class="fl">1</span>)</span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lines!</span>(windowedPos, windowedIndHet_standardized[i, <span class="op">:</span>])</span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a><span class="co">#lines!(windowedPos, windowedIndHet[21,:], linewidth = 10, color = :red)</span></span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(g)</span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Now graph the variance in standardized heterozygosity</span></span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a>varWindowedIndHet <span class="op">=</span> <span class="fu">getVarWindowedIndHet</span>(windowedIndHet_standardized)</span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a>plotTitle <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Variance in standardized het. among "</span>, <span class="fu">size</span>(windowedIndHet, <span class="fl">1</span>), <span class="st">" individuals"</span>)</span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a>titleSize <span class="op">=</span> <span class="fl">18</span></span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a>xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Location on scaffold "</span>, chr)</span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a>yLabelText <span class="op">=</span> <span class="st">"Heterozygosity variance (standardized)"</span></span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a>labelSize <span class="op">=</span> <span class="fl">18</span></span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a>h <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(h[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>plotTitle, titlesize<span class="op">=</span>titleSize,</span>
<span id="cb19-79"><a href="#cb19-79" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span>xLabelText, xlabelsize<span class="op">=</span>labelSize,</span>
<span id="cb19-80"><a href="#cb19-80" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span>yLabelText, ylabelsize<span class="op">=</span>labelSize)</span>
<span id="cb19-81"><a href="#cb19-81" aria-hidden="true" tabindex="-1"></a><span class="fu">lines!</span>(windowedPos, varWindowedIndHet)</span>
<span id="cb19-82"><a href="#cb19-82" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(h)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-12-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-12-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-12-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="make-heterozygosity-variance-plot-for-whole-genome" class="level2">
<h2 class="anchored" data-anchor-id="make-heterozygosity-variance-plot-for-whole-genome">Make heterozygosity variance plot for whole genome:</h2>
<div id="9bdcf1a9" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># option to select a subset of individuals</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>filterGroups <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># groupsToInclude = ["vir", "plumb", "plumb_vir"]</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># numIndsToPlot = [1000, 1000, 1000]</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co"># groupsToInclude = ["lud_PK", "lud_KS", "lud_central", "lud_Sath", "lud_ML","troch_west","troch_LN"]</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co"># numIndsToPlot = [1000, 1000, 1000, 1000, 1000, 1000, 1000]</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co"># groupsToInclude = ["troch_LN","troch_EM","obs","plumb_BJ"]</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># numIndsToPlot = [1000, 1000, 1000, 1000]</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>groupsToInclude <span class="op">=</span> [<span class="st">"vir"</span>]</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co">#groupsToInclude = ["troch_LN"]</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> [<span class="fl">1000</span>]</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> filterGroups</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    genosOnly_included, ind_with_metadata_included <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(groupsToInclude, numIndsToPlot, genosOnly_with_missing, ind_with_metadata_indFiltered)</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>    genosOnly_included <span class="op">=</span> genosOnly_with_missing</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>    ind_with_metadata_included <span class="op">=</span> ind_with_metadata_indFiltered</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="co">#scaffolds_for_HetVariance = ["gw26", "gw28"]</span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>scaffolds_for_HetVariance <span class="op">=</span> scaffolds_to_plot</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize data structures</span></span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>windowed_pos_all <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="dt">String</span>[], mean_position <span class="op">=</span> <span class="dt">Float64</span>[], first_position <span class="op">=</span> <span class="dt">Int</span>[], last_position <span class="op">=</span> <span class="dt">Int</span>[])</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>windowed_HetVariance_all <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{Float32}</span>(<span class="cn">undef</span>, <span class="fl">0</span>)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrom <span class="kw">in</span> scaffolds_for_HetVariance</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chrom)</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>    pos_region <span class="op">=</span> pos_SNP_filtered[loci_selection, <span class="op">:</span>]</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>    genotypes_region <span class="op">=</span> genosOnly_included[<span class="op">:</span>, loci_selection]</span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> chrom <span class="op">==</span> <span class="st">"gwZ"</span> <span class="co">#include only males (because females have only one Z)</span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>        genotypes_region_males <span class="op">=</span> genotypes_region[ind_with_metadata_indFiltered_chrgwZ_cleaned.sex <span class="op">.==</span> <span class="st">"M"</span>, <span class="op">:</span>]</span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>        windowedPos, windowedIndHet <span class="op">=</span> <span class="fu">getWindowedIndHet</span>(genotypes_region_males, pos_region, windowSize)</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span> <span class="co"># for all other chromosomes, include all individuals</span></span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>        windowedPos, windowedIndHet <span class="op">=</span> <span class="fu">getWindowedIndHet</span>(genotypes_region, pos_region, windowSize)</span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>    windowBoundaries <span class="op">=</span> <span class="fu">getWindowBoundaries</span>(pos_region.position, windowSize)</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a>    windowedIndHet_standardized <span class="op">=</span> <span class="fu">standardizeIndHet</span>(windowedIndHet)</span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a>    windowed_HetVariance_scaffold <span class="op">=</span> <span class="fu">getVarWindowedIndHet</span>(windowedIndHet_standardized)</span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a>    windowed_pos_chrom <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="fu">repeat</span>([chrom], <span class="fu">length</span>(windowedPos)), mean_position <span class="op">=</span> windowedPos, first_position <span class="op">=</span> windowBoundaries[<span class="op">:</span>,<span class="fl">1</span>], last_position <span class="op">=</span> windowBoundaries[<span class="op">:</span>,<span class="fl">2</span>])</span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a>    windowed_pos_all <span class="op">=</span> <span class="fu">vcat</span>(windowed_pos_all, windowed_pos_chrom)</span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a>    windowed_HetVariance_all <span class="op">=</span> [windowed_HetVariance_all; windowed_HetVariance_scaffold]</span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Quick test plot (overlapping all chromosomes, so not good)</span></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(windowed_pos_all.mean_position, windowed_HetVariance_all)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="identify-haploblock-regions-as-those-that-have-high-hetvar" class="level2">
<h2 class="anchored" data-anchor-id="identify-haploblock-regions-as-those-that-have-high-hetvar">Identify “haploblock regions” as those that have high HetVar</h2>
<div id="a54e3d4e" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>threshold_HetVar <span class="op">=</span> <span class="fl">0.4</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> windowed_HetVariance_all <span class="op">.&gt;=</span> threshold_HetVar</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>windowed_pos_all.high_HetVar <span class="op">=</span> selection <span class="co"># adds true/false column to dataframe indicating high HetVar windows</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Make list of contiguous high HetVar region:</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>highHetVarRegions <span class="op">=</span> <span class="fu">DataFrame</span>(regionChrom <span class="op">=</span> <span class="dt">String</span>[], regionStart <span class="op">=</span> <span class="dt">Int</span>[], regionEnd <span class="op">=</span> <span class="dt">Int</span>[])</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>i <span class="op">=</span> <span class="fl">1</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>lastWindow <span class="op">=</span> <span class="fu">nrow</span>(windowed_pos_all)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> i <span class="op">&lt;=</span> lastWindow <span class="co"># eachindex(windowed_pos_all[:,1])</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> windowed_pos_all.high_HetVar[i] <span class="op">==</span> <span class="cn">true</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>        regionChrom <span class="op">=</span> windowed_pos_all.chrom[i]</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>        regionStart <span class="op">=</span> windowed_pos_all.first_position[i]</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>        regionEnd <span class="op">=</span> windowed_pos_all.last_position[i]</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># check whether contiguous with next</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>        next <span class="op">=</span> <span class="fl">1</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">while</span> i <span class="op">+</span> next <span class="op">&lt;=</span> lastWindow <span class="op">&amp;&amp;</span> windowed_pos_all.chrom[i <span class="op">+</span> next] <span class="op">==</span> regionChrom</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> windowed_pos_all.high_HetVar[i <span class="op">+</span> next] <span class="op">==</span> <span class="cn">true</span></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>                regionEnd <span class="op">=</span> windowed_pos_all.last_position[i <span class="op">+</span> next]</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>                next <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>                <span class="cf">break</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>        highHetVarRegions <span class="op">=</span> <span class="fu">push!</span>(highHetVarRegions, [regionChrom, regionStart, regionEnd])</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>        i <span class="op">=</span> i <span class="op">+</span> next <span class="op">+</span> <span class="fl">1</span></span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>        i <span class="op">=</span> i <span class="op">+</span> <span class="fl">1</span></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span> </span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>    i</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>highHetVarRegions</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<div><div style="float: left;"><span>39×3 DataFrame</span></div><div style="float: right;"><span style="font-style: italic;">14 rows omitted</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw1</td>
<td style="text-align: right;">15689747</td>
<td style="text-align: right;">23478124</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw1A</td>
<td style="text-align: right;">4674</td>
<td style="text-align: right;">3771263</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">3</td>
<td style="text-align: left;">gw1A</td>
<td style="text-align: right;">23592559</td>
<td style="text-align: right;">30616953</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">4</td>
<td style="text-align: left;">gw2</td>
<td style="text-align: right;">54537375</td>
<td style="text-align: right;">59262130</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">5</td>
<td style="text-align: left;">gw2</td>
<td style="text-align: right;">60234161</td>
<td style="text-align: right;">61533451</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">6</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">101192949</td>
<td style="text-align: right;">103495514</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">7</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">104554714</td>
<td style="text-align: right;">108279595</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">8</td>
<td style="text-align: left;">gw4</td>
<td style="text-align: right;">5295912</td>
<td style="text-align: right;">5438270</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">9</td>
<td style="text-align: left;">gw4</td>
<td style="text-align: right;">14837641</td>
<td style="text-align: right;">16117455</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">10</td>
<td style="text-align: left;">gw4</td>
<td style="text-align: right;">20930552</td>
<td style="text-align: right;">23610800</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">11</td>
<td style="text-align: left;">gw4A</td>
<td style="text-align: right;">379058</td>
<td style="text-align: right;">730094</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">12</td>
<td style="text-align: left;">gw5</td>
<td style="text-align: right;">10095304</td>
<td style="text-align: right;">10956815</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">13</td>
<td style="text-align: left;">gw6</td>
<td style="text-align: right;">34584054</td>
<td style="text-align: right;">35259663</td>
</tr>
<tr class="even">
<td style="text-align: right;">⋮</td>
<td style="text-align: right;">⋮</td>
<td style="text-align: right;">⋮</td>
<td style="text-align: right;">⋮</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">28</td>
<td style="text-align: left;">gw19</td>
<td style="text-align: right;">43362</td>
<td style="text-align: right;">1006242</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">29</td>
<td style="text-align: left;">gw20</td>
<td style="text-align: right;">27354</td>
<td style="text-align: right;">721651</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">30</td>
<td style="text-align: left;">gw20</td>
<td style="text-align: right;">5852254</td>
<td style="text-align: right;">6671670</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">31</td>
<td style="text-align: left;">gw21</td>
<td style="text-align: right;">3275121</td>
<td style="text-align: right;">3731689</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">32</td>
<td style="text-align: left;">gw22</td>
<td style="text-align: right;">5214430</td>
<td style="text-align: right;">5775824</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">33</td>
<td style="text-align: left;">gw23</td>
<td style="text-align: right;">4135459</td>
<td style="text-align: right;">4774426</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">34</td>
<td style="text-align: left;">gw24</td>
<td style="text-align: right;">3468239</td>
<td style="text-align: right;">4001782</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">35</td>
<td style="text-align: left;">gw25</td>
<td style="text-align: right;">5185626</td>
<td style="text-align: right;">5473966</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">36</td>
<td style="text-align: left;">gw26</td>
<td style="text-align: right;">4153299</td>
<td style="text-align: right;">5549635</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">37</td>
<td style="text-align: left;">gw27</td>
<td style="text-align: right;">41621</td>
<td style="text-align: right;">541081</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">38</td>
<td style="text-align: left;">gw28</td>
<td style="text-align: right;">1822776</td>
<td style="text-align: right;">2522648</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">39</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">68372986</td>
<td style="text-align: right;">73749599</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section>
<section id="determine-fraction-of-genome-that-has-high-vishet" class="level2">
<h2 class="anchored" data-anchor-id="determine-fraction-of-genome-that-has-high-vishet">Determine fraction of genome that has high ViSHet</h2>
<div id="b1c35610" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get total length of high ViSHet regions </span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>sum_highHetVarRegions <span class="op">=</span> <span class="fu">sum</span>(highHetVarRegions.regionEnd <span class="op">.-</span> highHetVarRegions.regionStart)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># get total lengths of all scaffolds</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>sum_scaffold_lengths <span class="op">=</span> <span class="fl">0</span> </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> scaffold_name <span class="kw">in</span> scaffolds_to_plot</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(scaffold_name)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    sum_scaffold_lengths <span class="op">+=</span> scaffold_lengths[scaffold_name]</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>sum_scaffold_lengths</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate percent of genome in high ViSHet regions</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>percentGenomeHighViSHet <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> sum_highHetVarRegions <span class="op">/</span> sum_scaffold_lengths</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The percent of the genome in high ViSHet regions is </span><span class="sc">$</span>percentGenomeHighViSHet<span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>gw1
gw1A
gw2
gw3
gw4
gw4A
gw5
gw6
gw7
gw8
gw9
gw10
gw11
gw12
gw13
gw14
gw15
gw17
gw18
gw19
gw20
gw21
gw22
gw23
gw24
gw25
gw26
gw27
gw28
gwZ
The percent of the genome in high ViSHet regions is 5.799030328440733</code></pre>
</div>
</div>
</section>
<section id="plot-haploblock-regions-for-whole-genome" class="level2">
<h2 class="anchored" data-anchor-id="plot-haploblock-regions-for-whole-genome">Plot “haploblock regions” for whole genome</h2>
<div id="554a77d2" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">plotGenomeHetVariance</span>(scaffolds_to_plot, </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                                windowed_HetVariance_all,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                                windowed_pos_all;</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                                fillColor <span class="op">=</span> <span class="st">"purple"</span>,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>                                lineTransparency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>                                fillTransparency <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>                                figureSize <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">1200</span>),</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>                                plotRegions <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>                                regionsToPlot <span class="op">=</span> [],</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>                                regionColor <span class="op">=</span> <span class="st">"red"</span>)</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get sizes of chromosomes (note this actually isn't the true length, just the mean position of the rightmost window):</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    scaffoldLengths <span class="op">=</span> <span class="fu">repeat</span>([<span class="op">-</span><span class="fl">1</span>], <span class="fu">length</span>(scaffolds_to_plot))   <span class="co"># vector of Int64</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(scaffolds_to_plot)</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>        selection <span class="op">=</span> windowed_pos_all.chrom <span class="op">.==</span> scaffolds_to_plot[i]</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">sum</span>(selection) <span class="op">==</span> <span class="fl">0</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>            scaffoldLengths[i] <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>            scaffoldLengths[i] <span class="op">=</span> <span class="fu">round</span>(<span class="dt">Int</span>, <span class="fu">maximum</span>(windowed_pos_all.mean_position[selection]))</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># my plan is to use the layout feature of Makie to make one axis per row, and then plot</span></span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># multiple scaffolds to each row</span></span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    scaffoldPlottedAlready <span class="op">=</span> <span class="fu">falses</span>(<span class="fu">length</span>(scaffolds_to_plot))</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>    bpPerRow <span class="op">=</span> <span class="fl">200_000_000</span></span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>    bpGapBetweenScaffolds <span class="op">=</span> <span class="fl">10_000_000</span></span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>    scaffoldInRow <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{Vector{String}}</span>() <span class="co"># declare a vector of vectors to store the chromosome names per row</span></span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>    scaffoldBpStartInRow <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{Vector{Int}}</span>()</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>    row <span class="op">=</span> <span class="fl">1</span></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> <span class="fu">sum</span>(scaffoldPlottedAlready <span class="op">.==</span> <span class="cn">false</span>) <span class="op">&gt;</span> <span class="fl">0</span>  <span class="co"># repeat until all chr have a row to be plotted in</span></span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>        bpStart <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>        remainingRowLength <span class="op">=</span> bpPerRow</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">push!</span>(scaffoldInRow, []) <span class="co"># initialize an empty vector for the next row</span></span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">push!</span>(scaffoldBpStartInRow, []) <span class="co"># initialize an empty vector for the next row</span></span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>        placeInRow <span class="op">=</span> <span class="fl">1</span></span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>        <span class="co"># look through chromosomes and find some to go in this row</span></span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(scaffolds_to_plot)</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>            <span class="co"># if not plotted and short enough, add to row:</span></span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> scaffoldPlottedAlready[i] <span class="op">.==</span> <span class="cn">false</span> <span class="op">.&amp;&amp;</span> scaffoldLengths[i] <span class="op">.&lt;=</span> remainingRowLength</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>                <span class="fu">push!</span>(scaffoldInRow[row], scaffolds_to_plot[i])</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>                <span class="fu">push!</span>(scaffoldBpStartInRow[row], bpStart)</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>                remainingRowLength <span class="op">=</span> remainingRowLength <span class="op">-</span> scaffoldLengths[i] <span class="op">-</span> bpGapBetweenScaffolds</span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>                scaffoldPlottedAlready[i] <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>                placeInRow <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a>                bpStart <span class="op">=</span> bpStart <span class="op">+</span> scaffoldLengths[i] <span class="op">+</span> bpGapBetweenScaffolds</span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span> </span>
<span id="cb28-50"><a href="#cb28-50" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb28-51"><a href="#cb28-51" aria-hidden="true" tabindex="-1"></a>        row <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb28-52"><a href="#cb28-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb28-53"><a href="#cb28-53" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(scaffoldInRow)</span>
<span id="cb28-54"><a href="#cb28-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-55"><a href="#cb28-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># the above works. I have the scaffolds arranged for plotting.</span></span>
<span id="cb28-56"><a href="#cb28-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-57"><a href="#cb28-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot in the order defined above</span></span>
<span id="cb28-58"><a href="#cb28-58" aria-hidden="true" tabindex="-1"></a>    f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>(size<span class="op">=</span>figureSize)</span>
<span id="cb28-59"><a href="#cb28-59" aria-hidden="true" tabindex="-1"></a>    numRows <span class="op">=</span> <span class="fu">length</span>(scaffoldInRow)</span>
<span id="cb28-60"><a href="#cb28-60" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make one axis object per row:</span></span>
<span id="cb28-61"><a href="#cb28-61" aria-hidden="true" tabindex="-1"></a>    axs <span class="op">=</span> [<span class="fu">Axis</span>(f[i, <span class="fl">1</span>], limits<span class="op">=</span>(<span class="fl">0</span>, bpPerRow,</span>
<span id="cb28-62"><a href="#cb28-62" aria-hidden="true" tabindex="-1"></a>            <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">2.0</span>), ylabel <span class="op">=</span> <span class="st">"ViSHet"</span>) for i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRows]</span>
<span id="cb28-63"><a href="#cb28-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRows</span>
<span id="cb28-64"><a href="#cb28-64" aria-hidden="true" tabindex="-1"></a>        <span class="fu">hidexdecorations!</span>(axs[i]) <span class="co">#hide background lattice and axis labels</span></span>
<span id="cb28-65"><a href="#cb28-65" aria-hidden="true" tabindex="-1"></a>        <span class="fu">hidespines!</span>(axs[i], <span class="op">:</span>t, <span class="op">:</span>r, <span class="op">:</span>b) <span class="co"># hide box around plot</span></span>
<span id="cb28-66"><a href="#cb28-66" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb28-67"><a href="#cb28-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-68"><a href="#cb28-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> rowNum <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRows</span>
<span id="cb28-69"><a href="#cb28-69" aria-hidden="true" tabindex="-1"></a>        numScaffoldsInRow <span class="op">=</span> <span class="fu">length</span>(scaffoldInRow[rowNum])</span>
<span id="cb28-70"><a href="#cb28-70" aria-hidden="true" tabindex="-1"></a>        nextPlotShiftBP <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb28-71"><a href="#cb28-71" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> orderNum <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numScaffoldsInRow </span>
<span id="cb28-72"><a href="#cb28-72" aria-hidden="true" tabindex="-1"></a>            scaffoldName <span class="op">=</span> scaffoldInRow[rowNum][orderNum]</span>
<span id="cb28-73"><a href="#cb28-73" aria-hidden="true" tabindex="-1"></a>            plotBpLength <span class="op">=</span> scaffoldLengths[<span class="fu">findfirst</span>(scaffolds_to_plot <span class="op">.==</span> scaffoldName)]</span>
<span id="cb28-74"><a href="#cb28-74" aria-hidden="true" tabindex="-1"></a>            <span class="co"># select one scaffold from the whole genome:</span></span>
<span id="cb28-75"><a href="#cb28-75" aria-hidden="true" tabindex="-1"></a>            selection <span class="op">=</span> windowed_pos_all.chrom <span class="op">.==</span> scaffoldName </span>
<span id="cb28-76"><a href="#cb28-76" aria-hidden="true" tabindex="-1"></a>            windowed_pos_selected <span class="op">=</span> <span class="fu">view</span>(windowed_pos_all.mean_position, selection)</span>
<span id="cb28-77"><a href="#cb28-77" aria-hidden="true" tabindex="-1"></a>            windowed_HetVariance_selected <span class="op">=</span> <span class="fu">view</span>(windowed_HetVariance_all, selection)</span>
<span id="cb28-78"><a href="#cb28-78" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(windowed_pos_selected) <span class="op">&gt;</span> <span class="fl">0</span>  <span class="co"># only plot if there is data for that scaffold        </span></span>
<span id="cb28-79"><a href="#cb28-79" aria-hidden="true" tabindex="-1"></a>                <span class="fu">text!</span>(axs[rowNum], scaffoldBpStartInRow[rowNum][orderNum] <span class="op">+</span> <span class="fl">0.005</span>bpPerRow, <span class="op">-</span><span class="fl">0.2</span>; text<span class="op">=</span>scaffoldName, align<span class="op">=</span>(<span class="op">:</span>left, <span class="op">:</span>center), fontsize<span class="op">=</span><span class="fl">20</span>)</span>
<span id="cb28-80"><a href="#cb28-80" aria-hidden="true" tabindex="-1"></a>                xx <span class="op">=</span> <span class="fu">vcat</span>(<span class="fu">first</span>(windowed_pos_selected), windowed_pos_selected, <span class="fu">last</span>(windowed_pos_selected)) <span class="op">.+</span> scaffoldBpStartInRow[rowNum][orderNum]</span>
<span id="cb28-81"><a href="#cb28-81" aria-hidden="true" tabindex="-1"></a>                yy <span class="op">=</span> <span class="fu">vcat</span>(<span class="fl">0</span>, windowed_HetVariance_selected, <span class="fl">0</span>)</span>
<span id="cb28-82"><a href="#cb28-82" aria-hidden="true" tabindex="-1"></a>                <span class="fu">poly!</span>(axs[rowNum], <span class="fu">Point2f</span>.(xx, yy), color <span class="op">=</span> (fillColor, fillTransparency), strokecolor <span class="op">=</span> (fillColor, lineTransparency), strokewidth <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb28-83"><a href="#cb28-83" aria-hidden="true" tabindex="-1"></a>                <span class="co"># if indicating regions (e.g. that have high HetVar)</span></span>
<span id="cb28-84"><a href="#cb28-84" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> plotRegions</span>
<span id="cb28-85"><a href="#cb28-85" aria-hidden="true" tabindex="-1"></a>                    regionsToPlot_thisScaffold <span class="op">=</span> regionsToPlot[regionsToPlot.regionChrom <span class="op">.==</span> scaffoldName, <span class="op">:</span>]</span>
<span id="cb28-86"><a href="#cb28-86" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(regionsToPlot_thisScaffold.regionStart)</span>
<span id="cb28-87"><a href="#cb28-87" aria-hidden="true" tabindex="-1"></a>                        xValues <span class="op">=</span> [regionsToPlot_thisScaffold.regionStart[i], regionsToPlot_thisScaffold.regionEnd[i]] <span class="op">.+</span> scaffoldBpStartInRow[rowNum][orderNum]</span>
<span id="cb28-88"><a href="#cb28-88" aria-hidden="true" tabindex="-1"></a>                        yValues <span class="op">=</span> [<span class="fl">0.05</span>, <span class="fl">0.05</span>]</span>
<span id="cb28-89"><a href="#cb28-89" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">lines!</span>(axs[rowNum], xValues, yValues, color<span class="op">=</span>regionColor, linewidth<span class="op">=</span><span class="fl">10</span>)</span>
<span id="cb28-90"><a href="#cb28-90" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb28-91"><a href="#cb28-91" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb28-92"><a href="#cb28-92" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb28-93"><a href="#cb28-93" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb28-94"><a href="#cb28-94" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb28-95"><a href="#cb28-95" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(f)</span>
<span id="cb28-96"><a href="#cb28-96" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> f</span>
<span id="cb28-97"><a href="#cb28-97" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb28-98"><a href="#cb28-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-99"><a href="#cb28-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-100"><a href="#cb28-100" aria-hidden="true" tabindex="-1"></a>fig2 <span class="op">=</span> <span class="fu">plotGenomeHetVariance</span>(scaffolds_to_plot, </span>
<span id="cb28-101"><a href="#cb28-101" aria-hidden="true" tabindex="-1"></a>                        windowed_HetVariance_all,</span>
<span id="cb28-102"><a href="#cb28-102" aria-hidden="true" tabindex="-1"></a>                        windowed_pos_all;</span>
<span id="cb28-103"><a href="#cb28-103" aria-hidden="true" tabindex="-1"></a>                        fillColor <span class="op">=</span> <span class="st">"purple"</span>,</span>
<span id="cb28-104"><a href="#cb28-104" aria-hidden="true" tabindex="-1"></a>                        lineTransparency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span id="cb28-105"><a href="#cb28-105" aria-hidden="true" tabindex="-1"></a>                        fillTransparency <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span id="cb28-106"><a href="#cb28-106" aria-hidden="true" tabindex="-1"></a>                        figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">1200</span>),</span>
<span id="cb28-107"><a href="#cb28-107" aria-hidden="true" tabindex="-1"></a>                        plotRegions <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb28-108"><a href="#cb28-108" aria-hidden="true" tabindex="-1"></a>                        regionsToPlot <span class="op">=</span> highHetVarRegions,</span>
<span id="cb28-109"><a href="#cb28-109" aria-hidden="true" tabindex="-1"></a>                        regionColor <span class="op">=</span> <span class="st">"magenta"</span>)</span>
<span id="cb28-110"><a href="#cb28-110" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb28-111"><a href="#cb28-111" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="st">"Figure2_ViSHet_allGenome_fromJulia.png"</span></span>
<span id="cb28-112"><a href="#cb28-112" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(filename, fig2, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb28-113"><a href="#cb28-113" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Saved "</span>, filename)</span>
<span id="cb28-114"><a href="#cb28-114" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[["gw1", "gw4A", "gw6"], ["gw1A", "gw4", "gw9"], ["gw2", "gw8"], ["gw3", "gw5"], ["gw7", "gw10", "gw11", "gw12", "gw13", "gw14"], ["gw15", "gw17", "gw18", "gw19", "gw20", "gw21", "gw22", "gw23", "gw24", "gw25"], ["gw26", "gw27", "gw28", "gwZ"]]</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-16-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="make-a-pca-based-on-all-non-lhbrs-that-is-outside-of-the-identified-large-haploblock-regions" class="level2">
<h2 class="anchored" data-anchor-id="make-a-pca-based-on-all-non-lhbrs-that-is-outside-of-the-identified-large-haploblock-regions">Make a PCA based on all non-LHBRs (that is, outside of the identified Large HaploBlock Regions):</h2>
<section id="first-remove-the-lhbr-regions" class="level4">
<h4 class="anchored" data-anchor-id="first-remove-the-lhbr-regions">First remove the LHBR regions</h4>
<div id="c009dc4f" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cycle through the LHBRs, determining the SNPs within each to remove from the dataset:</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>lociToRemove <span class="op">=</span> <span class="fu">fill</span>(<span class="cn">false</span>, <span class="fu">nrow</span>(pos_SNP_filtered))</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachrow</span>(highHetVarRegions) </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># display(i.regionChrom)</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># display(i.regionStart)</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    lociWithinThisLHBR <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> i.regionChrom) <span class="op">.&amp;&amp;</span> </span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>                    (pos_SNP_filtered.position <span class="op">.&gt;=</span> i.regionStart) <span class="op">.&amp;&amp;</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>                    (pos_SNP_filtered.position <span class="op">.&lt;=</span> i.regionEnd)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>    lociToRemove <span class="op">=</span> lociToRemove <span class="op">.||</span> lociWithinThisLHBR</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co"># now actually remove them:</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>genosOnly_included_nonLHBR <span class="op">=</span> genosOnly_included[<span class="op">:</span>, .!lociToRemove]</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>pos_SNP_filtered_nonLHBR <span class="op">=</span> pos_SNP_filtered[.!lociToRemove, <span class="op">:</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<div><div style="float: left;"><span>969250×2 DataFrame</span></div><div style="float: right;"><span style="font-style: italic;">969225 rows omitted</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">chrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">position</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058517</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058520</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">3</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058521</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">4</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058526</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">5</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058528</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">6</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058535</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">7</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058551</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">8</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058562</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">9</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058566</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">10</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058573</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">11</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058593</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">12</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058594</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">13</td>
<td style="text-align: left;">gws103</td>
<td style="text-align: right;">3058597</td>
</tr>
<tr class="even">
<td style="text-align: right;">⋮</td>
<td style="text-align: right;">⋮</td>
<td style="text-align: right;">⋮</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969239</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574744</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969240</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574749</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969241</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574762</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969242</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574774</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969243</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574793</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969244</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574795</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969245</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574814</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969246</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574819</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969247</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574831</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969248</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574832</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969249</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574836</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">969250</td>
<td style="text-align: left;">gwZ</td>
<td style="text-align: right;">89574838</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section>
<section id="make-list-of-scaffolds-to-include-in-the-whole-genome-non_lhbr-pca" class="level4">
<h4 class="anchored" data-anchor-id="make-list-of-scaffolds-to-include-in-the-whole-genome-non_lhbr-pca">Make list of scaffolds to include in the whole-genome non_LHBR PCA:</h4>
<div id="69b0eb68" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>chromosomes_to_include <span class="op">=</span> <span class="st">"gw"</span> <span class="op">.*</span> <span class="fu">string</span>.(<span class="fu">vcat</span>(<span class="fl">28</span><span class="op">:-</span><span class="fl">1</span><span class="op">:</span><span class="fl">17</span>, <span class="fl">15</span><span class="op">:-</span><span class="fl">1</span><span class="op">:</span><span class="fl">1</span>))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">push!</span>(chromosomes_to_include, <span class="st">"gw1A"</span>, <span class="st">"gw4A"</span>, <span class="st">"gwZ"</span>)  <span class="co"># add two other scaffolds</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<pre><code>30-element Vector{String}:
 "gw28"
 "gw27"
 "gw26"
 "gw25"
 "gw24"
 "gw23"
 "gw22"
 "gw21"
 "gw20"
 "gw19"
 "gw18"
 "gw17"
 "gw15"
 ⋮
 "gw9"
 "gw8"
 "gw7"
 "gw6"
 "gw5"
 "gw4"
 "gw3"
 "gw2"
 "gw1"
 "gw1A"
 "gw4A"
 "gwZ"</code></pre>
</div>
</div>
</section>
<section id="imputation-using-knn" class="level4">
<h4 class="anchored" data-anchor-id="imputation-using-knn">Imputation using KNN</h4>
<p>Did this on 5Jan2025 and saved files, so inactivated this cell for now (can take up to several minutes for each big scaffold):</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_include)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_include[i]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom, <span class="st">"nonLHBR"</span>) <span class="co"># this is where "nonLHBR will be incorporated in file name"</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos_SNP_filtered_nonLHBR.chrom <span class="op">.==</span> chrom)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_nonLHBR_region <span class="op">=</span> pos_SNP_filtered_nonLHBR[loci_selection,<span class="op">:</span>]</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    genosOnly_nonLHBR_region_for_imputing <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genosOnly_included_nonLHBR[<span class="op">:</span>,loci_selection])</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@time</span> imputed_genos <span class="op">=</span> Impute.<span class="fu">knn</span>(genosOnly_nonLHBR_region_for_imputing; dims <span class="op">=</span> <span class="op">:</span>rows)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">jldsave</span>(filename; imputed_genos, ind_with_metadata_indFiltered, pos_SNP_filtered_nonLHBR_region)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": Saved real and imputed genotypes for non_LHBR parts of genome, for "</span>, <span class="fu">size</span>(pos_SNP_filtered_nonLHBR_region, <span class="fl">1</span>),<span class="st">" SNPs and "</span>, <span class="fu">size</span>(genosOnly_nonLHBR_region_for_imputing, <span class="fl">1</span>),<span class="st">" filtered individuals, in file </span><span class="sc">$</span>filename<span class="st">"</span>))</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Load saved imputed data for each chromosome (the non-LHBR part):</p>
<div id="d3e4dfd8" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize data structures</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>genos_imputed_loaded <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(<span class="cn">undef</span>, <span class="fu">nrow</span>(ind_with_metadata_indFiltered), <span class="fl">0</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>pos_SNP_loaded <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="dt">String</span>[], position <span class="op">=</span> <span class="dt">Int64</span>[])</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(chromosomes_to_include)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    chrom <span class="op">=</span> chromosomes_to_include[i]</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom, <span class="st">"nonLHBR"</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    filename <span class="op">=</span> <span class="fu">string</span>(baseName, tagName, regionText, <span class="st">".KNNimputedMissing.jld2"</span>)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>    imputed_genos_one_chr <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"imputed_genos"</span>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>    genos_imputed_loaded <span class="op">=</span> <span class="fu">hcat</span>(genos_imputed_loaded, imputed_genos_one_chr)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> ind_with_metadata_indFiltered.ind <span class="op">!=</span> <span class="fu">load</span>(filename, <span class="st">"ind_with_metadata_indFiltered"</span>)[<span class="op">:</span>, <span class="op">:</span>ind]</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"""Warning: "ind" columns in loaded data and memory data don't match."""</span>)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>    pos_SNP_filtered_region <span class="op">=</span> <span class="fu">load</span>(filename, <span class="st">"pos_SNP_filtered_nonLHBR_region"</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>    pos_SNP_loaded <span class="op">=</span> <span class="fu">vcat</span>(pos_SNP_loaded, pos_SNP_filtered_region)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(<span class="st">"Loaded "</span>,filename))</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="fu">string</span>(regionText, <span class="st">": "</span>, <span class="fu">size</span>(imputed_genos_one_chr,<span class="fl">2</span>), <span class="st">" SNPs from "</span>, <span class="fu">size</span>(imputed_genos_one_chr,<span class="fl">1</span>), <span class="st">" individuals"</span>))</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw28nonLHBR.KNNimputedMissing.jld2
chrgw28nonLHBR: 10180 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw27nonLHBR.KNNimputedMissing.jld2
chrgw27nonLHBR: 9184 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw26nonLHBR.KNNimputedMissing.jld2
chrgw26nonLHBR: 11803 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw25nonLHBR.KNNimputedMissing.jld2
chrgw25nonLHBR: 3294 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw24nonLHBR.KNNimputedMissing.jld2
chrgw24nonLHBR: 13321 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw23nonLHBR.KNNimputedMissing.jld2
chrgw23nonLHBR: 12949 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw22nonLHBR.KNNimputedMissing.jld2
chrgw22nonLHBR: 4973 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw21nonLHBR.KNNimputedMissing.jld2
chrgw21nonLHBR: 12821 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw20nonLHBR.KNNimputedMissing.jld2
chrgw20nonLHBR: 30239 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw19nonLHBR.KNNimputedMissing.jld2
chrgw19nonLHBR: 23914 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw18nonLHBR.KNNimputedMissing.jld2
chrgw18nonLHBR: 17359 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw17nonLHBR.KNNimputedMissing.jld2
chrgw17nonLHBR: 24313 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw15nonLHBR.KNNimputedMissing.jld2
chrgw15nonLHBR: 25517 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw14nonLHBR.KNNimputedMissing.jld2
chrgw14nonLHBR: 28469 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw13nonLHBR.KNNimputedMissing.jld2
chrgw13nonLHBR: 30543 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw12nonLHBR.KNNimputedMissing.jld2
chrgw12nonLHBR: 31794 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw11nonLHBR.KNNimputedMissing.jld2
chrgw11nonLHBR: 27183 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw10nonLHBR.KNNimputedMissing.jld2
chrgw10nonLHBR: 26462 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw9nonLHBR.KNNimputedMissing.jld2
chrgw9nonLHBR: 37680 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw8nonLHBR.KNNimputedMissing.jld2
chrgw8nonLHBR: 37318 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw7nonLHBR.KNNimputedMissing.jld2
chrgw7nonLHBR: 35575 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw6nonLHBR.KNNimputedMissing.jld2
chrgw6nonLHBR: 39675 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw5nonLHBR.KNNimputedMissing.jld2
chrgw5nonLHBR: 54829 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw4nonLHBR.KNNimputedMissing.jld2
chrgw4nonLHBR: 47980 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw3nonLHBR.KNNimputedMissing.jld2
chrgw3nonLHBR: 79372 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw2nonLHBR.KNNimputedMissing.jld2
chrgw2nonLHBR: 91292 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw1nonLHBR.KNNimputedMissing.jld2
chrgw1nonLHBR: 77362 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw1AnonLHBR.KNNimputedMissing.jld2
chrgw1AnonLHBR: 44551 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgw4AnonLHBR.KNNimputedMissing.jld2
chrgw4AnonLHBR: 17467 SNPs from 257 individuals
Loaded GW_genomics_2022_with_new_genome/GW2022_GBS_012NA_files/GW2022_all4plates.genotypes.SNPs_only.whole_genome.Jan2025.chrgwZnonLHBR.KNNimputedMissing.jld2
chrgwZnonLHBR: 50005 SNPs from 257 individuals</code></pre>
</div>
</div>
<p>Now do the PCA for non_LHBR parts of genome:</p>
<div id="a091acb4" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>PCA_wholeGenome <span class="op">=</span> <span class="fu">plotPCA</span>(genos_imputed_loaded, ind_with_metadata_indFiltered, </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>        groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> <span class="st">"non-LHBR wholeGenome"</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>        lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>        symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>totalObservationVariance <span class="op">=</span> <span class="fu">var</span>(PCA_wholeGenome.model) </span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>PC1_variance, PC2_variance <span class="op">=</span> <span class="fu">principalvars</span>(PCA_wholeGenome.model)[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>]</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>PC1_prop_variance <span class="op">=</span> PC1_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>PC2_prop_variance <span class="op">=</span> PC2_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"PC1 explains "</span>, <span class="fl">100</span><span class="op">*</span>PC1_prop_variance, <span class="st">"% of the total variance.</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>PC2 explains <span class="st">", 100*PC2_prop_variance, "</span><span class="op">%</span>.<span class="st">")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-20-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>PC1 explains 10.945476% of the total variance.
PC2 explains 5.529751%.</code></pre>
</div>
</div>
<p>The above looks quite similar to the whole-genome PCA including the LHBRs.</p>
</section>
<section id="do-non_lhbr-for-a-specific-chromosome" class="level4">
<h4 class="anchored" data-anchor-id="do-non_lhbr-for-a-specific-chromosome">Do non_LHBR for a specific chromosome:</h4>
<div id="9177fc37" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> pos_SNP_loaded.chrom <span class="op">.==</span> <span class="st">"gw4A"</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>pos_SNP_loaded_oneChr <span class="op">=</span> pos_SNP_loaded[selection, <span class="op">:</span>]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>genos_imputed_loaded_oneChr <span class="op">=</span> genos_imputed_loaded[<span class="op">:</span>, selection]</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>PCA_oneChr <span class="op">=</span> <span class="fu">plotPCA</span>(genos_imputed_loaded_oneChr, ind_with_metadata_indFiltered, </span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>        groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>        sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> <span class="st">"non-LHBR gw4A"</span>,</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>        flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>        lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>        symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>totalObservationVariance <span class="op">=</span> <span class="fu">var</span>(PCA_oneChr.model) </span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>PC1_variance, PC2_variance <span class="op">=</span> <span class="fu">principalvars</span>(PCA_oneChr.model)[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>]</span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>PC1_prop_variance <span class="op">=</span> PC1_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>PC2_prop_variance <span class="op">=</span> PC2_variance <span class="op">/</span> totalObservationVariance</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"PC1 explains "</span>, <span class="fl">100</span><span class="op">*</span>PC1_prop_variance, <span class="st">"% of the total variance.</span></span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>PC2 explains <span class="st">", 100*PC2_prop_variance, "</span><span class="op">%</span>.<span class="st">")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-21-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>PC1 explains 7.5557256% of the total variance.
PC2 explains 4.751407%.</code></pre>
</div>
</div>
<p>There is a whole lot of signal for population structure in the non-LHBR regions—maybe not suprising given there are high-Fst regions with low ViSHet.</p>
</section>
</section>
<section id="choose-low-het-individuals-in-a-high-hetvar-region" class="level2">
<h2 class="anchored" data-anchor-id="choose-low-het-individuals-in-a-high-hetvar-region">Choose low-het individuals in a high HetVar region</h2>
<p>Now that we have high HetVar regions indicated, we can automate the choosing of essential homozygous individuals in those regions. Below I will do this for chrZ (and later chr15 and chr18).</p>
<p>First, here’s two functions, first for getting the boundaries of the longest high HetVar region from a scaffold, second for getting a bunch of info about the region:</p>
<div id="ff1df3ba" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">getOneHighHetVarRegion</span>(highHetVarRegions, chr)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (highHetVarRegions.regionChrom <span class="op">.==</span> chr)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">sum</span>(selection) <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"Good news: 1 region on that scaffold"</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>        positionMin <span class="op">=</span> highHetVarRegions.regionStart[selection][<span class="fl">1</span>]</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>        positionMax <span class="op">=</span> highHetVarRegions.regionEnd[selection][<span class="fl">1</span>]</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>        regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,positionMin,<span class="st">" to "</span>,positionMax)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elseif</span> <span class="fu">sum</span>(selection) <span class="op">&gt;</span> <span class="fl">1</span></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"More than 1 region on that scaffold. Using just the longest one."</span>)</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>        highHetVarRegions_chr <span class="op">=</span> highHetVarRegions[selection, <span class="op">:</span>]</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">display</span>(highHetVarRegions_chr)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>        <span class="co"># get biggest region (first one if tied):</span></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>        regionSizes <span class="op">=</span> highHetVarRegions_chr.regionEnd <span class="op">.-</span> highHetVarRegions_chr.regionStart</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>        indexOfLongest <span class="op">=</span> <span class="fu">findfirst</span>(regionSizes <span class="op">.==</span> <span class="fu">maximum</span>(regionSizes))</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>        positionMin <span class="op">=</span> highHetVarRegions_chr.regionStart[indexOfLongest]</span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>        positionMax <span class="op">=</span> highHetVarRegions_chr.regionEnd[indexOfLongest]</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>        regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,positionMin,<span class="st">" to "</span>,positionMax)</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elseif</span> <span class="fu">sum</span>(selection) <span class="op">==</span> <span class="fl">0</span></span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"No high HetVar regions on that scaffold"</span>)</span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span></span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> positionMin, positionMax, regionText</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">getWindowedIndHetStanRegion</span>(genos, pos, </span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>                                    highHetVarRegions, chr;</span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>                                    windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># remake the windowedIndHet_standardized (done above in a different cell)</span></span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos.chrom <span class="op">.==</span> chr)</span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a>    pos_region <span class="op">=</span> pos[loci_selection, <span class="op">:</span>]</span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a>    genotypes_region <span class="op">=</span> genos[<span class="op">:</span>, loci_selection]</span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a>    windowedPos, windowedIndHet <span class="op">=</span> <span class="fu">getWindowedIndHet</span>(genotypes_region, pos_region, windowSize)</span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a>    windowedIndHet_standardized <span class="op">=</span> <span class="fu">standardizeIndHet</span>(windowedIndHet)</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># look up the boundaries of the high HetVar region:</span></span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a>    positionMin, positionMax, regionText <span class="op">=</span> <span class="fu">getOneHighHetVarRegion</span>(highHetVarRegions, chr)</span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># choose just the windows that are in the high HetVar region:</span></span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a>    window_selection <span class="op">=</span> (positionMin <span class="op">.&lt;</span> windowedPos <span class="op">.&lt;</span> positionMax)</span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion <span class="op">=</span> windowedIndHet_standardized[<span class="op">:</span>,window_selection]</span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a>    meanAcrossRegionIndHetStan <span class="op">=</span> <span class="fu">mean</span>.(<span class="fu">eachrow</span>(windowedIndHetStanRegion)) </span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># choose loci in region</span></span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>    lociSelection <span class="op">=</span> (positionMin <span class="op">.&lt;</span> pos_region.position <span class="op">.&lt;</span> positionMax)</span>
<span id="cb43-43"><a href="#cb43-43" aria-hidden="true" tabindex="-1"></a>    pos_highHetVarRegion <span class="op">=</span> pos_region[lociSelection, <span class="op">:</span>]</span>
<span id="cb43-44"><a href="#cb43-44" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion <span class="op">=</span> genotypes_region[<span class="op">:</span>, lociSelection]</span>
<span id="cb43-45"><a href="#cb43-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># convert `-1` genotypes (which indicates missing) to `missing`:</span></span>
<span id="cb43-46"><a href="#cb43-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">replace!</span>(genos_highHetVarRegion, <span class="op">-</span><span class="fl">1</span> <span class="op">=&gt;</span> <span class="cn">missing</span>)</span>
<span id="cb43-47"><a href="#cb43-47" aria-hidden="true" tabindex="-1"></a>    regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_highHetVarRegion, chr; positionMin<span class="op">=</span>positionMin, positionMax<span class="op">=</span>positionMax) <span class="co"># this makes appropriate text describing the region</span></span>
<span id="cb43-48"><a href="#cb43-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> positionMin, positionMax, regionText, </span>
<span id="cb43-49"><a href="#cb43-49" aria-hidden="true" tabindex="-1"></a>            windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb43-50"><a href="#cb43-50" aria-hidden="true" tabindex="-1"></a>            genos_highHetVarRegion, pos_highHetVarRegion, regionInfo</span>
<span id="cb43-51"><a href="#cb43-51" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>getWindowedIndHetStanRegion (generic function with 1 method)</code></pre>
</div>
</div>
<div id="3c93db89" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gwZ"</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-43"><a href="#cb45-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb45-44"><a href="#cb45-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb45-45"><a href="#cb45-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb45-46"><a href="#cb45-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb45-47"><a href="#cb45-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb45-48"><a href="#cb45-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb45-49"><a href="#cb45-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb45-50"><a href="#cb45-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb45-51"><a href="#cb45-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb45-52"><a href="#cb45-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb45-53"><a href="#cb45-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb45-54"><a href="#cb45-54" aria-hidden="true" tabindex="-1"></a><span class="co"># also flip PC3:</span></span>
<span id="cb45-55"><a href="#cb45-55" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb45-56"><a href="#cb45-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-57"><a href="#cb45-57" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb45-58"><a href="#cb45-58" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb45-59"><a href="#cb45-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-60"><a href="#cb45-60" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb45-61"><a href="#cb45-61" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">2</span>) </span>
<span id="cb45-62"><a href="#cb45-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-63"><a href="#cb45-63" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals, PC1 to PC2:</span></span>
<span id="cb45-64"><a href="#cb45-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-65"><a href="#cb45-65" aria-hidden="true" tabindex="-1"></a>fig_3A <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb45-66"><a href="#cb45-66" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(fig_3A[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb45-67"><a href="#cb45-67" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"gwZ LHBR PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb45-68"><a href="#cb45-68" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb45-69"><a href="#cb45-69" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb45-70"><a href="#cb45-70" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb45-71"><a href="#cb45-71" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb45-72"><a href="#cb45-72" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb45-73"><a href="#cb45-73" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb45-74"><a href="#cb45-74" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb45-75"><a href="#cb45-75" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb45-76"><a href="#cb45-76" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(fig_3A)</span>
<span id="cb45-77"><a href="#cb45-77" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb45-78"><a href="#cb45-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"Figure3A_from_Julia.png"</span>, fig_3A, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb45-79"><a href="#cb45-79" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-23-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-23-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="now-plot-pc1-vs.-pc3-for-low-vishet-individuals" class="level4">
<h4 class="anchored" data-anchor-id="now-plot-pc1-vs.-pc3-for-low-vishet-individuals">Now plot PC1 vs.&nbsp;PC3 for low ViSHet individuals</h4>
<div id="2a20c1c6" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals, PC1 to PC2:</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>fig_3B <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(fig_3B[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"gwZ LHBR PC1 vs. PC3, only low heterozygosity"</span>,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(fig_3B)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"Figure3B_from_Julia.png"</span>, fig_3B, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-24-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="save-the-individual-colors-in-the-metadata" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata">Save the individual colors in the metadata</h3>
<div id="0e3f5861" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="1b94b2ba" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>fig_3C <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(fig_3C[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"gwZ LHBR PC1 vs. PC2, all individuals"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(fig_3C)</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"Figure3C_from_Julia.png"</span>, fig_3C, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-26-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="b23c2d13" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>fig_3D <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(fig_3D[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"gwZ LHBR PC1 vs. PC3, all individuals"</span>,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(fig_3D)</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"Figure3D_from_Julia.png"</span>, fig_3D, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-27-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="now-it-is-clear-that-there-are-six-clear-haplogroups-of-z-high-hetvar-region.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="now-it-is-clear-that-there-are-six-clear-haplogroups-of-z-high-hetvar-region.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">Now it is clear that there are six clear haplogroups of Z high HetVar region. Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="d8b410ff" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect Z chromosome PCA of low IndHet (&lt; 2) individuals,</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and specify group boundaries:</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"seagreen4"</span>,</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a>vir <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4.5</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a>        (PCAmodelAll.metadata.PC3 <span class="op">.&gt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-21"><a href="#cb57-21" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-22"><a href="#cb57-22" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-25"><a href="#cb57-25" aria-hidden="true" tabindex="-1"></a>lud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4.5</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-26"><a href="#cb57-26" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span> </span>
<span id="cb57-27"><a href="#cb57-27" aria-hidden="true" tabindex="-1"></a>        (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-28"><a href="#cb57-28" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-29"><a href="#cb57-29" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-30"><a href="#cb57-30" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-31"><a href="#cb57-31" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-32"><a href="#cb57-32" aria-hidden="true" tabindex="-1"></a>        (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span> </span>
<span id="cb57-33"><a href="#cb57-33" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-34"><a href="#cb57-34" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb57-35"><a href="#cb57-35" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb57-36"><a href="#cb57-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-37"><a href="#cb57-37" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb57-38"><a href="#cb57-38" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[vir]</span>
<span id="cb57-39"><a href="#cb57-39" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit] <span class="co"># note there are two nitidus with nearly identical values</span></span>
<span id="cb57-40"><a href="#cb57-40" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[lud]</span>
<span id="cb57-41"><a href="#cb57-41" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb57-42"><a href="#cb57-42" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb57-43"><a href="#cb57-43" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb57-44"><a href="#cb57-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-45"><a href="#cb57-45" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir nit lud troch obs plumb]</span>
<span id="cb57-46"><a href="#cb57-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-47"><a href="#cb57-47" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb57-48"><a href="#cb57-48" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb57-49"><a href="#cb57-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-50"><a href="#cb57-50" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb57-51"><a href="#cb57-51" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb57-52"><a href="#cb57-52" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb57-53"><a href="#cb57-53" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb57-54"><a href="#cb57-54" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb57-55"><a href="#cb57-55" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb57-56"><a href="#cb57-56" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb57-57"><a href="#cb57-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-58"><a href="#cb57-58" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb57-59"><a href="#cb57-59" aria-hidden="true" tabindex="-1"></a><span class="st">    getFreqsAndSampleSizesBySexForZ(genoData, sex, indGroup, groupsToCalc)</span></span>
<span id="cb57-60"><a href="#cb57-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-61"><a href="#cb57-61" aria-hidden="true" tabindex="-1"></a><span class="st">Calculate allele frequencies and sample sizes for each group and SNP, taking into account sex for analysis of Z chromosome.</span></span>
<span id="cb57-62"><a href="#cb57-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-63"><a href="#cb57-63" aria-hidden="true" tabindex="-1"></a><span class="st">​# Arguments</span></span>
<span id="cb57-64"><a href="#cb57-64" aria-hidden="true" tabindex="-1"></a><span class="st">- `genoData`: The genotype matrix, where rows are individuals and columns are loci, with genotype codes 0,1,2 meaning homozygous reference, heterozygote, homozygous alternate, and missing genotypes can be either -1 or `missing`.</span></span>
<span id="cb57-65"><a href="#cb57-65" aria-hidden="true" tabindex="-1"></a><span class="st">- `sex`: Vector of sexes ('f` or `m`)</span></span>
<span id="cb57-66"><a href="#cb57-66" aria-hidden="true" tabindex="-1"></a><span class="st">- `indGroup`: A vector providing the group name each individual belongs to.</span></span>
<span id="cb57-67"><a href="#cb57-67" aria-hidden="true" tabindex="-1"></a><span class="st">- `groupsToCalc`: A list of group names to include in calculations.</span></span>
<span id="cb57-68"><a href="#cb57-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-69"><a href="#cb57-69" aria-hidden="true" tabindex="-1"></a><span class="st"># Notes</span></span>
<span id="cb57-70"><a href="#cb57-70" aria-hidden="true" tabindex="-1"></a><span class="st">Returns a tuple containing 1) a matrix of frequencies, and 2) a matrix of samples sizes (in both, rows are groups and columns are loci). </span></span>
<span id="cb57-71"><a href="#cb57-71" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb57-72"><a href="#cb57-72" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">getFreqsAndSampleSizesBySexForZ</span>(genoData, sex, indGroup, groupsToCalc)</span>
<span id="cb57-73"><a href="#cb57-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">any</span>(.!<span class="fu">map</span>(x <span class="op">-&gt;</span> x <span class="kw">in</span> [<span class="st">"F"</span>, <span class="st">"M"</span>], sex))</span>
<span id="cb57-74"><a href="#cb57-74" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"Warning: not all entries in sex vector are `F` or `M`"</span>)</span>
<span id="cb57-75"><a href="#cb57-75" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb57-76"><a href="#cb57-76" aria-hidden="true" tabindex="-1"></a>    genoData[<span class="fu">ismissing</span>.(genoData)] <span class="op">.=</span> <span class="op">-</span><span class="fl">1</span> <span class="co"># if "missing" datatype is use, convert to -1</span></span>
<span id="cb57-77"><a href="#cb57-77" aria-hidden="true" tabindex="-1"></a>    groupCount <span class="op">=</span> <span class="fu">length</span>(groupsToCalc)</span>
<span id="cb57-78"><a href="#cb57-78" aria-hidden="true" tabindex="-1"></a>    freqs <span class="op">=</span> <span class="fu">Array</span><span class="dt">{Float32,2}</span>(<span class="cn">undef</span>, groupCount, <span class="fu">size</span>(genoData, <span class="fl">2</span>))</span>
<span id="cb57-79"><a href="#cb57-79" aria-hidden="true" tabindex="-1"></a>    sampleSizes <span class="op">=</span> <span class="fu">Array</span><span class="dt">{Number,2}</span>(<span class="cn">undef</span>, groupCount, <span class="fu">size</span>(genoData, <span class="fl">2</span>))</span>
<span id="cb57-80"><a href="#cb57-80" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>groupCount</span>
<span id="cb57-81"><a href="#cb57-81" aria-hidden="true" tabindex="-1"></a>        <span class="co"># females:</span></span>
<span id="cb57-82"><a href="#cb57-82" aria-hidden="true" tabindex="-1"></a>        selection <span class="op">=</span> (indGroup <span class="op">.==</span> groupsToCalc[i]) <span class="op">.&amp;</span> (sex <span class="op">.==</span> <span class="st">"F"</span>) <span class="co"># gets the correct rows for individuals in the group </span></span>
<span id="cb57-83"><a href="#cb57-83" aria-hidden="true" tabindex="-1"></a>        geno0countsF <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">0</span>, dims<span class="op">=</span><span class="fl">1</span>) <span class="co"># count by column the number of 0 genotypes (homozygous ref)</span></span>
<span id="cb57-84"><a href="#cb57-84" aria-hidden="true" tabindex="-1"></a>        geno1countsF <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>) <span class="co"># same for 1 genotypes (heterozygous)</span></span>
<span id="cb57-85"><a href="#cb57-85" aria-hidden="true" tabindex="-1"></a>        geno2countsF <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">2</span>, dims<span class="op">=</span><span class="fl">1</span>) <span class="co"># same for 2 genotypes (homozygous alternate) </span></span>
<span id="cb57-86"><a href="#cb57-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-87"><a href="#cb57-87" aria-hidden="true" tabindex="-1"></a>        <span class="co"># males: </span></span>
<span id="cb57-88"><a href="#cb57-88" aria-hidden="true" tabindex="-1"></a>        selection <span class="op">=</span> (indGroup <span class="op">.==</span> groupsToCalc[i]) <span class="op">.&amp;</span> (sex <span class="op">.==</span> <span class="st">"M"</span>)</span>
<span id="cb57-89"><a href="#cb57-89" aria-hidden="true" tabindex="-1"></a>        geno0countsM <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">0</span>, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb57-90"><a href="#cb57-90" aria-hidden="true" tabindex="-1"></a>        geno1countsM <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">1</span>, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb57-91"><a href="#cb57-91" aria-hidden="true" tabindex="-1"></a>        geno2countsM <span class="op">=</span> <span class="fu">sum</span>(genoData[selection, <span class="op">:</span>] <span class="op">.==</span> <span class="fl">2</span>, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb57-92"><a href="#cb57-92" aria-hidden="true" tabindex="-1"></a>        allele0counts <span class="op">=</span> (<span class="fl">2</span> <span class="op">.*</span> geno0countsM) <span class="op">.+</span> geno1countsM <span class="op">.+</span> geno0countsF <span class="op">.+</span> (<span class="fl">0.5</span> <span class="op">.*</span> geno1countsF)</span>
<span id="cb57-93"><a href="#cb57-93" aria-hidden="true" tabindex="-1"></a>        allele2counts <span class="op">=</span> (<span class="fl">2</span> <span class="op">.*</span> geno2countsM) <span class="op">.+</span> geno1countsM <span class="op">.+</span> geno2countsF <span class="op">.+</span> (<span class="fl">0.5</span> <span class="op">.*</span> geno1countsF)</span>
<span id="cb57-94"><a href="#cb57-94" aria-hidden="true" tabindex="-1"></a>        sumAlleleCounts <span class="op">=</span> allele0counts <span class="op">.+</span> allele2counts</span>
<span id="cb57-95"><a href="#cb57-95" aria-hidden="true" tabindex="-1"></a>        sampleSizes[i, <span class="op">:</span>] <span class="op">=</span> <span class="fl">0.5</span> <span class="op">.*</span> sumAlleleCounts <span class="co"># sample size in number of individuals</span></span>
<span id="cb57-96"><a href="#cb57-96" aria-hidden="true" tabindex="-1"></a>        freqs[i, <span class="op">:</span>] <span class="op">=</span> allele2counts <span class="op">./</span> sumAlleleCounts</span>
<span id="cb57-97"><a href="#cb57-97" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb57-98"><a href="#cb57-98" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> freqs, sampleSizes</span>
<span id="cb57-99"><a href="#cb57-99" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb57-100"><a href="#cb57-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-101"><a href="#cb57-101" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb57-102"><a href="#cb57-102" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizesBySexForZ</span>(genos_highHetVarRegion, ind_with_metadata_included.sex, clusterMembership, clusterNames)</span>
<span id="cb57-103"><a href="#cb57-103" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb57-104"><a href="#cb57-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-105"><a href="#cb57-105" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb57-106"><a href="#cb57-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-107"><a href="#cb57-107" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb57-108"><a href="#cb57-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-109"><a href="#cb57-109" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb57-110"><a href="#cb57-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-111"><a href="#cb57-111" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb57-112"><a href="#cb57-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-113"><a href="#cb57-113" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb57-114"><a href="#cb57-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-115"><a href="#cb57-115" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb57-116"><a href="#cb57-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-117"><a href="#cb57-117" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb57-118"><a href="#cb57-118" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb57-119"><a href="#cb57-119" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb57-120"><a href="#cb57-120" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb57-121"><a href="#cb57-121" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb57-122"><a href="#cb57-122" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir      0.00706945</span></span>
<span id="cb57-123"><a href="#cb57-123" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.0035094</span></span>
<span id="cb57-124"><a href="#cb57-124" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ lud      0.00794828</span></span>
<span id="cb57-125"><a href="#cb57-125" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ troch    0.00968743</span></span>
<span id="cb57-126"><a href="#cb57-126" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ obs      0.0112686</span></span>
<span id="cb57-127"><a href="#cb57-127" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ plumb    0.0104236 =#</span></span>
<span id="cb57-128"><a href="#cb57-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-129"><a href="#cb57-129" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb57-130"><a href="#cb57-130" aria-hidden="true" tabindex="-1"></a><span class="co">#=</span></span>
<span id="cb57-131"><a href="#cb57-131" aria-hidden="true" tabindex="-1"></a><span class="co">15×2 DataFrame</span></span>
<span id="cb57-132"><a href="#cb57-132" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb57-133"><a href="#cb57-133" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb57-134"><a href="#cb57-134" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb57-135"><a href="#cb57-135" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir_nit       0.0388124</span></span>
<span id="cb57-136"><a href="#cb57-136" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ vir_lud       0.0173121</span></span>
<span id="cb57-137"><a href="#cb57-137" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ vir_troch     0.0410345</span></span>
<span id="cb57-138"><a href="#cb57-138" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ vir_obs       0.037509</span></span>
<span id="cb57-139"><a href="#cb57-139" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ vir_plumb     0.0466843</span></span>
<span id="cb57-140"><a href="#cb57-140" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_lud       0.0365198</span></span>
<span id="cb57-141"><a href="#cb57-141" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_troch     0.0464045</span></span>
<span id="cb57-142"><a href="#cb57-142" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ nit_obs       0.0427296</span></span>
<span id="cb57-143"><a href="#cb57-143" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ nit_plumb     0.0524538</span></span>
<span id="cb57-144"><a href="#cb57-144" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ lud_troch     0.038968</span></span>
<span id="cb57-145"><a href="#cb57-145" aria-hidden="true" tabindex="-1"></a><span class="co">  11 │ lud_obs       0.0354195</span></span>
<span id="cb57-146"><a href="#cb57-146" aria-hidden="true" tabindex="-1"></a><span class="co">  12 │ lud_plumb     0.0445131</span></span>
<span id="cb57-147"><a href="#cb57-147" aria-hidden="true" tabindex="-1"></a><span class="co">  13 │ troch_obs     0.0265615</span></span>
<span id="cb57-148"><a href="#cb57-148" aria-hidden="true" tabindex="-1"></a><span class="co">  14 │ troch_plumb   0.0404278</span></span>
<span id="cb57-149"><a href="#cb57-149" aria-hidden="true" tabindex="-1"></a><span class="co">  15 │ obs_plumb     0.0365139 =#</span></span>
<span id="cb57-150"><a href="#cb57-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-151"><a href="#cb57-151" aria-hidden="true" tabindex="-1"></a><span class="co"># It seems the distances are not very consistent with a bifurcating tree,</span></span>
<span id="cb57-152"><a href="#cb57-152" aria-hidden="true" tabindex="-1"></a><span class="co"># nor 1-D isolation by distance, but something more complex.</span></span>
<span id="cb57-153"><a href="#cb57-153" aria-hidden="true" tabindex="-1"></a><span class="co"># Obscuratus is closer to viridanus than troch is.</span></span>
<span id="cb57-154"><a href="#cb57-154" aria-hidden="true" tabindex="-1"></a><span class="co"># Nitidus quite distant but gets put in centre of PCA because off on its own axis.</span></span>
<span id="cb57-155"><a href="#cb57-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-156"><a href="#cb57-156" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb57-157"><a href="#cb57-157" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb57-158"><a href="#cb57-158" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb57-159"><a href="#cb57-159" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb57-160"><a href="#cb57-160" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb57-161"><a href="#cb57-161" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb57-162"><a href="#cb57-162" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb57-163"><a href="#cb57-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-164"><a href="#cb57-164" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb57-165"><a href="#cb57-165" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb57-166"><a href="#cb57-166" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb57-167"><a href="#cb57-167" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb57-168"><a href="#cb57-168" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb57-169"><a href="#cb57-169" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb57-170"><a href="#cb57-170" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb57-171"><a href="#cb57-171" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb57-172"><a href="#cb57-172" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb57-173"><a href="#cb57-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-174"><a href="#cb57-174" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb57-175"><a href="#cb57-175" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> [<span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">15</span>]</span>
<span id="cb57-176"><a href="#cb57-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-177"><a href="#cb57-177" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb57-178"><a href="#cb57-178" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb57-179"><a href="#cb57-179" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb57-180"><a href="#cb57-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-181"><a href="#cb57-181" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb57-182"><a href="#cb57-182" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb57-183"><a href="#cb57-183" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb57-184"><a href="#cb57-184" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [37 2 38 80 5 73] and the sum of those is 235
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-28-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="28">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [68372991, 68434641, 68434654, 68434681, 68506332, 68506385, 68506399, 68533951, 68682243, 68795208  …  73563085, 73563098, 73586599, 73680319, 73743035, 73743078, 73743083, 73743128, 73743136, 73743151], <span class="ansi-bold">67×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   3 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
   4 │ GW_Lane5_AA5               GW_Lane5_AA5               AA        vir_S
   5 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir     ⋯
   6 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
   7 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
   8 │ GW_Lane5_TL8               GW_Lane5_TL8               TL        vir
   9 │ GW_Lane5_YK5               GW_Lane5_YK5               YK        vir     ⋯
  10 │ GW_Lane5_YK6               GW_Lane5_YK6               YK        vir
  11 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  58 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
  59 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
  60 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  61 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
  62 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  63 │ GW_Armando_plate1_JF22G01  GW_Armando_plate1_JF22G01  MB        plumb   ⋯
  64 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  65 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  66 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  67 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  15 columns and 46 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="0f5619ca" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_lud"</span>,</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud"</span>,</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud_troch"</span>,</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>]</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lightseagreen"</span>,</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>]</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a>vir_lud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4.5</span>) <span class="op">.&amp;</span> </span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a>        (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="co"># Note this one has lowIndHetStan but is mix of vir and lud</span></span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a>lud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir vir_lud nit lud lud_troch troch obs plumb]</span>
<span id="cb60-27"><a href="#cb60-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-28"><a href="#cb60-28" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb60-29"><a href="#cb60-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-30"><a href="#cb60-30" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb60-31"><a href="#cb60-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb60-32"><a href="#cb60-32" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb60-33"><a href="#cb60-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb60-34"><a href="#cb60-34" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb60-35"><a href="#cb60-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-36"><a href="#cb60-36" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb60-37"><a href="#cb60-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-38"><a href="#cb60-38" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb60-39"><a href="#cb60-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-40"><a href="#cb60-40" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb60-41"><a href="#cb60-41" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb60-42"><a href="#cb60-42" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb60-43"><a href="#cb60-43" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb60-44"><a href="#cb60-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-45"><a href="#cb60-45" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb60-46"><a href="#cb60-46" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb60-47"><a href="#cb60-47" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb60-48"><a href="#cb60-48" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb60-49"><a href="#cb60-49" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb60-50"><a href="#cb60-50" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb60-51"><a href="#cb60-51" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb60-52"><a href="#cb60-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-53"><a href="#cb60-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb60-54"><a href="#cb60-54" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb60-55"><a href="#cb60-55" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb60-56"><a href="#cb60-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-57"><a href="#cb60-57" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb60-58"><a href="#cb60-58" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb60-59"><a href="#cb60-59" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb60-60"><a href="#cb60-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-61"><a href="#cb60-61" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb60-62"><a href="#cb60-62" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb60-63"><a href="#cb60-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-64"><a href="#cb60-64" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb60-65"><a href="#cb60-65" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb60-66"><a href="#cb60-66" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb60-67"><a href="#cb60-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-68"><a href="#cb60-68" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb60-69"><a href="#cb60-69" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb60-70"><a href="#cb60-70" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb60-71"><a href="#cb60-71" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb60-72"><a href="#cb60-72" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-29-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="29">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [68372991, 68434641, 68434654, 68434681, 68506332, 68506385, 68506399, 68533951, 68682243, 68795208  …  73563085, 73563098, 73586599, 73680319, 73743035, 73743078, 73743083, 73743128, 73743136, 73743151], <span class="ansi-bold">83×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   3 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
   4 │ GW_Lane5_AA5               GW_Lane5_AA5               AA        vir_S
   5 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir     ⋯
   6 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
   7 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
   8 │ GW_Lane5_TL8               GW_Lane5_TL8               TL        vir
   9 │ GW_Lane5_YK5               GW_Lane5_YK5               YK        vir     ⋯
  10 │ GW_Lane5_YK6               GW_Lane5_YK6               YK        vir
  11 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  74 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
  75 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
  76 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  77 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
  78 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  79 │ GW_Armando_plate1_JF22G01  GW_Armando_plate1_JF22G01  MB        plumb   ⋯
  80 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  81 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  82 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  83 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  15 columns and 62 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order">show GBI plot according to original groups and plot order</h3>
<div id="850047a3" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-30-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="30">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [68372991, 68434641, 68434654, 68434681, 68506332, 68506385, 68506399, 68533951, 68682243, 68795208  …  73563085, 73563098, 73586599, 73680319, 73743035, 73743078, 73743083, 73743128, 73743136, 73743151], <span class="ansi-bold">83×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   3 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   4 │ GW_Lane5_YK5               GW_Lane5_YK5               YK        vir
   5 │ GW_Lane5_YK6               GW_Lane5_YK6               YK        vir     ⋯
   6 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   7 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir
   8 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
   9 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir     ⋯
  10 │ GW_Lane5_TL8               GW_Lane5_TL8               TL        vir
  11 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  74 │ GW_Armando_plate1_JF09G01  GW_Armando_plate1_JF09G01  ST        plumb
  75 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb   ⋯
  76 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb
  77 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  78 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
  79 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb   ⋯
  80 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  81 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  82 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  83 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb   ⋯
<span class="ansi-cyan-fg">                                                  15 columns and 62 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals">show same but with all individuals</h3>
<div id="1be77217" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>                indFontSize<span class="op">=</span><span class="fl">6</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">1600</span>),</span>
<span id="cb65-14"><a href="#cb65-14" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb65-15"><a href="#cb65-15" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb65-16"><a href="#cb65-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-17"><a href="#cb65-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">true</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb65-18"><a href="#cb65-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"FigureS1_from_Julia.png"</span>, plotInfo[<span class="fl">1</span>], px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb65-19"><a href="#cb65-19" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-31-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="31">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus">Show just the west area (without nitidus)</h3>
<div id="cfaf41c3" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud"</span>,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud_troch"</span>,</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,]</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizesBySexForZ</span>(genos_selectedSNPs, ind_with_metadata_included.sex, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb68-21"><a href="#cb68-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb68-22"><a href="#cb68-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-23"><a href="#cb68-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb68-24"><a href="#cb68-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-25"><a href="#cb68-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb68-26"><a href="#cb68-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb68-27"><a href="#cb68-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb68-28"><a href="#cb68-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-29"><a href="#cb68-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb68-30"><a href="#cb68-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb68-31"><a href="#cb68-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb68-32"><a href="#cb68-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb68-33"><a href="#cb68-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-32-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="32">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 2 0; 0 0 … 2 0], [68434641, 68434654, 68434681, 68506332, 68506385, 68533951, 68682243, 68795208, 69062410, 69164105  …  73298332, 73298436, 73347645, 73504646, 73558384, 73563098, 73680319, 73743035, 73743128, 73743151], <span class="ansi-bold">176×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK5               GW_Lane5_YK5               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 167 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 168 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L ⋯
 169 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
 170 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L
 171 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L
 172 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L ⋯
 173 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
 174 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 175 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 176 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 15 columns and 155 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area">Show just the east area</h3>
<div id="964b823f" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizesBySexForZ</span>(genos_selectedSNPs, ind_with_metadata_included.sex, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb71-26"><a href="#cb71-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-27"><a href="#cb71-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb71-28"><a href="#cb71-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb71-29"><a href="#cb71-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb71-30"><a href="#cb71-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb71-31"><a href="#cb71-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-33-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="33">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 1 1 … 0 0; … ; 2 2 … 0 2; 2 2 … 0 2], [68434641, 68434654, 68533951, 68795208, 69010058, 69062405, 69103733, 69164105, 69293527, 69321751  …  73316610, 73344945, 73389057, 73471732, 73504381, 73504408, 73504704, 73743078, 73743083, 73743128], <span class="ansi-bold">158×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW70      GW_Armando_plate1_TTGW70      SA        l ⋯
   2 │ GW_Armando_plate1_TTGW72      GW_Armando_plate1_TTGW72      SA        l
   3 │ GW_Armando_plate1_TTGW78      GW_Armando_plate1_TTGW78      SA        l
   4 │ GW_Armando_plate2_TTGW77      GW_Armando_plate2_TTGW77      SA        l
   5 │ GW_Armando_plate2_TTGW79      GW_Armando_plate2_TTGW79      SA        l ⋯
   6 │ GW_Armando_plate2_TTGW80      GW_Armando_plate2_TTGW80      SA        l
   7 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l
   8 │ GW_Armando_plate2_TTGW75      GW_Armando_plate2_TTGW75      SA        l
   9 │ GW_Lane5_PT12                 GW_Lane5_PT12                 KL        l ⋯
  10 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l
  11 │ GW_Armando_plate2_TTGW28      GW_Armando_plate2_TTGW28      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 149 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
 150 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p ⋯
 151 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 152 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p
 153 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
 154 │ GW_Armando_plate2_JG02G01     GW_Armando_plate2_JG02G01     PR        p ⋯
 155 │ GW_Armando_plate2_JG02G03     GW_Armando_plate2_JG02G03     PR        p
 156 │ GW_Lane5_SL1                  GW_Lane5_SL1                  SL        p
 157 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p
 158 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                 15 columns and 137 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area">Show just the northern area</h3>
<div id="bd5dd9f8" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb74-15"><a href="#cb74-15" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb74-16"><a href="#cb74-16" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb74-17"><a href="#cb74-17" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb74-18"><a href="#cb74-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-19"><a href="#cb74-19" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb74-20"><a href="#cb74-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-21"><a href="#cb74-21" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb74-22"><a href="#cb74-22" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb74-23"><a href="#cb74-23" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb74-24"><a href="#cb74-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-25"><a href="#cb74-25" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb74-26"><a href="#cb74-26" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb74-27"><a href="#cb74-27" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb74-28"><a href="#cb74-28" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb74-29"><a href="#cb74-29" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-34-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="34">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [68434681, 68506332, 68506385, 68682243, 69010058, 69062410, 69103733, 69253971, 69267458, 69275355  …  73309732, 73316610, 73344945, 73347645, 73389057, 73471732, 73504381, 73558384, 73563098, 73743078], <span class="ansi-bold">110×18 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   4 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   5 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK5               GW_Lane5_YK5               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 101 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
 102 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb   ⋯
 103 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 104 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 105 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 106 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 107 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 108 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 109 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 110 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  15 columns and 89 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="do-a-pca-based-on-above-info-on-just-the-west-side-of-the-ring" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-above-info-on-just-the-west-side-of-the-ring">Do a PCA based on above info, on just the west side of the ring</h3>
<div id="cf3c1fa7" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA_westside <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>group_colors_PCA_westside <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="co"># without nitidus:</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA_westside <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>group_colors_PCA_westside <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA_westside, group_colors_PCA_westside; </span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers (west side)"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-35-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="35">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="do-a-pca-based-on-above-info-on-just-the-east-side-of-the-ring" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-above-info-on-just-the-east-side-of-the-ring">Do a PCA based on above info, on just the east side of the ring</h3>
<div id="6a7b4715" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA_eastside <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>group_colors_PCA_eastside <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>];</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA_eastside, group_colors_PCA_eastside; </span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers (east side)"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-36-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="36">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="show-pca-for-just-the-western-haploblocks-of-the-ring-for-just-chr-z" class="level3">
<h3 class="anchored" data-anchor-id="show-pca-for-just-the-western-haploblocks-of-the-ring-for-just-chr-z">Show PCA for just the western haploblocks of the ring, for just chr Z</h3>
<div id="55dac846" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included_temp <span class="op">=</span> <span class="fu">copy</span>(ind_with_metadata_included)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Leave out the individuals that don't have western haplogroup genotypes</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>hapGroups_to_plot_PCA_westside <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_lud"</span>, <span class="st">"nit"</span>, <span class="st">"lud"</span>]</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(hapGroups_to_plot_PCA_westside), clusterMembershipWithHets)</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included_temp.Fst_group[.!selection] <span class="op">.=</span> <span class="st">"ignore"</span> <span class="co"># write over the group name so function below won't plot that individual</span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA_westside <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>group_colors_PCA_westside <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>PCAmodelHapWest <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included_temp, </span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA_westside, group_colors_PCA_westside; </span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers west haps"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb83-17"><a href="#cb83-17" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb83-18"><a href="#cb83-18" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb83-19"><a href="#cb83-19" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb83-20"><a href="#cb83-20" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb83-21"><a href="#cb83-21" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb83-22"><a href="#cb83-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-23"><a href="#cb83-23" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelHapWest.PCAfig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-37-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="37">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="show-pca-for-just-the-eastern-haploblocks-of-the-ring-for-just-chr-z" class="level3">
<h3 class="anchored" data-anchor-id="show-pca-for-just-the-eastern-haploblocks-of-the-ring-for-just-chr-z">Show PCA for just the eastern haploblocks of the ring, for just chr Z</h3>
<div id="f9611a2e" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included_temp <span class="op">=</span> <span class="fu">copy</span>(ind_with_metadata_included)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Leave out the individuals that don't have western haplogroup genotypes</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>hapGroups_to_plot_PCA_eastside <span class="op">=</span> [<span class="st">"troch"</span>, <span class="st">"obs"</span>, <span class="st">"plumb"</span>]</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(hapGroups_to_plot_PCA_eastside), clusterMembershipWithHets)</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included_temp.Fst_group[.!selection] <span class="op">.=</span> <span class="st">"ignore"</span> <span class="co"># write over the group name so function below won't plot that individual</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>groups_to_plot_PCA_eastside <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>,<span class="st">"plumb_BJ"</span>]</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>group_colors_PCA_eastside <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>,<span class="st">"pink"</span>];</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>PCAmodelHapEast <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included_temp, </span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA_eastside, group_colors_PCA_eastside; </span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers east haps"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelHapEast.PCAfig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-38-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="do-a-pca-based-on-a-same-size-region-elsewhere-on-the-z-with-low-hetvar" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-a-same-size-region-elsewhere-on-the-z-with-low-hetvar">Do a PCA based on a same-size region elsewhere on the Z (with low HetVar):</h3>
<div id="c201cdcb" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get length of region</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>lengthHighHetVarRegion <span class="op">=</span> positionMax <span class="op">-</span> positionMin</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>leftLocus <span class="op">=</span> <span class="fl">1_000_000</span> <span class="co"># start at 1 Mb from left side</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>rightLocus <span class="op">=</span> leftLocus <span class="op">+</span> lengthHighHetVarRegion</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>regionText_lowHetVarRegion <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,leftLocus,<span class="st">" to "</span>,rightLocus)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>lociSelection <span class="op">=</span> (leftLocus <span class="op">.&lt;=</span> pos_region.position <span class="op">.&lt;=</span> rightLocus)</span>
<span id="cb89-9"><a href="#cb89-9" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion <span class="op">=</span> genotypes_region[<span class="op">:</span>, lociSelection]</span>
<span id="cb89-10"><a href="#cb89-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-11"><a href="#cb89-11" aria-hidden="true" tabindex="-1"></a><span class="co"># impute missing genotypes:</span></span>
<span id="cb89-12"><a href="#cb89-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-13"><a href="#cb89-13" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genotypes_lowHetVarRegion))</span>
<span id="cb89-14"><a href="#cb89-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-15"><a href="#cb89-15" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb89-16"><a href="#cb89-16" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb89-17"><a href="#cb89-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-18"><a href="#cb89-18" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genotypes_lowHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb89-19"><a href="#cb89-19" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb89-20"><a href="#cb89-20" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText_lowHetVarRegion,</span>
<span id="cb89-21"><a href="#cb89-21" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb89-22"><a href="#cb89-22" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb89-23"><a href="#cb89-23" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb89-24"><a href="#cb89-24" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb89-25"><a href="#cb89-25" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb89-26"><a href="#cb89-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-27"><a href="#cb89-27" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb89-28"><a href="#cb89-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb89-29"><a href="#cb89-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"FigureS2A_gwZ_nonHLBRarbitrary_from_Julia.png"</span>, PCAmodel.PCAfig, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb89-30"><a href="#cb89-30" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-39-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="do-similar-as-above-but-for-chr-15" class="level2">
<h2 class="anchored" data-anchor-id="do-similar-as-above-but-for-chr-15">Do similar as above but for chr 15:</h2>
<div id="4640fe84" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw15"</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-15"><a href="#cb91-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb91-16"><a href="#cb91-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb91-17"><a href="#cb91-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb91-18"><a href="#cb91-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb91-19"><a href="#cb91-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-20"><a href="#cb91-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb91-21"><a href="#cb91-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-22"><a href="#cb91-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb91-23"><a href="#cb91-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb91-24"><a href="#cb91-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-25"><a href="#cb91-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb91-26"><a href="#cb91-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-27"><a href="#cb91-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb91-28"><a href="#cb91-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-29"><a href="#cb91-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb91-30"><a href="#cb91-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb91-31"><a href="#cb91-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-32"><a href="#cb91-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb91-33"><a href="#cb91-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb91-34"><a href="#cb91-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb91-35"><a href="#cb91-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb91-36"><a href="#cb91-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb91-37"><a href="#cb91-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb91-38"><a href="#cb91-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb91-39"><a href="#cb91-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb91-40"><a href="#cb91-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-41"><a href="#cb91-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb91-42"><a href="#cb91-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-43"><a href="#cb91-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb91-44"><a href="#cb91-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb91-45"><a href="#cb91-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb91-46"><a href="#cb91-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb91-47"><a href="#cb91-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb91-48"><a href="#cb91-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb91-49"><a href="#cb91-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb91-50"><a href="#cb91-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb91-51"><a href="#cb91-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb91-52"><a href="#cb91-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb91-53"><a href="#cb91-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb91-54"><a href="#cb91-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb91-55"><a href="#cb91-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-56"><a href="#cb91-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb91-57"><a href="#cb91-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb91-58"><a href="#cb91-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-59"><a href="#cb91-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb91-60"><a href="#cb91-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.75</span>) </span>
<span id="cb91-61"><a href="#cb91-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-62"><a href="#cb91-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb91-63"><a href="#cb91-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-64"><a href="#cb91-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb91-65"><a href="#cb91-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb91-66"><a href="#cb91-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb91-67"><a href="#cb91-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb91-68"><a href="#cb91-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb91-69"><a href="#cb91-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb91-70"><a href="#cb91-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb91-71"><a href="#cb91-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb91-72"><a href="#cb91-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb91-73"><a href="#cb91-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb91-74"><a href="#cb91-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb91-75"><a href="#cb91-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-40-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-40-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="40">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-1" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-1">Save the individual colors in the metadata</h3>
<div id="3953b9f9" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-1" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-1">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="fc8bab87" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-42-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="42">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-1" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-1">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="415d9483" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-43-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="43">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-15-high-hetvar-region-there-are-only-5-clear-haplogroups-pc3-does-not-distinguish-vir-and-lud.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-15-high-hetvar-region-there-are-only-5-clear-haplogroups-pc3-does-not-distinguish-vir-and-lud.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 15 high HetVar region, there are only 5 clear haplogroups (PC3 does not distinguish vir and lud). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="0a689e04" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"green"</span>,</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellowgreen"</span>,</span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span> </span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3</span>) <span class="op">.&amp;</span> </span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">7</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-33"><a href="#cb104-33" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb104-34"><a href="#cb104-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-35"><a href="#cb104-35" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb104-36"><a href="#cb104-36" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb104-37"><a href="#cb104-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-38"><a href="#cb104-38" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb104-39"><a href="#cb104-39" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb104-40"><a href="#cb104-40" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb104-41"><a href="#cb104-41" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb104-42"><a href="#cb104-42" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb104-43"><a href="#cb104-43" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb104-44"><a href="#cb104-44" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb104-45"><a href="#cb104-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-46"><a href="#cb104-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb104-47"><a href="#cb104-47" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb104-48"><a href="#cb104-48" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb104-49"><a href="#cb104-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-50"><a href="#cb104-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb104-51"><a href="#cb104-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-52"><a href="#cb104-52" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb104-53"><a href="#cb104-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-54"><a href="#cb104-54" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb104-55"><a href="#cb104-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-56"><a href="#cb104-56" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb104-57"><a href="#cb104-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-58"><a href="#cb104-58" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb104-59"><a href="#cb104-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-60"><a href="#cb104-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb104-61"><a href="#cb104-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-62"><a href="#cb104-62" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb104-63"><a href="#cb104-63" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb104-64"><a href="#cb104-64" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb104-65"><a href="#cb104-65" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb104-66"><a href="#cb104-66" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb104-67"><a href="#cb104-67" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.00892738</span></span>
<span id="cb104-68"><a href="#cb104-68" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00677711</span></span>
<span id="cb104-69"><a href="#cb104-69" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00725483</span></span>
<span id="cb104-70"><a href="#cb104-70" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.0083953</span></span>
<span id="cb104-71"><a href="#cb104-71" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00673292 =#</span></span>
<span id="cb104-72"><a href="#cb104-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-73"><a href="#cb104-73" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb104-74"><a href="#cb104-74" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb104-75"><a href="#cb104-75" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb104-76"><a href="#cb104-76" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb104-77"><a href="#cb104-77" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb104-78"><a href="#cb104-78" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.032793</span></span>
<span id="cb104-79"><a href="#cb104-79" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0326016</span></span>
<span id="cb104-80"><a href="#cb104-80" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0334515</span></span>
<span id="cb104-81"><a href="#cb104-81" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.041869</span></span>
<span id="cb104-82"><a href="#cb104-82" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0389012</span></span>
<span id="cb104-83"><a href="#cb104-83" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0393449</span></span>
<span id="cb104-84"><a href="#cb104-84" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0476769</span></span>
<span id="cb104-85"><a href="#cb104-85" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0150895</span></span>
<span id="cb104-86"><a href="#cb104-86" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0294242</span></span>
<span id="cb104-87"><a href="#cb104-87" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0297807 =#</span></span>
<span id="cb104-88"><a href="#cb104-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-89"><a href="#cb104-89" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb104-90"><a href="#cb104-90" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb104-91"><a href="#cb104-91" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb104-92"><a href="#cb104-92" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb104-93"><a href="#cb104-93" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb104-94"><a href="#cb104-94" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb104-95"><a href="#cb104-95" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb104-96"><a href="#cb104-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-97"><a href="#cb104-97" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb104-98"><a href="#cb104-98" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb104-99"><a href="#cb104-99" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb104-100"><a href="#cb104-100" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb104-101"><a href="#cb104-101" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb104-102"><a href="#cb104-102" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb104-103"><a href="#cb104-103" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb104-104"><a href="#cb104-104" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb104-105"><a href="#cb104-105" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb104-106"><a href="#cb104-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-107"><a href="#cb104-107" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb104-108"><a href="#cb104-108" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb104-109"><a href="#cb104-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-110"><a href="#cb104-110" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb104-111"><a href="#cb104-111" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb104-112"><a href="#cb104-112" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb104-113"><a href="#cb104-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-114"><a href="#cb104-114" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb104-115"><a href="#cb104-115" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb104-116"><a href="#cb104-116" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb104-117"><a href="#cb104-117" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [78 2 71 5 70] and the sum of those is 226
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-44-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="44">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 1 … 0 0; … ; 0 2 … 2 0; 0 2 … 2 0], [5282897, 5282956, 5309350, 5337401, 5354821, 5354857, 5490417, 5522381, 5528891, 5684034  …  7539351, 7539368, 7571024, 7598479, 7598515, 7620519, 7678099, 7790537, 7813358, 7832029], <span class="ansi-bold">52×20 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   5 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l ⋯
   6 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   7 │ GW_Armando_plate1_TTGW59      GW_Armando_plate1_TTGW59      NG        l
   8 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   9 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
  10 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  11 │ GW_Armando_plate2_TTGW09      GW_Armando_plate2_TTGW09      SU        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  43 │ GW_Armando_plate1_JF13G01     GW_Armando_plate1_JF13G01     ST        p
  44 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p ⋯
  45 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p
  46 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  47 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  48 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p ⋯
  49 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  50 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  51 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  52 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p ⋯
<span class="ansi-cyan-fg">                                                  17 columns and 31 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-1" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-1">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="4a0d04c4" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb107-10"><a href="#cb107-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb107-11"><a href="#cb107-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb107-12"><a href="#cb107-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb107-13"><a href="#cb107-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb107-14"><a href="#cb107-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb107-15"><a href="#cb107-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb107-16"><a href="#cb107-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-17"><a href="#cb107-17" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb107-18"><a href="#cb107-18" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb107-19"><a href="#cb107-19" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb107-20"><a href="#cb107-20" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb107-21"><a href="#cb107-21" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5.5</span>) <span class="op">.&amp;</span></span>
<span id="cb107-22"><a href="#cb107-22" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb107-23"><a href="#cb107-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-24"><a href="#cb107-24" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit virLud_troch troch obs plumb vir_plumb]</span>
<span id="cb107-25"><a href="#cb107-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-26"><a href="#cb107-26" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb107-27"><a href="#cb107-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-28"><a href="#cb107-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb107-29"><a href="#cb107-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb107-30"><a href="#cb107-30" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb107-31"><a href="#cb107-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb107-32"><a href="#cb107-32" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb107-33"><a href="#cb107-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-34"><a href="#cb107-34" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb107-35"><a href="#cb107-35" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb107-36"><a href="#cb107-36" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb107-37"><a href="#cb107-37" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb107-38"><a href="#cb107-38" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb107-39"><a href="#cb107-39" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb107-40"><a href="#cb107-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb107-41"><a href="#cb107-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-42"><a href="#cb107-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb107-43"><a href="#cb107-43" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb107-44"><a href="#cb107-44" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb107-45"><a href="#cb107-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-46"><a href="#cb107-46" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb107-47"><a href="#cb107-47" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb107-48"><a href="#cb107-48" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb107-49"><a href="#cb107-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-50"><a href="#cb107-50" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb107-51"><a href="#cb107-51" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb107-52"><a href="#cb107-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-53"><a href="#cb107-53" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb107-54"><a href="#cb107-54" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb107-55"><a href="#cb107-55" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb107-56"><a href="#cb107-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-57"><a href="#cb107-57" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb107-58"><a href="#cb107-58" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb107-59"><a href="#cb107-59" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb107-60"><a href="#cb107-60" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb107-61"><a href="#cb107-61" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-45-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="45">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 1 … 0 0; … ; 0 1 … 1 0; 0 1 … 2 0], [5282897, 5282956, 5309350, 5337401, 5354821, 5354857, 5490417, 5522381, 5528891, 5684034  …  7539351, 7539368, 7571024, 7598479, 7598515, 7620519, 7678099, 7790537, 7813358, 7832029], <span class="ansi-bold">70×20 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   5 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l ⋯
   6 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   7 │ GW_Armando_plate1_TTGW59      GW_Armando_plate1_TTGW59      NG        l
   8 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   9 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
  10 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  11 │ GW_Armando_plate2_TTGW09      GW_Armando_plate2_TTGW09      SU        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  61 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  62 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p ⋯
  63 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  64 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  65 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  66 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
  67 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p
  68 │ GW_Armando_plate2_JF07G01     GW_Armando_plate2_JF07G01     ST        p
  69 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
  70 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                  17 columns and 49 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-1" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-1">Show just the west area (without nitidus)</h3>
<div id="a9ce5293" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-9"><a href="#cb110-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb110-10"><a href="#cb110-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb110-11"><a href="#cb110-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb110-12"><a href="#cb110-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb110-13"><a href="#cb110-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb110-14"><a href="#cb110-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb110-15"><a href="#cb110-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb110-16"><a href="#cb110-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb110-17"><a href="#cb110-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb110-18"><a href="#cb110-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb110-19"><a href="#cb110-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb110-20"><a href="#cb110-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-21"><a href="#cb110-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb110-22"><a href="#cb110-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-23"><a href="#cb110-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb110-24"><a href="#cb110-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb110-25"><a href="#cb110-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb110-26"><a href="#cb110-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-27"><a href="#cb110-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb110-28"><a href="#cb110-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb110-29"><a href="#cb110-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb110-30"><a href="#cb110-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb110-31"><a href="#cb110-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-46-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="46">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 1 … 0 0; … ; 2 2 … -1 2; 0 2 … 2 -1], [5282897, 5282956, 5309350, 5354821, 5522381, 5692997, 5693003, 5697689, 5774179, 5777672  …  7339982, 7371724, 7371744, 7539368, 7598479, 7598515, 7620519, 7678099, 7790537, 7813358], <span class="ansi-bold">177×20 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   5 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l ⋯
   6 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l
   7 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   8 │ GW_Armando_plate1_TTGW59      GW_Armando_plate1_TTGW59      NG        l
   9 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l ⋯
  10 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
  11 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 168 │ GW_Lane5_PT2                  GW_Lane5_PT2                  ML        l
 169 │ GW_Liz_GBS_Liz5190            GW_Liz_GBS_Liz5190            ML        l ⋯
 170 │ GW_Liz_GBS_Liz6008            GW_Liz_GBS_Liz6008            ML        l
 171 │ GW_Liz_GBS_Liz6478            GW_Liz_GBS_Liz6478            ML        l
 172 │ GW_Lane5_LN12                 GW_Lane5_LN12                 LN        t
 173 │ GW_Liz_GBS_Liz5178            GW_Liz_GBS_Liz5178            ML        l ⋯
 174 │ GW_Liz_GBS_Liz5179            GW_Liz_GBS_Liz5179            ML        l
 175 │ GW_Liz_GBS_Liz5194            GW_Liz_GBS_Liz5194            ML        l
 176 │ GW_Lane5_MN1                  GW_Lane5_MN1                  MN        t
 177 │ GW_Liz_GBS_Liz5144            GW_Liz_GBS_Liz5144            ML        l ⋯
<span class="ansi-cyan-fg">                                                 17 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-1" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-1">Show just the east area</h3>
<div id="6f1940ff" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb113-10"><a href="#cb113-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb113-11"><a href="#cb113-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb113-12"><a href="#cb113-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb113-13"><a href="#cb113-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb113-14"><a href="#cb113-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb113-15"><a href="#cb113-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb113-16"><a href="#cb113-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb113-17"><a href="#cb113-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb113-18"><a href="#cb113-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb113-19"><a href="#cb113-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb113-20"><a href="#cb113-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-21"><a href="#cb113-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb113-22"><a href="#cb113-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-23"><a href="#cb113-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb113-24"><a href="#cb113-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb113-25"><a href="#cb113-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb113-26"><a href="#cb113-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-27"><a href="#cb113-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb113-28"><a href="#cb113-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb113-29"><a href="#cb113-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb113-30"><a href="#cb113-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb113-31"><a href="#cb113-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-47-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="47">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[2 1 … 0 1; 0 0 … 0 1; … ; 2 2 … 0 2; 2 2 … 1 2], [5282897, 5309350, 5337401, 5490417, 5522381, 5528891, 5684034, 5692997, 5693003, 5693043  …  7120357, 7161643, 7189896, 7286852, 7293716, 7374312, 7414344, 7539351, 7571024, 7598515], <span class="ansi-bold">146×20 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                        </span><span class="ansi-bold"> ID                         </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group</span> ⋯
     │<span class="ansi-bright-black-fg"> String                     </span><span class="ansi-bright-black-fg"> String                     </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> Strin</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW01    GW_Armando_plate1_TTGW01    MN        troch ⋯
   2 │ GW_Armando_plate1_TTGW05r2  GW_Armando_plate1_TTGW05r2  MN        troch
   3 │ GW_Armando_plate1_TTGW63    GW_Armando_plate1_TTGW63    SP        lud_S
   4 │ GW_Armando_plate1_TTGW65    GW_Armando_plate1_TTGW65    SP        lud_S
   5 │ GW_Armando_plate1_TTGW70    GW_Armando_plate1_TTGW70    SA        lud_S ⋯
   6 │ GW_Armando_plate1_TTGW72    GW_Armando_plate1_TTGW72    SA        lud_S
   7 │ GW_Armando_plate1_TTGW78    GW_Armando_plate1_TTGW78    SA        lud_S
   8 │ GW_Armando_plate2_LN2       GW_Armando_plate2_LN2       LN        troch
   9 │ GW_Armando_plate2_TTGW03    GW_Armando_plate2_TTGW03    MN        troch ⋯
  10 │ GW_Armando_plate2_TTGW61    GW_Armando_plate2_TTGW61    SP        lud_S
  11 │ GW_Armando_plate2_TTGW62    GW_Armando_plate2_TTGW62    SP        lud_S
  ⋮  │             ⋮                           ⋮                  ⋮            ⋱
 137 │ GW_Lane5_SL2                GW_Lane5_SL2                SL        plumb
 138 │ GW_Lane5_ST12               GW_Lane5_ST12               ST        plumb ⋯
 139 │ GW_Lane5_ST3                GW_Lane5_ST3                ST        plumb
 140 │ GW_Lane5_TA1                GW_Lane5_TA1                TA        plumb
 141 │ GW_Armando_plate1_JF09G01   GW_Armando_plate1_JF09G01   ST        plumb
 142 │ GW_Armando_plate1_JG02G02   GW_Armando_plate1_JG02G02   PR        plumb ⋯
 143 │ GW_Armando_plate1_JG12G01   GW_Armando_plate1_JG12G01   ST        plumb
 144 │ GW_Armando_plate2_JF15G02   GW_Armando_plate2_JF15G02   DV        plumb
 145 │ GW_Lane5_AN1                GW_Lane5_AN1                AN        plumb
 146 │ GW_Lane5_UY6                GW_Lane5_UY6                UY        plumb ⋯
<span class="ansi-cyan-fg">                                                 17 columns and 125 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-1" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-1">Show just the northern area</h3>
<div id="844729c0" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb116-20"><a href="#cb116-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-21"><a href="#cb116-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb116-22"><a href="#cb116-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-23"><a href="#cb116-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb116-24"><a href="#cb116-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb116-25"><a href="#cb116-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb116-26"><a href="#cb116-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-27"><a href="#cb116-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb116-28"><a href="#cb116-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb116-29"><a href="#cb116-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb116-30"><a href="#cb116-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb116-31"><a href="#cb116-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-48-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="48">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 1 0 … 0 0; … ; 1 1 … 1 1; 1 1 … 1 2], [5282956, 5354821, 5490417, 5528891, 5684034, 5697689, 5701683, 5704379, 5774179, 5777672  …  7371744, 7414344, 7539351, 7539368, 7571024, 7598479, 7620519, 7678099, 7790537, 7813358], <span class="ansi-bold">151×20 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   5 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l ⋯
   6 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   7 │ GW_Armando_plate1_TTGW59      GW_Armando_plate1_TTGW59      NG        l
   8 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   9 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l ⋯
  10 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
  11 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 142 │ GW_Lane5_BK3                  GW_Lane5_BK3                  BK        p
 143 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p ⋯
 144 │ GW_Lane5_ST12                 GW_Lane5_ST12                 ST        p
 145 │ GW_Lane5_ST3                  GW_Lane5_ST3                  ST        p
 146 │ GW_Lane5_UY2                  GW_Lane5_UY2                  UY        p
 147 │ GW_Lane5_UY6                  GW_Lane5_UY6                  UY        p ⋯
 148 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 149 │ GW_Armando_plate2_JF07G01     GW_Armando_plate2_JF07G01     ST        p
 150 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
 151 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                 17 columns and 130 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="now-do-the-same-for-chr-28" class="level2">
<h2 class="anchored" data-anchor-id="now-do-the-same-for-chr-28">Now do the same for chr 28</h2>
<div id="b59f2cbd" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw28"</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-12"><a href="#cb119-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb119-13"><a href="#cb119-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb119-14"><a href="#cb119-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-15"><a href="#cb119-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb119-16"><a href="#cb119-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb119-17"><a href="#cb119-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb119-18"><a href="#cb119-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb119-19"><a href="#cb119-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-20"><a href="#cb119-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb119-21"><a href="#cb119-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-22"><a href="#cb119-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb119-23"><a href="#cb119-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb119-24"><a href="#cb119-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-25"><a href="#cb119-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb119-26"><a href="#cb119-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-27"><a href="#cb119-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb119-28"><a href="#cb119-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-29"><a href="#cb119-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb119-30"><a href="#cb119-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb119-31"><a href="#cb119-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-32"><a href="#cb119-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb119-33"><a href="#cb119-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb119-34"><a href="#cb119-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb119-35"><a href="#cb119-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb119-36"><a href="#cb119-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb119-37"><a href="#cb119-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb119-38"><a href="#cb119-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb119-39"><a href="#cb119-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb119-40"><a href="#cb119-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-41"><a href="#cb119-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb119-42"><a href="#cb119-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-43"><a href="#cb119-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb119-44"><a href="#cb119-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb119-45"><a href="#cb119-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb119-46"><a href="#cb119-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb119-47"><a href="#cb119-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb119-48"><a href="#cb119-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb119-49"><a href="#cb119-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb119-50"><a href="#cb119-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb119-51"><a href="#cb119-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb119-52"><a href="#cb119-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb119-53"><a href="#cb119-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb119-54"><a href="#cb119-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb119-55"><a href="#cb119-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-56"><a href="#cb119-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb119-57"><a href="#cb119-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb119-58"><a href="#cb119-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-59"><a href="#cb119-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb119-60"><a href="#cb119-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">2</span>) </span>
<span id="cb119-61"><a href="#cb119-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-62"><a href="#cb119-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb119-63"><a href="#cb119-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-64"><a href="#cb119-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb119-65"><a href="#cb119-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb119-66"><a href="#cb119-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb119-67"><a href="#cb119-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb119-68"><a href="#cb119-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb119-69"><a href="#cb119-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb119-70"><a href="#cb119-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb119-71"><a href="#cb119-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb119-72"><a href="#cb119-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb119-73"><a href="#cb119-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb119-74"><a href="#cb119-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb119-75"><a href="#cb119-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-49-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-49-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="49">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-2" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-2">Save the individual colors in the metadata</h3>
<div id="e440066a" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-2" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-2">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="8b66ec24" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb126-6"><a href="#cb126-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb126-7"><a href="#cb126-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb126-8"><a href="#cb126-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb126-9"><a href="#cb126-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb126-10"><a href="#cb126-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb126-11"><a href="#cb126-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb126-12"><a href="#cb126-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-51-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="51">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-2" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-2">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="2b2f9dfc" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb129-11"><a href="#cb129-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb129-12"><a href="#cb129-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-52-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="52">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-28-high-hetvar-region-there-are-only-5-clear-haplogroups-yellow-and-orange-are-close.-vir-and-lud-dont-distinguish-along-pc3-as-nitidus-varies-there.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-28-high-hetvar-region-there-are-only-5-clear-haplogroups-yellow-and-orange-are-close.-vir-and-lud-dont-distinguish-along-pc3-as-nitidus-varies-there.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 28 high HetVar region, there are only 5 clear haplogroups (yellow and orange are close). Vir and lud don’t distinguish along PC3, as nitidus varies there. Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="68006ed7" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb132-9"><a href="#cb132-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellowgreen"</span>,</span>
<span id="cb132-10"><a href="#cb132-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb132-11"><a href="#cb132-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb132-12"><a href="#cb132-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-13"><a href="#cb132-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb132-14"><a href="#cb132-14" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb132-15"><a href="#cb132-15" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span> </span>
<span id="cb132-16"><a href="#cb132-16" aria-hidden="true" tabindex="-1"></a>        (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3.2</span>) <span class="op">.&amp;</span> </span>
<span id="cb132-17"><a href="#cb132-17" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb132-18"><a href="#cb132-18" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span> </span>
<span id="cb132-19"><a href="#cb132-19" aria-hidden="true" tabindex="-1"></a>        (<span class="op">-</span><span class="fl">3.2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb132-20"><a href="#cb132-20" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb132-21"><a href="#cb132-21" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb132-22"><a href="#cb132-22" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.PC2) <span class="op">.&amp;</span></span>
<span id="cb132-23"><a href="#cb132-23" aria-hidden="true" tabindex="-1"></a>        indSelection_lowIndHetStan</span>
<span id="cb132-24"><a href="#cb132-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-25"><a href="#cb132-25" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb132-26"><a href="#cb132-26" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb132-27"><a href="#cb132-27" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb132-28"><a href="#cb132-28" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb132-29"><a href="#cb132-29" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb132-30"><a href="#cb132-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb132-31"><a href="#cb132-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-32"><a href="#cb132-32" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb132-33"><a href="#cb132-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-34"><a href="#cb132-34" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb132-35"><a href="#cb132-35" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb132-36"><a href="#cb132-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-37"><a href="#cb132-37" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb132-38"><a href="#cb132-38" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb132-39"><a href="#cb132-39" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb132-40"><a href="#cb132-40" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb132-41"><a href="#cb132-41" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb132-42"><a href="#cb132-42" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb132-43"><a href="#cb132-43" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb132-44"><a href="#cb132-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-45"><a href="#cb132-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb132-46"><a href="#cb132-46" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb132-47"><a href="#cb132-47" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb132-48"><a href="#cb132-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-49"><a href="#cb132-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb132-50"><a href="#cb132-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-51"><a href="#cb132-51" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb132-52"><a href="#cb132-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-53"><a href="#cb132-53" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb132-54"><a href="#cb132-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-55"><a href="#cb132-55" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb132-56"><a href="#cb132-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-57"><a href="#cb132-57" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb132-58"><a href="#cb132-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-59"><a href="#cb132-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb132-60"><a href="#cb132-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-61"><a href="#cb132-61" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb132-62"><a href="#cb132-62" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb132-63"><a href="#cb132-63" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb132-64"><a href="#cb132-64" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb132-65"><a href="#cb132-65" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb132-66"><a href="#cb132-66" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.00792304</span></span>
<span id="cb132-67"><a href="#cb132-67" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00320189</span></span>
<span id="cb132-68"><a href="#cb132-68" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00734994</span></span>
<span id="cb132-69"><a href="#cb132-69" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.0101536</span></span>
<span id="cb132-70"><a href="#cb132-70" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00270239 =#</span></span>
<span id="cb132-71"><a href="#cb132-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-72"><a href="#cb132-72" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb132-73"><a href="#cb132-73" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb132-74"><a href="#cb132-74" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb132-75"><a href="#cb132-75" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb132-76"><a href="#cb132-76" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb132-77"><a href="#cb132-77" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.0334156</span></span>
<span id="cb132-78"><a href="#cb132-78" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0318841</span></span>
<span id="cb132-79"><a href="#cb132-79" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0351279</span></span>
<span id="cb132-80"><a href="#cb132-80" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0330054</span></span>
<span id="cb132-81"><a href="#cb132-81" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0314387</span></span>
<span id="cb132-82"><a href="#cb132-82" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0344624</span></span>
<span id="cb132-83"><a href="#cb132-83" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0307517</span></span>
<span id="cb132-84"><a href="#cb132-84" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0188902</span></span>
<span id="cb132-85"><a href="#cb132-85" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0234771</span></span>
<span id="cb132-86"><a href="#cb132-86" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0265753 =#</span></span>
<span id="cb132-87"><a href="#cb132-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-88"><a href="#cb132-88" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb132-89"><a href="#cb132-89" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb132-90"><a href="#cb132-90" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb132-91"><a href="#cb132-91" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb132-92"><a href="#cb132-92" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb132-93"><a href="#cb132-93" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb132-94"><a href="#cb132-94" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb132-95"><a href="#cb132-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-96"><a href="#cb132-96" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb132-97"><a href="#cb132-97" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb132-98"><a href="#cb132-98" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb132-99"><a href="#cb132-99" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb132-100"><a href="#cb132-100" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb132-101"><a href="#cb132-101" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb132-102"><a href="#cb132-102" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb132-103"><a href="#cb132-103" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb132-104"><a href="#cb132-104" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb132-105"><a href="#cb132-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-106"><a href="#cb132-106" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb132-107"><a href="#cb132-107" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb132-108"><a href="#cb132-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-109"><a href="#cb132-109" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb132-110"><a href="#cb132-110" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb132-111"><a href="#cb132-111" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb132-112"><a href="#cb132-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-113"><a href="#cb132-113" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb132-114"><a href="#cb132-114" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb132-115"><a href="#cb132-115" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb132-116"><a href="#cb132-116" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [74 2 65 3 67] and the sum of those is 211
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-53-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="53">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 2 2; 0 2 … 2 2], [1825835, 1828808, 1835122, 1857465, 1857494, 1871762, 1872060, 1872112, 1872224, 1872436  …  2195993, 2200040, 2239778, 2242803, 2245870, 2264261, 2265267, 2424032, 2430336, 2430428], <span class="ansi-bold">50×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   3 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   4 │ GW_Armando_plate1_TTGW53      GW_Armando_plate1_TTGW53      NG        l
   5 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l ⋯
   6 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l
   7 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   8 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
  11 │ GW_Armando_plate2_TTGW08      GW_Armando_plate2_TTGW08      SU        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  41 │ GW_Armando_plate1_JF12G01     GW_Armando_plate1_JF12G01     ST        p
  42 │ GW_Armando_plate1_JF12G02     GW_Armando_plate1_JF12G02     ST        p ⋯
  43 │ GW_Armando_plate1_JF13G01     GW_Armando_plate1_JF13G01     ST        p
  44 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p
  45 │ GW_Armando_plate1_JF20G01     GW_Armando_plate1_JF20G01     MB        p
  46 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p ⋯
  47 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  48 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  49 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  50 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 29 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-2" class="level2">
<h2 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-2">Now show a GBI plot like above, but with heterozygotes:</h2>
<div id="a7d2522a" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_nit"</span>,</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obsHet"</span>,</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs_plumb"</span>,</span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb135-11"><a href="#cb135-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-12"><a href="#cb135-12" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb135-13"><a href="#cb135-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"slateblue1"</span>,</span>
<span id="cb135-14"><a href="#cb135-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb135-15"><a href="#cb135-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb135-16"><a href="#cb135-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb135-17"><a href="#cb135-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb135-18"><a href="#cb135-18" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"darkgoldenrod1"</span>,</span>
<span id="cb135-19"><a href="#cb135-19" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"darkorange1"</span>,</span>
<span id="cb135-20"><a href="#cb135-20" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb135-21"><a href="#cb135-21" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb135-22"><a href="#cb135-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-23"><a href="#cb135-23" aria-hidden="true" tabindex="-1"></a>virLud_nit <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb135-24"><a href="#cb135-24" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb135-25"><a href="#cb135-25" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb135-26"><a href="#cb135-26" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb135-27"><a href="#cb135-27" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb135-28"><a href="#cb135-28" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb135-29"><a href="#cb135-29" aria-hidden="true" tabindex="-1"></a>obsHet <span class="op">=</span> (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb135-30"><a href="#cb135-30" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb135-31"><a href="#cb135-31" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb135-32"><a href="#cb135-32" aria-hidden="true" tabindex="-1"></a>obs_plumb <span class="op">=</span> (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb135-33"><a href="#cb135-33" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb135-34"><a href="#cb135-34" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb135-35"><a href="#cb135-35" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb135-36"><a href="#cb135-36" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb135-37"><a href="#cb135-37" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb135-38"><a href="#cb135-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-39"><a href="#cb135-39" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb135-40"><a href="#cb135-40" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb135-41"><a href="#cb135-41" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud_nit]</span>
<span id="cb135-42"><a href="#cb135-42" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb135-43"><a href="#cb135-43" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud_troch]</span>
<span id="cb135-44"><a href="#cb135-44" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb135-45"><a href="#cb135-45" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb135-46"><a href="#cb135-46" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obsHet]</span>
<span id="cb135-47"><a href="#cb135-47" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs_plumb]</span>
<span id="cb135-48"><a href="#cb135-48" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb135-49"><a href="#cb135-49" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[vir_plumb]</span>
<span id="cb135-50"><a href="#cb135-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-51"><a href="#cb135-51" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud virLud_nit nit virLud_troch troch obs obsHet obs_plumb plumb vir_plumb]</span>
<span id="cb135-52"><a href="#cb135-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-53"><a href="#cb135-53" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb135-54"><a href="#cb135-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-55"><a href="#cb135-55" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb135-56"><a href="#cb135-56" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb135-57"><a href="#cb135-57" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb135-58"><a href="#cb135-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-59"><a href="#cb135-59" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb135-60"><a href="#cb135-60" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb135-61"><a href="#cb135-61" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb135-62"><a href="#cb135-62" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb135-63"><a href="#cb135-63" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb135-64"><a href="#cb135-64" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb135-65"><a href="#cb135-65" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb135-66"><a href="#cb135-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-67"><a href="#cb135-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb135-68"><a href="#cb135-68" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb135-69"><a href="#cb135-69" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb135-70"><a href="#cb135-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-71"><a href="#cb135-71" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb135-72"><a href="#cb135-72" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb135-73"><a href="#cb135-73" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb135-74"><a href="#cb135-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-75"><a href="#cb135-75" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb135-76"><a href="#cb135-76" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb135-77"><a href="#cb135-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-78"><a href="#cb135-78" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb135-79"><a href="#cb135-79" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb135-80"><a href="#cb135-80" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb135-81"><a href="#cb135-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-82"><a href="#cb135-82" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb135-83"><a href="#cb135-83" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb135-84"><a href="#cb135-84" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb135-85"><a href="#cb135-85" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb135-86"><a href="#cb135-86" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-54-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="54">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; -1 1 … 1 1; 0 1 … 1 1], [1825835, 1828808, 1835122, 1857465, 1857494, 1871762, 1872060, 1872112, 1872224, 1872436  …  2195993, 2200040, 2239778, 2242803, 2245870, 2264261, 2265267, 2424032, 2430336, 2430428], <span class="ansi-bold">74×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   3 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   4 │ GW_Armando_plate1_TTGW53      GW_Armando_plate1_TTGW53      NG        l
   5 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l ⋯
   6 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l
   7 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   8 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
  11 │ GW_Armando_plate2_TTGW08      GW_Armando_plate2_TTGW08      SU        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  65 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p
  66 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p ⋯
  67 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  68 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  69 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  70 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p ⋯
  71 │ GW_Armando_plate2_JF15G02     GW_Armando_plate2_JF15G02     DV        p
  72 │ GW_Armando_plate2_JF27G02     GW_Armando_plate2_JF27G02     ST        p
  73 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p
  74 │ GW_Armando_plate1_JF09G01     GW_Armando_plate1_JF09G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 53 rows omitted</span>)</pre>
</div>
</div>
</div>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-1" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-1">show GBI plot according to original groups and plot order</h3>
<div id="d2829b35" class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb138-6"><a href="#cb138-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-7"><a href="#cb138-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb138-8"><a href="#cb138-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb138-9"><a href="#cb138-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb138-10"><a href="#cb138-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb138-11"><a href="#cb138-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-55-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="55">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 1 … 1 1; -1 1 … 1 1], [1825835, 1828808, 1835122, 1857465, 1857494, 1871762, 1872060, 1872112, 1872224, 1872436  …  2195993, 2200040, 2239778, 2242803, 2245870, 2264261, 2265267, 2424032, 2430336, 2430428], <span class="ansi-bold">74×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   4 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   5 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir     ⋯
   6 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
   7 │ GW_Lane5_TU1               GW_Lane5_TU1               TU        nit
   8 │ GW_Lane5_TU2               GW_Lane5_TU2               TU        nit
   9 │ GW_Armando_plate1_TTGW70   GW_Armando_plate1_TTGW70   SA        lud_Sat ⋯
  10 │ GW_Armando_plate1_TTGW72   GW_Armando_plate1_TTGW72   SA        lud_Sat
  11 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
  65 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  66 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb   ⋯
  67 │ GW_Armando_plate2_JF27G02  GW_Armando_plate2_JF27G02  ST        plumb
  68 │ GW_Armando_plate1_JF09G01  GW_Armando_plate1_JF09G01  ST        plumb
  69 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  70 │ GW_Armando_plate2_JF15G02  GW_Armando_plate2_JF15G02  DV        plumb   ⋯
  71 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  72 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  73 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  74 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 53 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-1" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-1">show same but with all individuals</h3>
<div id="ab495a57" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb140-9"><a href="#cb140-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-10"><a href="#cb140-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb140-11"><a href="#cb140-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb140-12"><a href="#cb140-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb140-13"><a href="#cb140-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb140-14"><a href="#cb140-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-56-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="56">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 2 2; -1 1 … 1 1], [1825835, 1828808, 1835122, 1857465, 1857494, 1871762, 1872060, 1872112, 1872224, 1872436  …  2195993, 2200040, 2239778, 2242803, 2245870, 2264261, 2265267, 2424032, 2430336, 2430428], <span class="ansi-bold">257×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 249 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 252 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 253 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 19 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops">show same but with only vir and plumb pops</h3>
<div id="e5698492" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-20"><a href="#cb142-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb142-21"><a href="#cb142-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-22"><a href="#cb142-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb142-23"><a href="#cb142-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb142-24"><a href="#cb142-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-57-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="57">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 1 … 1 1], [1828808, 1871762, 1872060, 1872224, 1872436, 1877640, 1879000, 1879179, 1879284, 1946507  …  2111032, 2122636, 2188531, 2188581, 2190739, 2195551, 2195993, 2200040, 2430336, 2430428], <span class="ansi-bold">100×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-lud-troch-pops" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-lud-troch-pops">show same but with only vir lud troch pops</h3>
<div id="b01ea6b9" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"troch"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb144-8"><a href="#cb144-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb144-9"><a href="#cb144-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb144-10"><a href="#cb144-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb144-11"><a href="#cb144-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb144-12"><a href="#cb144-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb144-13"><a href="#cb144-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb144-14"><a href="#cb144-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-15"><a href="#cb144-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb144-16"><a href="#cb144-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb144-17"><a href="#cb144-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-18"><a href="#cb144-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb144-19"><a href="#cb144-19" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb144-20"><a href="#cb144-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-21"><a href="#cb144-21" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"vir_S"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>, <span class="st">"troch_west"</span>, <span class="st">"troch_LN"</span>]</span>
<span id="cb144-22"><a href="#cb144-22" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>, <span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb144-23"><a href="#cb144-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-24"><a href="#cb144-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb144-25"><a href="#cb144-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">10</span>, <span class="fu">length</span>(plotGroups))</span>
<span id="cb144-26"><a href="#cb144-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-27"><a href="#cb144-27" aria-hidden="true" tabindex="-1"></a>genosForGBI_limited, indMetadataforGBI_limited <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups,</span>
<span id="cb144-28"><a href="#cb144-28" aria-hidden="true" tabindex="-1"></a>                                            numIndsToPlotWithHets, </span>
<span id="cb144-29"><a href="#cb144-29" aria-hidden="true" tabindex="-1"></a>                                            genosForGBI, metadataForGBI;</span>
<span id="cb144-30"><a href="#cb144-30" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb144-31"><a href="#cb144-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-32"><a href="#cb144-32" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb144-33"><a href="#cb144-33" aria-hidden="true" tabindex="-1"></a>                genosForGBI_limited, indMetadataforGBI_limited, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb144-34"><a href="#cb144-34" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-58-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="58">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 1 … 2 2; 2 0 … 2 2], [1828808, 1857465, 1857494, 1871762, 1872112, 1872224, 1872436, 1877640, 1879179, 1879284  …  2111032, 2122636, 2147415, 2179072, 2188581, 2195551, 2195993, 2239778, 2424032, 2430428], <span class="ansi-bold">74×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   7 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
   8 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   9 │ GW_Lane5_AB2               GW_Lane5_AB2               AB        vir     ⋯
  10 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  11 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  65 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L
  66 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L ⋯
  67 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
  68 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
  69 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
  70 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L ⋯
  71 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
  72 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
  73 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
  74 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 53 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-troch-plumb-pops" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-troch-plumb-pops">show same but with only troch plumb pops</h3>
<div id="7c9bbefe" class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"troch"</span>, <span class="st">"obs"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb146-13"><a href="#cb146-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb146-14"><a href="#cb146-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-15"><a href="#cb146-15" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb146-16"><a href="#cb146-16" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb146-17"><a href="#cb146-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-18"><a href="#cb146-18" aria-hidden="true" tabindex="-1"></a><span class="co"># remove individuals that have vir haplotypes, as this could otherwise be mistaken for introgression from obscuratus:</span></span>
<span id="cb146-19"><a href="#cb146-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-20"><a href="#cb146-20" aria-hidden="true" tabindex="-1"></a>removeTheseInds <span class="op">=</span> [<span class="st">"GW_Armando_plate1_JF24G02"</span>, <span class="co"># gw19 hetero from plumb </span></span>
<span id="cb146-21"><a href="#cb146-21" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF07G03"</span>, <span class="co"># gw19 hetero from plumb</span></span>
<span id="cb146-22"><a href="#cb146-22" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF12G02"</span>, <span class="co"># gw19 hetero from plumb</span></span>
<span id="cb146-23"><a href="#cb146-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF09G01"</span>] <span class="co"># gw28 is hetero from plumb </span></span>
<span id="cb146-24"><a href="#cb146-24" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(removeTheseInds), metadataForGBI.ind)</span>
<span id="cb146-25"><a href="#cb146-25" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> metadataForGBI[.!selection, <span class="op">:</span>]</span>
<span id="cb146-26"><a href="#cb146-26" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genosForGBI[.!selection, <span class="op">:</span>]</span>
<span id="cb146-27"><a href="#cb146-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-28"><a href="#cb146-28" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb146-29"><a href="#cb146-29" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb146-30"><a href="#cb146-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-31"><a href="#cb146-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Set  limit on the number of individuals per group to plot</span></span>
<span id="cb146-32"><a href="#cb146-32" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(plotGroups))</span>
<span id="cb146-33"><a href="#cb146-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-34"><a href="#cb146-34" aria-hidden="true" tabindex="-1"></a><span class="co"># metadataForGBI[metadataForGBI.Fst_group .== "plumb", :]</span></span>
<span id="cb146-35"><a href="#cb146-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-36"><a href="#cb146-36" aria-hidden="true" tabindex="-1"></a>genosForGBI_limited, indMetadataforGBI_limited <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups,</span>
<span id="cb146-37"><a href="#cb146-37" aria-hidden="true" tabindex="-1"></a>                                            numIndsToPlotWithHets, </span>
<span id="cb146-38"><a href="#cb146-38" aria-hidden="true" tabindex="-1"></a>                                            genosForGBI, metadataForGBI;</span>
<span id="cb146-39"><a href="#cb146-39" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb146-40"><a href="#cb146-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-41"><a href="#cb146-41" aria-hidden="true" tabindex="-1"></a><span class="co"># indMetadataforGBI_limited[indMetadataforGBI_limited.Fst_group .== "plumb", :]</span></span>
<span id="cb146-42"><a href="#cb146-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-43"><a href="#cb146-43" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb146-44"><a href="#cb146-44" aria-hidden="true" tabindex="-1"></a>                genosForGBI_limited, indMetadataforGBI_limited, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb146-45"><a href="#cb146-45" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-59-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="59">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[1 1 … 0 0; 1 1 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [1857465, 1857494, 1872060, 1872112, 1879000, 1946511, 1948810, 1963744, 1963988, 1979848  …  2155255, 2179072, 2188531, 2190736, 2190739, 2200040, 2239778, 2245870, 2424032, 2430336], <span class="ansi-bold">37×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L ⋯
   2 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
   3 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
   4 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
   5 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L ⋯
   6 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L
   7 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
   8 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
   9 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L ⋯
  10 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
  11 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
  ⋮  │             ⋮                          ⋮                 ⋮         ⋮    ⋱
  28 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
  29 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
  30 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  31 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  32 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  33 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb   ⋯
  34 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  35 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  36 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  37 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb   ⋯
<span class="ansi-cyan-fg">                                                  19 columns and 16 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="show-just-the-west-area-without-nitidus-2" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-2">Show just the west area (without nitidus)</h2>
<div id="8a9c47a2" class="cell" data-execution_count="59">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-9"><a href="#cb148-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb148-10"><a href="#cb148-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb148-11"><a href="#cb148-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb148-12"><a href="#cb148-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb148-13"><a href="#cb148-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb148-14"><a href="#cb148-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb148-15"><a href="#cb148-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb148-16"><a href="#cb148-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb148-17"><a href="#cb148-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb148-18"><a href="#cb148-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb148-19"><a href="#cb148-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb148-20"><a href="#cb148-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-21"><a href="#cb148-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb148-22"><a href="#cb148-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-23"><a href="#cb148-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb148-24"><a href="#cb148-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb148-25"><a href="#cb148-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb148-26"><a href="#cb148-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-27"><a href="#cb148-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb148-28"><a href="#cb148-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb148-29"><a href="#cb148-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb148-30"><a href="#cb148-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb148-31"><a href="#cb148-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-60-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="60">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 1 … 0 0; … ; 2 0 … 2 2; 2 1 … 2 2], [1828808, 1857465, 1857494, 1871762, 1872112, 1872224, 1872436, 1877640, 1879179, 1879284  …  2111032, 2122636, 2147415, 2179072, 2188581, 2195551, 2195993, 2239778, 2424032, 2430428], <span class="ansi-bold">176×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 167 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
 168 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L ⋯
 169 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
 170 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 171 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 172 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L ⋯
 173 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L
 174 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L
 175 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 176 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 19 columns and 155 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-2" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-east-area-2">Show just the east area</h2>
<div id="14d7ae62" class="cell" data-execution_count="60">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obsHet"</span>,</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs_plumb"</span>,</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"darkgoldenrod1"</span>,</span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"darkorange1"</span>,</span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb151-18"><a href="#cb151-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb151-19"><a href="#cb151-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb151-20"><a href="#cb151-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb151-21"><a href="#cb151-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb151-22"><a href="#cb151-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb151-23"><a href="#cb151-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb151-24"><a href="#cb151-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-25"><a href="#cb151-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb151-26"><a href="#cb151-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-27"><a href="#cb151-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb151-28"><a href="#cb151-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb151-29"><a href="#cb151-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb151-30"><a href="#cb151-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-31"><a href="#cb151-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb151-32"><a href="#cb151-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb151-33"><a href="#cb151-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb151-34"><a href="#cb151-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb151-35"><a href="#cb151-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-61-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="61">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[2 2 … 0 0; 1 1 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [1857465, 1857494, 1872060, 1872112, 1879000, 1946511, 1948810, 1963744, 1963988, 1979848  …  2155255, 2179072, 2188531, 2190736, 2190739, 2200040, 2239778, 2245870, 2424032, 2430336], <span class="ansi-bold">138×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW70   GW_Armando_plate1_TTGW70   SA        lud_Sat ⋯
   2 │ GW_Armando_plate1_TTGW72   GW_Armando_plate1_TTGW72   SA        lud_Sat
   3 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
   4 │ GW_Armando_plate2_TTGW73   GW_Armando_plate2_TTGW73   SA        lud_Sat
   5 │ GW_Armando_plate2_TTGW75   GW_Armando_plate2_TTGW75   SA        lud_Sat ⋯
   6 │ GW_Armando_plate2_TTGW77   GW_Armando_plate2_TTGW77   SA        lud_Sat
   7 │ GW_Armando_plate2_TTGW80   GW_Armando_plate2_TTGW80   SA        lud_Sat
   8 │ GW_Armando_plate2_TTGW79   GW_Armando_plate2_TTGW79   SA        lud_Sat
   9 │ GW_Armando_plate1_TTGW55   GW_Armando_plate1_TTGW55   NG        lud_Nai ⋯
  10 │ GW_Armando_plate1_TTGW59   GW_Armando_plate1_TTGW59   NG        lud_Nai
  11 │ GW_Lane5_PT2               GW_Lane5_PT2               ML        lud_ML
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
 129 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
 130 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb   ⋯
 131 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 132 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 133 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 134 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb   ⋯
 135 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 136 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 137 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 138 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb   ⋯
<span class="ansi-cyan-fg">                                                 19 columns and 117 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-2" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-northern-area-2">Show just the northern area</h2>
<div id="630b7bfb" class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb154-4"><a href="#cb154-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-5"><a href="#cb154-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb154-6"><a href="#cb154-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb154-7"><a href="#cb154-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb154-8"><a href="#cb154-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-9"><a href="#cb154-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb154-10"><a href="#cb154-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb154-11"><a href="#cb154-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb154-12"><a href="#cb154-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb154-13"><a href="#cb154-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb154-14"><a href="#cb154-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb154-15"><a href="#cb154-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb154-16"><a href="#cb154-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb154-17"><a href="#cb154-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb154-18"><a href="#cb154-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb154-19"><a href="#cb154-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb154-20"><a href="#cb154-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-21"><a href="#cb154-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb154-22"><a href="#cb154-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-23"><a href="#cb154-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb154-24"><a href="#cb154-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb154-25"><a href="#cb154-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb154-26"><a href="#cb154-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-27"><a href="#cb154-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb154-28"><a href="#cb154-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb154-29"><a href="#cb154-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb154-30"><a href="#cb154-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb154-31"><a href="#cb154-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-62-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="62">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 1 … 1 1], [1828808, 1871762, 1872060, 1872224, 1872436, 1877640, 1879000, 1879179, 1879284, 1946507  …  2111032, 2122636, 2188531, 2188581, 2190739, 2195551, 2195993, 2200040, 2430336, 2430428], <span class="ansi-bold">146×22 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 137 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 138 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 139 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 140 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 141 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 142 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 143 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 144 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 145 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 146 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 19 columns and 125 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-do-the-same-for-chr-26" class="level2">
<h2 class="anchored" data-anchor-id="now-do-the-same-for-chr-26">Now do the same for chr 26</h2>
<div id="6bfe801f" class="cell" data-execution_count="62">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw26"</span></span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb157-7"><a href="#cb157-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb157-8"><a href="#cb157-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb157-9"><a href="#cb157-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb157-10"><a href="#cb157-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb157-11"><a href="#cb157-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-12"><a href="#cb157-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb157-13"><a href="#cb157-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb157-14"><a href="#cb157-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-15"><a href="#cb157-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb157-16"><a href="#cb157-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb157-17"><a href="#cb157-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb157-18"><a href="#cb157-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb157-19"><a href="#cb157-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-20"><a href="#cb157-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb157-21"><a href="#cb157-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-22"><a href="#cb157-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb157-23"><a href="#cb157-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb157-24"><a href="#cb157-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-25"><a href="#cb157-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb157-26"><a href="#cb157-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-27"><a href="#cb157-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb157-28"><a href="#cb157-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-29"><a href="#cb157-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb157-30"><a href="#cb157-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb157-31"><a href="#cb157-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-32"><a href="#cb157-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb157-33"><a href="#cb157-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb157-34"><a href="#cb157-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb157-35"><a href="#cb157-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb157-36"><a href="#cb157-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb157-37"><a href="#cb157-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb157-38"><a href="#cb157-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb157-39"><a href="#cb157-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb157-40"><a href="#cb157-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-41"><a href="#cb157-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb157-42"><a href="#cb157-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-43"><a href="#cb157-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb157-44"><a href="#cb157-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb157-45"><a href="#cb157-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb157-46"><a href="#cb157-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb157-47"><a href="#cb157-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb157-48"><a href="#cb157-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb157-49"><a href="#cb157-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb157-50"><a href="#cb157-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb157-51"><a href="#cb157-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb157-52"><a href="#cb157-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb157-53"><a href="#cb157-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb157-54"><a href="#cb157-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb157-55"><a href="#cb157-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-56"><a href="#cb157-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb157-57"><a href="#cb157-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb157-58"><a href="#cb157-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-59"><a href="#cb157-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb157-60"><a href="#cb157-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.5</span>) </span>
<span id="cb157-61"><a href="#cb157-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-62"><a href="#cb157-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb157-63"><a href="#cb157-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-64"><a href="#cb157-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb157-65"><a href="#cb157-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb157-66"><a href="#cb157-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb157-67"><a href="#cb157-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb157-68"><a href="#cb157-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb157-69"><a href="#cb157-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb157-70"><a href="#cb157-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb157-71"><a href="#cb157-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb157-72"><a href="#cb157-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb157-73"><a href="#cb157-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb157-74"><a href="#cb157-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb157-75"><a href="#cb157-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-63-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-63-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="63">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-3" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-3">Save the individual colors in the metadata</h3>
<div id="39f3be1a" class="cell" data-execution_count="63">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="64">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-3" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-3">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="41482589" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb164-5"><a href="#cb164-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb164-6"><a href="#cb164-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb164-7"><a href="#cb164-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb164-8"><a href="#cb164-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb164-9"><a href="#cb164-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb164-10"><a href="#cb164-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb164-11"><a href="#cb164-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb164-12"><a href="#cb164-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-65-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="65">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-3" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-3">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="75cdda84" class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb167-6"><a href="#cb167-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb167-7"><a href="#cb167-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb167-8"><a href="#cb167-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb167-9"><a href="#cb167-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb167-10"><a href="#cb167-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb167-11"><a href="#cb167-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb167-12"><a href="#cb167-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-66-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="66">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
</section>
<section id="at-chr-26-high-hetvar-region-there-are-only-5-clear-haplogroups-one-green-is-somewhat-away.-vir-and-lud-vary-along-pc3-but-cannot-be-clearly-grouped.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level2">
<h2 class="anchored" data-anchor-id="at-chr-26-high-hetvar-region-there-are-only-5-clear-haplogroups-one-green-is-somewhat-away.-vir-and-lud-vary-along-pc3-but-cannot-be-clearly-grouped.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 26 high HetVar region, there are only 5 clear haplogroups (one green is somewhat away). Vir and lud vary along PC3 but cannot be clearly grouped. Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h2>
<div id="c10860db" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect chromosome 26 PCA of low IndHet (&lt; 1.5) individuals,</span></span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and specify group boundaries:</span></span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-10"><a href="#cb170-10" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb170-11"><a href="#cb170-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb170-12"><a href="#cb170-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb170-13"><a href="#cb170-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb170-14"><a href="#cb170-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb170-15"><a href="#cb170-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-16"><a href="#cb170-16" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5.5</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb170-17"><a href="#cb170-17" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb170-18"><a href="#cb170-18" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">6</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb170-19"><a href="#cb170-19" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="op">-</span><span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb170-20"><a href="#cb170-20" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.PC2) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb170-21"><a href="#cb170-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-22"><a href="#cb170-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb170-23"><a href="#cb170-23" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb170-24"><a href="#cb170-24" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb170-25"><a href="#cb170-25" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb170-26"><a href="#cb170-26" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb170-27"><a href="#cb170-27" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb170-28"><a href="#cb170-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-29"><a href="#cb170-29" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb170-30"><a href="#cb170-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-31"><a href="#cb170-31" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb170-32"><a href="#cb170-32" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb170-33"><a href="#cb170-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-34"><a href="#cb170-34" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb170-35"><a href="#cb170-35" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb170-36"><a href="#cb170-36" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb170-37"><a href="#cb170-37" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb170-38"><a href="#cb170-38" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb170-39"><a href="#cb170-39" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb170-40"><a href="#cb170-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb170-41"><a href="#cb170-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-42"><a href="#cb170-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb170-43"><a href="#cb170-43" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb170-44"><a href="#cb170-44" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb170-45"><a href="#cb170-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-46"><a href="#cb170-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb170-47"><a href="#cb170-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-48"><a href="#cb170-48" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb170-49"><a href="#cb170-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-50"><a href="#cb170-50" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb170-51"><a href="#cb170-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-52"><a href="#cb170-52" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb170-53"><a href="#cb170-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-54"><a href="#cb170-54" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb170-55"><a href="#cb170-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-56"><a href="#cb170-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb170-57"><a href="#cb170-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-58"><a href="#cb170-58" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb170-59"><a href="#cb170-59" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb170-60"><a href="#cb170-60" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb170-61"><a href="#cb170-61" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb170-62"><a href="#cb170-62" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb170-63"><a href="#cb170-63" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.0135205</span></span>
<span id="cb170-64"><a href="#cb170-64" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00548557</span></span>
<span id="cb170-65"><a href="#cb170-65" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00975861</span></span>
<span id="cb170-66"><a href="#cb170-66" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.00902527</span></span>
<span id="cb170-67"><a href="#cb170-67" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00510553 =#</span></span>
<span id="cb170-68"><a href="#cb170-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-69"><a href="#cb170-69" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb170-70"><a href="#cb170-70" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb170-71"><a href="#cb170-71" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb170-72"><a href="#cb170-72" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb170-73"><a href="#cb170-73" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb170-74"><a href="#cb170-74" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.0243846</span></span>
<span id="cb170-75"><a href="#cb170-75" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0324256</span></span>
<span id="cb170-76"><a href="#cb170-76" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0332749</span></span>
<span id="cb170-77"><a href="#cb170-77" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0390193</span></span>
<span id="cb170-78"><a href="#cb170-78" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0341654</span></span>
<span id="cb170-79"><a href="#cb170-79" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0344734</span></span>
<span id="cb170-80"><a href="#cb170-80" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0403857</span></span>
<span id="cb170-81"><a href="#cb170-81" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0176458</span></span>
<span id="cb170-82"><a href="#cb170-82" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0296574</span></span>
<span id="cb170-83"><a href="#cb170-83" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0300157 =#</span></span>
<span id="cb170-84"><a href="#cb170-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-85"><a href="#cb170-85" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb170-86"><a href="#cb170-86" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb170-87"><a href="#cb170-87" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb170-88"><a href="#cb170-88" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb170-89"><a href="#cb170-89" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb170-90"><a href="#cb170-90" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb170-91"><a href="#cb170-91" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb170-92"><a href="#cb170-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-93"><a href="#cb170-93" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb170-94"><a href="#cb170-94" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb170-95"><a href="#cb170-95" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb170-96"><a href="#cb170-96" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb170-97"><a href="#cb170-97" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb170-98"><a href="#cb170-98" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb170-99"><a href="#cb170-99" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb170-100"><a href="#cb170-100" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb170-101"><a href="#cb170-101" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb170-102"><a href="#cb170-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-103"><a href="#cb170-103" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb170-104"><a href="#cb170-104" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb170-105"><a href="#cb170-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-106"><a href="#cb170-106" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb170-107"><a href="#cb170-107" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb170-108"><a href="#cb170-108" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb170-109"><a href="#cb170-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-110"><a href="#cb170-110" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb170-111"><a href="#cb170-111" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb170-112"><a href="#cb170-112" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb170-113"><a href="#cb170-113" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [71 2 62 3 67] and the sum of those is 205
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-67-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="67">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 0 … 2 2; 2 0 … 2 2], [4175655, 4190847, 4195006, 4201372, 4201950, 4201984, 4202662, 4220609, 4223822, 4269946  …  5519564, 5519744, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">50×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
   3 │ GW_Lane5_TL10                 GW_Lane5_TL10                 TL        v
   4 │ GW_Lane5_TL11                 GW_Lane5_TL11                 TL        v
   5 │ GW_Lane5_TL12                 GW_Lane5_TL12                 TL        v ⋯
   6 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v
   7 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   8 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   9 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l ⋯
  10 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  11 │ GW_Armando_plate2_TTGW-15-01  GW_Armando_plate2_TTGW-15-01  SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  41 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  42 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p ⋯
  43 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  44 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  45 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p
  46 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p ⋯
  47 │ GW_Armando_plate1_JG08G01     GW_Armando_plate1_JG08G01     ST        p
  48 │ GW_Armando_plate1_RF20G01     GW_Armando_plate1_RF20G01     BJ        o
  49 │ GW_Armando_plate1_RF29G02     GW_Armando_plate1_RF29G02     BJ        o
  50 │ GW_Armando_plate1_UY1         GW_Armando_plate1_UY1         UY        p ⋯
<span class="ansi-cyan-fg">                                                  21 columns and 29 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-3" class="level2">
<h2 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-3">Now show a GBI plot like above, but with heterozygotes:</h2>
<div id="79fdb768" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb173-5"><a href="#cb173-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb173-6"><a href="#cb173-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs_plumb"</span>,</span>
<span id="cb173-7"><a href="#cb173-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb173-8"><a href="#cb173-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb173-9"><a href="#cb173-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-10"><a href="#cb173-10" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb173-11"><a href="#cb173-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb173-12"><a href="#cb173-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb173-13"><a href="#cb173-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb173-14"><a href="#cb173-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb173-15"><a href="#cb173-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"darkorange1"</span>,</span>
<span id="cb173-16"><a href="#cb173-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb173-17"><a href="#cb173-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb173-18"><a href="#cb173-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-19"><a href="#cb173-19" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.2</span>) <span class="op">.&amp;</span></span>
<span id="cb173-20"><a href="#cb173-20" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb173-21"><a href="#cb173-21" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb173-22"><a href="#cb173-22" aria-hidden="true" tabindex="-1"></a>obs_plumb <span class="op">=</span> (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb173-23"><a href="#cb173-23" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb173-24"><a href="#cb173-24" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb173-25"><a href="#cb173-25" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb173-26"><a href="#cb173-26" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb173-27"><a href="#cb173-27" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb173-28"><a href="#cb173-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-29"><a href="#cb173-29" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb173-30"><a href="#cb173-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb173-31"><a href="#cb173-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb173-32"><a href="#cb173-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud_troch]</span>
<span id="cb173-33"><a href="#cb173-33" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb173-34"><a href="#cb173-34" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb173-35"><a href="#cb173-35" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs_plumb]</span>
<span id="cb173-36"><a href="#cb173-36" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb173-37"><a href="#cb173-37" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[vir_plumb]</span>
<span id="cb173-38"><a href="#cb173-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-39"><a href="#cb173-39" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit virLud_troch troch obs obs_plumb plumb vir_plumb]</span>
<span id="cb173-40"><a href="#cb173-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-41"><a href="#cb173-41" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb173-42"><a href="#cb173-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-43"><a href="#cb173-43" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb173-44"><a href="#cb173-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb173-45"><a href="#cb173-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb173-46"><a href="#cb173-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-47"><a href="#cb173-47" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb173-48"><a href="#cb173-48" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb173-49"><a href="#cb173-49" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb173-50"><a href="#cb173-50" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb173-51"><a href="#cb173-51" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb173-52"><a href="#cb173-52" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb173-53"><a href="#cb173-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb173-54"><a href="#cb173-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-55"><a href="#cb173-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb173-56"><a href="#cb173-56" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb173-57"><a href="#cb173-57" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb173-58"><a href="#cb173-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-59"><a href="#cb173-59" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb173-60"><a href="#cb173-60" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb173-61"><a href="#cb173-61" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb173-62"><a href="#cb173-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-63"><a href="#cb173-63" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb173-64"><a href="#cb173-64" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb173-65"><a href="#cb173-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-66"><a href="#cb173-66" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb173-67"><a href="#cb173-67" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb173-68"><a href="#cb173-68" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb173-69"><a href="#cb173-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-70"><a href="#cb173-70" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb173-71"><a href="#cb173-71" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb173-72"><a href="#cb173-72" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb173-73"><a href="#cb173-73" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb173-74"><a href="#cb173-74" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-68-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="68">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 0 … 1 1; 1 0 … 1 1], [4175655, 4190847, 4195006, 4201372, 4201950, 4201984, 4202662, 4220609, 4223822, 4269946  …  5519564, 5519744, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">72×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
   3 │ GW_Lane5_TL10                 GW_Lane5_TL10                 TL        v
   4 │ GW_Lane5_TL11                 GW_Lane5_TL11                 TL        v
   5 │ GW_Lane5_TL12                 GW_Lane5_TL12                 TL        v ⋯
   6 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v
   7 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l
   8 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   9 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l ⋯
  10 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  11 │ GW_Armando_plate2_TTGW-15-01  GW_Armando_plate2_TTGW-15-01  SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  63 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  64 │ GW_Armando_plate1_JG08G01     GW_Armando_plate1_JG08G01     ST        p ⋯
  65 │ GW_Armando_plate1_RF20G01     GW_Armando_plate1_RF20G01     BJ        o
  66 │ GW_Armando_plate1_RF29G02     GW_Armando_plate1_RF29G02     BJ        o
  67 │ GW_Armando_plate1_UY1         GW_Armando_plate1_UY1         UY        p
  68 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p ⋯
  69 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p
  70 │ GW_Armando_plate2_JF07G01     GW_Armando_plate2_JF07G01     ST        p
  71 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
  72 │ GW_Armando_plate1_JF09G01     GW_Armando_plate1_JF09G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  21 columns and 51 rows omitted</span>)</pre>
</div>
</div>
</div>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-2" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-2">show GBI plot according to original groups and plot order</h3>
<div id="7199ce13" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb176-5"><a href="#cb176-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb176-6"><a href="#cb176-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-7"><a href="#cb176-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb176-8"><a href="#cb176-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb176-9"><a href="#cb176-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb176-10"><a href="#cb176-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb176-11"><a href="#cb176-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-69-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="69">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 2 0; … ; 1 0 … 1 1; 2 0 … 0 0], [4175655, 4190847, 4195006, 4201372, 4201950, 4201984, 4202662, 4220609, 4223822, 4269946  …  5519564, 5519744, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">72×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   3 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   4 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
   5 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir     ⋯
   6 │ GW_Lane5_TL12              GW_Lane5_TL12              TL        vir
   7 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
   8 │ GW_Lane5_AA4               GW_Lane5_AA4               AA        vir_S
   9 │ GW_Lane5_TU2               GW_Lane5_TU2               TU        nit     ⋯
  10 │ GW_Lane5_TU1               GW_Lane5_TU1               TU        nit
  11 │ GW_Lane5_NA1               GW_Lane5_NA1               NR        lud_PK
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
  63 │ GW_Armando_plate1_JF29G03  GW_Armando_plate1_JF29G03  ST        plumb
  64 │ GW_Armando_plate1_JG08G01  GW_Armando_plate1_JG08G01  ST        plumb   ⋯
  65 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  66 │ GW_Armando_plate2_JF07G01  GW_Armando_plate2_JF07G01  ST        plumb
  67 │ GW_Armando_plate1_JF09G01  GW_Armando_plate1_JF09G01  ST        plumb
  68 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb   ⋯
  69 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  70 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  71 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  72 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  21 columns and 51 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-2" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-2">show same but with all individuals</h3>
<div id="99c86db8" class="cell" data-execution_count="69">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb178-4"><a href="#cb178-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb178-5"><a href="#cb178-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-6"><a href="#cb178-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb178-7"><a href="#cb178-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb178-8"><a href="#cb178-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb178-9"><a href="#cb178-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-10"><a href="#cb178-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb178-11"><a href="#cb178-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb178-12"><a href="#cb178-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb178-13"><a href="#cb178-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb178-14"><a href="#cb178-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-70-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="70">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 0 … 2 2; 2 0 … 0 0], [4175655, 4190847, 4195006, 4201372, 4201950, 4201984, 4202662, 4220609, 4223822, 4269946  …  5519564, 5519744, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">257×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK3               GW_Lane5_YK3               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 249 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 250 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 251 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 252 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 253 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 255 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 256 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 21 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops-1" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops-1">show same but with only vir and plumb pops</h3>
<div id="0e4e6f6f" class="cell" data-execution_count="70">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb180-8"><a href="#cb180-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb180-9"><a href="#cb180-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb180-10"><a href="#cb180-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb180-11"><a href="#cb180-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb180-12"><a href="#cb180-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb180-13"><a href="#cb180-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb180-14"><a href="#cb180-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-15"><a href="#cb180-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb180-16"><a href="#cb180-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb180-17"><a href="#cb180-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-18"><a href="#cb180-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb180-19"><a href="#cb180-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-20"><a href="#cb180-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb180-21"><a href="#cb180-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-22"><a href="#cb180-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb180-23"><a href="#cb180-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb180-24"><a href="#cb180-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-71-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="71">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 1 0; … ; 2 2 … 2 2; 2 1 … 0 0], [4175655, 4201372, 4202662, 4220609, 4269946, 4272732, 4280387, 4397231, 4397235, 4411591  …  5519560, 5519564, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">100×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  21 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="show-just-the-west-clusters-without-nitidus" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-west-clusters-without-nitidus">Show just the west clusters (without nitidus)</h2>
<div id="bb38e112" class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb182-3"><a href="#cb182-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb182-4"><a href="#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="#cb182-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb182-6"><a href="#cb182-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb182-7"><a href="#cb182-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb182-8"><a href="#cb182-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-9"><a href="#cb182-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb182-10"><a href="#cb182-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb182-11"><a href="#cb182-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb182-12"><a href="#cb182-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb182-13"><a href="#cb182-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb182-14"><a href="#cb182-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb182-15"><a href="#cb182-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb182-16"><a href="#cb182-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb182-17"><a href="#cb182-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb182-18"><a href="#cb182-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb182-19"><a href="#cb182-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb182-20"><a href="#cb182-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-21"><a href="#cb182-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb182-22"><a href="#cb182-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-23"><a href="#cb182-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb182-24"><a href="#cb182-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb182-25"><a href="#cb182-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb182-26"><a href="#cb182-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-27"><a href="#cb182-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb182-28"><a href="#cb182-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb182-29"><a href="#cb182-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb182-30"><a href="#cb182-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb182-31"><a href="#cb182-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-72-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="72">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; -1 2 … 2 2; 0 -1 … 0 0], [4175655, 4195006, 4201372, 4201950, 4201984, 4202662, 4220609, 4272842, 4280387, 4397235  …  5462347, 5485541, 5485558, 5493066, 5494683, 5519564, 5524191, 5524276, 5530753, 5549587], <span class="ansi-bold">178×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   3 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 169 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 170 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L ⋯
 171 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
 172 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 173 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L
 174 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L ⋯
 175 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
 176 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 177 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 178 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                 21 columns and 157 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-3" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-east-area-3">Show just the east area</h2>
<div id="82907607" class="cell" data-execution_count="72">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs_plumb"</span>,</span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>,]</span>
<span id="cb185-5"><a href="#cb185-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-6"><a href="#cb185-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb185-7"><a href="#cb185-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb185-8"><a href="#cb185-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"darkorange1"</span>,</span>
<span id="cb185-9"><a href="#cb185-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb185-10"><a href="#cb185-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-11"><a href="#cb185-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb185-12"><a href="#cb185-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb185-13"><a href="#cb185-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb185-14"><a href="#cb185-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb185-15"><a href="#cb185-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb185-16"><a href="#cb185-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb185-17"><a href="#cb185-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb185-18"><a href="#cb185-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb185-19"><a href="#cb185-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb185-20"><a href="#cb185-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb185-21"><a href="#cb185-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb185-22"><a href="#cb185-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-23"><a href="#cb185-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb185-24"><a href="#cb185-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-25"><a href="#cb185-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb185-26"><a href="#cb185-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb185-27"><a href="#cb185-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb185-28"><a href="#cb185-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-29"><a href="#cb185-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb185-30"><a href="#cb185-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb185-31"><a href="#cb185-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb185-32"><a href="#cb185-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb185-33"><a href="#cb185-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-73-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="73">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 2; … ; 2 2 … 2 2; -1 2 … 2 2], [4195006, 4201950, 4201984, 4269946, 4272732, 4272842, 4336840, 4397231, 4398942, 4411760  …  5471609, 5478884, 5499176, 5509392, 5519560, 5519744, 5530774, 5535422, 5540526, 5549602], <span class="ansi-bold">134×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW72      GW_Armando_plate1_TTGW72      SA        l ⋯
   2 │ GW_Armando_plate1_TTGW78      GW_Armando_plate1_TTGW78      SA        l
   3 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l
   4 │ GW_Armando_plate2_TTGW77      GW_Armando_plate2_TTGW77      SA        l
   5 │ GW_Armando_plate2_TTGW79      GW_Armando_plate2_TTGW79      SA        l ⋯
   6 │ GW_Armando_plate2_TTGW80      GW_Armando_plate2_TTGW80      SA        l
   7 │ GW_Armando_plate2_TTGW75      GW_Armando_plate2_TTGW75      SA        l
   8 │ GW_Armando_plate2_TTGW28      GW_Armando_plate2_TTGW28      SR        l
   9 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l ⋯
  10 │ GW_Armando_plate2_TTGW-15-04  GW_Armando_plate2_TTGW-15-04  SR        l
  11 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 125 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
 126 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p ⋯
 127 │ GW_Armando_plate2_JF24G01     GW_Armando_plate2_JF24G01     VB        p
 128 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p
 129 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p
 130 │ GW_Armando_plate2_JG02G01     GW_Armando_plate2_JG02G01     PR        p ⋯
 131 │ GW_Armando_plate2_JG02G03     GW_Armando_plate2_JG02G03     PR        p
 132 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 133 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p
 134 │ GW_Lane5_SL1                  GW_Lane5_SL1                  SL        p ⋯
<span class="ansi-cyan-fg">                                                 21 columns and 113 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-3" class="level2">
<h2 class="anchored" data-anchor-id="show-just-the-northern-area-3">Show just the northern area</h2>
<div id="34521bb8" class="cell" data-execution_count="73">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-5"><a href="#cb188-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb188-6"><a href="#cb188-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb188-7"><a href="#cb188-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb188-8"><a href="#cb188-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-9"><a href="#cb188-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb188-10"><a href="#cb188-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb188-11"><a href="#cb188-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb188-12"><a href="#cb188-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb188-13"><a href="#cb188-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb188-14"><a href="#cb188-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb188-15"><a href="#cb188-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb188-16"><a href="#cb188-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb188-17"><a href="#cb188-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb188-18"><a href="#cb188-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb188-19"><a href="#cb188-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb188-20"><a href="#cb188-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-21"><a href="#cb188-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb188-22"><a href="#cb188-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-23"><a href="#cb188-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb188-24"><a href="#cb188-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb188-25"><a href="#cb188-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb188-26"><a href="#cb188-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-27"><a href="#cb188-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb188-28"><a href="#cb188-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb188-29"><a href="#cb188-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb188-30"><a href="#cb188-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb188-31"><a href="#cb188-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-74-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="74">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 1 … 0 0], [4175655, 4201372, 4202662, 4220609, 4269946, 4272732, 4280387, 4397231, 4397235, 4411591  …  5519560, 5519564, 5524191, 5524276, 5530753, 5530774, 5535422, 5540526, 5549587, 5549602], <span class="ansi-bold">143×24 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   3 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   4 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   5 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK3               GW_Lane5_YK3               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 134 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 135 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 136 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 137 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 138 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 139 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb   ⋯
 140 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 141 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 142 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 143 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 21 columns and 122 rows omitted</span>)</pre>
</div>
</div>
</div>
<section id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw26-with-low-hetvar" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw26-with-low-hetvar">Do a PCA based on a same-size region elsewhere on gw26 (with low HetVar):</h3>
<div id="e6a5aec8" class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get length of region</span></span>
<span id="cb191-2"><a href="#cb191-2" aria-hidden="true" tabindex="-1"></a>lengthHighHetVarRegion <span class="op">=</span> positionMax <span class="op">-</span> positionMin</span>
<span id="cb191-3"><a href="#cb191-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-4"><a href="#cb191-4" aria-hidden="true" tabindex="-1"></a>leftLocus <span class="op">=</span> <span class="fl">1_000_000</span> <span class="co"># start at 1 Mb from left side</span></span>
<span id="cb191-5"><a href="#cb191-5" aria-hidden="true" tabindex="-1"></a>rightLocus <span class="op">=</span> leftLocus <span class="op">+</span> lengthHighHetVarRegion</span>
<span id="cb191-6"><a href="#cb191-6" aria-hidden="true" tabindex="-1"></a>regionText_lowHetVarRegion <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,leftLocus,<span class="st">" to "</span>,rightLocus)</span>
<span id="cb191-7"><a href="#cb191-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-8"><a href="#cb191-8" aria-hidden="true" tabindex="-1"></a>lociSelection <span class="op">=</span> (leftLocus <span class="op">.&lt;=</span> pos_region.position <span class="op">.&lt;=</span> rightLocus)</span>
<span id="cb191-9"><a href="#cb191-9" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion <span class="op">=</span> genotypes_region[<span class="op">:</span>, lociSelection]</span>
<span id="cb191-10"><a href="#cb191-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-11"><a href="#cb191-11" aria-hidden="true" tabindex="-1"></a><span class="co"># impute missing genotypes:</span></span>
<span id="cb191-12"><a href="#cb191-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-13"><a href="#cb191-13" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genotypes_lowHetVarRegion))</span>
<span id="cb191-14"><a href="#cb191-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-15"><a href="#cb191-15" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb191-16"><a href="#cb191-16" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb191-17"><a href="#cb191-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-18"><a href="#cb191-18" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genotypes_lowHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb191-19"><a href="#cb191-19" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb191-20"><a href="#cb191-20" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText_lowHetVarRegion,</span>
<span id="cb191-21"><a href="#cb191-21" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb191-22"><a href="#cb191-22" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb191-23"><a href="#cb191-23" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb191-24"><a href="#cb191-24" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb191-25"><a href="#cb191-25" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb191-26"><a href="#cb191-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-27"><a href="#cb191-27" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb191-28"><a href="#cb191-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb191-29"><a href="#cb191-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"FigureS2C_gw26_nonHLBRarbitrary_from_Julia.png"</span>, PCAmodel.PCAfig, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb191-30"><a href="#cb191-30" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-75-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="do-similar-as-above-but-for-chr-1a" class="level2">
<h2 class="anchored" data-anchor-id="do-similar-as-above-but-for-chr-1a">Do similar as above but for chr 1A:</h2>
<p>(Tried chr 1 but seems to be some recomb that makes it less clear to assign to groups)</p>
<div id="0514b40a" class="cell" data-execution_count="75">
<div class="sourceCode cell-code" id="cb193"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb193-2"><a href="#cb193-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw1A"</span></span>
<span id="cb193-3"><a href="#cb193-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-4"><a href="#cb193-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb193-5"><a href="#cb193-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb193-6"><a href="#cb193-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb193-7"><a href="#cb193-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb193-8"><a href="#cb193-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb193-9"><a href="#cb193-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb193-10"><a href="#cb193-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb193-11"><a href="#cb193-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-12"><a href="#cb193-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb193-13"><a href="#cb193-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb193-14"><a href="#cb193-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-15"><a href="#cb193-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb193-16"><a href="#cb193-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb193-17"><a href="#cb193-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb193-18"><a href="#cb193-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb193-19"><a href="#cb193-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-20"><a href="#cb193-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb193-21"><a href="#cb193-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-22"><a href="#cb193-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb193-23"><a href="#cb193-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb193-24"><a href="#cb193-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-25"><a href="#cb193-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb193-26"><a href="#cb193-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-27"><a href="#cb193-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb193-28"><a href="#cb193-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-29"><a href="#cb193-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb193-30"><a href="#cb193-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb193-31"><a href="#cb193-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-32"><a href="#cb193-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb193-33"><a href="#cb193-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb193-34"><a href="#cb193-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb193-35"><a href="#cb193-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb193-36"><a href="#cb193-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb193-37"><a href="#cb193-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb193-38"><a href="#cb193-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb193-39"><a href="#cb193-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb193-40"><a href="#cb193-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-41"><a href="#cb193-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb193-42"><a href="#cb193-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-43"><a href="#cb193-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb193-44"><a href="#cb193-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb193-45"><a href="#cb193-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb193-46"><a href="#cb193-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb193-47"><a href="#cb193-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb193-48"><a href="#cb193-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb193-49"><a href="#cb193-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb193-50"><a href="#cb193-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb193-51"><a href="#cb193-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb193-52"><a href="#cb193-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb193-53"><a href="#cb193-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb193-54"><a href="#cb193-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb193-55"><a href="#cb193-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-56"><a href="#cb193-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb193-57"><a href="#cb193-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb193-58"><a href="#cb193-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-59"><a href="#cb193-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb193-60"><a href="#cb193-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.5</span>) </span>
<span id="cb193-61"><a href="#cb193-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-62"><a href="#cb193-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb193-63"><a href="#cb193-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-64"><a href="#cb193-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb193-65"><a href="#cb193-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb193-66"><a href="#cb193-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb193-67"><a href="#cb193-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb193-68"><a href="#cb193-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb193-69"><a href="#cb193-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb193-70"><a href="#cb193-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb193-71"><a href="#cb193-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb193-72"><a href="#cb193-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb193-73"><a href="#cb193-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb193-74"><a href="#cb193-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb193-75"><a href="#cb193-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>2×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw1A</td>
<td style="text-align: right;">4674</td>
<td style="text-align: right;">3771263</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw1A</td>
<td style="text-align: right;">23592559</td>
<td style="text-align: right;">30616953</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-76-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-76-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="76">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-4" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-4">Save the individual colors in the metadata</h3>
<div id="bcc35d66" class="cell" data-execution_count="76">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb198-4"><a href="#cb198-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb198-5"><a href="#cb198-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb198-6"><a href="#cb198-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="77">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-4" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-4">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="03fd9c3a" class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb200-3"><a href="#cb200-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb200-4"><a href="#cb200-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb200-5"><a href="#cb200-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb200-6"><a href="#cb200-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb200-7"><a href="#cb200-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb200-8"><a href="#cb200-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb200-9"><a href="#cb200-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb200-10"><a href="#cb200-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb200-11"><a href="#cb200-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb200-12"><a href="#cb200-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-78-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="78">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-4" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-4">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="39bea636" class="cell" data-execution_count="78">
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb203-2"><a href="#cb203-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb203-3"><a href="#cb203-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb203-4"><a href="#cb203-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb203-5"><a href="#cb203-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb203-6"><a href="#cb203-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb203-7"><a href="#cb203-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb203-8"><a href="#cb203-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb203-9"><a href="#cb203-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb203-10"><a href="#cb203-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb203-11"><a href="#cb203-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb203-12"><a href="#cb203-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-79-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="79">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-1a-high-hetvar-region-there-are-only-5-clear-haplogroups-checked-that-vir-and-lud-dont-form-distinct-groups-on-pc3.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-1a-high-hetvar-region-there-are-only-5-clear-haplogroups-checked-that-vir-and-lud-dont-form-distinct-groups-on-pc3.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 1A high HetVar region, there are only 5 clear haplogroups (checked that vir and lud don’t form distinct groups on PC3). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="24e1bcff" class="cell" data-execution_count="79">
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb206-2"><a href="#cb206-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb206-3"><a href="#cb206-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb206-4"><a href="#cb206-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb206-5"><a href="#cb206-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb206-6"><a href="#cb206-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-7"><a href="#cb206-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"green"</span>,</span>
<span id="cb206-8"><a href="#cb206-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb206-9"><a href="#cb206-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellowgreen"</span>,</span>
<span id="cb206-10"><a href="#cb206-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb206-11"><a href="#cb206-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb206-12"><a href="#cb206-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-13"><a href="#cb206-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">8</span>) <span class="op">.&amp;</span> </span>
<span id="cb206-14"><a href="#cb206-14" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb206-15"><a href="#cb206-15" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">8</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb206-16"><a href="#cb206-16" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb206-17"><a href="#cb206-17" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">9</span>) <span class="op">.&amp;</span></span>
<span id="cb206-18"><a href="#cb206-18" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb206-19"><a href="#cb206-19" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb206-20"><a href="#cb206-20" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">6</span>) <span class="op">.&amp;</span></span>
<span id="cb206-21"><a href="#cb206-21" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span> </span>
<span id="cb206-22"><a href="#cb206-22" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb206-23"><a href="#cb206-23" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb206-24"><a href="#cb206-24" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">7.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb206-25"><a href="#cb206-25" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb206-26"><a href="#cb206-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-27"><a href="#cb206-27" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb206-28"><a href="#cb206-28" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb206-29"><a href="#cb206-29" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb206-30"><a href="#cb206-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb206-31"><a href="#cb206-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb206-32"><a href="#cb206-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb206-33"><a href="#cb206-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-34"><a href="#cb206-34" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb206-35"><a href="#cb206-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-36"><a href="#cb206-36" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb206-37"><a href="#cb206-37" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb206-38"><a href="#cb206-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-39"><a href="#cb206-39" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb206-40"><a href="#cb206-40" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb206-41"><a href="#cb206-41" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb206-42"><a href="#cb206-42" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb206-43"><a href="#cb206-43" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb206-44"><a href="#cb206-44" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb206-45"><a href="#cb206-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb206-46"><a href="#cb206-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-47"><a href="#cb206-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb206-48"><a href="#cb206-48" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb206-49"><a href="#cb206-49" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb206-50"><a href="#cb206-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-51"><a href="#cb206-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb206-52"><a href="#cb206-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-53"><a href="#cb206-53" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb206-54"><a href="#cb206-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-55"><a href="#cb206-55" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb206-56"><a href="#cb206-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-57"><a href="#cb206-57" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb206-58"><a href="#cb206-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-59"><a href="#cb206-59" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb206-60"><a href="#cb206-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-61"><a href="#cb206-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb206-62"><a href="#cb206-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-63"><a href="#cb206-63" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb206-64"><a href="#cb206-64" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb206-65"><a href="#cb206-65" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb206-66"><a href="#cb206-66" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb206-67"><a href="#cb206-67" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb206-68"><a href="#cb206-68" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.00559696</span></span>
<span id="cb206-69"><a href="#cb206-69" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00458482</span></span>
<span id="cb206-70"><a href="#cb206-70" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00470781</span></span>
<span id="cb206-71"><a href="#cb206-71" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.00524545</span></span>
<span id="cb206-72"><a href="#cb206-72" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00659452 =#</span></span>
<span id="cb206-73"><a href="#cb206-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-74"><a href="#cb206-74" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb206-75"><a href="#cb206-75" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb206-76"><a href="#cb206-76" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb206-77"><a href="#cb206-77" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb206-78"><a href="#cb206-78" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb206-79"><a href="#cb206-79" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.0234051</span></span>
<span id="cb206-80"><a href="#cb206-80" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0303858</span></span>
<span id="cb206-81"><a href="#cb206-81" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0285612</span></span>
<span id="cb206-82"><a href="#cb206-82" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0298279</span></span>
<span id="cb206-83"><a href="#cb206-83" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.036893</span></span>
<span id="cb206-84"><a href="#cb206-84" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0346282</span></span>
<span id="cb206-85"><a href="#cb206-85" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0363109</span></span>
<span id="cb206-86"><a href="#cb206-86" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0169886</span></span>
<span id="cb206-87"><a href="#cb206-87" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0274903</span></span>
<span id="cb206-88"><a href="#cb206-88" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0256253 =#</span></span>
<span id="cb206-89"><a href="#cb206-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-90"><a href="#cb206-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb206-91"><a href="#cb206-91" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb206-92"><a href="#cb206-92" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb206-93"><a href="#cb206-93" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb206-94"><a href="#cb206-94" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb206-95"><a href="#cb206-95" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb206-96"><a href="#cb206-96" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb206-97"><a href="#cb206-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-98"><a href="#cb206-98" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb206-99"><a href="#cb206-99" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb206-100"><a href="#cb206-100" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb206-101"><a href="#cb206-101" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb206-102"><a href="#cb206-102" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb206-103"><a href="#cb206-103" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb206-104"><a href="#cb206-104" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb206-105"><a href="#cb206-105" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb206-106"><a href="#cb206-106" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb206-107"><a href="#cb206-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-108"><a href="#cb206-108" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb206-109"><a href="#cb206-109" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb206-110"><a href="#cb206-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-111"><a href="#cb206-111" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb206-112"><a href="#cb206-112" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb206-113"><a href="#cb206-113" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb206-114"><a href="#cb206-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-115"><a href="#cb206-115" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb206-116"><a href="#cb206-116" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb206-117"><a href="#cb206-117" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb206-118"><a href="#cb206-118" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [75 2 65 5 68] and the sum of those is 215
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-80-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="80">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 0 0; 0 2 … 0 0], [23611461, 23653527, 23686609, 23686624, 23799577, 23812654, 23824078, 23832306, 23832308, 23868434  …  29356320, 29415681, 29627465, 29627471, 29738988, 29903377, 29993850, 30023875, 30023911, 30306067], <span class="ansi-bold">52×26 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   3 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   4 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l
   5 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l ⋯
   6 │ GW_Armando_plate1_TTGW65      GW_Armando_plate1_TTGW65      SP        l
   7 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   8 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
   9 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l ⋯
  10 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  11 │ GW_Armando_plate2_TTGW24      GW_Armando_plate2_TTGW24      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  43 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  44 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p ⋯
  45 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  46 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  47 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  48 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
  49 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p
  50 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  51 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  52 │ GW_Armando_plate1_JG08G01     GW_Armando_plate1_JG08G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  23 columns and 31 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-4" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-4">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="b369fdb9" class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb209-2"><a href="#cb209-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb209-3"><a href="#cb209-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb209-4"><a href="#cb209-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb209-5"><a href="#cb209-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb209-6"><a href="#cb209-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb209-7"><a href="#cb209-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb209-8"><a href="#cb209-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-9"><a href="#cb209-9" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb209-10"><a href="#cb209-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb209-11"><a href="#cb209-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb209-12"><a href="#cb209-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb209-13"><a href="#cb209-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb209-14"><a href="#cb209-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb209-15"><a href="#cb209-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb209-16"><a href="#cb209-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-17"><a href="#cb209-17" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb209-18"><a href="#cb209-18" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">8</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb209-19"><a href="#cb209-19" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb209-20"><a href="#cb209-20" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb209-21"><a href="#cb209-21" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb209-22"><a href="#cb209-22" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb209-23"><a href="#cb209-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-24"><a href="#cb209-24" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit virLud_troch troch obs plumb vir_plumb]</span>
<span id="cb209-25"><a href="#cb209-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-26"><a href="#cb209-26" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb209-27"><a href="#cb209-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-28"><a href="#cb209-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb209-29"><a href="#cb209-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb209-30"><a href="#cb209-30" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb209-31"><a href="#cb209-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb209-32"><a href="#cb209-32" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb209-33"><a href="#cb209-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-34"><a href="#cb209-34" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb209-35"><a href="#cb209-35" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb209-36"><a href="#cb209-36" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb209-37"><a href="#cb209-37" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb209-38"><a href="#cb209-38" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb209-39"><a href="#cb209-39" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb209-40"><a href="#cb209-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb209-41"><a href="#cb209-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-42"><a href="#cb209-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb209-43"><a href="#cb209-43" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb209-44"><a href="#cb209-44" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb209-45"><a href="#cb209-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-46"><a href="#cb209-46" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb209-47"><a href="#cb209-47" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb209-48"><a href="#cb209-48" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb209-49"><a href="#cb209-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-50"><a href="#cb209-50" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb209-51"><a href="#cb209-51" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb209-52"><a href="#cb209-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-53"><a href="#cb209-53" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb209-54"><a href="#cb209-54" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb209-55"><a href="#cb209-55" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb209-56"><a href="#cb209-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-57"><a href="#cb209-57" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb209-58"><a href="#cb209-58" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb209-59"><a href="#cb209-59" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb209-60"><a href="#cb209-60" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb209-61"><a href="#cb209-61" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-81-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="81">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 1 … 0 0; 0 1 … 0 0], [23611461, 23653527, 23686609, 23686624, 23799577, 23812654, 23824078, 23832306, 23832308, 23868434  …  29356320, 29415681, 29627465, 29627471, 29738988, 29903377, 29993850, 30023875, 30023911, 30306067], <span class="ansi-bold">72×26 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   3 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   4 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l
   5 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l ⋯
   6 │ GW_Armando_plate1_TTGW65      GW_Armando_plate1_TTGW65      SP        l
   7 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   8 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
   9 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l ⋯
  10 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  11 │ GW_Armando_plate2_TTGW24      GW_Armando_plate2_TTGW24      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  63 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  64 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p ⋯
  65 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  66 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  67 │ GW_Armando_plate1_JG08G01     GW_Armando_plate1_JG08G01     ST        p
  68 │ GW_Armando_plate2_JF19G01     GW_Armando_plate2_JF19G01     MB        p ⋯
  69 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p
  70 │ GW_Lane5_TA1                  GW_Lane5_TA1                  TA        p
  71 │ GW_Lane5_UY2                  GW_Lane5_UY2                  UY        p
  72 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                  23 columns and 51 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-3" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-3">Show just the west area (without nitidus)</h3>
<div id="21385e77" class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb212"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb212-1"><a href="#cb212-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb212-2"><a href="#cb212-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb212-3"><a href="#cb212-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb212-4"><a href="#cb212-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-5"><a href="#cb212-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb212-6"><a href="#cb212-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb212-7"><a href="#cb212-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb212-8"><a href="#cb212-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-9"><a href="#cb212-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb212-10"><a href="#cb212-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb212-11"><a href="#cb212-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb212-12"><a href="#cb212-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb212-13"><a href="#cb212-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb212-14"><a href="#cb212-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb212-15"><a href="#cb212-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb212-16"><a href="#cb212-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb212-17"><a href="#cb212-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb212-18"><a href="#cb212-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb212-19"><a href="#cb212-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb212-20"><a href="#cb212-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-21"><a href="#cb212-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb212-22"><a href="#cb212-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-23"><a href="#cb212-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb212-24"><a href="#cb212-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb212-25"><a href="#cb212-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb212-26"><a href="#cb212-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-27"><a href="#cb212-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb212-28"><a href="#cb212-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb212-29"><a href="#cb212-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb212-30"><a href="#cb212-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb212-31"><a href="#cb212-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-82-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="82">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 0; 2 2 … 1 0], [23653527, 23799577, 23812654, 23868434, 23880674, 23939747, 23970643, 24177835, 24200934, 24212017  …  28910551, 29072016, 29087210, 29415681, 29627465, 29738988, 29993850, 30023875, 30023911, 30306067], <span class="ansi-bold">177×26 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   3 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   4 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l
   5 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l ⋯
   6 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l
   7 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l
   8 │ GW_Armando_plate1_TTGW65      GW_Armando_plate1_TTGW65      SP        l
   9 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 168 │ GW_Lane5_LN8                  GW_Lane5_LN8                  LN        t
 169 │ GW_Liz_GBS_Liz5144            GW_Liz_GBS_Liz5144            ML        l ⋯
 170 │ GW_Liz_GBS_Liz5175            GW_Liz_GBS_Liz5175            ML        l
 171 │ GW_Armando_plate2_TTGW79      GW_Armando_plate2_TTGW79      SA        l
 172 │ GW_Lane5_LN10                 GW_Lane5_LN10                 LN        t
 173 │ GW_Lane5_MN8                  GW_Lane5_MN8                  MN        t ⋯
 174 │ GW_Lane5_MN1                  GW_Lane5_MN1                  MN        t
 175 │ GW_Lane5_LN12                 GW_Lane5_LN12                 LN        t
 176 │ GW_Lane5_MN9                  GW_Lane5_MN9                  MN        t
 177 │ GW_Lane5_MN3                  GW_Lane5_MN3                  MN        t ⋯
<span class="ansi-cyan-fg">                                                 23 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-4" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-4">Show just the east area</h3>
<div id="ba506400" class="cell" data-execution_count="82">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb215-2"><a href="#cb215-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb215-3"><a href="#cb215-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb215-4"><a href="#cb215-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-5"><a href="#cb215-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb215-6"><a href="#cb215-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb215-7"><a href="#cb215-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb215-8"><a href="#cb215-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-9"><a href="#cb215-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb215-10"><a href="#cb215-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb215-11"><a href="#cb215-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb215-12"><a href="#cb215-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb215-13"><a href="#cb215-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb215-14"><a href="#cb215-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb215-15"><a href="#cb215-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb215-16"><a href="#cb215-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb215-17"><a href="#cb215-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb215-18"><a href="#cb215-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb215-19"><a href="#cb215-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb215-20"><a href="#cb215-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-21"><a href="#cb215-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb215-22"><a href="#cb215-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-23"><a href="#cb215-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb215-24"><a href="#cb215-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb215-25"><a href="#cb215-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb215-26"><a href="#cb215-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-27"><a href="#cb215-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb215-28"><a href="#cb215-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb215-29"><a href="#cb215-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb215-30"><a href="#cb215-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb215-31"><a href="#cb215-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-83-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="83">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 2; 0 0 … 2 1; … ; 2 0 … 2 2; 2 0 … 2 2], [23824078, 23832308, 23868434, 23970643, 24176699, 24177835, 24185673, 24230071, 24246102, 24307431  …  29072016, 29256987, 29256988, 29356294, 29356307, 29627465, 29738988, 29903377, 30023911, 30306067], <span class="ansi-bold">138×26 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW01   GW_Armando_plate1_TTGW01   MN        troch_M ⋯
   2 │ GW_Armando_plate1_TTGW11   GW_Armando_plate1_TTGW11   SU        lud_Suk
   3 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
   4 │ GW_Armando_plate2_TTGW75   GW_Armando_plate2_TTGW75   SA        lud_Sat
   5 │ GW_Armando_plate2_TTGW77   GW_Armando_plate2_TTGW77   SA        lud_Sat ⋯
   6 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
   7 │ GW_Liz_GBS_Liz5179         GW_Liz_GBS_Liz5179         ML        lud_chi
   8 │ GW_Liz_GBS_Liz5191         GW_Liz_GBS_Liz5191         ML        lud_chi
   9 │ GW_Liz_GBS_Liz6204         GW_Liz_GBS_Liz6204         ML        lud_chi ⋯
  10 │ GW_Armando_plate1_TTGW06   GW_Armando_plate1_TTGW06   SU        lud_Suk
  11 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
 129 │ GW_Lane5_BK2               GW_Lane5_BK2               BK        plumb
 130 │ GW_Lane5_UY5               GW_Lane5_UY5               UY        plumb   ⋯
 131 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
 132 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
 133 │ GW_Lane5_ST1               GW_Lane5_ST1               ST        plumb
 134 │ GW_Lane5_UY3               GW_Lane5_UY3               UY        plumb   ⋯
 135 │ GW_Lane5_UY6               GW_Lane5_UY6               UY        plumb
 136 │ GW_Lane5_UY4               GW_Lane5_UY4               UY        plumb
 137 │ GW_Lane5_ST3               GW_Lane5_ST3               ST        plumb
 138 │ GW_Lane5_AN2               GW_Lane5_AN2               AN        plumb   ⋯
<span class="ansi-cyan-fg">                                                 23 columns and 117 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-4" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-4">Show just the northern area</h3>
<div id="d90eb994" class="cell" data-execution_count="83">
<div class="sourceCode cell-code" id="cb218"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb218-1"><a href="#cb218-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb218-2"><a href="#cb218-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb218-3"><a href="#cb218-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb218-4"><a href="#cb218-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-5"><a href="#cb218-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb218-6"><a href="#cb218-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb218-7"><a href="#cb218-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb218-8"><a href="#cb218-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-9"><a href="#cb218-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb218-10"><a href="#cb218-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb218-11"><a href="#cb218-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb218-12"><a href="#cb218-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb218-13"><a href="#cb218-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb218-14"><a href="#cb218-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb218-15"><a href="#cb218-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb218-16"><a href="#cb218-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb218-17"><a href="#cb218-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb218-18"><a href="#cb218-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb218-19"><a href="#cb218-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb218-20"><a href="#cb218-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-21"><a href="#cb218-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">100</span>]</span>
<span id="cb218-22"><a href="#cb218-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-23"><a href="#cb218-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb218-24"><a href="#cb218-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb218-25"><a href="#cb218-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb218-26"><a href="#cb218-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-27"><a href="#cb218-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb218-28"><a href="#cb218-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb218-29"><a href="#cb218-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb218-30"><a href="#cb218-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb218-31"><a href="#cb218-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-84-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="84">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 1 1; 1 1 … 1 -1], [23653527, 23799577, 23812654, 23824078, 23880674, 23939747, 24176699, 24200934, 24212017, 24246102  …  28504703, 28878541, 28910551, 29052457, 29087210, 29356307, 29415681, 29903377, 29993850, 30023875], <span class="ansi-bold">148×26 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   3 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   4 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l
   5 │ GW_Armando_plate1_TTGW19      GW_Armando_plate1_TTGW19      TH        l ⋯
   6 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l
   7 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l
   8 │ GW_Armando_plate1_TTGW65      GW_Armando_plate1_TTGW65      SP        l
   9 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 139 │ GW_Lane5_UY5                  GW_Lane5_UY5                  UY        p
 140 │ GW_Lane5_AN2                  GW_Lane5_AN2                  AN        p ⋯
 141 │ GW_Lane5_ST12                 GW_Lane5_ST12                 ST        p
 142 │ GW_Lane5_UY4                  GW_Lane5_UY4                  UY        p
 143 │ GW_Lane5_ST3                  GW_Lane5_ST3                  ST        p
 144 │ GW_Armando_plate2_JF19G01     GW_Armando_plate2_JF19G01     MB        p ⋯
 145 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p
 146 │ GW_Lane5_TA1                  GW_Lane5_TA1                  TA        p
 147 │ GW_Lane5_UY2                  GW_Lane5_UY2                  UY        p
 148 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                 23 columns and 127 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="do-similar-as-above-but-for-chr-2" class="level2">
<h2 class="anchored" data-anchor-id="do-similar-as-above-but-for-chr-2">Do similar as above but for chr 2:</h2>
<p>This one doesn’t look like a super clear pattern (in terms of assigning homozygous and heterozygous haploblock genotypes), but will see what it shows:</p>
<div id="f793dc2f" class="cell" data-execution_count="84">
<div class="sourceCode cell-code" id="cb221"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb221-1"><a href="#cb221-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb221-2"><a href="#cb221-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw2"</span></span>
<span id="cb221-3"><a href="#cb221-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-4"><a href="#cb221-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb221-5"><a href="#cb221-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb221-6"><a href="#cb221-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb221-7"><a href="#cb221-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb221-8"><a href="#cb221-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb221-9"><a href="#cb221-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb221-10"><a href="#cb221-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb221-11"><a href="#cb221-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-12"><a href="#cb221-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb221-13"><a href="#cb221-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb221-14"><a href="#cb221-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-15"><a href="#cb221-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb221-16"><a href="#cb221-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb221-17"><a href="#cb221-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb221-18"><a href="#cb221-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb221-19"><a href="#cb221-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-20"><a href="#cb221-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb221-21"><a href="#cb221-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-22"><a href="#cb221-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb221-23"><a href="#cb221-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb221-24"><a href="#cb221-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-25"><a href="#cb221-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb221-26"><a href="#cb221-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-27"><a href="#cb221-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb221-28"><a href="#cb221-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-29"><a href="#cb221-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb221-30"><a href="#cb221-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb221-31"><a href="#cb221-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-32"><a href="#cb221-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb221-33"><a href="#cb221-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb221-34"><a href="#cb221-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb221-35"><a href="#cb221-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb221-36"><a href="#cb221-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb221-37"><a href="#cb221-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb221-38"><a href="#cb221-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb221-39"><a href="#cb221-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb221-40"><a href="#cb221-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-41"><a href="#cb221-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb221-42"><a href="#cb221-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-43"><a href="#cb221-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb221-44"><a href="#cb221-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb221-45"><a href="#cb221-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb221-46"><a href="#cb221-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb221-47"><a href="#cb221-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb221-48"><a href="#cb221-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb221-49"><a href="#cb221-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb221-50"><a href="#cb221-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb221-51"><a href="#cb221-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb221-52"><a href="#cb221-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb221-53"><a href="#cb221-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb221-54"><a href="#cb221-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb221-55"><a href="#cb221-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-56"><a href="#cb221-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb221-57"><a href="#cb221-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb221-58"><a href="#cb221-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-59"><a href="#cb221-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb221-60"><a href="#cb221-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.25</span>) </span>
<span id="cb221-61"><a href="#cb221-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-62"><a href="#cb221-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb221-63"><a href="#cb221-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-64"><a href="#cb221-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb221-65"><a href="#cb221-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb221-66"><a href="#cb221-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb221-67"><a href="#cb221-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb221-68"><a href="#cb221-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb221-69"><a href="#cb221-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb221-70"><a href="#cb221-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb221-71"><a href="#cb221-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb221-72"><a href="#cb221-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb221-73"><a href="#cb221-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb221-74"><a href="#cb221-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb221-75"><a href="#cb221-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>2×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw2</td>
<td style="text-align: right;">54537375</td>
<td style="text-align: right;">59262130</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw2</td>
<td style="text-align: right;">60234161</td>
<td style="text-align: right;">61533451</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-85-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-85-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="85">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-5" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-5">Save the individual colors in the metadata</h3>
<div id="18d25094" class="cell" data-execution_count="85">
<div class="sourceCode cell-code" id="cb226"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb226-1"><a href="#cb226-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb226-2"><a href="#cb226-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb226-3"><a href="#cb226-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb226-4"><a href="#cb226-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb226-5"><a href="#cb226-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb226-6"><a href="#cb226-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="86">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-5" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-5">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="4953bc25" class="cell" data-execution_count="86">
<div class="sourceCode cell-code" id="cb228"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb228-1"><a href="#cb228-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb228-2"><a href="#cb228-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb228-3"><a href="#cb228-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb228-4"><a href="#cb228-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb228-5"><a href="#cb228-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb228-6"><a href="#cb228-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb228-7"><a href="#cb228-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb228-8"><a href="#cb228-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb228-9"><a href="#cb228-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb228-10"><a href="#cb228-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb228-11"><a href="#cb228-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb228-12"><a href="#cb228-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-87-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="87">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-5" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-5">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="474f0e6d" class="cell" data-execution_count="87">
<div class="sourceCode cell-code" id="cb231"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb231-1"><a href="#cb231-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb231-2"><a href="#cb231-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb231-3"><a href="#cb231-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb231-4"><a href="#cb231-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb231-5"><a href="#cb231-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb231-6"><a href="#cb231-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb231-7"><a href="#cb231-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb231-8"><a href="#cb231-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb231-9"><a href="#cb231-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb231-10"><a href="#cb231-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb231-11"><a href="#cb231-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb231-12"><a href="#cb231-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-88-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="88">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-2-high-hetvar-region-there-are-5-clear-haplogroups.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-2-high-hetvar-region-there-are-5-clear-haplogroups.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 2 high HetVar region, there are 5 clear haplogroups. Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="b207e661" class="cell" data-execution_count="88">
<div class="sourceCode cell-code" id="cb234"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb234-1"><a href="#cb234-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb234-2"><a href="#cb234-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb234-3"><a href="#cb234-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb234-4"><a href="#cb234-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb234-5"><a href="#cb234-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb234-6"><a href="#cb234-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-7"><a href="#cb234-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"green"</span>,</span>
<span id="cb234-8"><a href="#cb234-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb234-9"><a href="#cb234-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellowgreen"</span>,</span>
<span id="cb234-10"><a href="#cb234-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb234-11"><a href="#cb234-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb234-12"><a href="#cb234-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-13"><a href="#cb234-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> </span>
<span id="cb234-14"><a href="#cb234-14" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb234-15"><a href="#cb234-15" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb234-16"><a href="#cb234-16" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb234-17"><a href="#cb234-17" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb234-18"><a href="#cb234-18" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb234-19"><a href="#cb234-19" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb234-20"><a href="#cb234-20" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb234-21"><a href="#cb234-21" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb234-22"><a href="#cb234-22" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3</span>) <span class="op">.&amp;</span> </span>
<span id="cb234-23"><a href="#cb234-23" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb234-24"><a href="#cb234-24" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb234-25"><a href="#cb234-25" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb234-26"><a href="#cb234-26" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb234-27"><a href="#cb234-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-28"><a href="#cb234-28" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb234-29"><a href="#cb234-29" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb234-30"><a href="#cb234-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb234-31"><a href="#cb234-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb234-32"><a href="#cb234-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb234-33"><a href="#cb234-33" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb234-34"><a href="#cb234-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-35"><a href="#cb234-35" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb234-36"><a href="#cb234-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-37"><a href="#cb234-37" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb234-38"><a href="#cb234-38" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb234-39"><a href="#cb234-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-40"><a href="#cb234-40" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb234-41"><a href="#cb234-41" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb234-42"><a href="#cb234-42" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb234-43"><a href="#cb234-43" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb234-44"><a href="#cb234-44" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb234-45"><a href="#cb234-45" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb234-46"><a href="#cb234-46" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb234-47"><a href="#cb234-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-48"><a href="#cb234-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb234-49"><a href="#cb234-49" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb234-50"><a href="#cb234-50" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb234-51"><a href="#cb234-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-52"><a href="#cb234-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb234-53"><a href="#cb234-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-54"><a href="#cb234-54" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb234-55"><a href="#cb234-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-56"><a href="#cb234-56" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb234-57"><a href="#cb234-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-58"><a href="#cb234-58" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb234-59"><a href="#cb234-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-60"><a href="#cb234-60" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb234-61"><a href="#cb234-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-62"><a href="#cb234-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb234-63"><a href="#cb234-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-64"><a href="#cb234-64" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb234-65"><a href="#cb234-65" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb234-66"><a href="#cb234-66" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb234-67"><a href="#cb234-67" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb234-68"><a href="#cb234-68" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb234-69"><a href="#cb234-69" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.0123364</span></span>
<span id="cb234-70"><a href="#cb234-70" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00557103</span></span>
<span id="cb234-71"><a href="#cb234-71" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00911341</span></span>
<span id="cb234-72"><a href="#cb234-72" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.00891506</span></span>
<span id="cb234-73"><a href="#cb234-73" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.0086287 =#</span></span>
<span id="cb234-74"><a href="#cb234-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-75"><a href="#cb234-75" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb234-76"><a href="#cb234-76" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb234-77"><a href="#cb234-77" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb234-78"><a href="#cb234-78" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb234-79"><a href="#cb234-79" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb234-80"><a href="#cb234-80" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.0328534</span></span>
<span id="cb234-81"><a href="#cb234-81" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0337586</span></span>
<span id="cb234-82"><a href="#cb234-82" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0328064</span></span>
<span id="cb234-83"><a href="#cb234-83" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0416095</span></span>
<span id="cb234-84"><a href="#cb234-84" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0376123</span></span>
<span id="cb234-85"><a href="#cb234-85" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0363568</span></span>
<span id="cb234-86"><a href="#cb234-86" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0456889</span></span>
<span id="cb234-87"><a href="#cb234-87" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0144702</span></span>
<span id="cb234-88"><a href="#cb234-88" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0331178</span></span>
<span id="cb234-89"><a href="#cb234-89" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0318128 =#</span></span>
<span id="cb234-90"><a href="#cb234-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-91"><a href="#cb234-91" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb234-92"><a href="#cb234-92" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb234-93"><a href="#cb234-93" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb234-94"><a href="#cb234-94" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb234-95"><a href="#cb234-95" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb234-96"><a href="#cb234-96" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb234-97"><a href="#cb234-97" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb234-98"><a href="#cb234-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-99"><a href="#cb234-99" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb234-100"><a href="#cb234-100" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb234-101"><a href="#cb234-101" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb234-102"><a href="#cb234-102" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb234-103"><a href="#cb234-103" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb234-104"><a href="#cb234-104" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb234-105"><a href="#cb234-105" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb234-106"><a href="#cb234-106" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb234-107"><a href="#cb234-107" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb234-108"><a href="#cb234-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-109"><a href="#cb234-109" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb234-110"><a href="#cb234-110" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb234-111"><a href="#cb234-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-112"><a href="#cb234-112" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb234-113"><a href="#cb234-113" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb234-114"><a href="#cb234-114" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb234-115"><a href="#cb234-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-116"><a href="#cb234-116" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb234-117"><a href="#cb234-117" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb234-118"><a href="#cb234-118" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb234-119"><a href="#cb234-119" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [59 1 72 4 69] and the sum of those is 205
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-89-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="89">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 2 0; 0 2 … 2 0], [54765846, 54905388, 54954528, 54954575, 55151909, 55151945, 55176856, 55200024, 55200062, 55200114  …  59023396, 59058874, 59071396, 59075668, 59075686, 59133380, 59154238, 59154281, 59210193, 59210220], <span class="ansi-bold">50×28 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l ⋯
   2 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   3 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   4 │ GW_Armando_plate2_JF03G01     GW_Armando_plate2_JF03G01     ST_vi     v
   5 │ GW_Armando_plate2_TTGW25      GW_Armando_plate2_TTGW25      SR        l ⋯
   6 │ GW_Armando_plate2_TTGW27      GW_Armando_plate2_TTGW27      SR        l
   7 │ GW_Armando_plate2_TTGW56      GW_Armando_plate2_TTGW56      NG        l
   8 │ GW_Armando_plate2_TTGW-15-03  GW_Armando_plate2_TTGW-15-03  SR        l
   9 │ GW_Lane5_AA4                  GW_Lane5_AA4                  AA        v ⋯
  10 │ GW_Lane5_PT4                  GW_Lane5_PT4                  PA        l
  11 │ GW_Lane5_TL9                  GW_Lane5_TL9                  TL        v
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  41 │ GW_Armando_plate1_JF12G01     GW_Armando_plate1_JF12G01     ST        p
  42 │ GW_Armando_plate1_JF13G01     GW_Armando_plate1_JF13G01     ST        p ⋯
  43 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p
  44 │ GW_Armando_plate1_JF20G01     GW_Armando_plate1_JF20G01     MB        p
  45 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p
  46 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p ⋯
  47 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  48 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  49 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  50 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  25 columns and 29 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-5" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-5">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="4ccf4775" class="cell" data-execution_count="89">
<div class="sourceCode cell-code" id="cb237"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb237-1"><a href="#cb237-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb237-2"><a href="#cb237-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLudHet"</span>,</span>
<span id="cb237-3"><a href="#cb237-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb237-4"><a href="#cb237-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nitHet"</span>,</span>
<span id="cb237-5"><a href="#cb237-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb237-6"><a href="#cb237-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb237-7"><a href="#cb237-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochHet"</span>,</span>
<span id="cb237-8"><a href="#cb237-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb237-9"><a href="#cb237-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs_plumb"</span>,</span>
<span id="cb237-10"><a href="#cb237-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb237-11"><a href="#cb237-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumbHet"</span>,</span>
<span id="cb237-12"><a href="#cb237-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb237-13"><a href="#cb237-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-14"><a href="#cb237-14" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb237-15"><a href="#cb237-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"blue"</span>,</span>
<span id="cb237-16"><a href="#cb237-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb237-17"><a href="#cb237-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb237-18"><a href="#cb237-18" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb237-19"><a href="#cb237-19" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb237-20"><a href="#cb237-20" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb237-21"><a href="#cb237-21" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb237-22"><a href="#cb237-22" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"darkorange1"</span>,</span>
<span id="cb237-23"><a href="#cb237-23" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb237-24"><a href="#cb237-24" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb237-25"><a href="#cb237-25" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb237-26"><a href="#cb237-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-27"><a href="#cb237-27" aria-hidden="true" tabindex="-1"></a>virLudHet <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> </span>
<span id="cb237-28"><a href="#cb237-28" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb237-29"><a href="#cb237-29" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb237-30"><a href="#cb237-30" aria-hidden="true" tabindex="-1"></a>nitHet <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb237-31"><a href="#cb237-31" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb237-32"><a href="#cb237-32" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb237-33"><a href="#cb237-33" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb237-34"><a href="#cb237-34" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb237-35"><a href="#cb237-35" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb237-36"><a href="#cb237-36" aria-hidden="true" tabindex="-1"></a>trochHet <span class="op">=</span> (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb237-37"><a href="#cb237-37" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb237-38"><a href="#cb237-38" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb237-39"><a href="#cb237-39" aria-hidden="true" tabindex="-1"></a>obs_plumb <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb237-40"><a href="#cb237-40" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb237-41"><a href="#cb237-41" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb237-42"><a href="#cb237-42" aria-hidden="true" tabindex="-1"></a>plumbHet <span class="op">=</span> (<span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb237-43"><a href="#cb237-43" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb237-44"><a href="#cb237-44" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb237-45"><a href="#cb237-45" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb237-46"><a href="#cb237-46" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb237-47"><a href="#cb237-47" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb237-48"><a href="#cb237-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-49"><a href="#cb237-49" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud virLudHet nit nitHet virLud_troch troch trochHet obs obs_plumb plumb plumbHet vir_plumb]</span>
<span id="cb237-50"><a href="#cb237-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-51"><a href="#cb237-51" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb237-52"><a href="#cb237-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-53"><a href="#cb237-53" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb237-54"><a href="#cb237-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb237-55"><a href="#cb237-55" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb237-56"><a href="#cb237-56" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb237-57"><a href="#cb237-57" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb237-58"><a href="#cb237-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-59"><a href="#cb237-59" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb237-60"><a href="#cb237-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-61"><a href="#cb237-61" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb237-62"><a href="#cb237-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-63"><a href="#cb237-63" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb237-64"><a href="#cb237-64" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb237-65"><a href="#cb237-65" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb237-66"><a href="#cb237-66" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb237-67"><a href="#cb237-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-68"><a href="#cb237-68" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb237-69"><a href="#cb237-69" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb237-70"><a href="#cb237-70" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb237-71"><a href="#cb237-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb237-72"><a href="#cb237-72" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb237-73"><a href="#cb237-73" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb237-74"><a href="#cb237-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb237-75"><a href="#cb237-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-76"><a href="#cb237-76" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb237-77"><a href="#cb237-77" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb237-78"><a href="#cb237-78" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb237-79"><a href="#cb237-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-80"><a href="#cb237-80" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb237-81"><a href="#cb237-81" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb237-82"><a href="#cb237-82" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb237-83"><a href="#cb237-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-84"><a href="#cb237-84" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb237-85"><a href="#cb237-85" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb237-86"><a href="#cb237-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-87"><a href="#cb237-87" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb237-88"><a href="#cb237-88" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb237-89"><a href="#cb237-89" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb237-90"><a href="#cb237-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-91"><a href="#cb237-91" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb237-92"><a href="#cb237-92" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb237-93"><a href="#cb237-93" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb237-94"><a href="#cb237-94" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb237-95"><a href="#cb237-95" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-90-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="90">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 1 0; -1 0 … 2 0], [54765846, 54905388, 54954528, 54954575, 55151909, 55151945, 55176856, 55200024, 55200062, 55200114  …  59023396, 59058874, 59071396, 59075668, 59075686, 59133380, 59154238, 59154281, 59210193, 59210220], <span class="ansi-bold">88×28 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW13      GW_Armando_plate1_TTGW13      TH        l ⋯
   2 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   3 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   4 │ GW_Armando_plate2_JF03G01     GW_Armando_plate2_JF03G01     ST_vi     v
   5 │ GW_Armando_plate2_TTGW25      GW_Armando_plate2_TTGW25      SR        l ⋯
   6 │ GW_Armando_plate2_TTGW27      GW_Armando_plate2_TTGW27      SR        l
   7 │ GW_Armando_plate2_TTGW56      GW_Armando_plate2_TTGW56      NG        l
   8 │ GW_Armando_plate2_TTGW-15-03  GW_Armando_plate2_TTGW-15-03  SR        l
   9 │ GW_Lane5_AA4                  GW_Lane5_AA4                  AA        v ⋯
  10 │ GW_Lane5_PT4                  GW_Lane5_PT4                  PA        l
  11 │ GW_Lane5_TL9                  GW_Lane5_TL9                  TL        v
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  79 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  80 │ GW_Armando_plate2_TTGW-15-06  GW_Armando_plate2_TTGW-15-06  SR        l ⋯
  81 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p
  82 │ GW_Armando_plate2_TTGW28      GW_Armando_plate2_TTGW28      SR        l
  83 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p
  84 │ GW_Lane5_PT6                  GW_Lane5_PT6                  KL        l ⋯
  85 │ GW_Lane5_AA6                  GW_Lane5_AA6                  AA        v
  86 │ GW_Lane5_TL2                  GW_Lane5_TL2                  TL        v
  87 │ GW_Lane5_AB2                  GW_Lane5_AB2                  AB        v
  88 │ GW_Lane5_NA1                  GW_Lane5_NA1                  NR        l ⋯
<span class="ansi-cyan-fg">                                                  25 columns and 67 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-3" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-3">show GBI plot according to original groups and plot order</h3>
<div id="de98bcda" class="cell" data-execution_count="90">
<div class="sourceCode cell-code" id="cb240"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb240-1"><a href="#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PCAmodelAll.metadata.Fst_group = PCAmodelAll.metadata.original_Fst_group</span></span>
<span id="cb240-2"><a href="#cb240-2" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb240-3"><a href="#cb240-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb240-4"><a href="#cb240-4" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb240-5"><a href="#cb240-5" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb240-6"><a href="#cb240-6" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb240-7"><a href="#cb240-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb240-8"><a href="#cb240-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb240-9"><a href="#cb240-9" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb240-10"><a href="#cb240-10" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb240-11"><a href="#cb240-11" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb240-12"><a href="#cb240-12" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-91-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="91">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 1 … 1 0; 0 1 … 1 0], [54765846, 54905388, 54954528, 54954575, 55151909, 55151945, 55176856, 55200024, 55200062, 55200114  …  59023396, 59058874, 59071396, 59075668, 59075686, 59133380, 59154238, 59154281, 59210193, 59210220], <span class="ansi-bold">88×28 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   4 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
   6 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
   7 │ GW_Lane5_YK6               GW_Lane5_YK6               YK        vir
   8 │ GW_Lane5_AB2               GW_Lane5_AB2               AB        vir
   9 │ GW_Lane5_TL9               GW_Lane5_TL9               TL        vir     ⋯
  10 │ GW_Lane5_TL12              GW_Lane5_TL12              TL        vir
  11 │ GW_Lane5_TL2               GW_Lane5_TL2               TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  79 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  80 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb   ⋯
  81 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  82 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  83 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  84 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb   ⋯
  85 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  86 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  87 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
  88 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  25 columns and 67 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-3" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-3">show same but with all individuals</h3>
<div id="f6269000" class="cell" data-execution_count="91">
<div class="sourceCode cell-code" id="cb242"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb242-1"><a href="#cb242-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PCAmodelAll.metadata.Fst_group = PCAmodelAll.metadata.original_Fst_group</span></span>
<span id="cb242-2"><a href="#cb242-2" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb242-3"><a href="#cb242-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb242-4"><a href="#cb242-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb242-5"><a href="#cb242-5" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb242-6"><a href="#cb242-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb242-7"><a href="#cb242-7" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb242-8"><a href="#cb242-8" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb242-9"><a href="#cb242-9" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb242-10"><a href="#cb242-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb242-11"><a href="#cb242-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb242-12"><a href="#cb242-12" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb242-13"><a href="#cb242-13" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb242-14"><a href="#cb242-14" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb242-15"><a href="#cb242-15" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-92-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="92">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 1 … 1 0; 0 1 … 1 0], [54765846, 54905388, 54954528, 54954575, 55151909, 55151945, 55176856, 55200024, 55200062, 55200114  …  59023396, 59058874, 59071396, 59075668, 59075686, 59133380, 59154238, 59154281, 59210193, 59210220], <span class="ansi-bold">256×28 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 247 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 248 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb   ⋯
 249 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 250 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 251 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 252 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb   ⋯
 253 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 254 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 255 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 256 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 25 columns and 235 rows omitted</span>)</pre>
</div>
</div>
</div>
<p>Chr 2 is complicated, shows recomb and perhaps some haploblock sharing between east Siberia and the southwestern area. Hard to show in summary figure but should perhaps mention.</p>
</section>
</section>
<section id="same-for-chr-3" class="level2">
<h2 class="anchored" data-anchor-id="same-for-chr-3">Same for chr 3</h2>
<div id="ad17c5c3" class="cell" data-execution_count="92">
<div class="sourceCode cell-code" id="cb244"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb244-1"><a href="#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb244-2"><a href="#cb244-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw3"</span></span>
<span id="cb244-3"><a href="#cb244-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-4"><a href="#cb244-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb244-5"><a href="#cb244-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb244-6"><a href="#cb244-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb244-7"><a href="#cb244-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb244-8"><a href="#cb244-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb244-9"><a href="#cb244-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb244-10"><a href="#cb244-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb244-11"><a href="#cb244-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-12"><a href="#cb244-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb244-13"><a href="#cb244-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb244-14"><a href="#cb244-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-15"><a href="#cb244-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb244-16"><a href="#cb244-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb244-17"><a href="#cb244-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb244-18"><a href="#cb244-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb244-19"><a href="#cb244-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-20"><a href="#cb244-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb244-21"><a href="#cb244-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-22"><a href="#cb244-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb244-23"><a href="#cb244-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb244-24"><a href="#cb244-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-25"><a href="#cb244-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb244-26"><a href="#cb244-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-27"><a href="#cb244-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb244-28"><a href="#cb244-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-29"><a href="#cb244-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb244-30"><a href="#cb244-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb244-31"><a href="#cb244-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-32"><a href="#cb244-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb244-33"><a href="#cb244-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb244-34"><a href="#cb244-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb244-35"><a href="#cb244-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb244-36"><a href="#cb244-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb244-37"><a href="#cb244-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb244-38"><a href="#cb244-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb244-39"><a href="#cb244-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb244-40"><a href="#cb244-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-41"><a href="#cb244-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb244-42"><a href="#cb244-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-43"><a href="#cb244-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb244-44"><a href="#cb244-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb244-45"><a href="#cb244-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb244-46"><a href="#cb244-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb244-47"><a href="#cb244-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb244-48"><a href="#cb244-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb244-49"><a href="#cb244-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb244-50"><a href="#cb244-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb244-51"><a href="#cb244-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb244-52"><a href="#cb244-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb244-53"><a href="#cb244-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb244-54"><a href="#cb244-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb244-55"><a href="#cb244-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-56"><a href="#cb244-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb244-57"><a href="#cb244-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb244-58"><a href="#cb244-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-59"><a href="#cb244-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb244-60"><a href="#cb244-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.25</span>) </span>
<span id="cb244-61"><a href="#cb244-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-62"><a href="#cb244-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb244-63"><a href="#cb244-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-64"><a href="#cb244-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb244-65"><a href="#cb244-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb244-66"><a href="#cb244-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb244-67"><a href="#cb244-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb244-68"><a href="#cb244-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb244-69"><a href="#cb244-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb244-70"><a href="#cb244-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb244-71"><a href="#cb244-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb244-72"><a href="#cb244-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb244-73"><a href="#cb244-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb244-74"><a href="#cb244-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb244-75"><a href="#cb244-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>2×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">101192949</td>
<td style="text-align: right;">103495514</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">104554714</td>
<td style="text-align: right;">108279595</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-93-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-93-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="93">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-6" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-6">Save the individual colors in the metadata</h3>
<div id="53aea72f" class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb249"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb249-1"><a href="#cb249-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb249-2"><a href="#cb249-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb249-3"><a href="#cb249-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb249-4"><a href="#cb249-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb249-5"><a href="#cb249-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb249-6"><a href="#cb249-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="94">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-6" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-6">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="c8eb3c2c" class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb251"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb251-1"><a href="#cb251-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb251-2"><a href="#cb251-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb251-3"><a href="#cb251-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb251-4"><a href="#cb251-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb251-5"><a href="#cb251-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb251-6"><a href="#cb251-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb251-7"><a href="#cb251-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb251-8"><a href="#cb251-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb251-9"><a href="#cb251-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb251-10"><a href="#cb251-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb251-11"><a href="#cb251-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb251-12"><a href="#cb251-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-95-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="95">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-6" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-6">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="a571a2da" class="cell" data-execution_count="95">
<div class="sourceCode cell-code" id="cb254"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb254-1"><a href="#cb254-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb254-2"><a href="#cb254-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb254-3"><a href="#cb254-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb254-4"><a href="#cb254-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb254-5"><a href="#cb254-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb254-6"><a href="#cb254-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb254-7"><a href="#cb254-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb254-8"><a href="#cb254-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb254-9"><a href="#cb254-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb254-10"><a href="#cb254-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb254-11"><a href="#cb254-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb254-12"><a href="#cb254-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-96-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="96">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-3-high-hetvar-region-there-are-only-4-clear-haplogroups-vir-and-lud-separated-though-on-pc3-but-not-in-a-clean-way-that-i-can-distinguish-clearly.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-3-high-hetvar-region-there-are-only-4-clear-haplogroups-vir-and-lud-separated-though-on-pc3-but-not-in-a-clean-way-that-i-can-distinguish-clearly.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 3 high HetVar region, there are only 4 clear haplogroups (vir and lud separated though on PC3, but not in a clean way that I can distinguish clearly). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="3439b0e9" class="cell" data-execution_count="96">
<div class="sourceCode cell-code" id="cb257"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb257-1"><a href="#cb257-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb257-2"><a href="#cb257-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb257-3"><a href="#cb257-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochObs"</span>,</span>
<span id="cb257-4"><a href="#cb257-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb257-5"><a href="#cb257-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-6"><a href="#cb257-6" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb257-7"><a href="#cb257-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb257-8"><a href="#cb257-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb257-9"><a href="#cb257-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb257-10"><a href="#cb257-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-11"><a href="#cb257-11" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb257-12"><a href="#cb257-12" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb257-13"><a href="#cb257-13" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb257-14"><a href="#cb257-14" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb257-15"><a href="#cb257-15" aria-hidden="true" tabindex="-1"></a>trochObs <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb257-16"><a href="#cb257-16" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb257-17"><a href="#cb257-17" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb257-18"><a href="#cb257-18" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb257-19"><a href="#cb257-19" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb257-20"><a href="#cb257-20" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb257-21"><a href="#cb257-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-22"><a href="#cb257-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb257-23"><a href="#cb257-23" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb257-24"><a href="#cb257-24" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb257-25"><a href="#cb257-25" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[trochObs]</span>
<span id="cb257-26"><a href="#cb257-26" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb257-27"><a href="#cb257-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-28"><a href="#cb257-28" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit trochObs plumb]</span>
<span id="cb257-29"><a href="#cb257-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-30"><a href="#cb257-30" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb257-31"><a href="#cb257-31" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb257-32"><a href="#cb257-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-33"><a href="#cb257-33" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb257-34"><a href="#cb257-34" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb257-35"><a href="#cb257-35" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb257-36"><a href="#cb257-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb257-37"><a href="#cb257-37" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb257-38"><a href="#cb257-38" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb257-39"><a href="#cb257-39" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb257-40"><a href="#cb257-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-41"><a href="#cb257-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb257-42"><a href="#cb257-42" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb257-43"><a href="#cb257-43" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb257-44"><a href="#cb257-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-45"><a href="#cb257-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb257-46"><a href="#cb257-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-47"><a href="#cb257-47" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb257-48"><a href="#cb257-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-49"><a href="#cb257-49" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb257-50"><a href="#cb257-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-51"><a href="#cb257-51" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb257-52"><a href="#cb257-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-53"><a href="#cb257-53" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb257-54"><a href="#cb257-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-55"><a href="#cb257-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb257-56"><a href="#cb257-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-57"><a href="#cb257-57" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb257-58"><a href="#cb257-58" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb257-59"><a href="#cb257-59" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi         </span></span>
<span id="cb257-60"><a href="#cb257-60" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64    </span></span>
<span id="cb257-61"><a href="#cb257-61" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼──────────────────────</span></span>
<span id="cb257-62"><a href="#cb257-62" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.00950795</span></span>
<span id="cb257-63"><a href="#cb257-63" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.00509165</span></span>
<span id="cb257-64"><a href="#cb257-64" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ trochObs  0.00992915</span></span>
<span id="cb257-65"><a href="#cb257-65" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ plumb     0.00992294 =#</span></span>
<span id="cb257-66"><a href="#cb257-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-67"><a href="#cb257-67" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb257-68"><a href="#cb257-68" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb257-69"><a href="#cb257-69" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy       </span></span>
<span id="cb257-70"><a href="#cb257-70" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64   </span></span>
<span id="cb257-71"><a href="#cb257-71" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼────────────────────────────</span></span>
<span id="cb257-72"><a href="#cb257-72" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.0234761</span></span>
<span id="cb257-73"><a href="#cb257-73" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_trochObs  0.0309999</span></span>
<span id="cb257-74"><a href="#cb257-74" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_plumb     0.0345515</span></span>
<span id="cb257-75"><a href="#cb257-75" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_trochObs     0.0320461</span></span>
<span id="cb257-76"><a href="#cb257-76" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_plumb        0.0351086</span></span>
<span id="cb257-77"><a href="#cb257-77" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ trochObs_plumb   0.0305924 =#</span></span>
<span id="cb257-78"><a href="#cb257-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-79"><a href="#cb257-79" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb257-80"><a href="#cb257-80" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb257-81"><a href="#cb257-81" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb257-82"><a href="#cb257-82" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb257-83"><a href="#cb257-83" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb257-84"><a href="#cb257-84" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb257-85"><a href="#cb257-85" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb257-86"><a href="#cb257-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-87"><a href="#cb257-87" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb257-88"><a href="#cb257-88" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb257-89"><a href="#cb257-89" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb257-90"><a href="#cb257-90" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb257-91"><a href="#cb257-91" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb257-92"><a href="#cb257-92" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb257-93"><a href="#cb257-93" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb257-94"><a href="#cb257-94" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb257-95"><a href="#cb257-95" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb257-96"><a href="#cb257-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-97"><a href="#cb257-97" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb257-98"><a href="#cb257-98" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">150</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb257-99"><a href="#cb257-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-100"><a href="#cb257-100" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb257-101"><a href="#cb257-101" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb257-102"><a href="#cb257-102" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb257-103"><a href="#cb257-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-104"><a href="#cb257-104" aria-hidden="true" tabindex="-1"></a><span class="co"># sort based on original_plot_order, and then together with function below will arrange individuals in population order within clusters:</span></span>
<span id="cb257-105"><a href="#cb257-105" aria-hidden="true" tabindex="-1"></a>sortOrder <span class="op">=</span> <span class="fu">sortperm</span>(indMetadataforGBI.original_plot_order, rev<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb257-106"><a href="#cb257-106" aria-hidden="true" tabindex="-1"></a>indMetadataforGBI <span class="op">=</span> indMetadataforGBI[sortOrder, <span class="op">:</span>]</span>
<span id="cb257-107"><a href="#cb257-107" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genosForGBI[sortOrder, <span class="op">:</span>]</span>
<span id="cb257-108"><a href="#cb257-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-109"><a href="#cb257-109" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb257-110"><a href="#cb257-110" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb257-111"><a href="#cb257-111" aria-hidden="true" tabindex="-1"></a>                indFontSize<span class="op">=</span><span class="fl">6</span>, figureSize<span class="op">=</span>(<span class="fl">800</span>, <span class="fl">1800</span>),</span>
<span id="cb257-112"><a href="#cb257-112" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb257-113"><a href="#cb257-113" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb257-114"><a href="#cb257-114" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [64 2 72 63] and the sum of those is 201
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-97-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="97">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [104582339, 104582348, 104582352, 104654170, 104654177, 104654201, 104654235, 104655357, 104795328, 104795382  …  107808537, 107822070, 107822104, 107822186, 107920230, 108037413, 108050478, 108079656, 108089665, 108202755], <span class="ansi-bold">201×30 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  11 │ GW_Lane5_YK3               GW_Lane5_YK3               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 192 │ GW_Armando_plate2_JF15G02  GW_Armando_plate2_JF15G02  DV        plumb
 193 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb   ⋯
 194 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
 195 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
 196 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 197 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 198 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 199 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 200 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 201 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 27 columns and 180 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-6" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-6">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="c1d54c78" class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb260"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb260-1"><a href="#cb260-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb260-2"><a href="#cb260-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLudHet"</span>,</span>
<span id="cb260-3"><a href="#cb260-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb260-4"><a href="#cb260-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_trochObs"</span>,</span>
<span id="cb260-5"><a href="#cb260-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObs"</span>,</span>
<span id="cb260-6"><a href="#cb260-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObsHet"</span>,</span>
<span id="cb260-7"><a href="#cb260-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb260-8"><a href="#cb260-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumbHet"</span>,</span>
<span id="cb260-9"><a href="#cb260-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb260-10"><a href="#cb260-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-11"><a href="#cb260-11" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb260-12"><a href="#cb260-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"blue"</span>,</span>
<span id="cb260-13"><a href="#cb260-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb260-14"><a href="#cb260-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb260-15"><a href="#cb260-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb260-16"><a href="#cb260-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb260-17"><a href="#cb260-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb260-18"><a href="#cb260-18" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb260-19"><a href="#cb260-19" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb260-20"><a href="#cb260-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-21"><a href="#cb260-21" aria-hidden="true" tabindex="-1"></a>virLudHet <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span> </span>
<span id="cb260-22"><a href="#cb260-22" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb260-23"><a href="#cb260-23" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb260-24"><a href="#cb260-24" aria-hidden="true" tabindex="-1"></a>virLud_trochObs <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb260-25"><a href="#cb260-25" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb260-26"><a href="#cb260-26" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb260-27"><a href="#cb260-27" aria-hidden="true" tabindex="-1"></a>trochObsHet <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb260-28"><a href="#cb260-28" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb260-29"><a href="#cb260-29" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb260-30"><a href="#cb260-30" aria-hidden="true" tabindex="-1"></a>plumbHet <span class="op">=</span> (<span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb260-31"><a href="#cb260-31" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb260-32"><a href="#cb260-32" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb260-33"><a href="#cb260-33" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb260-34"><a href="#cb260-34" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb260-35"><a href="#cb260-35" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb260-36"><a href="#cb260-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-37"><a href="#cb260-37" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud virLudHet nit virLud_trochObs trochObs trochObsHet plumb plumbHet vir_plumb]</span>
<span id="cb260-38"><a href="#cb260-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-39"><a href="#cb260-39" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb260-40"><a href="#cb260-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-41"><a href="#cb260-41" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb260-42"><a href="#cb260-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb260-43"><a href="#cb260-43" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb260-44"><a href="#cb260-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb260-45"><a href="#cb260-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb260-46"><a href="#cb260-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-47"><a href="#cb260-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-48"><a href="#cb260-48" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb260-49"><a href="#cb260-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-50"><a href="#cb260-50" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb260-51"><a href="#cb260-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-52"><a href="#cb260-52" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb260-53"><a href="#cb260-53" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb260-54"><a href="#cb260-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb260-55"><a href="#cb260-55" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb260-56"><a href="#cb260-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-57"><a href="#cb260-57" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb260-58"><a href="#cb260-58" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb260-59"><a href="#cb260-59" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb260-60"><a href="#cb260-60" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb260-61"><a href="#cb260-61" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb260-62"><a href="#cb260-62" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb260-63"><a href="#cb260-63" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb260-64"><a href="#cb260-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-65"><a href="#cb260-65" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb260-66"><a href="#cb260-66" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb260-67"><a href="#cb260-67" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb260-68"><a href="#cb260-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-69"><a href="#cb260-69" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb260-70"><a href="#cb260-70" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb260-71"><a href="#cb260-71" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb260-72"><a href="#cb260-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-73"><a href="#cb260-73" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb260-74"><a href="#cb260-74" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb260-75"><a href="#cb260-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-76"><a href="#cb260-76" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb260-77"><a href="#cb260-77" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb260-78"><a href="#cb260-78" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb260-79"><a href="#cb260-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-80"><a href="#cb260-80" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb260-81"><a href="#cb260-81" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb260-82"><a href="#cb260-82" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb260-83"><a href="#cb260-83" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb260-84"><a href="#cb260-84" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-98-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="98">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [104582339, 104582348, 104582352, 104654170, 104654177, 104654201, 104654235, 104655357, 104795328, 104795382  …  107808537, 107822070, 107822104, 107822186, 107920230, 108037413, 108050478, 108079656, 108089665, 108202755], <span class="ansi-bold">82×30 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TTGW06      GW_Armando_plate1_TTGW06      SU        l
   5 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l ⋯
   6 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l
   7 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l
   8 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  11 │ GW_Armando_plate2_TTGW20      GW_Armando_plate2_TTGW20      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  73 │ GW_Armando_plate1_JF12G01     GW_Armando_plate1_JF12G01     ST        p
  74 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p ⋯
  75 │ GW_Armando_plate2_JF26G02     GW_Armando_plate2_JF26G02     ST        p
  76 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
  77 │ GW_Armando_plate2_JF11G02     GW_Armando_plate2_JF11G02     ST        p
  78 │ GW_Armando_plate2_JF24G01     GW_Armando_plate2_JF24G01     VB        p ⋯
  79 │ GW_Armando_plate1_JF09G02     GW_Armando_plate1_JF09G02     ST        p
  80 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  81 │ GW_Armando_plate2_JF14G02     GW_Armando_plate2_JF14G02     DV        p
  82 │ GW_Armando_plate2_JG10G03     GW_Armando_plate2_JG10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                  27 columns and 61 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-4" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-4">Show just the west area (without nitidus)</h3>
<div id="5d81113f" class="cell" data-execution_count="98">
<div class="sourceCode cell-code" id="cb263"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb263-1"><a href="#cb263-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb263-2"><a href="#cb263-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLudHet"</span>,</span>
<span id="cb263-3"><a href="#cb263-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_trochObs"</span>,</span>
<span id="cb263-4"><a href="#cb263-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObs"</span>,</span>
<span id="cb263-5"><a href="#cb263-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObsHet"</span>]</span>
<span id="cb263-6"><a href="#cb263-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-7"><a href="#cb263-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb263-8"><a href="#cb263-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"blue"</span>,</span>
<span id="cb263-9"><a href="#cb263-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellowgreen"</span>,</span>
<span id="cb263-10"><a href="#cb263-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb263-11"><a href="#cb263-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb263-12"><a href="#cb263-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-13"><a href="#cb263-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb263-14"><a href="#cb263-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb263-15"><a href="#cb263-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb263-16"><a href="#cb263-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb263-17"><a href="#cb263-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb263-18"><a href="#cb263-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb263-19"><a href="#cb263-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb263-20"><a href="#cb263-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb263-21"><a href="#cb263-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb263-22"><a href="#cb263-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb263-23"><a href="#cb263-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb263-24"><a href="#cb263-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-25"><a href="#cb263-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb263-26"><a href="#cb263-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-27"><a href="#cb263-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb263-28"><a href="#cb263-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb263-29"><a href="#cb263-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb263-30"><a href="#cb263-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-31"><a href="#cb263-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb263-32"><a href="#cb263-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb263-33"><a href="#cb263-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb263-34"><a href="#cb263-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb263-35"><a href="#cb263-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-99-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="99">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [104795328, 104795382, 104897991, 104924031, 105304008, 105412338, 105418462, 105460577, 105631673, 105659380  …  107406780, 107426628, 107480270, 107822070, 107822104, 107920230, 108037413, 108050478, 108079656, 108089665], <span class="ansi-bold">182×30 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   5 │ GW_Armando_plate1_TTGW06      GW_Armando_plate1_TTGW06      SU        l ⋯
   6 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l
   7 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l
   8 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l
   9 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
  10 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
  11 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 173 │ GW_Lane5_LN10                 GW_Lane5_LN10                 LN        t
 174 │ GW_Lane5_LN19                 GW_Lane5_LN19                 LN        t ⋯
 175 │ GW_Lane5_LN8                  GW_Lane5_LN8                  LN        t
 176 │ GW_Lane5_MN9                  GW_Lane5_MN9                  MN        t
 177 │ GW_Liz_GBS_Liz5190            GW_Liz_GBS_Liz5190            ML        l
 178 │ GW_Lane5_MN1                  GW_Lane5_MN1                  MN        t ⋯
 179 │ GW_Lane5_LN12                 GW_Lane5_LN12                 LN        t
 180 │ GW_Armando_plate2_TTGW60      GW_Armando_plate2_TTGW60      SP        l
 181 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l
 182 │ GW_Liz_GBS_Liz5193            GW_Liz_GBS_Liz5193            ML        l ⋯
<span class="ansi-cyan-fg">                                                 27 columns and 161 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-5" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-5">Show just the east area</h3>
<div id="7deaca58" class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb266"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb266-1"><a href="#cb266-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"trochObs"</span>,</span>
<span id="cb266-2"><a href="#cb266-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"trochObsHet"</span>,</span>
<span id="cb266-3"><a href="#cb266-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>,</span>
<span id="cb266-4"><a href="#cb266-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumbHet"</span>]</span>
<span id="cb266-5"><a href="#cb266-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb266-6"><a href="#cb266-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb266-7"><a href="#cb266-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"yellow"</span>,</span>
<span id="cb266-8"><a href="#cb266-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb266-9"><a href="#cb266-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb266-10"><a href="#cb266-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb266-11"><a href="#cb266-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb266-12"><a href="#cb266-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb266-13"><a href="#cb266-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb266-14"><a href="#cb266-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb266-15"><a href="#cb266-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb266-16"><a href="#cb266-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb266-17"><a href="#cb266-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb266-18"><a href="#cb266-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb266-19"><a href="#cb266-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb266-20"><a href="#cb266-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb266-21"><a href="#cb266-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb266-22"><a href="#cb266-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb266-23"><a href="#cb266-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb266-24"><a href="#cb266-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb266-25"><a href="#cb266-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb266-26"><a href="#cb266-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb266-27"><a href="#cb266-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb266-28"><a href="#cb266-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb266-29"><a href="#cb266-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb266-30"><a href="#cb266-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb266-31"><a href="#cb266-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb266-32"><a href="#cb266-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb266-33"><a href="#cb266-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-100-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="100">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 2 … 2 2; 1 2 … 2 2], [104582352, 104655357, 104864925, 104975314, 105059216, 105203277, 105386932, 105410394, 105596913, 105596914  …  107406686, 107406738, 107426628, 107480270, 107808537, 107822070, 107822104, 107920230, 108050478, 108089665], <span class="ansi-bold">145×30 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW01   GW_Armando_plate1_TTGW01   MN        troch_M ⋯
   2 │ GW_Armando_plate1_TTGW55   GW_Armando_plate1_TTGW55   NG        lud_Nai
   3 │ GW_Armando_plate1_TTGW63   GW_Armando_plate1_TTGW63   SP        lud_Spi
   4 │ GW_Armando_plate1_TTGW64   GW_Armando_plate1_TTGW64   SP        lud_Spi
   5 │ GW_Armando_plate1_TTGW65   GW_Armando_plate1_TTGW65   SP        lud_Spi ⋯
   6 │ GW_Armando_plate1_TTGW66   GW_Armando_plate1_TTGW66   SP        lud_Spi
   7 │ GW_Armando_plate1_TTGW68   GW_Armando_plate1_TTGW68   SP        lud_Spi
   8 │ GW_Armando_plate1_TTGW70   GW_Armando_plate1_TTGW70   SA        lud_Sat
   9 │ GW_Armando_plate1_TTGW72   GW_Armando_plate1_TTGW72   SA        lud_Sat ⋯
  10 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
  11 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
 136 │ GW_Lane5_AN1               GW_Lane5_AN1               AN        plumb
 137 │ GW_Lane5_BK2               GW_Lane5_BK2               BK        plumb   ⋯
 138 │ GW_Lane5_ST12              GW_Lane5_ST12              ST        plumb
 139 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
 140 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 141 │ GW_Armando_plate2_JF26G02  GW_Armando_plate2_JF26G02  ST        plumb   ⋯
 142 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 143 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
 144 │ GW_Armando_plate2_JF11G02  GW_Armando_plate2_JF11G02  ST        plumb
 145 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb   ⋯
<span class="ansi-cyan-fg">                                                 27 columns and 124 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-5" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-5">Show just the northern area</h3>
<div id="b823c359" class="cell" data-execution_count="100">
<div class="sourceCode cell-code" id="cb269"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb269-1"><a href="#cb269-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb269-2"><a href="#cb269-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLudHet"</span>,</span>
<span id="cb269-3"><a href="#cb269-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb269-4"><a href="#cb269-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>,</span>
<span id="cb269-5"><a href="#cb269-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumbHet"</span>]</span>
<span id="cb269-6"><a href="#cb269-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-7"><a href="#cb269-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb269-8"><a href="#cb269-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"blue"</span>,</span>
<span id="cb269-9"><a href="#cb269-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb269-10"><a href="#cb269-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb269-11"><a href="#cb269-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb269-12"><a href="#cb269-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-13"><a href="#cb269-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb269-14"><a href="#cb269-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb269-15"><a href="#cb269-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb269-16"><a href="#cb269-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb269-17"><a href="#cb269-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb269-18"><a href="#cb269-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb269-19"><a href="#cb269-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb269-20"><a href="#cb269-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb269-21"><a href="#cb269-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb269-22"><a href="#cb269-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb269-23"><a href="#cb269-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb269-24"><a href="#cb269-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-25"><a href="#cb269-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb269-26"><a href="#cb269-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-27"><a href="#cb269-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb269-28"><a href="#cb269-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb269-29"><a href="#cb269-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb269-30"><a href="#cb269-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-31"><a href="#cb269-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb269-32"><a href="#cb269-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb269-33"><a href="#cb269-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb269-34"><a href="#cb269-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb269-35"><a href="#cb269-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-101-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="101">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 1 … 0 0; … ; 0 1 … 1 1; 0 1 … 1 1], [104582352, 104655357, 104795328, 104795382, 104864925, 104897991, 104924031, 104975314, 105059216, 105203277  …  106994418, 106994484, 107109035, 107109045, 107251168, 107406738, 107406780, 107808537, 108037413, 108079656], <span class="ansi-bold">144×30 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_JF16G01     GW_Armando_plate1_JF16G01     DV_vi     p
   4 │ GW_Armando_plate1_TTGW06      GW_Armando_plate1_TTGW06      SU        l
   5 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l ⋯
   6 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l
   7 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l
   8 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  11 │ GW_Armando_plate2_TTGW20      GW_Armando_plate2_TTGW20      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 135 │ GW_Armando_plate1_JF12G01     GW_Armando_plate1_JF12G01     ST        p
 136 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p ⋯
 137 │ GW_Armando_plate2_JF26G02     GW_Armando_plate2_JF26G02     ST        p
 138 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
 139 │ GW_Armando_plate1_JF09G02     GW_Armando_plate1_JF09G02     ST        p
 140 │ GW_Armando_plate2_JF11G02     GW_Armando_plate2_JF11G02     ST        p ⋯
 141 │ GW_Armando_plate2_JF24G01     GW_Armando_plate2_JF24G01     VB        p
 142 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 143 │ GW_Armando_plate2_JF14G02     GW_Armando_plate2_JF14G02     DV        p
 144 │ GW_Armando_plate2_JG10G03     GW_Armando_plate2_JG10G03     ST        p ⋯
<span class="ansi-cyan-fg">                                                 27 columns and 123 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="same-for-chr-18" class="level2">
<h2 class="anchored" data-anchor-id="same-for-chr-18">Same for chr 18</h2>
<div id="e7021579" class="cell" data-execution_count="101">
<div class="sourceCode cell-code" id="cb272"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb272-1"><a href="#cb272-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb272-2"><a href="#cb272-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw18"</span></span>
<span id="cb272-3"><a href="#cb272-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-4"><a href="#cb272-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb272-5"><a href="#cb272-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb272-6"><a href="#cb272-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb272-7"><a href="#cb272-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb272-8"><a href="#cb272-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb272-9"><a href="#cb272-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb272-10"><a href="#cb272-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb272-11"><a href="#cb272-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-12"><a href="#cb272-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb272-13"><a href="#cb272-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb272-14"><a href="#cb272-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-15"><a href="#cb272-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb272-16"><a href="#cb272-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb272-17"><a href="#cb272-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb272-18"><a href="#cb272-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb272-19"><a href="#cb272-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-20"><a href="#cb272-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb272-21"><a href="#cb272-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-22"><a href="#cb272-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb272-23"><a href="#cb272-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb272-24"><a href="#cb272-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-25"><a href="#cb272-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb272-26"><a href="#cb272-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-27"><a href="#cb272-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb272-28"><a href="#cb272-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-29"><a href="#cb272-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb272-30"><a href="#cb272-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb272-31"><a href="#cb272-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-32"><a href="#cb272-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb272-33"><a href="#cb272-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb272-34"><a href="#cb272-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb272-35"><a href="#cb272-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb272-36"><a href="#cb272-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb272-37"><a href="#cb272-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb272-38"><a href="#cb272-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb272-39"><a href="#cb272-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb272-40"><a href="#cb272-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-41"><a href="#cb272-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb272-42"><a href="#cb272-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-43"><a href="#cb272-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb272-44"><a href="#cb272-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb272-45"><a href="#cb272-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb272-46"><a href="#cb272-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb272-47"><a href="#cb272-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb272-48"><a href="#cb272-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb272-49"><a href="#cb272-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb272-50"><a href="#cb272-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb272-51"><a href="#cb272-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb272-52"><a href="#cb272-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb272-53"><a href="#cb272-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb272-54"><a href="#cb272-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb272-55"><a href="#cb272-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-56"><a href="#cb272-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb272-57"><a href="#cb272-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb272-58"><a href="#cb272-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-59"><a href="#cb272-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb272-60"><a href="#cb272-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.55</span>) </span>
<span id="cb272-61"><a href="#cb272-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-62"><a href="#cb272-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb272-63"><a href="#cb272-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-64"><a href="#cb272-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb272-65"><a href="#cb272-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb272-66"><a href="#cb272-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb272-67"><a href="#cb272-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb272-68"><a href="#cb272-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb272-69"><a href="#cb272-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb272-70"><a href="#cb272-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb272-71"><a href="#cb272-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb272-72"><a href="#cb272-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb272-73"><a href="#cb272-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb272-74"><a href="#cb272-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb272-75"><a href="#cb272-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-102-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-102-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="102">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-7" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-7">Save the individual colors in the metadata</h3>
<div id="b64295fa" class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb277"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb277-1"><a href="#cb277-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb277-2"><a href="#cb277-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb277-3"><a href="#cb277-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb277-4"><a href="#cb277-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb277-5"><a href="#cb277-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb277-6"><a href="#cb277-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="103">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-7" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-7">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="b8a72fb5" class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb279"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb279-1"><a href="#cb279-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb279-2"><a href="#cb279-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb279-3"><a href="#cb279-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb279-4"><a href="#cb279-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb279-5"><a href="#cb279-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb279-6"><a href="#cb279-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb279-7"><a href="#cb279-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb279-8"><a href="#cb279-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb279-9"><a href="#cb279-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb279-10"><a href="#cb279-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb279-11"><a href="#cb279-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb279-12"><a href="#cb279-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-104-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="104">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-7" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-7">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="047a3edb" class="cell" data-execution_count="104">
<div class="sourceCode cell-code" id="cb282"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb282-1"><a href="#cb282-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb282-2"><a href="#cb282-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb282-3"><a href="#cb282-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb282-4"><a href="#cb282-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb282-5"><a href="#cb282-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb282-6"><a href="#cb282-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb282-7"><a href="#cb282-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb282-8"><a href="#cb282-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb282-9"><a href="#cb282-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb282-10"><a href="#cb282-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb282-11"><a href="#cb282-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb282-12"><a href="#cb282-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-105-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="105">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-18-high-hetvar-region-there-are-5-clear-haplogroups-vir-and-lud-separated-though-on-pc3-though-not-clearly-enough-to-indicate-as-different-in-summary-plot.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-18-high-hetvar-region-there-are-5-clear-haplogroups-vir-and-lud-separated-though-on-pc3-though-not-clearly-enough-to-indicate-as-different-in-summary-plot.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 18 high HetVar region, there are 5 clear haplogroups (vir and lud separated though on PC3, though not clearly enough to indicate as different in summary plot). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="85618a5a" class="cell" data-execution_count="105">
<div class="sourceCode cell-code" id="cb285"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb285-1"><a href="#cb285-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb285-2"><a href="#cb285-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb285-3"><a href="#cb285-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb285-4"><a href="#cb285-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb285-5"><a href="#cb285-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb285-6"><a href="#cb285-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-7"><a href="#cb285-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb285-8"><a href="#cb285-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb285-9"><a href="#cb285-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb285-10"><a href="#cb285-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb285-11"><a href="#cb285-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb285-12"><a href="#cb285-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-13"><a href="#cb285-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">7</span>) <span class="op">.&amp;</span> </span>
<span id="cb285-14"><a href="#cb285-14" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb285-15"><a href="#cb285-15" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb285-16"><a href="#cb285-16" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb285-17"><a href="#cb285-17" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb285-18"><a href="#cb285-18" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb285-19"><a href="#cb285-19" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb285-20"><a href="#cb285-20" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb285-21"><a href="#cb285-21" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb285-22"><a href="#cb285-22" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb285-23"><a href="#cb285-23" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb285-24"><a href="#cb285-24" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb285-25"><a href="#cb285-25" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb285-26"><a href="#cb285-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-27"><a href="#cb285-27" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb285-28"><a href="#cb285-28" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb285-29"><a href="#cb285-29" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb285-30"><a href="#cb285-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb285-31"><a href="#cb285-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb285-32"><a href="#cb285-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb285-33"><a href="#cb285-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-34"><a href="#cb285-34" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb285-35"><a href="#cb285-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-36"><a href="#cb285-36" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb285-37"><a href="#cb285-37" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb285-38"><a href="#cb285-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-39"><a href="#cb285-39" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb285-40"><a href="#cb285-40" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb285-41"><a href="#cb285-41" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb285-42"><a href="#cb285-42" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb285-43"><a href="#cb285-43" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb285-44"><a href="#cb285-44" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb285-45"><a href="#cb285-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb285-46"><a href="#cb285-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-47"><a href="#cb285-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb285-48"><a href="#cb285-48" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb285-49"><a href="#cb285-49" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb285-50"><a href="#cb285-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-51"><a href="#cb285-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb285-52"><a href="#cb285-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-53"><a href="#cb285-53" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb285-54"><a href="#cb285-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-55"><a href="#cb285-55" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb285-56"><a href="#cb285-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-57"><a href="#cb285-57" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb285-58"><a href="#cb285-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-59"><a href="#cb285-59" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb285-60"><a href="#cb285-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-61"><a href="#cb285-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb285-62"><a href="#cb285-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-63"><a href="#cb285-63" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb285-64"><a href="#cb285-64" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb285-65"><a href="#cb285-65" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb285-66"><a href="#cb285-66" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb285-67"><a href="#cb285-67" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb285-68"><a href="#cb285-68" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.0110074</span></span>
<span id="cb285-69"><a href="#cb285-69" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00453689</span></span>
<span id="cb285-70"><a href="#cb285-70" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00973106</span></span>
<span id="cb285-71"><a href="#cb285-71" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.0123218</span></span>
<span id="cb285-72"><a href="#cb285-72" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00925472 =#</span></span>
<span id="cb285-73"><a href="#cb285-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-74"><a href="#cb285-74" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb285-75"><a href="#cb285-75" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb285-76"><a href="#cb285-76" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb285-77"><a href="#cb285-77" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb285-78"><a href="#cb285-78" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb285-79"><a href="#cb285-79" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.0263493</span></span>
<span id="cb285-80"><a href="#cb285-80" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0361335</span></span>
<span id="cb285-81"><a href="#cb285-81" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0359267</span></span>
<span id="cb285-82"><a href="#cb285-82" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0395363</span></span>
<span id="cb285-83"><a href="#cb285-83" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0371472</span></span>
<span id="cb285-84"><a href="#cb285-84" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0377076</span></span>
<span id="cb285-85"><a href="#cb285-85" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.0400618</span></span>
<span id="cb285-86"><a href="#cb285-86" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0169656</span></span>
<span id="cb285-87"><a href="#cb285-87" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0287838</span></span>
<span id="cb285-88"><a href="#cb285-88" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0290661 =#</span></span>
<span id="cb285-89"><a href="#cb285-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-90"><a href="#cb285-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb285-91"><a href="#cb285-91" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb285-92"><a href="#cb285-92" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb285-93"><a href="#cb285-93" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb285-94"><a href="#cb285-94" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb285-95"><a href="#cb285-95" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb285-96"><a href="#cb285-96" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb285-97"><a href="#cb285-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-98"><a href="#cb285-98" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb285-99"><a href="#cb285-99" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb285-100"><a href="#cb285-100" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb285-101"><a href="#cb285-101" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb285-102"><a href="#cb285-102" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb285-103"><a href="#cb285-103" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb285-104"><a href="#cb285-104" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb285-105"><a href="#cb285-105" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb285-106"><a href="#cb285-106" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb285-107"><a href="#cb285-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-108"><a href="#cb285-108" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb285-109"><a href="#cb285-109" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb285-110"><a href="#cb285-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-111"><a href="#cb285-111" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb285-112"><a href="#cb285-112" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb285-113"><a href="#cb285-113" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb285-114"><a href="#cb285-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-115"><a href="#cb285-115" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb285-116"><a href="#cb285-116" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb285-117"><a href="#cb285-117" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb285-118"><a href="#cb285-118" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [72 2 76 4 70] and the sum of those is 224
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-106-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="106">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 0 … 0 2; 2 0 … 0 2], [6108876, 6190195, 6293642, 6345626, 6386615, 6414325, 6414334, 6456723, 6456819, 6477300  …  8750639, 8750642, 8772183, 8773241, 8773281, 8773283, 8784596, 8797460, 8829892, 8833826], <span class="ansi-bold">51×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l
   4 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l
   5 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l ⋯
   6 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l
   7 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   8 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   9 │ GW_Armando_plate2_TTGW08      GW_Armando_plate2_TTGW08      SU        l ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW24      GW_Armando_plate2_TTGW24      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  42 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p
  43 │ GW_Armando_plate1_JF20G01     GW_Armando_plate1_JF20G01     MB        p ⋯
  44 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p
  45 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  46 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  47 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p ⋯
  48 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  49 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  50 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  51 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p ⋯
<span class="ansi-cyan-fg">                                                  29 columns and 30 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-7" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-7">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="bd6c67ec" class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb288"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb288-1"><a href="#cb288-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb288-2"><a href="#cb288-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb288-3"><a href="#cb288-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb288-4"><a href="#cb288-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb288-5"><a href="#cb288-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb288-6"><a href="#cb288-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs_plumb"</span>,</span>
<span id="cb288-7"><a href="#cb288-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb288-8"><a href="#cb288-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb288-9"><a href="#cb288-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-10"><a href="#cb288-10" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb288-11"><a href="#cb288-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb288-12"><a href="#cb288-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb288-13"><a href="#cb288-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb288-14"><a href="#cb288-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb288-15"><a href="#cb288-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"darkorange1"</span>,</span>
<span id="cb288-16"><a href="#cb288-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb288-17"><a href="#cb288-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb288-18"><a href="#cb288-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-19"><a href="#cb288-19" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">0.5</span>) <span class="op">.&amp;</span></span>
<span id="cb288-20"><a href="#cb288-20" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">0.25</span>) <span class="op">.&amp;</span></span>
<span id="cb288-21"><a href="#cb288-21" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb288-22"><a href="#cb288-22" aria-hidden="true" tabindex="-1"></a><span class="co">#trochObsHet = (-2 .&lt; PCAmodelAll.metadata.PC1 .&lt; 2.5) .&amp;</span></span>
<span id="cb288-23"><a href="#cb288-23" aria-hidden="true" tabindex="-1"></a>            <span class="co"># (PCAmodelAll.metadata.PC2 .&lt; -3) .&amp;</span></span>
<span id="cb288-24"><a href="#cb288-24" aria-hidden="true" tabindex="-1"></a>            <span class="co"># .!indSelection_lowIndHetStan</span></span>
<span id="cb288-25"><a href="#cb288-25" aria-hidden="true" tabindex="-1"></a>obs_plumb <span class="op">=</span> (<span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span> </span>
<span id="cb288-26"><a href="#cb288-26" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb288-27"><a href="#cb288-27" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb288-28"><a href="#cb288-28" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb288-29"><a href="#cb288-29" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb288-30"><a href="#cb288-30" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb288-31"><a href="#cb288-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-32"><a href="#cb288-32" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit virLud_troch troch obs obs_plumb plumb vir_plumb]</span>
<span id="cb288-33"><a href="#cb288-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-34"><a href="#cb288-34" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb288-35"><a href="#cb288-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-36"><a href="#cb288-36" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb288-37"><a href="#cb288-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb288-38"><a href="#cb288-38" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb288-39"><a href="#cb288-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb288-40"><a href="#cb288-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb288-41"><a href="#cb288-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-42"><a href="#cb288-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-43"><a href="#cb288-43" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb288-44"><a href="#cb288-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-45"><a href="#cb288-45" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb288-46"><a href="#cb288-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-47"><a href="#cb288-47" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb288-48"><a href="#cb288-48" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb288-49"><a href="#cb288-49" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb288-50"><a href="#cb288-50" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb288-51"><a href="#cb288-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-52"><a href="#cb288-52" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb288-53"><a href="#cb288-53" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb288-54"><a href="#cb288-54" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb288-55"><a href="#cb288-55" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb288-56"><a href="#cb288-56" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb288-57"><a href="#cb288-57" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb288-58"><a href="#cb288-58" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb288-59"><a href="#cb288-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-60"><a href="#cb288-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb288-61"><a href="#cb288-61" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb288-62"><a href="#cb288-62" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb288-63"><a href="#cb288-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-64"><a href="#cb288-64" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb288-65"><a href="#cb288-65" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb288-66"><a href="#cb288-66" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb288-67"><a href="#cb288-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-68"><a href="#cb288-68" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb288-69"><a href="#cb288-69" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb288-70"><a href="#cb288-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-71"><a href="#cb288-71" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb288-72"><a href="#cb288-72" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb288-73"><a href="#cb288-73" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb288-74"><a href="#cb288-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-75"><a href="#cb288-75" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb288-76"><a href="#cb288-76" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb288-77"><a href="#cb288-77" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb288-78"><a href="#cb288-78" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb288-79"><a href="#cb288-79" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-107-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="107">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 0 … 0 1; 1 0 … 0 -1], [6108876, 6190195, 6293642, 6345626, 6386615, 6414325, 6414334, 6456723, 6456819, 6477300  …  8750639, 8750642, 8772183, 8773241, 8773281, 8773283, 8784596, 8797460, 8829892, 8833826], <span class="ansi-bold">70×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v
   3 │ GW_Armando_plate1_TTGW17      GW_Armando_plate1_TTGW17      TH        l
   4 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l
   5 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l ⋯
   6 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l
   7 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   8 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   9 │ GW_Armando_plate2_TTGW08      GW_Armando_plate2_TTGW08      SU        l ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW24      GW_Armando_plate2_TTGW24      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  61 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  62 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p ⋯
  63 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  64 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  65 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  66 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p ⋯
  67 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  68 │ GW_Armando_plate2_JF13G02     GW_Armando_plate2_JF13G02     ST        p
  69 │ GW_Armando_plate1_JF08G02     GW_Armando_plate1_JF08G02     ST        p
  70 │ GW_Lane5_UY4                  GW_Lane5_UY4                  UY        p ⋯
<span class="ansi-cyan-fg">                                                  29 columns and 49 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-4" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-4">show GBI plot according to original groups and plot order</h3>
<div id="68b4b4a2" class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb291"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb291-1"><a href="#cb291-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb291-2"><a href="#cb291-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-3"><a href="#cb291-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb291-4"><a href="#cb291-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb291-5"><a href="#cb291-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb291-6"><a href="#cb291-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-7"><a href="#cb291-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb291-8"><a href="#cb291-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb291-9"><a href="#cb291-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb291-10"><a href="#cb291-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb291-11"><a href="#cb291-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-108-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="108">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 0 … 0 2; 2 0 … 0 2], [6108876, 6190195, 6293642, 6345626, 6386615, 6414325, 6414334, 6456723, 6456819, 6477300  …  8750639, 8750642, 8772183, 8773241, 8773281, 8773283, 8784596, 8797460, 8829892, 8833826], <span class="ansi-bold">70×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   3 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v
   4 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v
   5 │ GW_Lane5_TU2                  GW_Lane5_TU2                  TU        n ⋯
   6 │ GW_Lane5_TU1                  GW_Lane5_TU1                  TU        n
   7 │ GW_Armando_plate1_TTGW70      GW_Armando_plate1_TTGW70      SA        l
   8 │ GW_Armando_plate1_TTGW72      GW_Armando_plate1_TTGW72      SA        l
   9 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l ⋯
  10 │ GW_Armando_plate2_TTGW75      GW_Armando_plate2_TTGW75      SA        l
  11 │ GW_Armando_plate1_TTGW29      GW_Armando_plate1_TTGW29      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  61 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  62 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
  63 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  64 │ GW_Armando_plate2_JF13G02     GW_Armando_plate2_JF13G02     ST        p
  65 │ GW_Armando_plate1_JF08G02     GW_Armando_plate1_JF08G02     ST        p
  66 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p ⋯
  67 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p
  68 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  69 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  70 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p ⋯
<span class="ansi-cyan-fg">                                                  29 columns and 49 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-4" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-4">show same but with all individuals</h3>
<div id="9a5b2dd5" class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb293"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb293-1"><a href="#cb293-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb293-2"><a href="#cb293-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb293-3"><a href="#cb293-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb293-4"><a href="#cb293-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb293-5"><a href="#cb293-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb293-6"><a href="#cb293-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb293-7"><a href="#cb293-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb293-8"><a href="#cb293-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb293-9"><a href="#cb293-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb293-10"><a href="#cb293-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb293-11"><a href="#cb293-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb293-12"><a href="#cb293-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb293-13"><a href="#cb293-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb293-14"><a href="#cb293-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-109-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="109">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 0 … 0 2; 2 0 … 0 2], [6108876, 6190195, 6293642, 6345626, 6386615, 6414325, 6414334, 6456723, 6456819, 6477300  …  8750639, 8750642, 8772183, 8773241, 8773281, 8773283, 8784596, 8797460, 8829892, 8833826], <span class="ansi-bold">257×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 249 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 252 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 253 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 29 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops-2" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops-2">show same but with only vir and plumb pops</h3>
<div id="c60886a1" class="cell" data-execution_count="109">
<div class="sourceCode cell-code" id="cb295"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb295-1"><a href="#cb295-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb295-2"><a href="#cb295-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-3"><a href="#cb295-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb295-4"><a href="#cb295-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb295-5"><a href="#cb295-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-6"><a href="#cb295-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb295-7"><a href="#cb295-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb295-8"><a href="#cb295-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb295-9"><a href="#cb295-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb295-10"><a href="#cb295-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb295-11"><a href="#cb295-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb295-12"><a href="#cb295-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb295-13"><a href="#cb295-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb295-14"><a href="#cb295-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-15"><a href="#cb295-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb295-16"><a href="#cb295-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb295-17"><a href="#cb295-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-18"><a href="#cb295-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb295-19"><a href="#cb295-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-20"><a href="#cb295-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb295-21"><a href="#cb295-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb295-22"><a href="#cb295-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb295-23"><a href="#cb295-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb295-24"><a href="#cb295-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-110-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="110">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 0 2; 2 2 … 1 2], [6108876, 6293642, 6414325, 6456723, 6456819, 6477300, 6631300, 6812016, 6812064, 6838561  …  8668147, 8681264, 8750618, 8750639, 8750642, 8772183, 8773241, 8773281, 8784596, 8833826], <span class="ansi-bold">100×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  29 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-5" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-5">Show just the west area (without nitidus)</h3>
<div id="ceb48521" class="cell" data-execution_count="110">
<div class="sourceCode cell-code" id="cb297"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb297-1"><a href="#cb297-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb297-2"><a href="#cb297-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb297-3"><a href="#cb297-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb297-4"><a href="#cb297-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb297-5"><a href="#cb297-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb297-6"><a href="#cb297-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb297-7"><a href="#cb297-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb297-8"><a href="#cb297-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb297-9"><a href="#cb297-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb297-10"><a href="#cb297-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb297-11"><a href="#cb297-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb297-12"><a href="#cb297-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb297-13"><a href="#cb297-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb297-14"><a href="#cb297-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb297-15"><a href="#cb297-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb297-16"><a href="#cb297-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb297-17"><a href="#cb297-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb297-18"><a href="#cb297-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb297-19"><a href="#cb297-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb297-20"><a href="#cb297-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb297-21"><a href="#cb297-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsWest))</span>
<span id="cb297-22"><a href="#cb297-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb297-23"><a href="#cb297-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb297-24"><a href="#cb297-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb297-25"><a href="#cb297-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb297-26"><a href="#cb297-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb297-27"><a href="#cb297-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb297-28"><a href="#cb297-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb297-29"><a href="#cb297-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb297-30"><a href="#cb297-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb297-31"><a href="#cb297-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-111-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="111">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 1 2; 2 2 … 2 2], [6108876, 6414325, 6414334, 6456723, 6485917, 6812016, 6812064, 6838561, 6838565, 6945805  …  8647967, 8661608, 8668147, 8681264, 8681284, 8750618, 8750639, 8750642, 8797460, 8829892], <span class="ansi-bold">177×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 168 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 169 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L ⋯
 170 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 171 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
 172 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L
 173 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L ⋯
 174 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 175 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 176 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
 177 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 29 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-6" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-6">Show just the east area</h3>
<div id="c9630e72" class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb300"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb300-1"><a href="#cb300-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"obs"</span>,</span>
<span id="cb300-2"><a href="#cb300-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs_plumb"</span>,</span>
<span id="cb300-3"><a href="#cb300-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb300-4"><a href="#cb300-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-5"><a href="#cb300-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb300-6"><a href="#cb300-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"darkorange1"</span>,</span>
<span id="cb300-7"><a href="#cb300-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb300-8"><a href="#cb300-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-9"><a href="#cb300-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb300-10"><a href="#cb300-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb300-11"><a href="#cb300-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb300-12"><a href="#cb300-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb300-13"><a href="#cb300-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb300-14"><a href="#cb300-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb300-15"><a href="#cb300-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb300-16"><a href="#cb300-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb300-17"><a href="#cb300-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb300-18"><a href="#cb300-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb300-19"><a href="#cb300-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb300-20"><a href="#cb300-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-21"><a href="#cb300-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb300-22"><a href="#cb300-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-23"><a href="#cb300-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb300-24"><a href="#cb300-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb300-25"><a href="#cb300-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb300-26"><a href="#cb300-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb300-27"><a href="#cb300-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb300-28"><a href="#cb300-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb300-29"><a href="#cb300-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb300-30"><a href="#cb300-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb300-31"><a href="#cb300-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-112-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="112">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 1; … ; 2 2 … 2 2; 2 2 … 2 2], [6293642, 6414334, 6456819, 6477300, 6485917, 6631300, 6812081, 7047936, 7172541, 7435797  …  8636680, 8643165, 8661565, 8739750, 8772183, 8773241, 8773281, 8784596, 8829892, 8833826], <span class="ansi-bold">75×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Lane5_EM1               GW_Lane5_EM1               EM        troch_E ⋯
   2 │ GW_Lane5_DA4               GW_Lane5_DA4               XN        obs
   3 │ GW_Lane5_DA7               GW_Lane5_DA7               XN        obs
   4 │ GW_Lane5_DA2               GW_Lane5_DA2               XN        obs
   5 │ GW_Lane5_DA3               GW_Lane5_DA3               XN        obs     ⋯
   6 │ GW_Armando_plate1_RF20G01  GW_Armando_plate1_RF20G01  BJ        obs_plu
   7 │ GW_Armando_plate1_RF29G02  GW_Armando_plate1_RF29G02  BJ        obs_plu
   8 │ GW_Armando_plate2_RF29G01  GW_Armando_plate2_RF29G01  BJ        obs_plu
   9 │ GW_Lane5_BK2               GW_Lane5_BK2               BK        plumb   ⋯
  10 │ GW_Lane5_BK3               GW_Lane5_BK3               BK        plumb
  11 │ GW_Lane5_AN2               GW_Lane5_AN2               AN        plumb
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  66 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  67 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  68 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  69 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  70 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
  71 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb   ⋯
  72 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  73 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  74 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
  75 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  29 columns and 54 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-6" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-6">Show just the northern area</h3>
<div id="683139a2" class="cell" data-execution_count="112">
<div class="sourceCode cell-code" id="cb303"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb303-1"><a href="#cb303-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb303-2"><a href="#cb303-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb303-3"><a href="#cb303-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb303-4"><a href="#cb303-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-5"><a href="#cb303-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb303-6"><a href="#cb303-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb303-7"><a href="#cb303-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb303-8"><a href="#cb303-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-9"><a href="#cb303-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb303-10"><a href="#cb303-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb303-11"><a href="#cb303-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb303-12"><a href="#cb303-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb303-13"><a href="#cb303-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb303-14"><a href="#cb303-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb303-15"><a href="#cb303-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb303-16"><a href="#cb303-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb303-17"><a href="#cb303-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb303-18"><a href="#cb303-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb303-19"><a href="#cb303-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb303-20"><a href="#cb303-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-21"><a href="#cb303-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb303-22"><a href="#cb303-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-23"><a href="#cb303-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb303-24"><a href="#cb303-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb303-25"><a href="#cb303-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb303-26"><a href="#cb303-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb303-27"><a href="#cb303-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb303-28"><a href="#cb303-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb303-29"><a href="#cb303-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb303-30"><a href="#cb303-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb303-31"><a href="#cb303-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-113-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="113">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 0 2; 2 2 … 1 2], [6108876, 6293642, 6414325, 6456723, 6456819, 6477300, 6631300, 6812016, 6812064, 6838561  …  8668147, 8681264, 8750618, 8750639, 8750642, 8772183, 8773241, 8773281, 8784596, 8833826], <span class="ansi-bold">145×32 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   4 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 136 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 137 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 138 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 139 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 140 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 141 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
 142 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 143 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 144 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 145 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 29 columns and 124 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="tried-chr-12-but-sort-of-a-mess.-the-ludlowi-samples-fall-in-all-clusters-even-plumb" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-12-but-sort-of-a-mess.-the-ludlowi-samples-fall-in-all-clusters-even-plumb">Tried chr 12 but sort of a mess. The ludlowi samples fall in all clusters, even plumb!</h2>
</section>
<section id="same-for-chr-13" class="level2">
<h2 class="anchored" data-anchor-id="same-for-chr-13">Same for chr 13</h2>
<div id="d09d2d5e" class="cell" data-execution_count="113">
<div class="sourceCode cell-code" id="cb306"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb306-1"><a href="#cb306-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb306-2"><a href="#cb306-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw13"</span></span>
<span id="cb306-3"><a href="#cb306-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-4"><a href="#cb306-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb306-5"><a href="#cb306-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb306-6"><a href="#cb306-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb306-7"><a href="#cb306-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb306-8"><a href="#cb306-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb306-9"><a href="#cb306-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb306-10"><a href="#cb306-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb306-11"><a href="#cb306-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-12"><a href="#cb306-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb306-13"><a href="#cb306-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb306-14"><a href="#cb306-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-15"><a href="#cb306-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb306-16"><a href="#cb306-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb306-17"><a href="#cb306-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb306-18"><a href="#cb306-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb306-19"><a href="#cb306-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-20"><a href="#cb306-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb306-21"><a href="#cb306-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-22"><a href="#cb306-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb306-23"><a href="#cb306-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb306-24"><a href="#cb306-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-25"><a href="#cb306-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb306-26"><a href="#cb306-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-27"><a href="#cb306-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb306-28"><a href="#cb306-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-29"><a href="#cb306-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb306-30"><a href="#cb306-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb306-31"><a href="#cb306-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-32"><a href="#cb306-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb306-33"><a href="#cb306-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb306-34"><a href="#cb306-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb306-35"><a href="#cb306-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb306-36"><a href="#cb306-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb306-37"><a href="#cb306-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb306-38"><a href="#cb306-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb306-39"><a href="#cb306-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb306-40"><a href="#cb306-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-41"><a href="#cb306-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb306-42"><a href="#cb306-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-43"><a href="#cb306-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb306-44"><a href="#cb306-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb306-45"><a href="#cb306-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb306-46"><a href="#cb306-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb306-47"><a href="#cb306-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb306-48"><a href="#cb306-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb306-49"><a href="#cb306-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb306-50"><a href="#cb306-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb306-51"><a href="#cb306-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb306-52"><a href="#cb306-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb306-53"><a href="#cb306-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb306-54"><a href="#cb306-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb306-55"><a href="#cb306-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-56"><a href="#cb306-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb306-57"><a href="#cb306-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb306-58"><a href="#cb306-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-59"><a href="#cb306-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb306-60"><a href="#cb306-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.75</span>) </span>
<span id="cb306-61"><a href="#cb306-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-62"><a href="#cb306-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb306-63"><a href="#cb306-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-64"><a href="#cb306-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb306-65"><a href="#cb306-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb306-66"><a href="#cb306-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb306-67"><a href="#cb306-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb306-68"><a href="#cb306-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb306-69"><a href="#cb306-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb306-70"><a href="#cb306-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb306-71"><a href="#cb306-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb306-72"><a href="#cb306-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb306-73"><a href="#cb306-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb306-74"><a href="#cb306-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb306-75"><a href="#cb306-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>3×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw13</td>
<td style="text-align: right;">13574177</td>
<td style="text-align: right;">13722280</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw13</td>
<td style="text-align: right;">14099239</td>
<td style="text-align: right;">15243036</td>
</tr>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">3</td>
<td style="text-align: left;">gw13</td>
<td style="text-align: right;">15413381</td>
<td style="text-align: right;">15607553</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-114-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-114-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="114">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-8" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-8">Save the individual colors in the metadata</h3>
<div id="21d678f2" class="cell" data-execution_count="114">
<div class="sourceCode cell-code" id="cb311"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb311-1"><a href="#cb311-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb311-2"><a href="#cb311-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb311-3"><a href="#cb311-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb311-4"><a href="#cb311-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb311-5"><a href="#cb311-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb311-6"><a href="#cb311-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="115">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-8" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-8">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="3baa9770" class="cell" data-execution_count="115">
<div class="sourceCode cell-code" id="cb313"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb313-1"><a href="#cb313-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb313-2"><a href="#cb313-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb313-3"><a href="#cb313-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb313-4"><a href="#cb313-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb313-5"><a href="#cb313-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb313-6"><a href="#cb313-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb313-7"><a href="#cb313-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb313-8"><a href="#cb313-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb313-9"><a href="#cb313-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb313-10"><a href="#cb313-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb313-11"><a href="#cb313-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb313-12"><a href="#cb313-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-116-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="116">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-8" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-8">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="98aa9229" class="cell" data-execution_count="116">
<div class="sourceCode cell-code" id="cb316"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb316-1"><a href="#cb316-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb316-2"><a href="#cb316-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb316-3"><a href="#cb316-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb316-4"><a href="#cb316-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb316-5"><a href="#cb316-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb316-6"><a href="#cb316-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb316-7"><a href="#cb316-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb316-8"><a href="#cb316-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb316-9"><a href="#cb316-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb316-10"><a href="#cb316-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb316-11"><a href="#cb316-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb316-12"><a href="#cb316-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-117-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="117">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-13-high-hetvar-region-there-are-6-clear-haplogroups-vir-and-lud-separated-cleanly-on-pc3-with-one-hetero-between-them.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-13-high-hetvar-region-there-are-6-clear-haplogroups-vir-and-lud-separated-cleanly-on-pc3-with-one-hetero-between-them.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 13 high HetVar region, there are 6 clear haplogroups (vir and lud separated cleanly on PC3, with one hetero between them). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="3b24e5d9" class="cell" data-execution_count="117">
<div class="sourceCode cell-code" id="cb319"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb319-1"><a href="#cb319-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb319-2"><a href="#cb319-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb319-3"><a href="#cb319-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb319-4"><a href="#cb319-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb319-5"><a href="#cb319-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb319-6"><a href="#cb319-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb319-7"><a href="#cb319-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-8"><a href="#cb319-8" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb319-9"><a href="#cb319-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb319-10"><a href="#cb319-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb319-11"><a href="#cb319-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb319-12"><a href="#cb319-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb319-13"><a href="#cb319-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb319-14"><a href="#cb319-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-15"><a href="#cb319-15" aria-hidden="true" tabindex="-1"></a>vir <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb319-16"><a href="#cb319-16" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3) <span class="op">.&amp;</span> </span>
<span id="cb319-17"><a href="#cb319-17" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-18"><a href="#cb319-18" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb319-19"><a href="#cb319-19" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">3.5</span>) <span class="op">.&amp;</span></span>
<span id="cb319-20"><a href="#cb319-20" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-21"><a href="#cb319-21" aria-hidden="true" tabindex="-1"></a>lud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb319-22"><a href="#cb319-22" aria-hidden="true" tabindex="-1"></a>        (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span> </span>
<span id="cb319-23"><a href="#cb319-23" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-24"><a href="#cb319-24" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="op">-</span><span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb319-25"><a href="#cb319-25" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5.5</span>) <span class="op">.&amp;</span></span>
<span id="cb319-26"><a href="#cb319-26" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-27"><a href="#cb319-27" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb319-28"><a href="#cb319-28" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb319-29"><a href="#cb319-29" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-30"><a href="#cb319-30" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb319-31"><a href="#cb319-31" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb319-32"><a href="#cb319-32" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb319-33"><a href="#cb319-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-34"><a href="#cb319-34" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb319-35"><a href="#cb319-35" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[vir]</span>
<span id="cb319-36"><a href="#cb319-36" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb319-37"><a href="#cb319-37" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[lud]</span>
<span id="cb319-38"><a href="#cb319-38" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb319-39"><a href="#cb319-39" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb319-40"><a href="#cb319-40" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb319-41"><a href="#cb319-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-42"><a href="#cb319-42" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir nit lud troch obs plumb]</span>
<span id="cb319-43"><a href="#cb319-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-44"><a href="#cb319-44" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb319-45"><a href="#cb319-45" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb319-46"><a href="#cb319-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-47"><a href="#cb319-47" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb319-48"><a href="#cb319-48" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb319-49"><a href="#cb319-49" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb319-50"><a href="#cb319-50" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb319-51"><a href="#cb319-51" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb319-52"><a href="#cb319-52" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb319-53"><a href="#cb319-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb319-54"><a href="#cb319-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-55"><a href="#cb319-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb319-56"><a href="#cb319-56" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb319-57"><a href="#cb319-57" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb319-58"><a href="#cb319-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-59"><a href="#cb319-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb319-60"><a href="#cb319-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-61"><a href="#cb319-61" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb319-62"><a href="#cb319-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-63"><a href="#cb319-63" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb319-64"><a href="#cb319-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-65"><a href="#cb319-65" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb319-66"><a href="#cb319-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-67"><a href="#cb319-67" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb319-68"><a href="#cb319-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-69"><a href="#cb319-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb319-70"><a href="#cb319-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-71"><a href="#cb319-71" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb319-72"><a href="#cb319-72" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb319-73"><a href="#cb319-73" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb319-74"><a href="#cb319-74" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb319-75"><a href="#cb319-75" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb319-76"><a href="#cb319-76" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir      0.00875059</span></span>
<span id="cb319-77"><a href="#cb319-77" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00517962</span></span>
<span id="cb319-78"><a href="#cb319-78" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ lud      0.00819617</span></span>
<span id="cb319-79"><a href="#cb319-79" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ troch    0.00565913</span></span>
<span id="cb319-80"><a href="#cb319-80" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ obs      0.0090813</span></span>
<span id="cb319-81"><a href="#cb319-81" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ plumb    0.00929977 =#</span></span>
<span id="cb319-82"><a href="#cb319-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-83"><a href="#cb319-83" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb319-84"><a href="#cb319-84" aria-hidden="true" tabindex="-1"></a><span class="co">#= 15×2 DataFrame</span></span>
<span id="cb319-85"><a href="#cb319-85" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb319-86"><a href="#cb319-86" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb319-87"><a href="#cb319-87" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb319-88"><a href="#cb319-88" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir_nit       0.035675</span></span>
<span id="cb319-89"><a href="#cb319-89" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ vir_lud       0.0188542</span></span>
<span id="cb319-90"><a href="#cb319-90" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ vir_troch     0.0297034</span></span>
<span id="cb319-91"><a href="#cb319-91" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ vir_obs       0.028434</span></span>
<span id="cb319-92"><a href="#cb319-92" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ vir_plumb     0.0382774</span></span>
<span id="cb319-93"><a href="#cb319-93" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_lud       0.0377189</span></span>
<span id="cb319-94"><a href="#cb319-94" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_troch     0.0437711</span></span>
<span id="cb319-95"><a href="#cb319-95" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ nit_obs       0.0424561</span></span>
<span id="cb319-96"><a href="#cb319-96" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ nit_plumb     0.0482994</span></span>
<span id="cb319-97"><a href="#cb319-97" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ lud_troch     0.0303352</span></span>
<span id="cb319-98"><a href="#cb319-98" aria-hidden="true" tabindex="-1"></a><span class="co">  11 │ lud_obs       0.0294719</span></span>
<span id="cb319-99"><a href="#cb319-99" aria-hidden="true" tabindex="-1"></a><span class="co">  12 │ lud_plumb     0.0394332</span></span>
<span id="cb319-100"><a href="#cb319-100" aria-hidden="true" tabindex="-1"></a><span class="co">  13 │ troch_obs     0.0124742</span></span>
<span id="cb319-101"><a href="#cb319-101" aria-hidden="true" tabindex="-1"></a><span class="co">  14 │ troch_plumb   0.0313941</span></span>
<span id="cb319-102"><a href="#cb319-102" aria-hidden="true" tabindex="-1"></a><span class="co">  15 │ obs_plumb     0.0300717 =#</span></span>
<span id="cb319-103"><a href="#cb319-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-104"><a href="#cb319-104" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb319-105"><a href="#cb319-105" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb319-106"><a href="#cb319-106" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb319-107"><a href="#cb319-107" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb319-108"><a href="#cb319-108" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb319-109"><a href="#cb319-109" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb319-110"><a href="#cb319-110" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb319-111"><a href="#cb319-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-112"><a href="#cb319-112" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb319-113"><a href="#cb319-113" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb319-114"><a href="#cb319-114" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb319-115"><a href="#cb319-115" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb319-116"><a href="#cb319-116" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb319-117"><a href="#cb319-117" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb319-118"><a href="#cb319-118" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb319-119"><a href="#cb319-119" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb319-120"><a href="#cb319-120" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb319-121"><a href="#cb319-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-122"><a href="#cb319-122" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb319-123"><a href="#cb319-123" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb319-124"><a href="#cb319-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-125"><a href="#cb319-125" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb319-126"><a href="#cb319-126" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb319-127"><a href="#cb319-127" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb319-128"><a href="#cb319-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-129"><a href="#cb319-129" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb319-130"><a href="#cb319-130" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb319-131"><a href="#cb319-131" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb319-132"><a href="#cb319-132" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [38 2 41 67 5 68] and the sum of those is 221
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-118-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="118">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 1; 0 0 … 0 0; … ; 0 2 … 2 2; 0 2 … 2 2], [14109120, 14109196, 14109223, 14113908, 14117707, 14117822, 14121642, 14124306, 14127880, 14135008  …  15108025, 15108049, 15108085, 15116504, 15127266, 15129890, 15134320, 15134324, 15177202, 15177253], <span class="ansi-bold">67×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   4 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
   5 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
   6 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   9 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir     ⋯
  10 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
  11 │ GW_Lane5_TL12              GW_Lane5_TL12              TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  58 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  59 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb   ⋯
  60 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  61 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  62 │ GW_Armando_plate1_JF29G02  GW_Armando_plate1_JF29G02  ST        plumb
  63 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb   ⋯
  64 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  65 │ GW_Armando_plate1_JG08G01  GW_Armando_plate1_JG08G01  ST        plumb
  66 │ GW_Armando_plate1_JG17G01  GW_Armando_plate1_JG17G01  ST        plumb_v
  67 │ GW_Armando_plate2_JF08G04  GW_Armando_plate2_JF08G04  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 46 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-8" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-8">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="34104721" class="cell" data-execution_count="118">
<div class="sourceCode cell-code" id="cb322"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb322-1"><a href="#cb322-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb322-2"><a href="#cb322-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"vir_lud"</span>,</span>
<span id="cb322-3"><a href="#cb322-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb322-4"><a href="#cb322-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb322-5"><a href="#cb322-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud_troch"</span>,</span>
<span id="cb322-6"><a href="#cb322-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb322-7"><a href="#cb322-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb322-8"><a href="#cb322-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>,</span>
<span id="cb322-9"><a href="#cb322-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumbHet"</span>,</span>
<span id="cb322-10"><a href="#cb322-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"vir_plumb"</span>]</span>
<span id="cb322-11"><a href="#cb322-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-12"><a href="#cb322-12" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb322-13"><a href="#cb322-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"seagreen"</span>,</span>
<span id="cb322-14"><a href="#cb322-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb322-15"><a href="#cb322-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb322-16"><a href="#cb322-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green2"</span>,</span>
<span id="cb322-17"><a href="#cb322-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb322-18"><a href="#cb322-18" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb322-19"><a href="#cb322-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb322-20"><a href="#cb322-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb322-21"><a href="#cb322-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">"purple"</span>]</span>
<span id="cb322-22"><a href="#cb322-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-23"><a href="#cb322-23" aria-hidden="true" tabindex="-1"></a>vir_lud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb322-24"><a href="#cb322-24" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="fl">1</span>)</span>
<span id="cb322-25"><a href="#cb322-25" aria-hidden="true" tabindex="-1"></a>lud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb322-26"><a href="#cb322-26" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb322-27"><a href="#cb322-27" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb322-28"><a href="#cb322-28" aria-hidden="true" tabindex="-1"></a>plumbHet <span class="op">=</span> (<span class="fl">7</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb322-29"><a href="#cb322-29" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb322-30"><a href="#cb322-30" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb322-31"><a href="#cb322-31" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb322-32"><a href="#cb322-32" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb322-33"><a href="#cb322-33" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb322-34"><a href="#cb322-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-35"><a href="#cb322-35" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir vir_lud nit lud lud_troch troch obs plumb plumbHet vir_plumb]</span>
<span id="cb322-36"><a href="#cb322-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-37"><a href="#cb322-37" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb322-38"><a href="#cb322-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-39"><a href="#cb322-39" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb322-40"><a href="#cb322-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb322-41"><a href="#cb322-41" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb322-42"><a href="#cb322-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb322-43"><a href="#cb322-43" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb322-44"><a href="#cb322-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-45"><a href="#cb322-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-46"><a href="#cb322-46" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb322-47"><a href="#cb322-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-48"><a href="#cb322-48" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb322-49"><a href="#cb322-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-50"><a href="#cb322-50" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb322-51"><a href="#cb322-51" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb322-52"><a href="#cb322-52" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb322-53"><a href="#cb322-53" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb322-54"><a href="#cb322-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-55"><a href="#cb322-55" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb322-56"><a href="#cb322-56" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb322-57"><a href="#cb322-57" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb322-58"><a href="#cb322-58" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb322-59"><a href="#cb322-59" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb322-60"><a href="#cb322-60" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb322-61"><a href="#cb322-61" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb322-62"><a href="#cb322-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-63"><a href="#cb322-63" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb322-64"><a href="#cb322-64" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb322-65"><a href="#cb322-65" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb322-66"><a href="#cb322-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-67"><a href="#cb322-67" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb322-68"><a href="#cb322-68" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb322-69"><a href="#cb322-69" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb322-70"><a href="#cb322-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-71"><a href="#cb322-71" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb322-72"><a href="#cb322-72" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb322-73"><a href="#cb322-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-74"><a href="#cb322-74" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb322-75"><a href="#cb322-75" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb322-76"><a href="#cb322-76" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb322-77"><a href="#cb322-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-78"><a href="#cb322-78" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb322-79"><a href="#cb322-79" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb322-80"><a href="#cb322-80" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb322-81"><a href="#cb322-81" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb322-82"><a href="#cb322-82" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-119-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="119">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 1; 0 0 … 0 0; … ; 0 1 … 1 1; 0 1 … 1 2], [14109120, 14109196, 14109223, 14113908, 14117707, 14117822, 14121642, 14124306, 14127880, 14135008  …  15108025, 15108049, 15108085, 15116504, 15127266, 15129890, 15134320, 15134324, 15177202, 15177253], <span class="ansi-bold">88×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   4 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
   5 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
   6 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   9 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir     ⋯
  10 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
  11 │ GW_Lane5_TL12              GW_Lane5_TL12              TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  79 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  80 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
  81 │ GW_Armando_plate1_JG08G01  GW_Armando_plate1_JG08G01  ST        plumb
  82 │ GW_Armando_plate1_JG17G01  GW_Armando_plate1_JG17G01  ST        plumb_v
  83 │ GW_Armando_plate2_JF08G04  GW_Armando_plate2_JF08G04  ST        plumb
  84 │ GW_Armando_plate1_JF20G01  GW_Armando_plate1_JF20G01  MB        plumb   ⋯
  85 │ GW_Armando_plate2_JF13G02  GW_Armando_plate2_JF13G02  ST        plumb
  86 │ GW_Armando_plate2_JF12G05  GW_Armando_plate2_JF12G05  ST        plumb
  87 │ GW_Armando_plate2_JF15G01  GW_Armando_plate2_JF15G01  DV        plumb
  88 │ GW_Armando_plate2_JG12G02  GW_Armando_plate2_JG12G02  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 67 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-5" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-5">show GBI plot according to original groups and plot order</h3>
<div id="14cd279f" class="cell" data-execution_count="119">
<div class="sourceCode cell-code" id="cb325"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb325-1"><a href="#cb325-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb325-2"><a href="#cb325-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb325-3"><a href="#cb325-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb325-4"><a href="#cb325-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb325-5"><a href="#cb325-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb325-6"><a href="#cb325-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb325-7"><a href="#cb325-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb325-8"><a href="#cb325-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb325-9"><a href="#cb325-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb325-10"><a href="#cb325-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb325-11"><a href="#cb325-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-120-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="120">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 2 2; 0 2 … 2 2], [14109120, 14109196, 14109223, 14113908, 14117707, 14117822, 14121642, 14124306, 14127880, 14135008  …  15108025, 15108049, 15108085, 15116504, 15127266, 15129890, 15134320, 15134324, 15177202, 15177253], <span class="ansi-bold">88×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   7 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
   8 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
   9 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
  10 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  11 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  79 │ GW_Armando_plate1_JG17G01  GW_Armando_plate1_JG17G01  ST        plumb_v
  80 │ GW_Armando_plate2_JF08G04  GW_Armando_plate2_JF08G04  ST        plumb   ⋯
  81 │ GW_Armando_plate2_JF13G02  GW_Armando_plate2_JF13G02  ST        plumb
  82 │ GW_Armando_plate2_JF12G05  GW_Armando_plate2_JF12G05  ST        plumb
  83 │ GW_Armando_plate2_JG12G02  GW_Armando_plate2_JG12G02  ST        plumb
  84 │ GW_Armando_plate2_JF15G01  GW_Armando_plate2_JF15G01  DV        plumb   ⋯
  85 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  86 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  87 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  88 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 67 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-5" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-5">show same but with all individuals</h3>
<div id="68aab847" class="cell" data-execution_count="120">
<div class="sourceCode cell-code" id="cb327"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb327-1"><a href="#cb327-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb327-2"><a href="#cb327-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-3"><a href="#cb327-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb327-4"><a href="#cb327-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb327-5"><a href="#cb327-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-6"><a href="#cb327-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb327-7"><a href="#cb327-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb327-8"><a href="#cb327-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb327-9"><a href="#cb327-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-10"><a href="#cb327-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb327-11"><a href="#cb327-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb327-12"><a href="#cb327-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb327-13"><a href="#cb327-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb327-14"><a href="#cb327-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-121-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="121">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 2 … 2 2; 0 2 … 2 2], [14109120, 14109196, 14109223, 14113908, 14117707, 14117822, 14121642, 14124306, 14127880, 14135008  …  15108025, 15108049, 15108085, 15116504, 15127266, 15129890, 15134320, 15134324, 15177202, 15177253], <span class="ansi-bold">257×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 249 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 252 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 253 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 31 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops-3" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops-3">show same but with only vir and plumb pops</h3>
<div id="3ab7d000" class="cell" data-execution_count="121">
<div class="sourceCode cell-code" id="cb329"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb329-1"><a href="#cb329-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb329-2"><a href="#cb329-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-3"><a href="#cb329-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb329-4"><a href="#cb329-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb329-5"><a href="#cb329-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-6"><a href="#cb329-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb329-7"><a href="#cb329-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb329-8"><a href="#cb329-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb329-9"><a href="#cb329-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb329-10"><a href="#cb329-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb329-11"><a href="#cb329-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb329-12"><a href="#cb329-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb329-13"><a href="#cb329-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb329-14"><a href="#cb329-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-15"><a href="#cb329-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb329-16"><a href="#cb329-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb329-17"><a href="#cb329-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-18"><a href="#cb329-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb329-19"><a href="#cb329-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-20"><a href="#cb329-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb329-21"><a href="#cb329-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-22"><a href="#cb329-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb329-23"><a href="#cb329-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb329-24"><a href="#cb329-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-122-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="122">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [14109196, 14121642, 14127880, 14152299, 14191715, 14212308, 14234576, 14246931, 14261718, 14282920  …  15059893, 15061446, 15061449, 15084470, 15089863, 15108049, 15127266, 15134320, 15177202, 15177253], <span class="ansi-bold">100×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-6" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-6">Show just the west area (without nitidus)</h3>
<div id="b83a005e" class="cell" data-execution_count="122">
<div class="sourceCode cell-code" id="cb331"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb331-1"><a href="#cb331-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb331-2"><a href="#cb331-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_lud"</span>,</span>
<span id="cb331-3"><a href="#cb331-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud"</span>,</span>
<span id="cb331-4"><a href="#cb331-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"lud_troch"</span>,</span>
<span id="cb331-5"><a href="#cb331-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>]</span>
<span id="cb331-6"><a href="#cb331-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-7"><a href="#cb331-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb331-8"><a href="#cb331-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"seagreen"</span>,</span>
<span id="cb331-9"><a href="#cb331-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb331-10"><a href="#cb331-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green2"</span>,</span>
<span id="cb331-11"><a href="#cb331-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb331-12"><a href="#cb331-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-13"><a href="#cb331-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb331-14"><a href="#cb331-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb331-15"><a href="#cb331-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb331-16"><a href="#cb331-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb331-17"><a href="#cb331-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb331-18"><a href="#cb331-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb331-19"><a href="#cb331-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb331-20"><a href="#cb331-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb331-21"><a href="#cb331-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb331-22"><a href="#cb331-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb331-23"><a href="#cb331-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb331-24"><a href="#cb331-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-25"><a href="#cb331-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsWest))</span>
<span id="cb331-26"><a href="#cb331-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-27"><a href="#cb331-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb331-28"><a href="#cb331-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb331-29"><a href="#cb331-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb331-30"><a href="#cb331-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-31"><a href="#cb331-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb331-32"><a href="#cb331-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb331-33"><a href="#cb331-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb331-34"><a href="#cb331-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb331-35"><a href="#cb331-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-123-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="123">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; -1 2 … 2 2; 0 2 … 2 2], [14109120, 14109196, 14109223, 14117707, 14135008, 14152299, 14206380, 14234576, 14234638, 14246931  …  15041158, 15053672, 15059893, 15061446, 15061449, 15090277, 15108025, 15134324, 15177202, 15177253], <span class="ansi-bold">177×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 168 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 169 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L ⋯
 170 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
 171 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 172 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 173 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L ⋯
 174 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
 175 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
 176 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L
 177 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 31 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-7" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-7">Show just the east area</h3>
<div id="fb6ffcfb" class="cell" data-execution_count="123">
<div class="sourceCode cell-code" id="cb334"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb334-1"><a href="#cb334-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"obs"</span>,</span>
<span id="cb334-2"><a href="#cb334-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>,</span>
<span id="cb334-3"><a href="#cb334-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumbHet"</span>]</span>
<span id="cb334-4"><a href="#cb334-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-5"><a href="#cb334-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"orange"</span>,</span>
<span id="cb334-6"><a href="#cb334-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb334-7"><a href="#cb334-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb334-8"><a href="#cb334-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-9"><a href="#cb334-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb334-10"><a href="#cb334-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb334-11"><a href="#cb334-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb334-12"><a href="#cb334-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb334-13"><a href="#cb334-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb334-14"><a href="#cb334-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb334-15"><a href="#cb334-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb334-16"><a href="#cb334-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb334-17"><a href="#cb334-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb334-18"><a href="#cb334-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb334-19"><a href="#cb334-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb334-20"><a href="#cb334-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-21"><a href="#cb334-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb334-22"><a href="#cb334-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-23"><a href="#cb334-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb334-24"><a href="#cb334-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb334-25"><a href="#cb334-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb334-26"><a href="#cb334-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-27"><a href="#cb334-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb334-28"><a href="#cb334-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb334-29"><a href="#cb334-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb334-30"><a href="#cb334-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb334-31"><a href="#cb334-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-124-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="124">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … -1 2], [14121642, 14127880, 14191715, 14206380, 14212308, 14234638, 14261718, 14318327, 14318334, 14327641  …  14985992, 15008548, 15029658, 15034885, 15041158, 15084470, 15089863, 15108049, 15127266, 15134320], <span class="ansi-bold">75×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Lane5_EM1               GW_Lane5_EM1               EM        troch_E ⋯
   2 │ GW_Lane5_DA3               GW_Lane5_DA3               XN        obs
   3 │ GW_Lane5_DA7               GW_Lane5_DA7               XN        obs
   4 │ GW_Lane5_DA4               GW_Lane5_DA4               XN        obs
   5 │ GW_Lane5_DA2               GW_Lane5_DA2               XN        obs     ⋯
   6 │ GW_Armando_plate1_RF20G01  GW_Armando_plate1_RF20G01  BJ        obs_plu
   7 │ GW_Armando_plate1_RF29G02  GW_Armando_plate1_RF29G02  BJ        obs_plu
   8 │ GW_Armando_plate2_RF29G01  GW_Armando_plate2_RF29G01  BJ        obs_plu
   9 │ GW_Lane5_BK3               GW_Lane5_BK3               BK        plumb   ⋯
  10 │ GW_Lane5_BK2               GW_Lane5_BK2               BK        plumb
  11 │ GW_Lane5_AN1               GW_Lane5_AN1               AN        plumb
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  66 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  67 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  68 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  69 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  70 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  71 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  72 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  73 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  74 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
  75 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 54 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-7" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-7">Show just the northern area</h3>
<div id="9131ba3a" class="cell" data-execution_count="124">
<div class="sourceCode cell-code" id="cb337"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb337-1"><a href="#cb337-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb337-2"><a href="#cb337-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb337-3"><a href="#cb337-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>,</span>
<span id="cb337-4"><a href="#cb337-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumbHet"</span>]</span>
<span id="cb337-5"><a href="#cb337-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-6"><a href="#cb337-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb337-7"><a href="#cb337-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb337-8"><a href="#cb337-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb337-9"><a href="#cb337-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb337-10"><a href="#cb337-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-11"><a href="#cb337-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb337-12"><a href="#cb337-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb337-13"><a href="#cb337-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb337-14"><a href="#cb337-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb337-15"><a href="#cb337-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb337-16"><a href="#cb337-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb337-17"><a href="#cb337-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb337-18"><a href="#cb337-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb337-19"><a href="#cb337-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb337-20"><a href="#cb337-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb337-21"><a href="#cb337-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb337-22"><a href="#cb337-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-23"><a href="#cb337-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb337-24"><a href="#cb337-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-25"><a href="#cb337-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb337-26"><a href="#cb337-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb337-27"><a href="#cb337-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb337-28"><a href="#cb337-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb337-29"><a href="#cb337-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb337-30"><a href="#cb337-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb337-31"><a href="#cb337-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb337-32"><a href="#cb337-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb337-33"><a href="#cb337-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-125-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="125">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [14109196, 14121642, 14127880, 14152299, 14191715, 14212308, 14234576, 14246931, 14261718, 14282920  …  15059893, 15061446, 15061449, 15084470, 15089863, 15108049, 15127266, 15134320, 15177202, 15177253], <span class="ansi-bold">111×34 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 102 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 103 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 104 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 105 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 106 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 107 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 108 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 109 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 110 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 111 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  31 columns and 90 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw13-with-low-hetvar" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw13-with-low-hetvar">Do a PCA based on a same-size region elsewhere on gw13 (with low HetVar):</h3>
<div id="0230d4e1" class="cell" data-execution_count="125">
<div class="sourceCode cell-code" id="cb340"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb340-1"><a href="#cb340-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get length of region</span></span>
<span id="cb340-2"><a href="#cb340-2" aria-hidden="true" tabindex="-1"></a>lengthHighHetVarRegion <span class="op">=</span> positionMax <span class="op">-</span> positionMin</span>
<span id="cb340-3"><a href="#cb340-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-4"><a href="#cb340-4" aria-hidden="true" tabindex="-1"></a>leftLocus <span class="op">=</span> <span class="fl">1_000_000</span> <span class="co"># start at 1 Mb from left side</span></span>
<span id="cb340-5"><a href="#cb340-5" aria-hidden="true" tabindex="-1"></a>rightLocus <span class="op">=</span> leftLocus <span class="op">+</span> lengthHighHetVarRegion</span>
<span id="cb340-6"><a href="#cb340-6" aria-hidden="true" tabindex="-1"></a>regionText_lowHetVarRegion <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,leftLocus,<span class="st">" to "</span>,rightLocus)</span>
<span id="cb340-7"><a href="#cb340-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-8"><a href="#cb340-8" aria-hidden="true" tabindex="-1"></a>lociSelection <span class="op">=</span> (leftLocus <span class="op">.&lt;=</span> pos_region.position <span class="op">.&lt;=</span> rightLocus)</span>
<span id="cb340-9"><a href="#cb340-9" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion <span class="op">=</span> genotypes_region[<span class="op">:</span>, lociSelection]</span>
<span id="cb340-10"><a href="#cb340-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-11"><a href="#cb340-11" aria-hidden="true" tabindex="-1"></a><span class="co"># impute missing genotypes:</span></span>
<span id="cb340-12"><a href="#cb340-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-13"><a href="#cb340-13" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genotypes_lowHetVarRegion))</span>
<span id="cb340-14"><a href="#cb340-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-15"><a href="#cb340-15" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb340-16"><a href="#cb340-16" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb340-17"><a href="#cb340-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-18"><a href="#cb340-18" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genotypes_lowHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb340-19"><a href="#cb340-19" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb340-20"><a href="#cb340-20" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText_lowHetVarRegion,</span>
<span id="cb340-21"><a href="#cb340-21" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb340-22"><a href="#cb340-22" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb340-23"><a href="#cb340-23" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb340-24"><a href="#cb340-24" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb340-25"><a href="#cb340-25" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb340-26"><a href="#cb340-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb340-27"><a href="#cb340-27" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb340-28"><a href="#cb340-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb340-29"><a href="#cb340-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"FigureS2B_gw13_nonHLBRarbitrary_from_Julia.png"</span>, PCAmodel.PCAfig, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb340-30"><a href="#cb340-30" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-126-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="tried-chr-14-but-not-very-clearseems-recombination-backcrosses-in-ludlowi" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-14-but-not-very-clearseems-recombination-backcrosses-in-ludlowi">Tried chr 14 but not very clear–seems recombination backcrosses in ludlowi</h2>
</section>
<section id="same-for-chr-17" class="level2">
<h2 class="anchored" data-anchor-id="same-for-chr-17">Same for chr 17</h2>
<section id="shows-a-pattern-seen-in-chr-2-and-others-where-ludlowi-samples-have-some-plumb-haplotypes" class="level3">
<h3 class="anchored" data-anchor-id="shows-a-pattern-seen-in-chr-2-and-others-where-ludlowi-samples-have-some-plumb-haplotypes">Shows a pattern seen in chr 2 (and others) where ludlowi samples have some plumb haplotypes</h3>
<div id="2fe6512b" class="cell" data-execution_count="126">
<div class="sourceCode cell-code" id="cb342"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb342-1"><a href="#cb342-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb342-2"><a href="#cb342-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw17"</span></span>
<span id="cb342-3"><a href="#cb342-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-4"><a href="#cb342-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb342-5"><a href="#cb342-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb342-6"><a href="#cb342-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb342-7"><a href="#cb342-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb342-8"><a href="#cb342-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb342-9"><a href="#cb342-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb342-10"><a href="#cb342-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb342-11"><a href="#cb342-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-12"><a href="#cb342-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb342-13"><a href="#cb342-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb342-14"><a href="#cb342-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-15"><a href="#cb342-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb342-16"><a href="#cb342-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb342-17"><a href="#cb342-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb342-18"><a href="#cb342-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb342-19"><a href="#cb342-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-20"><a href="#cb342-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb342-21"><a href="#cb342-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-22"><a href="#cb342-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb342-23"><a href="#cb342-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb342-24"><a href="#cb342-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-25"><a href="#cb342-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb342-26"><a href="#cb342-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-27"><a href="#cb342-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb342-28"><a href="#cb342-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-29"><a href="#cb342-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb342-30"><a href="#cb342-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb342-31"><a href="#cb342-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-32"><a href="#cb342-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb342-33"><a href="#cb342-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb342-34"><a href="#cb342-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb342-35"><a href="#cb342-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb342-36"><a href="#cb342-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb342-37"><a href="#cb342-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb342-38"><a href="#cb342-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb342-39"><a href="#cb342-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb342-40"><a href="#cb342-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-41"><a href="#cb342-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb342-42"><a href="#cb342-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-43"><a href="#cb342-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb342-44"><a href="#cb342-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb342-45"><a href="#cb342-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb342-46"><a href="#cb342-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb342-47"><a href="#cb342-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb342-48"><a href="#cb342-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb342-49"><a href="#cb342-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb342-50"><a href="#cb342-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb342-51"><a href="#cb342-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb342-52"><a href="#cb342-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb342-53"><a href="#cb342-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb342-54"><a href="#cb342-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb342-55"><a href="#cb342-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-56"><a href="#cb342-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb342-57"><a href="#cb342-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb342-58"><a href="#cb342-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-59"><a href="#cb342-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb342-60"><a href="#cb342-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.4</span>) </span>
<span id="cb342-61"><a href="#cb342-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-62"><a href="#cb342-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb342-63"><a href="#cb342-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb342-64"><a href="#cb342-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb342-65"><a href="#cb342-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb342-66"><a href="#cb342-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb342-67"><a href="#cb342-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb342-68"><a href="#cb342-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb342-69"><a href="#cb342-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb342-70"><a href="#cb342-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb342-71"><a href="#cb342-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb342-72"><a href="#cb342-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb342-73"><a href="#cb342-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb342-74"><a href="#cb342-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb342-75"><a href="#cb342-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-127-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-127-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="127">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="save-the-individual-colors-in-the-metadata-9" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-9">Save the individual colors in the metadata</h3>
<div id="41a38710" class="cell" data-execution_count="127">
<div class="sourceCode cell-code" id="cb347"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb347-1"><a href="#cb347-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb347-2"><a href="#cb347-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb347-3"><a href="#cb347-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb347-4"><a href="#cb347-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb347-5"><a href="#cb347-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb347-6"><a href="#cb347-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="128">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-9" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-9">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="623f2463" class="cell" data-execution_count="128">
<div class="sourceCode cell-code" id="cb349"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb349-1"><a href="#cb349-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb349-2"><a href="#cb349-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb349-3"><a href="#cb349-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb349-4"><a href="#cb349-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb349-5"><a href="#cb349-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb349-6"><a href="#cb349-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb349-7"><a href="#cb349-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb349-8"><a href="#cb349-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb349-9"><a href="#cb349-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb349-10"><a href="#cb349-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb349-11"><a href="#cb349-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb349-12"><a href="#cb349-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-129-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="129">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-9" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-9">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="0071514b" class="cell" data-execution_count="129">
<div class="sourceCode cell-code" id="cb352"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb352-1"><a href="#cb352-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb352-2"><a href="#cb352-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb352-3"><a href="#cb352-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb352-4"><a href="#cb352-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb352-5"><a href="#cb352-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb352-6"><a href="#cb352-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb352-7"><a href="#cb352-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb352-8"><a href="#cb352-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb352-9"><a href="#cb352-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb352-10"><a href="#cb352-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb352-11"><a href="#cb352-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb352-12"><a href="#cb352-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-130-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="130">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<div id="3b9906d8" class="cell" data-execution_count="130">
<div class="sourceCode cell-code" id="cb355"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb355-1"><a href="#cb355-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb355-2"><a href="#cb355-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb355-3"><a href="#cb355-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb355-4"><a href="#cb355-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb355-5"><a href="#cb355-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb355-6"><a href="#cb355-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-7"><a href="#cb355-7" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb355-8"><a href="#cb355-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb355-9"><a href="#cb355-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb355-10"><a href="#cb355-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb355-11"><a href="#cb355-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb355-12"><a href="#cb355-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-13"><a href="#cb355-13" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">6.5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-14"><a href="#cb355-14" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb355-15"><a href="#cb355-15" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb355-16"><a href="#cb355-16" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">6</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4.5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-17"><a href="#cb355-17" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-18"><a href="#cb355-18" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb355-19"><a href="#cb355-19" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">4.5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-20"><a href="#cb355-20" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb355-21"><a href="#cb355-21" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb355-22"><a href="#cb355-22" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-23"><a href="#cb355-23" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb355-24"><a href="#cb355-24" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb355-25"><a href="#cb355-25" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb355-26"><a href="#cb355-26" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb355-27"><a href="#cb355-27" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb355-28"><a href="#cb355-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-29"><a href="#cb355-29" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb355-30"><a href="#cb355-30" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb355-31"><a href="#cb355-31" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb355-32"><a href="#cb355-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb355-33"><a href="#cb355-33" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb355-34"><a href="#cb355-34" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb355-35"><a href="#cb355-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-36"><a href="#cb355-36" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obs plumb]</span>
<span id="cb355-37"><a href="#cb355-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-38"><a href="#cb355-38" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb355-39"><a href="#cb355-39" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb355-40"><a href="#cb355-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-41"><a href="#cb355-41" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb355-42"><a href="#cb355-42" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb355-43"><a href="#cb355-43" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb355-44"><a href="#cb355-44" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb355-45"><a href="#cb355-45" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb355-46"><a href="#cb355-46" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb355-47"><a href="#cb355-47" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb355-48"><a href="#cb355-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-49"><a href="#cb355-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb355-50"><a href="#cb355-50" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb355-51"><a href="#cb355-51" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb355-52"><a href="#cb355-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-53"><a href="#cb355-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb355-54"><a href="#cb355-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-55"><a href="#cb355-55" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb355-56"><a href="#cb355-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-57"><a href="#cb355-57" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb355-58"><a href="#cb355-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-59"><a href="#cb355-59" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb355-60"><a href="#cb355-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-61"><a href="#cb355-61" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb355-62"><a href="#cb355-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-63"><a href="#cb355-63" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb355-64"><a href="#cb355-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-65"><a href="#cb355-65" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb355-66"><a href="#cb355-66" aria-hidden="true" tabindex="-1"></a><span class="co">#= 5×2 DataFrame</span></span>
<span id="cb355-67"><a href="#cb355-67" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb355-68"><a href="#cb355-68" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb355-69"><a href="#cb355-69" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb355-70"><a href="#cb355-70" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud   0.0116354</span></span>
<span id="cb355-71"><a href="#cb355-71" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.0010142</span></span>
<span id="cb355-72"><a href="#cb355-72" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch    0.00706002</span></span>
<span id="cb355-73"><a href="#cb355-73" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obs      0.0162496</span></span>
<span id="cb355-74"><a href="#cb355-74" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ plumb    0.00402182 =#</span></span>
<span id="cb355-75"><a href="#cb355-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-76"><a href="#cb355-76" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb355-77"><a href="#cb355-77" aria-hidden="true" tabindex="-1"></a><span class="co">#= 10×2 DataFrame</span></span>
<span id="cb355-78"><a href="#cb355-78" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb355-79"><a href="#cb355-79" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb355-80"><a href="#cb355-80" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb355-81"><a href="#cb355-81" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit    0.021495</span></span>
<span id="cb355-82"><a href="#cb355-82" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch  0.0329751</span></span>
<span id="cb355-83"><a href="#cb355-83" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obs    0.0328026</span></span>
<span id="cb355-84"><a href="#cb355-84" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ virLud_plumb  0.0354124</span></span>
<span id="cb355-85"><a href="#cb355-85" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_troch     0.0339931</span></span>
<span id="cb355-86"><a href="#cb355-86" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_obs       0.0337221</span></span>
<span id="cb355-87"><a href="#cb355-87" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_plumb     0.036556</span></span>
<span id="cb355-88"><a href="#cb355-88" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ troch_obs     0.0201926</span></span>
<span id="cb355-89"><a href="#cb355-89" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ troch_plumb   0.0240156</span></span>
<span id="cb355-90"><a href="#cb355-90" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ obs_plumb     0.0195389 =#</span></span>
<span id="cb355-91"><a href="#cb355-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-92"><a href="#cb355-92" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb355-93"><a href="#cb355-93" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb355-94"><a href="#cb355-94" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb355-95"><a href="#cb355-95" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb355-96"><a href="#cb355-96" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb355-97"><a href="#cb355-97" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb355-98"><a href="#cb355-98" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb355-99"><a href="#cb355-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-100"><a href="#cb355-100" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb355-101"><a href="#cb355-101" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb355-102"><a href="#cb355-102" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb355-103"><a href="#cb355-103" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb355-104"><a href="#cb355-104" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb355-105"><a href="#cb355-105" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb355-106"><a href="#cb355-106" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb355-107"><a href="#cb355-107" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb355-108"><a href="#cb355-108" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb355-109"><a href="#cb355-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-110"><a href="#cb355-110" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb355-111"><a href="#cb355-111" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb355-112"><a href="#cb355-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-113"><a href="#cb355-113" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb355-114"><a href="#cb355-114" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb355-115"><a href="#cb355-115" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb355-116"><a href="#cb355-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-117"><a href="#cb355-117" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb355-118"><a href="#cb355-118" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb355-119"><a href="#cb355-119" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb355-120"><a href="#cb355-120" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [73 2 62 3 64] and the sum of those is 204
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-131-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="131">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 0 2; 2 2 … 0 2], [11771696, 11771805, 11824912, 11824930, 11840181, 11843613, 11854001, 11854002, 11862254, 11865895  …  12545202, 12545274, 12550749, 12556180, 12562225, 12562314, 12582178, 12584065, 12586755, 12623539], <span class="ansi-bold">50×36 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_TTGW53      GW_Armando_plate1_TTGW53      NG        l
   3 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l
   4 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l
   5 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
   6 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
   7 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   8 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
   9 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l ⋯
  10 │ GW_Armando_plate2_TTGW56      GW_Armando_plate2_TTGW56      NG        l
  11 │ GW_Armando_plate2_TTGW-15-03  GW_Armando_plate2_TTGW-15-03  SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  41 │ GW_Armando_plate1_JF12G02     GW_Armando_plate1_JF12G02     ST        p
  42 │ GW_Armando_plate1_JF13G01     GW_Armando_plate1_JF13G01     ST        p ⋯
  43 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p
  44 │ GW_Armando_plate1_JF22G01     GW_Armando_plate1_JF22G01     MB        p
  45 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  46 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p ⋯
  47 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  48 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p
  49 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  50 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p ⋯
<span class="ansi-cyan-fg">                                                  33 columns and 29 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-9" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-9">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="7575bb56" class="cell" data-execution_count="131">
<div class="sourceCode cell-code" id="cb358"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb358-1"><a href="#cb358-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb358-2"><a href="#cb358-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"nit"</span>,</span>
<span id="cb358-3"><a href="#cb358-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_troch"</span>,</span>
<span id="cb358-4"><a href="#cb358-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb358-5"><a href="#cb358-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_obs"</span>,</span>
<span id="cb358-6"><a href="#cb358-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obs"</span>,</span>
<span id="cb358-7"><a href="#cb358-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch_plumb"</span>,</span>
<span id="cb358-8"><a href="#cb358-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"plumb"</span>,</span>
<span id="cb358-9"><a href="#cb358-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"vir_plumb"</span>]</span>
<span id="cb358-10"><a href="#cb358-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-11"><a href="#cb358-11" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb358-12"><a href="#cb358-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"grey"</span>,</span>
<span id="cb358-13"><a href="#cb358-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb358-14"><a href="#cb358-14" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb358-15"><a href="#cb358-15" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"olive"</span>,</span>
<span id="cb358-16"><a href="#cb358-16" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"orange"</span>,</span>
<span id="cb358-17"><a href="#cb358-17" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"coral"</span>,</span>
<span id="cb358-18"><a href="#cb358-18" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"red"</span>,</span>
<span id="cb358-19"><a href="#cb358-19" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"purple"</span>]</span>
<span id="cb358-20"><a href="#cb358-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-21"><a href="#cb358-21" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb358-22"><a href="#cb358-22" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">0.7</span>) <span class="op">.&amp;</span></span>
<span id="cb358-23"><a href="#cb358-23" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb358-24"><a href="#cb358-24" aria-hidden="true" tabindex="-1"></a>virLud_obs <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb358-25"><a href="#cb358-25" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb358-26"><a href="#cb358-26" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb358-27"><a href="#cb358-27" aria-hidden="true" tabindex="-1"></a>troch_plumb <span class="op">=</span> (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5.5</span>) <span class="op">.&amp;</span></span>
<span id="cb358-28"><a href="#cb358-28" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb358-29"><a href="#cb358-29" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb358-30"><a href="#cb358-30" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb358-31"><a href="#cb358-31" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb358-32"><a href="#cb358-32" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb358-33"><a href="#cb358-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-34"><a href="#cb358-34" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit virLud_troch troch virLud_obs obs troch_plumb plumb vir_plumb]</span>
<span id="cb358-35"><a href="#cb358-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-36"><a href="#cb358-36" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb358-37"><a href="#cb358-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-38"><a href="#cb358-38" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb358-39"><a href="#cb358-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb358-40"><a href="#cb358-40" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb358-41"><a href="#cb358-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb358-42"><a href="#cb358-42" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb358-43"><a href="#cb358-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-44"><a href="#cb358-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-45"><a href="#cb358-45" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb358-46"><a href="#cb358-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-47"><a href="#cb358-47" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb358-48"><a href="#cb358-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-49"><a href="#cb358-49" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb358-50"><a href="#cb358-50" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb358-51"><a href="#cb358-51" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb358-52"><a href="#cb358-52" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb358-53"><a href="#cb358-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-54"><a href="#cb358-54" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb358-55"><a href="#cb358-55" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb358-56"><a href="#cb358-56" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb358-57"><a href="#cb358-57" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb358-58"><a href="#cb358-58" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb358-59"><a href="#cb358-59" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb358-60"><a href="#cb358-60" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb358-61"><a href="#cb358-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-62"><a href="#cb358-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb358-63"><a href="#cb358-63" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb358-64"><a href="#cb358-64" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb358-65"><a href="#cb358-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-66"><a href="#cb358-66" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb358-67"><a href="#cb358-67" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb358-68"><a href="#cb358-68" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb358-69"><a href="#cb358-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-70"><a href="#cb358-70" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb358-71"><a href="#cb358-71" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb358-72"><a href="#cb358-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-73"><a href="#cb358-73" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb358-74"><a href="#cb358-74" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb358-75"><a href="#cb358-75" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb358-76"><a href="#cb358-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb358-77"><a href="#cb358-77" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb358-78"><a href="#cb358-78" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb358-79"><a href="#cb358-79" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb358-80"><a href="#cb358-80" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb358-81"><a href="#cb358-81" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-132-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="132">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 0 1; 1 2 … 0 1], [11771696, 11771805, 11824912, 11824930, 11840181, 11843613, 11854001, 11854002, 11862254, 11865895  …  12545202, 12545274, 12550749, 12556180, 12562225, 12562314, 12582178, 12584065, 12586755, 12623539], <span class="ansi-bold">257×36 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_TTGW53      GW_Armando_plate1_TTGW53      NG        l
   3 │ GW_Armando_plate1_TTGW58      GW_Armando_plate1_TTGW58      NG        l
   4 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l
   5 │ GW_Armando_plate2_JE31G01     GW_Armando_plate2_JE31G01     VB_vi     v ⋯
   6 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
   7 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   8 │ GW_Armando_plate2_TTGW16      GW_Armando_plate2_TTGW16      TH        l
   9 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l ⋯
  10 │ GW_Armando_plate2_TTGW56      GW_Armando_plate2_TTGW56      NG        l
  11 │ GW_Armando_plate2_TTGW-15-03  GW_Armando_plate2_TTGW-15-03  SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 248 │ GW_Armando_plate2_JF10G02     GW_Armando_plate2_JF10G02     ST        p
 249 │ GW_Armando_plate2_JF14G01     GW_Armando_plate2_JF14G01     DV        p ⋯
 250 │ GW_Armando_plate2_JG12G02     GW_Armando_plate2_JG12G02     ST        p
 251 │ GW_Armando_plate2_TTGW24      GW_Armando_plate2_TTGW24      SR        l
 252 │ GW_Armando_plate2_TTGW25      GW_Armando_plate2_TTGW25      SR        l
 253 │ GW_Armando_plate2_TTGW-15-01  GW_Armando_plate2_TTGW-15-01  SR        l ⋯
 254 │ GW_Armando_plate2_JF08G04     GW_Armando_plate2_JF08G04     ST        p
 255 │ GW_Lane5_UY6                  GW_Lane5_UY6                  UY        p
 256 │ GW_Armando_plate2_TTGW-15-10  GW_Armando_plate2_TTGW-15-10  SR        l
 257 │ GW_Lane5_UY3                  GW_Lane5_UY3                  UY        p ⋯
<span class="ansi-cyan-fg">                                                 33 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="chr-17-above-shows-interesting-pattern-with-plumb-haplotype-more-widespreadfound-in-ludlowi-and-the-obscuratus-pattern-complex-with-recombination.-i-checked-chr-17-carefully-in-the-summary-plot-and-it-looks-good." class="level3">
<h3 class="anchored" data-anchor-id="chr-17-above-shows-interesting-pattern-with-plumb-haplotype-more-widespreadfound-in-ludlowi-and-the-obscuratus-pattern-complex-with-recombination.-i-checked-chr-17-carefully-in-the-summary-plot-and-it-looks-good.">Chr 17 above shows interesting pattern with plumb haplotype more widespread–found in ludlowi, and the obscuratus pattern complex with recombination. I checked chr 17 carefully in the summary plot, and it looks good.</h3>
</section>
</section>
<section id="same-for-chr-19" class="level2">
<h2 class="anchored" data-anchor-id="same-for-chr-19">Same for chr 19</h2>
<div id="bd73088f" class="cell" data-execution_count="132">
<div class="sourceCode cell-code" id="cb361"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb361-1"><a href="#cb361-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb361-2"><a href="#cb361-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw19"</span></span>
<span id="cb361-3"><a href="#cb361-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-4"><a href="#cb361-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb361-5"><a href="#cb361-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb361-6"><a href="#cb361-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb361-7"><a href="#cb361-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb361-8"><a href="#cb361-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb361-9"><a href="#cb361-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb361-10"><a href="#cb361-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb361-11"><a href="#cb361-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-12"><a href="#cb361-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb361-13"><a href="#cb361-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb361-14"><a href="#cb361-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-15"><a href="#cb361-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb361-16"><a href="#cb361-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb361-17"><a href="#cb361-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb361-18"><a href="#cb361-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb361-19"><a href="#cb361-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-20"><a href="#cb361-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb361-21"><a href="#cb361-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-22"><a href="#cb361-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb361-23"><a href="#cb361-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb361-24"><a href="#cb361-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-25"><a href="#cb361-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb361-26"><a href="#cb361-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-27"><a href="#cb361-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb361-28"><a href="#cb361-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-29"><a href="#cb361-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb361-30"><a href="#cb361-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb361-31"><a href="#cb361-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-32"><a href="#cb361-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb361-33"><a href="#cb361-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb361-34"><a href="#cb361-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb361-35"><a href="#cb361-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb361-36"><a href="#cb361-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb361-37"><a href="#cb361-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb361-38"><a href="#cb361-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb361-39"><a href="#cb361-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb361-40"><a href="#cb361-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-41"><a href="#cb361-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb361-42"><a href="#cb361-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-43"><a href="#cb361-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb361-44"><a href="#cb361-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb361-45"><a href="#cb361-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb361-46"><a href="#cb361-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb361-47"><a href="#cb361-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb361-48"><a href="#cb361-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb361-49"><a href="#cb361-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb361-50"><a href="#cb361-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb361-51"><a href="#cb361-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb361-52"><a href="#cb361-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb361-53"><a href="#cb361-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb361-54"><a href="#cb361-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb361-55"><a href="#cb361-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-56"><a href="#cb361-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb361-57"><a href="#cb361-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb361-58"><a href="#cb361-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-59"><a href="#cb361-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb361-60"><a href="#cb361-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.5</span>) </span>
<span id="cb361-61"><a href="#cb361-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-62"><a href="#cb361-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb361-63"><a href="#cb361-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb361-64"><a href="#cb361-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb361-65"><a href="#cb361-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb361-66"><a href="#cb361-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb361-67"><a href="#cb361-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb361-68"><a href="#cb361-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb361-69"><a href="#cb361-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb361-70"><a href="#cb361-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb361-71"><a href="#cb361-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb361-72"><a href="#cb361-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb361-73"><a href="#cb361-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb361-74"><a href="#cb361-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb361-75"><a href="#cb361-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-133-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-133-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="133">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-10" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-10">Save the individual colors in the metadata</h3>
<div id="9cfcf8f0" class="cell" data-execution_count="133">
<div class="sourceCode cell-code" id="cb366"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb366-1"><a href="#cb366-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb366-2"><a href="#cb366-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb366-3"><a href="#cb366-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb366-4"><a href="#cb366-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb366-5"><a href="#cb366-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb366-6"><a href="#cb366-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="134">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-10" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-10">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="4875396b" class="cell" data-execution_count="134">
<div class="sourceCode cell-code" id="cb368"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb368-1"><a href="#cb368-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb368-2"><a href="#cb368-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb368-3"><a href="#cb368-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb368-4"><a href="#cb368-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb368-5"><a href="#cb368-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb368-6"><a href="#cb368-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb368-7"><a href="#cb368-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb368-8"><a href="#cb368-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb368-9"><a href="#cb368-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb368-10"><a href="#cb368-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb368-11"><a href="#cb368-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb368-12"><a href="#cb368-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-135-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="135">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-10" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-10">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="a22fc3fe" class="cell" data-execution_count="135">
<div class="sourceCode cell-code" id="cb371"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb371-1"><a href="#cb371-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb371-2"><a href="#cb371-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb371-3"><a href="#cb371-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb371-4"><a href="#cb371-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb371-5"><a href="#cb371-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb371-6"><a href="#cb371-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb371-7"><a href="#cb371-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb371-8"><a href="#cb371-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb371-9"><a href="#cb371-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb371-10"><a href="#cb371-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb371-11"><a href="#cb371-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb371-12"><a href="#cb371-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-136-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="136">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-19-high-hetvar-region-there-are-4-clear-homozygous-haplogroups-vir-and-lud-separated-though-on-pc3-but-not-clearly-enough-to-show-in-summary-plot.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-19-high-hetvar-region-there-are-4-clear-homozygous-haplogroups-vir-and-lud-separated-though-on-pc3-but-not-clearly-enough-to-show-in-summary-plot.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 19 high HetVar region, there are 4 clear homozygous haplogroups (vir and lud separated though on PC3, but not clearly enough to show in summary plot). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="302c82bb" class="cell" data-execution_count="136">
<div class="sourceCode cell-code" id="cb374"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb374-1"><a href="#cb374-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb374-2"><a href="#cb374-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb374-3"><a href="#cb374-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochObs"</span>,</span>
<span id="cb374-4"><a href="#cb374-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb374-5"><a href="#cb374-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-6"><a href="#cb374-6" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb374-7"><a href="#cb374-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb374-8"><a href="#cb374-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb374-9"><a href="#cb374-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb374-10"><a href="#cb374-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-11"><a href="#cb374-11" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb374-12"><a href="#cb374-12" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb374-13"><a href="#cb374-13" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb374-14"><a href="#cb374-14" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb374-15"><a href="#cb374-15" aria-hidden="true" tabindex="-1"></a>        (<span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb374-16"><a href="#cb374-16" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb374-17"><a href="#cb374-17" aria-hidden="true" tabindex="-1"></a>trochObs <span class="op">=</span> (<span class="op">-</span><span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb374-18"><a href="#cb374-18" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3.5</span>) <span class="op">.&amp;</span></span>
<span id="cb374-19"><a href="#cb374-19" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb374-20"><a href="#cb374-20" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span> </span>
<span id="cb374-21"><a href="#cb374-21" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb374-22"><a href="#cb374-22" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb374-23"><a href="#cb374-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-24"><a href="#cb374-24" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb374-25"><a href="#cb374-25" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb374-26"><a href="#cb374-26" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb374-27"><a href="#cb374-27" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[trochObs]</span>
<span id="cb374-28"><a href="#cb374-28" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb374-29"><a href="#cb374-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-30"><a href="#cb374-30" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit trochObs plumb]</span>
<span id="cb374-31"><a href="#cb374-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-32"><a href="#cb374-32" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb374-33"><a href="#cb374-33" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb374-34"><a href="#cb374-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-35"><a href="#cb374-35" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb374-36"><a href="#cb374-36" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb374-37"><a href="#cb374-37" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb374-38"><a href="#cb374-38" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb374-39"><a href="#cb374-39" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb374-40"><a href="#cb374-40" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb374-41"><a href="#cb374-41" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb374-42"><a href="#cb374-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-43"><a href="#cb374-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb374-44"><a href="#cb374-44" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb374-45"><a href="#cb374-45" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb374-46"><a href="#cb374-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-47"><a href="#cb374-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb374-48"><a href="#cb374-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-49"><a href="#cb374-49" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb374-50"><a href="#cb374-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-51"><a href="#cb374-51" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb374-52"><a href="#cb374-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-53"><a href="#cb374-53" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb374-54"><a href="#cb374-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-55"><a href="#cb374-55" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb374-56"><a href="#cb374-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-57"><a href="#cb374-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb374-58"><a href="#cb374-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-59"><a href="#cb374-59" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb374-60"><a href="#cb374-60" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb374-61"><a href="#cb374-61" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi         </span></span>
<span id="cb374-62"><a href="#cb374-62" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64    </span></span>
<span id="cb374-63"><a href="#cb374-63" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼──────────────────────</span></span>
<span id="cb374-64"><a href="#cb374-64" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.0144925</span></span>
<span id="cb374-65"><a href="#cb374-65" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.0052608</span></span>
<span id="cb374-66"><a href="#cb374-66" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ trochObs  0.0150341</span></span>
<span id="cb374-67"><a href="#cb374-67" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ plumb     0.00320386 =#</span></span>
<span id="cb374-68"><a href="#cb374-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-69"><a href="#cb374-69" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb374-70"><a href="#cb374-70" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb374-71"><a href="#cb374-71" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy       </span></span>
<span id="cb374-72"><a href="#cb374-72" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64   </span></span>
<span id="cb374-73"><a href="#cb374-73" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼────────────────────────────</span></span>
<span id="cb374-74"><a href="#cb374-74" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.0291485</span></span>
<span id="cb374-75"><a href="#cb374-75" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_trochObs  0.0330435</span></span>
<span id="cb374-76"><a href="#cb374-76" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_plumb     0.0347335</span></span>
<span id="cb374-77"><a href="#cb374-77" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_trochObs     0.0359384</span></span>
<span id="cb374-78"><a href="#cb374-78" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_plumb        0.0373399</span></span>
<span id="cb374-79"><a href="#cb374-79" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ trochObs_plumb   0.0289202 =#</span></span>
<span id="cb374-80"><a href="#cb374-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-81"><a href="#cb374-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb374-82"><a href="#cb374-82" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb374-83"><a href="#cb374-83" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb374-84"><a href="#cb374-84" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb374-85"><a href="#cb374-85" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb374-86"><a href="#cb374-86" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb374-87"><a href="#cb374-87" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb374-88"><a href="#cb374-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-89"><a href="#cb374-89" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb374-90"><a href="#cb374-90" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb374-91"><a href="#cb374-91" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb374-92"><a href="#cb374-92" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb374-93"><a href="#cb374-93" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb374-94"><a href="#cb374-94" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb374-95"><a href="#cb374-95" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb374-96"><a href="#cb374-96" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb374-97"><a href="#cb374-97" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb374-98"><a href="#cb374-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-99"><a href="#cb374-99" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb374-100"><a href="#cb374-100" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb374-101"><a href="#cb374-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-102"><a href="#cb374-102" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb374-103"><a href="#cb374-103" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb374-104"><a href="#cb374-104" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb374-105"><a href="#cb374-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb374-106"><a href="#cb374-106" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb374-107"><a href="#cb374-107" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb374-108"><a href="#cb374-108" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb374-109"><a href="#cb374-109" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [70 2 67 66] and the sum of those is 205
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-137-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="137">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [63088, 70570, 71423, 74630, 78271, 89983, 89997, 132730, 153219, 167981  …  925752, 925762, 925928, 939829, 945566, 950361, 952095, 975932, 976062, 984894], <span class="ansi-bold">47×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   3 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l
   4 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   5 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l ⋯
   6 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   7 │ GW_Armando_plate2_JF03G01     GW_Armando_plate2_JF03G01     ST_vi     v
   8 │ GW_Armando_plate2_JF03G02     GW_Armando_plate2_JF03G02     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW14      GW_Armando_plate2_TTGW14      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  38 │ GW_Armando_plate1_JF15G03     GW_Armando_plate1_JF15G03     DV        p
  39 │ GW_Armando_plate1_JF23G01     GW_Armando_plate1_JF23G01     VB        p ⋯
  40 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
  41 │ GW_Armando_plate1_JF26G01     GW_Armando_plate1_JF26G01     ST        p
  42 │ GW_Armando_plate1_JF27G01     GW_Armando_plate1_JF27G01     ST        p
  43 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p ⋯
  44 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p
  45 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  46 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  47 │ GW_Armando_plate1_JG12G01     GW_Armando_plate1_JG12G01     ST        p ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 26 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-10" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-10">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="aa5954c0" class="cell" data-execution_count="137">
<div class="sourceCode cell-code" id="cb377"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb377-1"><a href="#cb377-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb377-2"><a href="#cb377-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"virLudHet"</span>,</span>
<span id="cb377-3"><a href="#cb377-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb377-4"><a href="#cb377-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"virLud_trochObs"</span>,</span>
<span id="cb377-5"><a href="#cb377-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochObs"</span>,</span>
<span id="cb377-6"><a href="#cb377-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochObsHet"</span>,</span>
<span id="cb377-7"><a href="#cb377-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochObs_plumb"</span>,</span>
<span id="cb377-8"><a href="#cb377-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>,</span>
<span id="cb377-9"><a href="#cb377-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"vir_plumb"</span>]</span>
<span id="cb377-10"><a href="#cb377-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-11"><a href="#cb377-11" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb377-12"><a href="#cb377-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"blue"</span>,</span>
<span id="cb377-13"><a href="#cb377-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb377-14"><a href="#cb377-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb377-15"><a href="#cb377-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb377-16"><a href="#cb377-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb377-17"><a href="#cb377-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb377-18"><a href="#cb377-18" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb377-19"><a href="#cb377-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">"purple"</span>]</span>
<span id="cb377-20"><a href="#cb377-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-21"><a href="#cb377-21" aria-hidden="true" tabindex="-1"></a>virLudHet <span class="op">=</span> (<span class="op">-</span><span class="fl">8</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb377-22"><a href="#cb377-22" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-23"><a href="#cb377-23" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb377-24"><a href="#cb377-24" aria-hidden="true" tabindex="-1"></a>virLud_trochObs <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb377-25"><a href="#cb377-25" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-26"><a href="#cb377-26" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb377-27"><a href="#cb377-27" aria-hidden="true" tabindex="-1"></a>trochObsHet <span class="op">=</span> (<span class="op">-</span><span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-28"><a href="#cb377-28" aria-hidden="true" tabindex="-1"></a>                (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3.5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-29"><a href="#cb377-29" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb377-30"><a href="#cb377-30" aria-hidden="true" tabindex="-1"></a>trochObs_plumb <span class="op">=</span> (<span class="fl">1.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-31"><a href="#cb377-31" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">3.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb377-32"><a href="#cb377-32" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb377-33"><a href="#cb377-33" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">1</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-34"><a href="#cb377-34" aria-hidden="true" tabindex="-1"></a>                (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb377-35"><a href="#cb377-35" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb377-36"><a href="#cb377-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-37"><a href="#cb377-37" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud virLudHet nit virLud_trochObs trochObs trochObsHet trochObs_plumb plumb vir_plumb]</span>
<span id="cb377-38"><a href="#cb377-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-39"><a href="#cb377-39" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb377-40"><a href="#cb377-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-41"><a href="#cb377-41" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb377-42"><a href="#cb377-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb377-43"><a href="#cb377-43" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb377-44"><a href="#cb377-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb377-45"><a href="#cb377-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb377-46"><a href="#cb377-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-47"><a href="#cb377-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-48"><a href="#cb377-48" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb377-49"><a href="#cb377-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-50"><a href="#cb377-50" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb377-51"><a href="#cb377-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-52"><a href="#cb377-52" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb377-53"><a href="#cb377-53" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb377-54"><a href="#cb377-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb377-55"><a href="#cb377-55" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb377-56"><a href="#cb377-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-57"><a href="#cb377-57" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb377-58"><a href="#cb377-58" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb377-59"><a href="#cb377-59" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb377-60"><a href="#cb377-60" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb377-61"><a href="#cb377-61" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb377-62"><a href="#cb377-62" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb377-63"><a href="#cb377-63" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb377-64"><a href="#cb377-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-65"><a href="#cb377-65" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb377-66"><a href="#cb377-66" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb377-67"><a href="#cb377-67" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb377-68"><a href="#cb377-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-69"><a href="#cb377-69" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb377-70"><a href="#cb377-70" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb377-71"><a href="#cb377-71" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb377-72"><a href="#cb377-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-73"><a href="#cb377-73" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb377-74"><a href="#cb377-74" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb377-75"><a href="#cb377-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-76"><a href="#cb377-76" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb377-77"><a href="#cb377-77" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb377-78"><a href="#cb377-78" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb377-79"><a href="#cb377-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-80"><a href="#cb377-80" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb377-81"><a href="#cb377-81" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb377-82"><a href="#cb377-82" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb377-83"><a href="#cb377-83" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb377-84"><a href="#cb377-84" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-138-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="138">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 1 1; 2 1 … -1 1], [63088, 70570, 71423, 74630, 78271, 89983, 89997, 132730, 153219, 167981  …  925752, 925762, 925928, 939829, 945566, 950361, 952095, 975932, 976062, 984894], <span class="ansi-bold">82×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1         GW_Armando_plate1_AB1         AB        v ⋯
   2 │ GW_Armando_plate1_TL3         GW_Armando_plate1_TL3         TL        v
   3 │ GW_Armando_plate1_TTGW10      GW_Armando_plate1_TTGW10      SU        l
   4 │ GW_Armando_plate1_TTGW22      GW_Armando_plate1_TTGW22      SR        l
   5 │ GW_Armando_plate1_TTGW57      GW_Armando_plate1_TTGW57      NG        l ⋯
   6 │ GW_Armando_plate1_TTGW-15-07  GW_Armando_plate1_TTGW-15-07  SR        l
   7 │ GW_Armando_plate2_JF03G01     GW_Armando_plate2_JF03G01     ST_vi     v
   8 │ GW_Armando_plate2_JF03G02     GW_Armando_plate2_JF03G02     VB_vi     v
   9 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p ⋯
  10 │ GW_Armando_plate2_TTGW12      GW_Armando_plate2_TTGW12      TH        l
  11 │ GW_Armando_plate2_TTGW14      GW_Armando_plate2_TTGW14      TH        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  73 │ GW_Armando_plate1_JF29G01     GW_Armando_plate1_JF29G01     ST        p
  74 │ GW_Armando_plate1_JF29G02     GW_Armando_plate1_JF29G02     ST        p ⋯
  75 │ GW_Armando_plate1_JF29G03     GW_Armando_plate1_JF29G03     ST        p
  76 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
  77 │ GW_Armando_plate1_JG12G01     GW_Armando_plate1_JG12G01     ST        p
  78 │ GW_Armando_plate1_JF07G03     GW_Armando_plate1_JF07G03     ST        p ⋯
  79 │ GW_Armando_plate1_JF10G03     GW_Armando_plate1_JF10G03     ST        p
  80 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p
  81 │ GW_Armando_plate2_JF15G01     GW_Armando_plate2_JF15G01     DV        p
  82 │ GW_Armando_plate1_JF12G02     GW_Armando_plate1_JF12G02     ST        p ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 61 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-6" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-6">show GBI plot according to original groups and plot order</h3>
<div id="c8ded28f" class="cell" data-execution_count="138">
<div class="sourceCode cell-code" id="cb380"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb380-1"><a href="#cb380-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb380-2"><a href="#cb380-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-3"><a href="#cb380-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb380-4"><a href="#cb380-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb380-5"><a href="#cb380-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb380-6"><a href="#cb380-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-7"><a href="#cb380-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb380-8"><a href="#cb380-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb380-9"><a href="#cb380-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb380-10"><a href="#cb380-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb380-11"><a href="#cb380-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-139-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="139">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 0 1 … 1 2], [63088, 70570, 71423, 74630, 78271, 89983, 89997, 132730, 153219, 167981  …  925752, 925762, 925928, 939829, 945566, 950361, 952095, 975932, 976062, 984894], <span class="ansi-bold">82×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   4 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
   5 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
   6 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
   7 │ GW_Lane5_TU1               GW_Lane5_TU1               TU        nit
   8 │ GW_Lane5_TU2               GW_Lane5_TU2               TU        nit
   9 │ GW_Armando_plate1_TTGW72   GW_Armando_plate1_TTGW72   SA        lud_Sat ⋯
  10 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
  11 │ GW_Armando_plate2_TTGW77   GW_Armando_plate2_TTGW77   SA        lud_Sat
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
  73 │ GW_Armando_plate1_JG12G01  GW_Armando_plate1_JG12G01  ST        plumb
  74 │ GW_Armando_plate1_JF07G03  GW_Armando_plate1_JF07G03  ST        plumb   ⋯
  75 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
  76 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  77 │ GW_Armando_plate2_JF15G01  GW_Armando_plate2_JF15G01  DV        plumb
  78 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb   ⋯
  79 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  80 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  81 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  82 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 61 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-6" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-6">show same but with all individuals</h3>
<div id="313fe5dc" class="cell" data-execution_count="139">
<div class="sourceCode cell-code" id="cb382"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb382-1"><a href="#cb382-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb382-2"><a href="#cb382-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb382-3"><a href="#cb382-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb382-4"><a href="#cb382-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb382-5"><a href="#cb382-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb382-6"><a href="#cb382-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb382-7"><a href="#cb382-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb382-8"><a href="#cb382-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb382-9"><a href="#cb382-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb382-10"><a href="#cb382-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb382-11"><a href="#cb382-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb382-12"><a href="#cb382-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb382-13"><a href="#cb382-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb382-14"><a href="#cb382-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-140-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="140">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 0 1 … 1 2], [63088, 70570, 71423, 74630, 78271, 89983, 89997, 132730, 153219, 167981  …  925752, 925762, 925928, 939829, 945566, 950361, 952095, 975932, 976062, 984894], <span class="ansi-bold">257×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 249 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 252 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 253 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops-4" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops-4">show same but with only vir and plumb pops</h3>
<div id="1fc3e92a" class="cell" data-execution_count="140">
<div class="sourceCode cell-code" id="cb384"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb384-1"><a href="#cb384-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb384-2"><a href="#cb384-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-3"><a href="#cb384-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb384-4"><a href="#cb384-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb384-5"><a href="#cb384-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-6"><a href="#cb384-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb384-7"><a href="#cb384-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb384-8"><a href="#cb384-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb384-9"><a href="#cb384-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb384-10"><a href="#cb384-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb384-11"><a href="#cb384-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb384-12"><a href="#cb384-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb384-13"><a href="#cb384-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb384-14"><a href="#cb384-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-15"><a href="#cb384-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb384-16"><a href="#cb384-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb384-17"><a href="#cb384-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-18"><a href="#cb384-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb384-19"><a href="#cb384-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-20"><a href="#cb384-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb384-21"><a href="#cb384-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-22"><a href="#cb384-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb384-23"><a href="#cb384-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb384-24"><a href="#cb384-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-141-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="141">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 0 1 … 1 2], [63088, 70570, 71423, 78271, 89983, 153219, 199863, 252463, 296231, 423565  …  882012, 900073, 925752, 925928, 939829, 945566, 952095, 975932, 976062, 984894], <span class="ansi-bold">100×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-lud-troch-pops-1" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-lud-troch-pops-1">show same but with only vir lud troch pops</h3>
<div id="94f02a57" class="cell" data-execution_count="141">
<div class="sourceCode cell-code" id="cb386"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb386-1"><a href="#cb386-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"trochObs"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb386-2"><a href="#cb386-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-3"><a href="#cb386-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb386-4"><a href="#cb386-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb386-5"><a href="#cb386-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-6"><a href="#cb386-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb386-7"><a href="#cb386-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb386-8"><a href="#cb386-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb386-9"><a href="#cb386-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb386-10"><a href="#cb386-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb386-11"><a href="#cb386-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb386-12"><a href="#cb386-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb386-13"><a href="#cb386-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb386-14"><a href="#cb386-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-15"><a href="#cb386-15" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb386-16"><a href="#cb386-16" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb386-17"><a href="#cb386-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-18"><a href="#cb386-18" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"vir_S"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>, <span class="st">"troch_west"</span>, <span class="st">"troch_LN"</span>]</span>
<span id="cb386-19"><a href="#cb386-19" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>, <span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>]</span>
<span id="cb386-20"><a href="#cb386-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-21"><a href="#cb386-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb386-22"><a href="#cb386-22" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">10</span>, <span class="fu">length</span>(plotGroups))</span>
<span id="cb386-23"><a href="#cb386-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-24"><a href="#cb386-24" aria-hidden="true" tabindex="-1"></a>genosForGBI_limited, indMetadataforGBI_limited <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups,</span>
<span id="cb386-25"><a href="#cb386-25" aria-hidden="true" tabindex="-1"></a>                                            numIndsToPlotWithHets, </span>
<span id="cb386-26"><a href="#cb386-26" aria-hidden="true" tabindex="-1"></a>                                            genosForGBI, metadataForGBI;</span>
<span id="cb386-27"><a href="#cb386-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb386-28"><a href="#cb386-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-29"><a href="#cb386-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb386-30"><a href="#cb386-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI_limited, indMetadataforGBI_limited, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb386-31"><a href="#cb386-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-142-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="142">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [63088, 153219, 167981, 204946, 252487, 296231, 367641, 423565, 431630, 443933  …  678802, 684399, 741883, 767533, 773736, 792896, 900073, 925762, 939829, 952095], <span class="ansi-bold">74×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   7 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
   8 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   9 │ GW_Lane5_AB2               GW_Lane5_AB2               AB        vir     ⋯
  10 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  11 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  65 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L
  66 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L ⋯
  67 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
  68 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
  69 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
  70 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L ⋯
  71 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
  72 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
  73 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
  74 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 53 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-troch-plumb-pops-1" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-troch-plumb-pops-1">show same but with only troch plumb pops</h3>
<div id="f449e433" class="cell" data-execution_count="142">
<div class="sourceCode cell-code" id="cb388"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb388-1"><a href="#cb388-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"trochObs"</span>, <span class="st">"plumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb388-2"><a href="#cb388-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-3"><a href="#cb388-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb388-4"><a href="#cb388-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb388-5"><a href="#cb388-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-6"><a href="#cb388-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb388-7"><a href="#cb388-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb388-8"><a href="#cb388-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb388-9"><a href="#cb388-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb388-10"><a href="#cb388-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb388-11"><a href="#cb388-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb388-12"><a href="#cb388-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb388-13"><a href="#cb388-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb388-14"><a href="#cb388-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-15"><a href="#cb388-15" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb388-16"><a href="#cb388-16" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb388-17"><a href="#cb388-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-18"><a href="#cb388-18" aria-hidden="true" tabindex="-1"></a><span class="co"># remove individuals that have vir haplotypes, as this could otherwise be mistaken for introgression from obscuratus:</span></span>
<span id="cb388-19"><a href="#cb388-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-20"><a href="#cb388-20" aria-hidden="true" tabindex="-1"></a>removeTheseInds <span class="op">=</span> [<span class="st">"GW_Armando_plate1_JF24G02"</span>, <span class="co"># gw19 hetero from plumb </span></span>
<span id="cb388-21"><a href="#cb388-21" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF07G03"</span>, <span class="co"># gw19 hetero from plumb</span></span>
<span id="cb388-22"><a href="#cb388-22" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF12G02"</span>, <span class="co"># gw19 hetero from plumb</span></span>
<span id="cb388-23"><a href="#cb388-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"GW_Armando_plate1_JF09G01"</span>] <span class="co"># gw28 is hetero from plumb  </span></span>
<span id="cb388-24"><a href="#cb388-24" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(removeTheseInds), metadataForGBI.ind)</span>
<span id="cb388-25"><a href="#cb388-25" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> metadataForGBI[.!selection, <span class="op">:</span>]</span>
<span id="cb388-26"><a href="#cb388-26" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genosForGBI[.!selection, <span class="op">:</span>]</span>
<span id="cb388-27"><a href="#cb388-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-28"><a href="#cb388-28" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb388-29"><a href="#cb388-29" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>]</span>
<span id="cb388-30"><a href="#cb388-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-31"><a href="#cb388-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Set  limit on the number of individuals per group to plot</span></span>
<span id="cb388-32"><a href="#cb388-32" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(plotGroups))</span>
<span id="cb388-33"><a href="#cb388-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-34"><a href="#cb388-34" aria-hidden="true" tabindex="-1"></a><span class="co"># metadataForGBI[metadataForGBI.Fst_group .== "plumb", :]</span></span>
<span id="cb388-35"><a href="#cb388-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-36"><a href="#cb388-36" aria-hidden="true" tabindex="-1"></a>genosForGBI_limited, indMetadataforGBI_limited <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups,</span>
<span id="cb388-37"><a href="#cb388-37" aria-hidden="true" tabindex="-1"></a>                                            numIndsToPlotWithHets, </span>
<span id="cb388-38"><a href="#cb388-38" aria-hidden="true" tabindex="-1"></a>                                            genosForGBI, metadataForGBI;</span>
<span id="cb388-39"><a href="#cb388-39" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb388-40"><a href="#cb388-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-41"><a href="#cb388-41" aria-hidden="true" tabindex="-1"></a><span class="co"># indMetadataforGBI_limited[indMetadataforGBI_limited.Fst_group .== "plumb", :]</span></span>
<span id="cb388-42"><a href="#cb388-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb388-43"><a href="#cb388-43" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb388-44"><a href="#cb388-44" aria-hidden="true" tabindex="-1"></a>                genosForGBI_limited, indMetadataforGBI_limited, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb388-45"><a href="#cb388-45" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-143-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="143">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [70570, 71423, 78271, 89983, 167981, 199863, 204946, 252463, 252487, 367641  …  792896, 867410, 881923, 882012, 925752, 925762, 925928, 945566, 976062, 984894], <span class="ansi-bold">37×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_LN2      GW_Armando_plate2_LN2      LN        troch_L ⋯
   2 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
   3 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
   4 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
   5 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L ⋯
   6 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L
   7 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
   8 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
   9 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L ⋯
  10 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
  11 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
  ⋮  │             ⋮                          ⋮                 ⋮         ⋮    ⋱
  28 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
  29 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
  30 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  31 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  32 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  33 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb   ⋯
  34 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  35 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  36 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  37 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb   ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 16 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-7" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-7">Show just the west area (without nitidus)</h3>
<div id="db96eefe" class="cell" data-execution_count="143">
<div class="sourceCode cell-code" id="cb390"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb390-1"><a href="#cb390-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb390-2"><a href="#cb390-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLudHet"</span>,</span>
<span id="cb390-3"><a href="#cb390-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"virLud_trochObs"</span>,</span>
<span id="cb390-4"><a href="#cb390-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObs"</span>,</span>
<span id="cb390-5"><a href="#cb390-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"trochObsHet"</span>]</span>
<span id="cb390-6"><a href="#cb390-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-7"><a href="#cb390-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb390-8"><a href="#cb390-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"blue"</span>,</span>
<span id="cb390-9"><a href="#cb390-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"green"</span>,</span>
<span id="cb390-10"><a href="#cb390-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>,</span>
<span id="cb390-11"><a href="#cb390-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"yellow"</span>]</span>
<span id="cb390-12"><a href="#cb390-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-13"><a href="#cb390-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb390-14"><a href="#cb390-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb390-15"><a href="#cb390-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb390-16"><a href="#cb390-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb390-17"><a href="#cb390-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb390-18"><a href="#cb390-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb390-19"><a href="#cb390-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb390-20"><a href="#cb390-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb390-21"><a href="#cb390-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb390-22"><a href="#cb390-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb390-23"><a href="#cb390-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb390-24"><a href="#cb390-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-25"><a href="#cb390-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsWest))</span>
<span id="cb390-26"><a href="#cb390-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-27"><a href="#cb390-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb390-28"><a href="#cb390-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb390-29"><a href="#cb390-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb390-30"><a href="#cb390-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-31"><a href="#cb390-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb390-32"><a href="#cb390-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb390-33"><a href="#cb390-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb390-34"><a href="#cb390-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb390-35"><a href="#cb390-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-144-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="144">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 0 0 … 0 0], [63088, 153219, 167981, 204946, 252487, 296231, 367641, 423565, 431630, 443933  …  678802, 684399, 741883, 767533, 773736, 792896, 900073, 925762, 939829, 952095], <span class="ansi-bold">180×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   2 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 171 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 172 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L ⋯
 173 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
 174 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
 175 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 176 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L ⋯
 177 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L
 178 │ GW_Lane5_EM1               GW_Lane5_EM1               EM        troch_E
 179 │ GW_Lane5_DA7               GW_Lane5_DA7               XN        obs
 180 │ GW_Lane5_UY4               GW_Lane5_UY4               UY        plumb   ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 159 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-8" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-8">Show just the east area</h3>
<div id="e27037ee" class="cell" data-execution_count="144">
<div class="sourceCode cell-code" id="cb393"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb393-1"><a href="#cb393-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"trochObs"</span>,</span>
<span id="cb393-2"><a href="#cb393-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"trochObsHet"</span>,</span>
<span id="cb393-3"><a href="#cb393-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"trochObs_plumb"</span>,</span>
<span id="cb393-4"><a href="#cb393-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb393-5"><a href="#cb393-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-6"><a href="#cb393-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb393-7"><a href="#cb393-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"yellow"</span>,</span>
<span id="cb393-8"><a href="#cb393-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb393-9"><a href="#cb393-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb393-10"><a href="#cb393-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-11"><a href="#cb393-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb393-12"><a href="#cb393-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb393-13"><a href="#cb393-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb393-14"><a href="#cb393-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb393-15"><a href="#cb393-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb393-16"><a href="#cb393-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb393-17"><a href="#cb393-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb393-18"><a href="#cb393-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb393-19"><a href="#cb393-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb393-20"><a href="#cb393-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb393-21"><a href="#cb393-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb393-22"><a href="#cb393-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-23"><a href="#cb393-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb393-24"><a href="#cb393-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-25"><a href="#cb393-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb393-26"><a href="#cb393-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb393-27"><a href="#cb393-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb393-28"><a href="#cb393-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-29"><a href="#cb393-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb393-30"><a href="#cb393-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb393-31"><a href="#cb393-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb393-32"><a href="#cb393-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb393-33"><a href="#cb393-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-145-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="145">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [70570, 71423, 78271, 89983, 167981, 199863, 204946, 252463, 252487, 367641  …  792896, 867410, 881923, 882012, 925752, 925762, 925928, 945566, 976062, 984894], <span class="ansi-bold">146×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW72      GW_Armando_plate1_TTGW72      SA        l ⋯
   2 │ GW_Armando_plate1_TTGW78      GW_Armando_plate1_TTGW78      SA        l
   3 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l
   4 │ GW_Armando_plate2_TTGW75      GW_Armando_plate2_TTGW75      SA        l
   5 │ GW_Armando_plate1_TTGW70      GW_Armando_plate1_TTGW70      SA        l ⋯
   6 │ GW_Armando_plate2_TTGW77      GW_Armando_plate2_TTGW77      SA        l
   7 │ GW_Armando_plate2_TTGW80      GW_Armando_plate2_TTGW80      SA        l
   8 │ GW_Armando_plate2_TTGW79      GW_Armando_plate2_TTGW79      SA        l
   9 │ GW_Armando_plate1_TTGW-15-08  GW_Armando_plate1_TTGW-15-08  SR        l ⋯
  10 │ GW_Armando_plate2_TTGW28      GW_Armando_plate2_TTGW28      SR        l
  11 │ GW_Armando_plate2_TTGW25      GW_Armando_plate2_TTGW25      SR        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 137 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
 138 │ GW_Armando_plate2_JF24G01     GW_Armando_plate2_JF24G01     VB        p ⋯
 139 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p
 140 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 141 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
 142 │ GW_Armando_plate2_JG02G01     GW_Armando_plate2_JG02G01     PR        p ⋯
 143 │ GW_Armando_plate2_JG02G03     GW_Armando_plate2_JG02G03     PR        p
 144 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p
 145 │ GW_Lane5_SL1                  GW_Lane5_SL1                  SL        p
 146 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 125 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-8" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-8">Show just the northern area</h3>
<div id="7f12df8c" class="cell" data-execution_count="145">
<div class="sourceCode cell-code" id="cb396"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb396-1"><a href="#cb396-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb396-2"><a href="#cb396-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLudHet"</span>,</span>
<span id="cb396-3"><a href="#cb396-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb396-4"><a href="#cb396-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb396-5"><a href="#cb396-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-6"><a href="#cb396-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb396-7"><a href="#cb396-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"blue"</span>,</span>
<span id="cb396-8"><a href="#cb396-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb396-9"><a href="#cb396-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb396-10"><a href="#cb396-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-11"><a href="#cb396-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb396-12"><a href="#cb396-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb396-13"><a href="#cb396-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb396-14"><a href="#cb396-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb396-15"><a href="#cb396-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb396-16"><a href="#cb396-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb396-17"><a href="#cb396-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb396-18"><a href="#cb396-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb396-19"><a href="#cb396-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb396-20"><a href="#cb396-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb396-21"><a href="#cb396-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb396-22"><a href="#cb396-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-23"><a href="#cb396-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb396-24"><a href="#cb396-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-25"><a href="#cb396-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb396-26"><a href="#cb396-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb396-27"><a href="#cb396-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb396-28"><a href="#cb396-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-29"><a href="#cb396-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb396-30"><a href="#cb396-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb396-31"><a href="#cb396-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb396-32"><a href="#cb396-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb396-33"><a href="#cb396-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-146-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="146">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 0 1 … 1 2], [63088, 70570, 71423, 78271, 89983, 153219, 199863, 252463, 296231, 423565  …  882012, 900073, 925752, 925928, 939829, 945566, 952095, 975932, 976062, 984894], <span class="ansi-bold">143×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   2 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 134 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 135 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 136 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 137 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 138 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 139 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb   ⋯
 140 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 141 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 142 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 143 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 122 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="add-chr-4a" class="level2">
<h2 class="anchored" data-anchor-id="add-chr-4a">Add chr 4A</h2>
<div id="f1b41881" class="cell" data-execution_count="146">
<div class="sourceCode cell-code" id="cb399"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb399-1"><a href="#cb399-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb399-2"><a href="#cb399-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw4A"</span></span>
<span id="cb399-3"><a href="#cb399-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-4"><a href="#cb399-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb399-5"><a href="#cb399-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb399-6"><a href="#cb399-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb399-7"><a href="#cb399-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb399-8"><a href="#cb399-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb399-9"><a href="#cb399-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb399-10"><a href="#cb399-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb399-11"><a href="#cb399-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-12"><a href="#cb399-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb399-13"><a href="#cb399-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb399-14"><a href="#cb399-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-15"><a href="#cb399-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb399-16"><a href="#cb399-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb399-17"><a href="#cb399-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb399-18"><a href="#cb399-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb399-19"><a href="#cb399-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-20"><a href="#cb399-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb399-21"><a href="#cb399-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-22"><a href="#cb399-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb399-23"><a href="#cb399-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb399-24"><a href="#cb399-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-25"><a href="#cb399-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb399-26"><a href="#cb399-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-27"><a href="#cb399-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb399-28"><a href="#cb399-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-29"><a href="#cb399-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb399-30"><a href="#cb399-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb399-31"><a href="#cb399-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-32"><a href="#cb399-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb399-33"><a href="#cb399-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb399-34"><a href="#cb399-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb399-35"><a href="#cb399-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb399-36"><a href="#cb399-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb399-37"><a href="#cb399-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb399-38"><a href="#cb399-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb399-39"><a href="#cb399-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb399-40"><a href="#cb399-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-41"><a href="#cb399-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb399-42"><a href="#cb399-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-43"><a href="#cb399-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb399-44"><a href="#cb399-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb399-45"><a href="#cb399-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb399-46"><a href="#cb399-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb399-47"><a href="#cb399-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb399-48"><a href="#cb399-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb399-49"><a href="#cb399-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb399-50"><a href="#cb399-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb399-51"><a href="#cb399-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb399-52"><a href="#cb399-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb399-53"><a href="#cb399-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb399-54"><a href="#cb399-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb399-55"><a href="#cb399-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-56"><a href="#cb399-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb399-57"><a href="#cb399-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb399-58"><a href="#cb399-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-59"><a href="#cb399-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb399-60"><a href="#cb399-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.5</span>) </span>
<span id="cb399-61"><a href="#cb399-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-62"><a href="#cb399-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb399-63"><a href="#cb399-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-64"><a href="#cb399-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb399-65"><a href="#cb399-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb399-66"><a href="#cb399-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb399-67"><a href="#cb399-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb399-68"><a href="#cb399-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb399-69"><a href="#cb399-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb399-70"><a href="#cb399-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb399-71"><a href="#cb399-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb399-72"><a href="#cb399-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb399-73"><a href="#cb399-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb399-74"><a href="#cb399-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb399-75"><a href="#cb399-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-147-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-147-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="147">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-11" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-11">Save the individual colors in the metadata</h3>
<div id="64331aba" class="cell" data-execution_count="147">
<div class="sourceCode cell-code" id="cb404"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb404-1"><a href="#cb404-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb404-2"><a href="#cb404-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb404-3"><a href="#cb404-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb404-4"><a href="#cb404-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb404-5"><a href="#cb404-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb404-6"><a href="#cb404-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="148">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-11" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-11">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="bd6d2a94" class="cell" data-execution_count="148">
<div class="sourceCode cell-code" id="cb406"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb406-1"><a href="#cb406-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb406-2"><a href="#cb406-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb406-3"><a href="#cb406-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb406-4"><a href="#cb406-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb406-5"><a href="#cb406-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb406-6"><a href="#cb406-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb406-7"><a href="#cb406-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb406-8"><a href="#cb406-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb406-9"><a href="#cb406-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb406-10"><a href="#cb406-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb406-11"><a href="#cb406-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb406-12"><a href="#cb406-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-149-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="149">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-11" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-11">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="e290ca43" class="cell" data-execution_count="149">
<div class="sourceCode cell-code" id="cb409"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb409-1"><a href="#cb409-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb409-2"><a href="#cb409-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb409-3"><a href="#cb409-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb409-4"><a href="#cb409-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb409-5"><a href="#cb409-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb409-6"><a href="#cb409-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb409-7"><a href="#cb409-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb409-8"><a href="#cb409-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb409-9"><a href="#cb409-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb409-10"><a href="#cb409-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb409-11"><a href="#cb409-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb409-12"><a href="#cb409-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-150-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="150">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-4a-high-hetvar-region-there-are-4-clear-haplogroups-no-discernment-of-lud-or-obs-on-pc3-which-is-driven-by-nitidus.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-4a-high-hetvar-region-there-are-4-clear-haplogroups-no-discernment-of-lud-or-obs-on-pc3-which-is-driven-by-nitidus.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 4A high HetVar region, there are 4 clear haplogroups (no discernment of lud or obs on PC3, which is driven by nitidus). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="6a6e3014" class="cell" data-execution_count="150">
<div class="sourceCode cell-code" id="cb412"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb412-1"><a href="#cb412-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb412-2"><a href="#cb412-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb412-3"><a href="#cb412-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb412-4"><a href="#cb412-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obsPlumb"</span>]</span>
<span id="cb412-5"><a href="#cb412-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-6"><a href="#cb412-6" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb412-7"><a href="#cb412-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb412-8"><a href="#cb412-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb412-9"><a href="#cb412-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb412-10"><a href="#cb412-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-11"><a href="#cb412-11" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb412-12"><a href="#cb412-12" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb412-13"><a href="#cb412-13" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb412-14"><a href="#cb412-14" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb412-15"><a href="#cb412-15" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb412-16"><a href="#cb412-16" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span></span>
<span id="cb412-17"><a href="#cb412-17" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb412-18"><a href="#cb412-18" aria-hidden="true" tabindex="-1"></a>obsPlumb <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span> </span>
<span id="cb412-19"><a href="#cb412-19" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb412-20"><a href="#cb412-20" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb412-21"><a href="#cb412-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-22"><a href="#cb412-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb412-23"><a href="#cb412-23" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[virLud]</span>
<span id="cb412-24"><a href="#cb412-24" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb412-25"><a href="#cb412-25" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb412-26"><a href="#cb412-26" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obsPlumb]</span>
<span id="cb412-27"><a href="#cb412-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-28"><a href="#cb412-28" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud nit troch obsPlumb]</span>
<span id="cb412-29"><a href="#cb412-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-30"><a href="#cb412-30" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb412-31"><a href="#cb412-31" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb412-32"><a href="#cb412-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-33"><a href="#cb412-33" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb412-34"><a href="#cb412-34" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb412-35"><a href="#cb412-35" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb412-36"><a href="#cb412-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb412-37"><a href="#cb412-37" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb412-38"><a href="#cb412-38" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb412-39"><a href="#cb412-39" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb412-40"><a href="#cb412-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-41"><a href="#cb412-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb412-42"><a href="#cb412-42" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb412-43"><a href="#cb412-43" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb412-44"><a href="#cb412-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-45"><a href="#cb412-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb412-46"><a href="#cb412-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-47"><a href="#cb412-47" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb412-48"><a href="#cb412-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-49"><a href="#cb412-49" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb412-50"><a href="#cb412-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-51"><a href="#cb412-51" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb412-52"><a href="#cb412-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-53"><a href="#cb412-53" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb412-54"><a href="#cb412-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-55"><a href="#cb412-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb412-56"><a href="#cb412-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-57"><a href="#cb412-57" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb412-58"><a href="#cb412-58" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb412-59"><a href="#cb412-59" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi         </span></span>
<span id="cb412-60"><a href="#cb412-60" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64    </span></span>
<span id="cb412-61"><a href="#cb412-61" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼──────────────────────</span></span>
<span id="cb412-62"><a href="#cb412-62" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.00664772</span></span>
<span id="cb412-63"><a href="#cb412-63" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.00609756</span></span>
<span id="cb412-64"><a href="#cb412-64" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch     0.00614846</span></span>
<span id="cb412-65"><a href="#cb412-65" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obsPlumb  0.00206023 =#</span></span>
<span id="cb412-66"><a href="#cb412-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-67"><a href="#cb412-67" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb412-68"><a href="#cb412-68" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb412-69"><a href="#cb412-69" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy       </span></span>
<span id="cb412-70"><a href="#cb412-70" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64   </span></span>
<span id="cb412-71"><a href="#cb412-71" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼────────────────────────────</span></span>
<span id="cb412-72"><a href="#cb412-72" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.0447451</span></span>
<span id="cb412-73"><a href="#cb412-73" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch     0.0343779</span></span>
<span id="cb412-74"><a href="#cb412-74" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obsPlumb  0.0217185</span></span>
<span id="cb412-75"><a href="#cb412-75" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_troch        0.0373178</span></span>
<span id="cb412-76"><a href="#cb412-76" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_obsPlumb     0.0317873</span></span>
<span id="cb412-77"><a href="#cb412-77" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ troch_obsPlumb   0.0237405 =#</span></span>
<span id="cb412-78"><a href="#cb412-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-79"><a href="#cb412-79" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb412-80"><a href="#cb412-80" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb412-81"><a href="#cb412-81" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb412-82"><a href="#cb412-82" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb412-83"><a href="#cb412-83" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb412-84"><a href="#cb412-84" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb412-85"><a href="#cb412-85" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb412-86"><a href="#cb412-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-87"><a href="#cb412-87" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb412-88"><a href="#cb412-88" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb412-89"><a href="#cb412-89" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb412-90"><a href="#cb412-90" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb412-91"><a href="#cb412-91" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb412-92"><a href="#cb412-92" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb412-93"><a href="#cb412-93" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb412-94"><a href="#cb412-94" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb412-95"><a href="#cb412-95" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb412-96"><a href="#cb412-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-97"><a href="#cb412-97" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb412-98"><a href="#cb412-98" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">150</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb412-99"><a href="#cb412-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-100"><a href="#cb412-100" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb412-101"><a href="#cb412-101" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb412-102"><a href="#cb412-102" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb412-103"><a href="#cb412-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-104"><a href="#cb412-104" aria-hidden="true" tabindex="-1"></a><span class="co"># sort based on original_plot_order, and then together with function below will arrange individuals in population order within clusters:</span></span>
<span id="cb412-105"><a href="#cb412-105" aria-hidden="true" tabindex="-1"></a>sortOrder <span class="op">=</span> <span class="fu">sortperm</span>(indMetadataforGBI.original_plot_order, rev<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb412-106"><a href="#cb412-106" aria-hidden="true" tabindex="-1"></a>indMetadataforGBI <span class="op">=</span> indMetadataforGBI[sortOrder, <span class="op">:</span>]</span>
<span id="cb412-107"><a href="#cb412-107" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genosForGBI[sortOrder, <span class="op">:</span>]</span>
<span id="cb412-108"><a href="#cb412-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-109"><a href="#cb412-109" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb412-110"><a href="#cb412-110" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb412-111"><a href="#cb412-111" aria-hidden="true" tabindex="-1"></a>                indFontSize<span class="op">=</span><span class="fl">6</span>, figureSize<span class="op">=</span>(<span class="fl">800</span>, <span class="fl">1800</span>),</span>
<span id="cb412-112"><a href="#cb412-112" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb412-113"><a href="#cb412-113" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb412-114"><a href="#cb412-114" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [40 2 62 91] and the sum of those is 195
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-151-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="151">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [387825, 397781, 397824, 405798, 408642, 419198, 425296, 437077, 447665, 454290  …  633803, 637531, 638456, 656257, 676526, 690700, 699577, 706480, 707103, 711862], <span class="ansi-bold">195×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 186 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
 187 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb   ⋯
 188 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
 189 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 190 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 191 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb   ⋯
 192 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 193 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 194 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 195 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb   ⋯
<span class="ansi-cyan-fg">                                                 37 columns and 174 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-11" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-11">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="b2ce344e" class="cell" data-execution_count="151">
<div class="sourceCode cell-code" id="cb415"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb415-1"><a href="#cb415-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb415-2"><a href="#cb415-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"virLudHet"</span>,</span>
<span id="cb415-3"><a href="#cb415-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb415-4"><a href="#cb415-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"virLud_troch"</span>,</span>
<span id="cb415-5"><a href="#cb415-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb415-6"><a href="#cb415-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"trochHet"</span>,</span>
<span id="cb415-7"><a href="#cb415-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch_obsPlumb"</span>,</span>
<span id="cb415-8"><a href="#cb415-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obsPlumb"</span>,</span>
<span id="cb415-9"><a href="#cb415-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obsPlumbHet"</span>,</span>
<span id="cb415-10"><a href="#cb415-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"virLud_obsPlumb"</span>]</span>
<span id="cb415-11"><a href="#cb415-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-12"><a href="#cb415-12" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb415-13"><a href="#cb415-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"blue"</span>,</span>
<span id="cb415-14"><a href="#cb415-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb415-15"><a href="#cb415-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb415-16"><a href="#cb415-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb415-17"><a href="#cb415-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb415-18"><a href="#cb415-18" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb415-19"><a href="#cb415-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb415-20"><a href="#cb415-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb415-21"><a href="#cb415-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">"purple"</span>]</span>
<span id="cb415-22"><a href="#cb415-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-23"><a href="#cb415-23" aria-hidden="true" tabindex="-1"></a>virLud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb415-24"><a href="#cb415-24" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-25"><a href="#cb415-25" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb415-26"><a href="#cb415-26" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb415-27"><a href="#cb415-27" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb415-28"><a href="#cb415-28" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span></span>
<span id="cb415-29"><a href="#cb415-29" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb415-30"><a href="#cb415-30" aria-hidden="true" tabindex="-1"></a>obsPlumb <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span> </span>
<span id="cb415-31"><a href="#cb415-31" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb415-32"><a href="#cb415-32" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb415-33"><a href="#cb415-33" aria-hidden="true" tabindex="-1"></a>virLudHet <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2</span>) <span class="op">.&amp;</span></span>
<span id="cb415-34"><a href="#cb415-34" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-35"><a href="#cb415-35" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb415-36"><a href="#cb415-36" aria-hidden="true" tabindex="-1"></a>virLud_troch <span class="op">=</span> (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb415-37"><a href="#cb415-37" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">0.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-38"><a href="#cb415-38" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb415-39"><a href="#cb415-39" aria-hidden="true" tabindex="-1"></a>trochHet <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1) <span class="op">.&amp;</span></span>
<span id="cb415-40"><a href="#cb415-40" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb415-41"><a href="#cb415-41" aria-hidden="true" tabindex="-1"></a>troch_obsPlumb <span class="op">=</span> (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">3</span>) <span class="op">.&amp;</span> </span>
<span id="cb415-42"><a href="#cb415-42" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-43"><a href="#cb415-43" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3) <span class="op">.&amp;</span></span>
<span id="cb415-44"><a href="#cb415-44" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb415-45"><a href="#cb415-45" aria-hidden="true" tabindex="-1"></a>obsPlumbHet <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span> </span>
<span id="cb415-46"><a href="#cb415-46" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb415-47"><a href="#cb415-47" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb415-48"><a href="#cb415-48" aria-hidden="true" tabindex="-1"></a>virLud_obsPlumb <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-49"><a href="#cb415-49" aria-hidden="true" tabindex="-1"></a>                (<span class="op">-</span><span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1.5</span>) <span class="op">.&amp;</span></span>
<span id="cb415-50"><a href="#cb415-50" aria-hidden="true" tabindex="-1"></a>                 .!indSelection_lowIndHetStan</span>
<span id="cb415-51"><a href="#cb415-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-52"><a href="#cb415-52" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [virLud virLudHet nit virLud_troch troch trochHet troch_obsPlumb obsPlumb obsPlumbHet virLud_obsPlumb]</span>
<span id="cb415-53"><a href="#cb415-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-54"><a href="#cb415-54" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb415-55"><a href="#cb415-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-56"><a href="#cb415-56" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb415-57"><a href="#cb415-57" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb415-58"><a href="#cb415-58" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb415-59"><a href="#cb415-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb415-60"><a href="#cb415-60" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb415-61"><a href="#cb415-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-62"><a href="#cb415-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-63"><a href="#cb415-63" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb415-64"><a href="#cb415-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-65"><a href="#cb415-65" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb415-66"><a href="#cb415-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-67"><a href="#cb415-67" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb415-68"><a href="#cb415-68" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb415-69"><a href="#cb415-69" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb415-70"><a href="#cb415-70" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb415-71"><a href="#cb415-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-72"><a href="#cb415-72" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb415-73"><a href="#cb415-73" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb415-74"><a href="#cb415-74" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb415-75"><a href="#cb415-75" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb415-76"><a href="#cb415-76" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb415-77"><a href="#cb415-77" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb415-78"><a href="#cb415-78" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb415-79"><a href="#cb415-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-80"><a href="#cb415-80" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb415-81"><a href="#cb415-81" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb415-82"><a href="#cb415-82" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb415-83"><a href="#cb415-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-84"><a href="#cb415-84" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb415-85"><a href="#cb415-85" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb415-86"><a href="#cb415-86" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb415-87"><a href="#cb415-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-88"><a href="#cb415-88" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb415-89"><a href="#cb415-89" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb415-90"><a href="#cb415-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-91"><a href="#cb415-91" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb415-92"><a href="#cb415-92" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb415-93"><a href="#cb415-93" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb415-94"><a href="#cb415-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-95"><a href="#cb415-95" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb415-96"><a href="#cb415-96" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb415-97"><a href="#cb415-97" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb415-98"><a href="#cb415-98" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb415-99"><a href="#cb415-99" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-152-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="152">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [387825, 397781, 397824, 405798, 408642, 419198, 425296, 437077, 447665, 454290  …  633803, 637531, 638456, 656257, 676526, 690700, 699577, 706480, 707103, 711862], <span class="ansi-bold">95×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04     GW_Armando_plate1_JF12G04     ST_vi     v ⋯
   2 │ GW_Armando_plate2_JF03G02     GW_Armando_plate2_JF03G02     VB_vi     v
   3 │ GW_Armando_plate2_JF16G02     GW_Armando_plate2_JF16G02     DV_vi     p
   4 │ GW_Armando_plate2_JF30G01     GW_Armando_plate2_JF30G01     ST_vi     v
   5 │ GW_Lane5_AA5                  GW_Lane5_AA5                  AA        v ⋯
   6 │ GW_Lane5_AA6                  GW_Lane5_AA6                  AA        v
   7 │ GW_Lane5_AA7                  GW_Lane5_AA7                  AA        v
   8 │ GW_Lane5_STvi1                GW_Lane5_STvi1                ST_vi     v
   9 │ GW_Lane5_STvi3                GW_Lane5_STvi3                ST_vi     v ⋯
  10 │ GW_Lane5_TL1                  GW_Lane5_TL1                  TL        v
  11 │ GW_Lane5_TL10                 GW_Lane5_TL10                 TL        v
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
  86 │ GW_Armando_plate1_TTGW53      GW_Armando_plate1_TTGW53      NG        l
  87 │ GW_Armando_plate1_TTGW-15-09  GW_Armando_plate1_TTGW-15-09  SR        l ⋯
  88 │ GW_Armando_plate2_JG01G01     GW_Armando_plate2_JG01G01     PR        p
  89 │ GW_Armando_plate2_TTGW18      GW_Armando_plate2_TTGW18      TH        l
  90 │ GW_Armando_plate2_TTGW20      GW_Armando_plate2_TTGW20      SR        l
  91 │ GW_Armando_plate2_TTGW56      GW_Armando_plate2_TTGW56      NG        l ⋯
  92 │ GW_Armando_plate2_TTGW-15-01  GW_Armando_plate2_TTGW-15-01  SR        l
  93 │ GW_Armando_plate2_TTGW-15-03  GW_Armando_plate2_TTGW-15-03  SR        l
  94 │ GW_Armando_plate2_TTGW-15-06  GW_Armando_plate2_TTGW-15-06  SR        l
  95 │ GW_Lane5_KS1                  GW_Lane5_KS1                  OV        l ⋯
<span class="ansi-cyan-fg">                                                  37 columns and 74 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-7" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-7">show GBI plot according to original groups and plot order</h3>
<div id="86de6ce5" class="cell" data-execution_count="152">
<div class="sourceCode cell-code" id="cb418"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb418-1"><a href="#cb418-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb418-2"><a href="#cb418-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb418-3"><a href="#cb418-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb418-4"><a href="#cb418-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb418-5"><a href="#cb418-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb418-6"><a href="#cb418-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb418-7"><a href="#cb418-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb418-8"><a href="#cb418-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb418-9"><a href="#cb418-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb418-10"><a href="#cb418-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb418-11"><a href="#cb418-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-153-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="153">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [387825, 397781, 397824, 405798, 408642, 419198, 425296, 437077, 447665, 454290  …  633803, 637531, 638456, 656257, 676526, 690700, 699577, 706480, 707103, 711862], <span class="ansi-bold">95×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
   7 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
   8 │ GW_Lane5_TL1               GW_Lane5_TL1               TL        vir
   9 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir     ⋯
  10 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
  11 │ GW_Lane5_TL12              GW_Lane5_TL12              TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  86 │ GW_Armando_plate1_JF29G01  GW_Armando_plate1_JF29G01  ST        plumb
  87 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb   ⋯
  88 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  89 │ GW_Armando_plate1_JF29G02  GW_Armando_plate1_JF29G02  ST        plumb
  90 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  91 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb   ⋯
  92 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  93 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  94 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb   ⋯
<span class="ansi-cyan-fg">                                                  37 columns and 74 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-7" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-7">show same but with all individuals</h3>
<div id="d0b8310f" class="cell" data-execution_count="153">
<div class="sourceCode cell-code" id="cb420"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb420-1"><a href="#cb420-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb420-2"><a href="#cb420-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb420-3"><a href="#cb420-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb420-4"><a href="#cb420-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb420-5"><a href="#cb420-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb420-6"><a href="#cb420-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb420-7"><a href="#cb420-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb420-8"><a href="#cb420-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb420-9"><a href="#cb420-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb420-10"><a href="#cb420-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb420-11"><a href="#cb420-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb420-12"><a href="#cb420-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb420-13"><a href="#cb420-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb420-14"><a href="#cb420-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-154-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="154">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 0 … 0 0], [387825, 397781, 397824, 405798, 408642, 419198, 425296, 437077, 447665, 454290  …  633803, 637531, 638456, 656257, 676526, 690700, 699577, 706480, 707103, 711862], <span class="ansi-bold">257×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 249 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 252 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 253 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 37 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-only-vir-and-plumb-pops-5" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-only-vir-and-plumb-pops-5">show same but with only vir and plumb pops</h3>
<div id="5740bb80" class="cell" data-execution_count="154">
<div class="sourceCode cell-code" id="cb422"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb422-1"><a href="#cb422-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>, <span class="st">"obsPlumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb422-2"><a href="#cb422-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-3"><a href="#cb422-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb422-4"><a href="#cb422-4" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb422-5"><a href="#cb422-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-6"><a href="#cb422-6" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb422-7"><a href="#cb422-7" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb422-8"><a href="#cb422-8" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb422-9"><a href="#cb422-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb422-10"><a href="#cb422-10" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb422-11"><a href="#cb422-11" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb422-12"><a href="#cb422-12" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb422-13"><a href="#cb422-13" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb422-14"><a href="#cb422-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-15"><a href="#cb422-15" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>, <span class="st">"plumb"</span>, <span class="st">"plumb_vir"</span>] <span class="co"># these are the original Fst_groups</span></span>
<span id="cb422-16"><a href="#cb422-16" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"purple"</span>]</span>
<span id="cb422-17"><a href="#cb422-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-18"><a href="#cb422-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb422-19"><a href="#cb422-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-20"><a href="#cb422-20" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb422-21"><a href="#cb422-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb422-22"><a href="#cb422-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb422-23"><a href="#cb422-23" aria-hidden="true" tabindex="-1"></a>                genosForGBI, metadataForGBI, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb422-24"><a href="#cb422-24" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-155-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="155">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 2 … 1 0], [397824, 447665, 454290, 505522, 510299, 520268, 527334, 531726, 578230, 582505, 587143, 614605, 617468, 617555, 621740, 633718, 633803, 690700], <span class="ansi-bold">100×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  37 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-whole-ring-but-not-nit" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-whole-ring-but-not-nit">show same but whole ring (but not nit)</h3>
<div id="f4b59a11" class="cell" data-execution_count="155">
<div class="sourceCode cell-code" id="cb424"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb424-1"><a href="#cb424-1" aria-hidden="true" tabindex="-1"></a>includeTheseClusters <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb424-2"><a href="#cb424-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"troch"</span>,</span>
<span id="cb424-3"><a href="#cb424-3" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"obsPlumb"</span>] <span class="co"># these are the haplotype clusters to include in the choice below of SNPs to show</span></span>
<span id="cb424-4"><a href="#cb424-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-5"><a href="#cb424-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb424-6"><a href="#cb424-6" aria-hidden="true" tabindex="-1"></a>freqs_local, sampleSizes_local <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembership, includeTheseClusters)</span>
<span id="cb424-7"><a href="#cb424-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-8"><a href="#cb424-8" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb424-9"><a href="#cb424-9" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb424-10"><a href="#cb424-10" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb424-11"><a href="#cb424-11" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb424-12"><a href="#cb424-12" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs_local, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb424-13"><a href="#cb424-13" aria-hidden="true" tabindex="-1"></a>genosForGBI <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb424-14"><a href="#cb424-14" aria-hidden="true" tabindex="-1"></a>posForGBI <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb424-15"><a href="#cb424-15" aria-hidden="true" tabindex="-1"></a>freqsForGBI <span class="op">=</span> freqs_local[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb424-16"><a href="#cb424-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-17"><a href="#cb424-17" aria-hidden="true" tabindex="-1"></a>metadataForGBI <span class="op">=</span> <span class="fu">copy</span>(PCAmodelAll.metadata)</span>
<span id="cb424-18"><a href="#cb424-18" aria-hidden="true" tabindex="-1"></a>metadataForGBI.Fst_group <span class="op">=</span> metadataForGBI.original_Fst_groups</span>
<span id="cb424-19"><a href="#cb424-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-20"><a href="#cb424-20" aria-hidden="true" tabindex="-1"></a><span class="co"># plotGroups = ["vir","vir_S","lud_PK", "lud_KS", "lud_central", "lud_Sath", "lud_ML","troch_west","troch_LN","troch_EM","obs","plumb_BJ","plumb","plumb_vir"]</span></span>
<span id="cb424-21"><a href="#cb424-21" aria-hidden="true" tabindex="-1"></a><span class="co"># plotGroupColors = ["blue","turquoise1","seagreen4","seagreen3","seagreen2","olivedrab3","olivedrab2","olivedrab1","yellow","gold","orange","pink","red","purple"]</span></span>
<span id="cb424-22"><a href="#cb424-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-23"><a href="#cb424-23" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"lud_PK"</span>,<span class="st">"lud_KS"</span>,<span class="st">"lud_central"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb424-24"><a href="#cb424-24" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>, <span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>]</span>
<span id="cb424-25"><a href="#cb424-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-26"><a href="#cb424-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Set  limit on the number of individuals per group to plot</span></span>
<span id="cb424-27"><a href="#cb424-27" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> [<span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">10</span>, <span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">10</span>, <span class="fl">1</span>] <span class="co"># maximum number of individuals to plot from each group</span></span>
<span id="cb424-28"><a href="#cb424-28" aria-hidden="true" tabindex="-1"></a><span class="co">#numIndsToPlotWithHets = fill(100, length(plotGroups))</span></span>
<span id="cb424-29"><a href="#cb424-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-30"><a href="#cb424-30" aria-hidden="true" tabindex="-1"></a><span class="co"># metadataForGBI[metadataForGBI.Fst_group .== "plumb", :]</span></span>
<span id="cb424-31"><a href="#cb424-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-32"><a href="#cb424-32" aria-hidden="true" tabindex="-1"></a>genosForGBI_limited, indMetadataforGBI_limited <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(plotGroups,</span>
<span id="cb424-33"><a href="#cb424-33" aria-hidden="true" tabindex="-1"></a>                                            numIndsToPlotWithHets, </span>
<span id="cb424-34"><a href="#cb424-34" aria-hidden="true" tabindex="-1"></a>                                            genosForGBI, metadataForGBI;</span>
<span id="cb424-35"><a href="#cb424-35" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb424-36"><a href="#cb424-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-37"><a href="#cb424-37" aria-hidden="true" tabindex="-1"></a><span class="co"># indMetadataforGBI_limited[indMetadataforGBI_limited.Fst_group .== "plumb", :]</span></span>
<span id="cb424-38"><a href="#cb424-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-39"><a href="#cb424-39" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, posForGBI,</span>
<span id="cb424-40"><a href="#cb424-40" aria-hidden="true" tabindex="-1"></a>                genosForGBI_limited, indMetadataforGBI_limited, freqsForGBI, plotGroups, plotGroupColors;</span>
<span id="cb424-41"><a href="#cb424-41" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-156-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="156">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 0 0 … 0 0; 0 1 … 0 0], [397781, 397824, 405798, 419198, 425296, 447665, 454290, 505522, 510299, 513889  …  617468, 617539, 617555, 621740, 633718, 633787, 633803, 676526, 690700, 711862], <span class="ansi-bold">61×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   7 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
   8 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   9 │ GW_Lane5_AB2               GW_Lane5_AB2               AB        vir     ⋯
  10 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  11 │ GW_Lane5_AA1               GW_Lane5_AA1               AA        vir_S
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  52 │ GW_Armando_plate1_JF07G03  GW_Armando_plate1_JF07G03  ST        plumb
  53 │ GW_Armando_plate1_JF07G04  GW_Armando_plate1_JF07G04  ST        plumb   ⋯
  54 │ GW_Armando_plate1_JF08G02  GW_Armando_plate1_JF08G02  ST        plumb
  55 │ GW_Armando_plate1_JF09G01  GW_Armando_plate1_JF09G01  ST        plumb
  56 │ GW_Armando_plate1_JF09G02  GW_Armando_plate1_JF09G02  ST        plumb
  57 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb   ⋯
  58 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  59 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb
  60 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  61 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  37 columns and 40 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-8" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-8">Show just the west area (without nitidus)</h3>
<div id="9d72cf67" class="cell" data-execution_count="156">
<div class="sourceCode cell-code" id="cb426"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb426-1"><a href="#cb426-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb426-2"><a href="#cb426-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLudHet"</span>,</span>
<span id="cb426-3"><a href="#cb426-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLud_troch"</span>,</span>
<span id="cb426-4"><a href="#cb426-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"troch"</span>,</span>
<span id="cb426-5"><a href="#cb426-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"trochHet"</span>]</span>
<span id="cb426-6"><a href="#cb426-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb426-7"><a href="#cb426-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb426-8"><a href="#cb426-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"blue"</span>,</span>
<span id="cb426-9"><a href="#cb426-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"green"</span>,</span>
<span id="cb426-10"><a href="#cb426-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"yellow"</span>,</span>
<span id="cb426-11"><a href="#cb426-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"yellow"</span>]</span>
<span id="cb426-12"><a href="#cb426-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb426-13"><a href="#cb426-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb426-14"><a href="#cb426-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb426-15"><a href="#cb426-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb426-16"><a href="#cb426-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb426-17"><a href="#cb426-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb426-18"><a href="#cb426-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb426-19"><a href="#cb426-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb426-20"><a href="#cb426-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb426-21"><a href="#cb426-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb426-22"><a href="#cb426-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb426-23"><a href="#cb426-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb426-24"><a href="#cb426-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb426-25"><a href="#cb426-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsWest))</span>
<span id="cb426-26"><a href="#cb426-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb426-27"><a href="#cb426-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb426-28"><a href="#cb426-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb426-29"><a href="#cb426-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb426-30"><a href="#cb426-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb426-31"><a href="#cb426-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb426-32"><a href="#cb426-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb426-33"><a href="#cb426-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb426-34"><a href="#cb426-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb426-35"><a href="#cb426-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-157-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="157">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 2 … 2 0; 0 0 … 0 0], [397781, 397824, 405798, 419198, 425296, 447665, 510299, 513889, 520268, 520285  …  614605, 614609, 617539, 617555, 621740, 633718, 633787, 676526, 690700, 711862], <span class="ansi-bold">119×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 110 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 111 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L ⋯
 112 │ GW_Lane5_LN1               GW_Lane5_LN1               LN        troch_L
 113 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
 114 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L
 115 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L ⋯
 116 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 117 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L
 118 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
 119 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  37 columns and 98 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-9" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-9">Show just the east area</h3>
<div id="669fd2ba" class="cell" data-execution_count="157">
<div class="sourceCode cell-code" id="cb429"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb429-1"><a href="#cb429-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb429-2"><a href="#cb429-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"trochHet"</span>,</span>
<span id="cb429-3"><a href="#cb429-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"troch_obsPlumb"</span>,</span>
<span id="cb429-4"><a href="#cb429-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obsPlumb"</span>,</span>
<span id="cb429-5"><a href="#cb429-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obsPlumbHet"</span>]</span>
<span id="cb429-6"><a href="#cb429-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-7"><a href="#cb429-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb429-8"><a href="#cb429-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"yellow"</span>,</span>
<span id="cb429-9"><a href="#cb429-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb429-10"><a href="#cb429-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb429-11"><a href="#cb429-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb429-12"><a href="#cb429-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-13"><a href="#cb429-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb429-14"><a href="#cb429-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb429-15"><a href="#cb429-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb429-16"><a href="#cb429-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb429-17"><a href="#cb429-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb429-18"><a href="#cb429-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb429-19"><a href="#cb429-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb429-20"><a href="#cb429-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb429-21"><a href="#cb429-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb429-22"><a href="#cb429-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb429-23"><a href="#cb429-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb429-24"><a href="#cb429-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-25"><a href="#cb429-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb429-26"><a href="#cb429-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-27"><a href="#cb429-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb429-28"><a href="#cb429-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb429-29"><a href="#cb429-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb429-30"><a href="#cb429-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-31"><a href="#cb429-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb429-32"><a href="#cb429-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb429-33"><a href="#cb429-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb429-34"><a href="#cb429-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb429-35"><a href="#cb429-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-158-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="158">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[2 2 … 2 2; 2 2 … 2 2; … ; 2 2 … 2 2; 2 2 … 2 2], [397781, 405798, 419198, 425296, 454290, 505522, 513889, 520285, 527255, 531214  …  553195, 582581, 587134, 614609, 617468, 617539, 633787, 633803, 676526, 711862], <span class="ansi-bold">180×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                          </span><span class="ansi-bold"> ID                           </span><span class="ansi-bold"> location </span><span class="ansi-bold"> g</span> ⋯
     │<span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String                       </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> S</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Lane5_NA3                  GW_Lane5_NA3                  NR        l ⋯
   2 │ GW_Lane5_SH1                  GW_Lane5_SH1                  SH        l
   3 │ GW_Lane5_SH2                  GW_Lane5_SH2                  SH        l
   4 │ GW_Lane5_SH4                  GW_Lane5_SH4                  SH        l
   5 │ GW_Lane5_SH5                  GW_Lane5_SH5                  SH        l ⋯
   6 │ GW_Armando_plate1_TTGW72      GW_Armando_plate1_TTGW72      SA        l
   7 │ GW_Armando_plate2_TTGW73      GW_Armando_plate2_TTGW73      SA        l
   8 │ GW_Armando_plate2_TTGW77      GW_Armando_plate2_TTGW77      SA        l
   9 │ GW_Armando_plate2_TTGW79      GW_Armando_plate2_TTGW79      SA        l ⋯
  10 │ GW_Armando_plate2_TTGW80      GW_Armando_plate2_TTGW80      SA        l
  11 │ GW_Armando_plate2_TTGW75      GW_Armando_plate2_TTGW75      SA        l
  ⋮  │              ⋮                             ⋮                   ⋮        ⋱
 171 │ GW_Armando_plate1_JF23G02     GW_Armando_plate1_JF23G02     VB        p
 172 │ GW_Armando_plate1_JF24G02     GW_Armando_plate1_JF24G02     VB        p ⋯
 173 │ GW_Armando_plate2_JF24G01     GW_Armando_plate2_JF24G01     VB        p
 174 │ GW_Armando_plate2_JF25G01     GW_Armando_plate2_JF25G01     VB        p
 175 │ GW_Armando_plate1_JG02G02     GW_Armando_plate1_JG02G02     PR        p
 176 │ GW_Armando_plate1_JG02G04     GW_Armando_plate1_JG02G04     PR        p ⋯
 177 │ GW_Armando_plate2_JG02G01     GW_Armando_plate2_JG02G01     PR        p
 178 │ GW_Armando_plate2_JG02G03     GW_Armando_plate2_JG02G03     PR        p
 179 │ GW_Lane5_SL1                  GW_Lane5_SL1                  SL        p
 180 │ GW_Lane5_SL2                  GW_Lane5_SL2                  SL        p ⋯
<span class="ansi-cyan-fg">                                                 37 columns and 159 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-9" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-9">Show just the northern area</h3>
<div id="f26b6117" class="cell" data-execution_count="158">
<div class="sourceCode cell-code" id="cb432"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb432-1"><a href="#cb432-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb432-2"><a href="#cb432-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLudHet"</span>,</span>
<span id="cb432-3"><a href="#cb432-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"virLud_obsPlumb"</span>,</span>
<span id="cb432-4"><a href="#cb432-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obsPlumb"</span>,</span>
<span id="cb432-5"><a href="#cb432-5" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obsPlumbHet"</span>]</span>
<span id="cb432-6"><a href="#cb432-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-7"><a href="#cb432-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb432-8"><a href="#cb432-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"blue"</span>,</span>
<span id="cb432-9"><a href="#cb432-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb432-10"><a href="#cb432-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>,</span>
<span id="cb432-11"><a href="#cb432-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb432-12"><a href="#cb432-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-13"><a href="#cb432-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb432-14"><a href="#cb432-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb432-15"><a href="#cb432-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb432-16"><a href="#cb432-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb432-17"><a href="#cb432-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb432-18"><a href="#cb432-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb432-19"><a href="#cb432-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb432-20"><a href="#cb432-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb432-21"><a href="#cb432-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb432-22"><a href="#cb432-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb432-23"><a href="#cb432-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb432-24"><a href="#cb432-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-25"><a href="#cb432-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb432-26"><a href="#cb432-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-27"><a href="#cb432-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb432-28"><a href="#cb432-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb432-29"><a href="#cb432-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb432-30"><a href="#cb432-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-31"><a href="#cb432-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb432-32"><a href="#cb432-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb432-33"><a href="#cb432-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb432-34"><a href="#cb432-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb432-35"><a href="#cb432-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-159-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="159">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 2 … 1 0], [397824, 447665, 454290, 505522, 510299, 520268, 527334, 531726, 578230, 582505, 587143, 614605, 617468, 617555, 621740, 633718, 633803, 690700], <span class="ansi-bold">157×40 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   3 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   4 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   5 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   8 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 148 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 149 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 150 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 151 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 152 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 153 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 154 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 155 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 156 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 157 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 37 columns and 136 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw4a-with-low-hetvar" class="level3">
<h3 class="anchored" data-anchor-id="do-a-pca-based-on-a-same-size-region-elsewhere-on-gw4a-with-low-hetvar">Do a PCA based on a same-size region elsewhere on gw4A (with low HetVar):</h3>
<div id="f6d51554" class="cell" data-execution_count="159">
<div class="sourceCode cell-code" id="cb435"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb435-1"><a href="#cb435-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get length of region</span></span>
<span id="cb435-2"><a href="#cb435-2" aria-hidden="true" tabindex="-1"></a>lengthHighHetVarRegion <span class="op">=</span> positionMax <span class="op">-</span> positionMin</span>
<span id="cb435-3"><a href="#cb435-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-4"><a href="#cb435-4" aria-hidden="true" tabindex="-1"></a><span class="co"># because the region is on the left side of chr 4A, will put the non-interesting region on the right side:</span></span>
<span id="cb435-5"><a href="#cb435-5" aria-hidden="true" tabindex="-1"></a>rightLocus <span class="op">=</span> scaffold_lengths[<span class="st">"gw4A"</span>] <span class="op">-</span> <span class="fl">1_000_000</span></span>
<span id="cb435-6"><a href="#cb435-6" aria-hidden="true" tabindex="-1"></a>leftLocus <span class="op">=</span> rightLocus <span class="op">-</span> lengthHighHetVarRegion</span>
<span id="cb435-7"><a href="#cb435-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-8"><a href="#cb435-8" aria-hidden="true" tabindex="-1"></a>regionText_lowHetVarRegion <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr "</span>, chr, <span class="st">" "</span>,leftLocus,<span class="st">" to "</span>,rightLocus)</span>
<span id="cb435-9"><a href="#cb435-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-10"><a href="#cb435-10" aria-hidden="true" tabindex="-1"></a>lociSelection <span class="op">=</span> (leftLocus <span class="op">.&lt;=</span> pos_region.position <span class="op">.&lt;=</span> rightLocus)</span>
<span id="cb435-11"><a href="#cb435-11" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion <span class="op">=</span> genotypes_region[<span class="op">:</span>, lociSelection]</span>
<span id="cb435-12"><a href="#cb435-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-13"><a href="#cb435-13" aria-hidden="true" tabindex="-1"></a><span class="co"># impute missing genotypes:</span></span>
<span id="cb435-14"><a href="#cb435-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-15"><a href="#cb435-15" aria-hidden="true" tabindex="-1"></a>genotypes_lowHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genotypes_lowHetVarRegion))</span>
<span id="cb435-16"><a href="#cb435-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-17"><a href="#cb435-17" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb435-18"><a href="#cb435-18" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb435-19"><a href="#cb435-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-20"><a href="#cb435-20" aria-hidden="true" tabindex="-1"></a>PCAmodel <span class="op">=</span> <span class="fu">plotPCA</span>(genotypes_lowHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb435-21"><a href="#cb435-21" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb435-22"><a href="#cb435-22" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText_lowHetVarRegion,</span>
<span id="cb435-23"><a href="#cb435-23" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb435-24"><a href="#cb435-24" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb435-25"><a href="#cb435-25" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb435-26"><a href="#cb435-26" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb435-27"><a href="#cb435-27" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb435-28"><a href="#cb435-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-29"><a href="#cb435-29" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodel.PCAfig)</span>
<span id="cb435-30"><a href="#cb435-30" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">true</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb435-31"><a href="#cb435-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"FigureS2D_gw4A_nonHLBRarbitrary_from_Julia.png"</span>, PCAmodel.PCAfig, px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb435-32"><a href="#cb435-32" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-160-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="160">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
<section id="chr-4a-shows-really-remarkable-patterns.-absolutely-must-be-selective-sweeps." class="level4">
<h4 class="anchored" data-anchor-id="chr-4a-shows-really-remarkable-patterns.-absolutely-must-be-selective-sweeps.">Chr 4A shows really remarkable patterns. Absolutely must be selective sweeps.</h4>
</section>
</section>
</section>
<section id="add-chr-20" class="level2">
<h2 class="anchored" data-anchor-id="add-chr-20">Add chr 20</h2>
<div id="07cda448" class="cell" data-execution_count="160">
<div class="sourceCode cell-code" id="cb438"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb438-1"><a href="#cb438-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb438-2"><a href="#cb438-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw20"</span></span>
<span id="cb438-3"><a href="#cb438-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-4"><a href="#cb438-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb438-5"><a href="#cb438-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb438-6"><a href="#cb438-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb438-7"><a href="#cb438-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb438-8"><a href="#cb438-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb438-9"><a href="#cb438-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb438-10"><a href="#cb438-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb438-11"><a href="#cb438-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-12"><a href="#cb438-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb438-13"><a href="#cb438-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb438-14"><a href="#cb438-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-15"><a href="#cb438-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb438-16"><a href="#cb438-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb438-17"><a href="#cb438-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb438-18"><a href="#cb438-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb438-19"><a href="#cb438-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-20"><a href="#cb438-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb438-21"><a href="#cb438-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-22"><a href="#cb438-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb438-23"><a href="#cb438-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb438-24"><a href="#cb438-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-25"><a href="#cb438-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb438-26"><a href="#cb438-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-27"><a href="#cb438-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb438-28"><a href="#cb438-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-29"><a href="#cb438-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">false</span></span>
<span id="cb438-30"><a href="#cb438-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb438-31"><a href="#cb438-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-32"><a href="#cb438-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb438-33"><a href="#cb438-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb438-34"><a href="#cb438-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb438-35"><a href="#cb438-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb438-36"><a href="#cb438-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb438-37"><a href="#cb438-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb438-38"><a href="#cb438-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb438-39"><a href="#cb438-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb438-40"><a href="#cb438-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-41"><a href="#cb438-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb438-42"><a href="#cb438-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-43"><a href="#cb438-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb438-44"><a href="#cb438-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb438-45"><a href="#cb438-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb438-46"><a href="#cb438-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb438-47"><a href="#cb438-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb438-48"><a href="#cb438-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb438-49"><a href="#cb438-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb438-50"><a href="#cb438-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb438-51"><a href="#cb438-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb438-52"><a href="#cb438-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb438-53"><a href="#cb438-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb438-54"><a href="#cb438-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb438-55"><a href="#cb438-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-56"><a href="#cb438-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb438-57"><a href="#cb438-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb438-58"><a href="#cb438-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-59"><a href="#cb438-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb438-60"><a href="#cb438-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.5</span>) </span>
<span id="cb438-61"><a href="#cb438-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-62"><a href="#cb438-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb438-63"><a href="#cb438-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-64"><a href="#cb438-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb438-65"><a href="#cb438-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb438-66"><a href="#cb438-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb438-67"><a href="#cb438-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb438-68"><a href="#cb438-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb438-69"><a href="#cb438-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb438-70"><a href="#cb438-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb438-71"><a href="#cb438-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb438-72"><a href="#cb438-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb438-73"><a href="#cb438-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb438-74"><a href="#cb438-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb438-75"><a href="#cb438-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span>
<span id="cb438-76"><a href="#cb438-76" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="cn">false</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb438-77"><a href="#cb438-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">save</span>(<span class="st">"Figure6_top_gw28GBIplotEast_from_Julia.png"</span>, plotInfo[<span class="fl">1</span>], px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb438-78"><a href="#cb438-78" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>2×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw20</td>
<td style="text-align: right;">27354</td>
<td style="text-align: right;">721651</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw20</td>
<td style="text-align: right;">5852254</td>
<td style="text-align: right;">6671670</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-161-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-161-output-6.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="save-the-individual-colors-in-the-metadata-12" class="level3">
<h3 class="anchored" data-anchor-id="save-the-individual-colors-in-the-metadata-12">Save the individual colors in the metadata</h3>
<div id="dc0c94bc" class="cell" data-execution_count="161">
<div class="sourceCode cell-code" id="cb442"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb442-1"><a href="#cb442-1" aria-hidden="true" tabindex="-1"></a>indColors <span class="op">=</span> <span class="fu">fill</span>(<span class="st">""</span>, <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))</span>
<span id="cb442-2"><a href="#cb442-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">axes</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb442-3"><a href="#cb442-3" aria-hidden="true" tabindex="-1"></a>    indColors[i] <span class="op">=</span> group_colors_PCA[<span class="fu">findfirst</span>(groups_to_plot_PCA <span class="op">.==</span> PCAmodelAll.metadata.Fst_group[i])]</span>
<span id="cb442-4"><a href="#cb442-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb442-5"><a href="#cb442-5" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorLeft <span class="op">=</span> indColors</span>
<span id="cb442-6"><a href="#cb442-6" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.indColorRight <span class="op">=</span> indColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="162">
<pre><code>257-element Vector{String}:
 "blue"
 "red"
 "red"
 "red"
 "red"
 "red"
 "red"
 "purple"
 "red"
 "red"
 "red"
 "blue"
 "red"
 ⋮
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"
 "olivedrab2"</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc2-12" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc2-12">Plot PC1 vs.&nbsp;PC2:</h3>
<div id="49d0d69b" class="cell" data-execution_count="162">
<div class="sourceCode cell-code" id="cb444"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb444-1"><a href="#cb444-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb444-2"><a href="#cb444-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb444-3"><a href="#cb444-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb444-4"><a href="#cb444-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb444-5"><a href="#cb444-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb444-6"><a href="#cb444-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb444-7"><a href="#cb444-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb444-8"><a href="#cb444-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb444-9"><a href="#cb444-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb444-10"><a href="#cb444-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb444-11"><a href="#cb444-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb444-12"><a href="#cb444-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-163-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="163">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="plot-pc1-vs.-pc3-12" class="level3">
<h3 class="anchored" data-anchor-id="plot-pc1-vs.-pc3-12">Plot PC1 vs.&nbsp;PC3:</h3>
<div id="65e331af" class="cell" data-execution_count="163">
<div class="sourceCode cell-code" id="cb447"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb447-1"><a href="#cb447-1" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb447-2"><a href="#cb447-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb447-3"><a href="#cb447-3" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC3"</span>,</span>
<span id="cb447-4"><a href="#cb447-4" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb447-5"><a href="#cb447-5" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC3"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb447-6"><a href="#cb447-6" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb447-7"><a href="#cb447-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb447-8"><a href="#cb447-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb447-9"><a href="#cb447-9" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb447-10"><a href="#cb447-10" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC3[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb447-11"><a href="#cb447-11" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb447-12"><a href="#cb447-12" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-164-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="164">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="at-chr-20-high-hetvar-region-there-are-6-clear-haplogroups-with-vir-and-lud-clearly-different-on-pc3.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between" class="level3">
<h3 class="anchored" data-anchor-id="at-chr-20-high-hetvar-region-there-are-6-clear-haplogroups-with-vir-and-lud-clearly-different-on-pc3.-divide-samples-into-those-groups-based-on-pca-scores-and-then-calculate-pi-within-and-pi-between">At chr 20 high HetVar region, there are 6 clear haplogroups (with vir and lud clearly different on PC3). Divide samples into those groups, based on PCA scores, and then calculate pi within and pi between</h3>
<div id="442d2d10" class="cell" data-execution_count="164">
<div class="sourceCode cell-code" id="cb450"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb450-1"><a href="#cb450-1" aria-hidden="true" tabindex="-1"></a>clusterNames <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb450-2"><a href="#cb450-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb450-3"><a href="#cb450-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb450-4"><a href="#cb450-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb450-5"><a href="#cb450-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb450-6"><a href="#cb450-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>]</span>
<span id="cb450-7"><a href="#cb450-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-8"><a href="#cb450-8" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb450-9"><a href="#cb450-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb450-10"><a href="#cb450-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb450-11"><a href="#cb450-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb450-12"><a href="#cb450-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb450-13"><a href="#cb450-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>]</span>
<span id="cb450-14"><a href="#cb450-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-15"><a href="#cb450-15" aria-hidden="true" tabindex="-1"></a>vir <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb450-16"><a href="#cb450-16" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">0.5</span>) <span class="op">.&amp;</span></span>
<span id="cb450-17"><a href="#cb450-17" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-18"><a href="#cb450-18" aria-hidden="true" tabindex="-1"></a>nit <span class="op">=</span> (<span class="op">-</span><span class="fl">5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">3</span>) <span class="op">.&amp;</span></span>
<span id="cb450-19"><a href="#cb450-19" aria-hidden="true" tabindex="-1"></a>        (PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb450-20"><a href="#cb450-20" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-21"><a href="#cb450-21" aria-hidden="true" tabindex="-1"></a>lud <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb450-22"><a href="#cb450-22" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb450-23"><a href="#cb450-23" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-24"><a href="#cb450-24" aria-hidden="true" tabindex="-1"></a>troch <span class="op">=</span> (<span class="fl">2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb450-25"><a href="#cb450-25" aria-hidden="true" tabindex="-1"></a>            (PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4.2</span>) <span class="op">.&amp;</span></span>
<span id="cb450-26"><a href="#cb450-26" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-27"><a href="#cb450-27" aria-hidden="true" tabindex="-1"></a>obs <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb450-28"><a href="#cb450-28" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">4.2</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">2.5</span>) <span class="op">.&amp;</span></span>
<span id="cb450-29"><a href="#cb450-29" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-30"><a href="#cb450-30" aria-hidden="true" tabindex="-1"></a>plumb <span class="op">=</span> (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">6</span>) <span class="op">.&amp;</span> </span>
<span id="cb450-31"><a href="#cb450-31" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2) <span class="op">.&amp;</span></span>
<span id="cb450-32"><a href="#cb450-32" aria-hidden="true" tabindex="-1"></a>            indSelection_lowIndHetStan</span>
<span id="cb450-33"><a href="#cb450-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-34"><a href="#cb450-34" aria-hidden="true" tabindex="-1"></a><span class="co"># check the individuals in each group</span></span>
<span id="cb450-35"><a href="#cb450-35" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[vir]</span>
<span id="cb450-36"><a href="#cb450-36" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[nit]</span>
<span id="cb450-37"><a href="#cb450-37" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[lud]</span>
<span id="cb450-38"><a href="#cb450-38" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[troch]</span>
<span id="cb450-39"><a href="#cb450-39" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[obs]</span>
<span id="cb450-40"><a href="#cb450-40" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group[plumb]</span>
<span id="cb450-41"><a href="#cb450-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-42"><a href="#cb450-42" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir nit lud troch obs plumb]</span>
<span id="cb450-43"><a href="#cb450-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-44"><a href="#cb450-44" aria-hidden="true" tabindex="-1"></a><span class="co"># show numbers in each group</span></span>
<span id="cb450-45"><a href="#cb450-45" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"The numbers in each group are </span><span class="sc">$</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>))<span class="st"> and the sum of those is </span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">"</span>)</span>
<span id="cb450-46"><a href="#cb450-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-47"><a href="#cb450-47" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb450-48"><a href="#cb450-48" aria-hidden="true" tabindex="-1"></a>clusterMembership <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb450-49"><a href="#cb450-49" aria-hidden="true" tabindex="-1"></a>plotOrder <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb450-50"><a href="#cb450-50" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb450-51"><a href="#cb450-51" aria-hidden="true" tabindex="-1"></a>    clusterMembership[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNames[i]</span>
<span id="cb450-52"><a href="#cb450-52" aria-hidden="true" tabindex="-1"></a>    plotOrder[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb450-53"><a href="#cb450-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb450-54"><a href="#cb450-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-55"><a href="#cb450-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb450-56"><a href="#cb450-56" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_highHetVarRegion, clusterMembership, clusterNames)</span>
<span id="cb450-57"><a href="#cb450-57" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb450-58"><a href="#cb450-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-59"><a href="#cb450-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb450-60"><a href="#cb450-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-61"><a href="#cb450-61" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb450-62"><a href="#cb450-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-63"><a href="#cb450-63" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb450-64"><a href="#cb450-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-65"><a href="#cb450-65" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames)</span>
<span id="cb450-66"><a href="#cb450-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-67"><a href="#cb450-67" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames; among<span class="op">=</span><span class="cn">false</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb450-68"><a href="#cb450-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-69"><a href="#cb450-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb450-70"><a href="#cb450-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-71"><a href="#cb450-71" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb450-72"><a href="#cb450-72" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb450-73"><a href="#cb450-73" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster  pi         </span></span>
<span id="cb450-74"><a href="#cb450-74" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String   Float64    </span></span>
<span id="cb450-75"><a href="#cb450-75" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────</span></span>
<span id="cb450-76"><a href="#cb450-76" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir      0.0132903</span></span>
<span id="cb450-77"><a href="#cb450-77" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit      0.00761773</span></span>
<span id="cb450-78"><a href="#cb450-78" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ lud      0.014873</span></span>
<span id="cb450-79"><a href="#cb450-79" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ troch    0.0101873</span></span>
<span id="cb450-80"><a href="#cb450-80" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ obs      0.00904222</span></span>
<span id="cb450-81"><a href="#cb450-81" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ plumb    0.00593251 =#</span></span>
<span id="cb450-82"><a href="#cb450-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-83"><a href="#cb450-83" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb450-84"><a href="#cb450-84" aria-hidden="true" tabindex="-1"></a><span class="co">#= 15×2 DataFrame</span></span>
<span id="cb450-85"><a href="#cb450-85" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair  Dxy       </span></span>
<span id="cb450-86"><a href="#cb450-86" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String        Float64   </span></span>
<span id="cb450-87"><a href="#cb450-87" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────</span></span>
<span id="cb450-88"><a href="#cb450-88" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ vir_nit       0.0280243</span></span>
<span id="cb450-89"><a href="#cb450-89" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ vir_lud       0.0204941</span></span>
<span id="cb450-90"><a href="#cb450-90" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ vir_troch     0.0394257</span></span>
<span id="cb450-91"><a href="#cb450-91" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ vir_obs       0.0403572</span></span>
<span id="cb450-92"><a href="#cb450-92" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ vir_plumb     0.0376188</span></span>
<span id="cb450-93"><a href="#cb450-93" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ nit_lud       0.0288021</span></span>
<span id="cb450-94"><a href="#cb450-94" aria-hidden="true" tabindex="-1"></a><span class="co">   7 │ nit_troch     0.0377964</span></span>
<span id="cb450-95"><a href="#cb450-95" aria-hidden="true" tabindex="-1"></a><span class="co">   8 │ nit_obs       0.0389254</span></span>
<span id="cb450-96"><a href="#cb450-96" aria-hidden="true" tabindex="-1"></a><span class="co">   9 │ nit_plumb     0.0359742</span></span>
<span id="cb450-97"><a href="#cb450-97" aria-hidden="true" tabindex="-1"></a><span class="co">  10 │ lud_troch     0.0390498</span></span>
<span id="cb450-98"><a href="#cb450-98" aria-hidden="true" tabindex="-1"></a><span class="co">  11 │ lud_obs       0.0398045</span></span>
<span id="cb450-99"><a href="#cb450-99" aria-hidden="true" tabindex="-1"></a><span class="co">  12 │ lud_plumb     0.0371989</span></span>
<span id="cb450-100"><a href="#cb450-100" aria-hidden="true" tabindex="-1"></a><span class="co">  13 │ troch_obs     0.015702</span></span>
<span id="cb450-101"><a href="#cb450-101" aria-hidden="true" tabindex="-1"></a><span class="co">  14 │ troch_plumb   0.0285113</span></span>
<span id="cb450-102"><a href="#cb450-102" aria-hidden="true" tabindex="-1"></a><span class="co">  15 │ obs_plumb     0.0286543 =#</span></span>
<span id="cb450-103"><a href="#cb450-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-104"><a href="#cb450-104" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a genotype-by-individual plot using all variable loci in the region,</span></span>
<span id="cb450-105"><a href="#cb450-105" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb450-106"><a href="#cb450-106" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb450-107"><a href="#cb450-107" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_Fst_groups <span class="op">=</span> PCAmodelAll.metadata.Fst_group <span class="co"># store the Fst_groups in this</span></span>
<span id="cb450-108"><a href="#cb450-108" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembership</span>
<span id="cb450-109"><a href="#cb450-109" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.original_plot_order <span class="op">=</span> PCAmodelAll.metadata.plot_order <span class="co"># store the original plot_order in this</span></span>
<span id="cb450-110"><a href="#cb450-110" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrder</span>
<span id="cb450-111"><a href="#cb450-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-112"><a href="#cb450-112" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb450-113"><a href="#cb450-113" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb450-114"><a href="#cb450-114" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb450-115"><a href="#cb450-115" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb450-116"><a href="#cb450-116" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb450-117"><a href="#cb450-117" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs <span class="op">=</span> genos_highHetVarRegion[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb450-118"><a href="#cb450-118" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs <span class="op">=</span> pos_highHetVarRegion[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb450-119"><a href="#cb450-119" aria-hidden="true" tabindex="-1"></a>Fst_selectedSNPs <span class="op">=</span> Fst[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb450-120"><a href="#cb450-120" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb450-121"><a href="#cb450-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-122"><a href="#cb450-122" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb450-123"><a href="#cb450-123" aria-hidden="true" tabindex="-1"></a>numIndsToPlot <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNames))</span>
<span id="cb450-124"><a href="#cb450-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-125"><a href="#cb450-125" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNames, numIndsToPlot, </span>
<span id="cb450-126"><a href="#cb450-126" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb450-127"><a href="#cb450-127" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb450-128"><a href="#cb450-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-129"><a href="#cb450-129" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb450-130"><a href="#cb450-130" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNames, clusterColors;</span>
<span id="cb450-131"><a href="#cb450-131" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb450-132"><a href="#cb450-132" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The numbers in each group are [38 2 29 68 4 66] and the sum of those is 207
Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-165-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="165">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 0 0; 2 2 … 0 0], [5852400, 5876520, 5876593, 5882337, 5882559, 5893275, 5901011, 5923576, 5923618, 5932214  …  6525018, 6533933, 6538226, 6559702, 6559912, 6582488, 6582518, 6594141, 6611772, 6618858], <span class="ansi-bold">66×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   4 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   5 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   6 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   7 │ GW_Lane5_AA3               GW_Lane5_AA3               AA        vir_S
   8 │ GW_Lane5_AA5               GW_Lane5_AA5               AA        vir_S
   9 │ GW_Lane5_AA9               GW_Lane5_AA9               AA        vir_S   ⋯
  10 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
  11 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  57 │ GW_Armando_plate1_JF11G01  GW_Armando_plate1_JF11G01  ST        plumb
  58 │ GW_Armando_plate1_JF12G02  GW_Armando_plate1_JF12G02  ST        plumb   ⋯
  59 │ GW_Armando_plate1_JF13G01  GW_Armando_plate1_JF13G01  ST        plumb
  60 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  61 │ GW_Armando_plate1_JF20G01  GW_Armando_plate1_JF20G01  MB        plumb
  62 │ GW_Armando_plate1_JF22G01  GW_Armando_plate1_JF22G01  MB        plumb   ⋯
  63 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  64 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  65 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb
  66 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  39 columns and 45 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-12" class="level3">
<h3 class="anchored" data-anchor-id="now-show-a-gbi-plot-like-above-but-with-heterozygotes-12">Now show a GBI plot like above, but with heterozygotes:</h3>
<div id="431ce76d" class="cell" data-execution_count="165">
<div class="sourceCode cell-code" id="cb453"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb453-1"><a href="#cb453-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHets <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb453-2"><a href="#cb453-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb453-3"><a href="#cb453-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb453-4"><a href="#cb453-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"ludHet"</span>,</span>
<span id="cb453-5"><a href="#cb453-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud_troch"</span>,</span>
<span id="cb453-6"><a href="#cb453-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb453-7"><a href="#cb453-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span>,</span>
<span id="cb453-8"><a href="#cb453-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs_plumb"</span>,</span>
<span id="cb453-9"><a href="#cb453-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span>,</span>
<span id="cb453-10"><a href="#cb453-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"vir_plumb"</span>]</span>
<span id="cb453-11"><a href="#cb453-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-12"><a href="#cb453-12" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHets <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb453-13"><a href="#cb453-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"grey"</span>,</span>
<span id="cb453-14"><a href="#cb453-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb453-15"><a href="#cb453-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb453-16"><a href="#cb453-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"seagreen"</span>,</span>
<span id="cb453-17"><a href="#cb453-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>,</span>
<span id="cb453-18"><a href="#cb453-18" aria-hidden="true" tabindex="-1"></a>                <span class="st">"orange"</span>,</span>
<span id="cb453-19"><a href="#cb453-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">"darkorange1"</span>,</span>
<span id="cb453-20"><a href="#cb453-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">"red"</span>,</span>
<span id="cb453-21"><a href="#cb453-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">"purple"</span>]</span>
<span id="cb453-22"><a href="#cb453-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-23"><a href="#cb453-23" aria-hidden="true" tabindex="-1"></a>ludHet <span class="op">=</span> (PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">4</span>) <span class="op">.&amp;</span></span>
<span id="cb453-24"><a href="#cb453-24" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC3 <span class="op">.&lt;</span> <span class="fl">7</span>) <span class="op">.&amp;</span></span>
<span id="cb453-25"><a href="#cb453-25" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb453-26"><a href="#cb453-26" aria-hidden="true" tabindex="-1"></a>lud_troch <span class="op">=</span> (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb453-27"><a href="#cb453-27" aria-hidden="true" tabindex="-1"></a>            (<span class="op">-</span><span class="fl">4</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb453-28"><a href="#cb453-28" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb453-29"><a href="#cb453-29" aria-hidden="true" tabindex="-1"></a>obs_plumb <span class="op">=</span> (<span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb453-30"><a href="#cb453-30" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">0</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">1</span>) <span class="op">.&amp;</span></span>
<span id="cb453-31"><a href="#cb453-31" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb453-32"><a href="#cb453-32" aria-hidden="true" tabindex="-1"></a>vir_plumb <span class="op">=</span> (<span class="op">-</span><span class="fl">3</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC1 <span class="op">.&lt;</span> <span class="fl">0</span>) <span class="op">.&amp;</span></span>
<span id="cb453-33"><a href="#cb453-33" aria-hidden="true" tabindex="-1"></a>            (<span class="fl">2.5</span> <span class="op">.&lt;</span> PCAmodelAll.metadata.PC2 <span class="op">.&lt;</span> <span class="fl">5</span>) <span class="op">.&amp;</span></span>
<span id="cb453-34"><a href="#cb453-34" aria-hidden="true" tabindex="-1"></a>            .!indSelection_lowIndHetStan</span>
<span id="cb453-35"><a href="#cb453-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-36"><a href="#cb453-36" aria-hidden="true" tabindex="-1"></a>clusterArray <span class="op">=</span> [vir nit lud ludHet lud_troch troch obs obs_plumb plumb vir_plumb]</span>
<span id="cb453-37"><a href="#cb453-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-38"><a href="#cb453-38" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)</span>
<span id="cb453-39"><a href="#cb453-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-40"><a href="#cb453-40" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">==</span> <span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>)</span>
<span id="cb453-41"><a href="#cb453-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Good news: Individuals included in a group matches total number of individuals"</span>)</span>
<span id="cb453-42"><a href="#cb453-42" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb453-43"><a href="#cb453-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Warning: Individuals included in a group (</span><span class="sc">$</span>(<span class="fu">sum</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">1</span>)))<span class="st">) do NOT match total number of individuals (</span><span class="sc">$</span>(<span class="fu">size</span>(PCAmodelAll.metadata, <span class="fl">1</span>))<span class="st">)"</span>)</span>
<span id="cb453-44"><a href="#cb453-44" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb453-45"><a href="#cb453-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-46"><a href="#cb453-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-47"><a href="#cb453-47" aria-hidden="true" tabindex="-1"></a><span class="co"># check which individuals left out:</span></span>
<span id="cb453-48"><a href="#cb453-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-49"><a href="#cb453-49" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>)</span>
<span id="cb453-50"><a href="#cb453-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-51"><a href="#cb453-51" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.ind[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb453-52"><a href="#cb453-52" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC1[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb453-53"><a href="#cb453-53" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC2[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb453-54"><a href="#cb453-54" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan[<span class="fu">vec</span>(<span class="fu">sum</span>(clusterArray, dims<span class="op">=</span><span class="fl">2</span>) <span class="op">.==</span> <span class="fl">0</span>)]</span>
<span id="cb453-55"><a href="#cb453-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-56"><a href="#cb453-56" aria-hidden="true" tabindex="-1"></a><span class="co"># create vectors that indicate the groups and plot order for this analysis:</span></span>
<span id="cb453-57"><a href="#cb453-57" aria-hidden="true" tabindex="-1"></a>clusterMembershipWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="st">"none"</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb453-58"><a href="#cb453-58" aria-hidden="true" tabindex="-1"></a>plotOrderWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">9</span>, <span class="fu">nrow</span>(PCAmodelAll.metadata))</span>
<span id="cb453-59"><a href="#cb453-59" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(clusterArray[<span class="fl">1</span>,<span class="op">:</span>])</span>
<span id="cb453-60"><a href="#cb453-60" aria-hidden="true" tabindex="-1"></a>    clusterMembershipWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> clusterNamesWithHets[i]</span>
<span id="cb453-61"><a href="#cb453-61" aria-hidden="true" tabindex="-1"></a>    plotOrderWithHets[clusterArray[<span class="op">:</span>,i]] <span class="op">.=</span> i</span>
<span id="cb453-62"><a href="#cb453-62" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb453-63"><a href="#cb453-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-64"><a href="#cb453-64" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to main metadata object containing the cluster membership for this highHet region:</span></span>
<span id="cb453-65"><a href="#cb453-65" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_cluster = clusterMembershipWithHets"</span></span>
<span id="cb453-66"><a href="#cb453-66" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb453-67"><a href="#cb453-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-68"><a href="#cb453-68" aria-hidden="true" tabindex="-1"></a><span class="co"># in metadata, replace `Fst_group` column with cluster info (needed for the function below):</span></span>
<span id="cb453-69"><a href="#cb453-69" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.Fst_group <span class="op">=</span> clusterMembershipWithHets</span>
<span id="cb453-70"><a href="#cb453-70" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> plotOrderWithHets</span>
<span id="cb453-71"><a href="#cb453-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-72"><a href="#cb453-72" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the number of individuals per group to plot</span></span>
<span id="cb453-73"><a href="#cb453-73" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">15</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb453-74"><a href="#cb453-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-75"><a href="#cb453-75" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb453-76"><a href="#cb453-76" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb453-77"><a href="#cb453-77" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb453-78"><a href="#cb453-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb453-79"><a href="#cb453-79" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb453-80"><a href="#cb453-80" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb453-81"><a href="#cb453-81" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb453-82"><a href="#cb453-82" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb453-83"><a href="#cb453-83" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: Individuals included in a group matches total number of individuals</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-166-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="166">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 0 0; 1 1 … 0 0], [5852400, 5876520, 5876593, 5882337, 5882559, 5893275, 5901011, 5923576, 5923618, 5932214  …  6525018, 6533933, 6538226, 6559702, 6559912, 6582488, 6582518, 6594141, 6611772, 6618858], <span class="ansi-bold">91×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir     ⋯
   2 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir
   3 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   4 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   5 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis ⋯
   6 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   7 │ GW_Lane5_AA3               GW_Lane5_AA3               AA        vir_S
   8 │ GW_Lane5_AA5               GW_Lane5_AA5               AA        vir_S
   9 │ GW_Lane5_AA9               GW_Lane5_AA9               AA        vir_S   ⋯
  10 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
  11 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  82 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  83 │ GW_Armando_plate1_JF26G01  GW_Armando_plate1_JF26G01  ST        plumb   ⋯
  84 │ GW_Armando_plate1_JF27G01  GW_Armando_plate1_JF27G01  ST        plumb
  85 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  86 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  87 │ GW_Armando_plate2_JF14G02  GW_Armando_plate2_JF14G02  DV        plumb   ⋯
  88 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v
  89 │ GW_Armando_plate2_JF15G01  GW_Armando_plate2_JF15G01  DV        plumb
  90 │ GW_Armando_plate2_JF24G03  GW_Armando_plate2_JF24G03  ST        plumb
  91 │ GW_Armando_plate2_JF26G02  GW_Armando_plate2_JF26G02  ST        plumb   ⋯
<span class="ansi-cyan-fg">                                                  39 columns and 70 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-gbi-plot-according-to-original-groups-and-plot-order-8" class="level3">
<h3 class="anchored" data-anchor-id="show-gbi-plot-according-to-original-groups-and-plot-order-8">show GBI plot according to original groups and plot order</h3>
<div id="3a8560b9" class="cell" data-execution_count="166">
<div class="sourceCode cell-code" id="cb456"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb456-1"><a href="#cb456-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb456-2"><a href="#cb456-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb456-3"><a href="#cb456-3" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb456-4"><a href="#cb456-4" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb456-5"><a href="#cb456-5" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb456-6"><a href="#cb456-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb456-7"><a href="#cb456-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb456-8"><a href="#cb456-8" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb456-9"><a href="#cb456-9" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb456-10"><a href="#cb456-10" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb456-11"><a href="#cb456-11" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-167-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="167">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 1 1 … 0 0; 1 1 … 0 0], [5852400, 5876520, 5876593, 5882337, 5882559, 5893275, 5901011, 5923576, 5923618, 5932214  …  6525018, 6533933, 6538226, 6559702, 6559912, 6582488, 6582518, 6594141, 6611772, 6618858], <span class="ansi-bold">91×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   5 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v ⋯
   6 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis
   7 │ GW_Armando_plate1_AB1      GW_Armando_plate1_AB1      AB        vir
   8 │ GW_Lane5_TL10              GW_Lane5_TL10              TL        vir
   9 │ GW_Lane5_TL11              GW_Lane5_TL11              TL        vir     ⋯
  10 │ GW_Lane5_TL9               GW_Lane5_TL9               TL        vir
  11 │ GW_Armando_plate1_TL3      GW_Armando_plate1_TL3      TL        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  82 │ GW_Armando_plate1_JF12G01  GW_Armando_plate1_JF12G01  ST        plumb
  83 │ GW_Armando_plate2_JF24G03  GW_Armando_plate2_JF24G03  ST        plumb   ⋯
  84 │ GW_Armando_plate2_JF26G02  GW_Armando_plate2_JF26G02  ST        plumb
  85 │ GW_Armando_plate1_JF15G03  GW_Armando_plate1_JF15G03  DV        plumb
  86 │ GW_Armando_plate2_JF14G02  GW_Armando_plate2_JF14G02  DV        plumb
  87 │ GW_Armando_plate2_JF15G01  GW_Armando_plate2_JF15G01  DV        plumb   ⋯
  88 │ GW_Armando_plate1_JF23G01  GW_Armando_plate1_JF23G01  VB        plumb
  89 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
  90 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  91 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  39 columns and 70 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-same-but-with-all-individuals-8" class="level3">
<h3 class="anchored" data-anchor-id="show-same-but-with-all-individuals-8">show same but with all individuals</h3>
<div id="9409a22d" class="cell" data-execution_count="167">
<div class="sourceCode cell-code" id="cb458"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb458-1"><a href="#cb458-1" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.plot_order <span class="op">=</span> PCAmodelAll.metadata.original_plot_order</span>
<span id="cb458-2"><a href="#cb458-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb458-3"><a href="#cb458-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set no limit (or high limit anyway) on the number of individuals per group to plot</span></span>
<span id="cb458-4"><a href="#cb458-4" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">1000</span>, <span class="fu">length</span>(clusterNamesWithHets))</span>
<span id="cb458-5"><a href="#cb458-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb458-6"><a href="#cb458-6" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHets, numIndsToPlotWithHets, </span>
<span id="cb458-7"><a href="#cb458-7" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs, PCAmodelAll.metadata;</span>
<span id="cb458-8"><a href="#cb458-8" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb458-9"><a href="#cb458-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb458-10"><a href="#cb458-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs,</span>
<span id="cb458-11"><a href="#cb458-11" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs, clusterNamesWithHets, clusterColorsWithHets;</span>
<span id="cb458-12"><a href="#cb458-12" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb458-13"><a href="#cb458-13" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb458-14"><a href="#cb458-14" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-168-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="168">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 0 0; 1 1 … 0 0], [5852400, 5876520, 5876593, 5882337, 5882559, 5893275, 5901011, 5923576, 5923618, 5932214  …  6525018, 6533933, 6538226, 6559702, 6559912, 6582488, 6582518, 6594141, 6611772, 6618858], <span class="ansi-bold">257×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK4               GW_Lane5_YK4               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 249 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 252 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 253 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 39 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-west-area-without-nitidus-9" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-west-area-without-nitidus-9">Show just the west area (without nitidus)</h3>
<div id="f3908769" class="cell" data-execution_count="168">
<div class="sourceCode cell-code" id="cb460"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb460-1"><a href="#cb460-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsWest <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb460-2"><a href="#cb460-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span>,</span>
<span id="cb460-3"><a href="#cb460-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"ludHet"</span>,</span>
<span id="cb460-4"><a href="#cb460-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud_troch"</span>,</span>
<span id="cb460-5"><a href="#cb460-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>]</span>
<span id="cb460-6"><a href="#cb460-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb460-7"><a href="#cb460-7" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsWest <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb460-8"><a href="#cb460-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb460-9"><a href="#cb460-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"green"</span>,</span>
<span id="cb460-10"><a href="#cb460-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"seagreen"</span>,</span>
<span id="cb460-11"><a href="#cb460-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"yellow"</span>]</span>
<span id="cb460-12"><a href="#cb460-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb460-13"><a href="#cb460-13" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb460-14"><a href="#cb460-14" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb460-15"><a href="#cb460-15" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb460-16"><a href="#cb460-16" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb460-17"><a href="#cb460-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb460-18"><a href="#cb460-18" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsWest)</span>
<span id="cb460-19"><a href="#cb460-19" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb460-20"><a href="#cb460-20" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb460-21"><a href="#cb460-21" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb460-22"><a href="#cb460-22" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb460-23"><a href="#cb460-23" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb460-24"><a href="#cb460-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb460-25"><a href="#cb460-25" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsWest))</span>
<span id="cb460-26"><a href="#cb460-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb460-27"><a href="#cb460-27" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsWest, numIndsToPlotWithHets, </span>
<span id="cb460-28"><a href="#cb460-28" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb460-29"><a href="#cb460-29" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb460-30"><a href="#cb460-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb460-31"><a href="#cb460-31" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb460-32"><a href="#cb460-32" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsWest, clusterColorsWithHetsWest;</span>
<span id="cb460-33"><a href="#cb460-33" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb460-34"><a href="#cb460-34" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb460-35"><a href="#cb460-35" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-169-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="169">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], [5852400, 5876593, 5882559, 5901011, 5923576, 5923618, 5945810, 5966423, 5970604, 5986806  …  6459128, 6459167, 6465882, 6465908, 6472163, 6472190, 6538226, 6582488, 6594141, 6611772], <span class="ansi-bold">177×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 168 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
 169 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L ⋯
 170 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
 171 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L
 172 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 173 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L ⋯
 174 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L
 175 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L
 176 │ GW_Lane5_LN10              GW_Lane5_LN10              LN        troch_L
 177 │ GW_Lane5_LN12              GW_Lane5_LN12              LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 39 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-east-area-10" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-east-area-10">Show just the east area</h3>
<div id="7d988e6b" class="cell" data-execution_count="169">
<div class="sourceCode cell-code" id="cb463"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb463-1"><a href="#cb463-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsEast <span class="op">=</span> [<span class="st">"troch"</span>,</span>
<span id="cb463-2"><a href="#cb463-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs"</span>,</span>
<span id="cb463-3"><a href="#cb463-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"obs_plumb"</span>,</span>
<span id="cb463-4"><a href="#cb463-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb463-5"><a href="#cb463-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb463-6"><a href="#cb463-6" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsEast <span class="op">=</span> [<span class="st">"yellow"</span>,</span>
<span id="cb463-7"><a href="#cb463-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"orange"</span>,</span>
<span id="cb463-8"><a href="#cb463-8" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"darkorange1"</span>,</span>
<span id="cb463-9"><a href="#cb463-9" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb463-10"><a href="#cb463-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb463-11"><a href="#cb463-11" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb463-12"><a href="#cb463-12" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb463-13"><a href="#cb463-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb463-14"><a href="#cb463-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb463-15"><a href="#cb463-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb463-16"><a href="#cb463-16" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsEast)</span>
<span id="cb463-17"><a href="#cb463-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb463-18"><a href="#cb463-18" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb463-19"><a href="#cb463-19" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb463-20"><a href="#cb463-20" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb463-21"><a href="#cb463-21" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb463-22"><a href="#cb463-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb463-23"><a href="#cb463-23" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHetsEast <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsEast))</span>
<span id="cb463-24"><a href="#cb463-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb463-25"><a href="#cb463-25" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsEast, numIndsToPlotWithHetsEast, </span>
<span id="cb463-26"><a href="#cb463-26" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb463-27"><a href="#cb463-27" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb463-28"><a href="#cb463-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb463-29"><a href="#cb463-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb463-30"><a href="#cb463-30" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsEast, clusterColorsWithHetsEast;</span>
<span id="cb463-31"><a href="#cb463-31" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb463-32"><a href="#cb463-32" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb463-33"><a href="#cb463-33" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-170-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="170">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 1; 0 0 … 2 0; … ; 2 2 … 0 2; 2 2 … 2 2], [5876520, 5882337, 5893275, 5923576, 5923618, 5932416, 5945810, 5966423, 5989088, 6012800  …  6433886, 6433927, 6498576, 6525018, 6533933, 6559702, 6559912, 6582488, 6582518, 6611772], <span class="ansi-bold">139×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_TTGW70   GW_Armando_plate1_TTGW70   SA        lud_Sat ⋯
   2 │ GW_Armando_plate1_TTGW72   GW_Armando_plate1_TTGW72   SA        lud_Sat
   3 │ GW_Armando_plate1_TTGW78   GW_Armando_plate1_TTGW78   SA        lud_Sat
   4 │ GW_Armando_plate2_TTGW73   GW_Armando_plate2_TTGW73   SA        lud_Sat
   5 │ GW_Armando_plate2_TTGW75   GW_Armando_plate2_TTGW75   SA        lud_Sat ⋯
   6 │ GW_Armando_plate2_TTGW80   GW_Armando_plate2_TTGW80   SA        lud_Sat
   7 │ GW_Armando_plate1_TTGW11   GW_Armando_plate1_TTGW11   SU        lud_Suk
   8 │ GW_Armando_plate2_TTGW09   GW_Armando_plate2_TTGW09   SU        lud_Suk
   9 │ GW_Liz_GBS_Liz10045        GW_Liz_GBS_Liz10045        ML        lud     ⋯
  10 │ GW_Liz_GBS_Liz10094        GW_Liz_GBS_Liz10094        ML        lud
  11 │ GW_Liz_GBS_Liz5163         GW_Liz_GBS_Liz5163         ML        lud_chi
  ⋮  │             ⋮                          ⋮                 ⋮            ⋮ ⋱
 130 │ GW_Armando_plate1_JF23G02  GW_Armando_plate1_JF23G02  VB        plumb
 131 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb   ⋯
 132 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
 133 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 134 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 135 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 136 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 137 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 138 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 139 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb   ⋯
<span class="ansi-cyan-fg">                                                 39 columns and 118 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="show-just-the-northern-area-10" class="level3">
<h3 class="anchored" data-anchor-id="show-just-the-northern-area-10">Show just the northern area</h3>
<div id="774e024f" class="cell" data-execution_count="170">
<div class="sourceCode cell-code" id="cb466"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb466-1"><a href="#cb466-1" aria-hidden="true" tabindex="-1"></a>clusterNamesWithHetsNorth <span class="op">=</span> [<span class="st">"vir"</span>,</span>
<span id="cb466-2"><a href="#cb466-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"vir_plumb"</span>,</span>
<span id="cb466-3"><a href="#cb466-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"plumb"</span>]</span>
<span id="cb466-4"><a href="#cb466-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb466-5"><a href="#cb466-5" aria-hidden="true" tabindex="-1"></a>clusterColorsWithHetsNorth <span class="op">=</span> [<span class="st">"blue"</span>,</span>
<span id="cb466-6"><a href="#cb466-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"purple"</span>,</span>
<span id="cb466-7"><a href="#cb466-7" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"red"</span>]</span>
<span id="cb466-8"><a href="#cb466-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb466-9"><a href="#cb466-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit the SNPs to those with variants greater than 50% in </span></span>
<span id="cb466-10"><a href="#cb466-10" aria-hidden="true" tabindex="-1"></a><span class="co"># at least one pop, and less than 50% in at least one pop.</span></span>
<span id="cb466-11"><a href="#cb466-11" aria-hidden="true" tabindex="-1"></a><span class="co"># (So for each column in `freqs`, the maximum should be &gt; 0.5 </span></span>
<span id="cb466-12"><a href="#cb466-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and the minimum should be &lt; 0.5)</span></span>
<span id="cb466-13"><a href="#cb466-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb466-14"><a href="#cb466-14" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genos_selectedSNPs, clusterMembershipWithHets, clusterNamesWithHetsNorth)</span>
<span id="cb466-15"><a href="#cb466-15" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb466-16"><a href="#cb466-16" aria-hidden="true" tabindex="-1"></a>selectedSNPs <span class="op">=</span> (<span class="fu">vec</span>(<span class="fu">maximum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&gt;</span> <span class="fl">0.5</span>) <span class="op">.&amp;</span> (<span class="fu">vec</span>(<span class="fu">minimum</span>(freqs, dims<span class="op">=</span><span class="fl">1</span>)) <span class="op">.&lt;</span> <span class="fl">0.5</span>)</span>
<span id="cb466-17"><a href="#cb466-17" aria-hidden="true" tabindex="-1"></a>genos_selectedSNPs2 <span class="op">=</span> genos_selectedSNPs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb466-18"><a href="#cb466-18" aria-hidden="true" tabindex="-1"></a>pos_selectedSNPs2 <span class="op">=</span> pos_selectedSNPs[selectedSNPs, <span class="op">:</span>]</span>
<span id="cb466-19"><a href="#cb466-19" aria-hidden="true" tabindex="-1"></a>freqs_selectedSNPs2 <span class="op">=</span> freqs[<span class="op">:</span>, selectedSNPs]</span>
<span id="cb466-20"><a href="#cb466-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb466-21"><a href="#cb466-21" aria-hidden="true" tabindex="-1"></a>numIndsToPlotWithHets <span class="op">=</span> <span class="fu">fill</span>(<span class="fl">100</span>, <span class="fu">length</span>(clusterNamesWithHetsNorth))</span>
<span id="cb466-22"><a href="#cb466-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb466-23"><a href="#cb466-23" aria-hidden="true" tabindex="-1"></a>genosForGBI, indMetadataforGBI <span class="op">=</span> <span class="fu">limitIndsToPlot</span>(clusterNamesWithHetsNorth, numIndsToPlotWithHets, </span>
<span id="cb466-24"><a href="#cb466-24" aria-hidden="true" tabindex="-1"></a>                                            genos_selectedSNPs2, PCAmodelAll.metadata;</span>
<span id="cb466-25"><a href="#cb466-25" aria-hidden="true" tabindex="-1"></a>                                            sortByMissing <span class="op">=</span> <span class="cn">true</span>)</span>
<span id="cb466-26"><a href="#cb466-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb466-27"><a href="#cb466-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenotypeByIndividual</span>(regionInfo, pos_selectedSNPs2,</span>
<span id="cb466-28"><a href="#cb466-28" aria-hidden="true" tabindex="-1"></a>                genosForGBI, indMetadataforGBI, freqs_selectedSNPs2, clusterNamesWithHetsNorth, clusterColorsWithHetsNorth;</span>
<span id="cb466-29"><a href="#cb466-29" aria-hidden="true" tabindex="-1"></a>                missingFractionAllowed <span class="op">=</span> missingFractionAllowed,</span>
<span id="cb466-30"><a href="#cb466-30" aria-hidden="true" tabindex="-1"></a>                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb466-31"><a href="#cb466-31" aria-hidden="true" tabindex="-1"></a>                indColorRightProvided <span class="op">=</span> <span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-171-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="171">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  2 Child Scenes:
    ├ Scene (768px, 960px)
    └ Scene (768px, 960px), Union{Missing, Int16}[0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 1 … 1 2], [5852400, 5876520, 5876593, 5882337, 5882559, 5893275, 5901011, 5970604, 5986806, 5994647  …  6459167, 6465882, 6465908, 6472163, 6472190, 6533933, 6538226, 6559702, 6559912, 6594141], <span class="ansi-bold">111×42 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK7               GW_Lane5_YK7               YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 102 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 103 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 104 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 105 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 106 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
 107 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb   ⋯
 108 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 109 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 110 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 111 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  39 columns and 90 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
</section>
<section id="try-chr-27" class="level2">
<h2 class="anchored" data-anchor-id="try-chr-27">Try chr 27</h2>
<div id="129a8207" class="cell" data-execution_count="171">
<div class="sourceCode cell-code" id="cb469"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb469-1"><a href="#cb469-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose scaffold</span></span>
<span id="cb469-2"><a href="#cb469-2" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw27"</span></span>
<span id="cb469-3"><a href="#cb469-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-4"><a href="#cb469-4" aria-hidden="true" tabindex="-1"></a>positionMin, positionMax, regionText, </span>
<span id="cb469-5"><a href="#cb469-5" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb469-6"><a href="#cb469-6" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb469-7"><a href="#cb469-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb469-8"><a href="#cb469-8" aria-hidden="true" tabindex="-1"></a>                            pos_SNP_filtered, </span>
<span id="cb469-9"><a href="#cb469-9" aria-hidden="true" tabindex="-1"></a>                            highHetVarRegions, chr;</span>
<span id="cb469-10"><a href="#cb469-10" aria-hidden="true" tabindex="-1"></a>                            windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb469-11"><a href="#cb469-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-12"><a href="#cb469-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect values for mean IndHetStan per individual for that high HetVar region</span></span>
<span id="cb469-13"><a href="#cb469-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(meanAcrossRegionIndHetStan)</span>
<span id="cb469-14"><a href="#cb469-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-15"><a href="#cb469-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column to metadata containing the regionIndHetStan for this highHet region:</span></span>
<span id="cb469-16"><a href="#cb469-16" aria-hidden="true" tabindex="-1"></a>command <span class="op">=</span> <span class="st">"ind_with_metadata_included."</span> <span class="op">*</span> chr <span class="op">*</span> <span class="st">"_regionIndHetStan = meanAcrossRegionIndHetStan"</span></span>
<span id="cb469-17"><a href="#cb469-17" aria-hidden="true" tabindex="-1"></a><span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb469-18"><a href="#cb469-18" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included.regionIndHetStan <span class="op">=</span> meanAcrossRegionIndHetStan</span>
<span id="cb469-19"><a href="#cb469-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-20"><a href="#cb469-20" aria-hidden="true" tabindex="-1"></a><span class="co">#names(ind_with_metadata_included)</span></span>
<span id="cb469-21"><a href="#cb469-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-22"><a href="#cb469-22" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether missing data related to heterozygosity (good news: not really)</span></span>
<span id="cb469-23"><a href="#cb469-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ind_with_metadata_included.numMissings, meanAcrossRegionIndHetStan)</span>
<span id="cb469-24"><a href="#cb469-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-25"><a href="#cb469-25" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA of all individuals:</span></span>
<span id="cb469-26"><a href="#cb469-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-27"><a href="#cb469-27" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="op">=</span> Impute.<span class="fu">svd</span>(<span class="fu">Matrix</span><span class="dt">{Union{Missing, Float32}}</span>(genos_highHetVarRegion))</span>
<span id="cb469-28"><a href="#cb469-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-29"><a href="#cb469-29" aria-hidden="true" tabindex="-1"></a>flipPC1 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb469-30"><a href="#cb469-30" aria-hidden="true" tabindex="-1"></a>flipPC2 <span class="op">=</span> <span class="cn">true</span></span>
<span id="cb469-31"><a href="#cb469-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-32"><a href="#cb469-32" aria-hidden="true" tabindex="-1"></a>PCAmodelAll <span class="op">=</span> <span class="fu">plotPCA</span>(genos_highHetVarRegion_imputed, ind_with_metadata_included, </span>
<span id="cb469-33"><a href="#cb469-33" aria-hidden="true" tabindex="-1"></a>            groups_to_plot_PCA, group_colors_PCA; </span>
<span id="cb469-34"><a href="#cb469-34" aria-hidden="true" tabindex="-1"></a>            sampleSet <span class="op">=</span> <span class="st">"greenish warblers"</span>, regionText <span class="op">=</span> regionText,</span>
<span id="cb469-35"><a href="#cb469-35" aria-hidden="true" tabindex="-1"></a>            flip1 <span class="op">=</span> flipPC1, flip2 <span class="op">=</span> flipPC2,</span>
<span id="cb469-36"><a href="#cb469-36" aria-hidden="true" tabindex="-1"></a>            lineOpacity <span class="op">=</span> <span class="fl">0.7</span>, fillOpacity <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span id="cb469-37"><a href="#cb469-37" aria-hidden="true" tabindex="-1"></a>            symbolSize <span class="op">=</span> <span class="fl">14</span>, showTitle <span class="op">=</span> <span class="cn">true</span>,</span>
<span id="cb469-38"><a href="#cb469-38" aria-hidden="true" tabindex="-1"></a>            xLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC1"</span>), yLabelText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Region PC2"</span>),</span>
<span id="cb469-39"><a href="#cb469-39" aria-hidden="true" tabindex="-1"></a>            showPlot <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb469-40"><a href="#cb469-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-41"><a href="#cb469-41" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(PCAmodelAll.PCAfig)</span>
<span id="cb469-42"><a href="#cb469-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-43"><a href="#cb469-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC values to metadata for individuals included in PCA above:</span></span>
<span id="cb469-44"><a href="#cb469-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC1</span>
<span id="cb469-45"><a href="#cb469-45" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb469-46"><a href="#cb469-46" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span> </span>
<span id="cb469-47"><a href="#cb469-47" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC1 <span class="op">=</span> PCAmodelAll.values[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb469-48"><a href="#cb469-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb469-49"><a href="#cb469-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> flipPC2</span>
<span id="cb469-50"><a href="#cb469-50" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">.*</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb469-51"><a href="#cb469-51" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb469-52"><a href="#cb469-52" aria-hidden="true" tabindex="-1"></a>    PCAmodelAll.metadata.PC2 <span class="op">=</span> PCAmodelAll.values[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb469-53"><a href="#cb469-53" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb469-54"><a href="#cb469-54" aria-hidden="true" tabindex="-1"></a>PCAmodelAll.metadata.PC3 <span class="op">=</span> PCAmodelAll.values[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb469-55"><a href="#cb469-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-56"><a href="#cb469-56" aria-hidden="true" tabindex="-1"></a><span class="co"># For the next bit to work with above, make sure that all individuals in the above `plotPCA` command</span></span>
<span id="cb469-57"><a href="#cb469-57" aria-hidden="true" tabindex="-1"></a><span class="co"># are included in the `groups_to_plot_PCA`</span></span>
<span id="cb469-58"><a href="#cb469-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-59"><a href="#cb469-59" aria-hidden="true" tabindex="-1"></a><span class="co"># choose inds with low IndHet in high HetVar region:</span></span>
<span id="cb469-60"><a href="#cb469-60" aria-hidden="true" tabindex="-1"></a>indSelection_lowIndHetStan <span class="op">=</span> (meanAcrossRegionIndHetStan <span class="op">.&lt;</span> <span class="fl">1.4</span>) </span>
<span id="cb469-61"><a href="#cb469-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-62"><a href="#cb469-62" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot only the lowIndHetStan individuals:</span></span>
<span id="cb469-63"><a href="#cb469-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb469-64"><a href="#cb469-64" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>();</span>
<span id="cb469-65"><a href="#cb469-65" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb469-66"><a href="#cb469-66" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2, only low heterozygosity"</span>,</span>
<span id="cb469-67"><a href="#cb469-67" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb469-68"><a href="#cb469-68" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb469-69"><a href="#cb469-69" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb469-70"><a href="#cb469-70" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb469-71"><a href="#cb469-71" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb469-72"><a href="#cb469-72" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> (PCAmodelAll.metadata.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]) <span class="op">.&amp;</span> indSelection_lowIndHetStan</span>
<span id="cb469-73"><a href="#cb469-73" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCAmodelAll.metadata.PC1[selection], PCAmodelAll.metadata.PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb469-74"><a href="#cb469-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb469-75"><a href="#cb469-75" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-172-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-172-output-5.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="172">
<pre><code>CairoMakie.Screen{IMAGE}</code></pre>
</div>
</div>
</section>
<section id="tried-chr-1-but-some-unclean-distinguishing-of-ludlowi-clusters-no-wide-sharing-of-plumb" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-1-but-some-unclean-distinguishing-of-ludlowi-clusters-no-wide-sharing-of-plumb">Tried chr 1 but some unclean distinguishing of ludlowi clusters (no wide sharing of plumb)</h2>
</section>
<section id="included-chr-1a" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-1a">Included chr 1A</h2>
</section>
<section id="tried-chr-2-but-some-unclean-distinguishing-of-ludlowi-clusters-with-some-sharing-of-plumb" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-2-but-some-unclean-distinguishing-of-ludlowi-clusters-with-some-sharing-of-plumb">Tried chr 2 but some unclean distinguishing of ludlowi clusters (with some sharing of plumb)</h2>
</section>
<section id="included-chr-3" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-3">Included chr 3</h2>
</section>
<section id="tried-chr-4-but-not-clean-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-4-but-not-clean-separation-of-high-vs.-low-indhet">Tried chr 4 but not clean separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="included-chr-4a" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-4a">Included chr 4A</h2>
</section>
<section id="tried-chr-5-but-not-clean-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-5-but-not-clean-separation-of-high-vs.-low-indhet">Tried chr 5 but not clean separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="tried-chr-6-but-not-clean-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-6-but-not-clean-separation-of-high-vs.-low-indhet">Tried chr 6 but not clean separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="tried-chr-7-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-7-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi">Tried chr 7 but not clean separation of high vs.&nbsp;low IndHet, recomb in ludlowi</h2>
</section>
<section id="tried-chr-8-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-8-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi">Tried chr 8 but not clean separation of high vs.&nbsp;low IndHet, recomb in ludlowi</h2>
</section>
<section id="tried-chr-9-but-not-clean-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-9-but-not-clean-separation-of-high-vs.-low-indhet">Tried chr 9 but not clean separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="tried-chr-10-but-not-clean-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-10-but-not-clean-separation-of-high-vs.-low-indhet">Tried chr 10 but not clean separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="tried-chr-11-and-has-potential-and-shows-an-obs-with-two-plumb-types-but-ludlowi-not-cleanly-distinguished-into-types" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-11-and-has-potential-and-shows-an-obs-with-two-plumb-types-but-ludlowi-not-cleanly-distinguished-into-types">Tried chr 11 and has potential, and shows an obs with two plumb types, but ludlowi not cleanly distinguished into types</h2>
</section>
<section id="tried-chr-12-but-not-a-very-clear-separation-of-high-vs.-low-indhet-shows-a-lot-of-sharing-of-plumb-haps-in-ludlowi" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-12-but-not-a-very-clear-separation-of-high-vs.-low-indhet-shows-a-lot-of-sharing-of-plumb-haps-in-ludlowi">Tried chr 12 but not a very clear separation of high vs.&nbsp;low IndHet (shows a lot of sharing of plumb haps in ludlowi)</h2>
</section>
<section id="tried-chr-14-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-14-but-not-clean-separation-of-high-vs.-low-indhet-recomb-in-ludlowi">Tried chr 14 but not clean separation of high vs.&nbsp;low IndHet, recomb in ludlowi</h2>
</section>
<section id="included-chr-15" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-15">Included chr 15</h2>
</section>
<section id="no-chr-16" class="level2">
<h2 class="anchored" data-anchor-id="no-chr-16">No chr 16</h2>
</section>
<section id="tried-chr-23-but-not-a-very-clear-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-23-but-not-a-very-clear-separation-of-high-vs.-low-indhet">Tried chr 23 but not a very clear separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="included-chr-17-chr-18-chr-19-chr-20" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-17-chr-18-chr-19-chr-20">Included chr 17, chr 18, chr 19, chr 20</h2>
</section>
<section id="tried-chr-21-but-not-a-very-clear-separation-of-high-vs.-low-indhet" class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-21-but-not-a-very-clear-separation-of-high-vs.-low-indhet">Tried chr 21 but not a very clear separation of high vs.&nbsp;low IndHet</h2>
</section>
<section id="tried-chr-22-but-not-a-very-clear-separation-of-high-vs.-low-indhet.-lud-is-all-over-the-place." class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-22-but-not-a-very-clear-separation-of-high-vs.-low-indhet.-lud-is-all-over-the-place.">Tried chr 22 but not a very clear separation of high vs.&nbsp;low IndHet. Lud is all over the place.</h2>
</section>
<section id="tried-chr-23-and-almost-included-but-a-few-inds-would-be-tough-to-categorize.-similar-pattern-as-some-others.-not-a-very-clear-separation-of-high-vs.-low-indhet." class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-23-and-almost-included-but-a-few-inds-would-be-tough-to-categorize.-similar-pattern-as-some-others.-not-a-very-clear-separation-of-high-vs.-low-indhet.">Tried chr 23 and almost included, but a few inds would be tough to categorize. Similar pattern as some others. Not a very clear separation of high vs.&nbsp;low IndHet.</h2>
</section>
<section id="tried-chr-24-but-not-a-very-clear-separation-of-high-vs.-low-indhet." class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-24-but-not-a-very-clear-separation-of-high-vs.-low-indhet.">Tried chr 24 but not a very clear separation of high vs.&nbsp;low IndHet.</h2>
</section>
<section id="tried-chr-25-but-not-a-very-clear-separation-of-high-vs.-low-indhet-and-hard-to-categorize-a-lot-of-inds." class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-25-but-not-a-very-clear-separation-of-high-vs.-low-indhet-and-hard-to-categorize-a-lot-of-inds.">Tried chr 25 but not a very clear separation of high vs.&nbsp;low IndHet, and hard to categorize a lot of inds.</h2>
</section>
<section id="included-chr-26" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-26">Included chr 26</h2>
</section>
<section id="tried-chr-27-but-not-a-very-clear-separation-of-high-vs.-low-indhet." class="level2">
<h2 class="anchored" data-anchor-id="tried-chr-27-but-not-a-very-clear-separation-of-high-vs.-low-indhet.">Tried chr 27 but not a very clear separation of high vs.&nbsp;low IndHet.</h2>
</section>
<section id="included-chr-28" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-28">Included chr 28</h2>
</section>
<section id="included-chr-z" class="level2">
<h2 class="anchored" data-anchor-id="included-chr-z">Included chr Z</h2>
</section>
<section id="make-a-summary-plot-for-the-cluster-types-at-different-chromosome-haploblocks-west-without-nitidus" class="level2">
<h2 class="anchored" data-anchor-id="make-a-summary-plot-for-the-cluster-types-at-different-chromosome-haploblocks-west-without-nitidus">Make a summary plot for the cluster types at different chromosome haploblocks (west without nitidus)</h2>
<p>Will modify the plotGenotypeByIndividual() function, but need to construct a genotype data structure based on the groups (determined above) for each haploblock.</p>
<p>For west side (without nitidus):</p>
<div id="7bb88bda" class="cell" data-execution_count="172">
<div class="sourceCode cell-code" id="cb474"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb474-1"><a href="#cb474-1" aria-hidden="true" tabindex="-1"></a><span class="co">#= # For debugging function:</span></span>
<span id="cb474-2"><a href="#cb474-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-3"><a href="#cb474-3" aria-hidden="true" tabindex="-1"></a><span class="co">indMetadata = ind_with_metadata_included</span></span>
<span id="cb474-4"><a href="#cb474-4" aria-hidden="true" tabindex="-1"></a><span class="co">plotGroups = plotGroupsForSummary</span></span>
<span id="cb474-5"><a href="#cb474-5" aria-hidden="true" tabindex="-1"></a><span class="co">plotGroupColors = groupColorsForSummary</span></span>
<span id="cb474-6"><a href="#cb474-6" aria-hidden="true" tabindex="-1"></a><span class="co">regionNames = HaploblockRegions</span></span>
<span id="cb474-7"><a href="#cb474-7" aria-hidden="true" tabindex="-1"></a><span class="co">indFontSize = 10</span></span>
<span id="cb474-8"><a href="#cb474-8" aria-hidden="true" tabindex="-1"></a><span class="co">figureSize = (1200, 1200)</span></span>
<span id="cb474-9"><a href="#cb474-9" aria-hidden="true" tabindex="-1"></a><span class="co">plotTitle = nothing</span></span>
<span id="cb474-10"><a href="#cb474-10" aria-hidden="true" tabindex="-1"></a><span class="co">indColorLeftProvided = false</span></span>
<span id="cb474-11"><a href="#cb474-11" aria-hidden="true" tabindex="-1"></a><span class="co">indColorRightProvided = false =#</span></span>
<span id="cb474-12"><a href="#cb474-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-13"><a href="#cb474-13" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb474-14"><a href="#cb474-14" aria-hidden="true" tabindex="-1"></a><span class="st">    plotHaploblockSummary(genosSummary, indMetadata,</span></span>
<span id="cb474-15"><a href="#cb474-15" aria-hidden="true" tabindex="-1"></a><span class="st">                            plotGroups, plotGroupColors;</span></span>
<span id="cb474-16"><a href="#cb474-16" aria-hidden="true" tabindex="-1"></a><span class="st">                            regionNames,</span></span>
<span id="cb474-17"><a href="#cb474-17" aria-hidden="true" tabindex="-1"></a><span class="st">                            indFontSize=10, figureSize=(1200, 1200),</span></span>
<span id="cb474-18"><a href="#cb474-18" aria-hidden="true" tabindex="-1"></a><span class="st">                            plotTitle = nothing,</span></span>
<span id="cb474-19"><a href="#cb474-19" aria-hidden="true" tabindex="-1"></a><span class="st">                            indColorLeftProvided = false,</span></span>
<span id="cb474-20"><a href="#cb474-20" aria-hidden="true" tabindex="-1"></a><span class="st">                            indColorRightProvided = false)</span></span>
<span id="cb474-21"><a href="#cb474-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-22"><a href="#cb474-22" aria-hidden="true" tabindex="-1"></a><span class="st">Construct a genotype-by-individual plot, with option to filter out SNPs with too much missing data. </span></span>
<span id="cb474-23"><a href="#cb474-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-24"><a href="#cb474-24" aria-hidden="true" tabindex="-1"></a><span class="st">Under the default setting, alleles are colored (dark purple vs. light purple) according to whichever allele is designated as `group1`. </span></span>
<span id="cb474-25"><a href="#cb474-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-26"><a href="#cb474-26" aria-hidden="true" tabindex="-1"></a><span class="st">​# Arguments</span></span>
<span id="cb474-27"><a href="#cb474-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-28"><a href="#cb474-28" aria-hidden="true" tabindex="-1"></a><span class="st">- `genosSummary`: Matrix containing summary genotype data (individuals in rows, loci in columns).</span></span>
<span id="cb474-29"><a href="#cb474-29" aria-hidden="true" tabindex="-1"></a><span class="st">- `indMetadata`: Matrix of metadata for individuals; must contain `Fst_group` and `plot_order` columns.</span></span>
<span id="cb474-30"><a href="#cb474-30" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotGroups`: Vector of group names to include in plot.</span></span>
<span id="cb474-31"><a href="#cb474-31" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotGroupColors`: Vector of plotting colors corresponding to the groups.</span></span>
<span id="cb474-32"><a href="#cb474-32" aria-hidden="true" tabindex="-1"></a><span class="st">- `regionNames`: Optional; Names of the genotyped regions.</span></span>
<span id="cb474-33"><a href="#cb474-33" aria-hidden="true" tabindex="-1"></a><span class="st">- `indFontSize`: Optional; the font size of the individual ID labels.</span></span>
<span id="cb474-34"><a href="#cb474-34" aria-hidden="true" tabindex="-1"></a><span class="st">- `figureSize`: Optional; the size of the figure; default is `(1200, 1200)`.  </span></span>
<span id="cb474-35"><a href="#cb474-35" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotTitle`: Optional; default will make a title. For no title, set to `""`.</span></span>
<span id="cb474-36"><a href="#cb474-36" aria-hidden="true" tabindex="-1"></a><span class="st">- `indColorLeftProvided`: Optional; Default is `false`. Set to `true` if there is a column labeled `indColorLeft` in the metadata providing color of each individual for plotting on left side.</span></span>
<span id="cb474-37"><a href="#cb474-37" aria-hidden="true" tabindex="-1"></a><span class="st">- `indColorRightProvided`: Optional; same as above but for right side (requires `indColorRight` column in metadata).</span></span>
<span id="cb474-38"><a href="#cb474-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-39"><a href="#cb474-39" aria-hidden="true" tabindex="-1"></a><span class="st"># Notes</span></span>
<span id="cb474-40"><a href="#cb474-40" aria-hidden="true" tabindex="-1"></a><span class="st">Returns a tuple containing:</span></span>
<span id="cb474-41"><a href="#cb474-41" aria-hidden="true" tabindex="-1"></a><span class="st">- the figure</span></span>
<span id="cb474-42"><a href="#cb474-42" aria-hidden="true" tabindex="-1"></a><span class="st">- the plotted genotypes</span></span>
<span id="cb474-43"><a href="#cb474-43" aria-hidden="true" tabindex="-1"></a><span class="st">- the sorted metadata matrix for the plotted individuals</span></span>
<span id="cb474-44"><a href="#cb474-44" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb474-45"><a href="#cb474-45" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">plotHaploblockSummary</span>(genosSummary, indMetadata,</span>
<span id="cb474-46"><a href="#cb474-46" aria-hidden="true" tabindex="-1"></a>                                plotGroups, plotGroupColors;</span>
<span id="cb474-47"><a href="#cb474-47" aria-hidden="true" tabindex="-1"></a>                                regionNames <span class="op">=</span> <span class="cn">nothing</span>,        </span>
<span id="cb474-48"><a href="#cb474-48" aria-hidden="true" tabindex="-1"></a>                                indFontSize<span class="op">=</span><span class="fl">10</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">1200</span>),</span>
<span id="cb474-49"><a href="#cb474-49" aria-hidden="true" tabindex="-1"></a>                                plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb474-50"><a href="#cb474-50" aria-hidden="true" tabindex="-1"></a>                                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb474-51"><a href="#cb474-51" aria-hidden="true" tabindex="-1"></a>                                indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb474-52"><a href="#cb474-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb474-53"><a href="#cb474-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if the genoData has missing values, then convert to -1:</span></span>
<span id="cb474-54"><a href="#cb474-54" aria-hidden="true" tabindex="-1"></a>    genosSummary[<span class="fu">ismissing</span>.(genosSummary)] <span class="op">.=</span> <span class="op">-</span><span class="fl">1</span></span>
<span id="cb474-55"><a href="#cb474-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-56"><a href="#cb474-56" aria-hidden="true" tabindex="-1"></a>    numRegions <span class="op">=</span> <span class="fu">size</span>(genosSummary, <span class="fl">2</span>)</span>
<span id="cb474-57"><a href="#cb474-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-58"><a href="#cb474-58" aria-hidden="true" tabindex="-1"></a>    genosSummary_subset <span class="op">=</span> genosSummary[indMetadata.Fst_group <span class="op">.∈</span> <span class="fu">Ref</span>(plotGroups), <span class="op">:</span>]</span>
<span id="cb474-59"><a href="#cb474-59" aria-hidden="true" tabindex="-1"></a>    indMetadata_subset <span class="op">=</span> indMetadata[indMetadata.Fst_group <span class="op">.∈</span> <span class="fu">Ref</span>(plotGroups), <span class="op">:</span>]</span>
<span id="cb474-60"><a href="#cb474-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-61"><a href="#cb474-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Choose sorting order by plot_order column in input metadata file</span></span>
<span id="cb474-62"><a href="#cb474-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-63"><a href="#cb474-63" aria-hidden="true" tabindex="-1"></a>    sorted_genosSummary_subset <span class="op">=</span> genosSummary_subset[<span class="fu">sortperm</span>(indMetadata_subset.plot_order, rev<span class="op">=</span><span class="cn">false</span>), <span class="op">:</span>]</span>
<span id="cb474-64"><a href="#cb474-64" aria-hidden="true" tabindex="-1"></a>    numInds <span class="op">=</span> <span class="fu">size</span>(sorted_genosSummary_subset, <span class="fl">1</span>)</span>
<span id="cb474-65"><a href="#cb474-65" aria-hidden="true" tabindex="-1"></a>    sorted_indMetadata_subset <span class="op">=</span> indMetadata_subset[<span class="fu">sortperm</span>(indMetadata_subset.plot_order, rev<span class="op">=</span><span class="cn">false</span>), <span class="op">:</span>]</span>
<span id="cb474-66"><a href="#cb474-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-67"><a href="#cb474-67" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set up the plot window:</span></span>
<span id="cb474-68"><a href="#cb474-68" aria-hidden="true" tabindex="-1"></a>    f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>(size<span class="op">=</span>figureSize)</span>
<span id="cb474-69"><a href="#cb474-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-70"><a href="#cb474-70" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">isnothing</span>(plotTitle)</span>
<span id="cb474-71"><a href="#cb474-71" aria-hidden="true" tabindex="-1"></a>        plotTitle <span class="op">=</span> <span class="st">"Summary of </span><span class="sc">$</span>numRegions<span class="st"> haploblock genotypes for </span><span class="sc">$</span>numInds<span class="st"> individuals"</span></span>
<span id="cb474-72"><a href="#cb474-72" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span> </span>
<span id="cb474-73"><a href="#cb474-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-74"><a href="#cb474-74" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set up the main Axis: </span></span>
<span id="cb474-75"><a href="#cb474-75" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb474-76"><a href="#cb474-76" aria-hidden="true" tabindex="-1"></a>        title <span class="op">=</span> plotTitle,</span>
<span id="cb474-77"><a href="#cb474-77" aria-hidden="true" tabindex="-1"></a>        titlesize<span class="op">=</span><span class="fl">30</span>,</span>
<span id="cb474-78"><a href="#cb474-78" aria-hidden="true" tabindex="-1"></a>        limits<span class="op">=</span>(<span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.09</span> <span class="op">*</span> (numRegions), <span class="fl">0.5</span> <span class="op">+</span> <span class="fl">1.09</span> <span class="op">*</span> (numRegions),</span>
<span id="cb474-79"><a href="#cb474-79" aria-hidden="true" tabindex="-1"></a>            <span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.3</span> <span class="op">*</span> numInds, <span class="fl">0.5</span> <span class="op">+</span> numInds)</span>
<span id="cb474-80"><a href="#cb474-80" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb474-81"><a href="#cb474-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hidedecorations!</span>(ax) <span class="co"># hide background lattice and axis labels</span></span>
<span id="cb474-82"><a href="#cb474-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hidespines!</span>(ax) <span class="co"># hide box around plot</span></span>
<span id="cb474-83"><a href="#cb474-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-84"><a href="#cb474-84" aria-hidden="true" tabindex="-1"></a>    genotypeColors <span class="op">=</span> [<span class="st">"#3f007d"</span>, <span class="st">"#807dba"</span>, <span class="st">"#dadaeb"</span>, <span class="st">"grey50"</span>]  <span class="co"># purple shades from colorbrewer</span></span>
<span id="cb474-85"><a href="#cb474-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-86"><a href="#cb474-86" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot evenly spaced by SNP order along chromosome:</span></span>
<span id="cb474-87"><a href="#cb474-87" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make top part of fig (genotypes for individuals)</span></span>
<span id="cb474-88"><a href="#cb474-88" aria-hidden="true" tabindex="-1"></a>    labelCushion <span class="op">=</span> numRegions <span class="op">/</span> <span class="fl">100</span></span>
<span id="cb474-89"><a href="#cb474-89" aria-hidden="true" tabindex="-1"></a>    label_x_left <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> labelCushion</span>
<span id="cb474-90"><a href="#cb474-90" aria-hidden="true" tabindex="-1"></a>    label_x_right <span class="op">=</span> <span class="fl">0.5</span> <span class="op">+</span> numRegions <span class="op">+</span> labelCushion</span>
<span id="cb474-91"><a href="#cb474-91" aria-hidden="true" tabindex="-1"></a>    colorBoxCushion <span class="op">=</span> <span class="fl">0.07</span> <span class="op">*</span> numRegions</span>
<span id="cb474-92"><a href="#cb474-92" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_left <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> colorBoxCushion</span>
<span id="cb474-93"><a href="#cb474-93" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_right <span class="op">=</span> <span class="fl">0.5</span> <span class="op">+</span> numRegions <span class="op">+</span> colorBoxCushion</span>
<span id="cb474-94"><a href="#cb474-94" aria-hidden="true" tabindex="-1"></a>    boxWidth <span class="op">=</span> <span class="fl">0.005</span> <span class="op">*</span> numRegions <span class="op">*</span> <span class="fl">2</span></span>
<span id="cb474-95"><a href="#cb474-95" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_left <span class="op">=</span> [<span class="op">-</span>boxWidth, <span class="op">-</span>boxWidth, boxWidth, boxWidth, <span class="op">-</span>boxWidth] <span class="op">.+</span> groupColorBox_x_left</span>
<span id="cb474-96"><a href="#cb474-96" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_right <span class="op">=</span> [<span class="op">-</span>boxWidth, <span class="op">-</span>boxWidth, boxWidth, boxWidth, <span class="op">-</span>boxWidth] <span class="op">.+</span> groupColorBox_x_right</span>
<span id="cb474-97"><a href="#cb474-97" aria-hidden="true" tabindex="-1"></a>    groupColorBox_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb474-98"><a href="#cb474-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-99"><a href="#cb474-99" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb474-100"><a href="#cb474-100" aria-hidden="true" tabindex="-1"></a>        y <span class="op">=</span> numInds <span class="op">+</span> <span class="fl">1</span> <span class="op">-</span> i  <span class="co"># y is location for plotting; this reverses order of plot top-bottom</span></span>
<span id="cb474-101"><a href="#cb474-101" aria-hidden="true" tabindex="-1"></a>        labelText <span class="op">=</span> <span class="fu">last</span>(<span class="fu">split</span>(sorted_indMetadata_subset.ID[i], <span class="st">"_"</span>))  <span class="co"># this gets the last part of the sample ID (usually the main ID part)</span></span>
<span id="cb474-102"><a href="#cb474-102" aria-hidden="true" tabindex="-1"></a>        <span class="co"># put sample label on left side:</span></span>
<span id="cb474-103"><a href="#cb474-103" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(label_x_left, y; text<span class="op">=</span>labelText, align<span class="op">=</span>(<span class="op">:</span>right, <span class="op">:</span>center), fontsize<span class="op">=</span>indFontSize)</span>
<span id="cb474-104"><a href="#cb474-104" aria-hidden="true" tabindex="-1"></a>        <span class="co"># put sample label on left side:</span></span>
<span id="cb474-105"><a href="#cb474-105" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(label_x_right, y; text<span class="op">=</span>labelText, align<span class="op">=</span>(<span class="op">:</span>left, <span class="op">:</span>center), fontsize<span class="op">=</span>indFontSize)</span>
<span id="cb474-106"><a href="#cb474-106" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> indColorLeftProvided</span>
<span id="cb474-107"><a href="#cb474-107" aria-hidden="true" tabindex="-1"></a>            boxColorLeft <span class="op">=</span> sorted_indMetadata_subset.indColorLeft[i]</span>
<span id="cb474-108"><a href="#cb474-108" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb474-109"><a href="#cb474-109" aria-hidden="true" tabindex="-1"></a>            boxColorLeft <span class="op">=</span> plotGroupColors[<span class="fu">findfirst</span>(plotGroups <span class="op">.==</span> sorted_indMetadata_subset.Fst_group[i])]</span>
<span id="cb474-110"><a href="#cb474-110" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb474-111"><a href="#cb474-111" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> indColorRightProvided</span>
<span id="cb474-112"><a href="#cb474-112" aria-hidden="true" tabindex="-1"></a>            boxColorRight <span class="op">=</span> sorted_indMetadata_subset.indColorRight[i]</span>
<span id="cb474-113"><a href="#cb474-113" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb474-114"><a href="#cb474-114" aria-hidden="true" tabindex="-1"></a>            boxColorRight <span class="op">=</span> plotGroupColors[<span class="fu">findfirst</span>(plotGroups <span class="op">.==</span> sorted_indMetadata_subset.Fst_group[i])]</span>
<span id="cb474-115"><a href="#cb474-115" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb474-116"><a href="#cb474-116" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.(groupColorBox_x_left, (y <span class="op">.+</span> groupColorBox_y)), color<span class="op">=</span>boxColorLeft)</span>
<span id="cb474-117"><a href="#cb474-117" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.(groupColorBox_x_right, (y <span class="op">.+</span> groupColorBox_y)), color<span class="op">=</span>boxColorRight)</span>
<span id="cb474-118"><a href="#cb474-118" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb474-119"><a href="#cb474-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-120"><a href="#cb474-120" aria-hidden="true" tabindex="-1"></a>    <span class="co"># generate my own plotting symbol (a rectangle)</span></span>
<span id="cb474-121"><a href="#cb474-121" aria-hidden="true" tabindex="-1"></a>    box_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb474-122"><a href="#cb474-122" aria-hidden="true" tabindex="-1"></a>    <span class="co">#box_x = [-0.5, -0.5, 0.5, 0.5, -0.5]</span></span>
<span id="cb474-123"><a href="#cb474-123" aria-hidden="true" tabindex="-1"></a>    box_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb474-124"><a href="#cb474-124" aria-hidden="true" tabindex="-1"></a>    <span class="co"># generate triangles for plotting heterozygotes</span></span>
<span id="cb474-125"><a href="#cb474-125" aria-hidden="true" tabindex="-1"></a>    triangle1_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb474-126"><a href="#cb474-126" aria-hidden="true" tabindex="-1"></a>    <span class="co">#triangle1_x = [-0.5, -0.5, 0.5, -0.5]</span></span>
<span id="cb474-127"><a href="#cb474-127" aria-hidden="true" tabindex="-1"></a>    triangle1_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb474-128"><a href="#cb474-128" aria-hidden="true" tabindex="-1"></a>    triangle2_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb474-129"><a href="#cb474-129" aria-hidden="true" tabindex="-1"></a>    <span class="co">#triangle2_x = [-0.5, 0.5, 0.5, -0.5]</span></span>
<span id="cb474-130"><a href="#cb474-130" aria-hidden="true" tabindex="-1"></a>    triangle2_y <span class="op">=</span> [<span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>]</span>
<span id="cb474-131"><a href="#cb474-131" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cycle through individuals, graphing each type of genotype:</span></span>
<span id="cb474-132"><a href="#cb474-132" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb474-133"><a href="#cb474-133" aria-hidden="true" tabindex="-1"></a>        y <span class="op">=</span> numInds <span class="op">+</span> <span class="fl">1</span> <span class="op">-</span> i  <span class="co"># y is location for plotting; this reverses order of plot top-bottom</span></span>
<span id="cb474-134"><a href="#cb474-134" aria-hidden="true" tabindex="-1"></a>        <span class="co">#CairoMakie.lines!([0.5, numRegions + 0.5], [y, y], color="grey40") # for lines across the individual rows</span></span>
<span id="cb474-135"><a href="#cb474-135" aria-hidden="true" tabindex="-1"></a>        genotypes <span class="op">=</span> sorted_genosSummary_subset[i, <span class="op">:</span>]</span>
<span id="cb474-136"><a href="#cb474-136" aria-hidden="true" tabindex="-1"></a>        hom_ref_locs <span class="op">=</span> <span class="fu">findall</span>(genotypes <span class="op">.==</span> <span class="fl">0</span>)</span>
<span id="cb474-137"><a href="#cb474-137" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">length</span>(hom_ref_locs) <span class="op">&gt;</span> <span class="fl">0</span></span>
<span id="cb474-138"><a href="#cb474-138" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> j <span class="kw">in</span> <span class="fu">eachindex</span>(hom_ref_locs)</span>
<span id="cb474-139"><a href="#cb474-139" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((hom_ref_locs[j] <span class="op">.+</span> box_x), (y <span class="op">.+</span> box_y)), color<span class="op">=</span>genotypeColors[<span class="fl">1</span>])</span>
<span id="cb474-140"><a href="#cb474-140" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb474-141"><a href="#cb474-141" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb474-142"><a href="#cb474-142" aria-hidden="true" tabindex="-1"></a>        het_locs <span class="op">=</span> <span class="fu">findall</span>(genotypes <span class="op">.==</span> <span class="fl">1</span>)</span>
<span id="cb474-143"><a href="#cb474-143" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">length</span>(het_locs) <span class="op">&gt;</span> <span class="fl">0</span></span>
<span id="cb474-144"><a href="#cb474-144" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> j <span class="kw">in</span> <span class="fu">eachindex</span>(het_locs)</span>
<span id="cb474-145"><a href="#cb474-145" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((het_locs[j] <span class="op">.+</span> triangle1_x), (y <span class="op">.+</span> triangle1_y)), color<span class="op">=</span>genotypeColors[<span class="fl">1</span>])</span>
<span id="cb474-146"><a href="#cb474-146" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((het_locs[j] <span class="op">.+</span> triangle2_x), (y <span class="op">.+</span> triangle2_y)), color<span class="op">=</span>genotypeColors[<span class="fl">3</span>])</span>
<span id="cb474-147"><a href="#cb474-147" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb474-148"><a href="#cb474-148" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb474-149"><a href="#cb474-149" aria-hidden="true" tabindex="-1"></a>        hom_alt_locs <span class="op">=</span> <span class="fu">findall</span>(genotypes <span class="op">.==</span> <span class="fl">2</span>)</span>
<span id="cb474-150"><a href="#cb474-150" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">length</span>(hom_alt_locs) <span class="op">&gt;</span> <span class="fl">0</span></span>
<span id="cb474-151"><a href="#cb474-151" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> j <span class="kw">in</span> <span class="fu">eachindex</span>(hom_alt_locs)</span>
<span id="cb474-152"><a href="#cb474-152" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((hom_alt_locs[j] <span class="op">.+</span> box_x), (y <span class="op">.+</span> box_y)), color<span class="op">=</span>genotypeColors[<span class="fl">3</span>])</span>
<span id="cb474-153"><a href="#cb474-153" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb474-154"><a href="#cb474-154" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb474-155"><a href="#cb474-155" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb474-156"><a href="#cb474-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-157"><a href="#cb474-157" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">isnothing</span>(regionNames)</span>
<span id="cb474-158"><a href="#cb474-158" aria-hidden="true" tabindex="-1"></a>        regionNames <span class="op">=</span> <span class="fu">string</span>.(<span class="fl">1</span><span class="op">:</span>numRegions)</span>
<span id="cb474-159"><a href="#cb474-159" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb474-160"><a href="#cb474-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-161"><a href="#cb474-161" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make labels on lower part</span></span>
<span id="cb474-162"><a href="#cb474-162" aria-hidden="true" tabindex="-1"></a>    y_label <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.025</span>numInds</span>
<span id="cb474-163"><a href="#cb474-163" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRegions</span>
<span id="cb474-164"><a href="#cb474-164" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(i, y_label; text <span class="op">=</span> regionNames[i], align<span class="op">=</span>(<span class="op">:</span>center, <span class="op">:</span>center), fontsize<span class="op">=</span><span class="fl">30</span>)</span>
<span id="cb474-165"><a href="#cb474-165" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb474-166"><a href="#cb474-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-167"><a href="#cb474-167" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(f)</span>
<span id="cb474-168"><a href="#cb474-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-169"><a href="#cb474-169" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> f, sorted_genosSummary_subset, sorted_indMetadata_subset</span>
<span id="cb474-170"><a href="#cb474-170" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb474-171"><a href="#cb474-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-172"><a href="#cb474-172" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up a data structure to store the key to converting, for each haploblock region, </span></span>
<span id="cb474-173"><a href="#cb474-173" aria-hidden="true" tabindex="-1"></a><span class="co"># the cluster names to genotype integers. This is a dictiionary of dictionaries:</span></span>
<span id="cb474-174"><a href="#cb474-174" aria-hidden="true" tabindex="-1"></a>regionHaplotypeCode_west <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String, Dict{String, Int}}</span>(</span>
<span id="cb474-175"><a href="#cb474-175" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw1A"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-176"><a href="#cb474-176" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw3"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLudHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_trochObs"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"trochObs"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"trochObsHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-177"><a href="#cb474-177" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw13"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-178"><a href="#cb474-178" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw15"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-179"><a href="#cb474-179" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw18"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-180"><a href="#cb474-180" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw19"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLudHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_trochObs"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"trochObs"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"trochObsHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-181"><a href="#cb474-181" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw26"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-182"><a href="#cb474-182" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw28"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb474-183"><a href="#cb474-183" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gwZ"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"vir"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"lud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"lud_troch"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"troch"</span><span class="op">=&gt;</span><span class="fl">2</span>)</span>
<span id="cb474-184"><a href="#cb474-184" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb474-185"><a href="#cb474-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-186"><a href="#cb474-186" aria-hidden="true" tabindex="-1"></a>haploblockRegions <span class="op">=</span> [<span class="st">"gw1A"</span>, <span class="st">"gw3"</span>, <span class="st">"gw13"</span>, <span class="st">"gw15"</span>, <span class="st">"gw18"</span>, <span class="st">"gw19"</span>, <span class="st">"gw26"</span>, <span class="st">"gw28"</span>, <span class="st">"gwZ"</span>]</span>
<span id="cb474-187"><a href="#cb474-187" aria-hidden="true" tabindex="-1"></a>numHaploblockRegions <span class="op">=</span> <span class="fu">length</span>(haploblockRegions)</span>
<span id="cb474-188"><a href="#cb474-188" aria-hidden="true" tabindex="-1"></a>numInds <span class="op">=</span> <span class="fu">size</span>(ind_with_metadata_included, <span class="fl">1</span>)</span>
<span id="cb474-189"><a href="#cb474-189" aria-hidden="true" tabindex="-1"></a><span class="co"># create genotype object and fill with missing (-1) genotypes</span></span>
<span id="cb474-190"><a href="#cb474-190" aria-hidden="true" tabindex="-1"></a>genosSummary <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">1</span>, (numInds, numHaploblockRegions)) </span>
<span id="cb474-191"><a href="#cb474-191" aria-hidden="true" tabindex="-1"></a><span class="co"># fill object with appropriate genotypes</span></span>
<span id="cb474-192"><a href="#cb474-192" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numHaploblockRegions</span>
<span id="cb474-193"><a href="#cb474-193" aria-hidden="true" tabindex="-1"></a>    region <span class="op">=</span> haploblockRegions[i]</span>
<span id="cb474-194"><a href="#cb474-194" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (key, value) <span class="kw">in</span> regionHaplotypeCode_west[region]</span>
<span id="cb474-195"><a href="#cb474-195" aria-hidden="true" tabindex="-1"></a>        command <span class="op">=</span> <span class="st">"""genosSummary[ind_with_metadata_included.</span><span class="sc">$</span>(region)<span class="st">_cluster .== "</span><span class="sc">$</span>(key)<span class="st">", </span><span class="sc">$</span>i<span class="st">] .= """</span> <span class="op">*</span> <span class="fu">string</span>(value)</span>
<span id="cb474-196"><a href="#cb474-196" aria-hidden="true" tabindex="-1"></a>        <span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb474-197"><a href="#cb474-197" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb474-198"><a href="#cb474-198" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb474-199"><a href="#cb474-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-200"><a href="#cb474-200" aria-hidden="true" tabindex="-1"></a><span class="co"># Must say I am pleased with the cleverness of above. Concise datastructure and code that does a lot. :) </span></span>
<span id="cb474-201"><a href="#cb474-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-202"><a href="#cb474-202" aria-hidden="true" tabindex="-1"></a>plotGroupsForSummaryWest <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>]</span>
<span id="cb474-203"><a href="#cb474-203" aria-hidden="true" tabindex="-1"></a>groupColorsForSummaryWest <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>] </span>
<span id="cb474-204"><a href="#cb474-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb474-205"><a href="#cb474-205" aria-hidden="true" tabindex="-1"></a><span class="fu">plotHaploblockSummary</span>(genosSummary, ind_with_metadata_included,</span>
<span id="cb474-206"><a href="#cb474-206" aria-hidden="true" tabindex="-1"></a>                            plotGroupsForSummaryWest, groupColorsForSummaryWest;</span>
<span id="cb474-207"><a href="#cb474-207" aria-hidden="true" tabindex="-1"></a>                            regionNames <span class="op">=</span> haploblockRegions,</span>
<span id="cb474-208"><a href="#cb474-208" aria-hidden="true" tabindex="-1"></a>                            indFontSize <span class="op">=</span> <span class="fl">8</span>, figureSize <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">1600</span>),</span>
<span id="cb474-209"><a href="#cb474-209" aria-hidden="true" tabindex="-1"></a>                            plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb474-210"><a href="#cb474-210" aria-hidden="true" tabindex="-1"></a>                            indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb474-211"><a href="#cb474-211" aria-hidden="true" tabindex="-1"></a>                            indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-173-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="173">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  1 Child Scene:
    └ Scene (768px, 960px), [0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], <span class="ansi-bold">177×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 168 │ GW_Lane5_LN14              GW_Lane5_LN14              LN        troch_L
 169 │ GW_Lane5_LN16              GW_Lane5_LN16              LN        troch_L ⋯
 170 │ GW_Lane5_LN18              GW_Lane5_LN18              LN        troch_L
 171 │ GW_Lane5_LN19              GW_Lane5_LN19              LN        troch_L
 172 │ GW_Lane5_LN20              GW_Lane5_LN20              LN        troch_L
 173 │ GW_Lane5_LN3               GW_Lane5_LN3               LN        troch_L ⋯
 174 │ GW_Lane5_LN4               GW_Lane5_LN4               LN        troch_L
 175 │ GW_Lane5_LN6               GW_Lane5_LN6               LN        troch_L
 176 │ GW_Lane5_LN7               GW_Lane5_LN7               LN        troch_L
 177 │ GW_Lane5_LN8               GW_Lane5_LN8               LN        troch_L ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 156 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="make-a-summary-plot-for-the-east-side" class="level2">
<h2 class="anchored" data-anchor-id="make-a-summary-plot-for-the-east-side">Make a summary plot for the east side</h2>
<div id="ff5a8f4c" class="cell" data-execution_count="173">
<div class="sourceCode cell-code" id="cb476"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb476-1"><a href="#cb476-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up a data structure to store the key to converting, for each haploblock region, </span></span>
<span id="cb476-2"><a href="#cb476-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the cluster names to genotype integers. This is a dictiionary of dictionaries:</span></span>
<span id="cb476-3"><a href="#cb476-3" aria-hidden="true" tabindex="-1"></a>regionHaplotypeCode_east <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String, Dict{String, Int}}</span>(</span>
<span id="cb476-4"><a href="#cb476-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw1A"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-5"><a href="#cb476-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw3"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"trochObs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"trochObsHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"plumbHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-6"><a href="#cb476-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw13"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"plumbHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-7"><a href="#cb476-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw15"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-8"><a href="#cb476-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw18"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"obs_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-9"><a href="#cb476-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw19"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"trochObs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"trochObs_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-10"><a href="#cb476-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw26"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"obs_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-11"><a href="#cb476-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw28"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"obsHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"obs_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb476-12"><a href="#cb476-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gwZ"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"obs"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>)</span>
<span id="cb476-13"><a href="#cb476-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb476-14"><a href="#cb476-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb476-15"><a href="#cb476-15" aria-hidden="true" tabindex="-1"></a>haploblockRegions <span class="op">=</span> [<span class="st">"gw1A"</span>, <span class="st">"gw3"</span>, <span class="st">"gw13"</span>, <span class="st">"gw15"</span>, <span class="st">"gw18"</span>, <span class="st">"gw19"</span>, <span class="st">"gw26"</span>, <span class="st">"gw28"</span>, <span class="st">"gwZ"</span>]</span>
<span id="cb476-16"><a href="#cb476-16" aria-hidden="true" tabindex="-1"></a>numHaploblockRegions <span class="op">=</span> <span class="fu">length</span>(haploblockRegions)</span>
<span id="cb476-17"><a href="#cb476-17" aria-hidden="true" tabindex="-1"></a>numInds <span class="op">=</span> <span class="fu">size</span>(ind_with_metadata_included, <span class="fl">1</span>)</span>
<span id="cb476-18"><a href="#cb476-18" aria-hidden="true" tabindex="-1"></a><span class="co"># create genotype object and fill with missing (-1) genotypes</span></span>
<span id="cb476-19"><a href="#cb476-19" aria-hidden="true" tabindex="-1"></a>genosSummary <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">1</span>, (numInds, numHaploblockRegions)) </span>
<span id="cb476-20"><a href="#cb476-20" aria-hidden="true" tabindex="-1"></a><span class="co"># fill object with appropriate genotypes</span></span>
<span id="cb476-21"><a href="#cb476-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numHaploblockRegions</span>
<span id="cb476-22"><a href="#cb476-22" aria-hidden="true" tabindex="-1"></a>    region <span class="op">=</span> haploblockRegions[i]</span>
<span id="cb476-23"><a href="#cb476-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (key, value) <span class="kw">in</span> regionHaplotypeCode_east[region]</span>
<span id="cb476-24"><a href="#cb476-24" aria-hidden="true" tabindex="-1"></a>        command <span class="op">=</span> <span class="st">"""genosSummary[ind_with_metadata_included.</span><span class="sc">$</span>(region)<span class="st">_cluster .== "</span><span class="sc">$</span>(key)<span class="st">", </span><span class="sc">$</span>i<span class="st">] .= """</span> <span class="op">*</span> <span class="fu">string</span>(value)</span>
<span id="cb476-25"><a href="#cb476-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb476-26"><a href="#cb476-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb476-27"><a href="#cb476-27" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb476-28"><a href="#cb476-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb476-29"><a href="#cb476-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Must say I am pleased with the cleverness of above. Concise datastructure and code that does a lot. :) </span></span>
<span id="cb476-30"><a href="#cb476-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb476-31"><a href="#cb476-31" aria-hidden="true" tabindex="-1"></a>plotGroupsForSummaryEast <span class="op">=</span> [<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb476-32"><a href="#cb476-32" aria-hidden="true" tabindex="-1"></a>groupColorsForSummaryEast <span class="op">=</span> [<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>] </span>
<span id="cb476-33"><a href="#cb476-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb476-34"><a href="#cb476-34" aria-hidden="true" tabindex="-1"></a><span class="fu">plotHaploblockSummary</span>(genosSummary, ind_with_metadata_included,</span>
<span id="cb476-35"><a href="#cb476-35" aria-hidden="true" tabindex="-1"></a>                            plotGroupsForSummaryEast, groupColorsForSummaryEast;</span>
<span id="cb476-36"><a href="#cb476-36" aria-hidden="true" tabindex="-1"></a>                            regionNames <span class="op">=</span> haploblockRegions,</span>
<span id="cb476-37"><a href="#cb476-37" aria-hidden="true" tabindex="-1"></a>                            indFontSize <span class="op">=</span> <span class="fl">8</span>, figureSize <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">1600</span>),</span>
<span id="cb476-38"><a href="#cb476-38" aria-hidden="true" tabindex="-1"></a>                            plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb476-39"><a href="#cb476-39" aria-hidden="true" tabindex="-1"></a>                            indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb476-40"><a href="#cb476-40" aria-hidden="true" tabindex="-1"></a>                            indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-174-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="174">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  1 Child Scene:
    └ Scene (768px, 960px), [0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 2 2 … 2 2], <span class="ansi-bold">76×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Lane5_DA2               GW_Lane5_DA2               XN        obs     ⋯
   2 │ GW_Lane5_DA3               GW_Lane5_DA3               XN        obs
   3 │ GW_Lane5_DA4               GW_Lane5_DA4               XN        obs
   4 │ GW_Lane5_DA7               GW_Lane5_DA7               XN        obs
   5 │ GW_Armando_plate1_RF20G01  GW_Armando_plate1_RF20G01  BJ        obs_plu ⋯
   6 │ GW_Armando_plate1_RF29G02  GW_Armando_plate1_RF29G02  BJ        obs_plu
   7 │ GW_Armando_plate2_RF29G01  GW_Armando_plate2_RF29G01  BJ        obs_plu
   8 │ GW_Lane5_BK2               GW_Lane5_BK2               BK        plumb
   9 │ GW_Lane5_BK3               GW_Lane5_BK3               BK        plumb   ⋯
  10 │ GW_Lane5_AN1               GW_Lane5_AN1               AN        plumb
  11 │ GW_Lane5_AN2               GW_Lane5_AN2               AN        plumb
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  67 │ GW_Armando_plate1_JF24G02  GW_Armando_plate1_JF24G02  VB        plumb
  68 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb   ⋯
  69 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb
  70 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  71 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  72 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb   ⋯
  73 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb
  74 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  75 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  76 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb   ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 55 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="make-a-summary-plot-for-the-north-side" class="level2">
<h2 class="anchored" data-anchor-id="make-a-summary-plot-for-the-north-side">Make a summary plot for the north side</h2>
<div id="1b3dc7ba" class="cell" data-execution_count="174">
<div class="sourceCode cell-code" id="cb478"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb478-1"><a href="#cb478-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up a data structure to store the code to converting, for each haploblock region, </span></span>
<span id="cb478-2"><a href="#cb478-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the cluster names to genotype integers. This is a dictiionary of dictionaries:</span></span>
<span id="cb478-3"><a href="#cb478-3" aria-hidden="true" tabindex="-1"></a>regionHaplotypeCode_north <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String, Dict{String, Int}}</span>(</span>
<span id="cb478-4"><a href="#cb478-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw1A"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-5"><a href="#cb478-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw3"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLudHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"plumbHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-6"><a href="#cb478-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw13"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>, <span class="st">"plumbHet"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-7"><a href="#cb478-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw15"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-8"><a href="#cb478-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw18"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-9"><a href="#cb478-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw19"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"virLudHet"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-10"><a href="#cb478-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw26"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-11"><a href="#cb478-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw28"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"vir_plumb"</span><span class="op">=&gt;</span><span class="fl">1</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>),</span>
<span id="cb478-12"><a href="#cb478-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gwZ"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"vir"</span><span class="op">=&gt;</span><span class="fl">0</span>, <span class="st">"plumb"</span><span class="op">=&gt;</span><span class="fl">2</span>)</span>
<span id="cb478-13"><a href="#cb478-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb478-14"><a href="#cb478-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-15"><a href="#cb478-15" aria-hidden="true" tabindex="-1"></a>haploblockRegions <span class="op">=</span> [<span class="st">"gw1A"</span>, <span class="st">"gw3"</span>, <span class="st">"gw13"</span>, <span class="st">"gw15"</span>, <span class="st">"gw18"</span>, <span class="st">"gw19"</span>, <span class="st">"gw26"</span>, <span class="st">"gw28"</span>, <span class="st">"gwZ"</span>]</span>
<span id="cb478-16"><a href="#cb478-16" aria-hidden="true" tabindex="-1"></a>numHaploblockRegions <span class="op">=</span> <span class="fu">length</span>(haploblockRegions)</span>
<span id="cb478-17"><a href="#cb478-17" aria-hidden="true" tabindex="-1"></a>numInds <span class="op">=</span> <span class="fu">size</span>(ind_with_metadata_included, <span class="fl">1</span>)</span>
<span id="cb478-18"><a href="#cb478-18" aria-hidden="true" tabindex="-1"></a><span class="co"># create genotype object and fill with missing (-1) genotypes</span></span>
<span id="cb478-19"><a href="#cb478-19" aria-hidden="true" tabindex="-1"></a>genosSummary <span class="op">=</span> <span class="fu">fill</span>(<span class="op">-</span><span class="fl">1</span>, (numInds, numHaploblockRegions)) </span>
<span id="cb478-20"><a href="#cb478-20" aria-hidden="true" tabindex="-1"></a><span class="co"># fill object with appropriate genotypes</span></span>
<span id="cb478-21"><a href="#cb478-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numHaploblockRegions</span>
<span id="cb478-22"><a href="#cb478-22" aria-hidden="true" tabindex="-1"></a>    region <span class="op">=</span> haploblockRegions[i]</span>
<span id="cb478-23"><a href="#cb478-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (key, value) <span class="kw">in</span> regionHaplotypeCode_north[region]</span>
<span id="cb478-24"><a href="#cb478-24" aria-hidden="true" tabindex="-1"></a>        command <span class="op">=</span> <span class="st">"""genosSummary[ind_with_metadata_included.</span><span class="sc">$</span>(region)<span class="st">_cluster .== "</span><span class="sc">$</span>(key)<span class="st">", </span><span class="sc">$</span>i<span class="st">] .= """</span> <span class="op">*</span> <span class="fu">string</span>(value)</span>
<span id="cb478-25"><a href="#cb478-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb478-26"><a href="#cb478-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb478-27"><a href="#cb478-27" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb478-28"><a href="#cb478-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-29"><a href="#cb478-29" aria-hidden="true" tabindex="-1"></a>plotGroupsForSummaryNorth <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"plumb_vir"</span>,<span class="st">"plumb"</span>]</span>
<span id="cb478-30"><a href="#cb478-30" aria-hidden="true" tabindex="-1"></a>groupColorsForSummaryNorth <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"purple"</span>,<span class="st">"red"</span>] </span>
<span id="cb478-31"><a href="#cb478-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-32"><a href="#cb478-32" aria-hidden="true" tabindex="-1"></a><span class="fu">plotHaploblockSummary</span>(genosSummary, ind_with_metadata_included,</span>
<span id="cb478-33"><a href="#cb478-33" aria-hidden="true" tabindex="-1"></a>                            plotGroupsForSummaryNorth, groupColorsForSummaryNorth;</span>
<span id="cb478-34"><a href="#cb478-34" aria-hidden="true" tabindex="-1"></a>                            regionNames <span class="op">=</span> haploblockRegions,</span>
<span id="cb478-35"><a href="#cb478-35" aria-hidden="true" tabindex="-1"></a>                            indFontSize <span class="op">=</span> <span class="fl">8</span>, figureSize <span class="op">=</span> (<span class="fl">1200</span>, <span class="fl">1600</span>),</span>
<span id="cb478-36"><a href="#cb478-36" aria-hidden="true" tabindex="-1"></a>                            plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb478-37"><a href="#cb478-37" aria-hidden="true" tabindex="-1"></a>                            indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb478-38"><a href="#cb478-38" aria-hidden="true" tabindex="-1"></a>                            indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-175-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="175">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  1 Child Scene:
    └ Scene (768px, 960px), [0 0 … 0 0; 0 0 … 0 0; … ; 2 2 … 2 2; 1 2 … 1 2], <span class="ansi-bold">100×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
  91 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
  92 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
  93 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
  94 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
  95 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
  96 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
  97 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
  98 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
  99 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 100 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                  35 columns and 79 rows omitted</span>)</pre>
</div>
</div>
</div>
</section>
<section id="make-a-summary-plot-for-the-whole-ring" class="level2">
<h2 class="anchored" data-anchor-id="make-a-summary-plot-for-the-whole-ring">Make a summary plot for the whole ring</h2>
<div id="60465b27" class="cell" data-execution_count="175">
<div class="sourceCode cell-code" id="cb480"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb480-1"><a href="#cb480-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up a code converting integers to colors. </span></span>
<span id="cb480-2"><a href="#cb480-2" aria-hidden="true" tabindex="-1"></a><span class="co"># These will be used for all chromosome regions below.</span></span>
<span id="cb480-3"><a href="#cb480-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-4"><a href="#cb480-4" aria-hidden="true" tabindex="-1"></a>integerToColorCodes <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{Int, String}</span>(</span>
<span id="cb480-5"><a href="#cb480-5" aria-hidden="true" tabindex="-1"></a>    <span class="fl">1</span> <span class="op">=&gt;</span> <span class="st">"blue"</span>, <span class="co"># vir</span></span>
<span id="cb480-6"><a href="#cb480-6" aria-hidden="true" tabindex="-1"></a>    <span class="fl">2</span> <span class="op">=&gt;</span> <span class="st">"turquoise1"</span>, <span class="co"># vir south</span></span>
<span id="cb480-7"><a href="#cb480-7" aria-hidden="true" tabindex="-1"></a>    <span class="fl">3</span> <span class="op">=&gt;</span> <span class="st">"grey"</span>, <span class="co"># nit</span></span>
<span id="cb480-8"><a href="#cb480-8" aria-hidden="true" tabindex="-1"></a>    <span class="fl">4</span> <span class="op">=&gt;</span> <span class="st">"green"</span>, <span class="co"># lud</span></span>
<span id="cb480-9"><a href="#cb480-9" aria-hidden="true" tabindex="-1"></a>    <span class="fl">5</span> <span class="op">=&gt;</span> <span class="st">"yellow"</span>, <span class="co"># troch</span></span>
<span id="cb480-10"><a href="#cb480-10" aria-hidden="true" tabindex="-1"></a>    <span class="fl">6</span> <span class="op">=&gt;</span> <span class="st">"orange"</span>, <span class="co"># obs</span></span>
<span id="cb480-11"><a href="#cb480-11" aria-hidden="true" tabindex="-1"></a>    <span class="fl">7</span> <span class="op">=&gt;</span> <span class="st">"red"</span>, <span class="co"># plumb</span></span>
<span id="cb480-12"><a href="#cb480-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb480-13"><a href="#cb480-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-14"><a href="#cb480-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up a data structure to store the code to converting, for each haploblock region, </span></span>
<span id="cb480-15"><a href="#cb480-15" aria-hidden="true" tabindex="-1"></a><span class="co"># the cluster names to genotype integers corresponding to colors above. This is a dictionary of dictionaries.</span></span>
<span id="cb480-16"><a href="#cb480-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Each genotype will be encoded with a tuple representing the alleles.</span></span>
<span id="cb480-17"><a href="#cb480-17" aria-hidden="true" tabindex="-1"></a>regionHaplotypeCode_all <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String, Dict{String, Tuple{Int, Int}}}</span>(</span>
<span id="cb480-18"><a href="#cb480-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw1A"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>), </span>
<span id="cb480-19"><a href="#cb480-19" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-20"><a href="#cb480-20" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>), </span>
<span id="cb480-21"><a href="#cb480-21" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-22"><a href="#cb480-22" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-23"><a href="#cb480-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-24"><a href="#cb480-24" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-25"><a href="#cb480-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw3"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>), </span>
<span id="cb480-26"><a href="#cb480-26" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLudHet"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-27"><a href="#cb480-27" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-28"><a href="#cb480-28" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_trochObs"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-29"><a href="#cb480-29" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObs"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-30"><a href="#cb480-30" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObsHet"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-31"><a href="#cb480-31" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-32"><a href="#cb480-32" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumbHet"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-33"><a href="#cb480-33" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-34"><a href="#cb480-34" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw4A"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-35"><a href="#cb480-35" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLudHet"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-36"><a href="#cb480-36" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-37"><a href="#cb480-37" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-38"><a href="#cb480-38" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-39"><a href="#cb480-39" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochHet"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-40"><a href="#cb480-40" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch_obsPlumb"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">7</span>),</span>
<span id="cb480-41"><a href="#cb480-41" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obsPlumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-42"><a href="#cb480-42" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obsPlumbHet"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-43"><a href="#cb480-43" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_obsPlumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-44"><a href="#cb480-44" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw13"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"vir"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-45"><a href="#cb480-45" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_lud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">4</span>),</span>
<span id="cb480-46"><a href="#cb480-46" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-47"><a href="#cb480-47" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"lud"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">4</span>),</span>
<span id="cb480-48"><a href="#cb480-48" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"lud_troch"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">5</span>),</span>
<span id="cb480-49"><a href="#cb480-49" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-50"><a href="#cb480-50" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-51"><a href="#cb480-51" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-52"><a href="#cb480-52" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumbHet"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-53"><a href="#cb480-53" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),      </span>
<span id="cb480-54"><a href="#cb480-54" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw15"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-55"><a href="#cb480-55" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-56"><a href="#cb480-56" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-57"><a href="#cb480-57" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-58"><a href="#cb480-58" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-59"><a href="#cb480-59" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-60"><a href="#cb480-60" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-61"><a href="#cb480-61" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw17"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-62"><a href="#cb480-62" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-63"><a href="#cb480-63" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-64"><a href="#cb480-64" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-65"><a href="#cb480-65" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_obs"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">6</span>),</span>
<span id="cb480-66"><a href="#cb480-66" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-67"><a href="#cb480-67" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch_plumb"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">7</span>),</span>
<span id="cb480-68"><a href="#cb480-68" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-69"><a href="#cb480-69" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-70"><a href="#cb480-70" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw18"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-71"><a href="#cb480-71" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-72"><a href="#cb480-72" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-73"><a href="#cb480-73" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-74"><a href="#cb480-74" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-75"><a href="#cb480-75" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs_plumb"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">7</span>),</span>
<span id="cb480-76"><a href="#cb480-76" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-77"><a href="#cb480-77" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-78"><a href="#cb480-78" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw19"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-79"><a href="#cb480-79" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLudHet"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-80"><a href="#cb480-80" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-81"><a href="#cb480-81" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_trochObs"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-82"><a href="#cb480-82" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObs"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-83"><a href="#cb480-83" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObsHet"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-84"><a href="#cb480-84" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObs_plumb"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">7</span>),</span>
<span id="cb480-85"><a href="#cb480-85" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-86"><a href="#cb480-86" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-87"><a href="#cb480-87" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw20"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"vir"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-88"><a href="#cb480-88" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-89"><a href="#cb480-89" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">4</span>),</span>
<span id="cb480-90"><a href="#cb480-90" aria-hidden="true" tabindex="-1"></a>                <span class="st">"ludHet"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">4</span>),</span>
<span id="cb480-91"><a href="#cb480-91" aria-hidden="true" tabindex="-1"></a>                <span class="st">"lud_troch"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">5</span>),</span>
<span id="cb480-92"><a href="#cb480-92" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-93"><a href="#cb480-93" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-94"><a href="#cb480-94" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obs_plumb"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">7</span>),</span>
<span id="cb480-95"><a href="#cb480-95" aria-hidden="true" tabindex="-1"></a>                <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-96"><a href="#cb480-96" aria-hidden="true" tabindex="-1"></a>                <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-97"><a href="#cb480-97" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw26"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-98"><a href="#cb480-98" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-99"><a href="#cb480-99" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-100"><a href="#cb480-100" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-101"><a href="#cb480-101" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-102"><a href="#cb480-102" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs_plumb"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">7</span>),</span>
<span id="cb480-103"><a href="#cb480-103" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-104"><a href="#cb480-104" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-105"><a href="#cb480-105" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gw28"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"virLud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-106"><a href="#cb480-106" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_nit"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">3</span>),</span>
<span id="cb480-107"><a href="#cb480-107" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-108"><a href="#cb480-108" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"virLud_troch"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">5</span>),</span>
<span id="cb480-109"><a href="#cb480-109" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-110"><a href="#cb480-110" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-111"><a href="#cb480-111" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obsHet"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-112"><a href="#cb480-112" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs_plumb"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">7</span>),</span>
<span id="cb480-113"><a href="#cb480-113" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>),</span>
<span id="cb480-114"><a href="#cb480-114" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_plumb"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">7</span>)),</span>
<span id="cb480-115"><a href="#cb480-115" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gwZ"</span> <span class="op">=&gt;</span> <span class="fu">Dict</span>(<span class="st">"vir"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">1</span>),</span>
<span id="cb480-116"><a href="#cb480-116" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"vir_lud"</span><span class="op">=&gt;</span>(<span class="fl">1</span>,<span class="fl">4</span>),</span>
<span id="cb480-117"><a href="#cb480-117" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span><span class="op">=&gt;</span>(<span class="fl">3</span>,<span class="fl">3</span>),</span>
<span id="cb480-118"><a href="#cb480-118" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"lud"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">4</span>),</span>
<span id="cb480-119"><a href="#cb480-119" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"lud_troch"</span><span class="op">=&gt;</span>(<span class="fl">4</span>,<span class="fl">5</span>),</span>
<span id="cb480-120"><a href="#cb480-120" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"troch"</span><span class="op">=&gt;</span>(<span class="fl">5</span>,<span class="fl">5</span>),</span>
<span id="cb480-121"><a href="#cb480-121" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"obs"</span><span class="op">=&gt;</span>(<span class="fl">6</span>,<span class="fl">6</span>),</span>
<span id="cb480-122"><a href="#cb480-122" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span><span class="op">=&gt;</span>(<span class="fl">7</span>,<span class="fl">7</span>))</span>
<span id="cb480-123"><a href="#cb480-123" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb480-124"><a href="#cb480-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-125"><a href="#cb480-125" aria-hidden="true" tabindex="-1"></a>haploblockRegions <span class="op">=</span> [<span class="st">"gwZ"</span>, <span class="st">"gw1A"</span>, <span class="st">"gw3"</span>, <span class="st">"gw4A"</span>, <span class="st">"gw13"</span>, <span class="st">"gw15"</span>, <span class="st">"gw17"</span>, <span class="st">"gw18"</span>, <span class="st">"gw19"</span>, <span class="st">"gw20"</span>,<span class="st">"gw26"</span>, <span class="st">"gw28"</span>]</span>
<span id="cb480-126"><a href="#cb480-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-127"><a href="#cb480-127" aria-hidden="true" tabindex="-1"></a>numHaploblockRegions <span class="op">=</span> <span class="fu">length</span>(haploblockRegions)</span>
<span id="cb480-128"><a href="#cb480-128" aria-hidden="true" tabindex="-1"></a>numInds <span class="op">=</span> <span class="fu">size</span>(ind_with_metadata_included, <span class="fl">1</span>)</span>
<span id="cb480-129"><a href="#cb480-129" aria-hidden="true" tabindex="-1"></a><span class="co"># create genotype object and fill with missing (-1) genotypes</span></span>
<span id="cb480-130"><a href="#cb480-130" aria-hidden="true" tabindex="-1"></a>genosSummary <span class="op">=</span> <span class="fu">fill</span>((<span class="op">-</span><span class="fl">9</span>, <span class="op">-</span><span class="fl">9</span>), numInds, numHaploblockRegions)</span>
<span id="cb480-131"><a href="#cb480-131" aria-hidden="true" tabindex="-1"></a><span class="co"># fill object with appropriate genotypes</span></span>
<span id="cb480-132"><a href="#cb480-132" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numHaploblockRegions</span>
<span id="cb480-133"><a href="#cb480-133" aria-hidden="true" tabindex="-1"></a>    region <span class="op">=</span> haploblockRegions[i]</span>
<span id="cb480-134"><a href="#cb480-134" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (key, value) <span class="kw">in</span> regionHaplotypeCode_all[region]</span>
<span id="cb480-135"><a href="#cb480-135" aria-hidden="true" tabindex="-1"></a>        command <span class="op">=</span> <span class="st">"""genosSummary[ind_with_metadata_included.</span><span class="sc">$</span>(region)<span class="st">_cluster .== "</span><span class="sc">$</span>(key)<span class="st">", </span><span class="sc">$</span>i<span class="st">] .= (</span><span class="sc">$</span>(<span class="fu">string</span>(value))<span class="st">,)"""</span>  <span class="co"># the construction at the end "protects" the tuple within a tuple, so it broadcasts correctly to each element on the left</span></span>
<span id="cb480-136"><a href="#cb480-136" aria-hidden="true" tabindex="-1"></a>        <span class="fu">eval</span>(<span class="bu">Meta</span>.<span class="fu">parse</span>(command)) <span class="co"># this executes the command constructed above</span></span>
<span id="cb480-137"><a href="#cb480-137" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb480-138"><a href="#cb480-138" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb480-139"><a href="#cb480-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-140"><a href="#cb480-140" aria-hidden="true" tabindex="-1"></a>plotGroupsForSummary_all <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb480-141"><a href="#cb480-141" aria-hidden="true" tabindex="-1"></a>groupColorsForSummary_all <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>] </span>
<span id="cb480-142"><a href="#cb480-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-143"><a href="#cb480-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-144"><a href="#cb480-144" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb480-145"><a href="#cb480-145" aria-hidden="true" tabindex="-1"></a><span class="st">    plotHaploblockSummary(integerToColorCodes::Dict{Int, String},</span></span>
<span id="cb480-146"><a href="#cb480-146" aria-hidden="true" tabindex="-1"></a><span class="st">                            genosSummary::Matrix{Tuple{Int64, Int64}},</span></span>
<span id="cb480-147"><a href="#cb480-147" aria-hidden="true" tabindex="-1"></a><span class="st">                            indMetadata,</span></span>
<span id="cb480-148"><a href="#cb480-148" aria-hidden="true" tabindex="-1"></a><span class="st">                            plotGroups, plotGroupColors;</span></span>
<span id="cb480-149"><a href="#cb480-149" aria-hidden="true" tabindex="-1"></a><span class="st">                            regionNames,</span></span>
<span id="cb480-150"><a href="#cb480-150" aria-hidden="true" tabindex="-1"></a><span class="st">                            indFontSize=10, figureSize=(1200, 1200),</span></span>
<span id="cb480-151"><a href="#cb480-151" aria-hidden="true" tabindex="-1"></a><span class="st">                            plotTitle = nothing,</span></span>
<span id="cb480-152"><a href="#cb480-152" aria-hidden="true" tabindex="-1"></a><span class="st">                            indColorLeftProvided = false,</span></span>
<span id="cb480-153"><a href="#cb480-153" aria-hidden="true" tabindex="-1"></a><span class="st">                            indColorRightProvided = false)</span></span>
<span id="cb480-154"><a href="#cb480-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-155"><a href="#cb480-155" aria-hidden="true" tabindex="-1"></a><span class="st">Construct a genotype-by-individual plot, with option to filter out SNPs with too much missing data. </span></span>
<span id="cb480-156"><a href="#cb480-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-157"><a href="#cb480-157" aria-hidden="true" tabindex="-1"></a><span class="st">In this version, more than two haplotype alleles can be plotted, using colors provided according to the first argument.</span></span>
<span id="cb480-158"><a href="#cb480-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-159"><a href="#cb480-159" aria-hidden="true" tabindex="-1"></a><span class="st">​# Arguments</span></span>
<span id="cb480-160"><a href="#cb480-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-161"><a href="#cb480-161" aria-hidden="true" tabindex="-1"></a><span class="st">- `integerToColorCodes`: The code matching integer haploblock types to colors.</span></span>
<span id="cb480-162"><a href="#cb480-162" aria-hidden="true" tabindex="-1"></a><span class="st">- `genosSummary`: Matrix containing summary genotype data (individuals in rows, loci in columns), with each genotype represented by a tuple of 2 integers.</span></span>
<span id="cb480-163"><a href="#cb480-163" aria-hidden="true" tabindex="-1"></a><span class="st">- `indMetadata`: Matrix of metadata for individuals; must contain `Fst_group` and `plot_order` columns.</span></span>
<span id="cb480-164"><a href="#cb480-164" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotGroups`: Vector of group names to include in plot.</span></span>
<span id="cb480-165"><a href="#cb480-165" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotGroupColors`: Vector of plotting colors corresponding to the groups.</span></span>
<span id="cb480-166"><a href="#cb480-166" aria-hidden="true" tabindex="-1"></a><span class="st">- `regionNames`: Optional; Names of the genotyped regions.</span></span>
<span id="cb480-167"><a href="#cb480-167" aria-hidden="true" tabindex="-1"></a><span class="st">- `indFontSize`: Optional; the font size of the individual ID labels.</span></span>
<span id="cb480-168"><a href="#cb480-168" aria-hidden="true" tabindex="-1"></a><span class="st">- `figureSize`: Optional; the size of the figure; default is `(1200, 1200)`.  </span></span>
<span id="cb480-169"><a href="#cb480-169" aria-hidden="true" tabindex="-1"></a><span class="st">- `plotTitle`: Optional; default will make a title. For no title, set to `""`.</span></span>
<span id="cb480-170"><a href="#cb480-170" aria-hidden="true" tabindex="-1"></a><span class="st">- `indColorLeftProvided`: Optional; Default is `false`. Set to `true` if there is a column labeled `indColorLeft` in the metadata providing color of each individual for plotting on left side.</span></span>
<span id="cb480-171"><a href="#cb480-171" aria-hidden="true" tabindex="-1"></a><span class="st">- `indColorRightProvided`: Optional; same as above but for right side (requires `indColorRight` column in metadata).</span></span>
<span id="cb480-172"><a href="#cb480-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-173"><a href="#cb480-173" aria-hidden="true" tabindex="-1"></a><span class="st"># Notes</span></span>
<span id="cb480-174"><a href="#cb480-174" aria-hidden="true" tabindex="-1"></a><span class="st">Returns a tuple containing:</span></span>
<span id="cb480-175"><a href="#cb480-175" aria-hidden="true" tabindex="-1"></a><span class="st">- the figure</span></span>
<span id="cb480-176"><a href="#cb480-176" aria-hidden="true" tabindex="-1"></a><span class="st">- the plotted genotypes</span></span>
<span id="cb480-177"><a href="#cb480-177" aria-hidden="true" tabindex="-1"></a><span class="st">- the sorted metadata matrix for the plotted individuals</span></span>
<span id="cb480-178"><a href="#cb480-178" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb480-179"><a href="#cb480-179" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">plotHaploblockSummary</span>(integerToColorCodes<span class="op">::</span><span class="dt">Dict{Int, String}</span>,</span>
<span id="cb480-180"><a href="#cb480-180" aria-hidden="true" tabindex="-1"></a>                                genosSummary<span class="op">::</span><span class="dt">Matrix{Tuple{Int64, Int64}}</span>, </span>
<span id="cb480-181"><a href="#cb480-181" aria-hidden="true" tabindex="-1"></a>                                indMetadata,</span>
<span id="cb480-182"><a href="#cb480-182" aria-hidden="true" tabindex="-1"></a>                                plotGroups, plotGroupColors;</span>
<span id="cb480-183"><a href="#cb480-183" aria-hidden="true" tabindex="-1"></a>                                regionNames <span class="op">=</span> <span class="cn">nothing</span>,        </span>
<span id="cb480-184"><a href="#cb480-184" aria-hidden="true" tabindex="-1"></a>                                indFontSize<span class="op">=</span><span class="fl">10</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">1200</span>),</span>
<span id="cb480-185"><a href="#cb480-185" aria-hidden="true" tabindex="-1"></a>                                plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb480-186"><a href="#cb480-186" aria-hidden="true" tabindex="-1"></a>                                indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb480-187"><a href="#cb480-187" aria-hidden="true" tabindex="-1"></a>                                indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span>
<span id="cb480-188"><a href="#cb480-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-189"><a href="#cb480-189" aria-hidden="true" tabindex="-1"></a>    numRegions <span class="op">=</span> <span class="fu">size</span>(genosSummary, <span class="fl">2</span>)</span>
<span id="cb480-190"><a href="#cb480-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-191"><a href="#cb480-191" aria-hidden="true" tabindex="-1"></a>    genosSummary_subset <span class="op">=</span> genosSummary[indMetadata.Fst_group <span class="op">.∈</span> <span class="fu">Ref</span>(plotGroups), <span class="op">:</span>]</span>
<span id="cb480-192"><a href="#cb480-192" aria-hidden="true" tabindex="-1"></a>    indMetadata_subset <span class="op">=</span> indMetadata[indMetadata.Fst_group <span class="op">.∈</span> <span class="fu">Ref</span>(plotGroups), <span class="op">:</span>]</span>
<span id="cb480-193"><a href="#cb480-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-194"><a href="#cb480-194" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Choose sorting order by plot_order column in input metadata file</span></span>
<span id="cb480-195"><a href="#cb480-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-196"><a href="#cb480-196" aria-hidden="true" tabindex="-1"></a>    sorted_genosSummary_subset <span class="op">=</span> genosSummary_subset[<span class="fu">sortperm</span>(indMetadata_subset.plot_order, rev<span class="op">=</span><span class="cn">false</span>), <span class="op">:</span>]</span>
<span id="cb480-197"><a href="#cb480-197" aria-hidden="true" tabindex="-1"></a>    numInds <span class="op">=</span> <span class="fu">size</span>(sorted_genosSummary_subset, <span class="fl">1</span>)</span>
<span id="cb480-198"><a href="#cb480-198" aria-hidden="true" tabindex="-1"></a>    sorted_indMetadata_subset <span class="op">=</span> indMetadata_subset[<span class="fu">sortperm</span>(indMetadata_subset.plot_order, rev<span class="op">=</span><span class="cn">false</span>), <span class="op">:</span>]</span>
<span id="cb480-199"><a href="#cb480-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-200"><a href="#cb480-200" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set up the plot window:</span></span>
<span id="cb480-201"><a href="#cb480-201" aria-hidden="true" tabindex="-1"></a>    f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>(size<span class="op">=</span>figureSize)</span>
<span id="cb480-202"><a href="#cb480-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-203"><a href="#cb480-203" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">isnothing</span>(plotTitle)</span>
<span id="cb480-204"><a href="#cb480-204" aria-hidden="true" tabindex="-1"></a>        plotTitle <span class="op">=</span> <span class="st">"Summary of </span><span class="sc">$</span>numRegions<span class="st"> haploblock genotypes for </span><span class="sc">$</span>numInds<span class="st"> individuals"</span></span>
<span id="cb480-205"><a href="#cb480-205" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span> </span>
<span id="cb480-206"><a href="#cb480-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-207"><a href="#cb480-207" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set up the main Axis: </span></span>
<span id="cb480-208"><a href="#cb480-208" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb480-209"><a href="#cb480-209" aria-hidden="true" tabindex="-1"></a>        title <span class="op">=</span> plotTitle,</span>
<span id="cb480-210"><a href="#cb480-210" aria-hidden="true" tabindex="-1"></a>        titlesize<span class="op">=</span><span class="fl">30</span>,</span>
<span id="cb480-211"><a href="#cb480-211" aria-hidden="true" tabindex="-1"></a>        limits<span class="op">=</span>(<span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.09</span> <span class="op">*</span> (numRegions), <span class="fl">0.5</span> <span class="op">+</span> <span class="fl">1.09</span> <span class="op">*</span> (numRegions),</span>
<span id="cb480-212"><a href="#cb480-212" aria-hidden="true" tabindex="-1"></a>            <span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.3</span> <span class="op">*</span> numInds, <span class="fl">0.5</span> <span class="op">+</span> numInds)</span>
<span id="cb480-213"><a href="#cb480-213" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb480-214"><a href="#cb480-214" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hidedecorations!</span>(ax) <span class="co"># hide background lattice and axis labels</span></span>
<span id="cb480-215"><a href="#cb480-215" aria-hidden="true" tabindex="-1"></a>    <span class="fu">hidespines!</span>(ax) <span class="co"># hide box around plot</span></span>
<span id="cb480-216"><a href="#cb480-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-217"><a href="#cb480-217" aria-hidden="true" tabindex="-1"></a>    genotypeColors <span class="op">=</span> [<span class="st">"#3f007d"</span>, <span class="st">"#807dba"</span>, <span class="st">"#dadaeb"</span>, <span class="st">"grey50"</span>]  <span class="co"># purple shades from colorbrewer</span></span>
<span id="cb480-218"><a href="#cb480-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-219"><a href="#cb480-219" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot evenly spaced by SNP order along chromosome:</span></span>
<span id="cb480-220"><a href="#cb480-220" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make top part of fig (genotypes for individuals)</span></span>
<span id="cb480-221"><a href="#cb480-221" aria-hidden="true" tabindex="-1"></a>    labelCushion <span class="op">=</span> numRegions <span class="op">/</span> <span class="fl">100</span></span>
<span id="cb480-222"><a href="#cb480-222" aria-hidden="true" tabindex="-1"></a>    label_x_left <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> labelCushion</span>
<span id="cb480-223"><a href="#cb480-223" aria-hidden="true" tabindex="-1"></a>    label_x_right <span class="op">=</span> <span class="fl">0.5</span> <span class="op">+</span> numRegions <span class="op">+</span> labelCushion</span>
<span id="cb480-224"><a href="#cb480-224" aria-hidden="true" tabindex="-1"></a>    colorBoxCushion <span class="op">=</span> <span class="fl">0.07</span> <span class="op">*</span> numRegions</span>
<span id="cb480-225"><a href="#cb480-225" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_left <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> colorBoxCushion</span>
<span id="cb480-226"><a href="#cb480-226" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_right <span class="op">=</span> <span class="fl">0.5</span> <span class="op">+</span> numRegions <span class="op">+</span> colorBoxCushion</span>
<span id="cb480-227"><a href="#cb480-227" aria-hidden="true" tabindex="-1"></a>    boxWidth <span class="op">=</span> <span class="fl">0.005</span> <span class="op">*</span> numRegions <span class="op">*</span> <span class="fl">2</span></span>
<span id="cb480-228"><a href="#cb480-228" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_left <span class="op">=</span> [<span class="op">-</span>boxWidth, <span class="op">-</span>boxWidth, boxWidth, boxWidth, <span class="op">-</span>boxWidth] <span class="op">.+</span> groupColorBox_x_left</span>
<span id="cb480-229"><a href="#cb480-229" aria-hidden="true" tabindex="-1"></a>    groupColorBox_x_right <span class="op">=</span> [<span class="op">-</span>boxWidth, <span class="op">-</span>boxWidth, boxWidth, boxWidth, <span class="op">-</span>boxWidth] <span class="op">.+</span> groupColorBox_x_right</span>
<span id="cb480-230"><a href="#cb480-230" aria-hidden="true" tabindex="-1"></a>    groupColorBox_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb480-231"><a href="#cb480-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-232"><a href="#cb480-232" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb480-233"><a href="#cb480-233" aria-hidden="true" tabindex="-1"></a>        y <span class="op">=</span> numInds <span class="op">+</span> <span class="fl">1</span> <span class="op">-</span> i  <span class="co"># y is location for plotting; this reverses order of plot top-bottom</span></span>
<span id="cb480-234"><a href="#cb480-234" aria-hidden="true" tabindex="-1"></a>        labelText <span class="op">=</span> <span class="fu">last</span>(<span class="fu">split</span>(sorted_indMetadata_subset.ID[i], <span class="st">"_"</span>))  <span class="co"># this gets the last part of the sample ID (usually the main ID part)</span></span>
<span id="cb480-235"><a href="#cb480-235" aria-hidden="true" tabindex="-1"></a>        <span class="co"># put sample label on left side:</span></span>
<span id="cb480-236"><a href="#cb480-236" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(label_x_left, y; text<span class="op">=</span>labelText, align<span class="op">=</span>(<span class="op">:</span>right, <span class="op">:</span>center), fontsize<span class="op">=</span>indFontSize)</span>
<span id="cb480-237"><a href="#cb480-237" aria-hidden="true" tabindex="-1"></a>        <span class="co"># put sample label on left side:</span></span>
<span id="cb480-238"><a href="#cb480-238" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(label_x_right, y; text<span class="op">=</span>labelText, align<span class="op">=</span>(<span class="op">:</span>left, <span class="op">:</span>center), fontsize<span class="op">=</span>indFontSize)</span>
<span id="cb480-239"><a href="#cb480-239" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> indColorLeftProvided</span>
<span id="cb480-240"><a href="#cb480-240" aria-hidden="true" tabindex="-1"></a>            boxColorLeft <span class="op">=</span> sorted_indMetadata_subset.indColorLeft[i]</span>
<span id="cb480-241"><a href="#cb480-241" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb480-242"><a href="#cb480-242" aria-hidden="true" tabindex="-1"></a>            boxColorLeft <span class="op">=</span> plotGroupColors[<span class="fu">findfirst</span>(plotGroups <span class="op">.==</span> sorted_indMetadata_subset.Fst_group[i])]</span>
<span id="cb480-243"><a href="#cb480-243" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb480-244"><a href="#cb480-244" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> indColorRightProvided</span>
<span id="cb480-245"><a href="#cb480-245" aria-hidden="true" tabindex="-1"></a>            boxColorRight <span class="op">=</span> sorted_indMetadata_subset.indColorRight[i]</span>
<span id="cb480-246"><a href="#cb480-246" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb480-247"><a href="#cb480-247" aria-hidden="true" tabindex="-1"></a>            boxColorRight <span class="op">=</span> plotGroupColors[<span class="fu">findfirst</span>(plotGroups <span class="op">.==</span> sorted_indMetadata_subset.Fst_group[i])]</span>
<span id="cb480-248"><a href="#cb480-248" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb480-249"><a href="#cb480-249" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.(groupColorBox_x_left, (y <span class="op">.+</span> groupColorBox_y)), color<span class="op">=</span>boxColorLeft)</span>
<span id="cb480-250"><a href="#cb480-250" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.(groupColorBox_x_right, (y <span class="op">.+</span> groupColorBox_y)), color<span class="op">=</span>boxColorRight)</span>
<span id="cb480-251"><a href="#cb480-251" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb480-252"><a href="#cb480-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-253"><a href="#cb480-253" aria-hidden="true" tabindex="-1"></a>    <span class="co"># generate my own plotting symbol (a rectangle)</span></span>
<span id="cb480-254"><a href="#cb480-254" aria-hidden="true" tabindex="-1"></a>    box_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb480-255"><a href="#cb480-255" aria-hidden="true" tabindex="-1"></a>    <span class="co">#box_x = [-0.5, -0.5, 0.5, 0.5, -0.5]</span></span>
<span id="cb480-256"><a href="#cb480-256" aria-hidden="true" tabindex="-1"></a>    box_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb480-257"><a href="#cb480-257" aria-hidden="true" tabindex="-1"></a>    <span class="co"># generate triangles for plotting heterozygotes</span></span>
<span id="cb480-258"><a href="#cb480-258" aria-hidden="true" tabindex="-1"></a>    triangle1_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb480-259"><a href="#cb480-259" aria-hidden="true" tabindex="-1"></a>    <span class="co">#triangle1_x = [-0.5, -0.5, 0.5, -0.5]</span></span>
<span id="cb480-260"><a href="#cb480-260" aria-hidden="true" tabindex="-1"></a>    triangle1_y <span class="op">=</span> [<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="fl">0.4</span>]</span>
<span id="cb480-261"><a href="#cb480-261" aria-hidden="true" tabindex="-1"></a>    triangle2_x <span class="op">=</span> [<span class="op">-</span><span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="fl">0.45</span>, <span class="op">-</span><span class="fl">0.45</span>]</span>
<span id="cb480-262"><a href="#cb480-262" aria-hidden="true" tabindex="-1"></a>    <span class="co">#triangle2_x = [-0.5, 0.5, 0.5, -0.5]</span></span>
<span id="cb480-263"><a href="#cb480-263" aria-hidden="true" tabindex="-1"></a>    triangle2_y <span class="op">=</span> [<span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>]</span>
<span id="cb480-264"><a href="#cb480-264" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cycle through individuals, graphing each type of genotype:</span></span>
<span id="cb480-265"><a href="#cb480-265" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb480-266"><a href="#cb480-266" aria-hidden="true" tabindex="-1"></a>        y <span class="op">=</span> numInds <span class="op">+</span> <span class="fl">1</span> <span class="op">-</span> i  <span class="co"># y is location for plotting; this reverses order of plot top-bottom</span></span>
<span id="cb480-267"><a href="#cb480-267" aria-hidden="true" tabindex="-1"></a>        <span class="co">#CairoMakie.lines!([0.5, numRegions + 0.5], [y, y], color="grey40") # for lines across the individual rows</span></span>
<span id="cb480-268"><a href="#cb480-268" aria-hidden="true" tabindex="-1"></a>        <span class="co"># cycle through regions for this individual</span></span>
<span id="cb480-269"><a href="#cb480-269" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRegions</span>
<span id="cb480-270"><a href="#cb480-270" aria-hidden="true" tabindex="-1"></a>            genotype <span class="op">=</span> sorted_genosSummary_subset[i, j]</span>
<span id="cb480-271"><a href="#cb480-271" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> genotype[<span class="fl">1</span>] <span class="op">==</span> genotype[<span class="fl">2</span>]  <span class="co"># homozygous</span></span>
<span id="cb480-272"><a href="#cb480-272" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((j <span class="op">.+</span> box_x), (y <span class="op">.+</span> box_y)), color<span class="op">=</span>integerToColorCodes[genotype[<span class="fl">1</span>]])</span>
<span id="cb480-273"><a href="#cb480-273" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span> <span class="co"># heterozygous</span></span>
<span id="cb480-274"><a href="#cb480-274" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((j <span class="op">.+</span> triangle1_x), (y <span class="op">.+</span> triangle1_y)), color<span class="op">=</span>integerToColorCodes[genotype[<span class="fl">1</span>]])</span>
<span id="cb480-275"><a href="#cb480-275" aria-hidden="true" tabindex="-1"></a>                CairoMakie.<span class="fu">poly!</span>(<span class="fu">Point2f</span>.((j <span class="op">.+</span> triangle2_x), (y <span class="op">.+</span> triangle2_y)), color<span class="op">=</span>integerToColorCodes[genotype[<span class="fl">2</span>]])</span>
<span id="cb480-276"><a href="#cb480-276" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb480-277"><a href="#cb480-277" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb480-278"><a href="#cb480-278" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb480-279"><a href="#cb480-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-280"><a href="#cb480-280" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">isnothing</span>(regionNames)</span>
<span id="cb480-281"><a href="#cb480-281" aria-hidden="true" tabindex="-1"></a>        regionNames <span class="op">=</span> <span class="fu">string</span>.(<span class="fl">1</span><span class="op">:</span>numRegions)</span>
<span id="cb480-282"><a href="#cb480-282" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb480-283"><a href="#cb480-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-284"><a href="#cb480-284" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make labels on lower part</span></span>
<span id="cb480-285"><a href="#cb480-285" aria-hidden="true" tabindex="-1"></a>    y_label <span class="op">=</span> <span class="fl">0.5</span> <span class="op">-</span> <span class="fl">0.025</span>numInds</span>
<span id="cb480-286"><a href="#cb480-286" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numRegions</span>
<span id="cb480-287"><a href="#cb480-287" aria-hidden="true" tabindex="-1"></a>        CairoMakie.<span class="fu">text!</span>(i, y_label; text <span class="op">=</span> regionNames[i], align<span class="op">=</span>(<span class="op">:</span>center, <span class="op">:</span>center), fontsize<span class="op">=</span><span class="fl">24</span>)</span>
<span id="cb480-288"><a href="#cb480-288" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb480-289"><a href="#cb480-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-290"><a href="#cb480-290" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display</span>(f)</span>
<span id="cb480-291"><a href="#cb480-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-292"><a href="#cb480-292" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> f, sorted_genosSummary_subset, sorted_indMetadata_subset</span>
<span id="cb480-293"><a href="#cb480-293" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb480-294"><a href="#cb480-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-295"><a href="#cb480-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb480-296"><a href="#cb480-296" aria-hidden="true" tabindex="-1"></a><span class="fu">plotHaploblockSummary</span>(integerToColorCodes,</span>
<span id="cb480-297"><a href="#cb480-297" aria-hidden="true" tabindex="-1"></a>                            genosSummary, </span>
<span id="cb480-298"><a href="#cb480-298" aria-hidden="true" tabindex="-1"></a>                            ind_with_metadata_included,</span>
<span id="cb480-299"><a href="#cb480-299" aria-hidden="true" tabindex="-1"></a>                            plotGroupsForSummary_all, groupColorsForSummary_all;</span>
<span id="cb480-300"><a href="#cb480-300" aria-hidden="true" tabindex="-1"></a>                            regionNames <span class="op">=</span> haploblockRegions,</span>
<span id="cb480-301"><a href="#cb480-301" aria-hidden="true" tabindex="-1"></a>                            indFontSize <span class="op">=</span> <span class="fl">8</span>, figureSize <span class="op">=</span> (<span class="fl">1000</span>, <span class="fl">2000</span>),</span>
<span id="cb480-302"><a href="#cb480-302" aria-hidden="true" tabindex="-1"></a>                            plotTitle <span class="op">=</span> <span class="cn">nothing</span>,</span>
<span id="cb480-303"><a href="#cb480-303" aria-hidden="true" tabindex="-1"></a>                            indColorLeftProvided <span class="op">=</span> <span class="cn">false</span>,</span>
<span id="cb480-304"><a href="#cb480-304" aria-hidden="true" tabindex="-1"></a>                            indColorRightProvided <span class="op">=</span> <span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-176-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="176">
<div class="ansi-escaped-output">
<pre>(Scene (768px, 960px):
  0 Plots
  1 Child Scene:
    └ Scene (768px, 960px), [(1, 1) (1, 1) … (1, 1) (1, 1); (1, 1) (1, 1) … (1, 1) (1, 1); … ; (7, 7) (7, 7) … (7, 7) (7, 7); (7, 7) (1, 7) … (1, 7) (1, 7)], <span class="ansi-bold">257×38 DataFrame</span>
<span class="ansi-bold"> Row </span>│<span class="ansi-bold"> ind                       </span><span class="ansi-bold"> ID                        </span><span class="ansi-bold"> location </span><span class="ansi-bold"> group  </span> ⋯
     │<span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String                    </span><span class="ansi-bright-black-fg"> String7  </span><span class="ansi-bright-black-fg"> String1</span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_JF12G04  GW_Armando_plate1_JF12G04  ST_vi     vir     ⋯
   2 │ GW_Armando_plate2_JF03G01  GW_Armando_plate2_JF03G01  ST_vi     vir_mis
   3 │ GW_Armando_plate2_JF30G01  GW_Armando_plate2_JF30G01  ST_vi     vir_mis
   4 │ GW_Lane5_STvi1             GW_Lane5_STvi1             ST_vi     vir
   5 │ GW_Lane5_STvi2             GW_Lane5_STvi2             ST_vi     vir     ⋯
   6 │ GW_Lane5_STvi3             GW_Lane5_STvi3             ST_vi     vir
   7 │ GW_Armando_plate1_JF16G01  GW_Armando_plate1_JF16G01  DV_vi     plumb_v
   8 │ GW_Armando_plate2_JF16G02  GW_Armando_plate2_JF16G02  DV_vi     plumb_v
   9 │ GW_Armando_plate2_JE31G01  GW_Armando_plate2_JE31G01  VB_vi     vir_mis ⋯
  10 │ GW_Armando_plate2_JF03G02  GW_Armando_plate2_JF03G02  VB_vi     vir_mis
  11 │ GW_Lane5_YK11              GW_Lane5_YK11              YK        vir
  ⋮  │             ⋮                          ⋮                 ⋮          ⋮   ⋱
 248 │ GW_Armando_plate2_JF24G01  GW_Armando_plate2_JF24G01  VB        plumb
 249 │ GW_Armando_plate2_JF25G01  GW_Armando_plate2_JF25G01  VB        plumb   ⋯
 250 │ GW_Armando_plate1_JG02G02  GW_Armando_plate1_JG02G02  PR        plumb
 251 │ GW_Armando_plate1_JG02G04  GW_Armando_plate1_JG02G04  PR        plumb
 252 │ GW_Armando_plate2_JG01G01  GW_Armando_plate2_JG01G01  PR        plumb
 253 │ GW_Armando_plate2_JG02G01  GW_Armando_plate2_JG02G01  PR        plumb   ⋯
 254 │ GW_Armando_plate2_JG02G03  GW_Armando_plate2_JG02G03  PR        plumb
 255 │ GW_Lane5_SL1               GW_Lane5_SL1               SL        plumb
 256 │ GW_Lane5_SL2               GW_Lane5_SL2               SL        plumb
 257 │ GW_Armando_plate1_JF10G03  GW_Armando_plate1_JF10G03  ST        plumb_v ⋯
<span class="ansi-cyan-fg">                                                 35 columns and 236 rows omitted</span>)</pre>
</div>
</div>
</div>
<section id="section-1" class="level25">
<p class="heading"></p>
<p>EDIT / REVISE THIS PART BELOW TO PROPERLY INCORPORATE THE Z CHROM FILTERING</p>
</section>
</section>
<section id="produce-fst-plot-across-genome" class="level2">
<h2 class="anchored" data-anchor-id="produce-fst-plot-across-genome">Produce Fst plot across genome</h2>
<section id="calculate-allele-freqs-and-sample-sizes-use-column-fst_group" class="level4">
<h4 class="anchored" data-anchor-id="calculate-allele-freqs-and-sample-sizes-use-column-fst_group">Calculate allele freqs and sample sizes (use column Fst_group)</h4>
<div id="8c5c6cb8" class="cell" data-execution_count="176">
<div class="sourceCode cell-code" id="cb482"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb482-1"><a href="#cb482-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb482-2"><a href="#cb482-2" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_included, ind_with_metadata_included.Fst_group, groups)</span>
<span id="cb482-3"><a href="#cb482-3" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes</code></pre>
</div>
</div>
</section>
<section id="calculate-fst-for-each-snp" class="level4">
<h4 class="anchored" data-anchor-id="calculate-fst-for-each-snp">calculate Fst for each SNP</h4>
<div id="885878ce" class="cell" data-execution_count="177">
<div class="sourceCode cell-code" id="cb484"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb484-1"><a href="#cb484-1" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb484-2"><a href="#cb484-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated Fst values</code></pre>
</div>
</div>
</section>
<section id="make-list-of-main-scaffolds-to-include-in-fst-plot-across-genome" class="level4">
<h4 class="anchored" data-anchor-id="make-list-of-main-scaffolds-to-include-in-fst-plot-across-genome">Make list of main scaffolds to include in Fst plot across genome:</h4>
<div id="499e7504" class="cell" data-execution_count="178">
<div class="sourceCode cell-code" id="cb486"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb486-1"><a href="#cb486-1" aria-hidden="true" tabindex="-1"></a>scaffolds_for_Fst <span class="op">=</span> <span class="st">"gw"</span> <span class="op">.*</span> <span class="fu">string</span>.(<span class="fu">vcat</span>(<span class="fl">1</span>, <span class="st">"1A"</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, <span class="st">"4A"</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">17</span><span class="op">:</span><span class="fl">28</span>, <span class="st">"Z"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="179">
<pre><code>30-element Vector{String}:
 "gw1"
 "gw1A"
 "gw2"
 "gw3"
 "gw4"
 "gw4A"
 "gw5"
 "gw6"
 "gw7"
 "gw8"
 "gw9"
 "gw10"
 "gw11"
 ⋮
 "gw18"
 "gw19"
 "gw20"
 "gw21"
 "gw22"
 "gw23"
 "gw24"
 "gw25"
 "gw26"
 "gw27"
 "gw28"
 "gwZ"</code></pre>
</div>
</div>
</section>
<section id="calculate-windowed-fst" class="level4">
<h4 class="anchored" data-anchor-id="calculate-windowed-fst">calculate windowed Fst</h4>
<p>This is calculated according to Weir&amp;Cockerham1984 (with sample size and pop number correction), calculated as windowed numerator over windowed denominator, in whole windows starting on left side of chromosome.</p>
<div id="046e2dc2" class="cell" data-execution_count="179">
<div class="sourceCode cell-code" id="cb488"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb488-1"><a href="#cb488-1" aria-hidden="true" tabindex="-1"></a>windowSize <span class="op">=</span> <span class="fl">500</span></span>
<span id="cb488-2"><a href="#cb488-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb488-3"><a href="#cb488-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate windowed Fst across all scaffolds:</span></span>
<span id="cb488-4"><a href="#cb488-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb488-5"><a href="#cb488-5" aria-hidden="true" tabindex="-1"></a>windowed_pos_all <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="dt">String</span>[], mean_position <span class="op">=</span> <span class="dt">Float64</span>[])</span>
<span id="cb488-6"><a href="#cb488-6" aria-hidden="true" tabindex="-1"></a>windowed_Fst_all <span class="op">=</span> <span class="fu">Array</span><span class="dt">{Float32, 2}</span>(<span class="cn">undef</span>, <span class="fu">size</span>(FstNumerator, <span class="fl">1</span>), <span class="fl">0</span>)</span>
<span id="cb488-7"><a href="#cb488-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrom <span class="kw">in</span> scaffolds_for_Fst</span>
<span id="cb488-8"><a href="#cb488-8" aria-hidden="true" tabindex="-1"></a>    regionText <span class="op">=</span> <span class="fu">string</span>(<span class="st">"chr"</span>, chrom)</span>
<span id="cb488-9"><a href="#cb488-9" aria-hidden="true" tabindex="-1"></a>    loci_selection <span class="op">=</span> (pos_SNP_filtered.chrom <span class="op">.==</span> chrom)</span>
<span id="cb488-10"><a href="#cb488-10" aria-hidden="true" tabindex="-1"></a>    pos_region <span class="op">=</span> pos_SNP_filtered[loci_selection, <span class="op">:</span>]</span>
<span id="cb488-11"><a href="#cb488-11" aria-hidden="true" tabindex="-1"></a>    FstNumerator_region <span class="op">=</span> FstNumerator[<span class="op">:</span>, loci_selection]</span>
<span id="cb488-12"><a href="#cb488-12" aria-hidden="true" tabindex="-1"></a>    FstDenominator_region <span class="op">=</span> FstDenominator[<span class="op">:</span>, loci_selection]</span>
<span id="cb488-13"><a href="#cb488-13" aria-hidden="true" tabindex="-1"></a>    windowedPos, windowedFst <span class="op">=</span> <span class="fu">getWindowedFst</span>(FstNumerator_region, FstDenominator_region, pos_region, windowSize)</span>
<span id="cb488-14"><a href="#cb488-14" aria-hidden="true" tabindex="-1"></a>    windowed_pos_chrom <span class="op">=</span> <span class="fu">DataFrame</span>(chrom <span class="op">=</span> <span class="fu">repeat</span>([chrom], <span class="fu">length</span>(windowedPos)), mean_position <span class="op">=</span> windowedPos)</span>
<span id="cb488-15"><a href="#cb488-15" aria-hidden="true" tabindex="-1"></a>    windowed_pos_all <span class="op">=</span> <span class="fu">vcat</span>(windowed_pos_all, windowed_pos_chrom)</span>
<span id="cb488-16"><a href="#cb488-16" aria-hidden="true" tabindex="-1"></a>    windowed_Fst_all <span class="op">=</span> <span class="fu">hcat</span>(windowed_Fst_all, windowedFst)</span>
<span id="cb488-17"><a href="#cb488-17" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb488-18"><a href="#cb488-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb488-19"><a href="#cb488-19" aria-hidden="true" tabindex="-1"></a><span class="co"># The below is just a test plot, showing nothing useful really (as it overlaps all chromosomes onto one x axis):</span></span>
<span id="cb488-20"><a href="#cb488-20" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(windowed_pos_all.mean_position, windowed_Fst_all[1,:])</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The above has produced windowed Fst values across the whole genome, for each population comparison. These are stored in <code>windowed_Fst_all</code> and the location info is stored in <code>windowed_pos_all</code>.</p>
<p>Now make a plot of windowed Fst across all scaffolds:</p>
<div id="11676ea4" class="cell" data-execution_count="180">
<div class="sourceCode cell-code" id="cb489"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb489-1"><a href="#cb489-1" aria-hidden="true" tabindex="-1"></a>scaffolds_to_plot <span class="op">=</span> scaffolds_for_Fst</span>
<span id="cb489-2"><a href="#cb489-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb489-3"><a href="#cb489-3" aria-hidden="true" tabindex="-1"></a>groupsToPlotFst <span class="op">=</span> [<span class="st">"vir_troch_LN"</span>, <span class="st">"troch_LN_plumb"</span>, <span class="st">"vir_plumb"</span>]</span>
<span id="cb489-4"><a href="#cb489-4" aria-hidden="true" tabindex="-1"></a>groupColorsFst <span class="op">=</span> [<span class="st">"green3"</span>, <span class="st">"orange"</span>, <span class="st">"purple"</span>]</span>
<span id="cb489-5"><a href="#cb489-5" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb489-6"><a href="#cb489-6" aria-hidden="true" tabindex="-1"></a>figHandle_GenomeFst3 <span class="op">=</span> <span class="fu">plotGenomeFst</span>(scaffolds_to_plot, </span>
<span id="cb489-7"><a href="#cb489-7" aria-hidden="true" tabindex="-1"></a>                                    windowed_Fst_all,</span>
<span id="cb489-8"><a href="#cb489-8" aria-hidden="true" tabindex="-1"></a>                                    pairwiseNamesFst,</span>
<span id="cb489-9"><a href="#cb489-9" aria-hidden="true" tabindex="-1"></a>                                    windowed_pos_all,</span>
<span id="cb489-10"><a href="#cb489-10" aria-hidden="true" tabindex="-1"></a>                                    groupsToPlotFst,</span>
<span id="cb489-11"><a href="#cb489-11" aria-hidden="true" tabindex="-1"></a>                                    groupColorsFst;</span>
<span id="cb489-12"><a href="#cb489-12" aria-hidden="true" tabindex="-1"></a>                                    lineTransparency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span id="cb489-13"><a href="#cb489-13" aria-hidden="true" tabindex="-1"></a>                                    fillTransparency <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span id="cb489-14"><a href="#cb489-14" aria-hidden="true" tabindex="-1"></a>                                    figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">1200</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[["gw1", "gw4A", "gw6"], ["gw1A", "gw4", "gw9"], ["gw2", "gw8"], ["gw3", "gw5"], ["gw7", "gw10", "gw11", "gw12", "gw13", "gw14"], ["gw15", "gw17", "gw18", "gw19", "gw20", "gw21", "gw22", "gw23", "gw24", "gw25"], ["gw26", "gw27", "gw28", "gwZ"]]</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-181-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="181">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-181-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Now do one with just the vir_plumb connection</p>
<div id="bac1f9e0" class="cell" data-execution_count="181">
<div class="sourceCode cell-code" id="cb492"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb492-1"><a href="#cb492-1" aria-hidden="true" tabindex="-1"></a>groupsToPlotFst <span class="op">=</span> [<span class="st">"vir_plumb"</span>]</span>
<span id="cb492-2"><a href="#cb492-2" aria-hidden="true" tabindex="-1"></a>groupColorsFst <span class="op">=</span> [<span class="st">"purple"</span>]</span>
<span id="cb492-3"><a href="#cb492-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb492-4"><a href="#cb492-4" aria-hidden="true" tabindex="-1"></a>figHandle_GenomeFst1 <span class="op">=</span> <span class="fu">plotGenomeFst</span>(scaffolds_to_plot, </span>
<span id="cb492-5"><a href="#cb492-5" aria-hidden="true" tabindex="-1"></a>                                    windowed_Fst_all,</span>
<span id="cb492-6"><a href="#cb492-6" aria-hidden="true" tabindex="-1"></a>                                    pairwiseNamesFst,</span>
<span id="cb492-7"><a href="#cb492-7" aria-hidden="true" tabindex="-1"></a>                                    windowed_pos_all,</span>
<span id="cb492-8"><a href="#cb492-8" aria-hidden="true" tabindex="-1"></a>                                    groupsToPlotFst,</span>
<span id="cb492-9"><a href="#cb492-9" aria-hidden="true" tabindex="-1"></a>                                    groupColorsFst;</span>
<span id="cb492-10"><a href="#cb492-10" aria-hidden="true" tabindex="-1"></a>                                    lineTransparency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span id="cb492-11"><a href="#cb492-11" aria-hidden="true" tabindex="-1"></a>                                    fillTransparency <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span id="cb492-12"><a href="#cb492-12" aria-hidden="true" tabindex="-1"></a>                                    figureSize<span class="op">=</span>(<span class="fl">1200</span>, <span class="fl">800</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[["gw1", "gw4A", "gw6"], ["gw1A", "gw4", "gw9"], ["gw2", "gw8"], ["gw3", "gw5"], ["gw7", "gw10", "gw11", "gw12", "gw13", "gw14"], ["gw15", "gw17", "gw18", "gw19", "gw20", "gw21", "gw22", "gw23", "gw24", "gw25"], ["gw26", "gw27", "gw28", "gwZ"]]</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-182-output-3.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="182">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-182-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="plot-vishet-vs.-fst" class="level2">
<h2 class="anchored" data-anchor-id="plot-vishet-vs.-fst">Plot ViSHet vs.&nbsp;Fst</h2>
<div id="ecb61dd5" class="cell" data-execution_count="182">
<div class="sourceCode cell-code" id="cb495"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb495-1"><a href="#cb495-1" aria-hidden="true" tabindex="-1"></a>groupsToCompareUsingFst <span class="op">=</span> <span class="st">"vir_plumb"</span></span>
<span id="cb495-2"><a href="#cb495-2" aria-hidden="true" tabindex="-1"></a>FstRow <span class="op">=</span> <span class="fu">findfirst</span>(pairwiseNamesFst <span class="op">.==</span> groupsToCompareUsingFst)</span>
<span id="cb495-3"><a href="#cb495-3" aria-hidden="true" tabindex="-1"></a>windowedFstValues <span class="op">=</span> windowed_Fst_all[FstRow, <span class="op">:</span>]</span>
<span id="cb495-4"><a href="#cb495-4" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(windowedFstValues, windowed_HetVariance_all)</span></span>
<span id="cb495-5"><a href="#cb495-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb495-6"><a href="#cb495-6" aria-hidden="true" tabindex="-1"></a>fillOpacity <span class="op">=</span> <span class="fl">0.3</span></span>
<span id="cb495-7"><a href="#cb495-7" aria-hidden="true" tabindex="-1"></a>lineOpacity <span class="op">=</span> <span class="fl">0.8</span></span>
<span id="cb495-8"><a href="#cb495-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb495-9"><a href="#cb495-9" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> <span class="fu">Figure</span>()</span>
<span id="cb495-10"><a href="#cb495-10" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb495-11"><a href="#cb495-11" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"windowed Fst"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb495-12"><a href="#cb495-12" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"windowed VisHet"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>)</span>
<span id="cb495-13"><a href="#cb495-13" aria-hidden="true" tabindex="-1"></a><span class="co"># hidedecorations!(ax, label = false, ticklabels = false, ticks = false) # hide background lattice</span></span>
<span id="cb495-14"><a href="#cb495-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot!</span>(ax, windowedFstValues, windowed_HetVariance_all,</span>
<span id="cb495-15"><a href="#cb495-15" aria-hidden="true" tabindex="-1"></a>    marker <span class="op">=</span> <span class="op">:</span>circle, color <span class="op">=</span> (<span class="st">"black"</span>, fillOpacity), markersize <span class="op">=</span> <span class="fl">8</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, lineOpacity))</span>
<span id="cb495-16"><a href="#cb495-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb495-17"><a href="#cb495-17" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span>
<span id="cb495-18"><a href="#cb495-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb495-19"><a href="#cb495-19" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(windowedFstValues)</span>
<span id="cb495-20"><a href="#cb495-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb495-21"><a href="#cb495-21" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(windowed_HetVariance_all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-183-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238
┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="183">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-183-output-4.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="examine-chromosome-4a-large-haploblock-region-lhbr-with-invariant-sites-included" class="level2">
<h2 class="anchored" data-anchor-id="examine-chromosome-4a-large-haploblock-region-lhbr-with-invariant-sites-included">Examine chromosome 4A Large HaploBlock Region (LHBR) with invariant sites included</h2>
<p>Before running below, I need to change format of 012NA file to # Before running below, changed 012NA file back into 012minus1 file, using commands like below, so can be read as integer:</p>
<div class="sourceCode" id="cb498"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb498-1"><a href="#cb498-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> /Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw4A.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab.012NA <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/NA/-1/g'</span> <span class="op">&gt;</span> /Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw4A.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab.012minus1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="3f1ffa23" class="cell" data-execution_count="183">
<div class="sourceCode cell-code" id="cb499"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb499-1"><a href="#cb499-1" aria-hidden="true" tabindex="-1"></a>baseName <span class="op">=</span> <span class="st">"/Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw4A.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab"</span></span>
<span id="cb499-2"><a href="#cb499-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load metadata</span></span>
<span id="cb499-3"><a href="#cb499-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span>
<span id="cb499-4"><a href="#cb499-4" aria-hidden="true" tabindex="-1"></a>metadata_chr4A <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(metadataFile)) <span class="co"># the CSV.File function interprets the correct delimiter</span></span>
<span id="cb499-5"><a href="#cb499-5" aria-hidden="true" tabindex="-1"></a>num_metadata_cols_chr4A <span class="op">=</span> <span class="fu">ncol</span>(metadata_chr4A)</span>
<span id="cb499-6"><a href="#cb499-6" aria-hidden="true" tabindex="-1"></a>num_individuals_chr4A <span class="op">=</span> <span class="fu">nrow</span>(metadata_chr4A) </span>
<span id="cb499-7"><a href="#cb499-7" aria-hidden="true" tabindex="-1"></a><span class="co"># read in individual names for this dataset</span></span>
<span id="cb499-8"><a href="#cb499-8" aria-hidden="true" tabindex="-1"></a>individuals_file_name_chr4A <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012.indv"</span>)</span>
<span id="cb499-9"><a href="#cb499-9" aria-hidden="true" tabindex="-1"></a>ind_chr4A <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(individuals_file_name_chr4A; header<span class="op">=</span>[<span class="st">"ind"</span>], types<span class="op">=</span>[<span class="dt">String</span>])) </span>
<span id="cb499-10"><a href="#cb499-10" aria-hidden="true" tabindex="-1"></a>indNum_chr4A <span class="op">=</span> <span class="fu">size</span>(ind_chr4A, <span class="fl">1</span>) <span class="co"># number of individuals</span></span>
<span id="cb499-11"><a href="#cb499-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> num_individuals_chr4A <span class="op">!=</span> indNum_chr4A</span>
<span id="cb499-12"><a href="#cb499-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: number of rows in metadata file different than number of individuals in .indv file"</span>)</span>
<span id="cb499-13"><a href="#cb499-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb499-14"><a href="#cb499-14" aria-hidden="true" tabindex="-1"></a><span class="co"># read in position data for this dataset</span></span>
<span id="cb499-15"><a href="#cb499-15" aria-hidden="true" tabindex="-1"></a>position_file_name_chr4A <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012.pos"</span>)</span>
<span id="cb499-16"><a href="#cb499-16" aria-hidden="true" tabindex="-1"></a>pos_chr4A <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(position_file_name_chr4A; header<span class="op">=</span>[<span class="st">"chrom"</span>, <span class="st">"position"</span>], types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int</span>]))</span>
<span id="cb499-17"><a href="#cb499-17" aria-hidden="true" tabindex="-1"></a><span class="co"># read in genotype data</span></span>
<span id="cb499-18"><a href="#cb499-18" aria-hidden="true" tabindex="-1"></a>genotype_file_name_chr4A <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012minus1"</span>) </span>
<span id="cb499-19"><a href="#cb499-19" aria-hidden="true" tabindex="-1"></a><span class="pp">@time</span> <span class="cf">if</span> <span class="fl">1</span> <span class="op">&lt;=</span> indNum_chr4A <span class="op">&lt;=</span> <span class="fl">127</span>   </span>
<span id="cb499-20"><a href="#cb499-20" aria-hidden="true" tabindex="-1"></a>    geno_chr4A <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name_chr4A, <span class="ch">'\t'</span>, <span class="dt">Int8</span>, <span class="ch">'\n'</span>); <span class="co"># this has been sped up dramatically, by first converting "NA" to -1</span></span>
<span id="cb499-21"><a href="#cb499-21" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> <span class="fl">128</span> <span class="op">&lt;=</span> indNum_chr4A <span class="op">&lt;=</span> <span class="fl">32767</span></span>
<span id="cb499-22"><a href="#cb499-22" aria-hidden="true" tabindex="-1"></a>    geno_chr4A <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name_chr4A, <span class="ch">'\t'</span>, <span class="dt">Int16</span>, <span class="ch">'\n'</span>); <span class="co"># this needed for first column, which is number of individual; Int16 not much slower on import than Int8</span></span>
<span id="cb499-23"><a href="#cb499-23" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb499-24"><a href="#cb499-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Error: Number of individuals in .indv appears outside of range from 1 to 32767"</span>)</span>
<span id="cb499-25"><a href="#cb499-25" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb499-26"><a href="#cb499-26" aria-hidden="true" tabindex="-1"></a>loci_count_chr4A <span class="op">=</span> <span class="fu">size</span>(geno_chr4A, <span class="fl">2</span>) <span class="op">-</span> <span class="fl">1</span>   <span class="co"># because the first column is not a SNP (just a count from zero)</span></span>
<span id="cb499-27"><a href="#cb499-27" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">string</span>(<span class="st">"Read in genotypic data at "</span>, loci_count_chr4A,<span class="st">" loci for "</span>, indNum_chr4A, <span class="st">" individuals. </span><span class="sc">\n</span><span class="st">"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 66.922885 seconds (340.65 M allocations: 9.787 GiB, 73.95% gc time, 1.13% compilation time)
Read in genotypic data at 364640 loci for 310 individuals. </code></pre>
</div>
</div>
<section id="check-that-individuals-are-same-in-genotype-data-and-metadata" class="level3">
<h3 class="anchored" data-anchor-id="check-that-individuals-are-same-in-genotype-data-and-metadata">Check that individuals are same in genotype data and metadata</h3>
<div id="680f5703" class="cell" data-execution_count="184">
<div class="sourceCode cell-code" id="cb501"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb501-1"><a href="#cb501-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr4A <span class="op">=</span> <span class="fu">hcat</span>(ind_chr4A, metadata_chr4A)</span>
<span id="cb501-2"><a href="#cb501-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(ind_with_metadata_chr4A)</span>
<span id="cb501-3"><a href="#cb501-3" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>()  <span class="co"># prints a line break </span></span>
<span id="cb501-4"><a href="#cb501-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">isequal</span>(ind_with_metadata_chr4A.ind, ind_with_metadata_chr4A.ID)</span>
<span id="cb501-5"><a href="#cb501-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly."</span>)</span>
<span id="cb501-6"><a href="#cb501-6" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb501-7"><a href="#cb501-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: names of individuals in metadata file and genotype ind file do not completely match."</span>)</span>
<span id="cb501-8"><a href="#cb501-8" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>310×6 DataFrame
 Row │ ind                             ID                              location  group           Fst_group       plot_order 
     │ String                          String31                        String7   String15        String15        Float64    
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1           GW_Armando_plate1_AB1           AB        vir             vir                 20.01
   2 │ GW_Armando_plate1_JF07G02       GW_Armando_plate1_JF07G02       ST        plumb           plumb              108.0
   3 │ GW_Armando_plate1_JF07G03       GW_Armando_plate1_JF07G03       ST        plumb           plumb              109.0
   4 │ GW_Armando_plate1_JF07G04       GW_Armando_plate1_JF07G04       ST        plumb           plumb              110.0
   5 │ GW_Armando_plate1_JF08G02       GW_Armando_plate1_JF08G02       ST        plumb           plumb              111.0
   6 │ GW_Armando_plate1_JF09G01       GW_Armando_plate1_JF09G01       ST        plumb           plumb              112.0
   7 │ GW_Armando_plate1_JF09G02       GW_Armando_plate1_JF09G02       ST        plumb           plumb              113.0
   8 │ GW_Armando_plate1_JF10G03       GW_Armando_plate1_JF10G03       ST        plumb_vir       plumb_vir          170.0
   9 │ GW_Armando_plate1_JF11G01       GW_Armando_plate1_JF11G01       ST        plumb           plumb              114.0
  10 │ GW_Armando_plate1_JF12G01       GW_Armando_plate1_JF12G01       ST        plumb           plumb              115.0
  11 │ GW_Armando_plate1_JF12G02       GW_Armando_plate1_JF12G02       ST        plumb           plumb              116.0
  12 │ GW_Armando_plate1_JF12G04       GW_Armando_plate1_JF12G04       ST_vi     vir             vir                 24.001
  13 │ GW_Armando_plate1_JF13G01       GW_Armando_plate1_JF13G01       ST        plumb           plumb              117.0
  14 │ GW_Armando_plate1_JF15G03       GW_Armando_plate1_JF15G03       DV        plumb           plumb              103.0
  15 │ GW_Armando_plate1_JF16G01       GW_Armando_plate1_JF16G01       DV_vi     plumb_vir       vir                 24.041
  16 │ GW_Armando_plate1_JF20G01       GW_Armando_plate1_JF20G01       MB        plumb           plumb               94.0
  17 │ GW_Armando_plate1_JF22G01       GW_Armando_plate1_JF22G01       MB        plumb           plumb               95.0
  18 │ GW_Armando_plate1_JF23G01       GW_Armando_plate1_JF23G01       VB        plumb           plumb               98.0
  19 │ GW_Armando_plate1_JF23G02       GW_Armando_plate1_JF23G02       VB        plumb           plumb               99.0
  20 │ GW_Armando_plate1_JF24G02       GW_Armando_plate1_JF24G02       VB        plumb           plumb              100.0
  21 │ GW_Armando_plate1_JF26G01       GW_Armando_plate1_JF26G01       ST        plumb           plumb              118.0
  22 │ GW_Armando_plate1_JF27G01       GW_Armando_plate1_JF27G01       ST        plumb           plumb              119.0
  23 │ GW_Armando_plate1_JF29G01       GW_Armando_plate1_JF29G01       ST        plumb           plumb              120.0
  24 │ GW_Armando_plate1_JF29G02       GW_Armando_plate1_JF29G02       ST        plumb           plumb              121.0
  25 │ GW_Armando_plate1_JF29G03       GW_Armando_plate1_JF29G03       ST        plumb           plumb              122.0
  26 │ GW_Armando_plate1_JG02G02       GW_Armando_plate1_JG02G02       PR        plumb           plumb              145.0
  27 │ GW_Armando_plate1_JG02G04       GW_Armando_plate1_JG02G04       PR        plumb           plumb              146.0
  28 │ GW_Armando_plate1_JG08G01       GW_Armando_plate1_JG08G01       ST        plumb           plumb              123.0
  29 │ GW_Armando_plate1_JG08G02       GW_Armando_plate1_JG08G02       ST        plumb           plumb              124.0
  30 │ GW_Armando_plate1_JG10G01       GW_Armando_plate1_JG10G01       ST        plumb           plumb              125.0
  31 │ GW_Armando_plate1_JG12G01       GW_Armando_plate1_JG12G01       ST        plumb           plumb              126.0
  32 │ GW_Armando_plate1_JG17G01       GW_Armando_plate1_JG17G01       ST        plumb_vir       plumb              127.0
  33 │ GW_Armando_plate1_NO_BC_TTGW05  GW_Armando_plate1_NO_BC_TTGW05  blank     blank           blank              -99.0
  34 │ GW_Armando_plate1_NO_DNA        GW_Armando_plate1_NO_DNA        blank     blank           blank              -99.0
  35 │ GW_Armando_plate1_RF20G01       GW_Armando_plate1_RF20G01       BJ        obs_plumb       plumb_BJ            77.501
  36 │ GW_Armando_plate1_RF29G02       GW_Armando_plate1_RF29G02       BJ        obs_plumb       plumb_BJ            77.502
  37 │ GW_Armando_plate1_TL3           GW_Armando_plate1_TL3           TL        vir             vir                 11.01
  38 │ GW_Armando_plate1_TTGW01        GW_Armando_plate1_TTGW01        MN        troch_MN        troch_west          53.0
  39 │ GW_Armando_plate1_TTGW05_rep1   GW_Armando_plate1_TTGW05_rep1   MN_rep    troch_MN_rep    troch_west_rep      53.0
  40 │ GW_Armando_plate1_TTGW05_rep2   GW_Armando_plate1_TTGW05_rep2   MN        troch_MN        troch_west          53.0
  41 │ GW_Armando_plate1_TTGW06        GW_Armando_plate1_TTGW06        SU        lud_Sukhto      lud_central         47.0
  42 │ GW_Armando_plate1_TTGW07        GW_Armando_plate1_TTGW07        SU        lud_Sukhto      lud_central         47.0
  43 │ GW_Armando_plate1_TTGW10        GW_Armando_plate1_TTGW10        SU        lud_Sukhto      lud_central         47.0
  44 │ GW_Armando_plate1_TTGW11        GW_Armando_plate1_TTGW11        SU        lud_Sukhto      lud_central         47.0
  45 │ GW_Armando_plate1_TTGW13        GW_Armando_plate1_TTGW13        TH        lud_Thallighar  lud_central         43.0
  46 │ GW_Armando_plate1_TTGW17        GW_Armando_plate1_TTGW17        TH        lud_Thallighar  lud_central         43.0
  47 │ GW_Armando_plate1_TTGW19        GW_Armando_plate1_TTGW19        TH        lud_Thallighar  lud_central         43.0
  48 │ GW_Armando_plate1_TTGW21        GW_Armando_plate1_TTGW21        SR        lud_Sural       lud_central         45.0
  49 │ GW_Armando_plate1_TTGW22        GW_Armando_plate1_TTGW22        SR        lud_Sural       lud_central         45.0
  50 │ GW_Armando_plate1_TTGW23        GW_Armando_plate1_TTGW23        SR        lud_Sural       lud_central         45.0
  51 │ GW_Armando_plate1_TTGW29        GW_Armando_plate1_TTGW29        SR        lud_Sural       lud_central         45.0
  52 │ GW_Armando_plate1_TTGW52        GW_Armando_plate1_TTGW52        NG        lud_Nainaghar   lud_central         49.0
  53 │ GW_Armando_plate1_TTGW53        GW_Armando_plate1_TTGW53        NG        lud_Nainaghar   lud_central         49.0
  54 │ GW_Armando_plate1_TTGW55        GW_Armando_plate1_TTGW55        NG        lud_Nainaghar   lud_central         49.0
  55 │ GW_Armando_plate1_TTGW57        GW_Armando_plate1_TTGW57        NG        lud_Nainaghar   lud_central         49.0
  56 │ GW_Armando_plate1_TTGW58        GW_Armando_plate1_TTGW58        NG        lud_Nainaghar   lud_central         49.0
  57 │ GW_Armando_plate1_TTGW59        GW_Armando_plate1_TTGW59        NG        lud_Nainaghar   lud_central         49.0
  58 │ GW_Armando_plate1_TTGW63        GW_Armando_plate1_TTGW63        SP        lud_Spiti       troch_west          55.0
  59 │ GW_Armando_plate1_TTGW64        GW_Armando_plate1_TTGW64        SP        lud_Spiti       troch_west          55.0
  60 │ GW_Armando_plate1_TTGW65        GW_Armando_plate1_TTGW65        SP        lud_Spiti       troch_west          55.0
  61 │ GW_Armando_plate1_TTGW66        GW_Armando_plate1_TTGW66        SP        lud_Spiti       troch_west          55.0
  62 │ GW_Armando_plate1_TTGW68        GW_Armando_plate1_TTGW68        SP        lud_Spiti       troch_west          55.0
  63 │ GW_Armando_plate1_TTGW70        GW_Armando_plate1_TTGW70        SA        lud_Sathrundi   lud_Sath            41.0
  64 │ GW_Armando_plate1_TTGW71        GW_Armando_plate1_TTGW71        SA        lud_Sathrundi   lud_Sath            41.0
  65 │ GW_Armando_plate1_TTGW72        GW_Armando_plate1_TTGW72        SA        lud_Sathrundi   lud_Sath            41.0
  66 │ GW_Armando_plate1_TTGW74        GW_Armando_plate1_TTGW74        SA        lud_Sathrundi   lud_Sath            41.0
  67 │ GW_Armando_plate1_TTGW78        GW_Armando_plate1_TTGW78        SA        lud_Sathrundi   lud_Sath            41.0
  68 │ GW_Armando_plate1_TTGW_15_05    GW_Armando_plate1_TTGW_15_05    SR        lud_Sural       lud_central         45.0
  69 │ GW_Armando_plate1_TTGW_15_07    GW_Armando_plate1_TTGW_15_07    SR        lud_Sural       lud_central         45.0
  70 │ GW_Armando_plate1_TTGW_15_08    GW_Armando_plate1_TTGW_15_08    SR        lud_Sural       lud_central         45.0
  71 │ GW_Armando_plate1_TTGW_15_09    GW_Armando_plate1_TTGW_15_09    SR        lud_Sural       lud_central         45.0
  72 │ GW_Armando_plate1_UY1           GW_Armando_plate1_UY1           UY        plumb           plumb               87.0
  73 │ GW_Armando_plate2_IL2           GW_Armando_plate2_IL2           IL_rep    plumb_rep       plumb_rep           84.0
  74 │ GW_Armando_plate2_JE31G01       GW_Armando_plate2_JE31G01       VB_vi     vir_misID       vir                 24.002
  75 │ GW_Armando_plate2_JF03G01       GW_Armando_plate2_JF03G01       ST_vi     vir_misID       vir                 24.003
  76 │ GW_Armando_plate2_JF03G02       GW_Armando_plate2_JF03G02       VB_vi     vir_misID       vir                 24.004
  77 │ GW_Armando_plate2_JF07G01       GW_Armando_plate2_JF07G01       ST        plumb           plumb              128.0
  78 │ GW_Armando_plate2_JF08G04       GW_Armando_plate2_JF08G04       ST        plumb           plumb              129.0
  79 │ GW_Armando_plate2_JF10G02       GW_Armando_plate2_JF10G02       ST        plumb           plumb              130.0
  80 │ GW_Armando_plate2_JF11G02       GW_Armando_plate2_JF11G02       ST        plumb           plumb              131.0
  81 │ GW_Armando_plate2_JF12G03       GW_Armando_plate2_JF12G03       ST        plumb           plumb              132.0
  82 │ GW_Armando_plate2_JF12G05       GW_Armando_plate2_JF12G05       ST        plumb           plumb              133.0
  83 │ GW_Armando_plate2_JF13G02       GW_Armando_plate2_JF13G02       ST        plumb           plumb              134.0
  84 │ GW_Armando_plate2_JF14G01       GW_Armando_plate2_JF14G01       DV        plumb           plumb              104.0
  85 │ GW_Armando_plate2_JF14G02       GW_Armando_plate2_JF14G02       DV        plumb           plumb              105.0
  86 │ GW_Armando_plate2_JF15G01       GW_Armando_plate2_JF15G01       DV        plumb           plumb              106.0
  87 │ GW_Armando_plate2_JF15G02       GW_Armando_plate2_JF15G02       DV        plumb           plumb              107.0
  88 │ GW_Armando_plate2_JF16G02       GW_Armando_plate2_JF16G02       DV_vi     plumb_vir       vir                 24.042
  89 │ GW_Armando_plate2_JF19G01       GW_Armando_plate2_JF19G01       MB        plumb           plumb               96.0
  90 │ GW_Armando_plate2_JF20G02       GW_Armando_plate2_JF20G02       MB        plumb           plumb               97.0
  91 │ GW_Armando_plate2_JF24G01       GW_Armando_plate2_JF24G01       VB        plumb           plumb              101.0
  92 │ GW_Armando_plate2_JF24G03       GW_Armando_plate2_JF24G03       ST        plumb           plumb              135.0
  93 │ GW_Armando_plate2_JF25G01       GW_Armando_plate2_JF25G01       VB        plumb           plumb              102.0
  94 │ GW_Armando_plate2_JF26G02       GW_Armando_plate2_JF26G02       ST        plumb           plumb              136.0
  95 │ GW_Armando_plate2_JF27G02       GW_Armando_plate2_JF27G02       ST        plumb           plumb              137.0
  96 │ GW_Armando_plate2_JF30G01       GW_Armando_plate2_JF30G01       ST_vi     vir_misID       vir                 24.005
  97 │ GW_Armando_plate2_JG01G01       GW_Armando_plate2_JG01G01       PR        plumb           plumb              147.0
  98 │ GW_Armando_plate2_JG02G01       GW_Armando_plate2_JG02G01       PR        plumb           plumb              148.0
  99 │ GW_Armando_plate2_JG02G03       GW_Armando_plate2_JG02G03       PR        plumb           plumb              149.0
 100 │ GW_Armando_plate2_JG10G02       GW_Armando_plate2_JG10G02       ST        plumb           plumb              138.0
 101 │ GW_Armando_plate2_JG10G03       GW_Armando_plate2_JG10G03       ST        plumb           plumb              139.0
 102 │ GW_Armando_plate2_JG12G02       GW_Armando_plate2_JG12G02       ST        plumb           plumb              140.0
 103 │ GW_Armando_plate2_JG12G03       GW_Armando_plate2_JG12G03       ST        plumb           plumb              141.0
 104 │ GW_Armando_plate2_LN11          GW_Armando_plate2_LN11          LN_rep    troch_LN_rep    troch_LN_rep        65.01
 105 │ GW_Armando_plate2_LN2           GW_Armando_plate2_LN2           LN        troch_LN        troch_LN            58.01
 106 │ GW_Armando_plate2_NO_BC_TTGW05  GW_Armando_plate2_NO_BC_TTGW05  blank     blank           blank              -99.0
 107 │ GW_Armando_plate2_NO_DNA        GW_Armando_plate2_NO_DNA        blank     blank           blank              -99.0
 108 │ GW_Armando_plate2_RF29G01       GW_Armando_plate2_RF29G01       BJ        obs_plumb       plumb_BJ            77.503
 109 │ GW_Armando_plate2_TTGW02        GW_Armando_plate2_TTGW02        MN        troch_MN        troch_west          53.0
 110 │ GW_Armando_plate2_TTGW03        GW_Armando_plate2_TTGW03        MN        troch_MN        troch_west          53.0
 111 │ GW_Armando_plate2_TTGW05_rep3   GW_Armando_plate2_TTGW05_rep3   MN_rep    troch_MN_rep    troch_west_rep      53.0
 112 │ GW_Armando_plate2_TTGW05_rep4   GW_Armando_plate2_TTGW05_rep4   MN_rep    troch_MN_rep    troch_west_rep      53.0
 113 │ GW_Armando_plate2_TTGW08        GW_Armando_plate2_TTGW08        SU        lud_Sukhto      lud_central         47.0
 114 │ GW_Armando_plate2_TTGW09        GW_Armando_plate2_TTGW09        SU        lud_Sukhto      lud_central         47.0
 115 │ GW_Armando_plate2_TTGW12        GW_Armando_plate2_TTGW12        TH        lud_Thallighar  lud_central         43.0
 116 │ GW_Armando_plate2_TTGW14        GW_Armando_plate2_TTGW14        TH        lud_Thallighar  lud_central         43.0
 117 │ GW_Armando_plate2_TTGW15        GW_Armando_plate2_TTGW15        TH        lud_Thallighar  lud_central         43.0
 118 │ GW_Armando_plate2_TTGW16        GW_Armando_plate2_TTGW16        TH        lud_Thallighar  lud_central         43.0
 119 │ GW_Armando_plate2_TTGW18        GW_Armando_plate2_TTGW18        TH        lud_Thallighar  lud_central         43.0
 120 │ GW_Armando_plate2_TTGW20        GW_Armando_plate2_TTGW20        SR        lud_Sural       lud_central         45.0
 121 │ GW_Armando_plate2_TTGW24        GW_Armando_plate2_TTGW24        SR        lud_Sural       lud_central         45.0
 122 │ GW_Armando_plate2_TTGW25        GW_Armando_plate2_TTGW25        SR        lud_Sural       lud_central         45.0
 123 │ GW_Armando_plate2_TTGW27        GW_Armando_plate2_TTGW27        SR        lud_Sural       lud_central         45.0
 124 │ GW_Armando_plate2_TTGW28        GW_Armando_plate2_TTGW28        SR        lud_Sural       lud_central         45.0
 125 │ GW_Armando_plate2_TTGW50        GW_Armando_plate2_TTGW50        NG        lud_Nainaghar   lud_central         49.0
 126 │ GW_Armando_plate2_TTGW51        GW_Armando_plate2_TTGW51        NG        lud_Nainaghar   lud_central         49.0
 127 │ GW_Armando_plate2_TTGW54        GW_Armando_plate2_TTGW54        NG        lud_Nainaghar   lud_central         49.0
 128 │ GW_Armando_plate2_TTGW56        GW_Armando_plate2_TTGW56        NG        lud_Nainaghar   lud_central         49.0
 129 │ GW_Armando_plate2_TTGW60        GW_Armando_plate2_TTGW60        SP        lud_Spiti       troch_west          55.0
 130 │ GW_Armando_plate2_TTGW61        GW_Armando_plate2_TTGW61        SP        lud_Spiti       troch_west          55.0
 131 │ GW_Armando_plate2_TTGW62        GW_Armando_plate2_TTGW62        SP        lud_Spiti       troch_west          55.0
 132 │ GW_Armando_plate2_TTGW67        GW_Armando_plate2_TTGW67        SP        lud_Spiti       troch_west          55.0
 133 │ GW_Armando_plate2_TTGW69        GW_Armando_plate2_TTGW69        SP        lud_Spiti       troch_west          55.0
 134 │ GW_Armando_plate2_TTGW73        GW_Armando_plate2_TTGW73        SA        lud_Sathrundi   lud_Sath            41.0
 135 │ GW_Armando_plate2_TTGW75        GW_Armando_plate2_TTGW75        SA        lud_Sathrundi   lud_Sath            41.0
 136 │ GW_Armando_plate2_TTGW77        GW_Armando_plate2_TTGW77        SA        lud_Sathrundi   lud_Sath            41.0
 137 │ GW_Armando_plate2_TTGW79        GW_Armando_plate2_TTGW79        SA        lud_Sathrundi   lud_Sath            41.0
 138 │ GW_Armando_plate2_TTGW80        GW_Armando_plate2_TTGW80        SA        lud_Sathrundi   lud_Sath            41.0
 139 │ GW_Armando_plate2_TTGW_15_01    GW_Armando_plate2_TTGW_15_01    SR        lud_Sural       lud_central         45.0
 140 │ GW_Armando_plate2_TTGW_15_02    GW_Armando_plate2_TTGW_15_02    SR        lud_Sural       lud_central         45.0
 141 │ GW_Armando_plate2_TTGW_15_03    GW_Armando_plate2_TTGW_15_03    SR        lud_Sural       lud_central         45.0
 142 │ GW_Armando_plate2_TTGW_15_04    GW_Armando_plate2_TTGW_15_04    SR        lud_Sural       lud_central         45.0
 143 │ GW_Armando_plate2_TTGW_15_06    GW_Armando_plate2_TTGW_15_06    SR        lud_Sural       lud_central         45.0
 144 │ GW_Armando_plate2_TTGW_15_10    GW_Armando_plate2_TTGW_15_10    SR        lud_Sural       lud_central         45.0
 145 │ GW_Lane5_AA1                    GW_Lane5_AA1                    AA        vir_S           vir_S               25.0
 146 │ GW_Lane5_AA10                   GW_Lane5_AA10                   AA        vir_S           vir_S               33.0
 147 │ GW_Lane5_AA11                   GW_Lane5_AA11                   AA        vir_S           vir_S               34.0
 148 │ GW_Lane5_AA3                    GW_Lane5_AA3                    AA        vir_S           vir_S               26.0
 149 │ GW_Lane5_AA4                    GW_Lane5_AA4                    AA        vir_S           vir_S               27.0
 150 │ GW_Lane5_AA5                    GW_Lane5_AA5                    AA        vir_S           vir_S               28.0
 151 │ GW_Lane5_AA6                    GW_Lane5_AA6                    AA        vir_S           vir_S               29.0
 152 │ GW_Lane5_AA7                    GW_Lane5_AA7                    AA        vir_S           vir_S               30.0
 153 │ GW_Lane5_AA8                    GW_Lane5_AA8                    AA        vir_S           vir_S               31.0
 154 │ GW_Lane5_AA9                    GW_Lane5_AA9                    AA        vir_S           vir_S               32.0
 155 │ GW_Lane5_AB1                    GW_Lane5_AB1                    AB_rep    vir_rep         vir_rep             20.0
 156 │ GW_Lane5_AB2                    GW_Lane5_AB2                    AB        vir             vir                 21.0
 157 │ GW_Lane5_AN1                    GW_Lane5_AN1                    AN        plumb           plumb               80.0
 158 │ GW_Lane5_AN2                    GW_Lane5_AN2                    AN        plumb           plumb               81.0
 159 │ GW_Lane5_BK2                    GW_Lane5_BK2                    BK        plumb           plumb               78.0
 160 │ GW_Lane5_BK3                    GW_Lane5_BK3                    BK        plumb           plumb               79.0
 161 │ GW_Lane5_DA2                    GW_Lane5_DA2                    XN        obs             obs                 73.0
 162 │ GW_Lane5_DA3                    GW_Lane5_DA3                    XN        obs             obs                 74.0
 163 │ GW_Lane5_DA4                    GW_Lane5_DA4                    XN        obs             obs                 75.0
 164 │ GW_Lane5_DA6                    GW_Lane5_DA6                    XN        obs             low_reads           76.0
 165 │ GW_Lane5_DA7                    GW_Lane5_DA7                    XN        obs             obs                 77.0
 166 │ GW_Lane5_EM1                    GW_Lane5_EM1                    EM        troch_EM        troch_EM            72.0
 167 │ GW_Lane5_IL1                    GW_Lane5_IL1                    IL        plumb           plumb               82.0
 168 │ GW_Lane5_IL2                    GW_Lane5_IL2                    IL_rep    plumb_rep       plumb_rep           85.0
 169 │ GW_Lane5_IL4                    GW_Lane5_IL4                    IL        plumb           plumb               83.0
 170 │ GW_Lane5_KS1                    GW_Lane5_KS1                    OV        lud_KS          lud_KS              40.0
 171 │ GW_Lane5_KS2                    GW_Lane5_KS2                    OV        lud_KS          lud_KS              40.0
 172 │ GW_Lane5_LN1                    GW_Lane5_LN1                    LN        troch_LN        troch_LN            57.0
 173 │ GW_Lane5_LN10                   GW_Lane5_LN10                   LN        troch_LN        troch_LN            64.0
 174 │ GW_Lane5_LN11                   GW_Lane5_LN11                   LN        troch_LN        troch_LN            65.0
 175 │ GW_Lane5_LN12                   GW_Lane5_LN12                   LN        troch_LN        troch_LN            66.0
 176 │ GW_Lane5_LN14                   GW_Lane5_LN14                   LN        troch_LN        troch_LN            67.0
 177 │ GW_Lane5_LN16                   GW_Lane5_LN16                   LN        troch_LN        troch_LN            68.0
 178 │ GW_Lane5_LN18                   GW_Lane5_LN18                   LN        troch_LN        troch_LN            69.0
 179 │ GW_Lane5_LN19                   GW_Lane5_LN19                   LN        troch_LN        troch_LN            70.0
 180 │ GW_Lane5_LN2                    GW_Lane5_LN2                    LN_rep    troch_LN_rep    troch_LN_rep        58.0
 181 │ GW_Lane5_LN20                   GW_Lane5_LN20                   LN        troch_LN        troch_LN            71.0
 182 │ GW_Lane5_LN3                    GW_Lane5_LN3                    LN        troch_LN        troch_LN            59.0
 183 │ GW_Lane5_LN4                    GW_Lane5_LN4                    LN        troch_LN        troch_LN            60.0
 184 │ GW_Lane5_LN6                    GW_Lane5_LN6                    LN        troch_LN        troch_LN            61.0
 185 │ GW_Lane5_LN7                    GW_Lane5_LN7                    LN        troch_LN        troch_LN            62.0
 186 │ GW_Lane5_LN8                    GW_Lane5_LN8                    LN        troch_LN        troch_LN            63.0
 187 │ GW_Lane5_MN1                    GW_Lane5_MN1                    MN        troch_MN        troch_west          51.0
 188 │ GW_Lane5_MN12                   GW_Lane5_MN12                   MN        troch_MN        troch_west          56.0
 189 │ GW_Lane5_MN3                    GW_Lane5_MN3                    MN        troch_MN        troch_west          52.0
 190 │ GW_Lane5_MN5                    GW_Lane5_MN5                    MN        troch_MN        troch_west          53.0
 191 │ GW_Lane5_MN8                    GW_Lane5_MN8                    MN        troch_MN        troch_west          54.0
 192 │ GW_Lane5_MN9                    GW_Lane5_MN9                    MN        troch_MN        troch_west          55.0
 193 │ GW_Lane5_NA1                    GW_Lane5_NA1                    NR        lud_PK          lud_PK              39.2
 194 │ GW_Lane5_NA3-3ul                GW_Lane5_NA3-3ul                NR        lud_PK          lud_PK              39.2
 195 │ GW_Lane5_PT11                   GW_Lane5_PT11                   KL        lud_KL          lud_central         42.0
 196 │ GW_Lane5_PT12                   GW_Lane5_PT12                   KL        lud_KL          lud_central         42.0
 197 │ GW_Lane5_PT2                    GW_Lane5_PT2                    ML        lud_ML          lud_ML              51.0
 198 │ GW_Lane5_PT3                    GW_Lane5_PT3                    PA        lud_PA          lud_central         46.0
 199 │ GW_Lane5_PT4                    GW_Lane5_PT4                    PA        lud_PA          lud_central         46.0
 200 │ GW_Lane5_PT6                    GW_Lane5_PT6                    KL        lud_KL          lud_central         42.0
 201 │ GW_Lane5_SH1                    GW_Lane5_SH1                    SH        lud_PK          lud_PK              39.1
 202 │ GW_Lane5_SH2                    GW_Lane5_SH2                    SH        lud_PK          lud_PK              39.1
 203 │ GW_Lane5_SH4                    GW_Lane5_SH4                    SH        lud_PK          lud_PK              39.1
 204 │ GW_Lane5_SH5                    GW_Lane5_SH5                    SH        lud_PK          lud_PK              39.1
 205 │ GW_Lane5_SL1                    GW_Lane5_SL1                    SL        plumb           plumb              150.0
 206 │ GW_Lane5_SL2                    GW_Lane5_SL2                    SL        plumb           plumb              151.0
 207 │ GW_Lane5_ST1                    GW_Lane5_ST1                    ST        plumb           plumb              142.0
 208 │ GW_Lane5_ST12                   GW_Lane5_ST12                   ST        plumb           plumb              144.0
 209 │ GW_Lane5_ST3                    GW_Lane5_ST3                    ST        plumb           plumb              143.0
 210 │ GW_Lane5_STvi1                  GW_Lane5_STvi1                  ST_vi     vir             vir                 22.0
 211 │ GW_Lane5_STvi2                  GW_Lane5_STvi2                  ST_vi     vir             vir                 23.0
 212 │ GW_Lane5_STvi3                  GW_Lane5_STvi3                  ST_vi     vir             vir                 24.0
 213 │ GW_Lane5_TA1                    GW_Lane5_TA1                    TA        plumb           plumb               86.0
 214 │ GW_Lane5_TL1                    GW_Lane5_TL1                    TL        vir             vir                  9.0
 215 │ GW_Lane5_TL10                   GW_Lane5_TL10                   TL        vir             vir                 17.0
 216 │ GW_Lane5_TL11                   GW_Lane5_TL11                   TL        vir             vir                 18.0
 217 │ GW_Lane5_TL12                   GW_Lane5_TL12                   TL        vir             vir                 19.0
 218 │ GW_Lane5_TL2                    GW_Lane5_TL2                    TL        vir             vir                 10.0
 219 │ GW_Lane5_TL3                    GW_Lane5_TL3                    TL_rep    vir_rep         vir_rep             11.0
 220 │ GW_Lane5_TL4                    GW_Lane5_TL4                    TL        vir             vir                 12.0
 221 │ GW_Lane5_TL5                    GW_Lane5_TL5                    TL        vir             vir                 13.0
 222 │ GW_Lane5_TL7                    GW_Lane5_TL7                    TL        vir             vir                 14.0
 223 │ GW_Lane5_TL8                    GW_Lane5_TL8                    TL        vir             vir                 15.0
 224 │ GW_Lane5_TL9                    GW_Lane5_TL9                    TL        vir             vir                 16.0
 225 │ GW_Lane5_TU1                    GW_Lane5_TU1                    TU        nit             nit                 35.0
 226 │ GW_Lane5_TU2                    GW_Lane5_TU2                    TU        nit             nit                 36.0
 227 │ GW_Lane5_UY1                    GW_Lane5_UY1                    UY_rep    plumb_rep       plumb_rep           93.0
 228 │ GW_Lane5_UY2                    GW_Lane5_UY2                    UY        plumb           plumb               88.0
 229 │ GW_Lane5_UY3                    GW_Lane5_UY3                    UY        plumb           plumb               89.0
 230 │ GW_Lane5_UY4                    GW_Lane5_UY4                    UY        plumb           plumb               90.0
 231 │ GW_Lane5_UY5                    GW_Lane5_UY5                    UY        plumb           plumb               91.0
 232 │ GW_Lane5_UY6                    GW_Lane5_UY6                    UY        plumb           plumb               92.0
 233 │ GW_Lane5_YK1                    GW_Lane5_YK1                    YK        vir             vir                  1.0
 234 │ GW_Lane5_YK11                   GW_Lane5_YK11                   YK        vir             vir                  8.0
 235 │ GW_Lane5_YK3                    GW_Lane5_YK3                    YK        vir             vir                  2.0
 236 │ GW_Lane5_YK4                    GW_Lane5_YK4                    YK        vir             vir                  3.0
 237 │ GW_Lane5_YK5                    GW_Lane5_YK5                    YK        vir             vir                  4.0
 238 │ GW_Lane5_YK6                    GW_Lane5_YK6                    YK        vir             vir                  5.0
 239 │ GW_Lane5_YK7                    GW_Lane5_YK7                    YK        vir             vir                  6.0
 240 │ GW_Lane5_YK9                    GW_Lane5_YK9                    YK        vir             vir                  7.0
 241 │ GW_Liz_GBS_Liz10045             GW_Liz_GBS_Liz10045             ML        lud             lud_ML              51.01
 242 │ GW_Liz_GBS_Liz10094             GW_Liz_GBS_Liz10094             ML        lud             lud_ML              51.02
 243 │ GW_Liz_GBS_Liz5101              GW_Liz_GBS_Liz5101              ML        lud             lud_ML              51.03
 244 │ GW_Liz_GBS_Liz5101_R            GW_Liz_GBS_Liz5101_R            ML_rep    lud_rep         lud_ML_rep          51.04
 245 │ GW_Liz_GBS_Liz5118              GW_Liz_GBS_Liz5118              ML        lud             lud_ML              51.05
 246 │ GW_Liz_GBS_Liz5139              GW_Liz_GBS_Liz5139              ML        lud             lud_ML              51.06
 247 │ GW_Liz_GBS_Liz5142              GW_Liz_GBS_Liz5142              ML        lud             lud_ML              51.07
 248 │ GW_Liz_GBS_Liz5144              GW_Liz_GBS_Liz5144              ML        lud             lud_ML              51.08
 249 │ GW_Liz_GBS_Liz5150              GW_Liz_GBS_Liz5150              ML        lud             lud_ML              51.09
 250 │ GW_Liz_GBS_Liz5159              GW_Liz_GBS_Liz5159              ML        lud_chick       lud_ML              51.1
 251 │ GW_Liz_GBS_Liz5162              GW_Liz_GBS_Liz5162              ML        lud_chick       lud_ML              51.11
 252 │ GW_Liz_GBS_Liz5163              GW_Liz_GBS_Liz5163              ML        lud_chick       lud_ML              51.12
 253 │ GW_Liz_GBS_Liz5164              GW_Liz_GBS_Liz5164              ML        lud_chick       lud_ML              51.13
 254 │ GW_Liz_GBS_Liz5165              GW_Liz_GBS_Liz5165              ML        lud             lud_ML              51.14
 255 │ GW_Liz_GBS_Liz5167              GW_Liz_GBS_Liz5167              ML        lud_chick       lud_ML              51.15
 256 │ GW_Liz_GBS_Liz5168              GW_Liz_GBS_Liz5168              ML        lud_chick       lud_ML              51.16
 257 │ GW_Liz_GBS_Liz5169              GW_Liz_GBS_Liz5169              ML        lud_chick       lud_ML              51.17
 258 │ GW_Liz_GBS_Liz5171              GW_Liz_GBS_Liz5171              ML        lud             lud_ML              51.18
 259 │ GW_Liz_GBS_Liz5172              GW_Liz_GBS_Liz5172              ML        lud_chick       lud_ML              51.19
 260 │ GW_Liz_GBS_Liz5173              GW_Liz_GBS_Liz5173              ML        lud_chick       lud_ML              51.2
 261 │ GW_Liz_GBS_Liz5174              GW_Liz_GBS_Liz5174              ML        lud             lud_ML              51.21
 262 │ GW_Liz_GBS_Liz5175              GW_Liz_GBS_Liz5175              ML        lud             lud_ML              51.22
 263 │ GW_Liz_GBS_Liz5176              GW_Liz_GBS_Liz5176              ML        lud             lud_ML              51.23
 264 │ GW_Liz_GBS_Liz5177              GW_Liz_GBS_Liz5177              ML        lud_chick       lud_ML              51.24
 265 │ GW_Liz_GBS_Liz5178              GW_Liz_GBS_Liz5178              ML        lud_chick       lud_ML              51.25
 266 │ GW_Liz_GBS_Liz5179              GW_Liz_GBS_Liz5179              ML        lud_chick       lud_ML              51.26
 267 │ GW_Liz_GBS_Liz5180              GW_Liz_GBS_Liz5180              ML        lud             lud_ML              51.27
 268 │ GW_Liz_GBS_Liz5182              GW_Liz_GBS_Liz5182              ML        lud_chick       lud_ML              51.28
 269 │ GW_Liz_GBS_Liz5184              GW_Liz_GBS_Liz5184              ML        lud_chick       lud_ML              51.29
 270 │ GW_Liz_GBS_Liz5185              GW_Liz_GBS_Liz5185              ML        lud             lud_ML              51.3
 271 │ GW_Liz_GBS_Liz5186              GW_Liz_GBS_Liz5186              ML        lud_chick       lud_ML              51.31
 272 │ GW_Liz_GBS_Liz5187              GW_Liz_GBS_Liz5187              ML        lud_chick       lud_ML              51.32
 273 │ GW_Liz_GBS_Liz5188              GW_Liz_GBS_Liz5188              ML        lud             lud_ML              51.33
 274 │ GW_Liz_GBS_Liz5189              GW_Liz_GBS_Liz5189              ML        lud_chick       lud_ML              51.34
 275 │ GW_Liz_GBS_Liz5190              GW_Liz_GBS_Liz5190              ML        lud_chick       lud_ML              51.35
 276 │ GW_Liz_GBS_Liz5191              GW_Liz_GBS_Liz5191              ML        lud_chick       lud_ML              51.36
 277 │ GW_Liz_GBS_Liz5192              GW_Liz_GBS_Liz5192              ML        lud_chick       lud_ML              51.37
 278 │ GW_Liz_GBS_Liz5193              GW_Liz_GBS_Liz5193              ML        lud_chick       lud_ML              51.38
 279 │ GW_Liz_GBS_Liz5194              GW_Liz_GBS_Liz5194              ML        lud_chick       lud_ML              51.39
 280 │ GW_Liz_GBS_Liz5195              GW_Liz_GBS_Liz5195              ML        lud             lud_ML              51.4
 281 │ GW_Liz_GBS_Liz5197              GW_Liz_GBS_Liz5197              ML        lud             lud_ML              51.41
 282 │ GW_Liz_GBS_Liz5199              GW_Liz_GBS_Liz5199              ML        lud_chick       lud_ML              51.42
 283 │ GW_Liz_GBS_Liz6002              GW_Liz_GBS_Liz6002              ML        lud             lud_ML              51.43
 284 │ GW_Liz_GBS_Liz6006              GW_Liz_GBS_Liz6006              ML        lud             lud_ML              51.44
 285 │ GW_Liz_GBS_Liz6008              GW_Liz_GBS_Liz6008              ML        lud             lud_ML              51.45
 286 │ GW_Liz_GBS_Liz6009              GW_Liz_GBS_Liz6009              ML        lud             lud_ML              51.46
 287 │ GW_Liz_GBS_Liz6010              GW_Liz_GBS_Liz6010              ML        lud             lud_ML              51.47
 288 │ GW_Liz_GBS_Liz6012              GW_Liz_GBS_Liz6012              ML        lud             lud_ML              51.48
 289 │ GW_Liz_GBS_Liz6014              GW_Liz_GBS_Liz6014              ML        lud             lud_ML              51.49
 290 │ GW_Liz_GBS_Liz6055              GW_Liz_GBS_Liz6055              ML        lud             lud_ML              51.5
 291 │ GW_Liz_GBS_Liz6057              GW_Liz_GBS_Liz6057              ML        lud             lud_ML              51.51
 292 │ GW_Liz_GBS_Liz6060              GW_Liz_GBS_Liz6060              ML        lud             lud_ML              51.52
 293 │ GW_Liz_GBS_Liz6062              GW_Liz_GBS_Liz6062              ML        lud             lud_ML              51.53
 294 │ GW_Liz_GBS_Liz6063              GW_Liz_GBS_Liz6063              ML        lud             lud_ML              51.54
 295 │ GW_Liz_GBS_Liz6066              GW_Liz_GBS_Liz6066              ML        lud             lud_ML              51.55
 296 │ GW_Liz_GBS_Liz6072              GW_Liz_GBS_Liz6072              ML        lud             lud_ML              51.56
 297 │ GW_Liz_GBS_Liz6079              GW_Liz_GBS_Liz6079              ML        lud             lud_ML              51.57
 298 │ GW_Liz_GBS_Liz6203              GW_Liz_GBS_Liz6203              ML        lud_chick       lud_ML              51.58
 299 │ GW_Liz_GBS_Liz6204              GW_Liz_GBS_Liz6204              ML        lud_chick       lud_ML              51.59
 300 │ GW_Liz_GBS_Liz6461              GW_Liz_GBS_Liz6461              ML        lud             lud_ML              51.6
 301 │ GW_Liz_GBS_Liz6472              GW_Liz_GBS_Liz6472              ML        lud             lud_ML              51.61
 302 │ GW_Liz_GBS_Liz6478              GW_Liz_GBS_Liz6478              ML        lud             lud_ML              51.62
 303 │ GW_Liz_GBS_Liz6766              GW_Liz_GBS_Liz6766              ML        lud             lud_ML              51.63
 304 │ GW_Liz_GBS_Liz6776              GW_Liz_GBS_Liz6776              ML        lud             lud_ML              51.64
 305 │ GW_Liz_GBS_Liz6794              GW_Liz_GBS_Liz6794              ML        lud             lud_ML              51.65
 306 │ GW_Liz_GBS_P_fusc               GW_Liz_GBS_P_fusc               fusc      fusc            fusc               201.0
 307 │ GW_Liz_GBS_P_h_man              GW_Liz_GBS_P_h_man              hmand     hmand           hmand              202.0
 308 │ GW_Liz_GBS_P_humei              GW_Liz_GBS_P_humei              hume      hume            hume               203.0
 309 │ GW_Liz_GBS_P_inor               GW_Liz_GBS_P_inor               inor      inor            inor               204.0
 310 │ GW_Liz_GBS_S_burk               GW_Liz_GBS_S_burk               burk      burk            burk               205.0

GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly.</code></pre>
</div>
</div>
</section>
<section id="polish-a-few-individual-names-to-match-those-in-other-metadata-object-above-and-make-more-readable-graphs" class="level3">
<h3 class="anchored" data-anchor-id="polish-a-few-individual-names-to-match-those-in-other-metadata-object-above-and-make-more-readable-graphs">Polish a few individual names (to match those in other metadata object above, and make more readable graphs):</h3>
<div id="c1dad207" class="cell" data-execution_count="185">
<div class="sourceCode cell-code" id="cb503"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb503-1"><a href="#cb503-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr4A.ind <span class="op">=</span> <span class="fu">correctNames</span>(ind_with_metadata_chr4A.ind)</span>
<span id="cb503-2"><a href="#cb503-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb503-3"><a href="#cb503-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr4A.ID <span class="op">=</span> <span class="fu">correctNames</span>(ind_with_metadata_chr4A.ID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="186">
<pre><code>310-element Vector{String}:
 "GW_Armando_plate1_AB1"
 "GW_Armando_plate1_JF07G02"
 "GW_Armando_plate1_JF07G03"
 "GW_Armando_plate1_JF07G04"
 "GW_Armando_plate1_JF08G02"
 "GW_Armando_plate1_JF09G01"
 "GW_Armando_plate1_JF09G02"
 "GW_Armando_plate1_JF10G03"
 "GW_Armando_plate1_JF11G01"
 "GW_Armando_plate1_JF12G01"
 "GW_Armando_plate1_JF12G02"
 "GW_Armando_plate1_JF12G04"
 "GW_Armando_plate1_JF13G01"
 ⋮
 "GW_Liz_GBS_Liz6204"
 "GW_Liz_GBS_Liz6461"
 "GW_Liz_GBS_Liz6472"
 "GW_Liz_GBS_Liz6478"
 "GW_Liz_GBS_Liz6766"
 "GW_Liz_GBS_Liz6776"
 "GW_Liz_GBS_Liz6794"
 "GW_Liz_GBS_P_fusc"
 "GW_Liz_GBS_P_h_man"
 "GW_Liz_GBS_P_humei"
 "GW_Liz_GBS_P_inor"
 "GW_Liz_GBS_S_burk"</code></pre>
</div>
</div>
</section>
<section id="filter-to-just-the-individuals-also-included-in-the-analysis-of-lhbrs-above" class="level3">
<h3 class="anchored" data-anchor-id="filter-to-just-the-individuals-also-included-in-the-analysis-of-lhbrs-above">Filter to just the individuals also included in the analysis of LHBRs above</h3>
<div id="d7e22b26" class="cell" data-execution_count="186">
<div class="sourceCode cell-code" id="cb505"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb505-1"><a href="#cb505-1" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(ind_with_metadata_included.ind), ind_with_metadata_chr4A.ind)</span>
<span id="cb505-2"><a href="#cb505-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb505-3"><a href="#cb505-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr4A_included <span class="op">=</span> ind_with_metadata_chr4A[selection, <span class="op">:</span>]</span>
<span id="cb505-4"><a href="#cb505-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb505-5"><a href="#cb505-5" aria-hidden="true" tabindex="-1"></a><span class="co"># select genotypes of just the included individuals, and ignore first column</span></span>
<span id="cb505-6"><a href="#cb505-6" aria-hidden="true" tabindex="-1"></a>geno_chr4A_included <span class="op">=</span> geno_chr4A[selection, <span class="fl">2</span><span class="op">:</span><span class="kw">end</span>]</span>
<span id="cb505-7"><a href="#cb505-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb505-8"><a href="#cb505-8" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb505-9"><a href="#cb505-9" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(ind_with_metadata_included.gw4A_cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>["virLud", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud", "virLud", "obsPlumb", "virLud", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "virLud_obsPlumb", "obsPlumb", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLudHet", "troch", "virLud_troch", "obsPlumb", "virLud_troch", "troch_obsPlumb", "troch_obsPlumb", "troch_obsPlumb", "virLudHet", "virLud_obsPlumb", "virLudHet", "obsPlumb", "virLud_obsPlumb", "virLud_troch", "virLud_obsPlumb", "obsPlumb", "troch_obsPlumb", "troch_obsPlumb", "troch_obsPlumb", "troch", "troch_obsPlumb", "troch", "troch", "troch", "virLud_troch", "troch", "trochHet", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "obsPlumb", "virLud", "virLud", "virLud", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud", "virLud_obsPlumb", "obsPlumb", "obsPlumbHet", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "troch", "obsPlumb", "troch", "troch", "troch_obsPlumb", "virLud_troch", "obsPlumb", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "troch_obsPlumb", "obsPlumb", "virLud_obsPlumb", "trochHet", "troch", "troch", "troch_obsPlumb", "troch", "troch", "troch", "troch", "troch", "troch", "virLud_obsPlumb", "virLud_troch", "virLud_obsPlumb", "obsPlumb", "virLud_obsPlumb", "virLud_obsPlumb", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "troch_obsPlumb", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "virLud_obsPlumb", "troch", "troch", "troch", "trochHet", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "troch", "virLud", "obsPlumb", "virLud", "obsPlumb", "troch", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "obsPlumb", "virLud", "virLud", "virLud", "obsPlumb", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "nit", "nit", "obsPlumb", "obsPlumb", "virLud_obsPlumb", "obsPlumb", "obsPlumb", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "troch_obsPlumb", "troch", "troch", "troch", "troch_obsPlumb", "troch", "troch", "troch", "troch", "troch", "troch", "troch_obsPlumb", "troch", "troch", "troch_obsPlumb", "troch_obsPlumb", "troch", "troch_obsPlumb", "troch", "virLud_obsPlumb", "troch_obsPlumb", "troch_obsPlumb", "troch", "virLud_troch", "troch", "troch_obsPlumb", "troch", "virLud_troch", "virLud_troch", "troch_obsPlumb", "troch", "troch", "virLud_troch", "troch", "troch_obsPlumb", "troch", "troch", "troch", "trochHet", "troch", "troch", "troch", "troch"]</code></pre>
</div>
</div>
</section>
<section id="look-up-the-chr4a-individual-membership-in-homozygous-clusters-and-calculate-pi-and-dxy" class="level3">
<h3 class="anchored" data-anchor-id="look-up-the-chr4a-individual-membership-in-homozygous-clusters-and-calculate-pi-and-dxy">Look up the chr4A individual membership in homozygous clusters, and calculate pi and Dxy</h3>
<div id="60298a43" class="cell" data-execution_count="187">
<div class="sourceCode cell-code" id="cb507"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb507-1"><a href="#cb507-1" aria-hidden="true" tabindex="-1"></a>indClusterMembership_gw4A <span class="op">=</span> ind_with_metadata_included.gw4A_cluster</span>
<span id="cb507-2"><a href="#cb507-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-3"><a href="#cb507-3" aria-hidden="true" tabindex="-1"></a>clusterNames_gw4A <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb507-4"><a href="#cb507-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"nit"</span>,</span>
<span id="cb507-5"><a href="#cb507-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"troch"</span>,</span>
<span id="cb507-6"><a href="#cb507-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"obsPlumb"</span>]</span>
<span id="cb507-7"><a href="#cb507-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-8"><a href="#cb507-8" aria-hidden="true" tabindex="-1"></a><span class="co"># get boundaries of gw4A LHBR:</span></span>
<span id="cb507-9"><a href="#cb507-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-10"><a href="#cb507-10" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw4A"</span></span>
<span id="cb507-11"><a href="#cb507-11" aria-hidden="true" tabindex="-1"></a>positionMin_chr4A_LHBR, positionMax_chr4A_LHBR, regionText, </span>
<span id="cb507-12"><a href="#cb507-12" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb507-13"><a href="#cb507-13" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb507-14"><a href="#cb507-14" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb507-15"><a href="#cb507-15" aria-hidden="true" tabindex="-1"></a>                                                    pos_SNP_filtered, </span>
<span id="cb507-16"><a href="#cb507-16" aria-hidden="true" tabindex="-1"></a>                                                    highHetVarRegions, chr;</span>
<span id="cb507-17"><a href="#cb507-17" aria-hidden="true" tabindex="-1"></a>                                                    windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb507-18"><a href="#cb507-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-19"><a href="#cb507-19" aria-hidden="true" tabindex="-1"></a><span class="co"># select the loci within the gw4A LHBR:</span></span>
<span id="cb507-20"><a href="#cb507-20" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (positionMin_chr4A_LHBR <span class="op">.&lt;=</span> pos_chr4A.position <span class="op">.&lt;=</span> positionMax_chr4A_LHBR) </span>
<span id="cb507-21"><a href="#cb507-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-22"><a href="#cb507-22" aria-hidden="true" tabindex="-1"></a>geno_chr4A_included_LHBR <span class="op">=</span> geno_chr4A_included[<span class="op">:</span>, selection]</span>
<span id="cb507-23"><a href="#cb507-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-24"><a href="#cb507-24" aria-hidden="true" tabindex="-1"></a>pos_chr4A_LHBR <span class="op">=</span> pos_chr4A[selection, <span class="op">:</span>]</span>
<span id="cb507-25"><a href="#cb507-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-26"><a href="#cb507-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb507-27"><a href="#cb507-27" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(geno_chr4A_included_LHBR, indClusterMembership_gw4A, clusterNames_gw4A)</span>
<span id="cb507-28"><a href="#cb507-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-29"><a href="#cb507-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb507-30"><a href="#cb507-30" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb507-31"><a href="#cb507-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-32"><a href="#cb507-32" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb507-33"><a href="#cb507-33" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames_gw4A)</span>
<span id="cb507-34"><a href="#cb507-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-35"><a href="#cb507-35" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames_gw4A; among<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb507-36"><a href="#cb507-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-37"><a href="#cb507-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb507-38"><a href="#cb507-38" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames_gw4A, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb507-39"><a href="#cb507-39" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb507-40"><a href="#cb507-40" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi          </span></span>
<span id="cb507-41"><a href="#cb507-41" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64     </span></span>
<span id="cb507-42"><a href="#cb507-42" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼───────────────────────</span></span>
<span id="cb507-43"><a href="#cb507-43" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.000956575</span></span>
<span id="cb507-44"><a href="#cb507-44" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.000332204</span></span>
<span id="cb507-45"><a href="#cb507-45" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch     0.000613901</span></span>
<span id="cb507-46"><a href="#cb507-46" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obsPlumb  0.000261819 =#</span></span>
<span id="cb507-47"><a href="#cb507-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-48"><a href="#cb507-48" aria-hidden="true" tabindex="-1"></a><span class="co"># average pi (for chr 4A LHBR) among three major groups:</span></span>
<span id="cb507-49"><a href="#cb507-49" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.000956575</span> <span class="op">+</span> <span class="fl">0.000613901</span> <span class="op">+</span> <span class="fl">0.000261819</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb507-50"><a href="#cb507-50" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.000610765</span></span>
<span id="cb507-51"><a href="#cb507-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-52"><a href="#cb507-52" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb507-53"><a href="#cb507-53" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb507-54"><a href="#cb507-54" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy        </span></span>
<span id="cb507-55"><a href="#cb507-55" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64    </span></span>
<span id="cb507-56"><a href="#cb507-56" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────────</span></span>
<span id="cb507-57"><a href="#cb507-57" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.00325609</span></span>
<span id="cb507-58"><a href="#cb507-58" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch     0.0031813</span></span>
<span id="cb507-59"><a href="#cb507-59" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obsPlumb  0.00241666</span></span>
<span id="cb507-60"><a href="#cb507-60" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_troch        0.00286634</span></span>
<span id="cb507-61"><a href="#cb507-61" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_obsPlumb     0.00249507</span></span>
<span id="cb507-62"><a href="#cb507-62" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ troch_obsPlumb   0.00305931 =#</span></span>
<span id="cb507-63"><a href="#cb507-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-64"><a href="#cb507-64" aria-hidden="true" tabindex="-1"></a><span class="co"># average Dxy (for chr 4A LHBR) among three major groups:</span></span>
<span id="cb507-65"><a href="#cb507-65" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0031813</span> <span class="op">+</span> <span class="fl">0.00241666</span> <span class="op">+</span> <span class="fl">0.00305931</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb507-66"><a href="#cb507-66" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.0028857566666666674</span></span>
<span id="cb507-67"><a href="#cb507-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb507-68"><a href="#cb507-68" aria-hidden="true" tabindex="-1"></a><span class="co"># Drawing phylogeny (in Illustrator) based on above, between three major groups.</span></span>
<span id="cb507-69"><a href="#cb507-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Ignoring nit, the most recent connection is between virLud and obsPlumb (0.00241666).</span></span>
<span id="cb507-70"><a href="#cb507-70" aria-hidden="true" tabindex="-1"></a><span class="co"># For deeper branch length, am using: </span></span>
<span id="cb507-71"><a href="#cb507-71" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculation for average Dxy between troch and (virLud, obsPlumb):</span></span>
<span id="cb507-72"><a href="#cb507-72" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0031813</span> <span class="op">+</span> <span class="fl">0.00305931</span>) <span class="op">/</span> <span class="fl">2</span></span>
<span id="cb507-73"><a href="#cb507-73" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.003120305</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Good news: 1 region on that scaffold</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="188">
<pre><code>0.0031203050000000003</code></pre>
</div>
</div>
<p>Wow, that is an amazing difference between pi within the obsPlumb haplotype and Dxy between that and others (roughly 10x).</p>
</section>
<section id="calculate-pi-and-dxy-outside-of-the-chr-4a-lhbr-grouped-by-the-lhbr-homozygous-groups" class="level3">
<h3 class="anchored" data-anchor-id="calculate-pi-and-dxy-outside-of-the-chr-4a-lhbr-grouped-by-the-lhbr-homozygous-groups">Calculate pi and Dxy outside of the chr 4A LHBR (grouped by the LHBR homozygous groups)</h3>
<div id="6e3b1ad5" class="cell" data-execution_count="188">
<div class="sourceCode cell-code" id="cb510"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb510-1"><a href="#cb510-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the loci outside of the gw4A LHBR:</span></span>
<span id="cb510-2"><a href="#cb510-2" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> .!(positionMin_chr4A_LHBR <span class="op">.&lt;=</span> pos_chr4A.position <span class="op">.&lt;=</span> positionMax_chr4A_LHBR) </span>
<span id="cb510-3"><a href="#cb510-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-4"><a href="#cb510-4" aria-hidden="true" tabindex="-1"></a>geno_chr4A_included_nonLHBR <span class="op">=</span> geno_chr4A_included[<span class="op">:</span>, selection]</span>
<span id="cb510-5"><a href="#cb510-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-6"><a href="#cb510-6" aria-hidden="true" tabindex="-1"></a>pos_chr4A_nonLHBR <span class="op">=</span> pos_chr4A[selection, <span class="op">:</span>]</span>
<span id="cb510-7"><a href="#cb510-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-8"><a href="#cb510-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb510-9"><a href="#cb510-9" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(geno_chr4A_included_nonLHBR, indClusterMembership_gw4A, clusterNames_gw4A)</span>
<span id="cb510-10"><a href="#cb510-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-11"><a href="#cb510-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb510-12"><a href="#cb510-12" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb510-13"><a href="#cb510-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-14"><a href="#cb510-14" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb510-15"><a href="#cb510-15" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames_gw4A)</span>
<span id="cb510-16"><a href="#cb510-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-17"><a href="#cb510-17" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames_gw4A; among<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb510-18"><a href="#cb510-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-19"><a href="#cb510-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb510-20"><a href="#cb510-20" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames_gw4A, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb510-21"><a href="#cb510-21" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb510-22"><a href="#cb510-22" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi         </span></span>
<span id="cb510-23"><a href="#cb510-23" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64    </span></span>
<span id="cb510-24"><a href="#cb510-24" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼──────────────────────</span></span>
<span id="cb510-25"><a href="#cb510-25" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.0041321</span></span>
<span id="cb510-26"><a href="#cb510-26" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.00196343</span></span>
<span id="cb510-27"><a href="#cb510-27" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ troch     0.00551821</span></span>
<span id="cb510-28"><a href="#cb510-28" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ obsPlumb  0.0055897 =#</span></span>
<span id="cb510-29"><a href="#cb510-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-30"><a href="#cb510-30" aria-hidden="true" tabindex="-1"></a><span class="co"># average pi (for chr 4A NOT in LHBR) among three major groups:</span></span>
<span id="cb510-31"><a href="#cb510-31" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0041321</span> <span class="op">+</span> <span class="fl">0.00551821</span> <span class="op">+</span> <span class="fl">0.0055897</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb510-32"><a href="#cb510-32" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.005080003333333333</span></span>
<span id="cb510-33"><a href="#cb510-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-34"><a href="#cb510-34" aria-hidden="true" tabindex="-1"></a><span class="co">#ratio of average pi outside to average pi within chr 4A LHBR:</span></span>
<span id="cb510-35"><a href="#cb510-35" aria-hidden="true" tabindex="-1"></a><span class="fl">0.005080003333333333</span> <span class="op">/</span> <span class="fl">0.000610765</span></span>
<span id="cb510-36"><a href="#cb510-36" aria-hidden="true" tabindex="-1"></a><span class="co"># 8.317443424776032</span></span>
<span id="cb510-37"><a href="#cb510-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-38"><a href="#cb510-38" aria-hidden="true" tabindex="-1"></a><span class="co"># percent lower that average pi is within compared to outside LHBR:</span></span>
<span id="cb510-39"><a href="#cb510-39" aria-hidden="true" tabindex="-1"></a><span class="fl">100</span> <span class="op">*</span> (<span class="fl">8.317443424776032</span> <span class="op">-</span> <span class="fl">1</span>) <span class="op">/</span> <span class="fl">8.317443424776032</span></span>
<span id="cb510-40"><a href="#cb510-40" aria-hidden="true" tabindex="-1"></a><span class="co"># 87.97707481819238</span></span>
<span id="cb510-41"><a href="#cb510-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-42"><a href="#cb510-42" aria-hidden="true" tabindex="-1"></a><span class="co"># for obsPlumb haplotype, ratio of pi outside to pi within chr 4A LHBR:</span></span>
<span id="cb510-43"><a href="#cb510-43" aria-hidden="true" tabindex="-1"></a><span class="fl">0.0055897</span> <span class="op">/</span> <span class="fl">0.000261819</span></span>
<span id="cb510-44"><a href="#cb510-44" aria-hidden="true" tabindex="-1"></a><span class="co"># 21.349481893980194</span></span>
<span id="cb510-45"><a href="#cb510-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-46"><a href="#cb510-46" aria-hidden="true" tabindex="-1"></a><span class="co"># percent lower that pi of obsPlumb haplotype is within vs. outside of HLBR:</span></span>
<span id="cb510-47"><a href="#cb510-47" aria-hidden="true" tabindex="-1"></a><span class="fl">100</span> <span class="op">*</span> (<span class="fl">21.349481893980194</span> <span class="op">-</span> <span class="fl">1</span>) <span class="op">/</span> <span class="fl">21.349481893980194</span></span>
<span id="cb510-48"><a href="#cb510-48" aria-hidden="true" tabindex="-1"></a><span class="co"># 95.31604558384171</span></span>
<span id="cb510-49"><a href="#cb510-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-50"><a href="#cb510-50" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb510-51"><a href="#cb510-51" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb510-52"><a href="#cb510-52" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy        </span></span>
<span id="cb510-53"><a href="#cb510-53" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64    </span></span>
<span id="cb510-54"><a href="#cb510-54" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────────</span></span>
<span id="cb510-55"><a href="#cb510-55" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.00440681</span></span>
<span id="cb510-56"><a href="#cb510-56" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_troch     0.0055599</span></span>
<span id="cb510-57"><a href="#cb510-57" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_obsPlumb  0.00556595</span></span>
<span id="cb510-58"><a href="#cb510-58" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_troch        0.00520058</span></span>
<span id="cb510-59"><a href="#cb510-59" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_obsPlumb     0.00548082</span></span>
<span id="cb510-60"><a href="#cb510-60" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ troch_obsPlumb   0.00601295 =#</span></span>
<span id="cb510-61"><a href="#cb510-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-62"><a href="#cb510-62" aria-hidden="true" tabindex="-1"></a><span class="co"># average Dxy (for OUTSIDE of chr 4A LHBR) among three major groups:</span></span>
<span id="cb510-63"><a href="#cb510-63" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0055599</span> <span class="op">+</span> <span class="fl">0.00556595</span> <span class="op">+</span> <span class="fl">0.00601295</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb510-64"><a href="#cb510-64" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.005712933333333333</span></span>
<span id="cb510-65"><a href="#cb510-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-66"><a href="#cb510-66" aria-hidden="true" tabindex="-1"></a><span class="co">#ratio of average Dxy outside to average Dxy within chr 4A LHBR among 3 major groups:</span></span>
<span id="cb510-67"><a href="#cb510-67" aria-hidden="true" tabindex="-1"></a><span class="fl">0.005712933333333333</span> <span class="op">/</span> <span class="fl">0.0028857566666666674</span></span>
<span id="cb510-68"><a href="#cb510-68" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.9797002981309344</span></span>
<span id="cb510-69"><a href="#cb510-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-70"><a href="#cb510-70" aria-hidden="true" tabindex="-1"></a><span class="co"># percent lower that average Dxy is within compared to outside LHBR:</span></span>
<span id="cb510-71"><a href="#cb510-71" aria-hidden="true" tabindex="-1"></a><span class="fl">100</span> <span class="op">*</span> (<span class="fl">1.9797002981309344</span> <span class="op">-</span> <span class="fl">1</span>) <span class="op">/</span> <span class="fl">1.9797002981309344</span></span>
<span id="cb510-72"><a href="#cb510-72" aria-hidden="true" tabindex="-1"></a><span class="co"># 49.487303661866626</span></span>
<span id="cb510-73"><a href="#cb510-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb510-74"><a href="#cb510-74" aria-hidden="true" tabindex="-1"></a><span class="co"># Drawing phylogeny (in Illustrator) based on above, between three major groups.</span></span>
<span id="cb510-75"><a href="#cb510-75" aria-hidden="true" tabindex="-1"></a><span class="co"># In this case, the virLud_troch is the lower Dxy so am connecting those more recently.</span></span>
<span id="cb510-76"><a href="#cb510-76" aria-hidden="true" tabindex="-1"></a><span class="co"># For deeper brancha length, using this:</span></span>
<span id="cb510-77"><a href="#cb510-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculation of average distance between obsPlumb and (virLud, troch)</span></span>
<span id="cb510-78"><a href="#cb510-78" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.00556595</span> <span class="op">+</span> <span class="fl">0.00601295</span>) <span class="op">/</span> <span class="fl">2</span></span>
<span id="cb510-79"><a href="#cb510-79" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.00578945</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="189">
<pre><code>0.00578945</code></pre>
</div>
</div>
<p>Remarkable differences between pi and Dxy in the gw4A LHBR, and between LHBR and non-LHBR part of that chromosome!</p>
<p>Do same with chr 3, which also shows 3 clear haplotype groups, but a very different biogeographic pattern than 4A:</p>
</section>
</section>
<section id="examine-chromosome-3-large-haploblock-region-lhbr-with-invariant-sites-included" class="level2">
<h2 class="anchored" data-anchor-id="examine-chromosome-3-large-haploblock-region-lhbr-with-invariant-sites-included">Examine chromosome 3 Large HaploBlock Region (LHBR) with invariant sites included</h2>
<p>Before running below, I need to change format of 012NA file to # Before running below, changed 012NA file back into 012minus1 file, using commands like below, so can be read as integer:</p>
<div class="sourceCode" id="cb512"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb512-1"><a href="#cb512-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> /Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw3.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab.012NA <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/NA/-1/g'</span> <span class="op">&gt;</span> /Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw3.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab.012minus1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="1cf9f3ef" class="cell" data-execution_count="189">
<div class="sourceCode cell-code" id="cb513"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb513-1"><a href="#cb513-1" aria-hidden="true" tabindex="-1"></a>baseName <span class="op">=</span> <span class="st">"/Users/darrenirwin/GW_data_from_cedar_Feb2024/GW2022_cedar/infoSites_vcfs/GW2022_all4plates.genotypes.allSites.chrgw3.infoSites.max2allele_noindel.maxmiss60.MQ20.lowHet.tab"</span></span>
<span id="cb513-2"><a href="#cb513-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load metadata</span></span>
<span id="cb513-3"><a href="#cb513-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(dataDirectory)</span>
<span id="cb513-4"><a href="#cb513-4" aria-hidden="true" tabindex="-1"></a>metadata_chr3 <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(metadataFile)) <span class="co"># the CSV.File function interprets the correct delimiter</span></span>
<span id="cb513-5"><a href="#cb513-5" aria-hidden="true" tabindex="-1"></a>num_metadata_cols_chr3 <span class="op">=</span> <span class="fu">ncol</span>(metadata_chr3)</span>
<span id="cb513-6"><a href="#cb513-6" aria-hidden="true" tabindex="-1"></a>num_individuals_chr3 <span class="op">=</span> <span class="fu">nrow</span>(metadata_chr3) </span>
<span id="cb513-7"><a href="#cb513-7" aria-hidden="true" tabindex="-1"></a><span class="co"># read in individual names for this dataset</span></span>
<span id="cb513-8"><a href="#cb513-8" aria-hidden="true" tabindex="-1"></a>individuals_file_name_chr3 <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012.indv"</span>)</span>
<span id="cb513-9"><a href="#cb513-9" aria-hidden="true" tabindex="-1"></a>ind_chr3 <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(individuals_file_name_chr3; header<span class="op">=</span>[<span class="st">"ind"</span>], types<span class="op">=</span>[<span class="dt">String</span>])) </span>
<span id="cb513-10"><a href="#cb513-10" aria-hidden="true" tabindex="-1"></a>indNum_chr3 <span class="op">=</span> <span class="fu">size</span>(ind_chr3, <span class="fl">1</span>) <span class="co"># number of individuals</span></span>
<span id="cb513-11"><a href="#cb513-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> num_individuals_chr3 <span class="op">!=</span> indNum_chr3</span>
<span id="cb513-12"><a href="#cb513-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: number of rows in metadata file different than number of individuals in .indv file"</span>)</span>
<span id="cb513-13"><a href="#cb513-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb513-14"><a href="#cb513-14" aria-hidden="true" tabindex="-1"></a><span class="co"># read in position data for this dataset</span></span>
<span id="cb513-15"><a href="#cb513-15" aria-hidden="true" tabindex="-1"></a>position_file_name_chr3 <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012.pos"</span>)</span>
<span id="cb513-16"><a href="#cb513-16" aria-hidden="true" tabindex="-1"></a>pos_chr3 <span class="op">=</span> <span class="fu">DataFrame</span>(CSV.<span class="fu">File</span>(position_file_name_chr3; header<span class="op">=</span>[<span class="st">"chrom"</span>, <span class="st">"position"</span>], types<span class="op">=</span>[<span class="dt">String</span>, <span class="dt">Int</span>]))</span>
<span id="cb513-17"><a href="#cb513-17" aria-hidden="true" tabindex="-1"></a><span class="co"># read in genotype data</span></span>
<span id="cb513-18"><a href="#cb513-18" aria-hidden="true" tabindex="-1"></a>genotype_file_name_chr3 <span class="op">=</span> <span class="fu">string</span>(baseName, <span class="st">".012minus1"</span>) </span>
<span id="cb513-19"><a href="#cb513-19" aria-hidden="true" tabindex="-1"></a><span class="pp">@time</span> <span class="cf">if</span> <span class="fl">1</span> <span class="op">&lt;=</span> indNum_chr3 <span class="op">&lt;=</span> <span class="fl">127</span>   </span>
<span id="cb513-20"><a href="#cb513-20" aria-hidden="true" tabindex="-1"></a>    geno_chr3 <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name_chr3, <span class="ch">'\t'</span>, <span class="dt">Int8</span>, <span class="ch">'\n'</span>); <span class="co"># this has been sped up dramatically, by first converting "NA" to -1</span></span>
<span id="cb513-21"><a href="#cb513-21" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> <span class="fl">128</span> <span class="op">&lt;=</span> indNum_chr3 <span class="op">&lt;=</span> <span class="fl">32767</span></span>
<span id="cb513-22"><a href="#cb513-22" aria-hidden="true" tabindex="-1"></a>    geno_chr3 <span class="op">=</span> <span class="fu">readdlm</span>(genotype_file_name_chr3, <span class="ch">'\t'</span>, <span class="dt">Int16</span>, <span class="ch">'\n'</span>); <span class="co"># this needed for first column, which is number of individual; Int16 not much slower on import than Int8</span></span>
<span id="cb513-23"><a href="#cb513-23" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb513-24"><a href="#cb513-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Error: Number of individuals in .indv appears outside of range from 1 to 32767"</span>)</span>
<span id="cb513-25"><a href="#cb513-25" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb513-26"><a href="#cb513-26" aria-hidden="true" tabindex="-1"></a>loci_count_chr3 <span class="op">=</span> <span class="fu">size</span>(geno_chr3, <span class="fl">2</span>) <span class="op">-</span> <span class="fl">1</span>   <span class="co"># because the first column is not a SNP (just a count from zero)</span></span>
<span id="cb513-27"><a href="#cb513-27" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">string</span>(<span class="st">"Read in genotypic data at "</span>, loci_count_chr3,<span class="st">" loci for "</span>, indNum_chr3, <span class="st">" individuals. </span><span class="sc">\n</span><span class="st">"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 52.088845 seconds (2.97 M allocations: 15.221 GiB, 4.08% gc time, 0.16% compilation time)
Read in genotypic data at 1855532 loci for 310 individuals. </code></pre>
</div>
</div>
<section id="check-that-individuals-are-same-in-genotype-data-and-metadata-1" class="level3">
<h3 class="anchored" data-anchor-id="check-that-individuals-are-same-in-genotype-data-and-metadata-1">Check that individuals are same in genotype data and metadata</h3>
<div id="07b97bb8" class="cell" data-execution_count="190">
<div class="sourceCode cell-code" id="cb515"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb515-1"><a href="#cb515-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr3 <span class="op">=</span> <span class="fu">hcat</span>(ind_chr3, metadata_chr3)</span>
<span id="cb515-2"><a href="#cb515-2" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(ind_with_metadata_chr3)</span>
<span id="cb515-3"><a href="#cb515-3" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>()  <span class="co"># prints a line break </span></span>
<span id="cb515-4"><a href="#cb515-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="fu">isequal</span>(ind_with_metadata_chr3.ind, ind_with_metadata_chr3.ID)</span>
<span id="cb515-5"><a href="#cb515-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly."</span>)</span>
<span id="cb515-6"><a href="#cb515-6" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb515-7"><a href="#cb515-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"WARNING: names of individuals in metadata file and genotype ind file do not completely match."</span>)</span>
<span id="cb515-8"><a href="#cb515-8" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>310×6 DataFrame
 Row │ ind                             ID                              location  group           Fst_group       plot_order 
     │ String                          String31                        String7   String15        String15        Float64    
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ GW_Armando_plate1_AB1           GW_Armando_plate1_AB1           AB        vir             vir                 20.01
   2 │ GW_Armando_plate1_JF07G02       GW_Armando_plate1_JF07G02       ST        plumb           plumb              108.0
   3 │ GW_Armando_plate1_JF07G03       GW_Armando_plate1_JF07G03       ST        plumb           plumb              109.0
   4 │ GW_Armando_plate1_JF07G04       GW_Armando_plate1_JF07G04       ST        plumb           plumb              110.0
   5 │ GW_Armando_plate1_JF08G02       GW_Armando_plate1_JF08G02       ST        plumb           plumb              111.0
   6 │ GW_Armando_plate1_JF09G01       GW_Armando_plate1_JF09G01       ST        plumb           plumb              112.0
   7 │ GW_Armando_plate1_JF09G02       GW_Armando_plate1_JF09G02       ST        plumb           plumb              113.0
   8 │ GW_Armando_plate1_JF10G03       GW_Armando_plate1_JF10G03       ST        plumb_vir       plumb_vir          170.0
   9 │ GW_Armando_plate1_JF11G01       GW_Armando_plate1_JF11G01       ST        plumb           plumb              114.0
  10 │ GW_Armando_plate1_JF12G01       GW_Armando_plate1_JF12G01       ST        plumb           plumb              115.0
  11 │ GW_Armando_plate1_JF12G02       GW_Armando_plate1_JF12G02       ST        plumb           plumb              116.0
  12 │ GW_Armando_plate1_JF12G04       GW_Armando_plate1_JF12G04       ST_vi     vir             vir                 24.001
  13 │ GW_Armando_plate1_JF13G01       GW_Armando_plate1_JF13G01       ST        plumb           plumb              117.0
  14 │ GW_Armando_plate1_JF15G03       GW_Armando_plate1_JF15G03       DV        plumb           plumb              103.0
  15 │ GW_Armando_plate1_JF16G01       GW_Armando_plate1_JF16G01       DV_vi     plumb_vir       vir                 24.041
  16 │ GW_Armando_plate1_JF20G01       GW_Armando_plate1_JF20G01       MB        plumb           plumb               94.0
  17 │ GW_Armando_plate1_JF22G01       GW_Armando_plate1_JF22G01       MB        plumb           plumb               95.0
  18 │ GW_Armando_plate1_JF23G01       GW_Armando_plate1_JF23G01       VB        plumb           plumb               98.0
  19 │ GW_Armando_plate1_JF23G02       GW_Armando_plate1_JF23G02       VB        plumb           plumb               99.0
  20 │ GW_Armando_plate1_JF24G02       GW_Armando_plate1_JF24G02       VB        plumb           plumb              100.0
  21 │ GW_Armando_plate1_JF26G01       GW_Armando_plate1_JF26G01       ST        plumb           plumb              118.0
  22 │ GW_Armando_plate1_JF27G01       GW_Armando_plate1_JF27G01       ST        plumb           plumb              119.0
  23 │ GW_Armando_plate1_JF29G01       GW_Armando_plate1_JF29G01       ST        plumb           plumb              120.0
  24 │ GW_Armando_plate1_JF29G02       GW_Armando_plate1_JF29G02       ST        plumb           plumb              121.0
  25 │ GW_Armando_plate1_JF29G03       GW_Armando_plate1_JF29G03       ST        plumb           plumb              122.0
  26 │ GW_Armando_plate1_JG02G02       GW_Armando_plate1_JG02G02       PR        plumb           plumb              145.0
  27 │ GW_Armando_plate1_JG02G04       GW_Armando_plate1_JG02G04       PR        plumb           plumb              146.0
  28 │ GW_Armando_plate1_JG08G01       GW_Armando_plate1_JG08G01       ST        plumb           plumb              123.0
  29 │ GW_Armando_plate1_JG08G02       GW_Armando_plate1_JG08G02       ST        plumb           plumb              124.0
  30 │ GW_Armando_plate1_JG10G01       GW_Armando_plate1_JG10G01       ST        plumb           plumb              125.0
  31 │ GW_Armando_plate1_JG12G01       GW_Armando_plate1_JG12G01       ST        plumb           plumb              126.0
  32 │ GW_Armando_plate1_JG17G01       GW_Armando_plate1_JG17G01       ST        plumb_vir       plumb              127.0
  33 │ GW_Armando_plate1_NO_BC_TTGW05  GW_Armando_plate1_NO_BC_TTGW05  blank     blank           blank              -99.0
  34 │ GW_Armando_plate1_NO_DNA        GW_Armando_plate1_NO_DNA        blank     blank           blank              -99.0
  35 │ GW_Armando_plate1_RF20G01       GW_Armando_plate1_RF20G01       BJ        obs_plumb       plumb_BJ            77.501
  36 │ GW_Armando_plate1_RF29G02       GW_Armando_plate1_RF29G02       BJ        obs_plumb       plumb_BJ            77.502
  37 │ GW_Armando_plate1_TL3           GW_Armando_plate1_TL3           TL        vir             vir                 11.01
  38 │ GW_Armando_plate1_TTGW01        GW_Armando_plate1_TTGW01        MN        troch_MN        troch_west          53.0
  39 │ GW_Armando_plate1_TTGW05_rep1   GW_Armando_plate1_TTGW05_rep1   MN_rep    troch_MN_rep    troch_west_rep      53.0
  40 │ GW_Armando_plate1_TTGW05_rep2   GW_Armando_plate1_TTGW05_rep2   MN        troch_MN        troch_west          53.0
  41 │ GW_Armando_plate1_TTGW06        GW_Armando_plate1_TTGW06        SU        lud_Sukhto      lud_central         47.0
  42 │ GW_Armando_plate1_TTGW07        GW_Armando_plate1_TTGW07        SU        lud_Sukhto      lud_central         47.0
  43 │ GW_Armando_plate1_TTGW10        GW_Armando_plate1_TTGW10        SU        lud_Sukhto      lud_central         47.0
  44 │ GW_Armando_plate1_TTGW11        GW_Armando_plate1_TTGW11        SU        lud_Sukhto      lud_central         47.0
  45 │ GW_Armando_plate1_TTGW13        GW_Armando_plate1_TTGW13        TH        lud_Thallighar  lud_central         43.0
  46 │ GW_Armando_plate1_TTGW17        GW_Armando_plate1_TTGW17        TH        lud_Thallighar  lud_central         43.0
  47 │ GW_Armando_plate1_TTGW19        GW_Armando_plate1_TTGW19        TH        lud_Thallighar  lud_central         43.0
  48 │ GW_Armando_plate1_TTGW21        GW_Armando_plate1_TTGW21        SR        lud_Sural       lud_central         45.0
  49 │ GW_Armando_plate1_TTGW22        GW_Armando_plate1_TTGW22        SR        lud_Sural       lud_central         45.0
  50 │ GW_Armando_plate1_TTGW23        GW_Armando_plate1_TTGW23        SR        lud_Sural       lud_central         45.0
  51 │ GW_Armando_plate1_TTGW29        GW_Armando_plate1_TTGW29        SR        lud_Sural       lud_central         45.0
  52 │ GW_Armando_plate1_TTGW52        GW_Armando_plate1_TTGW52        NG        lud_Nainaghar   lud_central         49.0
  53 │ GW_Armando_plate1_TTGW53        GW_Armando_plate1_TTGW53        NG        lud_Nainaghar   lud_central         49.0
  54 │ GW_Armando_plate1_TTGW55        GW_Armando_plate1_TTGW55        NG        lud_Nainaghar   lud_central         49.0
  55 │ GW_Armando_plate1_TTGW57        GW_Armando_plate1_TTGW57        NG        lud_Nainaghar   lud_central         49.0
  56 │ GW_Armando_plate1_TTGW58        GW_Armando_plate1_TTGW58        NG        lud_Nainaghar   lud_central         49.0
  57 │ GW_Armando_plate1_TTGW59        GW_Armando_plate1_TTGW59        NG        lud_Nainaghar   lud_central         49.0
  58 │ GW_Armando_plate1_TTGW63        GW_Armando_plate1_TTGW63        SP        lud_Spiti       troch_west          55.0
  59 │ GW_Armando_plate1_TTGW64        GW_Armando_plate1_TTGW64        SP        lud_Spiti       troch_west          55.0
  60 │ GW_Armando_plate1_TTGW65        GW_Armando_plate1_TTGW65        SP        lud_Spiti       troch_west          55.0
  61 │ GW_Armando_plate1_TTGW66        GW_Armando_plate1_TTGW66        SP        lud_Spiti       troch_west          55.0
  62 │ GW_Armando_plate1_TTGW68        GW_Armando_plate1_TTGW68        SP        lud_Spiti       troch_west          55.0
  63 │ GW_Armando_plate1_TTGW70        GW_Armando_plate1_TTGW70        SA        lud_Sathrundi   lud_Sath            41.0
  64 │ GW_Armando_plate1_TTGW71        GW_Armando_plate1_TTGW71        SA        lud_Sathrundi   lud_Sath            41.0
  65 │ GW_Armando_plate1_TTGW72        GW_Armando_plate1_TTGW72        SA        lud_Sathrundi   lud_Sath            41.0
  66 │ GW_Armando_plate1_TTGW74        GW_Armando_plate1_TTGW74        SA        lud_Sathrundi   lud_Sath            41.0
  67 │ GW_Armando_plate1_TTGW78        GW_Armando_plate1_TTGW78        SA        lud_Sathrundi   lud_Sath            41.0
  68 │ GW_Armando_plate1_TTGW_15_05    GW_Armando_plate1_TTGW_15_05    SR        lud_Sural       lud_central         45.0
  69 │ GW_Armando_plate1_TTGW_15_07    GW_Armando_plate1_TTGW_15_07    SR        lud_Sural       lud_central         45.0
  70 │ GW_Armando_plate1_TTGW_15_08    GW_Armando_plate1_TTGW_15_08    SR        lud_Sural       lud_central         45.0
  71 │ GW_Armando_plate1_TTGW_15_09    GW_Armando_plate1_TTGW_15_09    SR        lud_Sural       lud_central         45.0
  72 │ GW_Armando_plate1_UY1           GW_Armando_plate1_UY1           UY        plumb           plumb               87.0
  73 │ GW_Armando_plate2_IL2           GW_Armando_plate2_IL2           IL_rep    plumb_rep       plumb_rep           84.0
  74 │ GW_Armando_plate2_JE31G01       GW_Armando_plate2_JE31G01       VB_vi     vir_misID       vir                 24.002
  75 │ GW_Armando_plate2_JF03G01       GW_Armando_plate2_JF03G01       ST_vi     vir_misID       vir                 24.003
  76 │ GW_Armando_plate2_JF03G02       GW_Armando_plate2_JF03G02       VB_vi     vir_misID       vir                 24.004
  77 │ GW_Armando_plate2_JF07G01       GW_Armando_plate2_JF07G01       ST        plumb           plumb              128.0
  78 │ GW_Armando_plate2_JF08G04       GW_Armando_plate2_JF08G04       ST        plumb           plumb              129.0
  79 │ GW_Armando_plate2_JF10G02       GW_Armando_plate2_JF10G02       ST        plumb           plumb              130.0
  80 │ GW_Armando_plate2_JF11G02       GW_Armando_plate2_JF11G02       ST        plumb           plumb              131.0
  81 │ GW_Armando_plate2_JF12G03       GW_Armando_plate2_JF12G03       ST        plumb           plumb              132.0
  82 │ GW_Armando_plate2_JF12G05       GW_Armando_plate2_JF12G05       ST        plumb           plumb              133.0
  83 │ GW_Armando_plate2_JF13G02       GW_Armando_plate2_JF13G02       ST        plumb           plumb              134.0
  84 │ GW_Armando_plate2_JF14G01       GW_Armando_plate2_JF14G01       DV        plumb           plumb              104.0
  85 │ GW_Armando_plate2_JF14G02       GW_Armando_plate2_JF14G02       DV        plumb           plumb              105.0
  86 │ GW_Armando_plate2_JF15G01       GW_Armando_plate2_JF15G01       DV        plumb           plumb              106.0
  87 │ GW_Armando_plate2_JF15G02       GW_Armando_plate2_JF15G02       DV        plumb           plumb              107.0
  88 │ GW_Armando_plate2_JF16G02       GW_Armando_plate2_JF16G02       DV_vi     plumb_vir       vir                 24.042
  89 │ GW_Armando_plate2_JF19G01       GW_Armando_plate2_JF19G01       MB        plumb           plumb               96.0
  90 │ GW_Armando_plate2_JF20G02       GW_Armando_plate2_JF20G02       MB        plumb           plumb               97.0
  91 │ GW_Armando_plate2_JF24G01       GW_Armando_plate2_JF24G01       VB        plumb           plumb              101.0
  92 │ GW_Armando_plate2_JF24G03       GW_Armando_plate2_JF24G03       ST        plumb           plumb              135.0
  93 │ GW_Armando_plate2_JF25G01       GW_Armando_plate2_JF25G01       VB        plumb           plumb              102.0
  94 │ GW_Armando_plate2_JF26G02       GW_Armando_plate2_JF26G02       ST        plumb           plumb              136.0
  95 │ GW_Armando_plate2_JF27G02       GW_Armando_plate2_JF27G02       ST        plumb           plumb              137.0
  96 │ GW_Armando_plate2_JF30G01       GW_Armando_plate2_JF30G01       ST_vi     vir_misID       vir                 24.005
  97 │ GW_Armando_plate2_JG01G01       GW_Armando_plate2_JG01G01       PR        plumb           plumb              147.0
  98 │ GW_Armando_plate2_JG02G01       GW_Armando_plate2_JG02G01       PR        plumb           plumb              148.0
  99 │ GW_Armando_plate2_JG02G03       GW_Armando_plate2_JG02G03       PR        plumb           plumb              149.0
 100 │ GW_Armando_plate2_JG10G02       GW_Armando_plate2_JG10G02       ST        plumb           plumb              138.0
 101 │ GW_Armando_plate2_JG10G03       GW_Armando_plate2_JG10G03       ST        plumb           plumb              139.0
 102 │ GW_Armando_plate2_JG12G02       GW_Armando_plate2_JG12G02       ST        plumb           plumb              140.0
 103 │ GW_Armando_plate2_JG12G03       GW_Armando_plate2_JG12G03       ST        plumb           plumb              141.0
 104 │ GW_Armando_plate2_LN11          GW_Armando_plate2_LN11          LN_rep    troch_LN_rep    troch_LN_rep        65.01
 105 │ GW_Armando_plate2_LN2           GW_Armando_plate2_LN2           LN        troch_LN        troch_LN            58.01
 106 │ GW_Armando_plate2_NO_BC_TTGW05  GW_Armando_plate2_NO_BC_TTGW05  blank     blank           blank              -99.0
 107 │ GW_Armando_plate2_NO_DNA        GW_Armando_plate2_NO_DNA        blank     blank           blank              -99.0
 108 │ GW_Armando_plate2_RF29G01       GW_Armando_plate2_RF29G01       BJ        obs_plumb       plumb_BJ            77.503
 109 │ GW_Armando_plate2_TTGW02        GW_Armando_plate2_TTGW02        MN        troch_MN        troch_west          53.0
 110 │ GW_Armando_plate2_TTGW03        GW_Armando_plate2_TTGW03        MN        troch_MN        troch_west          53.0
 111 │ GW_Armando_plate2_TTGW05_rep3   GW_Armando_plate2_TTGW05_rep3   MN_rep    troch_MN_rep    troch_west_rep      53.0
 112 │ GW_Armando_plate2_TTGW05_rep4   GW_Armando_plate2_TTGW05_rep4   MN_rep    troch_MN_rep    troch_west_rep      53.0
 113 │ GW_Armando_plate2_TTGW08        GW_Armando_plate2_TTGW08        SU        lud_Sukhto      lud_central         47.0
 114 │ GW_Armando_plate2_TTGW09        GW_Armando_plate2_TTGW09        SU        lud_Sukhto      lud_central         47.0
 115 │ GW_Armando_plate2_TTGW12        GW_Armando_plate2_TTGW12        TH        lud_Thallighar  lud_central         43.0
 116 │ GW_Armando_plate2_TTGW14        GW_Armando_plate2_TTGW14        TH        lud_Thallighar  lud_central         43.0
 117 │ GW_Armando_plate2_TTGW15        GW_Armando_plate2_TTGW15        TH        lud_Thallighar  lud_central         43.0
 118 │ GW_Armando_plate2_TTGW16        GW_Armando_plate2_TTGW16        TH        lud_Thallighar  lud_central         43.0
 119 │ GW_Armando_plate2_TTGW18        GW_Armando_plate2_TTGW18        TH        lud_Thallighar  lud_central         43.0
 120 │ GW_Armando_plate2_TTGW20        GW_Armando_plate2_TTGW20        SR        lud_Sural       lud_central         45.0
 121 │ GW_Armando_plate2_TTGW24        GW_Armando_plate2_TTGW24        SR        lud_Sural       lud_central         45.0
 122 │ GW_Armando_plate2_TTGW25        GW_Armando_plate2_TTGW25        SR        lud_Sural       lud_central         45.0
 123 │ GW_Armando_plate2_TTGW27        GW_Armando_plate2_TTGW27        SR        lud_Sural       lud_central         45.0
 124 │ GW_Armando_plate2_TTGW28        GW_Armando_plate2_TTGW28        SR        lud_Sural       lud_central         45.0
 125 │ GW_Armando_plate2_TTGW50        GW_Armando_plate2_TTGW50        NG        lud_Nainaghar   lud_central         49.0
 126 │ GW_Armando_plate2_TTGW51        GW_Armando_plate2_TTGW51        NG        lud_Nainaghar   lud_central         49.0
 127 │ GW_Armando_plate2_TTGW54        GW_Armando_plate2_TTGW54        NG        lud_Nainaghar   lud_central         49.0
 128 │ GW_Armando_plate2_TTGW56        GW_Armando_plate2_TTGW56        NG        lud_Nainaghar   lud_central         49.0
 129 │ GW_Armando_plate2_TTGW60        GW_Armando_plate2_TTGW60        SP        lud_Spiti       troch_west          55.0
 130 │ GW_Armando_plate2_TTGW61        GW_Armando_plate2_TTGW61        SP        lud_Spiti       troch_west          55.0
 131 │ GW_Armando_plate2_TTGW62        GW_Armando_plate2_TTGW62        SP        lud_Spiti       troch_west          55.0
 132 │ GW_Armando_plate2_TTGW67        GW_Armando_plate2_TTGW67        SP        lud_Spiti       troch_west          55.0
 133 │ GW_Armando_plate2_TTGW69        GW_Armando_plate2_TTGW69        SP        lud_Spiti       troch_west          55.0
 134 │ GW_Armando_plate2_TTGW73        GW_Armando_plate2_TTGW73        SA        lud_Sathrundi   lud_Sath            41.0
 135 │ GW_Armando_plate2_TTGW75        GW_Armando_plate2_TTGW75        SA        lud_Sathrundi   lud_Sath            41.0
 136 │ GW_Armando_plate2_TTGW77        GW_Armando_plate2_TTGW77        SA        lud_Sathrundi   lud_Sath            41.0
 137 │ GW_Armando_plate2_TTGW79        GW_Armando_plate2_TTGW79        SA        lud_Sathrundi   lud_Sath            41.0
 138 │ GW_Armando_plate2_TTGW80        GW_Armando_plate2_TTGW80        SA        lud_Sathrundi   lud_Sath            41.0
 139 │ GW_Armando_plate2_TTGW_15_01    GW_Armando_plate2_TTGW_15_01    SR        lud_Sural       lud_central         45.0
 140 │ GW_Armando_plate2_TTGW_15_02    GW_Armando_plate2_TTGW_15_02    SR        lud_Sural       lud_central         45.0
 141 │ GW_Armando_plate2_TTGW_15_03    GW_Armando_plate2_TTGW_15_03    SR        lud_Sural       lud_central         45.0
 142 │ GW_Armando_plate2_TTGW_15_04    GW_Armando_plate2_TTGW_15_04    SR        lud_Sural       lud_central         45.0
 143 │ GW_Armando_plate2_TTGW_15_06    GW_Armando_plate2_TTGW_15_06    SR        lud_Sural       lud_central         45.0
 144 │ GW_Armando_plate2_TTGW_15_10    GW_Armando_plate2_TTGW_15_10    SR        lud_Sural       lud_central         45.0
 145 │ GW_Lane5_AA1                    GW_Lane5_AA1                    AA        vir_S           vir_S               25.0
 146 │ GW_Lane5_AA10                   GW_Lane5_AA10                   AA        vir_S           vir_S               33.0
 147 │ GW_Lane5_AA11                   GW_Lane5_AA11                   AA        vir_S           vir_S               34.0
 148 │ GW_Lane5_AA3                    GW_Lane5_AA3                    AA        vir_S           vir_S               26.0
 149 │ GW_Lane5_AA4                    GW_Lane5_AA4                    AA        vir_S           vir_S               27.0
 150 │ GW_Lane5_AA5                    GW_Lane5_AA5                    AA        vir_S           vir_S               28.0
 151 │ GW_Lane5_AA6                    GW_Lane5_AA6                    AA        vir_S           vir_S               29.0
 152 │ GW_Lane5_AA7                    GW_Lane5_AA7                    AA        vir_S           vir_S               30.0
 153 │ GW_Lane5_AA8                    GW_Lane5_AA8                    AA        vir_S           vir_S               31.0
 154 │ GW_Lane5_AA9                    GW_Lane5_AA9                    AA        vir_S           vir_S               32.0
 155 │ GW_Lane5_AB1                    GW_Lane5_AB1                    AB_rep    vir_rep         vir_rep             20.0
 156 │ GW_Lane5_AB2                    GW_Lane5_AB2                    AB        vir             vir                 21.0
 157 │ GW_Lane5_AN1                    GW_Lane5_AN1                    AN        plumb           plumb               80.0
 158 │ GW_Lane5_AN2                    GW_Lane5_AN2                    AN        plumb           plumb               81.0
 159 │ GW_Lane5_BK2                    GW_Lane5_BK2                    BK        plumb           plumb               78.0
 160 │ GW_Lane5_BK3                    GW_Lane5_BK3                    BK        plumb           plumb               79.0
 161 │ GW_Lane5_DA2                    GW_Lane5_DA2                    XN        obs             obs                 73.0
 162 │ GW_Lane5_DA3                    GW_Lane5_DA3                    XN        obs             obs                 74.0
 163 │ GW_Lane5_DA4                    GW_Lane5_DA4                    XN        obs             obs                 75.0
 164 │ GW_Lane5_DA6                    GW_Lane5_DA6                    XN        obs             low_reads           76.0
 165 │ GW_Lane5_DA7                    GW_Lane5_DA7                    XN        obs             obs                 77.0
 166 │ GW_Lane5_EM1                    GW_Lane5_EM1                    EM        troch_EM        troch_EM            72.0
 167 │ GW_Lane5_IL1                    GW_Lane5_IL1                    IL        plumb           plumb               82.0
 168 │ GW_Lane5_IL2                    GW_Lane5_IL2                    IL_rep    plumb_rep       plumb_rep           85.0
 169 │ GW_Lane5_IL4                    GW_Lane5_IL4                    IL        plumb           plumb               83.0
 170 │ GW_Lane5_KS1                    GW_Lane5_KS1                    OV        lud_KS          lud_KS              40.0
 171 │ GW_Lane5_KS2                    GW_Lane5_KS2                    OV        lud_KS          lud_KS              40.0
 172 │ GW_Lane5_LN1                    GW_Lane5_LN1                    LN        troch_LN        troch_LN            57.0
 173 │ GW_Lane5_LN10                   GW_Lane5_LN10                   LN        troch_LN        troch_LN            64.0
 174 │ GW_Lane5_LN11                   GW_Lane5_LN11                   LN        troch_LN        troch_LN            65.0
 175 │ GW_Lane5_LN12                   GW_Lane5_LN12                   LN        troch_LN        troch_LN            66.0
 176 │ GW_Lane5_LN14                   GW_Lane5_LN14                   LN        troch_LN        troch_LN            67.0
 177 │ GW_Lane5_LN16                   GW_Lane5_LN16                   LN        troch_LN        troch_LN            68.0
 178 │ GW_Lane5_LN18                   GW_Lane5_LN18                   LN        troch_LN        troch_LN            69.0
 179 │ GW_Lane5_LN19                   GW_Lane5_LN19                   LN        troch_LN        troch_LN            70.0
 180 │ GW_Lane5_LN2                    GW_Lane5_LN2                    LN_rep    troch_LN_rep    troch_LN_rep        58.0
 181 │ GW_Lane5_LN20                   GW_Lane5_LN20                   LN        troch_LN        troch_LN            71.0
 182 │ GW_Lane5_LN3                    GW_Lane5_LN3                    LN        troch_LN        troch_LN            59.0
 183 │ GW_Lane5_LN4                    GW_Lane5_LN4                    LN        troch_LN        troch_LN            60.0
 184 │ GW_Lane5_LN6                    GW_Lane5_LN6                    LN        troch_LN        troch_LN            61.0
 185 │ GW_Lane5_LN7                    GW_Lane5_LN7                    LN        troch_LN        troch_LN            62.0
 186 │ GW_Lane5_LN8                    GW_Lane5_LN8                    LN        troch_LN        troch_LN            63.0
 187 │ GW_Lane5_MN1                    GW_Lane5_MN1                    MN        troch_MN        troch_west          51.0
 188 │ GW_Lane5_MN12                   GW_Lane5_MN12                   MN        troch_MN        troch_west          56.0
 189 │ GW_Lane5_MN3                    GW_Lane5_MN3                    MN        troch_MN        troch_west          52.0
 190 │ GW_Lane5_MN5                    GW_Lane5_MN5                    MN        troch_MN        troch_west          53.0
 191 │ GW_Lane5_MN8                    GW_Lane5_MN8                    MN        troch_MN        troch_west          54.0
 192 │ GW_Lane5_MN9                    GW_Lane5_MN9                    MN        troch_MN        troch_west          55.0
 193 │ GW_Lane5_NA1                    GW_Lane5_NA1                    NR        lud_PK          lud_PK              39.2
 194 │ GW_Lane5_NA3-3ul                GW_Lane5_NA3-3ul                NR        lud_PK          lud_PK              39.2
 195 │ GW_Lane5_PT11                   GW_Lane5_PT11                   KL        lud_KL          lud_central         42.0
 196 │ GW_Lane5_PT12                   GW_Lane5_PT12                   KL        lud_KL          lud_central         42.0
 197 │ GW_Lane5_PT2                    GW_Lane5_PT2                    ML        lud_ML          lud_ML              51.0
 198 │ GW_Lane5_PT3                    GW_Lane5_PT3                    PA        lud_PA          lud_central         46.0
 199 │ GW_Lane5_PT4                    GW_Lane5_PT4                    PA        lud_PA          lud_central         46.0
 200 │ GW_Lane5_PT6                    GW_Lane5_PT6                    KL        lud_KL          lud_central         42.0
 201 │ GW_Lane5_SH1                    GW_Lane5_SH1                    SH        lud_PK          lud_PK              39.1
 202 │ GW_Lane5_SH2                    GW_Lane5_SH2                    SH        lud_PK          lud_PK              39.1
 203 │ GW_Lane5_SH4                    GW_Lane5_SH4                    SH        lud_PK          lud_PK              39.1
 204 │ GW_Lane5_SH5                    GW_Lane5_SH5                    SH        lud_PK          lud_PK              39.1
 205 │ GW_Lane5_SL1                    GW_Lane5_SL1                    SL        plumb           plumb              150.0
 206 │ GW_Lane5_SL2                    GW_Lane5_SL2                    SL        plumb           plumb              151.0
 207 │ GW_Lane5_ST1                    GW_Lane5_ST1                    ST        plumb           plumb              142.0
 208 │ GW_Lane5_ST12                   GW_Lane5_ST12                   ST        plumb           plumb              144.0
 209 │ GW_Lane5_ST3                    GW_Lane5_ST3                    ST        plumb           plumb              143.0
 210 │ GW_Lane5_STvi1                  GW_Lane5_STvi1                  ST_vi     vir             vir                 22.0
 211 │ GW_Lane5_STvi2                  GW_Lane5_STvi2                  ST_vi     vir             vir                 23.0
 212 │ GW_Lane5_STvi3                  GW_Lane5_STvi3                  ST_vi     vir             vir                 24.0
 213 │ GW_Lane5_TA1                    GW_Lane5_TA1                    TA        plumb           plumb               86.0
 214 │ GW_Lane5_TL1                    GW_Lane5_TL1                    TL        vir             vir                  9.0
 215 │ GW_Lane5_TL10                   GW_Lane5_TL10                   TL        vir             vir                 17.0
 216 │ GW_Lane5_TL11                   GW_Lane5_TL11                   TL        vir             vir                 18.0
 217 │ GW_Lane5_TL12                   GW_Lane5_TL12                   TL        vir             vir                 19.0
 218 │ GW_Lane5_TL2                    GW_Lane5_TL2                    TL        vir             vir                 10.0
 219 │ GW_Lane5_TL3                    GW_Lane5_TL3                    TL_rep    vir_rep         vir_rep             11.0
 220 │ GW_Lane5_TL4                    GW_Lane5_TL4                    TL        vir             vir                 12.0
 221 │ GW_Lane5_TL5                    GW_Lane5_TL5                    TL        vir             vir                 13.0
 222 │ GW_Lane5_TL7                    GW_Lane5_TL7                    TL        vir             vir                 14.0
 223 │ GW_Lane5_TL8                    GW_Lane5_TL8                    TL        vir             vir                 15.0
 224 │ GW_Lane5_TL9                    GW_Lane5_TL9                    TL        vir             vir                 16.0
 225 │ GW_Lane5_TU1                    GW_Lane5_TU1                    TU        nit             nit                 35.0
 226 │ GW_Lane5_TU2                    GW_Lane5_TU2                    TU        nit             nit                 36.0
 227 │ GW_Lane5_UY1                    GW_Lane5_UY1                    UY_rep    plumb_rep       plumb_rep           93.0
 228 │ GW_Lane5_UY2                    GW_Lane5_UY2                    UY        plumb           plumb               88.0
 229 │ GW_Lane5_UY3                    GW_Lane5_UY3                    UY        plumb           plumb               89.0
 230 │ GW_Lane5_UY4                    GW_Lane5_UY4                    UY        plumb           plumb               90.0
 231 │ GW_Lane5_UY5                    GW_Lane5_UY5                    UY        plumb           plumb               91.0
 232 │ GW_Lane5_UY6                    GW_Lane5_UY6                    UY        plumb           plumb               92.0
 233 │ GW_Lane5_YK1                    GW_Lane5_YK1                    YK        vir             vir                  1.0
 234 │ GW_Lane5_YK11                   GW_Lane5_YK11                   YK        vir             vir                  8.0
 235 │ GW_Lane5_YK3                    GW_Lane5_YK3                    YK        vir             vir                  2.0
 236 │ GW_Lane5_YK4                    GW_Lane5_YK4                    YK        vir             vir                  3.0
 237 │ GW_Lane5_YK5                    GW_Lane5_YK5                    YK        vir             vir                  4.0
 238 │ GW_Lane5_YK6                    GW_Lane5_YK6                    YK        vir             vir                  5.0
 239 │ GW_Lane5_YK7                    GW_Lane5_YK7                    YK        vir             vir                  6.0
 240 │ GW_Lane5_YK9                    GW_Lane5_YK9                    YK        vir             vir                  7.0
 241 │ GW_Liz_GBS_Liz10045             GW_Liz_GBS_Liz10045             ML        lud             lud_ML              51.01
 242 │ GW_Liz_GBS_Liz10094             GW_Liz_GBS_Liz10094             ML        lud             lud_ML              51.02
 243 │ GW_Liz_GBS_Liz5101              GW_Liz_GBS_Liz5101              ML        lud             lud_ML              51.03
 244 │ GW_Liz_GBS_Liz5101_R            GW_Liz_GBS_Liz5101_R            ML_rep    lud_rep         lud_ML_rep          51.04
 245 │ GW_Liz_GBS_Liz5118              GW_Liz_GBS_Liz5118              ML        lud             lud_ML              51.05
 246 │ GW_Liz_GBS_Liz5139              GW_Liz_GBS_Liz5139              ML        lud             lud_ML              51.06
 247 │ GW_Liz_GBS_Liz5142              GW_Liz_GBS_Liz5142              ML        lud             lud_ML              51.07
 248 │ GW_Liz_GBS_Liz5144              GW_Liz_GBS_Liz5144              ML        lud             lud_ML              51.08
 249 │ GW_Liz_GBS_Liz5150              GW_Liz_GBS_Liz5150              ML        lud             lud_ML              51.09
 250 │ GW_Liz_GBS_Liz5159              GW_Liz_GBS_Liz5159              ML        lud_chick       lud_ML              51.1
 251 │ GW_Liz_GBS_Liz5162              GW_Liz_GBS_Liz5162              ML        lud_chick       lud_ML              51.11
 252 │ GW_Liz_GBS_Liz5163              GW_Liz_GBS_Liz5163              ML        lud_chick       lud_ML              51.12
 253 │ GW_Liz_GBS_Liz5164              GW_Liz_GBS_Liz5164              ML        lud_chick       lud_ML              51.13
 254 │ GW_Liz_GBS_Liz5165              GW_Liz_GBS_Liz5165              ML        lud             lud_ML              51.14
 255 │ GW_Liz_GBS_Liz5167              GW_Liz_GBS_Liz5167              ML        lud_chick       lud_ML              51.15
 256 │ GW_Liz_GBS_Liz5168              GW_Liz_GBS_Liz5168              ML        lud_chick       lud_ML              51.16
 257 │ GW_Liz_GBS_Liz5169              GW_Liz_GBS_Liz5169              ML        lud_chick       lud_ML              51.17
 258 │ GW_Liz_GBS_Liz5171              GW_Liz_GBS_Liz5171              ML        lud             lud_ML              51.18
 259 │ GW_Liz_GBS_Liz5172              GW_Liz_GBS_Liz5172              ML        lud_chick       lud_ML              51.19
 260 │ GW_Liz_GBS_Liz5173              GW_Liz_GBS_Liz5173              ML        lud_chick       lud_ML              51.2
 261 │ GW_Liz_GBS_Liz5174              GW_Liz_GBS_Liz5174              ML        lud             lud_ML              51.21
 262 │ GW_Liz_GBS_Liz5175              GW_Liz_GBS_Liz5175              ML        lud             lud_ML              51.22
 263 │ GW_Liz_GBS_Liz5176              GW_Liz_GBS_Liz5176              ML        lud             lud_ML              51.23
 264 │ GW_Liz_GBS_Liz5177              GW_Liz_GBS_Liz5177              ML        lud_chick       lud_ML              51.24
 265 │ GW_Liz_GBS_Liz5178              GW_Liz_GBS_Liz5178              ML        lud_chick       lud_ML              51.25
 266 │ GW_Liz_GBS_Liz5179              GW_Liz_GBS_Liz5179              ML        lud_chick       lud_ML              51.26
 267 │ GW_Liz_GBS_Liz5180              GW_Liz_GBS_Liz5180              ML        lud             lud_ML              51.27
 268 │ GW_Liz_GBS_Liz5182              GW_Liz_GBS_Liz5182              ML        lud_chick       lud_ML              51.28
 269 │ GW_Liz_GBS_Liz5184              GW_Liz_GBS_Liz5184              ML        lud_chick       lud_ML              51.29
 270 │ GW_Liz_GBS_Liz5185              GW_Liz_GBS_Liz5185              ML        lud             lud_ML              51.3
 271 │ GW_Liz_GBS_Liz5186              GW_Liz_GBS_Liz5186              ML        lud_chick       lud_ML              51.31
 272 │ GW_Liz_GBS_Liz5187              GW_Liz_GBS_Liz5187              ML        lud_chick       lud_ML              51.32
 273 │ GW_Liz_GBS_Liz5188              GW_Liz_GBS_Liz5188              ML        lud             lud_ML              51.33
 274 │ GW_Liz_GBS_Liz5189              GW_Liz_GBS_Liz5189              ML        lud_chick       lud_ML              51.34
 275 │ GW_Liz_GBS_Liz5190              GW_Liz_GBS_Liz5190              ML        lud_chick       lud_ML              51.35
 276 │ GW_Liz_GBS_Liz5191              GW_Liz_GBS_Liz5191              ML        lud_chick       lud_ML              51.36
 277 │ GW_Liz_GBS_Liz5192              GW_Liz_GBS_Liz5192              ML        lud_chick       lud_ML              51.37
 278 │ GW_Liz_GBS_Liz5193              GW_Liz_GBS_Liz5193              ML        lud_chick       lud_ML              51.38
 279 │ GW_Liz_GBS_Liz5194              GW_Liz_GBS_Liz5194              ML        lud_chick       lud_ML              51.39
 280 │ GW_Liz_GBS_Liz5195              GW_Liz_GBS_Liz5195              ML        lud             lud_ML              51.4
 281 │ GW_Liz_GBS_Liz5197              GW_Liz_GBS_Liz5197              ML        lud             lud_ML              51.41
 282 │ GW_Liz_GBS_Liz5199              GW_Liz_GBS_Liz5199              ML        lud_chick       lud_ML              51.42
 283 │ GW_Liz_GBS_Liz6002              GW_Liz_GBS_Liz6002              ML        lud             lud_ML              51.43
 284 │ GW_Liz_GBS_Liz6006              GW_Liz_GBS_Liz6006              ML        lud             lud_ML              51.44
 285 │ GW_Liz_GBS_Liz6008              GW_Liz_GBS_Liz6008              ML        lud             lud_ML              51.45
 286 │ GW_Liz_GBS_Liz6009              GW_Liz_GBS_Liz6009              ML        lud             lud_ML              51.46
 287 │ GW_Liz_GBS_Liz6010              GW_Liz_GBS_Liz6010              ML        lud             lud_ML              51.47
 288 │ GW_Liz_GBS_Liz6012              GW_Liz_GBS_Liz6012              ML        lud             lud_ML              51.48
 289 │ GW_Liz_GBS_Liz6014              GW_Liz_GBS_Liz6014              ML        lud             lud_ML              51.49
 290 │ GW_Liz_GBS_Liz6055              GW_Liz_GBS_Liz6055              ML        lud             lud_ML              51.5
 291 │ GW_Liz_GBS_Liz6057              GW_Liz_GBS_Liz6057              ML        lud             lud_ML              51.51
 292 │ GW_Liz_GBS_Liz6060              GW_Liz_GBS_Liz6060              ML        lud             lud_ML              51.52
 293 │ GW_Liz_GBS_Liz6062              GW_Liz_GBS_Liz6062              ML        lud             lud_ML              51.53
 294 │ GW_Liz_GBS_Liz6063              GW_Liz_GBS_Liz6063              ML        lud             lud_ML              51.54
 295 │ GW_Liz_GBS_Liz6066              GW_Liz_GBS_Liz6066              ML        lud             lud_ML              51.55
 296 │ GW_Liz_GBS_Liz6072              GW_Liz_GBS_Liz6072              ML        lud             lud_ML              51.56
 297 │ GW_Liz_GBS_Liz6079              GW_Liz_GBS_Liz6079              ML        lud             lud_ML              51.57
 298 │ GW_Liz_GBS_Liz6203              GW_Liz_GBS_Liz6203              ML        lud_chick       lud_ML              51.58
 299 │ GW_Liz_GBS_Liz6204              GW_Liz_GBS_Liz6204              ML        lud_chick       lud_ML              51.59
 300 │ GW_Liz_GBS_Liz6461              GW_Liz_GBS_Liz6461              ML        lud             lud_ML              51.6
 301 │ GW_Liz_GBS_Liz6472              GW_Liz_GBS_Liz6472              ML        lud             lud_ML              51.61
 302 │ GW_Liz_GBS_Liz6478              GW_Liz_GBS_Liz6478              ML        lud             lud_ML              51.62
 303 │ GW_Liz_GBS_Liz6766              GW_Liz_GBS_Liz6766              ML        lud             lud_ML              51.63
 304 │ GW_Liz_GBS_Liz6776              GW_Liz_GBS_Liz6776              ML        lud             lud_ML              51.64
 305 │ GW_Liz_GBS_Liz6794              GW_Liz_GBS_Liz6794              ML        lud             lud_ML              51.65
 306 │ GW_Liz_GBS_P_fusc               GW_Liz_GBS_P_fusc               fusc      fusc            fusc               201.0
 307 │ GW_Liz_GBS_P_h_man              GW_Liz_GBS_P_h_man              hmand     hmand           hmand              202.0
 308 │ GW_Liz_GBS_P_humei              GW_Liz_GBS_P_humei              hume      hume            hume               203.0
 309 │ GW_Liz_GBS_P_inor               GW_Liz_GBS_P_inor               inor      inor            inor               204.0
 310 │ GW_Liz_GBS_S_burk               GW_Liz_GBS_S_burk               burk      burk            burk               205.0

GOOD NEWS: names of individuals in metadata file and genotype ind file match perfectly.</code></pre>
</div>
</div>
</section>
<section id="polish-a-few-individual-names-to-match-those-in-other-metadata-object-above-and-make-more-readable-graphs-1" class="level3">
<h3 class="anchored" data-anchor-id="polish-a-few-individual-names-to-match-those-in-other-metadata-object-above-and-make-more-readable-graphs-1">Polish a few individual names (to match those in other metadata object above, and make more readable graphs):</h3>
<div id="c76f8403" class="cell" data-execution_count="191">
<div class="sourceCode cell-code" id="cb517"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb517-1"><a href="#cb517-1" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr3.ind <span class="op">=</span> <span class="fu">correctNames</span>(ind_with_metadata_chr3.ind)</span>
<span id="cb517-2"><a href="#cb517-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb517-3"><a href="#cb517-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr3.ID <span class="op">=</span> <span class="fu">correctNames</span>(ind_with_metadata_chr3.ID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="192">
<pre><code>310-element Vector{String}:
 "GW_Armando_plate1_AB1"
 "GW_Armando_plate1_JF07G02"
 "GW_Armando_plate1_JF07G03"
 "GW_Armando_plate1_JF07G04"
 "GW_Armando_plate1_JF08G02"
 "GW_Armando_plate1_JF09G01"
 "GW_Armando_plate1_JF09G02"
 "GW_Armando_plate1_JF10G03"
 "GW_Armando_plate1_JF11G01"
 "GW_Armando_plate1_JF12G01"
 "GW_Armando_plate1_JF12G02"
 "GW_Armando_plate1_JF12G04"
 "GW_Armando_plate1_JF13G01"
 ⋮
 "GW_Liz_GBS_Liz6204"
 "GW_Liz_GBS_Liz6461"
 "GW_Liz_GBS_Liz6472"
 "GW_Liz_GBS_Liz6478"
 "GW_Liz_GBS_Liz6766"
 "GW_Liz_GBS_Liz6776"
 "GW_Liz_GBS_Liz6794"
 "GW_Liz_GBS_P_fusc"
 "GW_Liz_GBS_P_h_man"
 "GW_Liz_GBS_P_humei"
 "GW_Liz_GBS_P_inor"
 "GW_Liz_GBS_S_burk"</code></pre>
</div>
</div>
</section>
<section id="filter-to-just-the-individuals-also-included-in-the-analysis-of-lhbrs-above-1" class="level3">
<h3 class="anchored" data-anchor-id="filter-to-just-the-individuals-also-included-in-the-analysis-of-lhbrs-above-1">Filter to just the individuals also included in the analysis of LHBRs above</h3>
<div id="c821c11d" class="cell" data-execution_count="192">
<div class="sourceCode cell-code" id="cb519"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb519-1"><a href="#cb519-1" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(ind_with_metadata_included.ind), ind_with_metadata_chr3.ind)</span>
<span id="cb519-2"><a href="#cb519-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb519-3"><a href="#cb519-3" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chr3_included <span class="op">=</span> ind_with_metadata_chr3[selection, <span class="op">:</span>]</span>
<span id="cb519-4"><a href="#cb519-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb519-5"><a href="#cb519-5" aria-hidden="true" tabindex="-1"></a><span class="co"># select genotypes of just the included individuals, and ignore first column</span></span>
<span id="cb519-6"><a href="#cb519-6" aria-hidden="true" tabindex="-1"></a>geno_chr3_included <span class="op">=</span> geno_chr3[selection, <span class="fl">2</span><span class="op">:</span><span class="kw">end</span>]</span>
<span id="cb519-7"><a href="#cb519-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb519-8"><a href="#cb519-8" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb519-9"><a href="#cb519-9" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(ind_with_metadata_included.gw3_cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>["virLud", "plumb", "plumb", "plumb", "plumb", "plumb", "plumbHet", "plumb", "plumb", "plumbHet", "plumb", "virLud", "plumb", "plumb", "virLud", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "vir_plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "plumb", "virLud", "trochObs", "trochObs", "virLud", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "virLud", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "trochObs", "virLud", "virLudHet", "virLud_trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "virLud_trochObs", "virLud_trochObs", "virLud", "virLud_trochObs", "plumb", "virLud", "virLudHet", "virLud", "plumb", "plumb", "plumb", "plumbHet", "plumb", "plumb", "plumb", "plumb", "vir_plumb", "plumb", "plumb", "virLud", "plumb", "plumb", "plumbHet", "plumb", "plumbHet", "plumbHet", "plumb", "virLud", "plumbHet", "plumb", "plumb", "plumb", "vir_plumb", "plumb", "plumb", "trochObs", "plumb", "trochObs", "trochObs", "virLudHet", "virLud_trochObs", "virLudHet", "virLud_trochObs", "virLud", "virLud", "virLud", "virLudHet", "virLud", "virLudHet", "virLudHet", "virLud_trochObs", "virLud", "virLud", "trochObsHet", "trochObs", "trochObs", "trochObs", "trochObs", "trochObsHet", "trochObs", "trochObs", "trochObs", "trochObs", "virLud", "virLud", "virLud", "virLud", "virLud_trochObs", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "plumb", "plumb", "plumb", "plumb", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "plumb", "plumb", "virLud", "virLud", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "virLud_trochObs", "trochObs", "trochObs", "virLud", "virLud", "virLud_trochObs", "virLud", "trochObs", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "plumb", "plumb", "plumb", "plumb", "plumb", "virLud", "virLud", "virLud", "plumb", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "nit", "nit", "plumb", "plumb", "plumb", "plumb", "plumb", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "virLud", "trochObs", "trochObs", "trochObs", "trochObs", "virLud_trochObs", "virLud_trochObs", "trochObs", "virLud_trochObs", "trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "trochObs", "virLud_trochObs", "trochObs", "virLud_trochObs", "trochObs", "trochObs", "virLud_trochObs", "trochObsHet", "trochObs", "virLud_trochObs", "virLud_trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "virLud_trochObs", "virLud_trochObs", "virLud_trochObs", "trochObs", "virLud_trochObs", "trochObs", "trochObs", "trochObs", "virLud_trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs", "trochObs"]</code></pre>
</div>
</div>
</section>
<section id="look-up-the-chr-3-individual-membership-in-homozygous-clusters-and-calculate-pi-and-dxy" class="level3">
<h3 class="anchored" data-anchor-id="look-up-the-chr-3-individual-membership-in-homozygous-clusters-and-calculate-pi-and-dxy">Look up the chr 3 individual membership in homozygous clusters, and calculate pi and Dxy</h3>
<div id="6f0caea2" class="cell" data-execution_count="193">
<div class="sourceCode cell-code" id="cb521"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb521-1"><a href="#cb521-1" aria-hidden="true" tabindex="-1"></a>indClusterMembership_gw3 <span class="op">=</span> ind_with_metadata_included.gw3_cluster</span>
<span id="cb521-2"><a href="#cb521-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-3"><a href="#cb521-3" aria-hidden="true" tabindex="-1"></a>clusterNames_gw3 <span class="op">=</span> [<span class="st">"virLud"</span>,</span>
<span id="cb521-4"><a href="#cb521-4" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"nit"</span>,</span>
<span id="cb521-5"><a href="#cb521-5" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"trochObs"</span>,</span>
<span id="cb521-6"><a href="#cb521-6" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"plumb"</span>]</span>
<span id="cb521-7"><a href="#cb521-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-8"><a href="#cb521-8" aria-hidden="true" tabindex="-1"></a><span class="co"># get boundaries of gw3 LHBR:</span></span>
<span id="cb521-9"><a href="#cb521-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-10"><a href="#cb521-10" aria-hidden="true" tabindex="-1"></a>chr <span class="op">=</span> <span class="st">"gw3"</span></span>
<span id="cb521-11"><a href="#cb521-11" aria-hidden="true" tabindex="-1"></a>positionMin_chr3_LHBR, positionMax_chr3_LHBR, regionText, </span>
<span id="cb521-12"><a href="#cb521-12" aria-hidden="true" tabindex="-1"></a>    windowedIndHetStanRegion, meanAcrossRegionIndHetStan,</span>
<span id="cb521-13"><a href="#cb521-13" aria-hidden="true" tabindex="-1"></a>    genos_highHetVarRegion, pos_highHetVarRegion, regionInfo <span class="op">=</span> </span>
<span id="cb521-14"><a href="#cb521-14" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">getWindowedIndHetStanRegion</span>(genosOnly_included, </span>
<span id="cb521-15"><a href="#cb521-15" aria-hidden="true" tabindex="-1"></a>                                                    pos_SNP_filtered, </span>
<span id="cb521-16"><a href="#cb521-16" aria-hidden="true" tabindex="-1"></a>                                                    highHetVarRegions, chr;</span>
<span id="cb521-17"><a href="#cb521-17" aria-hidden="true" tabindex="-1"></a>                                                    windowSize <span class="op">=</span> <span class="fl">500</span>)</span>
<span id="cb521-18"><a href="#cb521-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-19"><a href="#cb521-19" aria-hidden="true" tabindex="-1"></a><span class="co"># select the loci within the gw3 LHBR:</span></span>
<span id="cb521-20"><a href="#cb521-20" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (positionMin_chr3_LHBR <span class="op">.&lt;=</span> pos_chr3.position <span class="op">.&lt;=</span> positionMax_chr3_LHBR) </span>
<span id="cb521-21"><a href="#cb521-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-22"><a href="#cb521-22" aria-hidden="true" tabindex="-1"></a>geno_chr3_included_LHBR <span class="op">=</span> geno_chr3_included[<span class="op">:</span>, selection]</span>
<span id="cb521-23"><a href="#cb521-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-24"><a href="#cb521-24" aria-hidden="true" tabindex="-1"></a>pos_chr3_LHBR <span class="op">=</span> pos_chr3[selection, <span class="op">:</span>]</span>
<span id="cb521-25"><a href="#cb521-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-26"><a href="#cb521-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb521-27"><a href="#cb521-27" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(geno_chr3_included_LHBR, indClusterMembership_gw3, clusterNames_gw3)</span>
<span id="cb521-28"><a href="#cb521-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-29"><a href="#cb521-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb521-30"><a href="#cb521-30" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb521-31"><a href="#cb521-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-32"><a href="#cb521-32" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb521-33"><a href="#cb521-33" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames_gw3)</span>
<span id="cb521-34"><a href="#cb521-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-35"><a href="#cb521-35" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames_gw3; among<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb521-36"><a href="#cb521-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-37"><a href="#cb521-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for chr 3 LHBR:</span></span>
<span id="cb521-38"><a href="#cb521-38" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames_gw3, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb521-39"><a href="#cb521-39" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb521-40"><a href="#cb521-40" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi          </span></span>
<span id="cb521-41"><a href="#cb521-41" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64     </span></span>
<span id="cb521-42"><a href="#cb521-42" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼───────────────────────</span></span>
<span id="cb521-43"><a href="#cb521-43" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.0012486</span></span>
<span id="cb521-44"><a href="#cb521-44" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.000697639</span></span>
<span id="cb521-45"><a href="#cb521-45" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ trochObs  0.00136251</span></span>
<span id="cb521-46"><a href="#cb521-46" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ plumb     0.00111764 =#</span></span>
<span id="cb521-47"><a href="#cb521-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-48"><a href="#cb521-48" aria-hidden="true" tabindex="-1"></a><span class="co"># average pi (for chr 3 LHBR) among three major groups:</span></span>
<span id="cb521-49"><a href="#cb521-49" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0012486</span> <span class="op">+</span> <span class="fl">0.00136251</span> <span class="op">+</span> <span class="fl">0.00111764</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb521-50"><a href="#cb521-50" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.0012429166</span></span>
<span id="cb521-51"><a href="#cb521-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-52"><a href="#cb521-52" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb521-53"><a href="#cb521-53" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb521-54"><a href="#cb521-54" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy        </span></span>
<span id="cb521-55"><a href="#cb521-55" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64    </span></span>
<span id="cb521-56"><a href="#cb521-56" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────────</span></span>
<span id="cb521-57"><a href="#cb521-57" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.0026694</span></span>
<span id="cb521-58"><a href="#cb521-58" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_trochObs  0.00354486</span></span>
<span id="cb521-59"><a href="#cb521-59" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_plumb     0.00398499</span></span>
<span id="cb521-60"><a href="#cb521-60" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_trochObs     0.00335481</span></span>
<span id="cb521-61"><a href="#cb521-61" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_plumb        0.00400019</span></span>
<span id="cb521-62"><a href="#cb521-62" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ trochObs_plumb   0.00328016 =#</span></span>
<span id="cb521-63"><a href="#cb521-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-64"><a href="#cb521-64" aria-hidden="true" tabindex="-1"></a><span class="co"># average Dxy (for chr 3 LHBR) among three major groups:</span></span>
<span id="cb521-65"><a href="#cb521-65" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.00354486</span> <span class="op">+</span> <span class="fl">0.00398499</span> <span class="op">+</span> <span class="fl">0.00328016</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb521-66"><a href="#cb521-66" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.0036033366666666663</span></span>
<span id="cb521-67"><a href="#cb521-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb521-68"><a href="#cb521-68" aria-hidden="true" tabindex="-1"></a><span class="co"># Drawing phylogeny (in Illustrator) based on above, between three major groups.</span></span>
<span id="cb521-69"><a href="#cb521-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Lowest Dxy is between trochObs and plumb (0.00328016).</span></span>
<span id="cb521-70"><a href="#cb521-70" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculation of deepest split, an average of virLud diff with (trochObs, plumb):</span></span>
<span id="cb521-71"><a href="#cb521-71" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.00354486</span> <span class="op">+</span> <span class="fl">0.00398499</span>) <span class="op">/</span> <span class="fl">2</span></span>
<span id="cb521-72"><a href="#cb521-72" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.003764925</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>More than 1 region on that scaffold. Using just the longest one.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div><div style="float: left;"><span>2×3 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;">
<table class="data-frame caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;">Row</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionChrom</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionStart</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">regionEnd</th>
</tr>
<tr class="subheader headerLastRow even">
<th class="rowNumber" data-quarto-table-cell-role="th" style="text-align: right; font-weight: bold;"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="String">String</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
<th style="text-align: left;" data-quarto-table-cell-role="th" title="Int64">Int64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="rowNumber" style="text-align: right; font-weight: bold;">1</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">101192949</td>
<td style="text-align: right;">103495514</td>
</tr>
<tr class="even">
<td class="rowNumber" style="text-align: right; font-weight: bold;">2</td>
<td style="text-align: left;">gw3</td>
<td style="text-align: right;">104554714</td>
<td style="text-align: right;">108279595</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="194">
<pre><code>0.0037649249999999997</code></pre>
</div>
</div>
</section>
<section id="calculate-pi-and-dxy-outside-of-the-chr-3-lhbr-grouped-by-the-lhbr-homozygous-groups" class="level3">
<h3 class="anchored" data-anchor-id="calculate-pi-and-dxy-outside-of-the-chr-3-lhbr-grouped-by-the-lhbr-homozygous-groups">Calculate pi and Dxy outside of the chr 3 LHBR (grouped by the LHBR homozygous groups)</h3>
<div id="7ed429f4" class="cell" data-execution_count="194">
<div class="sourceCode cell-code" id="cb524"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb524-1"><a href="#cb524-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the loci outside of the gw3 LHBR:</span></span>
<span id="cb524-2"><a href="#cb524-2" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> .!(positionMin_chr3_LHBR <span class="op">.&lt;=</span> pos_chr3.position <span class="op">.&lt;=</span> positionMax_chr3_LHBR) </span>
<span id="cb524-3"><a href="#cb524-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-4"><a href="#cb524-4" aria-hidden="true" tabindex="-1"></a>geno_chr3_included_nonLHBR <span class="op">=</span> geno_chr3_included[<span class="op">:</span>, selection]</span>
<span id="cb524-5"><a href="#cb524-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-6"><a href="#cb524-6" aria-hidden="true" tabindex="-1"></a>pos_chr3_nonLHBR <span class="op">=</span> pos_chr3[selection, <span class="op">:</span>]</span>
<span id="cb524-7"><a href="#cb524-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-8"><a href="#cb524-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes</span></span>
<span id="cb524-9"><a href="#cb524-9" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(geno_chr3_included_nonLHBR, indClusterMembership_gw3, clusterNames_gw3)</span>
<span id="cb524-10"><a href="#cb524-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-11"><a href="#cb524-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate per-site pi (within-group nucleotide distance)</span></span>
<span id="cb524-12"><a href="#cb524-12" aria-hidden="true" tabindex="-1"></a>sitePi <span class="op">=</span> <span class="fu">getSitePi</span>(freqs, sampleSizes)</span>
<span id="cb524-13"><a href="#cb524-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-14"><a href="#cb524-14" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate pairwise Dxy per site, using data in "freqs" and groups in "groups"</span></span>
<span id="cb524-15"><a href="#cb524-15" aria-hidden="true" tabindex="-1"></a>Dxy, pairwiseDxyClusterNames <span class="op">=</span> <span class="fu">getDxy</span>(freqs, clusterNames_gw3)</span>
<span id="cb524-16"><a href="#cb524-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-17"><a href="#cb524-17" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseFstClusterNames <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, clusterNames_gw3; among<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb524-18"><a href="#cb524-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-19"><a href="#cb524-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get averages of pi and Dxy for whole region:</span></span>
<span id="cb524-20"><a href="#cb524-20" aria-hidden="true" tabindex="-1"></a>regionPiTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster <span class="op">=</span> clusterNames_gw3, <span class="cn">pi</span> <span class="op">=</span> <span class="fu">getRegionPi</span>(sitePi))</span>
<span id="cb524-21"><a href="#cb524-21" aria-hidden="true" tabindex="-1"></a><span class="co">#= 4×2 DataFrame</span></span>
<span id="cb524-22"><a href="#cb524-22" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster   pi         </span></span>
<span id="cb524-23"><a href="#cb524-23" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String    Float64    </span></span>
<span id="cb524-24"><a href="#cb524-24" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼──────────────────────</span></span>
<span id="cb524-25"><a href="#cb524-25" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud    0.00456714</span></span>
<span id="cb524-26"><a href="#cb524-26" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ nit       0.00161186</span></span>
<span id="cb524-27"><a href="#cb524-27" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ trochObs  0.00568622</span></span>
<span id="cb524-28"><a href="#cb524-28" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ plumb     0.00554501 =#</span></span>
<span id="cb524-29"><a href="#cb524-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-30"><a href="#cb524-30" aria-hidden="true" tabindex="-1"></a><span class="co"># average pi (for chr 3 NOT in LHBR) among three major groups:</span></span>
<span id="cb524-31"><a href="#cb524-31" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.00456714</span> <span class="op">+</span> <span class="fl">0.00568622</span> <span class="op">+</span> <span class="fl">0.00554501</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb524-32"><a href="#cb524-32" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.005266123</span></span>
<span id="cb524-33"><a href="#cb524-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-34"><a href="#cb524-34" aria-hidden="true" tabindex="-1"></a><span class="co">#ratio of average pi outside to average pi within chr 3 LHBR:</span></span>
<span id="cb524-35"><a href="#cb524-35" aria-hidden="true" tabindex="-1"></a><span class="fl">0.005266123</span> <span class="op">/</span> <span class="fl">0.0012429166</span></span>
<span id="cb524-36"><a href="#cb524-36" aria-hidden="true" tabindex="-1"></a><span class="co"># 4.236907769998406</span></span>
<span id="cb524-37"><a href="#cb524-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-38"><a href="#cb524-38" aria-hidden="true" tabindex="-1"></a><span class="co"># percent lower that average pi is within compared to outside LHBR:</span></span>
<span id="cb524-39"><a href="#cb524-39" aria-hidden="true" tabindex="-1"></a><span class="fl">100</span> <span class="op">*</span> (<span class="fl">4.2369078</span> <span class="op">-</span> <span class="fl">1</span>) <span class="op">/</span> <span class="fl">4.2369078</span></span>
<span id="cb524-40"><a href="#cb524-40" aria-hidden="true" tabindex="-1"></a><span class="co"># 76.39788149272448</span></span>
<span id="cb524-41"><a href="#cb524-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-42"><a href="#cb524-42" aria-hidden="true" tabindex="-1"></a><span class="co">#ratio of pi outside to within LHBR:</span></span>
<span id="cb524-43"><a href="#cb524-43" aria-hidden="true" tabindex="-1"></a><span class="fl">0.00456714</span> <span class="op">/</span> <span class="fl">0.0012486</span></span>
<span id="cb524-44"><a href="#cb524-44" aria-hidden="true" tabindex="-1"></a><span class="co"># 3.6578087457952906</span></span>
<span id="cb524-45"><a href="#cb524-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-46"><a href="#cb524-46" aria-hidden="true" tabindex="-1"></a><span class="fl">0.00568622</span> <span class="op">/</span> <span class="fl">0.00136251</span></span>
<span id="cb524-47"><a href="#cb524-47" aria-hidden="true" tabindex="-1"></a><span class="co"># 4.173341847032315</span></span>
<span id="cb524-48"><a href="#cb524-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-49"><a href="#cb524-49" aria-hidden="true" tabindex="-1"></a><span class="fl">0.00554501</span> <span class="op">/</span> <span class="fl">0.00111764</span></span>
<span id="cb524-50"><a href="#cb524-50" aria-hidden="true" tabindex="-1"></a><span class="co"># 4.961356071722558</span></span>
<span id="cb524-51"><a href="#cb524-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-52"><a href="#cb524-52" aria-hidden="true" tabindex="-1"></a>((<span class="fl">0.00456714</span> <span class="op">/</span> <span class="fl">0.0012486</span>) <span class="op">+</span> (<span class="fl">0.00568622</span> <span class="op">/</span> <span class="fl">0.00136251</span>) <span class="op">+</span> (<span class="fl">0.00554501</span> <span class="op">/</span> <span class="fl">0.00111764</span>)) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb524-53"><a href="#cb524-53" aria-hidden="true" tabindex="-1"></a><span class="co"># 4.264168888183388</span></span>
<span id="cb524-54"><a href="#cb524-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-55"><a href="#cb524-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-56"><a href="#cb524-56" aria-hidden="true" tabindex="-1"></a>regionDxyTable <span class="op">=</span> <span class="fu">DataFrame</span>(cluster_pair <span class="op">=</span> pairwiseDxyClusterNames, Dxy <span class="op">=</span> <span class="fu">getRegionDxy</span>(Dxy))</span>
<span id="cb524-57"><a href="#cb524-57" aria-hidden="true" tabindex="-1"></a><span class="co">#= 6×2 DataFrame</span></span>
<span id="cb524-58"><a href="#cb524-58" aria-hidden="true" tabindex="-1"></a><span class="co"> Row │ cluster_pair     Dxy        </span></span>
<span id="cb524-59"><a href="#cb524-59" aria-hidden="true" tabindex="-1"></a><span class="co">     │ String           Float64    </span></span>
<span id="cb524-60"><a href="#cb524-60" aria-hidden="true" tabindex="-1"></a><span class="co">─────┼─────────────────────────────</span></span>
<span id="cb524-61"><a href="#cb524-61" aria-hidden="true" tabindex="-1"></a><span class="co">   1 │ virLud_nit       0.00439336</span></span>
<span id="cb524-62"><a href="#cb524-62" aria-hidden="true" tabindex="-1"></a><span class="co">   2 │ virLud_trochObs  0.0059857</span></span>
<span id="cb524-63"><a href="#cb524-63" aria-hidden="true" tabindex="-1"></a><span class="co">   3 │ virLud_plumb     0.00660034</span></span>
<span id="cb524-64"><a href="#cb524-64" aria-hidden="true" tabindex="-1"></a><span class="co">   4 │ nit_trochObs     0.00548855</span></span>
<span id="cb524-65"><a href="#cb524-65" aria-hidden="true" tabindex="-1"></a><span class="co">   5 │ nit_plumb        0.00608182</span></span>
<span id="cb524-66"><a href="#cb524-66" aria-hidden="true" tabindex="-1"></a><span class="co">   6 │ trochObs_plumb   0.00652593 =#</span></span>
<span id="cb524-67"><a href="#cb524-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-68"><a href="#cb524-68" aria-hidden="true" tabindex="-1"></a><span class="co"># average Dxy (for OUTSIDE of chr 3 LHBR) among three major groups:</span></span>
<span id="cb524-69"><a href="#cb524-69" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.0059857</span> <span class="op">+</span> <span class="fl">0.00660034</span> <span class="op">+</span> <span class="fl">0.00652593</span>) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb524-70"><a href="#cb524-70" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.006370656666666666</span></span>
<span id="cb524-71"><a href="#cb524-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-72"><a href="#cb524-72" aria-hidden="true" tabindex="-1"></a><span class="co">#ratio of average Dxy outside to average Dxy within LHBR among 3 major groups:</span></span>
<span id="cb524-73"><a href="#cb524-73" aria-hidden="true" tabindex="-1"></a><span class="fl">0.006370656666666666</span> <span class="op">/</span> <span class="fl">0.0036033366666666663</span></span>
<span id="cb524-74"><a href="#cb524-74" aria-hidden="true" tabindex="-1"></a><span class="co"># 1.76798818872508</span></span>
<span id="cb524-75"><a href="#cb524-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-76"><a href="#cb524-76" aria-hidden="true" tabindex="-1"></a><span class="co"># percent lower that average Dxy is within compared to outside LHBR:</span></span>
<span id="cb524-77"><a href="#cb524-77" aria-hidden="true" tabindex="-1"></a><span class="fl">100</span> <span class="op">*</span> (<span class="fl">1.76798818872508</span> <span class="op">-</span> <span class="fl">1</span>) <span class="op">/</span> <span class="fl">1.76798818872508</span></span>
<span id="cb524-78"><a href="#cb524-78" aria-hidden="true" tabindex="-1"></a><span class="co"># 43.43853616346196</span></span>
<span id="cb524-79"><a href="#cb524-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-80"><a href="#cb524-80" aria-hidden="true" tabindex="-1"></a><span class="co"># Drawing phylogeny (in Illustrator) based on above, between three major groups.</span></span>
<span id="cb524-81"><a href="#cb524-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Lowest Dxy is between virLud and trochObs (0.0059857).</span></span>
<span id="cb524-82"><a href="#cb524-82" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculation of deepest split, an average of plumb diff with (virLud, trochObs):</span></span>
<span id="cb524-83"><a href="#cb524-83" aria-hidden="true" tabindex="-1"></a>(<span class="fl">0.00660034</span> <span class="op">+</span> <span class="fl">0.00652593</span>) <span class="op">/</span> <span class="fl">2</span></span>
<span id="cb524-84"><a href="#cb524-84" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.006563135</span></span>
<span id="cb524-85"><a href="#cb524-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb524-86"><a href="#cb524-86" aria-hidden="true" tabindex="-1"></a>((<span class="fl">0.00354486</span> <span class="op">/</span> <span class="fl">0.0059857</span>) <span class="op">+</span> (<span class="fl">0.00328016</span> <span class="op">/</span> <span class="fl">0.00652593</span>) <span class="op">+</span> (<span class="fl">0.00398499</span> <span class="op">/</span> <span class="fl">0.00660034</span>)) <span class="op">/</span> <span class="fl">3</span></span>
<span id="cb524-87"><a href="#cb524-87" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.5662038652462132</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="195">
<pre><code>0.5662038652462132</code></pre>
</div>
</div>
<p>Really neat results above. More diversity at chr 3 LHBR than at 4A.</p>
</section>
<section id="copying-from-the-main-page-and-modifying-the-bit-below-to-make-supplemental-gbi-plots" class="level3">
<h3 class="anchored" data-anchor-id="copying-from-the-main-page-and-modifying-the-bit-below-to-make-supplemental-gbi-plots">Copying (from the main page) and modifying the bit below to make Supplemental GBI plots</h3>
<section id="make-list-of-scaffolds-to-plot-1" class="level5">
<h5 class="anchored" data-anchor-id="make-list-of-scaffolds-to-plot-1">Make list of scaffolds to plot:</h5>
<div id="85a49d53" class="cell" data-execution_count="195">
<div class="sourceCode cell-code" id="cb526"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb526-1"><a href="#cb526-1" aria-hidden="true" tabindex="-1"></a>scaffolds_to_plot <span class="op">=</span> <span class="st">"gw"</span> <span class="op">.*</span> <span class="fu">string</span>.(<span class="fu">vcat</span>(<span class="fl">28</span><span class="op">:-</span><span class="fl">1</span><span class="op">:</span><span class="fl">17</span>, <span class="fl">15</span><span class="op">:-</span><span class="fl">1</span><span class="op">:</span><span class="fl">1</span>))</span>
<span id="cb526-2"><a href="#cb526-2" aria-hidden="true" tabindex="-1"></a><span class="fu">push!</span>(scaffolds_to_plot, <span class="st">"gw1A"</span>, <span class="st">"gw4A"</span>)  <span class="co"># add two other scaffolds</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="196">
<pre><code>29-element Vector{String}:
 "gw28"
 "gw27"
 "gw26"
 "gw25"
 "gw24"
 "gw23"
 "gw22"
 "gw21"
 "gw20"
 "gw19"
 "gw18"
 "gw17"
 "gw15"
 ⋮
 "gw10"
 "gw9"
 "gw8"
 "gw7"
 "gw6"
 "gw5"
 "gw4"
 "gw3"
 "gw2"
 "gw1"
 "gw1A"
 "gw4A"</code></pre>
</div>
</div>
</section>
<section id="do-other-setup" class="level5">
<h5 class="anchored" data-anchor-id="do-other-setup">Do other setup:</h5>
<div id="00d6aacb" class="cell" data-execution_count="196">
<div class="sourceCode cell-code" id="cb528"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb528-1"><a href="#cb528-1" aria-hidden="true" tabindex="-1"></a>groups_to_plot_all <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"vir_S"</span>,<span class="st">"nit"</span>, <span class="st">"lud_PK"</span>, <span class="st">"lud_KS"</span>, <span class="st">"lud_central"</span>, <span class="st">"lud_Sath"</span>, <span class="st">"lud_ML"</span>,<span class="st">"troch_west"</span>,<span class="st">"troch_LN"</span>,<span class="st">"troch_EM"</span>,<span class="st">"obs"</span>,<span class="st">"plumb_BJ"</span>,<span class="st">"plumb"</span>,<span class="st">"plumb_vir"</span>]</span>
<span id="cb528-2"><a href="#cb528-2" aria-hidden="true" tabindex="-1"></a>group_colors_all <span class="op">=</span> [<span class="st">"blue"</span>,<span class="st">"turquoise1"</span>,<span class="st">"grey"</span>,<span class="st">"seagreen4"</span>,<span class="st">"seagreen3"</span>,<span class="st">"seagreen2"</span>,<span class="st">"olivedrab3"</span>,<span class="st">"olivedrab2"</span>,<span class="st">"olivedrab1"</span>,<span class="st">"yellow"</span>,<span class="st">"gold"</span>,<span class="st">"orange"</span>,<span class="st">"pink"</span>,<span class="st">"red"</span>,<span class="st">"purple"</span>];</span>
<span id="cb528-3"><a href="#cb528-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb528-4"><a href="#cb528-4" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> [<span class="st">"vir"</span>,<span class="st">"troch_LN"</span>,<span class="st">"plumb"</span>] <span class="co"># for purpose of calculating pairwise Fst and Fst_group (to determine SNPs)   </span></span>
<span id="cb528-5"><a href="#cb528-5" aria-hidden="true" tabindex="-1"></a>plotGroups <span class="op">=</span> groups_to_plot_all </span>
<span id="cb528-6"><a href="#cb528-6" aria-hidden="true" tabindex="-1"></a>plotGroupColors <span class="op">=</span> group_colors_all</span>
<span id="cb528-7"><a href="#cb528-7" aria-hidden="true" tabindex="-1"></a>group1 <span class="op">=</span> <span class="st">"vir"</span>   <span class="co"># these groups will determine the color used in the graph</span></span>
<span id="cb528-8"><a href="#cb528-8" aria-hidden="true" tabindex="-1"></a>group2 <span class="op">=</span> <span class="st">"plumb"</span></span>
<span id="cb528-9"><a href="#cb528-9" aria-hidden="true" tabindex="-1"></a>groupsToCompare <span class="op">=</span> [<span class="st">"vir_plumb"</span>, <span class="st">"vir_troch_LN"</span>, <span class="st">"troch_LN_plumb"</span>]     <span class="co"># "Fst_among" #"vir_plumb" </span></span>
<span id="cb528-10"><a href="#cb528-10" aria-hidden="true" tabindex="-1"></a>missingFractionAllowed <span class="op">=</span> <span class="fl">0.2</span>  <span class="co"># only show SNPs with less than this fraction of missing data among individuals</span></span>
<span id="cb528-11"><a href="#cb528-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb528-12"><a href="#cb528-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate allele freqs and sample sizes (use column Fst_group)</span></span>
<span id="cb528-13"><a href="#cb528-13" aria-hidden="true" tabindex="-1"></a>freqs, sampleSizes <span class="op">=</span> <span class="fu">getFreqsAndSampleSizes</span>(genosOnly_with_missing, ind_with_metadata_indFiltered.Fst_group, groups)</span>
<span id="cb528-14"><a href="#cb528-14" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated population allele frequencies and sample sizes"</span>)</span>
<span id="cb528-15"><a href="#cb528-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb528-16"><a href="#cb528-16" aria-hidden="true" tabindex="-1"></a>Fst, FstNumerator, FstDenominator, pairwiseNamesFst <span class="op">=</span> <span class="fu">getFst</span>(freqs, sampleSizes, groups; among<span class="op">=</span><span class="cn">true</span>)  <span class="co"># set among to FALSE if no among Fst wanted (some things won't work without it) </span></span>
<span id="cb528-17"><a href="#cb528-17" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Calculated Fst values"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculated population allele frequencies and sample sizes
Calculated Fst values</code></pre>
</div>
</div>
</section>
<section id="loop-through-scaffolds-and-make-plots-and-adding-a-line-for-the-lhbrs" class="level5">
<h5 class="anchored" data-anchor-id="loop-through-scaffolds-and-make-plots-and-adding-a-line-for-the-lhbrs">Loop through scaffolds and make plots, and adding a line for the LHBRs:</h5>
<p>Making inactive because plots were already made:</p>
<div class="sourceCode" id="cb530"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb530-1"><a href="#cb530-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for autosomes, Fst &gt; 0.8</span></span>
<span id="cb530-2"><a href="#cb530-2" aria-hidden="true" tabindex="-1"></a>Fst_cutoff <span class="op">=</span> <span class="fl">0.8</span></span>
<span id="cb530-3"><a href="#cb530-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(scaffolds_to_plot)</span>
<span id="cb530-4"><a href="#cb530-4" aria-hidden="true" tabindex="-1"></a>    chr <span class="op">=</span> scaffolds_to_plot[i]</span>
<span id="cb530-5"><a href="#cb530-5" aria-hidden="true" tabindex="-1"></a>    regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb530-6"><a href="#cb530-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get info for lines for LHBRs</span></span>
<span id="cb530-7"><a href="#cb530-7" aria-hidden="true" tabindex="-1"></a>    highHetVarRegions_thisScaffold <span class="op">=</span> highHetVarRegions[highHetVarRegions.regionChrom <span class="op">.==</span> chr, <span class="op">:</span>]</span>
<span id="cb530-8"><a href="#cb530-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb530-9"><a href="#cb530-9" aria-hidden="true" tabindex="-1"></a>    plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividualWithFst</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb530-10"><a href="#cb530-10" aria-hidden="true" tabindex="-1"></a>        missingFractionAllowed, regionInfo,</span>
<span id="cb530-11"><a href="#cb530-11" aria-hidden="true" tabindex="-1"></a>        pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb530-12"><a href="#cb530-12" aria-hidden="true" tabindex="-1"></a>        genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb530-13"><a href="#cb530-13" aria-hidden="true" tabindex="-1"></a>        plotGroups, plotGroupColors;</span>
<span id="cb530-14"><a href="#cb530-14" aria-hidden="true" tabindex="-1"></a>        indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>), plotTitle <span class="op">=</span> <span class="st">""</span>,</span>
<span id="cb530-15"><a href="#cb530-15" aria-hidden="true" tabindex="-1"></a>        highlightRegionStarts <span class="op">=</span> highHetVarRegions_thisScaffold.regionStart,</span>
<span id="cb530-16"><a href="#cb530-16" aria-hidden="true" tabindex="-1"></a>        highlightRegionEnds <span class="op">=</span> highHetVarRegions_thisScaffold.regionEnd,</span>
<span id="cb530-17"><a href="#cb530-17" aria-hidden="true" tabindex="-1"></a>        highlightRegionColor <span class="op">=</span> <span class="st">"magenta"</span>)</span>
<span id="cb530-18"><a href="#cb530-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb530-19"><a href="#cb530-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Completed the figure for "</span>, chr, <span class="st">"."</span>)</span>
<span id="cb530-20"><a href="#cb530-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="cn">true</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb530-21"><a href="#cb530-21" aria-hidden="true" tabindex="-1"></a>        filename <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Figure_"</span>, chr, <span class="st">"_Fst3groupsGBIallInds_Fst"</span>,Fst_cutoff,<span class="st">"_fromJulia.png"</span>)</span>
<span id="cb530-22"><a href="#cb530-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">save</span>(filename, plotInfo[<span class="fl">1</span>], px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb530-23"><a href="#cb530-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"Saved "</span>, filename)</span>
<span id="cb530-24"><a href="#cb530-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span> </span>
<span id="cb530-25"><a href="#cb530-25" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="now-for-z-chromosome-with-fst-0.9" class="level5">
<h5 class="anchored" data-anchor-id="now-for-z-chromosome-with-fst-0.9">Now for Z chromosome, with Fst &gt; 0.9</h5>
<div id="e071583a" class="cell" data-execution_count="197">
<div class="sourceCode cell-code" id="cb531"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb531-1"><a href="#cb531-1" aria-hidden="true" tabindex="-1"></a>scaffolds_to_plot <span class="op">=</span> [<span class="st">"gwZ"</span>]; Fst_cutoff <span class="op">=</span> <span class="fl">0.9</span></span>
<span id="cb531-2"><a href="#cb531-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(scaffolds_to_plot)</span>
<span id="cb531-3"><a href="#cb531-3" aria-hidden="true" tabindex="-1"></a>    chr <span class="op">=</span> scaffolds_to_plot[i]</span>
<span id="cb531-4"><a href="#cb531-4" aria-hidden="true" tabindex="-1"></a>    regionInfo <span class="op">=</span> <span class="fu">chooseChrRegion</span>(pos_SNP_filtered, chr; positionMin<span class="op">=</span><span class="fl">1</span>, positionMax<span class="op">=</span>scaffold_lengths[chr]) <span class="co"># this gets the maximum position for the chromosome</span></span>
<span id="cb531-5"><a href="#cb531-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get info for lines for LHBRs</span></span>
<span id="cb531-6"><a href="#cb531-6" aria-hidden="true" tabindex="-1"></a>    highHetVarRegions_thisScaffold <span class="op">=</span> highHetVarRegions[highHetVarRegions.regionChrom <span class="op">.==</span> chr, <span class="op">:</span>]</span>
<span id="cb531-7"><a href="#cb531-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb531-8"><a href="#cb531-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#= ##### for testing function</span></span>
<span id="cb531-9"><a href="#cb531-9" aria-hidden="true" tabindex="-1"></a><span class="co">    pos=pos_SNP_filtered</span></span>
<span id="cb531-10"><a href="#cb531-10" aria-hidden="true" tabindex="-1"></a><span class="co">    genoData=genosOnly_with_missing</span></span>
<span id="cb531-11"><a href="#cb531-11" aria-hidden="true" tabindex="-1"></a><span class="co">    indMetadata= ind_with_metadata_indFiltered</span></span>
<span id="cb531-12"><a href="#cb531-12" aria-hidden="true" tabindex="-1"></a><span class="co">    colorAllelesByGroup = true</span></span>
<span id="cb531-13"><a href="#cb531-13" aria-hidden="true" tabindex="-1"></a><span class="co">    group1 = plotGroups[1]</span></span>
<span id="cb531-14"><a href="#cb531-14" aria-hidden="true" tabindex="-1"></a><span class="co">    indFontSize=4</span></span>
<span id="cb531-15"><a href="#cb531-15" aria-hidden="true" tabindex="-1"></a><span class="co">    figureSize=(1200,1600)</span></span>
<span id="cb531-16"><a href="#cb531-16" aria-hidden="true" tabindex="-1"></a><span class="co">    show_SNP_density = true </span></span>
<span id="cb531-17"><a href="#cb531-17" aria-hidden="true" tabindex="-1"></a><span class="co">    densityPlotColor = :steelblue1</span></span>
<span id="cb531-18"><a href="#cb531-18" aria-hidden="true" tabindex="-1"></a><span class="co">    plotTitle = ""</span></span>
<span id="cb531-19"><a href="#cb531-19" aria-hidden="true" tabindex="-1"></a><span class="co">    highlightRegionStarts = highHetVarRegions_thisScaffold.regionStart</span></span>
<span id="cb531-20"><a href="#cb531-20" aria-hidden="true" tabindex="-1"></a><span class="co">    highlightRegionEnds = highHetVarRegions_thisScaffold.regionEnd</span></span>
<span id="cb531-21"><a href="#cb531-21" aria-hidden="true" tabindex="-1"></a><span class="co">    ############ =#</span></span>
<span id="cb531-22"><a href="#cb531-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb531-23"><a href="#cb531-23" aria-hidden="true" tabindex="-1"></a>    plotInfo <span class="op">=</span> <span class="fu">plotGenotypeByIndividualWithFst</span>(groupsToCompare, Fst_cutoff, </span>
<span id="cb531-24"><a href="#cb531-24" aria-hidden="true" tabindex="-1"></a>        missingFractionAllowed, regionInfo,</span>
<span id="cb531-25"><a href="#cb531-25" aria-hidden="true" tabindex="-1"></a>        pos_SNP_filtered, Fst, pairwiseNamesFst, </span>
<span id="cb531-26"><a href="#cb531-26" aria-hidden="true" tabindex="-1"></a>        genosOnly_with_missing, ind_with_metadata_indFiltered, freqs, </span>
<span id="cb531-27"><a href="#cb531-27" aria-hidden="true" tabindex="-1"></a>        plotGroups, plotGroupColors;</span>
<span id="cb531-28"><a href="#cb531-28" aria-hidden="true" tabindex="-1"></a>        indFontSize<span class="op">=</span><span class="fl">4</span>, figureSize<span class="op">=</span>(<span class="fl">1200</span>,<span class="fl">1600</span>), plotTitle <span class="op">=</span> <span class="st">""</span>,</span>
<span id="cb531-29"><a href="#cb531-29" aria-hidden="true" tabindex="-1"></a>        highlightRegionStarts <span class="op">=</span> highHetVarRegions_thisScaffold.regionStart,</span>
<span id="cb531-30"><a href="#cb531-30" aria-hidden="true" tabindex="-1"></a>        highlightRegionEnds <span class="op">=</span> highHetVarRegions_thisScaffold.regionEnd,</span>
<span id="cb531-31"><a href="#cb531-31" aria-hidden="true" tabindex="-1"></a>        highlightRegionColor <span class="op">=</span> <span class="st">"magenta"</span>)</span>
<span id="cb531-32"><a href="#cb531-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb531-33"><a href="#cb531-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">println</span>(<span class="st">"Completed the figure for "</span>, chr, <span class="st">"."</span>)</span>
<span id="cb531-34"><a href="#cb531-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="cn">true</span>  <span class="co"># set to true to save plot</span></span>
<span id="cb531-35"><a href="#cb531-35" aria-hidden="true" tabindex="-1"></a>        filename <span class="op">=</span> <span class="fu">string</span>(<span class="st">"Figure_"</span>, chr, <span class="st">"_Fst3groupsGBIallInds_fromJulia.png"</span>)</span>
<span id="cb531-36"><a href="#cb531-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">save</span>(filename, plotInfo[<span class="fl">1</span>], px_per_unit <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb531-37"><a href="#cb531-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(<span class="st">"Saved "</span>, filename)</span>
<span id="cb531-38"><a href="#cb531-38" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span> </span>
<span id="cb531-39"><a href="#cb531-39" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>┌ Warning: Found `resolution` in the theme when creating a `Scene`. The `resolution` keyword for `Scene`s and `Figure`s has been deprecated. Use `Figure(; size = ...` or `Scene(; size = ...)` instead, which better reflects that this is a unitless size and not a pixel resolution. The key could also come from `set_theme!` calls or related theming functions.
└ @ Makie ~/.julia/packages/Makie/pFPBw/src/scenes.jl:238</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GW_Heterozygosity_Variance_files/figure-html/cell-198-output-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed the figure for gwZ.
Saved Figure_gwZ_Fst3groupsGBIallInds_fromJulia.png</code></pre>
</div>
</div>
</section>
</section>
<section id="old-stuff-below-can-probably-delete-but-might-be-some-good-bits-to-keep" class="level3">
<h3 class="anchored" data-anchor-id="old-stuff-below-can-probably-delete-but-might-be-some-good-bits-to-keep">OLD STUFF BELOW (CAN PROBABLY DELETE, BUT MIGHT BE SOME GOOD BITS TO KEEP)</h3>
</section>
<section id="actually-new-plan-just-do-the-pca-on-all-individuals-then-show-a-graph-without-the-heterozygotes.-will-do-this-above." class="level3">
<h3 class="anchored" data-anchor-id="actually-new-plan-just-do-the-pca-on-all-individuals-then-show-a-graph-without-the-heterozygotes.-will-do-this-above.">Actually new plan: Just do the PCA on all individuals, then show a graph without the heterozygotes. Will do this above.</h3>
</section>
</section>
<section id="adding-28june2024-starting-with-pca-of-just-the-homozygotes-above-apply-those-axes-to-other-individuals-to-add-them-to-the-pca" class="level1">
<h1>Adding 28June2024: starting with PCA of just the homozygotes (above), apply those axes to other individuals to add them to the PCA:</h1>
<div class="sourceCode" id="cb534"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb534-1"><a href="#cb534-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb534-2"><a href="#cb534-2" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_included <span class="co"># this is the list of all individuals prior to choosing low IndHet</span></span>
<span id="cb534-3"><a href="#cb534-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb534-4"><a href="#cb534-4" aria-hidden="true" tabindex="-1"></a>genos_highHetVarRegion_imputed <span class="co"># these are the genotypes of all individuals (imputed)</span></span>
<span id="cb534-5"><a href="#cb534-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb534-6"><a href="#cb534-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb534-7"><a href="#cb534-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the key command, from another place in my code (to expand to multiple samples, matrix should have individuals as columns):</span></span>
<span id="cb534-8"><a href="#cb534-8" aria-hidden="true" tabindex="-1"></a>matrixForPCAplot <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Float32}</span>(<span class="fu">transpose</span>(genos_highHetVarRegion_imputed))</span>
<span id="cb534-9"><a href="#cb534-9" aria-hidden="true" tabindex="-1"></a>PCvalues <span class="op">=</span> <span class="fu">transpose</span>(<span class="fu">predict</span>(PCAmodel.model, matrixForPCAplot))</span>
<span id="cb534-10"><a href="#cb534-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb534-11"><a href="#cb534-11" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb534-12"><a href="#cb534-12" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb534-13"><a href="#cb534-13" aria-hidden="true" tabindex="-1"></a>    title <span class="op">=</span> <span class="st">"PC1 vs. PC2"</span>,</span>
<span id="cb534-14"><a href="#cb534-14" aria-hidden="true" tabindex="-1"></a>    xlabel <span class="op">=</span> <span class="st">"Region PC1"</span>, xlabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb534-15"><a href="#cb534-15" aria-hidden="true" tabindex="-1"></a>    ylabel <span class="op">=</span> <span class="st">"Region PC2"</span>, ylabelsize <span class="op">=</span> <span class="fl">24</span>,</span>
<span id="cb534-16"><a href="#cb534-16" aria-hidden="true" tabindex="-1"></a>    autolimitaspect <span class="op">=</span> <span class="fl">1</span>)</span>
<span id="cb534-17"><a href="#cb534-17" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb534-18"><a href="#cb534-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb534-19"><a href="#cb534-19" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> ind_with_metadata_included.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb534-20"><a href="#cb534-20" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PCvalues[selection, <span class="fl">1</span>], PCvalues[selection, <span class="fl">2</span>], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], <span class="fl">0.6</span>), markersize <span class="op">=</span> <span class="fl">14</span>, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, <span class="fl">0.7</span>))</span>
<span id="cb534-21"><a href="#cb534-21" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb534-22"><a href="#cb534-22" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="section-2" class="level10">
<p class="heading"></p>
</section>
</section>
<section id="below-what-i-tried-first-here-is-to-calculate-a-matrix-of-pairwise-genetic-distances-and-then-use-principal-coordinates-analysis-metric-multidimensional-scaling.-the-below-works.-then-i-realized-that-it-is-perhaps-better-to-stick-with-pca-like-in-the-rest-of-the-paper.-so-will-do-that-above." class="level1">
<h1>Below, what I tried first here is to calculate a matrix of pairwise genetic distances and then use Principal Coordinates Analysis (Metric Multidimensional Scaling). The below works. Then I realized that it is perhaps better to stick with PCA like in the rest of the paper. So will do that above.</h1>
<div class="sourceCode" id="cb535"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb535-1"><a href="#cb535-1" aria-hidden="true" tabindex="-1"></a>indGenDistances <span class="op">=</span> <span class="fu">getIndGenDistances</span>(genos_lowIndHetStan)</span>
<span id="cb535-2"><a href="#cb535-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-3"><a href="#cb535-3" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> <span class="fu">fit</span>(MDS, indGenDistances; distances<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb535-4"><a href="#cb535-4" aria-hidden="true" tabindex="-1"></a>pco_coords <span class="op">=</span> <span class="fu">predict</span>(mds)</span>
<span id="cb535-5"><a href="#cb535-5" aria-hidden="true" tabindex="-1"></a>PCO1 <span class="op">=</span> pco_coords[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb535-6"><a href="#cb535-6" aria-hidden="true" tabindex="-1"></a>PCO2 <span class="op">=</span> pco_coords[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb535-7"><a href="#cb535-7" aria-hidden="true" tabindex="-1"></a>PCO3 <span class="op">=</span> pco_coords[<span class="fl">3</span>,<span class="op">:</span>]</span>
<span id="cb535-8"><a href="#cb535-8" aria-hidden="true" tabindex="-1"></a>PCO4 <span class="op">=</span> pco_coords[<span class="fl">4</span>,<span class="op">:</span>]</span>
<span id="cb535-9"><a href="#cb535-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-10"><a href="#cb535-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-11"><a href="#cb535-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCO1, PCO2)</span>
<span id="cb535-12"><a href="#cb535-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-13"><a href="#cb535-13" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(PCO1, PCO3)</span></span>
<span id="cb535-14"><a href="#cb535-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-15"><a href="#cb535-15" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(PCO1, PCO4)</span></span>
<span id="cb535-16"><a href="#cb535-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-17"><a href="#cb535-17" aria-hidden="true" tabindex="-1"></a>pcoResults <span class="op">=</span> <span class="fu">DataFrame</span>(ind <span class="op">=</span> ind_with_metadata_lowIndHetStan.ind, Fst_group <span class="op">=</span> ind_with_metadata_lowIndHetStan.Fst_group, PCo1 <span class="op">=</span> PCO1, PCo2 <span class="op">=</span> PCO2)</span>
<span id="cb535-18"><a href="#cb535-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-19"><a href="#cb535-19" aria-hidden="true" tabindex="-1"></a><span class="co"># code taken from GenomicDiversity.plotPCA :</span></span>
<span id="cb535-20"><a href="#cb535-20" aria-hidden="true" tabindex="-1"></a>autolimitaspect_setting <span class="op">=</span> <span class="fl">1</span>; lineOpacity <span class="op">=</span> <span class="fl">0.8</span>; fillOpacity <span class="op">=</span> <span class="fl">0.2</span>;</span>
<span id="cb535-21"><a href="#cb535-21" aria-hidden="true" tabindex="-1"></a>    symbolSize <span class="op">=</span> <span class="fl">10</span>; plotTitle <span class="op">=</span> <span class="st">"PCo plot"</span>; showTitle <span class="op">=</span> <span class="cn">true</span>;</span>
<span id="cb535-22"><a href="#cb535-22" aria-hidden="true" tabindex="-1"></a>    xLabelText <span class="op">=</span> <span class="st">"Genomic PC1"</span>; yLabelText <span class="op">=</span> <span class="st">"Genomic PC2"</span>;</span>
<span id="cb535-23"><a href="#cb535-23" aria-hidden="true" tabindex="-1"></a>    labelSize <span class="op">=</span> <span class="fl">24</span></span>
<span id="cb535-24"><a href="#cb535-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-25"><a href="#cb535-25" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_forPlot <span class="op">=</span> ind_with_metadata_lowIndHetStan</span>
<span id="cb535-26"><a href="#cb535-26" aria-hidden="true" tabindex="-1"></a>PC1 <span class="op">=</span> PCO1; PC2 <span class="op">=</span> PCO2</span>
<span id="cb535-27"><a href="#cb535-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb535-28"><a href="#cb535-28" aria-hidden="true" tabindex="-1"></a>f <span class="op">=</span> CairoMakie.<span class="fu">Figure</span>()</span>
<span id="cb535-29"><a href="#cb535-29" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> <span class="fu">Axis</span>(f[<span class="fl">1</span>, <span class="fl">1</span>],</span>
<span id="cb535-30"><a href="#cb535-30" aria-hidden="true" tabindex="-1"></a>            title <span class="op">=</span> plotTitle,</span>
<span id="cb535-31"><a href="#cb535-31" aria-hidden="true" tabindex="-1"></a>            xlabel <span class="op">=</span> xLabelText, xlabelsize <span class="op">=</span> labelSize,</span>
<span id="cb535-32"><a href="#cb535-32" aria-hidden="true" tabindex="-1"></a>            ylabel <span class="op">=</span> yLabelText, ylabelsize <span class="op">=</span> labelSize,</span>
<span id="cb535-33"><a href="#cb535-33" aria-hidden="true" tabindex="-1"></a>            autolimitaspect <span class="op">=</span> autolimitaspect_setting)</span>
<span id="cb535-34"><a href="#cb535-34" aria-hidden="true" tabindex="-1"></a><span class="fu">hidedecorations!</span>(ax, label <span class="op">=</span> <span class="cn">false</span>, ticklabels <span class="op">=</span> <span class="cn">false</span>, ticks <span class="op">=</span> <span class="cn">false</span>) <span class="co"># hide background lattice</span></span>
<span id="cb535-35"><a href="#cb535-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fu">eachindex</span>(groups_to_plot_PCA) </span>
<span id="cb535-36"><a href="#cb535-36" aria-hidden="true" tabindex="-1"></a>    selection <span class="op">=</span> ind_with_metadata_forPlot.Fst_group <span class="op">.==</span> groups_to_plot_PCA[i]</span>
<span id="cb535-37"><a href="#cb535-37" aria-hidden="true" tabindex="-1"></a>    CairoMakie.<span class="fu">scatter!</span>(ax, PC1[selection], PC2[selection], marker <span class="op">=</span> <span class="op">:</span>diamond, color <span class="op">=</span> (group_colors_PCA[i], fillOpacity), markersize <span class="op">=</span> symbolSize, strokewidth<span class="op">=</span><span class="fl">0.5</span>, strokecolor <span class="op">=</span> (<span class="st">"black"</span>, lineOpacity))</span>
<span id="cb535-38"><a href="#cb535-38" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb535-39"><a href="#cb535-39" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="notes" class="level4">
<h4 class="anchored" data-anchor-id="notes">Notes:</h4>
</section>
</section>
<section id="get-all-samples-with-regionindhetstan-.-0.75-or-maybe-another-value" class="level1">
<h1>- Get all samples with regionIndHetStan .&lt; 0.75 (or maybe another value)</h1>
</section>
<section id="construct-distance-matrix" class="level1">
<h1>- Construct distance matrix</h1>
</section>
<section id="do-pcoa-on-distance-matrix-also-known-as-metric-multidimensional-scaling" class="level1">
<h1>- Do PCoA on distance matrix (also known as Metric Multidimensional Scaling)</h1>
</section>
<section id="mds-fitmds-d-distancestrue-maxoutdimsized1-1" class="level1">
<h1>mds = fit(MDS, D; distances=true, maxoutdim=size(D,1)-1)</h1>
</section>
<section id="httpsjuliastats.orgmultivariatestats.jldevmds" class="level1">
<h1>https://juliastats.org/MultivariateStats.jl/dev/mds/</h1>
</section>
<section id="pick-a-small-set-of-individuals-to-examine-genetic-distances" class="level1">
<h1>Pick a small set of individuals to examine genetic distances</h1>
<div class="sourceCode" id="cb536"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb536-1"><a href="#cb536-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for demonstration haplotypes, choose individuals with numMissings &lt; 200_000</span></span>
<span id="cb536-2"><a href="#cb536-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and regionIndHetStan &lt; 0.5:</span></span>
<span id="cb536-3"><a href="#cb536-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-4"><a href="#cb536-4" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (ind_with_metadata_included.numMissings <span class="op">.&lt;</span> <span class="fl">200_000</span>) <span class="op">.&amp;</span> (ind_with_metadata_included.regionIndHetStan <span class="op">.&lt;</span> <span class="fl">0.75</span>)</span>
<span id="cb536-5"><a href="#cb536-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-6"><a href="#cb536-6" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_haplotypeCandidates <span class="op">=</span> ind_with_metadata_included[selection, <span class="op">:</span>]</span>
<span id="cb536-7"><a href="#cb536-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-8"><a href="#cb536-8" aria-hidden="true" tabindex="-1"></a><span class="co"># But the above cuts out the obscuratus and nitidus individuals (which have about 700k missing genome-wide at least in early filtering calculations)</span></span>
<span id="cb536-9"><a href="#cb536-9" aria-hidden="true" tabindex="-1"></a><span class="co"># So don't worry about missings for now--just sort and choose low ones for each haplogroup:</span></span>
<span id="cb536-10"><a href="#cb536-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-11"><a href="#cb536-11" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate2_TTGW73  #lud_Sath</span></span>
<span id="cb536-12"><a href="#cb536-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-13"><a href="#cb536-13" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate2_LN2 #troch_LN</span></span>
<span id="cb536-14"><a href="#cb536-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-15"><a href="#cb536-15" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate2_TTGW-15-04 #lud_Sural</span></span>
<span id="cb536-16"><a href="#cb536-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-17"><a href="#cb536-17" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate1_JG02G02 #plumb</span></span>
<span id="cb536-18"><a href="#cb536-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-19"><a href="#cb536-19" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate2_JE31G01 #vir</span></span>
<span id="cb536-20"><a href="#cb536-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-21"><a href="#cb536-21" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Armando_plate1_RF20G01 #plumb_BJ</span></span>
<span id="cb536-22"><a href="#cb536-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-23"><a href="#cb536-23" aria-hidden="true" tabindex="-1"></a><span class="co"># GW_Lane5_TU1 #nit</span></span>
<span id="cb536-24"><a href="#cb536-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-25"><a href="#cb536-25" aria-hidden="true" tabindex="-1"></a>indsToChoose <span class="op">=</span> [<span class="st">"GW_Armando_plate2_TTGW73"</span>,</span>
<span id="cb536-26"><a href="#cb536-26" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Armando_plate2_LN2"</span>,</span>
<span id="cb536-27"><a href="#cb536-27" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Armando_plate2_TTGW-15-04"</span>,</span>
<span id="cb536-28"><a href="#cb536-28" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Armando_plate1_JG02G02"</span>,</span>
<span id="cb536-29"><a href="#cb536-29" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Armando_plate2_JE31G01"</span>,</span>
<span id="cb536-30"><a href="#cb536-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Armando_plate1_RF20G01"</span>,</span>
<span id="cb536-31"><a href="#cb536-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">"GW_Lane5_TU1"</span>]</span>
<span id="cb536-32"><a href="#cb536-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-33"><a href="#cb536-33" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> <span class="fu">map</span>(<span class="fu">in</span>(indsToChoose), ind_with_metadata_included.ind)</span>
<span id="cb536-34"><a href="#cb536-34" aria-hidden="true" tabindex="-1"></a>filtered_in <span class="op">=</span> ind_with_metadata_included.ind[selection]</span>
<span id="cb536-35"><a href="#cb536-35" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chosenInds <span class="op">=</span> ind_with_metadata_included[selection, <span class="op">:</span>]</span>
<span id="cb536-36"><a href="#cb536-36" aria-hidden="true" tabindex="-1"></a>geno_indFiltered <span class="op">=</span> <span class="fu">view</span>(genotypes_region, selection, <span class="op">:</span>)</span>
<span id="cb536-37"><a href="#cb536-37" aria-hidden="true" tabindex="-1"></a><span class="fu">println</span>(<span class="st">"Specific individuals chosen as requested: "</span>)</span>
<span id="cb536-38"><a href="#cb536-38" aria-hidden="true" tabindex="-1"></a><span class="fu">display</span>(filtered_in)</span>
<span id="cb536-39"><a href="#cb536-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-40"><a href="#cb536-40" aria-hidden="true" tabindex="-1"></a><span class="co"># choose the loci in the high HetVar region:</span></span>
<span id="cb536-41"><a href="#cb536-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-42"><a href="#cb536-42" aria-hidden="true" tabindex="-1"></a>selection <span class="op">=</span> (positionMin <span class="op">.&lt;</span> pos_region.position <span class="op">.&lt;</span> positionMax)</span>
<span id="cb536-43"><a href="#cb536-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-44"><a href="#cb536-44" aria-hidden="true" tabindex="-1"></a>geno_indFiltered_highHetVarRegion <span class="op">=</span> geno_indFiltered[<span class="op">:</span>, selection]</span>
<span id="cb536-45"><a href="#cb536-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-46"><a href="#cb536-46" aria-hidden="true" tabindex="-1"></a><span class="co"># put in plot order:</span></span>
<span id="cb536-47"><a href="#cb536-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-48"><a href="#cb536-48" aria-hidden="true" tabindex="-1"></a>indexOrder <span class="op">=</span> <span class="fu">sortperm</span>(ind_with_metadata_chosenInds.plot_order)</span>
<span id="cb536-49"><a href="#cb536-49" aria-hidden="true" tabindex="-1"></a>geno_indFiltered_highHetVarRegion_ordered <span class="op">=</span> geno_indFiltered_highHetVarRegion[indexOrder, <span class="op">:</span>]</span>
<span id="cb536-50"><a href="#cb536-50" aria-hidden="true" tabindex="-1"></a>ind_with_metadata_chosenInds_ordered <span class="op">=</span> ind_with_metadata_chosenInds[indexOrder, <span class="op">:</span>]</span>
<span id="cb536-51"><a href="#cb536-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-52"><a href="#cb536-52" aria-hidden="true" tabindex="-1"></a><span class="co"># ABOVE HAVE GENOTYPES FOR JUST THESE INDIVIDUALS IN JUST THE HIGH HETVAR REGION ON THE ONE CHR. NEXT DO GENETIC DISTANCE BETWEEN ALL PAIRS OF INDIVIDUALS FOR THIS REGION. </span></span>
<span id="cb536-53"><a href="#cb536-53" aria-hidden="true" tabindex="-1"></a><span class="co"># count mismatches and divide by total number of sites.</span></span>
<span id="cb536-54"><a href="#cb536-54" aria-hidden="true" tabindex="-1"></a><span class="co"># mismatches of genotype 0 to genotype 2 count as 1 mismatch.</span></span>
<span id="cb536-55"><a href="#cb536-55" aria-hidden="true" tabindex="-1"></a><span class="co"># genotype 0 or 2 to genotype 1 (het) counts as 0.5 mismatch.</span></span>
<span id="cb536-56"><a href="#cb536-56" aria-hidden="true" tabindex="-1"></a><span class="co"># genotype 1 to genotype 1 counts as 0 mismatch (debatable whether it should be 0 or 0.5, but I want F1 hybrids to not have distance between them).</span></span>
<span id="cb536-57"><a href="#cb536-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-58"><a href="#cb536-58" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">getIndGenDistances</span>(genotypeMatrix)</span>
<span id="cb536-59"><a href="#cb536-59" aria-hidden="true" tabindex="-1"></a>    genotypeMatrix[<span class="fu">ismissing</span>.(genotypeMatrix)] <span class="op">.=</span> <span class="op">-</span><span class="fl">1</span> <span class="co"># converts `missing` to -1</span></span>
<span id="cb536-60"><a href="#cb536-60" aria-hidden="true" tabindex="-1"></a>    numInds <span class="op">=</span> <span class="fu">size</span>(genotypeMatrix)[<span class="fl">1</span>]</span>
<span id="cb536-61"><a href="#cb536-61" aria-hidden="true" tabindex="-1"></a>    distanceMatrix <span class="op">=</span> <span class="fu">Matrix</span><span class="dt">{Float64}</span>(<span class="cn">undef</span>, numInds, numInds)</span>
<span id="cb536-62"><a href="#cb536-62" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb536-63"><a href="#cb536-63" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>numInds</span>
<span id="cb536-64"><a href="#cb536-64" aria-hidden="true" tabindex="-1"></a>            firstGeno <span class="op">=</span> genotypeMatrix[i,<span class="op">:</span>]</span>
<span id="cb536-65"><a href="#cb536-65" aria-hidden="true" tabindex="-1"></a>            secondGeno <span class="op">=</span> genotypeMatrix[j,<span class="op">:</span>]</span>
<span id="cb536-66"><a href="#cb536-66" aria-hidden="true" tabindex="-1"></a>            fullMismatches <span class="op">=</span> <span class="fu">sum</span>((firstGeno <span class="op">.==</span> <span class="fl">0</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">2</span>) <span class="op">.||</span></span>
<span id="cb536-67"><a href="#cb536-67" aria-hidden="true" tabindex="-1"></a>                                (firstGeno <span class="op">.==</span> <span class="fl">2</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">0</span>))</span>
<span id="cb536-68"><a href="#cb536-68" aria-hidden="true" tabindex="-1"></a>            partialMismatches <span class="op">=</span> <span class="fu">sum</span>((firstGeno <span class="op">.==</span> <span class="fl">0</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">1</span>) <span class="op">.||</span> </span>
<span id="cb536-69"><a href="#cb536-69" aria-hidden="true" tabindex="-1"></a>                                    (firstGeno <span class="op">.==</span> <span class="fl">1</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">0</span>) <span class="op">.||</span> </span>
<span id="cb536-70"><a href="#cb536-70" aria-hidden="true" tabindex="-1"></a>                                    (firstGeno <span class="op">.==</span> <span class="fl">2</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">1</span>) <span class="op">.||</span> </span>
<span id="cb536-71"><a href="#cb536-71" aria-hidden="true" tabindex="-1"></a>                                    (firstGeno <span class="op">.==</span> <span class="fl">1</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">2</span>))</span>
<span id="cb536-72"><a href="#cb536-72" aria-hidden="true" tabindex="-1"></a>            bothHet <span class="op">=</span> <span class="fu">sum</span>(firstGeno <span class="op">.==</span> <span class="fl">1</span> <span class="op">.&amp;&amp;</span> secondGeno <span class="op">.==</span> <span class="fl">1</span>)</span>
<span id="cb536-73"><a href="#cb536-73" aria-hidden="true" tabindex="-1"></a>            nonMissingGenotypes <span class="op">=</span> <span class="fu">sum</span>((firstGeno <span class="op">.!=</span> <span class="op">-</span><span class="fl">1</span>) <span class="op">.&amp;&amp;</span> (secondGeno <span class="op">.!=</span> <span class="op">-</span><span class="fl">1</span>))</span>
<span id="cb536-74"><a href="#cb536-74" aria-hidden="true" tabindex="-1"></a>            distanceMatrix[i,j] <span class="op">=</span> (fullMismatches <span class="op">+</span> <span class="fl">0.5</span>partialMismatches <span class="op">+</span> <span class="fl">0</span><span class="op">*</span>bothHet) <span class="op">/</span> nonMissingGenotypes</span>
<span id="cb536-75"><a href="#cb536-75" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb536-76"><a href="#cb536-76" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb536-77"><a href="#cb536-77" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> distanceMatrix</span>
<span id="cb536-78"><a href="#cb536-78" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb536-79"><a href="#cb536-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-80"><a href="#cb536-80" aria-hidden="true" tabindex="-1"></a>indGenDistances <span class="op">=</span> <span class="fu">getIndGenDistances</span>(geno_indFiltered_highHetVarRegion_ordered)</span>
<span id="cb536-81"><a href="#cb536-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-82"><a href="#cb536-82" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> <span class="fu">fit</span>(MDS, indGenDistances; distances<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb536-83"><a href="#cb536-83" aria-hidden="true" tabindex="-1"></a>pco_coords <span class="op">=</span> <span class="fu">predict</span>(mds)</span>
<span id="cb536-84"><a href="#cb536-84" aria-hidden="true" tabindex="-1"></a>PCO1 <span class="op">=</span> pco_coords[<span class="fl">1</span>,<span class="op">:</span>]</span>
<span id="cb536-85"><a href="#cb536-85" aria-hidden="true" tabindex="-1"></a>PCO2 <span class="op">=</span> pco_coords[<span class="fl">2</span>,<span class="op">:</span>]</span>
<span id="cb536-86"><a href="#cb536-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb536-87"><a href="#cb536-87" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(PCO1, PCO2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>